
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>Barcoded-subamplicon sequencing &#8212; dms_tools2 2.4.16 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Amino-acid preferences" href="prefs.html" />
    <link rel="prev" title="Examples" href="examples.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="barcoded-subamplicon-sequencing">
<span id="bcsubamp"></span><h1>Barcoded-subamplicon sequencing<a class="headerlink" href="#barcoded-subamplicon-sequencing" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="contents">
<ul class="simple">
<li><p><a class="reference internal" href="#overview" id="id3">Overview</a></p></li>
<li><p><a class="reference internal" href="#how-does-barcoded-subamplicon-sequencing-work" id="id4">How does barcoded-subamplicon sequencing work?</a></p></li>
<li><p><a class="reference internal" href="#workflow" id="id5">Workflow</a></p></li>
<li><p><a class="reference internal" href="#programs-and-examples" id="id6">Programs and examples</a></p></li>
<li><p><a class="reference internal" href="#more-details-and-notes" id="id7">More details and notes</a></p></li>
</ul>
</div>
<div class="section" id="overview">
<h2><a class="toc-backref" href="#id3">Overview</a><a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>Barcoded-subamplicon sequencing is a strategy to increase the accuracy of Illumina deep sequencing.
This strategy is useful for deep mutational scanning, since it enables better identification of rare mutations without the confounding effects of sequencing errors.</p>
<p>The <a class="reference external" href="http://jbloomlab.github.io/dms_tools2">dms_tools2</a> software contains programs to process the deep-sequencing data generated by this sequencing strategy.
Typically you would want to analyze your data using <a class="reference internal" href="dms2_batch_bcsubamp.html#dms2-batch-bcsubamp"><span class="std std-ref">dms2_batch_bcsubamp</span></a>, which in turn calls <a class="reference internal" href="dms2_bcsubamp.html#dms2-bcsubamp"><span class="std std-ref">dms2_bcsubamp</span></a>.
The <a class="reference internal" href="examples.html#examples"><span class="std std-ref">Examples</span></a> include some analyses of this type of deep-sequencing data; for instance see the <a class="reference external" href="https://github.com/jbloomlab/dms_tools2/blob/master/examples/Doud2016/analysis_notebook.ipynb">Doud2016 example</a>.</p>
</div>
<div class="section" id="how-does-barcoded-subamplicon-sequencing-work">
<h2><a class="toc-backref" href="#id4">How does barcoded-subamplicon sequencing work?</a><a class="headerlink" href="#how-does-barcoded-subamplicon-sequencing-work" title="Permalink to this headline">¶</a></h2>
<p>Barcoded-subamplicon sequencing is an Illumina library preparation strategy that increases the sequencing accuracy by attaching unique random barcodes to molecules during library preparation, and then grouping sequencing reads by barcode to correct for sequencing errors.
To our knowledge, this basic concept was first described by <a class="reference external" href="https://www.nature.com/nmeth/journal/v7/n2/full/nmeth.1416.html">Hiatt et al (2010)</a>, <a class="reference external" href="http://www.pnas.org/content/108/23/9530.short">Kinde et al (2011)</a>, and <a class="reference external" href="http://www.pnas.org/content/108/50/20166.short">Jabara et al (2011)</a>.
The concept was first applied for use in deep mutational scanning by <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/24820965">Wu et al (2014)</a>.
<a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/26868371">Zhang et al (2016)</a> reported that barcoded-subamplicon sequencing is a better error-correction strategy than overlapping paired-end reads, and our results in the <a class="reference external" href="https://research.fhcrc.org/bloom/en.html">Bloom lab</a> are consistent with their report.
The exact implementation of barcoded-subamplicon sequencing used by the <a class="reference external" href="https://research.fhcrc.org/bloom/en.html">Bloom lab</a> (and the implementation that <a class="reference external" href="http://jbloomlab.github.io/dms_tools2">dms_tools2</a> is designed to analyze) is described in <a class="reference external" href="http://www.mdpi.com/1999-4915/8/6/155">Doud and Bloom (2016)</a>.</p>
<div class="figure align-default" id="id1" style="width: 90%">
<span id="barcodedsubamplicondoud2016"></span><a class="reference internal image-reference" href="_images/BarcodedSubampliconDoud2016.jpg"><img alt="schematic of barcoded subamplicon sequencing, Doud and Bloom 2016" src="_images/BarcodedSubampliconDoud2016.jpg" style="width: 90%;" /></a>
<p class="caption"><span class="caption-number">Fig. 1 </span><span class="caption-text">Schematic of barcoded-subamplicon sequencing of influenza HA as done by <a class="reference external" href="http://www.mdpi.com/1999-4915/8/6/155">Doud and Bloom (2016)</a>.</span><a class="headerlink" href="#id1" title="Permalink to this image">¶</a></p>
</div>
<div class="figure align-default" id="id2" style="width: 90%">
<span id="barcodedsubamplicondingens2017"></span><a class="reference internal image-reference" href="_images/BarcodedSubampliconDingens2017.jpg"><img alt="schematic of barcoded subamplicon sequencing, Dingens et al 2017" src="_images/BarcodedSubampliconDingens2017.jpg" style="width: 90%;" /></a>
<p class="caption"><span class="caption-number">Fig. 2 </span><span class="caption-text">Schematic of barcoded-subamplicon sequencing of HIV Env as done by <a class="reference external" href="http://dx.doi.org/10.1016/j.chom.2017.05.003">Dingens et al (2017)</a>.</span><a class="headerlink" href="#id2" title="Permalink to this image">¶</a></p>
</div>
<p>Schematics of barcoded-subamplicon sequencing are shown in <a class="reference internal" href="#barcodedsubamplicondoud2016"><span class="std std-numref">Fig. 1</span></a> and <a class="reference internal" href="#barcodedsubamplicondingens2017"><span class="std std-numref">Fig. 2</span></a>.
Essentially, barcoded-subamplicon sequencing involves using PCR to make a series of subamplicons from the gene to be sequenced.
Each of these subamplicons contains a random barcode (a string of <code class="docutils literal notranslate"><span class="pre">N</span></code> nucleotides) at each end that provides a unique identifier.
The barcoded subamplicons are then diluted such that the number of unique ssDNA molecules is less than the read depth, and then this dilution is amplified by primers that also add the Illumina adaptors.
Finally, these molecules are sequenced to a depth such that most barcodes are observed multiple times with different reads.</p>
<p>In analyzing the sequencing, we identify barcodes that are sequenced multiple times, and build a consensus sequence for the subamplicon from these multiple reads.
We only call mutations if multiple reads at a site all agree, thereby ameliorating the effect of sequencing errors that affect a single read.
We then align each consensus subamplicon to the reference sequence to call mutations.
There are more details in the <a class="reference internal" href="dms2_bcsubamp.html#bcsubamp-algorithm"><span class="std std-ref">Algorithm for assembling and aligning subamplicons</span></a> described in the documentation for <a class="reference internal" href="dms2_bcsubamp.html#dms2-bcsubamp"><span class="std std-ref">dms2_bcsubamp</span></a>.</p>
</div>
<div class="section" id="workflow">
<h2><a class="toc-backref" href="#id5">Workflow</a><a class="headerlink" href="#workflow" title="Permalink to this headline">¶</a></h2>
<p>To understand in detail how barcoded-subamplicon sequencing is done, example primers are shown below for sequencing of the hemagglutinin (HA) from A/WSN/1933 (H1N1) influenza:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Rnd1for1</span><span class="p">:</span> <span class="mi">5</span><span class="s1">&#39;-CTTTCCCTACACGACGCTCTTCCGATCTNNNNNNNNaagcaggggaaaataaaaacaaccaaa-3&#39;</span>                                                                                                                                                                                                                                                                                                                                                             <span class="n">Rnd1for427</span><span class="p">:</span> <span class="mi">5</span><span class="s1">&#39;-CTTTCCCTACACGACGCTCTTCCGATCTNNNNNNNNccaaggaaagttcatggcccaac-3&#39;</span>                                                                                                                                                                                                                                                                                                                                                           <span class="n">Rnd1for850</span><span class="p">:</span> <span class="mi">5</span><span class="s1">&#39;-CTTTCCCTACACGACGCTCTTCCGATCTNNNNNNNNgtttgagtccggcatcatcacc-3&#39;</span>                                                                                                                                                                                                                                                                                                                                              <span class="n">Rnd1for1276</span><span class="p">:</span> <span class="mi">5</span><span class="s1">&#39;-CTTTCCCTACACGACGCTCTTCCGATCTNNNNNNNNcaacaacttagaaaaaaggatggaaaatttaaataaa-3&#39;</span>
    <span class="n">WSN</span> <span class="n">HA</span> <span class="k">with</span> <span class="n">coding</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">caps</span><span class="p">:</span> <span class="mi">5</span><span class="s1">&#39;-agcaaaagcaggggaaaataaaaacaaccaaaATGAAGGCAAAACTACTGGTCCTGTTATATGCATTTGTAGCTACAGATGCAGACACAATATGTATAGGCTACCATGCGAACAACTCAACCGACACTGTTGACACAATACTCGAGAAGAATGTGGCAGTGACACATTCTGTTAACCTGCTCGAAGACAGCCACAACGGGAAACTATGTAAATTAAAAGGAATAGCCCCACTACAATTGGGGAAATGTAACATCACCGGATGGCTCTTGGGAAATCCAGAATGCGACTCACTGCTTCCAGCGAGATCATGGTCCTACATTGTAGAAACACCAAACTCTGAGAATGGAGCATGTTATCCAGGAGATCTCATCGACTATGAGGAACTGAGGGAGCAATTGAGCTCAGTATCATCATTAGAAAGATTCGAAATATTTCCCAAGGAAAGTTCATGGCCCAACCACACATTCAACGGAGTAACAGTATCATGCTCCCATAGGGGAAAAAGCAGTTTTTACAGAAATTTGCTATGGCTGACGAAGAAGGGGGATTCATACCCAAAGCTGACCAATTCCTATGTGAACAATAAAGGGAAAGAAGTCCTTGTACTATGGGGTGTTCATCACCCGTCTAGCAGTGATGAGCAACAGAGTCTCTATAGTAATGGAAATGCTTATGTCTCTGTAGCGTCTTCAAATTATAACAGGAGATTCACCCCGGAAATAGCTGCAAGGCCCAAAGTAAGAGATCAACATGGGAGGATGAACTATTACTGGACCTTGCTAGAACCCGGAGACACAATAATATTTGAGGCAACTGGTAATCTAATAGCACCATGGTATGCTTTCGCACTGAGTAGAGGGTTTGAGTCCGGCATCATCACCTCAAACGCGTCAATGCATGAGTGTAACACGAAGTGTCAAACACCCCAGGGAGCTATAAACAGCAATCTCCCTTTCCAGAATATACACCCAGTCACAATAGGAGAGTGCCCAAAATATGTCAGGAGTACCAAATTGAGGATGGTTACAGGACTAAGAAACATCCCATCCATTCAATACAGAGGTCTATTTGGAGCCATTGCTGGTTTTATTGAGGGGGGATGGACTGGAATGATAGATGGATGGTATGGTTATCATCATCAGAATGAACAGGGATCAGGCTATGCAGCGGATCAAAAAAGCACACAAAATGCCATTAACGGGATTACAAACAAGGTGAACTCTGTTATCGAGAAAATGAACACTCAATTCACAGCTGTGGGTAAAGAATTCAACAACTTAGAAAAAAGGATGGAAAATTTAAATAAAAAAGTTGATGATGGGTTTCTGGACATTTGGACATATAATGCAGAATTGTTAGTTCTACTGGAAAATGAAAGGACTTTGGATTTCCATGACTTAAATGTGAAGAATCTGTACGAGAAAGTAAAAAGCCAATTAAAGAATAATGCCAAAGAAATCGGAAATGGGTGTTTTGAGTTCTACCACAAGTGTGACAATGAATGCATGGAAAGTGTAAGAAATGGGACTTATGATTATCCAAAATATTCAGAAGAATCAAAGTTGAACAGGGAAAAGATAGATGGAGTGAAATTGGAATCAATGGGGGTGTATCAGATTCTGGCGATCTACTCAACTGTCGCCAGTTCACTGGTGCTTTTGGTCTCCCTGGGGGCAATCAGTTTCTGGATGTGTTCTAATGGGTCTTTGCAGTGCAGAATATGCATCTGAgattaggatttcagaaatataaggaaaaacacccttgtttctact-3&#39;</span>
    <span class="n">WSN</span> <span class="n">HA</span> <span class="k">with</span> <span class="n">coding</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">caps</span><span class="p">:</span> <span class="mi">3</span><span class="s1">&#39;-tcgttttcgtccccttttatttttgttggtttTACTTCCGTTTTGATGACCAGGACAATATACGTAAACATCGATGTCTACGTCTGTGTTATACATATCCGATGGTACGCTTGTTGAGTTGGCTGTGACAACTGTGTTATGAGCTCTTCTTACACCGTCACTGTGTAAGACAATTGGACGAGCTTCTGTCGGTGTTGCCCTTTGATACATTTAATTTTCCTTATCGGGGTGATGTTAACCCCTTTACATTGTAGTGGCCTACCGAGAACCCTTTAGGTCTTACGCTGAGTGACGAAGGTCGCTCTAGTACCAGGATGTAACATCTTTGTGGTTTGAGACTCTTACCTCGTACAATAGGTCCTCTAGAGTAGCTGATACTCCTTGACTCCCTCGTTAACTCGAGTCATAGTAGTAATCTTTCTAAGCTTTATAAAGGGTTCCTTTCAAGTACCGGGTTGGTGTGTAAGTTGCCTCATTGTCATAGTACGAGGGTATCCCCTTTTTCGTCAAAAATGTCTTTAAACGATACCGACTGCTTCTTCCCCCTAAGTATGGGTTTCGACTGGTTAAGGATACACTTGTTATTTCCCTTTCTTCAGGAACATGATACCCCACAAGTAGTGGGCAGATCGTCACTACTCGTTGTCTCAGAGATATCATTACCTTTACGAATACAGAGACATCGCAGAAGTTTAATATTGTCCTCTAAGTGGGGCCTTTATCGACGTTCCGGGTTTCATTCTCTAGTTGTACCCTCCTACTTGATAATGACCTGGAACGATCTTGGGCCTCTGTGTTATTATAAACTCCGTTGACCATTAGATTATCGTGGTACCATACGAAAGCGTGACTCATCTCCCAAACTCAGGCCGTAGTAGTGGAGTTTGCGCAGTTACGTACTCACATTGTGCTTCACAGTTTGTGGGGTCCCTCGATATTTGTCGTTAGAGGGAAAGGTCTTATATGTGGGTCAGTGTTATCCTCTCACGGGTTTTATACAGTCCTCATGGTTTAACTCCTACCAATGTCCTGATTCTTTGTAGGGTAGGTAAGTTATGTCTCCAGATAAACCTCGGTAACGACCAAAATAACTCCCCCCTACCTGACCTTACTATCTACCTACCATACCAATAGTAGTAGTCTTACTTGTCCCTAGTCCGATACGTCGCCTAGTTTTTTCGTGTGTTTTACGGTAATTGCCCTAATGTTTGTTCCACTTGAGACAATAGCTCTTTTACTTGTGAGTTAAGTGTCGACACCCATTTCTTAAGTTGTTGAATCTTTTTTCCTACCTTTTAAATTTATTTTTTCAACTACTACCCAAAGACCTGTAAACCTGTATATTACGTCTTAACAATCAAGATGACCTTTTACTTTCCTGAAACCTAAAGGTACTGAATTTACACTTCTTAGACATGCTCTTTCATTTTTCGGTTAATTTCTTATTACGGTTTCTTTAGCCTTTACCCACAAAACTCAAGATGGTGTTCACACTGTTACTTACGTACCTTTCACATTCTTTACCCTGAATACTAATAGGTTTTATAAGTCTTCTTAGTTTCAACTTGTCCCTTTTCTATCTACCTCACTTTAACCTTAGTTACCCCCACATAGTCTAAGACCGCTAGATGAGTTGACAGCGGTCAAGTGACCACGAAAACCAGAGGGACCCCCGTTAGTCAAAGACCTACACAAGATTACCCAGAAACGTCACGTCTTATACGTAGACTctaatcctaaagtctttatattcctttttgtgggaacaaagatga-5&#39;</span>
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       <span class="n">Rnd1rev426</span><span class="p">:</span> <span class="mi">3</span><span class="s1">&#39;-gtgtgtaagttgcctcattgtcatagtacNNNNNNNNTCTAGCCTTCTCGTGTGCAGACTTGAGG-5&#39;</span>                                                                                                                                                                                                                                                                                                                                                    <span class="n">Rnd1rev849</span><span class="p">:</span> <span class="mi">3</span><span class="s1">&#39;-agtttgcgcagttacgtactcacNNNNNNNNTCTAGCCTTCTCGTGTGCAGACTTGAGG-5&#39;</span>                                                                                                                                                                                                                                                                                                                                                                 <span class="n">Rnd1rev1275</span><span class="p">:</span> <span class="mi">3</span><span class="s1">&#39;-actactacccaaagacctgtaaaNNNNNNNNTCTAGCCTTCTCGTGTGCAGACTTGAGG-5&#39;</span>                                                                                                                                                                                                                                                                                                                                                    <span class="n">Rnd1rev1698</span><span class="p">:</span> <span class="mi">3</span><span class="s1">&#39;-ctaatcctaaagtctttatattcctttttgtgggaaNNNNNNNNTCTAGCCTTCTCGTGTGCAGACTTGAGG-5&#39;</span>
</pre></div>
</div>
<p>Briefly, the steps are as follows:</p>
<blockquote>
<div><ol class="arabic">
<li><p>The library of HA genes is subjected to PCR with primers to create subamplicons of <span class="math notranslate nohighlight">\(\approx 425\)</span> internal nucleotides (or less) as well as additional adaptor and primer-binding sequences. In the example above, PCRs are performed with the following primer pairs:</p>
<blockquote>
<div><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">Rnd1for1</span></code> and <code class="docutils literal notranslate"><span class="pre">Rnd1rev426</span></code>: these primers create a subamplicon spanning nucleotides 1 to 426 (codons 1 to 142) of the WSN HA, which is the part of the sequence shown in caps above.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Rnd1for427</span></code> and <code class="docutils literal notranslate"><span class="pre">Rnd1rev849</span></code>: these primers create a subamplicon spanning nucleotides 427 to 849 (codons 143 to 283).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Rnd1for850</span></code> and <code class="docutils literal notranslate"><span class="pre">Rnd1rev1275</span></code>: these primers create a subamplicon spanning nucleotides 850 to 1275 (codons 284 to 425).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Rnd1for1276</span></code> and <code class="docutils literal notranslate"><span class="pre">Rnd1rev1698</span></code>: these primers create a subamplicon spanning nucleotides 1276 to 1698 (codons 426 to 566).</p></li>
</ul>
</div></blockquote>
<p><strong>Note that each of these subamplicons begins and ends on a full codon.</strong> This is important; design your primers to do the same.</p>
</li>
<li><p>The subamplicons are diluted so that the total number of ssDNA molecules is less than the sequencing depth. Since each subamplicon molecule has 8 <code class="docutils literal notranslate"><span class="pre">N</span></code> nucleotides at each end, there are <span class="math notranslate nohighlight">\(4^{16} \approx 4.3 \times 10^9\)</span> unique barcodes. Since the number of unique barcodes is much greater than the total number of molecules in the dilution, each barcode will generally be associated with just one molecule.</p></li>
<li><p>The barcoded subamplicons are further amplified by primers that add the Illumina adaptors, and perhaps Illumina indices for multiplexing. For instance, here are two example primers for this second round of PCR:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">Rnd2forUniversal</span><span class="p">:</span> <span class="n">forward</span> <span class="n">primer</span> <span class="k">for</span> <span class="n">the</span> <span class="nb">round</span> <span class="mi">2</span> <span class="n">PCRs</span> <span class="n">that</span> <span class="n">adds</span> <span class="n">the</span> <span class="n">Illumina</span> <span class="n">adaptor</span>
<span class="n">AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCC</span>

<span class="o">&gt;</span><span class="n">Rnd2revIndex</span><span class="p">:</span> <span class="n">reverse</span> <span class="n">primer</span> <span class="k">for</span> <span class="n">the</span> <span class="nb">round</span> <span class="mi">2</span> <span class="n">PCRs</span> <span class="n">that</span> <span class="n">adds</span> <span class="n">the</span> <span class="n">TruSeq</span> <span class="n">index</span> <span class="mi">1</span> <span class="p">(</span><span class="ow">in</span> <span class="n">lower</span> <span class="n">case</span><span class="p">)</span> <span class="k">as</span> <span class="n">well</span> <span class="k">as</span> <span class="n">the</span> <span class="n">Illumina</span> <span class="n">adaptor</span>
<span class="n">CAAGCAGAAGACGGCATACGAGATcgtgatGTGACTGGAGTTCAGACGTGTGCTCTTCC</span>
</pre></div>
</div>
</li>
<li><p>The products of the round 2 PCRs are sequenced with overlapping paired-end reads of sufficient length to cover the entire subamplicon, and the sequencing data are analyzed with the standard Illumina pipeline to generate FASTQ files with the R1 and R2 reads.</p></li>
<li><p>The FASTQ files are analyzed with <a class="reference internal" href="dms2_batch_bcsubamp.html#dms2-batch-bcsubamp"><span class="std std-ref">dms2_batch_bcsubamp</span></a>. For the example shown above, the <code class="docutils literal notranslate"><span class="pre">--alignspecs</span></code> parameter would be <code class="docutils literal notranslate"><span class="pre">1,426,36,38</span> <span class="pre">427,849,32,32</span> <span class="pre">850,1275,31,37</span> <span class="pre">1276,1698,46,45</span></code>.</p></li>
</ol>
</div></blockquote>
<p>Some important considerations if you are designing such an experiment:</p>
<blockquote>
<div><ul class="simple">
<li><p>Make sure that each primer begins / ends on a full codon – you do not want a codon split among subamplicons.</p></li>
<li><p>Make sure that your subamplicons are not excessively long, as the Illumina sequencing quality falls off near the end of reads. Although the example above uses four subamplicons to sequencing the 1.7 kb HA gene, in the actual experiments of <a class="reference external" href="http://www.mdpi.com/1999-4915/8/6/155">Doud and Bloom (2016)</a> on this gene, they instead chose to use six subamplicons so that they did not need high quality at the end of the reads.</p></li>
</ul>
</div></blockquote>
</div>
<div class="section" id="programs-and-examples">
<h2><a class="toc-backref" href="#id6">Programs and examples</a><a class="headerlink" href="#programs-and-examples" title="Permalink to this headline">¶</a></h2>
<p>The <a class="reference external" href="http://jbloomlab.github.io/dms_tools2">dms_tools2</a> software contains programs to analyze the data generated by barcoded-subamplicon sequencing:</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="dms2_bcsubamp.html#dms2-bcsubamp"><span class="std std-ref">dms2_bcsubamp</span></a> processes the FASTQ data to count mutations.</p></li>
<li><p><a class="reference internal" href="dms2_batch_bcsubamp.html#dms2-batch-bcsubamp"><span class="std std-ref">dms2_batch_bcsubamp</span></a> runs <a class="reference internal" href="dms2_bcsubamp.html#dms2-bcsubamp"><span class="std std-ref">dms2_bcsubamp</span></a> on a set of samples and creates summary plots. In general, you will have an easier time if you use <a class="reference internal" href="dms2_batch_bcsubamp.html#dms2-batch-bcsubamp"><span class="std std-ref">dms2_batch_bcsubamp</span></a> rather than runnning <a class="reference internal" href="dms2_bcsubamp.html#dms2-bcsubamp"><span class="std std-ref">dms2_bcsubamp</span></a> directly.</p></li>
</ul>
</div></blockquote>
<p>Here are some example analyses:</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference external" href="https://github.com/jbloomlab/dms_tools2/blob/master/examples/Doud2016/analysis_notebook.ipynb">Doud2016 example</a></p></li>
<li><p><a class="reference external" href="https://github.com/jbloomlab/dms_tools2/blob/master/examples/Doud2017/analysis_notebook.ipynb">Doud2017 example</a></p></li>
<li><p><a class="reference external" href="https://github.com/jbloomlab/dms_tools2/blob/master/examples/Dingens2017/analysis_notebook.ipynb">Dingens2017 example</a></p></li>
<li><p><a class="reference external" href="https://github.com/jbloomlab/dms_tools2/blob/master/examples/Doud2018/analysis_notebook.ipynb">Doud2018 example</a></p></li>
<li><p><a class="reference external" href="https://github.com/jbloomlab/dms_tools2/tree/master/examples/Haddox2018">Haddox2018 example</a></p></li>
<li><p><a class="reference external" href="https://github.com/jbloomlab/dms_tools2/blob/master/examples/Dingens2018/analysis_notebook.ipynb">Dingens2018 example</a></p></li>
<li><p><a class="reference external" href="https://github.com/jbloomlab/dms_tools2/tree/master/examples/Lee2018">Lee2018 example</a></p></li>
</ul>
</div></blockquote>
</div>
<div class="section" id="more-details-and-notes">
<h2><a class="toc-backref" href="#id7">More details and notes</a><a class="headerlink" href="#more-details-and-notes" title="Permalink to this headline">¶</a></h2>
<p>The <a class="reference external" href="https://research.fhcrc.org/bloom/en.html">Bloom lab</a> has now published a number of papers using this deep sequencing strategy.
The methods sections of those papers have more details.
For instance, see:</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference external" href="http://www.mdpi.com/1999-4915/8/6/155">Doud and Bloom (2016)</a></p></li>
<li><p><a class="reference external" href="http://dx.doi.org/10.1371/journal.ppat.1006114">Haddox et al (2016)</a></p></li>
<li><p><a class="reference external" href="http://journals.plos.org/plospathogens/article?id=10.1371/journal.ppat.1006271">Doud et al (2017)</a></p></li>
<li><p><a class="reference external" href="http://dx.doi.org/10.1016/j.chom.2017.05.003">Dingens et al (2017)</a></p></li>
<li><p><a class="reference external" href="https://www.nature.com/articles/s41467-018-03665-3">Doud et al (2018)</a></p></li>
<li><p><a class="reference external" href="https://elifesciences.org/articles/34420">Haddox et al (2018)</a></p></li>
<li><p><a class="reference external" href="https://doi.org/10.1371/journal.ppat.1007159">Dingens et al (2018)</a></p></li>
<li><p><a class="reference external" href="http://dx.doi.org/10.1101/298364">Lee et al (2018)</a></p></li>
</ul>
</div></blockquote>
<p>If you want <strong>really detailed</strong> information, you can look at some lab notes. Here are <a class="reference external" href="_static/BarcodedSubampliconNotes_Jesse.pdf">Jesse Bloom’s lab notes</a> from an early iteration of this strategy on influenza HA, and here are <a class="reference external" href="_static/BarcodedSubampliconNotes_Mike.pdf">Mike Doud’s lab notes</a> from an experiment similar to that described in <a class="reference external" href="http://journals.plos.org/plospathogens/article?id=10.1371/journal.ppat.1006271">Doud et al (2017)</a>.</p>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">dms_tools2</h1>
    
  </a>
</p>



<p class="blurb">Tools for analyzing deep mutational scanning data.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=dms_tools2&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a class="badge" href="https://travis-ci.org/jbloomlab/dms_tools2">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Barcoded-subamplicon sequencing</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="#how-does-barcoded-subamplicon-sequencing-work">How does barcoded-subamplicon sequencing work?</a></li>
<li class="toctree-l2"><a class="reference internal" href="#workflow">Workflow</a></li>
<li class="toctree-l2"><a class="reference internal" href="#programs-and-examples">Programs and examples</a></li>
<li class="toctree-l2"><a class="reference internal" href="#more-details-and-notes">More details and notes</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="prefs.html">Amino-acid preferences</a></li>
<li class="toctree-l1"><a class="reference internal" href="diffsel.html">Differential selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="fracsurvive.html">Fraction surviving</a></li>
<li class="toctree-l1"><a class="reference internal" href="programs.html">Executable programs</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">Python API</a></li>
<li class="toctree-l1"><a class="reference internal" href="citations.html">Citations</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="examples.html" title="previous chapter">Examples</a></li>
      <li>Next: <a href="prefs.html" title="next chapter">Amino-acid preferences</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017--2019, the Bloom lab.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.1.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/bcsubamp.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/dms_tools2" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>