
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Example analysis of barcoded codon variants &#8212; dms_tools2 2.4.9 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="example-analysis-of-barcoded-codon-variants">
<h1>Example analysis of barcoded codon variants<a class="headerlink" href="#example-analysis-of-barcoded-codon-variants" title="Permalink to this headline">¶</a></h1>
<a class="reference download internal" href="codonvariant_analysis_demo_full.ipynb">Download this page as a Jupyter notebook (with outputs)</a><p>Here is an example of analyzing barcoded coding variants of a gene. The
analysis first simulates an “plausible” data set. It then analyzes the
simulated data.</p>
<p>Most of the analysis is done by using a
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html">CodonVariantTable</a>
to store and analyze the variants and their sequencing counts.</p>
<div class="contents local topic" id="table-of-contents">
<p class="topic-title first">Table of Contents</p>
<ul class="simple">
<li><a class="reference internal" href="#setup-for-analysis" id="id3">Setup for analysis</a></li>
<li><a class="reference internal" href="#initialize-a-codonvarianttable" id="id4">Initialize a CodonVariantTable</a><ul>
<li><a class="reference internal" href="#simulate-variants" id="id5">Simulate variants</a></li>
<li><a class="reference internal" href="#analyze-the-codon-variant-table" id="id6">Analyze the codon variant table</a></li>
</ul>
</li>
<li><a class="reference internal" href="#simulate-variant-counts" id="id7">Simulate variant counts</a><ul>
<li><a class="reference internal" href="#define-phenotype-function" id="id8">Define phenotype function</a></li>
<li><a class="reference internal" href="#simulate-variant-counts-for-samples" id="id9">Simulate variant counts for samples</a></li>
<li><a class="reference internal" href="#add-counts-to-variant-table" id="id10">Add counts to variant table</a></li>
</ul>
</li>
<li><a class="reference internal" href="#analyze-variant-counts-for-samples" id="id11">Analyze variant counts for samples</a><ul>
<li><a class="reference internal" href="#directly-access-variant-counts" id="id12">Directly access variant counts</a></li>
<li><a class="reference internal" href="#distribution-of-variant-counts" id="id13">Distribution of variant counts</a></li>
<li><a class="reference internal" href="#mutation-frequencies-in-each-sample" id="id14">Mutation frequencies in each sample</a></li>
<li><a class="reference internal" href="#mutation-sampling" id="id15">Mutation sampling</a></li>
</ul>
</li>
<li><a class="reference internal" href="#functional-scores-for-variants" id="id16">Functional scores for variants</a><ul>
<li><a class="reference internal" href="#calculate-functional-scores" id="id17">Calculate functional scores</a></li>
<li><a class="reference internal" href="#distribution-of-functional-scores" id="id18">Distribution of functional scores</a></li>
<li><a class="reference internal" href="#functional-scores-versus-phenotype" id="id19">Functional scores versus phenotype</a></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="setup-for-analysis">
<h2><a class="toc-backref" href="#id3">Setup for analysis</a><a class="headerlink" href="#setup-for-analysis" title="Permalink to this headline">¶</a></h2>
<p>We import key classes and functions from
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html">dms_tools2.codonvarianttable</a>.</p>
<p>We use <a class="reference external" href="https://plotnine.readthedocs.io">plotnine</a> for ggplot2-like
plotting syntax:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">collections</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">os</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">math</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">tempfile</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">random</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">scipy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">plotnine</span> <span class="k">import</span> <span class="o">*</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">dms_tools2</span> <span class="k">import</span> <span class="n">CODONS</span><span class="p">,</span> <span class="n">CODON_TO_AA</span><span class="p">,</span> <span class="n">NTS</span><span class="p">,</span> <span class="n">AAS_WITHSTOP</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">NONSTOP_CODONS</span> <span class="o">=</span> <span class="p">[</span><span class="n">c</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">CODONS</span> <span class="k">if</span> <span class="n">CODON_TO_AA</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;*&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">dms_tools2.codonvarianttable</span> <span class="k">import</span> <span class="n">CodonVariantTable</span><span class="p">,</span> \
<span class="gp">... </span>                                         <span class="n">simulateSampleCounts</span><span class="p">,</span> \
<span class="gp">... </span>                                         <span class="n">PhenotypeSimulator</span><span class="p">,</span> \
<span class="gp">... </span>                                         <span class="n">CBPALETTE</span>
</pre></div>
</div>
<p>Seed random number generators for reproducible output:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">scipy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
<p>Set pandas display options to show large chunks of Data Frames in this
example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s1">&#39;display.max_columns&#39;</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s1">&#39;display.width&#39;</span><span class="p">,</span> <span class="mi">500</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="initialize-a-codonvarianttable">
<h2><a class="toc-backref" href="#id4">Initialize a CodonVariantTable</a><a class="headerlink" href="#initialize-a-codonvarianttable" title="Permalink to this headline">¶</a></h2>
<p>We create a
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html">CodonVariantTable</a>
with simulated data.</p>
<div class="section" id="simulate-variants">
<h3><a class="toc-backref" href="#id5">Simulate variants</a><a class="headerlink" href="#simulate-variants" title="Permalink to this headline">¶</a></h3>
<p>Define a wildtype gene sequence of 50 codons. We will then build a table
of variants of this wildtype sequence.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">genelength</span> <span class="o">=</span> <span class="mi">50</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">geneseq</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">choices</span><span class="p">(</span><span class="n">NONSTOP_CODONS</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="n">genelength</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">geneseq</span>
<span class="go">&#39;AGATCCGTGATTCTGCGTGCTTACACCAACTCACGGGTGAAACGTGTAATCTTATGCAACAACGACTTACCTATCCGCAACATCCGGCTGATGATGATCCTACACAACTCCGACGCTAGTTTTTCGACTCCAGTAGGTTTACGCTCAGGA&#39;</span>
</pre></div>
</div>
<p>We now initialize a
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html">CodonVariantTable</a>.
Normally you would do this using the constructor for that class
described in the link above, which takes as input a
CSV file giving the barcodes and nucleotide substitutions in
the gene (in 1, 2, … numbering) for each variant, as well as the library
to which each variant belongs to (we always have at least
one library, but hopefully you have replicate libraries), and the
variant call support. The variant call support is confident you are that
the variant is called correctly, and might be the number of PacBio CCSs
that support that variant.</p>
<p>However, here we are just simulating data, so we create the
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html">CodonVariantTable</a>
using its <a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.from_simulation">from_simulation</a> method.
We simulate the variants for two libraries. Each library has a
number of variants that is 400-times the codon-length of the gene, and
barcodes are of length 16. We will have a Poisson distributed number of
codon mutations per variant with a mean of 1.3:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">libs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;lib_1&#39;</span><span class="p">,</span> <span class="s1">&#39;lib_2&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variants_per_lib</span> <span class="o">=</span> <span class="mi">400</span> <span class="o">*</span> <span class="n">genelength</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span> <span class="o">=</span> <span class="n">CodonVariantTable</span><span class="o">.</span><span class="n">from_simulation</span><span class="p">(</span>
<span class="gp">... </span>            <span class="n">geneseq</span><span class="o">=</span><span class="n">geneseq</span><span class="p">,</span>
<span class="gp">... </span>            <span class="n">bclen</span><span class="o">=</span><span class="mi">16</span><span class="p">,</span>
<span class="gp">... </span>            <span class="n">library_specs</span><span class="o">=</span><span class="p">{</span><span class="n">lib</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;avgmuts&#39;</span><span class="p">:</span> <span class="mf">1.3</span><span class="p">,</span> <span class="s1">&#39;nvariants&#39;</span><span class="p">:</span> <span class="n">variants_per_lib</span><span class="p">}</span>
<span class="gp">... </span>                           <span class="k">for</span> <span class="n">lib</span> <span class="ow">in</span> <span class="n">libs</span><span class="p">},</span>
<span class="gp">... </span>            <span class="n">variant_call_support</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span>
<span class="gp">... </span>            <span class="n">seed</span><span class="o">=</span><span class="kc">None</span>
<span class="gp">... </span>            <span class="p">)</span>
</pre></div>
</div>
<p>Now <cite>variants</cite> is a
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html">CodonVariantTable</a>
with the data on our barcodes and variants. We can get basic information
about the wildtype codon and amino-acid identities at each site using
the <cite>sites</cite>, <cite>codons</cite>, and <cite>aas</cite> attributes of the table. For
instance:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">sites</span><span class="p">[</span> <span class="p">:</span> <span class="mi">10</span><span class="p">]</span>
<span class="go">[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">list</span><span class="p">(</span><span class="n">variants</span><span class="o">.</span><span class="n">codons</span><span class="p">[</span><span class="n">r</span><span class="p">]</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">sites</span><span class="p">[</span> <span class="p">:</span> <span class="mi">10</span><span class="p">])</span>
<span class="go">[&#39;AGA&#39;, &#39;TCC&#39;, &#39;GTG&#39;, &#39;ATT&#39;, &#39;CTG&#39;, &#39;CGT&#39;, &#39;GCT&#39;, &#39;TAC&#39;, &#39;ACC&#39;, &#39;AAC&#39;]</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">list</span><span class="p">(</span><span class="n">variants</span><span class="o">.</span><span class="n">aas</span><span class="p">[</span><span class="n">r</span><span class="p">]</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">sites</span><span class="p">[</span> <span class="p">:</span> <span class="mi">10</span><span class="p">])</span>
<span class="go">[&#39;R&#39;, &#39;S&#39;, &#39;V&#39;, &#39;I&#39;, &#39;L&#39;, &#39;R&#39;, &#39;A&#39;, &#39;Y&#39;, &#39;T&#39;, &#39;N&#39;]</span>
</pre></div>
</div>
<p>We can get a list of the libraries for which we have barcodes:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">libraries</span>
<span class="go">[&#39;lib_1&#39;, &#39;lib_2&#39;]</span>
</pre></div>
</div>
<p>We can also get a Data Frame that includes the information
about the variants along with additional columns containing amino-acid
mutations and mutation counts via the <cite>barcode_variant_df</cite> attribute
of our
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html">CodonVariantTable</a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">barcode_variant_df</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
<span class="go">  library           barcode  variant_call_support                 codon_substitutions    aa_substitutions  n_codon_substitutions  n_aa_substitutions</span>
<span class="go">0   lib_1  AAAAAACGTTTTGTCC                     1                             AGA1TGT                 R1C                      1                   1</span>
<span class="go">1   lib_1  AAAAAAGACGACCCAT                     1                   AAC20TAA CGC26CAG           N20* R26Q                      2                   2</span>
<span class="go">2   lib_1  AAAAAAGCTTCATTTG                     3  AGA1CCG CGC26TGG AAC36TTA GGA50ATC  R1P R26W N36L G50I                      4                   4</span>
<span class="go">3   lib_1  AAAAAAGGTGACAATA                     3                            GTA16TCA                V16S                      1                   1</span>
</pre></div>
</div>
</div>
<div class="section" id="analyze-the-codon-variant-table">
<h3><a class="toc-backref" href="#id6">Analyze the codon variant table</a><a class="headerlink" href="#analyze-the-codon-variant-table" title="Permalink to this headline">¶</a></h3>
<p>The
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html">CodonVariantTable</a>
has a large number of methods to get or plot information about the
variants (look at the
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html">documentation</a>
for details on these methods).</p>
<p>Eventually, we will add the counts of different variants for specific
samples. But for now, we haven’t added any samples. However, we can use
the analysis methods of our
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html">CodonVariantTable</a>
to get information about the variant composition itself by calling them
with <cite>samples=None</cite>. When we do that, we simply get</p>
<p>Get the number of variants in each library using
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.n_variants_df">CodonVariantTable.n_variants_df</a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">n_variants_df</span><span class="p">(</span><span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="go">         library             sample  count</span>
<span class="go">0          lib_1  barcoded variants  20000</span>
<span class="go">1          lib_2  barcoded variants  20000</span>
<span class="go">2  all libraries  barcoded variants  40000</span>
</pre></div>
</div>
<p>Get the set of valid barcodes for each library using
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.valid_barcodes">CodonVariantTable.valid_barcodes</a>.
Here we just show the first few valid barcodes for the first library:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">sorted</span><span class="p">(</span><span class="n">variants</span><span class="o">.</span><span class="n">valid_barcodes</span><span class="p">(</span><span class="s1">&#39;lib_1&#39;</span><span class="p">))[</span> <span class="p">:</span> <span class="mi">3</span><span class="p">]</span>
<span class="go">[&#39;AAAAAACGTTTTGTCC&#39;, &#39;AAAAAAGACGACCCAT&#39;, &#39;AAAAAAGCTTCATTTG&#39;]</span>
</pre></div>
</div>
<p>Plots the number of mutations per variant using
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.plotNumMutsHistogram">CodonVariantTable.plotNumMutsHistogram</a>.
Here we do that for amino-acid mutations:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotNumMutsHistogram</span><span class="p">(</span><span class="s1">&#39;aa&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
<p>(<a class="reference external" href=".//codonvariant_analysis_demo-13.png">png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-13.hires.png">hires.png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-13.pdf">pdf</a>)</p>
<div class="figure">
<img alt="_images/codonvariant_analysis_demo-13.png" src="_images/codonvariant_analysis_demo-13.png" />
</div>
<p>We can make a similar plot for codon mutations:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotNumMutsHistogram</span><span class="p">(</span><span class="s1">&#39;codon&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
<p>(<a class="reference external" href=".//codonvariant_analysis_demo-14.png">png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-14.hires.png">hires.png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-14.pdf">pdf</a>)</p>
<div class="figure">
<img alt="_images/codonvariant_analysis_demo-14.png" src="_images/codonvariant_analysis_demo-14.png" />
</div>
<p>Examine the average frequency of codon mutations of each type per
variant using
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.plotNumCodonMutsByType">CodonVariantTable.plotNumCodonMutsByType</a>.
First, we just do that for single-codon variants and wildtype:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotNumCodonMutsByType</span><span class="p">(</span><span class="s1">&#39;single&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
<p>(<a class="reference external" href=".//codonvariant_analysis_demo-15.png">png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-15.hires.png">hires.png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-15.pdf">pdf</a>)</p>
<div class="figure">
<img alt="_images/codonvariant_analysis_demo-15.png" src="_images/codonvariant_analysis_demo-15.png" />
</div>
<p>Now for variants with any number of mutations:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotNumCodonMutsByType</span><span class="p">(</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
<p>(<a class="reference external" href=".//codonvariant_analysis_demo-16.png">png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-16.hires.png">hires.png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-16.pdf">pdf</a>)</p>
<div class="figure">
<img alt="_images/codonvariant_analysis_demo-16.png" src="_images/codonvariant_analysis_demo-16.png" />
</div>
<p>Examine how well mutations are sampled in the library by looking at the
fraction of mutations seen &lt;= some number of times using
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.plotCumulMutCoverage">CodonVariantTable.plotCumulMutCoverage</a>.
Here we do that for codon mutations, looking at just for the
single-codon mutants:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotCumulMutCoverage</span><span class="p">(</span><span class="s1">&#39;single&#39;</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
<p>(<a class="reference external" href=".//codonvariant_analysis_demo-17.png">png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-17.hires.png">hires.png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-17.pdf">pdf</a>)</p>
<div class="figure">
<img alt="_images/codonvariant_analysis_demo-17.png" src="_images/codonvariant_analysis_demo-17.png" />
</div>
<p>We can also get the numerical information plotted above using
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.mutCounts">CodonVariantTable.mutCounts</a></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">mutCounts</span><span class="p">(</span><span class="s1">&#39;single&#39;</span><span class="p">,</span> <span class="s1">&#39;aa&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
<span class="go">  library             sample mutation  count  mutation_type  site</span>
<span class="go">0   lib_1  barcoded variants      R1S     24  nonsynonymous     1</span>
<span class="go">1   lib_1  barcoded variants     I28R     23  nonsynonymous    28</span>
<span class="go">2   lib_1  barcoded variants      V3R     23  nonsynonymous     3</span>
<span class="go">3   lib_1  barcoded variants     P24S     22  nonsynonymous    24</span>
</pre></div>
</div>
<p>Here are the frequencies of mutations along the gene in our library
plotted using
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.plotMutFreqs">CodonVariantTable.plotMutFreqs</a>,
looking at just single amino-acid variants:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotMutFreqs</span><span class="p">(</span><span class="s1">&#39;single&#39;</span><span class="p">,</span> <span class="s1">&#39;aa&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
<p>(<a class="reference external" href=".//codonvariant_analysis_demo-19.png">png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-19.hires.png">hires.png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-19.pdf">pdf</a>)</p>
<div class="figure">
<img alt="_images/codonvariant_analysis_demo-19.png" src="_images/codonvariant_analysis_demo-19.png" />
</div>
<p>We can also look at comparable information in a heat-map form using
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.plotMutHeatMap">CodonVariantTable.plotMutHeatMap</a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotMutHeatmap</span><span class="p">(</span><span class="s1">&#39;single&#39;</span><span class="p">,</span> <span class="s1">&#39;aa&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
<p>(<a class="reference external" href=".//codonvariant_analysis_demo-20.png">png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-20.hires.png">hires.png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-20.pdf">pdf</a>)</p>
<div class="figure">
<img alt="_images/codonvariant_analysis_demo-20.png" src="_images/codonvariant_analysis_demo-20.png" />
</div>
</div>
</div>
<div class="section" id="simulate-variant-counts">
<h2><a class="toc-backref" href="#id7">Simulate variant counts</a><a class="headerlink" href="#simulate-variant-counts" title="Permalink to this headline">¶</a></h2>
<p>An experiment involves looking at how the counts of the different
variants changes under selection. In a <strong>real</strong> experiment, you would
obtain the variant counts by sequencing the barcodes pre- and
post-selection.</p>
<p>However, here we <strong>simulate</strong> an experiment by simulating codon counts
based on a known phenotype for each variant to show demonstrate the
analysis.</p>
<p>We use
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.simulateSampleCounts">simulateSampleCounts</a>
to simulate data for an experiment. This function simulates the counts
of variants pre- and post-selection; the change in variant counts after
selection indicates the effect of a mutation.</p>
<div class="section" id="define-phenotype-function">
<h3><a class="toc-backref" href="#id8">Define phenotype function</a><a class="headerlink" href="#define-phenotype-function" title="Permalink to this headline">¶</a></h3>
<p>First, we define a “phenotype” function. We will do this using a
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.phenotypeSimulator">PhenotypeSimulator</a>,
which follow the concepts of <a class="reference external" href="https://doi.org/10.1073/pnas.1804015115">Otwinoski et
al</a> and <a class="reference external" href="https://doi.org/10.1534/genetics.116.195214">Sailer and
Harms</a> to define the
phenotype in two steps: an underlying latent phenotype that mutations
affect additively, and then a non-linear relationship between this
latent phenotype and the observed phenotype.</p>
<p>First, we initialize a
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.phenotypeSimulator">PhenotypeSimulator</a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">phenosimulator</span> <span class="o">=</span> <span class="n">PhenotypeSimulator</span><span class="p">(</span><span class="n">geneseq</span><span class="p">)</span>
</pre></div>
</div>
<p>We plot the relationship of the latent and observed phenotypes, with a
dashed vertical line indicating the wildtype phenotype:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">phenosimulator</span><span class="o">.</span><span class="n">plotLatentVersusObservedPhenotype</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
<p>(<a class="reference external" href=".//codonvariant_analysis_demo-22.png">png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-22.hires.png">hires.png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-22.pdf">pdf</a>)</p>
<div class="figure">
<img alt="_images/codonvariant_analysis_demo-22.png" src="_images/codonvariant_analysis_demo-22.png" />
</div>
<p>Next, we plot the latent and observed phenotypes of all single mutants,
with a dashed vertical line indicating the wildtype phenotype:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">phenotype</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;latent&#39;</span><span class="p">,</span> <span class="s1">&#39;observed&#39;</span><span class="p">]:</span>
<span class="gp">... </span>    <span class="n">p</span> <span class="o">=</span> <span class="n">phenosimulator</span><span class="o">.</span><span class="n">plotMutsHistogram</span><span class="p">(</span><span class="n">phenotype</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
<div class="figure" id="id1">
<img alt="_images/codonvariant_analysis_demo-23_00.png" src="_images/codonvariant_analysis_demo-23_00.png" />
<p class="caption"><span class="caption-text">(<a class="reference external" href=".//codonvariant_analysis_demo-23_00.png">png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-23_00.hires.png">hires.png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-23_00.pdf">pdf</a>)</span></p>
</div>
<div class="figure" id="id2">
<img alt="_images/codonvariant_analysis_demo-23_01.png" src="_images/codonvariant_analysis_demo-23_01.png" />
<p class="caption"><span class="caption-text">(<a class="reference external" href=".//codonvariant_analysis_demo-23_01.png">png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-23_01.hires.png">hires.png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-23_01.pdf">pdf</a>)</span></p>
</div>
</div>
<div class="section" id="simulate-variant-counts-for-samples">
<h3><a class="toc-backref" href="#id9">Simulate variant counts for samples</a><a class="headerlink" href="#simulate-variant-counts-for-samples" title="Permalink to this headline">¶</a></h3>
<p>Now we use
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.simulateSampleCounts">simulateSampleCounts</a>
to simulate the counts of variants assuming selection on each variant
proportional to its observed phenotype as defined above. This function
allows us to tune various aspects of the simulations to make them
reflect real experiments, such as by tweaking the error-rate in variant
calling, the sequencing depth, the bottlenecks during the experiment,
and non-uniformity in library composition.</p>
<p>We simulate under a range of conditions spanning reasonable values of
the parameters that are likely to affect the outcome of a real
experiment:</p>
<blockquote>
<div><ul class="simple">
<li>The case where the mutations in each variant are called
perfectly with no errors, and where 5% of them contain an erroneously
mis-called codon mutation.</li>
<li>Where there is a low
(small) or big (loose) bottleneck between the pre- and post-selection
steps. These bottlenecks are equal to 5 or 50 times the total number of
variants in each library.</li>
</ul>
</div></blockquote>
<p>For all simulations, we assume variant composition of the library is
reasonably but not completely uniform.
Although the code in the next cell is set up to simulate sequencing
to several different depths, here we only simulate “deep” depth
of 500 times the number of variants in the library.</p>
<p>Simulate sample counts under all of these conditions, and put them in a
data frame:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">post_to_pre</span> <span class="o">=</span> <span class="p">{}</span> <span class="c1"># dict mapping each post-selection sample to its pre-selection sample</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts_df</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1"># list of data frames for each selection</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">err_str</span><span class="p">,</span> <span class="n">err</span> <span class="ow">in</span> <span class="p">[(</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;err_&#39;</span><span class="p">,</span> <span class="mf">0.05</span><span class="p">)]:</span>
<span class="gp">...</span>
<span class="gp">... </span>    <span class="k">for</span> <span class="n">depth_str</span><span class="p">,</span> <span class="n">depth</span> <span class="ow">in</span> <span class="p">[(</span><span class="s1">&#39;deep&#39;</span><span class="p">,</span> <span class="mi">500</span> <span class="o">*</span> <span class="n">variants_per_lib</span><span class="p">)]:</span>
<span class="gp">...</span>
<span class="gp">... </span>        <span class="n">sample_prefix</span> <span class="o">=</span> <span class="n">f</span><span class="s2">&quot;</span><span class="si">{err_str}{depth_str}</span><span class="s2">_&quot;</span>
<span class="gp">... </span>        <span class="n">pre_sample_name</span> <span class="o">=</span> <span class="n">sample_prefix</span> <span class="o">+</span> <span class="s1">&#39;pre&#39;</span>
<span class="gp">... </span>        <span class="n">post_lowbottle_sample_name</span> <span class="o">=</span> <span class="n">sample_prefix</span> <span class="o">+</span> <span class="s1">&#39;lowbottle&#39;</span>
<span class="gp">... </span>        <span class="n">post_bigbottle_sample_name</span> <span class="o">=</span> <span class="n">sample_prefix</span> <span class="o">+</span> <span class="s1">&#39;bigbottle&#39;</span>
<span class="gp">... </span>        <span class="n">post_to_pre</span><span class="p">[</span><span class="n">post_lowbottle_sample_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">pre_sample_name</span>
<span class="gp">... </span>        <span class="n">post_to_pre</span><span class="p">[</span><span class="n">post_bigbottle_sample_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">pre_sample_name</span>
<span class="gp">...</span>
<span class="gp">... </span>        <span class="n">counts_df</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
<span class="gp">... </span>                <span class="n">simulateSampleCounts</span><span class="p">(</span>
<span class="gp">... </span>                        <span class="n">variants</span><span class="o">=</span><span class="n">variants</span><span class="p">,</span>
<span class="gp">... </span>                        <span class="n">phenotype_func</span><span class="o">=</span><span class="n">phenosimulator</span><span class="o">.</span><span class="n">observedPhenotype</span><span class="p">,</span>
<span class="gp">... </span>                        <span class="n">variant_error_rate</span><span class="o">=</span><span class="n">err</span><span class="p">,</span>
<span class="gp">... </span>                        <span class="n">pre_sample</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;total_count&#39;</span><span class="p">:</span><span class="n">depth</span><span class="p">,</span>
<span class="gp">... </span>                                    <span class="s1">&#39;uniformity&#39;</span><span class="p">:</span><span class="mi">10</span><span class="p">},</span>
<span class="gp">... </span>                        <span class="n">pre_sample_name</span><span class="o">=</span><span class="n">pre_sample_name</span><span class="p">,</span>
<span class="gp">... </span>                        <span class="n">post_samples</span><span class="o">=</span><span class="p">{</span>
<span class="gp">... </span>                            <span class="n">post_bigbottle_sample_name</span><span class="p">:{</span>
<span class="gp">... </span>                                    <span class="s1">&#39;total_count&#39;</span><span class="p">:</span><span class="n">depth</span><span class="p">,</span>
<span class="gp">... </span>                                    <span class="s1">&#39;noise&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span>
<span class="gp">... </span>                                    <span class="s1">&#39;bottleneck&#39;</span><span class="p">:</span><span class="mi">50</span> <span class="o">*</span> <span class="n">variants_per_lib</span><span class="p">},</span>
<span class="gp">... </span>                            <span class="n">post_lowbottle_sample_name</span><span class="p">:{</span>
<span class="gp">... </span>                                    <span class="s1">&#39;total_count&#39;</span><span class="p">:</span><span class="n">depth</span><span class="p">,</span>
<span class="gp">... </span>                                    <span class="s1">&#39;noise&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">,</span>
<span class="gp">... </span>                                    <span class="s1">&#39;bottleneck&#39;</span><span class="p">:</span><span class="mi">5</span> <span class="o">*</span> <span class="n">variants_per_lib</span><span class="p">},</span>
<span class="gp">... </span>                            <span class="p">},</span>
<span class="gp">... </span>                        <span class="p">)</span>
<span class="gp">... </span>                 <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># combine all the simulated data frames</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">counts_df</span><span class="p">)</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">post_to_pre</span><span class="o">.</span><span class="n">items</span><span class="p">()),</span>
<span class="gp">... </span>                          <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;post-selection&#39;</span><span class="p">,</span> <span class="s1">&#39;pre-selection&#39;</span><span class="p">])</span>
<span class="go">       post-selection pre-selection</span>
<span class="go">0      deep_lowbottle      deep_pre</span>
<span class="go">1      deep_bigbottle      deep_pre</span>
<span class="go">2  err_deep_lowbottle  err_deep_pre</span>
<span class="go">3  err_deep_bigbottle  err_deep_pre</span>
</pre></div>
</div>
<p>Here are the first few lines of the data frame with the simulated
counts:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts_df</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="go">  library           barcode    sample  count</span>
<span class="go">0   lib_1  AAAAAACGTTTTGTCC  deep_pre    490</span>
<span class="go">1   lib_1  AAAAAAGACGACCCAT  deep_pre    831</span>
<span class="go">2   lib_1  AAAAAAGCTTCATTTG  deep_pre    424</span>
<span class="go">3   lib_1  AAAAAAGGTGACAATA  deep_pre    701</span>
<span class="go">4   lib_1  AAAAAATACGGTCAGC  deep_pre    394</span>
</pre></div>
</div>
</div>
<div class="section" id="add-counts-to-variant-table">
<h3><a class="toc-backref" href="#id10">Add counts to variant table</a><a class="headerlink" href="#add-counts-to-variant-table" title="Permalink to this headline">¶</a></h3>
<p>We now add the simulated counts to the
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html">CodonVariantTable</a>
using its
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.addSampleCounts">addSampleCounts</a>
method:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">counts_df</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
<span class="gp">... </span>    <span class="k">for</span> <span class="n">lib</span> <span class="ow">in</span> <span class="n">libs</span><span class="p">:</span>
<span class="gp">... </span>        <span class="n">icounts</span> <span class="o">=</span> <span class="n">counts_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;sample == @sample &amp; library == @lib&#39;</span><span class="p">)</span>
<span class="gp">... </span>        <span class="n">variants</span><span class="o">.</span><span class="n">addSampleCounts</span><span class="p">(</span><span class="n">lib</span><span class="p">,</span> <span class="n">sample</span><span class="p">,</span> <span class="n">icounts</span><span class="p">)</span>
</pre></div>
</div>
<p>We use the
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.n_variants_df">CodonVariantTable.n_variants_df</a>
method to confirm that we have the expected number of variant counts for
each sample:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">n_variants_df</span><span class="p">()</span>
<span class="go">          library              sample     count</span>
<span class="go">0           lib_1            deep_pre  10000000</span>
<span class="go">1           lib_1      deep_bigbottle  10000000</span>
<span class="go">2           lib_1      deep_lowbottle  10000000</span>
<span class="go">3           lib_1        err_deep_pre  10000000</span>
<span class="go">4           lib_1  err_deep_bigbottle  10000000</span>
<span class="go">5           lib_1  err_deep_lowbottle  10000000</span>
<span class="go">6           lib_2            deep_pre  10000000</span>
<span class="go">7           lib_2      deep_bigbottle  10000000</span>
<span class="go">8           lib_2      deep_lowbottle  10000000</span>
<span class="go">9           lib_2        err_deep_pre  10000000</span>
<span class="go">10          lib_2  err_deep_bigbottle  10000000</span>
<span class="go">11          lib_2  err_deep_lowbottle  10000000</span>
<span class="go">12  all libraries            deep_pre  20000000</span>
<span class="go">13  all libraries      deep_bigbottle  20000000</span>
<span class="go">14  all libraries      deep_lowbottle  20000000</span>
<span class="go">15  all libraries        err_deep_pre  20000000</span>
<span class="go">16  all libraries  err_deep_bigbottle  20000000</span>
<span class="go">17  all libraries  err_deep_lowbottle  20000000</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="analyze-variant-counts-for-samples">
<h2><a class="toc-backref" href="#id11">Analyze variant counts for samples</a><a class="headerlink" href="#analyze-variant-counts-for-samples" title="Permalink to this headline">¶</a></h2>
<p>Now we analyze the variant counts for the samples that we have added to
our
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html">CodonVariantTable</a>.</p>
<div class="section" id="directly-access-variant-counts">
<h3><a class="toc-backref" href="#id12">Directly access variant counts</a><a class="headerlink" href="#directly-access-variant-counts" title="Permalink to this headline">¶</a></h3>
<p>The variant counts are in the <cite>variant_count_df</cite> attribute of the
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html">CodonVariantTable</a>.
Here is what the first few lines of that Data Frame look like:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">variant_count_df</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
<span class="go">            barcode  count library    sample  variant_call_support                          codon_substitutions    aa_substitutions  n_codon_substitutions  n_aa_substitutions</span>
<span class="go">0  AAATCACTAATATGGG   1308   lib_1  deep_pre                     1                                                                                       0                   0</span>
<span class="go">1  TTAATGAAAATAAGTC   1301   lib_1  deep_pre                     1  GTG3ACA GTG13CGA CCT24GAG GAC38GAT CCA44ATC  V3T V13R P24E P44I                      5                   4</span>
<span class="go">2  TGATAGAAATTATTTG   1294   lib_1  deep_pre                     3                                     ATC17CGA                I17R                      1                   1</span>
<span class="go">3  TTATCAATTGCGGATG   1294   lib_1  deep_pre                     3                             GTG3TAC AAA14CTC            V3Y K14L                      2                   2</span>
</pre></div>
</div>
</div>
<div class="section" id="distribution-of-variant-counts">
<h3><a class="toc-backref" href="#id13">Distribution of variant counts</a><a class="headerlink" href="#distribution-of-variant-counts" title="Permalink to this headline">¶</a></h3>
<p>We plot the number of counts for each variant in each sample. The
horizontal line shows the total number of variants. The plot shows that
all variants are well-sampled in the pre-selection libraries, but that
post-selection some variants are not sampled or sampled very little.
This is as expected since selection will decrease and increase the
frequency of variants.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotCumulVariantCounts</span><span class="p">(</span><span class="n">heightscale</span><span class="o">=</span><span class="mf">1.3</span><span class="p">,</span> <span class="n">widthscale</span><span class="o">=</span><span class="mf">1.2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
<p>(<a class="reference external" href=".//codonvariant_analysis_demo-29.png">png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-29.hires.png">hires.png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-29.pdf">pdf</a>)</p>
<div class="figure">
<img alt="_images/codonvariant_analysis_demo-29.png" src="_images/codonvariant_analysis_demo-29.png" />
</div>
</div>
<div class="section" id="mutation-frequencies-in-each-sample">
<h3><a class="toc-backref" href="#id14">Mutation frequencies in each sample</a><a class="headerlink" href="#mutation-frequencies-in-each-sample" title="Permalink to this headline">¶</a></h3>
<p>We look at the counts of each variant as a function of the number of
amino-acid mutations. The plots
below show that selection enriches for variants with fewer mutations,
which is as expected since many mutations are deleterious:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotNumMutsHistogram</span><span class="p">(</span><span class="n">mut_type</span><span class="o">=</span><span class="s1">&#39;aa&#39;</span><span class="p">,</span> <span class="n">heightscale</span><span class="o">=</span><span class="mf">1.3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
<p>(<a class="reference external" href=".//codonvariant_analysis_demo-30.png">png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-30.hires.png">hires.png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-30.pdf">pdf</a>)</p>
<div class="figure">
<img alt="_images/codonvariant_analysis_demo-30.png" src="_images/codonvariant_analysis_demo-30.png" />
</div>
<p>Next we plot the average number of codon mutations of each type
(nonsynonymous, synonymous, and stop). We do this for <em>all</em> variants (as
opposed to just single mutants). We see that as expected, there is
strong selection against stop codons, and some selection against
nonsynonymous ones:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotNumCodonMutsByType</span><span class="p">(</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">heightscale</span><span class="o">=</span><span class="mf">1.5</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
<p>(<a class="reference external" href=".//codonvariant_analysis_demo-31.png">png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-31.hires.png">hires.png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-31.pdf">pdf</a>)</p>
<div class="figure">
<img alt="_images/codonvariant_analysis_demo-31.png" src="_images/codonvariant_analysis_demo-31.png" />
</div>
<p>Here are the mutation frequencies as a function of primary sequence.
This plot is for all variants, showing codon mutations:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotMutFreqs</span><span class="p">(</span><span class="n">variant_type</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">mut_type</span><span class="o">=</span><span class="s1">&#39;codon&#39;</span><span class="p">,</span> <span class="n">heightscale</span><span class="o">=</span><span class="mf">1.9</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
<p>(<a class="reference external" href=".//codonvariant_analysis_demo-32.png">png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-32.hires.png">hires.png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-32.pdf">pdf</a>)</p>
<div class="figure">
<img alt="_images/codonvariant_analysis_demo-32.png" src="_images/codonvariant_analysis_demo-32.png" />
</div>
</div>
<div class="section" id="mutation-sampling">
<h3><a class="toc-backref" href="#id15">Mutation sampling</a><a class="headerlink" href="#mutation-sampling" title="Permalink to this headline">¶</a></h3>
<p>We look at how thoroughly the different possible mutations are sampled
in our libraries by plotting the fraction of mutations found &lt;= an
indicated number of times. We can do this separately for single-mutants
only or for all variants, and for codon or amino-acid mutations. Here we
do it all variants, showing codon mutations. As the plot below shows,
sampling of stop and to a lesser degree nonsynonymous mutations drops
strongly after selection, as expected since many of these are
deleterious:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotCumulMutCoverage</span><span class="p">(</span><span class="n">variant_type</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">mut_type</span><span class="o">=</span><span class="s1">&#39;codon&#39;</span><span class="p">,</span>
<span class="gp">... </span>                                  <span class="n">heightscale</span><span class="o">=</span><span class="mf">1.8</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
<p>(<a class="reference external" href=".//codonvariant_analysis_demo-33.png">png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-33.hires.png">hires.png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-33.pdf">pdf</a>)</p>
<div class="figure">
<img alt="_images/codonvariant_analysis_demo-33.png" src="_images/codonvariant_analysis_demo-33.png" />
</div>
</div>
</div>
<div class="section" id="functional-scores-for-variants">
<h2><a class="toc-backref" href="#id16">Functional scores for variants</a><a class="headerlink" href="#functional-scores-for-variants" title="Permalink to this headline">¶</a></h2>
<p>We want to estimate a functional score (similar to “fitness” in the
experimental assay) for each variant based on its change in frequency
post-selection relative to its initial pre-selection frequency.</p>
<div class="section" id="calculate-functional-scores">
<h3><a class="toc-backref" href="#id17">Calculate functional scores</a><a class="headerlink" href="#calculate-functional-scores" title="Permalink to this headline">¶</a></h3>
<p>We can calculate functional scores using the
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.func_scores">func_scores</a>
method of the
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html">CodonVariantTable</a>.
The documentation for that method explains exactly how the scores are
calculated.</p>
<p>In order to calculate functional scores, we need to pair each
post-selection sample with a pre-selection one. We use the pairing
described above when we simulated the samples, which we re-print here:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">post_to_pre</span><span class="o">.</span><span class="n">items</span><span class="p">()),</span>
<span class="gp">... </span>                          <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;post-selection&#39;</span><span class="p">,</span> <span class="s1">&#39;pre-selection&#39;</span><span class="p">])</span>
<span class="go">       post-selection pre-selection</span>
<span class="go">0      deep_lowbottle      deep_pre</span>
<span class="go">1      deep_bigbottle      deep_pre</span>
<span class="go">2  err_deep_lowbottle  err_deep_pre</span>
<span class="go">3  err_deep_bigbottle  err_deep_pre</span>
</pre></div>
</div>
<p>Now we calculate the functional scores for each barcoded variant:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">func_scores</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">func_scores</span><span class="p">(</span><span class="n">post_to_pre</span><span class="p">)</span>
</pre></div>
</div>
<p>We can see that the resulting data frame has a functional score (and its
variance) for each barcoded variant:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">func_scores</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
<span class="go">  library pre_sample     post_sample           barcode  func_score  func_score_var  pre_count  post_count  pre_count_wt  post_count_wt  pseudocount                 codon_substitutions  n_codon_substitutions    aa_substitutions  n_aa_substitutions</span>
<span class="go">0   lib_1   deep_pre  deep_bigbottle  AAAAAACGTTTTGTCC   -0.233554        0.007166        490         712       2735936        4672599          0.5                             AGA1TGT                      1                 R1C                   1</span>
<span class="go">1   lib_1   deep_pre  deep_bigbottle  AAAAAAGACGACCCAT   -9.149835        0.835052        831           2       2735936        4672599          0.5                   AAC20TAA CGC26CAG                      2           N20* R26Q                   2</span>
<span class="go">2   lib_1   deep_pre  deep_bigbottle  AAAAAAGCTTCATTTG  -10.501811        4.167642        424           0       2735936        4672599          0.5  AGA1CCG CGC26TGG AAC36TTA GGA50ATC                      4  R1P R26W N36L G50I                   4</span>
<span class="go">3   lib_1   deep_pre  deep_bigbottle  AAAAAAGGTGACAATA   -0.323361        0.005142        701         957       2735936        4672599          0.5                            GTA16TCA                      1                V16S                   1</span>
</pre></div>
</div>
<p>We can also calculate functional scores at the level of amino-acid or
codon substitutions rather than at the level of variants. The difference
is that this calculation groups all variants with a given set of
substitutions before calculating the functional score.</p>
<p>Here are are scores grouping by amino-acid substitutions. Note that
when aggregating at the level of amino-acid substitutions, it makes more
sense to calculate the functional scores relative to all variants with
wild-type amino acid sequences by using <cite>syn_as_wt</cite> (by default, the
calculation is relative to variants with wildtype codon sequences)</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">func_scores_aa</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">func_scores</span><span class="p">(</span><span class="n">post_to_pre</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="s1">&#39;aa_substitutions&#39;</span><span class="p">,</span>
<span class="gp">... </span>                                      <span class="n">syn_as_wt</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">func_scores_aa</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
<span class="go">  library pre_sample     post_sample aa_substitutions  func_score  func_score_var  pre_count  post_count  pre_count_wt  post_count_wt  pseudocount  n_aa_substitutions</span>
<span class="go">0   lib_1   deep_pre  deep_bigbottle                     0.000000        0.000002    2901007     4955429       2901007        4955429          0.5                   0</span>
<span class="go">1   lib_1   deep_pre  deep_bigbottle             A39*   -8.420165        0.097292       4311          21       2901007        4955429          0.5                   1</span>
<span class="go">2   lib_1   deep_pre  deep_bigbottle        A39* F41H   -8.528679        0.836400        540           2       2901007        4955429          0.5                   2</span>
<span class="go">3   lib_1   deep_pre  deep_bigbottle        A39* G50S  -10.815483        4.166685        527           0       2901007        4955429          0.5                   2</span>
</pre></div>
</div>
<p>Since all libraries have the same samples, we can also calculate functional
scores aggregating across libraries using the <cite>combine_libs</cite> option,
although we don’t do that here.</p>
</div>
<div class="section" id="distribution-of-functional-scores">
<h3><a class="toc-backref" href="#id18">Distribution of functional scores</a><a class="headerlink" href="#distribution-of-functional-scores" title="Permalink to this headline">¶</a></h3>
<p>We can plot the distribution of the functional scores (here taking the
scores at the level of barcoded variants).</p>
<p>Such plots are most informative if we additionally classify variants
by the “types” of mutations they have, which we do here using the
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.classifyVariants">CodonVariantTable.classifyVariants</a> method, which adds a <cite>variant_class</cite> column to the data frame:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">func_scores</span> <span class="o">=</span> <span class="n">CodonVariantTable</span><span class="o">.</span><span class="n">classifyVariants</span><span class="p">(</span><span class="n">func_scores</span><span class="p">)</span>
</pre></div>
</div>
<p>Now we plot the distributions of scores, coloring by the variant class:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
<span class="gp">... </span>    <span class="n">ggplot</span><span class="p">(</span><span class="n">func_scores</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="s1">&#39;variant_class&#39;</span><span class="p">,</span> <span class="s1">&#39;func_score&#39;</span><span class="p">))</span> <span class="o">+</span>
<span class="gp">... </span>    <span class="n">geom_violin</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">fill</span><span class="o">=</span><span class="s1">&#39;variant_class&#39;</span><span class="p">))</span> <span class="o">+</span>
<span class="gp">... </span>    <span class="n">ylab</span><span class="p">(</span><span class="s1">&#39;functional score&#39;</span><span class="p">)</span> <span class="o">+</span>
<span class="gp">... </span>    <span class="n">xlab</span><span class="p">(</span><span class="s1">&#39;sample&#39;</span><span class="p">)</span> <span class="o">+</span>
<span class="gp">... </span>    <span class="n">facet_grid</span><span class="p">(</span><span class="s1">&#39;post_sample ~ library&#39;</span><span class="p">)</span> <span class="o">+</span>
<span class="gp">... </span>    <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">7</span><span class="p">,</span> <span class="mi">9</span><span class="p">),</span>
<span class="gp">... </span>          <span class="n">axis_text_x</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="mi">90</span><span class="p">))</span> <span class="o">+</span>
<span class="gp">... </span>    <span class="n">scale_fill_manual</span><span class="p">(</span><span class="n">values</span><span class="o">=</span><span class="n">CBPALETTE</span><span class="p">[</span><span class="mi">1</span> <span class="p">:],</span> <span class="n">guide</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">... </span>    <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
<p>(<a class="reference external" href=".//codonvariant_analysis_demo-39.png">png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-39.hires.png">hires.png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-39.pdf">pdf</a>)</p>
<div class="figure">
<img alt="_images/codonvariant_analysis_demo-39.png" src="_images/codonvariant_analysis_demo-39.png" />
</div>
<p>We see the expected behavior: stop codon variants have low scores,
nonsynonymous variants have a spread of scores, and wildtype /
synonymous variants have scores around zero. Note, however, that when
there are low bottlenecks or sequencing errors then sometimes even
wildtype / synonymous variants end up with low scores due to the errors
or bottlenecking noise.</p>
</div>
<div class="section" id="functional-scores-versus-phenotype">
<h3><a class="toc-backref" href="#id19">Functional scores versus phenotype</a><a class="headerlink" href="#functional-scores-versus-phenotype" title="Permalink to this headline">¶</a></h3>
<p>Since we simulated the data, we know the <em>true</em> observed phenotype for
each variant. We can therefore correlate the true phenotype with the
observed functional score. Note that given how we defined things above,
we expect the <span class="math notranslate nohighlight">\(\log_2\)</span> of the true phenotype to equal the observed
functional score.</p>
<p>We add the true phenotypes to our data frame of functional scores:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">func_scores</span> <span class="o">=</span> <span class="p">(</span>
<span class="gp">... </span>    <span class="n">func_scores</span>
<span class="gp">... </span>    <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">observed_phenotype</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
<span class="gp">... </span>                <span class="n">phenosimulator</span><span class="o">.</span><span class="n">observedPhenotype</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span>
<span class="gp">... </span>            <span class="n">latent_phenotype</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
<span class="gp">... </span>                <span class="n">phenosimulator</span><span class="o">.</span><span class="n">latentPhenotype</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span>
<span class="gp">... </span>            <span class="n">log_observed_phenotype</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">scipy</span><span class="o">.</span><span class="n">log</span><span class="p">(</span>
<span class="gp">... </span>                <span class="n">x</span><span class="o">.</span><span class="n">observed_phenotype</span><span class="p">)</span> <span class="o">/</span> <span class="n">scipy</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mi">2</span><span class="p">),</span>
<span class="gp">... </span>            <span class="p">)</span>
<span class="gp">... </span>    <span class="p">)</span>
</pre></div>
</div>
<p>For highly deleterious mutations at some point it becomes irrelevant to
estimate the exact score as long as we know that it’s very low. In fact,
we don’t expect to be able to accurately estimate the score for highly
impaired variants with very low scores.</p>
<p>So we also calculate a “clipped” score where we simply assign all
functional scores and log observed phenotypes less than -8 to a value of
-8. The reason is that the simulated experiments don’t appear to
estimate values much lower than -8:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">score_clip</span> <span class="o">=</span> <span class="o">-</span><span class="mi">8</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">func_scores</span> <span class="o">=</span> <span class="p">(</span>
<span class="gp">... </span>    <span class="n">func_scores</span>
<span class="gp">... </span>    <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">func_score_clipped</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">scipy</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span>
<span class="gp">... </span>                    <span class="n">x</span><span class="o">.</span><span class="n">func_score</span><span class="p">,</span> <span class="n">score_clip</span><span class="p">,</span> <span class="kc">None</span><span class="p">),</span>
<span class="gp">... </span>            <span class="n">log_observed_phenotype_clipped</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">scipy</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span>
<span class="gp">... </span>                    <span class="n">x</span><span class="o">.</span><span class="n">log_observed_phenotype</span><span class="p">,</span> <span class="n">score_clip</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
<span class="gp">... </span>            <span class="p">)</span>
<span class="gp">... </span>    <span class="p">)</span>
</pre></div>
</div>
<p>Now plot the correlation after this clipping at the low end:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
<span class="gp">... </span>    <span class="n">ggplot</span><span class="p">(</span><span class="n">func_scores</span><span class="p">,</span>
<span class="gp">... </span>           <span class="n">aes</span><span class="p">(</span><span class="s1">&#39;log_observed_phenotype_clipped&#39;</span><span class="p">,</span> <span class="s1">&#39;func_score_clipped&#39;</span><span class="p">))</span> <span class="o">+</span>
<span class="gp">... </span>    <span class="n">geom_point</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span> <span class="o">+</span>
<span class="gp">... </span>    <span class="n">facet_grid</span><span class="p">(</span><span class="s1">&#39;post_sample ~ library&#39;</span><span class="p">)</span> <span class="o">+</span>
<span class="gp">... </span>    <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">9</span><span class="p">))</span> <span class="o">+</span>
<span class="gp">... </span>    <span class="n">xlab</span><span class="p">(</span><span class="s1">&#39;true phenotype&#39;</span><span class="p">)</span> <span class="o">+</span>
<span class="gp">... </span>    <span class="n">ylab</span><span class="p">(</span><span class="s1">&#39;functional score from experiment&#39;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
<p>(<a class="reference external" href=".//codonvariant_analysis_demo-42.png">png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-42.hires.png">hires.png</a>, <a class="reference external" href=".//codonvariant_analysis_demo-42.pdf">pdf</a>)</p>
<div class="figure">
<img alt="_images/codonvariant_analysis_demo-42.png" src="_images/codonvariant_analysis_demo-42.png" />
</div>
<p>Finally, calculate the Pearson correlation coefficients for the clipped
data. We see below that a small (low) bottleneck and sequencing errors
decrease the accuracy with which the measured
functional scores correlate with the true phenotype:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">corr</span> <span class="o">=</span> <span class="p">(</span>
<span class="gp">... </span>    <span class="n">func_scores</span>
<span class="gp">... </span>     <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;post_sample&#39;</span><span class="p">])</span>
<span class="gp">... </span>     <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;func_score_clipped&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span>
<span class="gp">... </span>                      <span class="n">x</span><span class="p">[</span><span class="s1">&#39;log_observed_phenotype_clipped&#39;</span><span class="p">],</span>
<span class="gp">... </span>                      <span class="n">method</span><span class="o">=</span><span class="s1">&#39;pearson&#39;</span><span class="p">))</span>
<span class="gp">... </span>     <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="s1">&#39;Pearson correlation&#39;</span><span class="p">)</span>
<span class="gp">... </span>     <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
<span class="gp">... </span>     <span class="o">.</span><span class="n">pivot_table</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="s1">&#39;post_sample&#39;</span><span class="p">,</span>
<span class="gp">... </span>                  <span class="n">values</span><span class="o">=</span><span class="s1">&#39;Pearson correlation&#39;</span><span class="p">,</span>
<span class="gp">... </span>                  <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;library&#39;</span><span class="p">)</span>
<span class="gp">... </span>     <span class="p">)</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">corr</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
<span class="go">library             lib_1  lib_2</span>
<span class="go">post_sample</span>
<span class="go">deep_bigbottle      0.996  0.996</span>
<span class="go">deep_lowbottle      0.943  0.945</span>
<span class="go">err_deep_bigbottle  0.976  0.973</span>
<span class="go">err_deep_lowbottle  0.925  0.921</span>
</pre></div>
</div>
</div>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">dms_tools2</h1>
    
  </a>
</p>



<p class="blurb">Tools for analyzing deep mutational scanning data.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=dms_tools2&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a class="badge" href="https://travis-ci.org/jbloomlab/dms_tools2">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="bcsubamp.html">Barcoded-subamplicon sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="prefs.html">Amino-acid preferences</a></li>
<li class="toctree-l1"><a class="reference internal" href="diffsel.html">Differential selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="fracsurvive.html">Fraction surviving</a></li>
<li class="toctree-l1"><a class="reference internal" href="programs.html">Executable programs</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">Python API</a></li>
<li class="toctree-l1"><a class="reference internal" href="citations.html">Citations</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017--2019, the Bloom lab.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/codonvariant_analysis_demo.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/dms_tools2" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>