
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Differential selection &#8212; dms_tools2 2.4.0 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Fraction surviving" href="fracsurvive.html" />
    <link rel="prev" title="Amino-acid preferences" href="prefs.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="differential-selection">
<span id="diffsel"></span><h1>Differential selection<a class="headerlink" href="#differential-selection" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="contents">
<ul class="simple">
<li><a class="reference internal" href="#overview" id="id3">Overview</a></li>
<li><a class="reference internal" href="#definition-of-differential-selection" id="id4">Definition of differential selection</a><ul>
<li><a class="reference internal" href="#formula-for-differential-selection" id="id5">Formula for differential selection</a></li>
<li><a class="reference internal" href="#error-correction" id="id6">Error correction</a></li>
</ul>
</li>
<li><a class="reference internal" href="#programs-and-examples" id="id7">Programs and examples</a></li>
</ul>
</div>
<div class="section" id="overview">
<h2><a class="toc-backref" href="#id3">Overview</a><a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>Deep mutational scanning libraries are often subject to multiple different selection pressures with the goal of identifying mutations that are differentially selected in one condition versus the other.
For instance, you might select a library of viral mutants with a monoclonal antibody, and also perform a mock selection without antibody.</p>
<p>One way to quantify the differential selection on mutations in the selected versus mock condition is in terms of the <em>differential selection</em> measure defined here.
A visualization of differential selection is shown in <a class="reference internal" href="#diffselsnippetfig"><span class="std std-numref">Fig. 5</span></a>.
Essentially, most mutations are <strong>not</strong> selected differently in the selected (e.g., antibody) treatment versus the mock selection – but a small number are under strong differential selection.</p>
<div class="figure align-center" id="id1" style="width: 75%">
<span id="diffselsnippetfig"></span><img alt="Logo plot visualization of differential selection." src="_images/DiffSelSnippet.jpg" />
<p class="caption"><span class="caption-number">Fig. 5 </span><span class="caption-text">Visualization of differential selection created using <a class="reference internal" href="dms2_logoplot.html#dms2-logoplot"><span class="std std-ref">dms2_logoplot</span></a>.
The height of each letter above the center line is proportional to the differential selection for that amino acid.
For instance, at site 154 there is differential selection mostly for negatively charged amino acids (aspartic acid - D, glutamic acid - E), whereas at site 157 there is differential selection for most mutations.
At other sites (such as 151), there is no differential selection for any mutation.
The scale bar indicates a differential selection of 8, which corresponds to a <span class="math notranslate nohighlight">\(2^8 = 256\)</span>-fold enrichment of the mutation.
Figure from <a class="reference external" href="http://journals.plos.org/plospathogens/article?id=10.1371/journal.ppat.1006271">Doud et al (2017)</a>.
Note that this figure only shows <strong>positive</strong> differential selection (mutations that are favored in the selected versus mock condition).</span></p>
</div>
<p>The <a class="reference external" href="http://jbloomlab.github.io/dms_tools2">dms_tools2</a> software contains programs to estimate the differential selection from the counts of mutations generated from the deep sequencing data.
Typically you would want to estimate the differential selection using <a class="reference internal" href="dms2_batch_diffsel.html#dms2-batch-diffsel"><span class="std std-ref">dms2_batch_diffsel</span></a>, which in turn calls <a class="reference internal" href="dms2_diffsel.html#dms2-diffsel"><span class="std std-ref">dms2_diffsel</span></a>.
These programs create visualizations of the results, and you can also visualize the differential selection with <a class="reference internal" href="dms2_logoplot.html#dms2-logoplot"><span class="std std-ref">dms2_logoplot</span></a> as in <a class="reference internal" href="#diffselsnippetfig"><span class="std std-numref">Fig. 5</span></a>.</p>
</div>
<div class="section" id="definition-of-differential-selection">
<h2><a class="toc-backref" href="#id4">Definition of differential selection</a><a class="headerlink" href="#definition-of-differential-selection" title="Permalink to this headline">¶</a></h2>
<p>The definition of <em>differential selection</em> used here was originally introduced in <a class="reference external" href="http://journals.plos.org/plospathogens/article?id=10.1371/journal.ppat.1006271">Doud et al (2017)</a>.
A schematic of the overall integrated experimental and computational workflow is shown in <a class="reference internal" href="#diffselworkflowfig"><span class="std std-numref">Fig. 6</span></a>.</p>
<div class="figure align-center" id="id2" style="width: 90%">
<span id="diffselworkflowfig"></span><img alt="Schematic of differential selection." src="_images/DiffSelWorkflow.jpg" />
<p class="caption"><span class="caption-number">Fig. 6 </span><span class="caption-text">Schematic of differential selection.
<strong>(A)</strong> A library of mutants is subjected to a selection (antibody neutralization) and also a mock-selection (no-antibody control).
The mutants that survive each selection are deep sequenced, and the counts of mutations relative to wildtype are analyzed to identify mutations enriched in the antibody versus no-antibody treatment.
<strong>(B)</strong> A differential selection value is computed for each mutation, and can be displayed in logo plots using <a class="reference internal" href="dms2_logoplot.html#dms2-logoplot"><span class="std std-ref">dms2_logoplot</span></a>.
The height of each letter is proportional to the differential selection for that mutation, and is termed the <em>mutation differential selection</em>.
The overall height of the letter stack is termed the <em>site differential selection</em>.
Figure from <a class="reference external" href="http://journals.plos.org/plospathogens/article?id=10.1371/journal.ppat.1006271">Doud et al (2017)</a>.</span></p>
</div>
<div class="section" id="formula-for-differential-selection">
<h3><a class="toc-backref" href="#id5">Formula for differential selection</a><a class="headerlink" href="#formula-for-differential-selection" title="Permalink to this headline">¶</a></h3>
<p>For each site <span class="math notranslate nohighlight">\(r\)</span> in the gene, let <span class="math notranslate nohighlight">\(n_{r,x}^{\rm{mock}}\)</span> be the number of observed counts of character <span class="math notranslate nohighlight">\(x\)</span> (e.g., an amino acid) at site <span class="math notranslate nohighlight">\(r\)</span> in the mock-treated condition, and let <span class="math notranslate nohighlight">\(n_{r,x}^{\rm{selected}}\)</span> be the number of observed counts of <span class="math notranslate nohighlight">\(x\)</span> at <span class="math notranslate nohighlight">\(r\)</span> in the selected condition.
Let <span class="math notranslate nohighlight">\(\operatorname{wt}\left(r\right)\)</span> be denote the wildtype character at <span class="math notranslate nohighlight">\(r\)</span>.
Then the relative enrichment of the mutant relative to the wildtype after selection is</p>
<div class="math notranslate nohighlight" id="equation-e-rx">
<span class="eqno">(18)<a class="headerlink" href="#equation-e-rx" title="Permalink to this equation">¶</a></span>\[ E_{r,x} = \frac{\left(n_{r,x}^{\rm{selected}} + f_{r, \rm{selected}} \times P\right) / \left(n_{r,\operatorname{wt}\left(r\right)}^{\rm{selected}} + f_{r, \rm{selected}} \times P\right)}{\left(n_{r,x}^{\rm{mock}} + f_{r, \rm{mock}} \times P\right) / \left(n_{r,\operatorname{wt}\left(r\right)}^{\rm{mock}} + f_{r, \rm{mock}} \times P\right)}\]</div>
<p>where <span class="math notranslate nohighlight">\(P &gt; 0\)</span> is a pseudocount that is added to each observed count (specified by <code class="docutils literal notranslate"><span class="pre">--pseudocount</span></code> option to <a class="reference internal" href="dms2_diffsel.html#dms2-diffsel"><span class="std std-ref">dms2_diffsel</span></a> / <a class="reference internal" href="dms2_batch_diffsel.html#dms2-batch-diffsel"><span class="std std-ref">dms2_batch_diffsel</span></a>), and <span class="math notranslate nohighlight">\(f_{r, \rm{selected}}\)</span> and <span class="math notranslate nohighlight">\(f_{r, \rm{mock}}\)</span> are defined so that the pseudocount is scaled up for the library (<em>mock</em> or <em>selected</em>) with higher depth at site <span class="math notranslate nohighlight">\(r\)</span>:</p>
<div class="math notranslate nohighlight" id="equation-f-rselected">
<span class="eqno">(19)<a class="headerlink" href="#equation-f-rselected" title="Permalink to this equation">¶</a></span>\[f_{r, \rm{selected}} = \max\left[1, \left(\sum_x n_{r,x}^{\rm{selected}}\right) / \left(\sum_x n_{r,x}^{\rm{mock}}\right)\right]\]</div>
<div class="math notranslate nohighlight" id="equation-f-rmock">
<span class="eqno">(20)<a class="headerlink" href="#equation-f-rmock" title="Permalink to this equation">¶</a></span>\[f_{r, \rm{mock}} = \max\left[1, \left(\sum_x n_{r,x}^{\rm{mock}}\right) / \left(\sum_x n_{r,x}^{\rm{selected}}\right)\right].\]</div>
<p>The reason for scaling the pseudocount by library depth is that the <em>mock</em> and <em>selected</em> libraries may often be sequenced at different depths.
If the same pseudocount is added to both, then estimates of <span class="math notranslate nohighlight">\(E_{r,x}\)</span> will be systematically different than one even if the relative counts for the wildtype and mutant amino acid are the same in both two conditions.
Scaling the pseudocounts by the ratio of depths fixes this problem.
Larger values of the pseudocount <span class="math notranslate nohighlight">\(P\)</span> will avoid spuriously estimating differential selection when there is just statistical noise due to low counts, but this comes at the cost of reducing sensitivity.
If you are using <code class="docutils literal notranslate"><span class="pre">--chartype</span></code> of <code class="docutils literal notranslate"><span class="pre">codon_to_aa</span></code>, the counts for amino acids are aggregated <strong>before</strong> doing the calculations above.</p>
<p>Note that by definition, <span class="math notranslate nohighlight">\(E_{r,\operatorname{wt}\left(r\right)}\)</span> is always one.</p>
<p>We quantify the <em>differential selection</em> <span class="math notranslate nohighlight">\(s_{r,x}\)</span> for the mutation to <span class="math notranslate nohighlight">\(x\)</span> at site <span class="math notranslate nohighlight">\(r\)</span> in the selected versus control condition as:</p>
<div class="math notranslate nohighlight" id="equation-mutdiffsel">
<span class="eqno">(21)<a class="headerlink" href="#equation-mutdiffsel" title="Permalink to this equation">¶</a></span>\[s_{r,x} = \log_2 E_{r,x}\]</div>
<p>These <span class="math notranslate nohighlight">\(s_{r,x}\)</span> are reported as the <em>mutation differential selection</em> values by <a class="reference internal" href="dms2_diffsel.html#dms2-diffsel"><span class="std std-ref">dms2_diffsel</span></a> / <a class="reference internal" href="dms2_batch_diffsel.html#dms2-batch-diffsel"><span class="std std-ref">dms2_batch_diffsel</span></a>, and can be plotted in the logo plots above using <a class="reference internal" href="dms2_logoplot.html#dms2-logoplot"><span class="std std-ref">dms2_logoplot</span></a>.
When <span class="math notranslate nohighlight">\(s_{r,x} &gt; 0\)</span>, then <span class="math notranslate nohighlight">\(x\)</span> is more favored in the selection versus mock condition.
When <span class="math notranslate nohighlight">\(s_{r,x} &lt; 0\)</span>, then <span class="math notranslate nohighlight">\(x\)</span> is less favored in the selection versus mock condition.
By definition, <span class="math notranslate nohighlight">\(s_{r,\operatorname{wt}\left(r\right)}\)</span> is always zero.</p>
<p>There is a mutation differential selection value for each mutation at a site – for instance, there will be 19 such values for amino acid characters (the wildtype value is always one by definition).
For visualization and analysis, it is sometimes useful to summarize the mutation differential selection values at a site with a single number.
This can be done in several ways:</p>
<ol class="arabic">
<li><p class="first">We define <em>absolute site differential selection</em> in terms of the <em>mutation differential selection</em> values at that site as</p>
<div class="math notranslate nohighlight" id="equation-abs-diffsel">
<span class="eqno">(22)<a class="headerlink" href="#equation-abs-diffsel" title="Permalink to this equation">¶</a></span>\[\mbox{abs_diffsel} = \sum_x \left|s_{r,x}\right|.\]</div>
<p>The absolute site differential selection is equivalent to the total height of the letter stack in both the positive and negative directions in logoplots such as <a class="reference internal" href="#diffselsnippetfig"><span class="std std-numref">Fig. 5</span></a>.</p>
</li>
<li><p class="first">We define <em>positive site differential selection</em> in terms of the <em>mutation differential selection</em> values at that site as</p>
<div class="math notranslate nohighlight" id="equation-positive-diffsel">
<span class="eqno">(23)<a class="headerlink" href="#equation-positive-diffsel" title="Permalink to this equation">¶</a></span>\[\mbox{positive_diffsel} = \sum_x \max\left(0, s_{r,x}\right).\]</div>
<p>The positive site differential selection is equivalent to the total height of the letter stack in just the positive direction in logoplots such as <a class="reference internal" href="#diffselsnippetfig"><span class="std std-numref">Fig. 5</span></a>.</p>
</li>
<li><p class="first">We define <em>negative site differential selection</em> in terms of the <em>mutation differential selection</em> values at that site as</p>
<div class="math notranslate nohighlight" id="equation-negative-diffsel">
<span class="eqno">(24)<a class="headerlink" href="#equation-negative-diffsel" title="Permalink to this equation">¶</a></span>\[\mbox{negative_diffsel} = \sum_x \min\left(0, s_{r,x}\right).\]</div>
<p>The negative site differential selection is equivalent to the total height of the letter stack in just the negative direction in logoplots such as <a class="reference internal" href="#diffselsnippetfig"><span class="std std-numref">Fig. 5</span></a>.</p>
</li>
<li><p class="first">We define <em>maximum site differential selection</em> in terms of the <em>mutation differential selection</em> values at that site as</p>
<div class="math notranslate nohighlight" id="equation-max-diffsel">
<span class="eqno">(25)<a class="headerlink" href="#equation-max-diffsel" title="Permalink to this equation">¶</a></span>\[\mbox{max_diffsel} = \sum_x \max\left(0, \max_x\left(s_{r,x}\right)\right).\]</div>
<p>The maximum site differential selection is equivalent to the tallest letter in the positive direction in logoplots such as <a class="reference internal" href="#diffselsnippetfig"><span class="std std-numref">Fig. 5</span></a>, or zero if no letters are in the positive direction.</p>
</li>
<li><p class="first">We define <em>minimum site differential selection</em> in terms of the <em>mutation differential selection</em> values at that site as</p>
<div class="math notranslate nohighlight" id="equation-min-diffsel">
<span class="eqno">(26)<a class="headerlink" href="#equation-min-diffsel" title="Permalink to this equation">¶</a></span>\[\mbox{min_diffsel} = \sum_x \min\left(0, \min_x\left(s_{r,x}\right)\right).\]</div>
<p>The maximum site differential selection is equivalent to the tallest letter in the negative direction in logoplots such as <a class="reference internal" href="#diffselsnippetfig"><span class="std std-numref">Fig. 5</span></a>, or zero if no letters are in the negative direction.</p>
</li>
</ol>
</div>
<div class="section" id="error-correction">
<h3><a class="toc-backref" href="#id6">Error correction</a><a class="headerlink" href="#error-correction" title="Permalink to this headline">¶</a></h3>
<p>You can optionally correct for the potential inflation of some counts by sequencing errors by using the <code class="docutils literal notranslate"><span class="pre">--err</span></code> option to <a class="reference internal" href="dms2_diffsel.html#dms2-diffsel"><span class="std std-ref">dms2_diffsel</span></a> / <a class="reference internal" href="dms2_batch_diffsel.html#dms2-batch-diffsel"><span class="std std-ref">dms2_batch_diffsel</span></a>.
To do this, you should use some control where the <strong>only</strong> mutations will come from sequencing / library-prep-induced errors that are presumed to occur at the same rate as the actual samples being analyzed.
Typically, this would be deep sequencing of a wildtype control.</p>
<p>With an error control, the counts are adjusted as follows.
Let <span class="math notranslate nohighlight">\(n_{r,x}\)</span> be the counts for character <span class="math notranslate nohighlight">\(x\)</span> at site <span class="math notranslate nohighlight">\(r\)</span> in the <em>mock</em> or <em>selected</em> sample.
Let <span class="math notranslate nohighlight">\(n^{\rm{err}}_{r,x}\)</span> be the number of counts of <span class="math notranslate nohighlight">\(x\)</span> at site <span class="math notranslate nohighlight">\(r\)</span> in the error control.
Define</p>
<div class="math notranslate nohighlight" id="equation-epsilon">
<span class="eqno">(27)<a class="headerlink" href="#equation-epsilon" title="Permalink to this equation">¶</a></span>\[\epsilon_{r,x} = \left(n^{\rm{err}}_{r,x}\right) / \left(\sum_y n^{\rm{err}}_{r,y}\right).\]</div>
<p>When <span class="math notranslate nohighlight">\(x \ne \operatorname{wt}\left(r\right)\)</span> then <span class="math notranslate nohighlight">\(\epsilon_{r,x}\)</span> is the rate of errors to <span class="math notranslate nohighlight">\(x\)</span> at site <span class="math notranslate nohighlight">\(r\)</span>, and when <span class="math notranslate nohighlight">\(x = \operatorname{wt}\left(r\right)\)</span> then <span class="math notranslate nohighlight">\(\epsilon_{r,x}\)</span> is one minus the rate of errors away from the wildtype at site <span class="math notranslate nohighlight">\(r\)</span>.</p>
<p>We then adjust observed counts <span class="math notranslate nohighlight">\(n_{r,x}\)</span> to the error-corrected counts <span class="math notranslate nohighlight">\(\hat{n}_{r,x}\)</span> by</p>
<div class="math notranslate nohighlight" id="equation-n-rx-adjusted">
<span class="eqno">(28)<a class="headerlink" href="#equation-n-rx-adjusted" title="Permalink to this equation">¶</a></span>\[\begin{split}\hat{n}_{r,x} = \begin{cases}
\max\left[\left(\sum_y n_{r,y}\right) \left(\frac{n_{r,x}}{\sum_y n_{r,y}} - \epsilon_{r,x}\right), 0\right] &amp; \mbox{if } x \ne \operatorname{wt}\left(r\right) \\
n_{r,x} / \epsilon_{r,x} &amp; \mbox{if } x = \operatorname{wt}\left(r\right) \\
\end{cases}\end{split}\]</div>
<p>These adjusted counts are then used in place of the un-adjusted counts in Equation <a class="reference internal" href="#equation-e-rx">(18)</a> above.</p>
<p>Note that if you are using <code class="docutils literal notranslate"><span class="pre">--chartype</span></code> of <code class="docutils literal notranslate"><span class="pre">codon_to_aa</span></code>, the corrections are done on the codon counts, and these corrected codon counts are then aggregated into the amino acid counts before computing the differential selection.</p>
</div>
</div>
<div class="section" id="programs-and-examples">
<h2><a class="toc-backref" href="#id7">Programs and examples</a><a class="headerlink" href="#programs-and-examples" title="Permalink to this headline">¶</a></h2>
<p>The <a class="reference external" href="http://jbloomlab.github.io/dms_tools2">dms_tools2</a> software contains programs to analyze differential selection:</p>
<blockquote>
<div><ul class="simple">
<li><a class="reference internal" href="dms2_diffsel.html#dms2-diffsel"><span class="std std-ref">dms2_diffsel</span></a> estimates differential selection from counts files.</li>
<li><a class="reference internal" href="dms2_batch_diffsel.html#dms2-batch-diffsel"><span class="std std-ref">dms2_batch_diffsel</span></a> runs <a class="reference internal" href="dms2_diffsel.html#dms2-diffsel"><span class="std std-ref">dms2_diffsel</span></a> on a set of samples and summarizes the results. If you have multiple related replicates, it is easier to use <a class="reference internal" href="dms2_batch_diffsel.html#dms2-batch-diffsel"><span class="std std-ref">dms2_batch_diffsel</span></a> rather than runnning <a class="reference internal" href="dms2_diffsel.html#dms2-diffsel"><span class="std std-ref">dms2_diffsel</span></a> directly.</li>
<li><a class="reference internal" href="dms2_logoplot.html#dms2-logoplot"><span class="std std-ref">dms2_logoplot</span></a> can be used to create logo plots that visualize differential selection.</li>
</ul>
</div></blockquote>
<p>Here are some example analyses:</p>
<blockquote>
<div><ul class="simple">
<li><a class="reference external" href="https://github.com/jbloomlab/dms_tools2/blob/master/examples/Doud2017/analysis_notebook.ipynb">Doud2017 example</a></li>
<li><a class="reference external" href="https://github.com/jbloomlab/dms_tools2/blob/master/examples/Dingens2017/analysis_notebook.ipynb">Dingens2017 example</a></li>
<li><a class="reference external" href="https://github.com/jbloomlab/dms_tools2/blob/master/examples/Dingens2018/analysis_notebook.ipynb">Dingens2018 example</a></li>
</ul>
</div></blockquote>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">dms_tools2</h1>
    
  </a>
</p>



<p class="blurb">Tools for analyzing deep mutational scanning data.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=dms_tools2&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a href="https://travis-ci.org/jbloomlab/dms_tools2">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="bcsubamp.html">Barcoded-subamplicon sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="prefs.html">Amino-acid preferences</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Differential selection</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="#definition-of-differential-selection">Definition of differential selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="#programs-and-examples">Programs and examples</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="fracsurvive.html">Fraction surviving</a></li>
<li class="toctree-l1"><a class="reference internal" href="programs.html">Executable programs</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">Python API</a></li>
<li class="toctree-l1"><a class="reference internal" href="citations.html">Citations</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="prefs.html" title="previous chapter">Amino-acid preferences</a></li>
      <li>Next: <a href="fracsurvive.html" title="next chapter">Fraction surviving</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017--2018, the Bloom lab.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.11</a>
      
      |
      <a href="_sources/diffsel.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/dms_tools2" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>