
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>dms2_batch_fracsurvive &#8212; dms_tools2 2.6.1 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="dms2_logoplot" href="dms2_logoplot.html" />
    <link rel="prev" title="dms2_fracsurvive" href="dms2_fracsurvive.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="dms2-batch-fracsurvive">
<span id="id1"></span><h1><code class="docutils literal notranslate"><span class="pre">dms2_batch_fracsurvive</span></code><a class="headerlink" href="#dms2-batch-fracsurvive" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="contents">
<ul class="simple">
<li><p><a class="reference internal" href="#overview" id="id3">Overview</a></p></li>
<li><p><a class="reference internal" href="#command-line-usage" id="id4">Command-line usage</a></p>
<ul>
<li><p><a class="reference internal" href="#Named Arguments" id="id5">Named Arguments</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#output-files" id="id6">Output files</a></p>
<ul>
<li><p><a class="reference internal" href="#results-for-each-sample" id="id7">Results for each sample</a></p></li>
<li><p><a class="reference internal" href="#mean-and-median-fraction-surviving" id="id8">Mean and median fraction surviving</a></p></li>
<li><p><a class="reference internal" href="#correlation-plots" id="id9">Correlation plots</a></p></li>
<li><p><a class="reference internal" href="#fracsurvive-plots" id="id10">Fracsurvive plots</a></p></li>
<li><p><a class="reference internal" href="#log-file" id="id11">Log file</a></p></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="overview">
<h2><a class="toc-backref" href="#id3">Overview</a><a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">dms2_batch_fracsurvive</span></code> program can be used to estimate <a class="reference internal" href="fracsurvive.html#fracsurvive"><span class="std std-ref">Fraction surviving</span></a> for each mutation following a stringent selection such as antbody treatment.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">dms2_batch_fracsurvive</span></code> program runs <a class="reference internal" href="dms2_fracsurvive.html#dms2-fracsurvive"><span class="std std-ref">dms2_fracsurvive</span></a> for each sample listed in a batch file specified by <code class="docutils literal notranslate"><span class="pre">--batchfile</span></code>.
Specifically, as described in <a class="reference internal" href="dms2_batch_diffsel.html#batch-diffsel-commandlineusage"><span class="std std-ref">Command-line usage</span></a>, you can specify a few sample-specific arguments in the <code class="docutils literal notranslate"><span class="pre">--batchfile</span></code>.
All other arguments are specified using the normal option syntax (e.g., <code class="docutils literal notranslate"><span class="pre">--indir</span> <span class="pre">INDIR</span></code>) and are shared between all samples specified in <code class="docutils literal notranslate"><span class="pre">--batchfile</span></code>.
The result is the output for each individual run of <a class="reference internal" href="dms2_fracsurvive.html#dms2-fracsurvive"><span class="std std-ref">dms2_fracsurvive</span></a> plus the summary plots described in <a class="reference internal" href="#output-files">Output files</a>.
It then creates the summary plots described in <a class="reference internal" href="#output-files">Output files</a>.</p>
<p>Because <code class="docutils literal notranslate"><span class="pre">dms2_batch_fracsurvive</span></code> simply runs <a class="reference internal" href="dms2_fracsurvive.html#dms2-fracsurvive"><span class="std std-ref">dms2_fracsurvive</span></a> on each sample specified by the <code class="docutils literal notranslate"><span class="pre">--batchfile</span></code> argument described below, see the <a class="reference internal" href="dms2_fracsurvive.html#dms2-fracsurvive"><span class="std std-ref">dms2_fracsurvive</span></a> <a class="reference internal" href="dms2_fracsurvive.html#fracsurvive-commandlineusage"><span class="std std-ref">Command-line usage</span></a> for details that are helpful for understanding some of the arguments in the <code class="docutils literal notranslate"><span class="pre">dms2_batch_fracsurvive</span></code> <a class="reference internal" href="#batch-fracsurvive-commandlineusage"><span class="std std-ref">Command-line usage</span></a> below.</p>
</div>
<div class="section" id="command-line-usage">
<span id="batch-fracsurvive-commandlineusage"></span><h2><a class="toc-backref" href="#id4">Command-line usage</a><a class="headerlink" href="#command-line-usage" title="Permalink to this headline">¶</a></h2>
<p><p>Perform runs of <code class="docutils literal notranslate"><span class="pre">dms2_fracsurvive</span></code> and summarize results. Part of <a class="reference external" href="http://jbloomlab.github.io/dms_tools2">dms_tools2</a> (version 2.6.1) written by <a class="reference external" href="https://github.com/jbloomlab/dms_tools2/graphs/contributors">the Bloom Lab</a>.</p>
</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">dms2_batch_fracsurvive</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">outdir</span> <span class="n">OUTDIR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">ncpus</span> <span class="n">NCPUS</span><span class="p">]</span>
                              <span class="p">[</span><span class="o">--</span><span class="n">use_existing</span> <span class="p">{</span><span class="n">yes</span><span class="p">,</span><span class="n">no</span><span class="p">}]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">indir</span> <span class="n">INDIR</span><span class="p">]</span>
                              <span class="p">[</span><span class="o">--</span><span class="n">chartype</span> <span class="p">{</span><span class="n">codon_to_aa</span><span class="p">}]</span> <span class="p">[</span><span class="o">--</span><span class="n">aboveavg</span> <span class="p">{</span><span class="n">yes</span><span class="p">,</span><span class="n">no</span><span class="p">}]</span>
                              <span class="p">[</span><span class="o">--</span><span class="n">excludestop</span> <span class="p">{</span><span class="n">yes</span><span class="p">,</span><span class="n">no</span><span class="p">}]</span>
                              <span class="p">[</span><span class="o">--</span><span class="n">pseudocount</span> <span class="n">PSEUDOCOUNT</span><span class="p">]</span>
                              <span class="p">[</span><span class="o">--</span><span class="n">mincount</span> <span class="n">MINCOUNT</span><span class="p">]</span> <span class="o">--</span><span class="n">batchfile</span> <span class="n">BATCHFILE</span>
                              <span class="o">--</span><span class="n">summaryprefix</span> <span class="n">SUMMARYPREFIX</span>
</pre></div>
</div>
<div class="section" id="Named Arguments">
<h3><a class="toc-backref" href="#id5">Named Arguments</a><a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h3>
<dl class="option-list">
<dt><kbd>--outdir</kbd></dt>
<dd><p>Output files to this directory (create if needed).</p>
</dd>
<dt><kbd>--ncpus</kbd></dt>
<dd><p>Number of CPUs to use, -1 is all available.</p>
<p>Default: -1</p>
</dd>
<dt><kbd>--use_existing</kbd></dt>
<dd><p>Possible choices: yes, no</p>
<p>If files with names of expected output already exist, do not re-run.</p>
<p>Default: “no”</p>
</dd>
<dt><kbd>-v, --version</kbd></dt>
<dd><p>show program’s version number and exit</p>
</dd>
<dt><kbd>--indir</kbd></dt>
<dd><p>Input counts files in this directory.</p>
</dd>
<dt><kbd>--chartype</kbd></dt>
<dd><p>Possible choices: codon_to_aa</p>
<p>Characters for which fraction surviving selection is estimated. <cite>codon_to_aa</cite> = amino acids from codon counts.</p>
<p>Default: “codon_to_aa”</p>
</dd>
<dt><kbd>--aboveavg</kbd></dt>
<dd><p>Possible choices: yes, no</p>
<p>Report fracsurvive <strong>above</strong> the library average rather than direct fracsurvive values.</p>
<p>Default: “no”</p>
</dd>
<dt><kbd>--excludestop</kbd></dt>
<dd><p>Possible choices: yes, no</p>
<p>Exclude stop codons as a possible amino acid?</p>
<p>Default: “yes”</p>
</dd>
<dt><kbd>--pseudocount</kbd></dt>
<dd><p>Pseudocount added to each count for sample with smaller depth; pseudocount for other sample scaled by relative depth.</p>
<p>Default: 5</p>
</dd>
<dt><kbd>--mincount</kbd></dt>
<dd><p>Report as <cite>NaN</cite> the fracsurvive of mutations for which both selected and mock-selected samples have &lt; this many counts.</p>
<p>Default: 0</p>
</dd>
<dt><kbd>--batchfile</kbd></dt>
<dd><p>CSV file specifying each <code class="docutils literal notranslate"><span class="pre">dms2_fracsurvive</span></code> run. Must have these columns: <cite>name</cite>, <cite>sel</cite>, <cite>mock</cite>, <cite>libfracsurvive</cite>. Can also have these <cite>err</cite>, <cite>group</cite>, <cite>grouplabel</cite>. If <cite>group</cite> is used, samples are grouped in summary plots labeled by <cite>group</cite>, or by <cite>grouplabel</cite> if it is specified. Other columns are ignored, so other <code class="docutils literal notranslate"><span class="pre">dms2_fracsurvive</span></code> args should be passed as separate command line args rather than in <code class="docutils literal notranslate"><span class="pre">--batchfile</span></code>.</p>
<p>Each of the arguments <cite>name</cite>, <cite>sel</cite>, <cite>mock</cite>, <cite>libfracsurvive</cite>, and optionally <cite>err</cite> gives the value of the same parameter passed to <code class="docutils literal notranslate"><span class="pre">dms2_fracsurvive</span></code>.
If <cite>group</cite> is being used, then the <cite>group</cite> is pre-pended to <cite>name</cite> for that sample.
In addition, <cite>group</cite> is used to organize output for similar runs that should be grouped when calculating means / medians and plotting.</p>
<p>If you are running with no error-control counts, then do <strong>not</strong> specify <code class="docutils literal notranslate"><span class="pre">--err</span></code>.</p>
</dd>
<dt><kbd>--summaryprefix</kbd></dt>
<dd><p>Prefix of output summary files and plots.</p>
<p>As detailed in <cite>Output files</cite> below, <code class="docutils literal notranslate"><span class="pre">dms2_batch_fracsurvive</span></code> creates a variety of plots summarizing the output.
These files are in the directory specified by <code class="docutils literal notranslate"><span class="pre">--outdir</span></code>, and have the prefix specified here.
This name should only contain letters, numbers, dashes, and spaces.
Underscores are <strong>not</strong> allowed as they are a LaTex special character.</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="output-files">
<h2><a class="toc-backref" href="#id6">Output files</a><a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h2>
<p>Running <code class="docutils literal notranslate"><span class="pre">dms2_batch_fracsurvive</span></code> produces output files in the directory specified by <code class="docutils literal notranslate"><span class="pre">--outdir</span></code>.</p>
<div class="section" id="results-for-each-sample">
<h3><a class="toc-backref" href="#id7">Results for each sample</a><a class="headerlink" href="#results-for-each-sample" title="Permalink to this headline">¶</a></h3>
<p>The program <a class="reference internal" href="dms2_fracsurvive.html#dms2-fracsurvive"><span class="std std-ref">dms2_fracsurvive</span></a> is run on each sample specified by <code class="docutils literal notranslate"><span class="pre">--batchfile</span></code>, so you will create all of the <a class="reference internal" href="dms2_fracsurvive.html#dms2-fracsurvive"><span class="std std-ref">dms2_fracsurvive</span></a> <a class="reference internal" href="dms2_fracsurvive.html#fracsurvive-outputfiles"><span class="std std-ref">Output files</span></a>.</p>
<p>If you are using the <cite>group</cite> entry in <code class="docutils literal notranslate"><span class="pre">--batchfile</span></code>, then for each sample we create a name by pre-pending the group to the name.
For instance, if <code class="docutils literal notranslate"><span class="pre">--batchfile</span></code> is:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">group</span><span class="p">,</span><span class="n">name</span><span class="p">,</span><span class="n">sel</span><span class="p">,</span><span class="n">mock</span>
<span class="n">antibody</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="n">replicate</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="n">sel_1_1</span><span class="p">,</span><span class="n">mock_1</span>
<span class="n">antibody</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="n">replicate</span><span class="o">-</span><span class="mi">2</span><span class="p">,</span><span class="n">sel_1_2</span><span class="p">,</span><span class="n">mock_2</span>
<span class="n">antibody</span><span class="o">-</span><span class="mi">2</span><span class="p">,</span><span class="n">replicate</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="n">sel_2_1</span><span class="p">,</span><span class="n">mock_1</span>
<span class="n">antibody</span><span class="o">-</span><span class="mi">2</span><span class="p">,</span><span class="n">replicate</span><span class="o">-</span><span class="mi">2</span><span class="p">,</span><span class="n">sel_2_2</span><span class="p">,</span><span class="n">mock_2</span>
</pre></div>
</div>
<p>then the output files for the individual samples will have prefixes like <code class="docutils literal notranslate"><span class="pre">antibody-1-replicate-1_*</span></code>, <code class="docutils literal notranslate"><span class="pre">antibody-1-replicate-2_*</span></code>, etc.</p>
<p>On the other hand, if <code class="docutils literal notranslate"><span class="pre">--batchfile</span></code> does not specify groups, then the name for each sample is just given by the <cite>name</cite> column.
So if <code class="docutils literal notranslate"><span class="pre">--batchfile</span></code> is:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">name</span><span class="p">,</span><span class="n">sel</span><span class="p">,</span><span class="n">mock</span>
<span class="n">replicate</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="n">sel_1_1</span><span class="p">,</span><span class="n">mock_1</span>
<span class="n">replicate</span><span class="o">-</span><span class="mi">2</span><span class="p">,</span><span class="n">sel_1_2</span><span class="p">,</span><span class="n">mock_2</span>
</pre></div>
</div>
<p>then the output files will have prefixes like <code class="docutils literal notranslate"><span class="pre">replicate-1_*</span></code>, <code class="docutils literal notranslate"><span class="pre">replicate-2_*</span></code>.</p>
</div>
<div class="section" id="mean-and-median-fraction-surviving">
<h3><a class="toc-backref" href="#id8">Mean and median fraction surviving</a><a class="headerlink" href="#mean-and-median-fraction-surviving" title="Permalink to this headline">¶</a></h3>
<p>The program computes the mean and median fraction surviving values for each group (if there are groups), or for all samples.
Note that the means and medians are computed on the mutation fraction surviving values, and then the site values are computed from these mean / median mutation selections.
The files are in the same format as those created by <a class="reference internal" href="dms2_fracsurvive.html#dms2-fracsurvive"><span class="std std-ref">dms2_fracsurvive</span></a>.</p>
<p>For instance, for the first example <code class="docutils literal notranslate"><span class="pre">--batchfile</span></code> in the section above (the one with a <cite>group</cite> column), we would get the following files if we used <code class="docutils literal notranslate"><span class="pre">--summaryprefix</span> <span class="pre">summary</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">summary_antibody</span><span class="o">-</span><span class="mi">1</span><span class="o">-</span><span class="n">meanmutfracsurvive</span><span class="o">.</span><span class="n">csv</span>
<span class="n">summary_antibody</span><span class="o">-</span><span class="mi">2</span><span class="o">-</span><span class="n">meanmutfracsurvive</span><span class="o">.</span><span class="n">csv</span>
<span class="n">summary_antibody</span><span class="o">-</span><span class="mi">1</span><span class="o">-</span><span class="n">medianmutfracsurvive</span><span class="o">.</span><span class="n">csv</span>
<span class="n">summary_antibody</span><span class="o">-</span><span class="mi">2</span><span class="o">-</span><span class="n">medianmutfracsurvive</span><span class="o">.</span><span class="n">csv</span>
<span class="n">summary_antibody</span><span class="o">-</span><span class="mi">1</span><span class="o">-</span><span class="n">meansitefracsurvive</span><span class="o">.</span><span class="n">csv</span>
<span class="n">summary_antibody</span><span class="o">-</span><span class="mi">2</span><span class="o">-</span><span class="n">meansitefracsurvive</span><span class="o">.</span><span class="n">csv</span>
<span class="n">summary_antibody</span><span class="o">-</span><span class="mi">1</span><span class="o">-</span><span class="n">mediansitefracsurvive</span><span class="o">.</span><span class="n">csv</span>
<span class="n">summary_antibody</span><span class="o">-</span><span class="mi">2</span><span class="o">-</span><span class="n">mediansitefracsurvive</span><span class="o">.</span><span class="n">csv</span>
</pre></div>
</div>
<p>For the second example <code class="docutils literal notranslate"><span class="pre">--batchfile</span></code> (the one without a <cite>group</cite> column), we would get the following files:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">summary_meanmutfracsurvive</span><span class="o">.</span><span class="n">csv</span>
<span class="n">summary_medianmutfracsurvive</span><span class="o">.</span><span class="n">csv</span>
<span class="n">summary_meansitefracsurvive</span><span class="o">.</span><span class="n">csv</span>
<span class="n">summary_mediansitefracsurvive</span><span class="o">.</span><span class="n">csv</span>
</pre></div>
</div>
<p>It is often useful to visualize the mean or median <cite>mutfracsurvive</cite> files with <a class="reference internal" href="dms2_logoplot.html#dms2-logoplot"><span class="std std-ref">dms2_logoplot</span></a>.</p>
</div>
<div class="section" id="correlation-plots">
<h3><a class="toc-backref" href="#id9">Correlation plots</a><a class="headerlink" href="#correlation-plots" title="Permalink to this headline">¶</a></h3>
<p>Scatter plots are created that show the correlations among samples within the same <cite>group</cite>, or among all samples if there are not any groups.</p>
<p>Separate plots are made for the <cite>mutfracsurvive</cite>, the <cite>avgfracsurvive</cite> (averaged across all mutations at each site), and the maximum <cite>mutfracsurvive</cite> at each site.
The names will have the form:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">summary_antibody</span><span class="o">-</span><span class="mi">1</span><span class="o">-</span><span class="n">mutfracsurvivecorr</span><span class="o">.</span><span class="n">pdf</span>
<span class="n">summary_antibody</span><span class="o">-</span><span class="mi">1</span><span class="o">-</span><span class="n">avgfracsurvivecorr</span><span class="o">.</span><span class="n">pdf</span>
<span class="n">summary_antibody</span><span class="o">-</span><span class="mi">1</span><span class="o">-</span><span class="n">maxfracsurvivecorr</span><span class="o">.</span><span class="n">pdf</span>
</pre></div>
</div>
</div>
<div class="section" id="fracsurvive-plots">
<h3><a class="toc-backref" href="#id10">Fracsurvive plots</a><a class="headerlink" href="#fracsurvive-plots" title="Permalink to this headline">¶</a></h3>
<p>Plots are made that show the site average and maximum fraction surviving as a function of the primary sequence.
These plots show the mean and median values for each group, and are faceted by <cite>group</cite> (if there are groups).
If you run with <code class="docutils literal notranslate"><span class="pre">--summaryprefix</span> <span class="pre">summary</span></code>, then the plots will be:</p>
<blockquote>
<div><ul class="simple">
<li><p><cite>avgfracsurvive</cite>: files <code class="docutils literal notranslate"><span class="pre">summary_meanavgfracsurvive.pdf</span></code> and <code class="docutils literal notranslate"><span class="pre">summary_medianavgfracsurvive.pdf</span></code> show the average <cite>mutfracsurvive</cite> across all mutations for each site.</p></li>
<li><p><cite>maxfracsurvive</cite>: files <code class="docutils literal notranslate"><span class="pre">summary_meanmaxfracsurvive.pdf</span></code> and <code class="docutils literal notranslate"><span class="pre">summary_medianmaxfracsurvive.pdf</span></code> show the maximum <cite>mutfracsurvive</cite> across all mutations for each site.</p></li>
</ul>
</div></blockquote>
</div>
<div class="section" id="log-file">
<h3><a class="toc-backref" href="#id11">Log file</a><a class="headerlink" href="#log-file" title="Permalink to this headline">¶</a></h3>
<p>A log file is created that summarizes the output.
For instance, if you run <code class="docutils literal notranslate"><span class="pre">dms2_batch_fracsurvive</span></code> with the arguments <code class="docutils literal notranslate"><span class="pre">--outdir</span> <span class="pre">results</span> <span class="pre">--summaryprefix</span> <span class="pre">summary</span></code> then the log will be <code class="docutils literal notranslate"><span class="pre">./results/summary.log</span></code>.</p>
</div>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">dms_tools2</h1>
    
  </a>
</p>



<p class="blurb">Tools for analyzing deep mutational scanning data.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=dms_tools2&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a class="badge" href="https://travis-ci.org/jbloomlab/dms_tools2">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="bcsubamp.html">Barcoded-subamplicon sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="prefs.html">Amino-acid preferences</a></li>
<li class="toctree-l1"><a class="reference internal" href="diffsel.html">Differential selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="fracsurvive.html">Fraction surviving</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="programs.html">Executable programs</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="dms2_bcsubamp.html"><code class="docutils literal notranslate"><span class="pre">dms2_bcsubamp</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="dms2_batch_bcsubamp.html"><code class="docutils literal notranslate"><span class="pre">dms2_batch_bcsubamp</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="dms2_prefs.html"><code class="docutils literal notranslate"><span class="pre">dms2_prefs</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="dms2_batch_prefs.html"><code class="docutils literal notranslate"><span class="pre">dms2_batch_prefs</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="dms2_diffsel.html"><code class="docutils literal notranslate"><span class="pre">dms2_diffsel</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="dms2_batch_diffsel.html"><code class="docutils literal notranslate"><span class="pre">dms2_batch_diffsel</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="dms2_fracsurvive.html"><code class="docutils literal notranslate"><span class="pre">dms2_fracsurvive</span></code></a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#"><code class="docutils literal notranslate"><span class="pre">dms2_batch_fracsurvive</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="dms2_logoplot.html"><code class="docutils literal notranslate"><span class="pre">dms2_logoplot</span></code></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="api.html">Python API</a></li>
<li class="toctree-l1"><a class="reference internal" href="citations.html">Citations</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="programs.html">Executable programs</a><ul>
      <li>Previous: <a href="dms2_fracsurvive.html" title="previous chapter"><code class="docutils literal notranslate"><span class="pre">dms2_fracsurvive</span></code></a></li>
      <li>Next: <a href="dms2_logoplot.html" title="next chapter"><code class="docutils literal notranslate"><span class="pre">dms2_logoplot</span></code></a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017--2019, the Bloom lab.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/dms2_batch_fracsurvive.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/dms_tools2" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>