
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>dms2_bcsubamp &#8212; dms_tools2 2.6.2 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="dms2_batch_bcsubamp" href="dms2_batch_bcsubamp.html" />
    <link rel="prev" title="Executable programs" href="programs.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="dms2-bcsubamp">
<span id="id1"></span><h1><code class="docutils literal notranslate"><span class="pre">dms2_bcsubamp</span></code><a class="headerlink" href="#dms2-bcsubamp" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="contents">
<ul class="simple">
<li><p><a class="reference internal" href="#overview" id="id2">Overview</a></p></li>
<li><p><a class="reference internal" href="#algorithm-for-assembling-and-aligning-subamplicons" id="id3">Algorithm for assembling and aligning subamplicons</a></p></li>
<li><p><a class="reference internal" href="#command-line-usage" id="id4">Command-line usage</a></p></li>
<li><p><a class="reference internal" href="#output-files" id="id5">Output files</a></p>
<ul>
<li><p><a class="reference internal" href="#log-file" id="id6">Log file</a></p></li>
<li><p><a class="reference internal" href="#read-statistics-file" id="id7">Read statistics file</a></p></li>
<li><p><a class="reference internal" href="#reads-per-barcode-file" id="id8">Reads-per-barcode file</a></p></li>
<li><p><a class="reference internal" href="#barcode-statistics-file" id="id9">Barcode statistics file</a></p></li>
<li><p><a class="reference internal" href="#counts-file" id="id10">Counts file</a></p></li>
<li><p><a class="reference internal" href="#detailed-barcode-information-file" id="id11">Detailed barcode information file</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#memory-usage" id="id12">Memory usage</a></p></li>
</ul>
</div>
<div class="section" id="overview">
<h2><a class="toc-backref" href="#id2">Overview</a><a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">dms2_bcsubamp</span></code> program processes the FASTQ files generated by <a class="reference internal" href="bcsubamp.html#bcsubamp"><span class="std std-ref">Barcoded-subamplicon sequencing</span></a> to count the frequencies of each identity at each site and generate summary statistics.</p>
<p>In general, you will probably have an easier time if you use the <a class="reference internal" href="dms2_batch_bcsubamp.html#dms2-batch-bcsubamp"><span class="std std-ref">dms2_batch_bcsubamp</span></a> program rather than running <code class="docutils literal notranslate"><span class="pre">dms2_bcsubamp</span></code> directly.</p>
</div>
<div class="section" id="algorithm-for-assembling-and-aligning-subamplicons">
<span id="bcsubamp-algorithm"></span><h2><a class="toc-backref" href="#id3">Algorithm for assembling and aligning subamplicons</a><a class="headerlink" href="#algorithm-for-assembling-and-aligning-subamplicons" title="Permalink to this headline">¶</a></h2>
<p>The algorithm implemented by <code class="docutils literal notranslate"><span class="pre">dms2_bcsubamp</span></code> is as follows:</p>
<ol class="arabic">
<li><p>Read pairs are discarded if either read fails the Illumina chastity filter.</p></li>
<li><p>All sites in each read are classified as “high quality” if the site is not an <code class="docutils literal notranslate"><span class="pre">N</span></code> (ambiguous nucleotide) and has a Q-score of at least <code class="docutils literal notranslate"><span class="pre">--minq</span></code>, or as “low quality” otherwise. At low-quality sites, the identity in the read is then set to <code class="docutils literal notranslate"><span class="pre">N</span></code> for the remainder of the analysis.</p></li>
<li><p>We discard any read pairs for which there are any <code class="docutils literal notranslate"><span class="pre">N</span></code> nucleotides in the barcode.</p></li>
<li><p>We then collect all remaining read pairs for each barcode (concatenating the barcodes on the paired reads for a total barcode length that is twice <code class="docutils literal notranslate"><span class="pre">--bclen</span></code>) and perform all subsequent operations on these barcodes. Note that if you have different length barcodes on the two reads (or all the barcode on one read), you can use the <code class="docutils literal notranslate"><span class="pre">--bclen2</span></code> option below.</p></li>
<li><p>Any barcode with less than <code class="docutils literal notranslate"><span class="pre">--minreads</span></code> read pairs is discarded.</p></li>
<li><p>The R1 and R2 reads for each barcode are each assembled into consensus sequences. The consensus sequences is built by applying the following algorithm to each site:</p>
<blockquote>
<div><ol class="loweralpha simple">
<li><p>We collect the identities for all reads, ignoring <code class="docutils literal notranslate"><span class="pre">N</span></code> nucleotides.</p></li>
<li><p>If there are less than <code class="docutils literal notranslate"><span class="pre">--minreads</span></code> called (non <code class="docutils literal notranslate"><span class="pre">N</span></code>) nucleotides, set the consensus identity for that site to <code class="docutils literal notranslate"><span class="pre">N</span></code>.</p></li>
<li><p>If there are at least <code class="docutils literal notranslate"><span class="pre">--minreads</span></code> called nucleotides, set the consensus identity to <code class="docutils literal notranslate"><span class="pre">N</span></code> if the fraction of called identities that are identical is less than <code class="docutils literal notranslate"><span class="pre">--minconcur</span></code>, and set it to the consensus called identity if at least <code class="docutils literal notranslate"><span class="pre">--minconcur</span></code> of the reads are identical at the site.</p></li>
</ol>
</div></blockquote>
</li>
<li><p>We attempt to align the consensus read sequences for each barcode to subamplicon positions specified by <code class="docutils literal notranslate"><span class="pre">--alignspecs</span></code>. The attempted alignment does <strong>not</strong> accommodate gaps; the consensus sequence must align gaplessly at the position specified by <code class="docutils literal notranslate"><span class="pre">--alignspecs</span></code>. In any region where the reads overlap (which may happen near the center of the subamplicon), if both consensus reads report high quality nucleotides, the identity is considered ambiguous if the sequences disagree. If one consensus read is ambiguous, we take the identity for the non-ambiguous read. The alignment procedure is as follows:</p>
<blockquote>
<div><ol class="loweralpha simple">
<li><p>We trim the reads from the 3’ end using the <code class="docutils literal notranslate"><span class="pre">--R1trim</span></code> and <code class="docutils literal notranslate"><span class="pre">--R2trim</span></code> parameters for that subamplicon.</p></li>
<li><p>We then attempt to align the consensus reads at each position indicated in <code class="docutils literal notranslate"><span class="pre">--alignspecs</span></code>. We consider an alignment acceptable if there are no more than <code class="docutils literal notranslate"><span class="pre">--maxmuts</span></code> mutations relative to <code class="docutils literal notranslate"><span class="pre">refseq</span></code> (with mutations counted in terms of character <code class="docutils literal notranslate"><span class="pre">--chartype</span></code>) and if the fraction of called nucleotides in the aligned consensus is at least <code class="docutils literal notranslate"><span class="pre">--minfraccall</span></code>. We take the first valid alignment that we find. But unless you are working with a very repetitive sequence or specify a very large <code class="docutils literal notranslate"><span class="pre">--maxmuts</span></code>, there should be at most one valid alignment.</p></li>
</ol>
</div></blockquote>
</li>
<li><p>For each aligned subamplicon, we count the identities at each site. We call an identity if at least <code class="docutils literal notranslate"><span class="pre">--minreads</span></code> reads with non-ambiguous identities span that site, and at least <code class="docutils literal notranslate"><span class="pre">--minconcur</span></code> of these agree on the identity. If both reads span a site and they disagree, we call that site as <code class="docutils literal notranslate"><span class="pre">N</span></code>.</p></li>
<li><p>The results are reported in the <a class="reference internal" href="#output-files">Output files</a>.</p></li>
</ol>
</div>
<div class="section" id="command-line-usage">
<span id="bcsubamp-commandlineusage"></span><h2><a class="toc-backref" href="#id4">Command-line usage</a><a class="headerlink" href="#command-line-usage" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="output-files">
<span id="bcsubamp-outputfiles"></span><h2><a class="toc-backref" href="#id5">Output files</a><a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h2>
<p>The program produces a variety of output files.
These files all have the prefix specified by <code class="docutils literal notranslate"><span class="pre">--outdir</span></code> and <code class="docutils literal notranslate"><span class="pre">--name</span></code>.
For instance, if you use <code class="docutils literal notranslate"><span class="pre">--outdir</span> <span class="pre">results</span> <span class="pre">--name</span> <span class="pre">sample-1</span></code>, then the output files will have the prefix <code class="docutils literal notranslate"><span class="pre">./results/sample-1</span></code> and the suffixes described below.</p>
<p>Here are the specific output files:</p>
<div class="section" id="log-file">
<h3><a class="toc-backref" href="#id6">Log file</a><a class="headerlink" href="#log-file" title="Permalink to this headline">¶</a></h3>
<p>This file has the suffix <code class="docutils literal notranslate"><span class="pre">.log</span></code>.
It is a text file that logs the progress of the program.</p>
</div>
<div class="section" id="read-statistics-file">
<h3><a class="toc-backref" href="#id7">Read statistics file</a><a class="headerlink" href="#read-statistics-file" title="Permalink to this headline">¶</a></h3>
<p>This file has the suffix <code class="docutils literal notranslate"><span class="pre">_readstats.csv</span></code>.
It gives the statistics on all the processed reads.
For instance:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">fail</span> <span class="nb">filter</span><span class="p">,</span><span class="n">low</span> <span class="n">Q</span> <span class="n">barcode</span><span class="p">,</span><span class="n">total</span>
<span class="mi">0</span><span class="p">,</span><span class="mi">785663</span><span class="p">,</span><span class="mi">12691176</span>
</pre></div>
</div>
</div>
<div class="section" id="reads-per-barcode-file">
<h3><a class="toc-backref" href="#id8">Reads-per-barcode file</a><a class="headerlink" href="#reads-per-barcode-file" title="Permalink to this headline">¶</a></h3>
<p>This file has the suffix <code class="docutils literal notranslate"><span class="pre">_readsperbc.csv</span></code>.
It gives the number of reads associated with each barcode.
For instance:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">number</span> <span class="n">of</span> <span class="n">reads</span><span class="p">,</span><span class="n">number</span> <span class="n">of</span> <span class="n">barcodes</span>
<span class="mi">1</span><span class="p">,</span><span class="mi">1248294</span>
<span class="mi">2</span><span class="p">,</span><span class="mi">400579</span>
<span class="mi">3</span><span class="p">,</span><span class="mi">471893</span>
<span class="mi">4</span><span class="p">,</span><span class="mi">456661</span>
<span class="mi">5</span><span class="p">,</span><span class="mi">376158</span>
<span class="mi">6</span><span class="p">,</span><span class="mi">272020</span>
<span class="mi">7</span><span class="p">,</span><span class="mi">174295</span>
<span class="mi">8</span><span class="p">,</span><span class="mi">102660</span>
<span class="mi">9</span><span class="p">,</span><span class="mi">56302</span>
<span class="mi">10</span><span class="p">,</span><span class="mi">28079</span>
<span class="mi">11</span><span class="p">,</span><span class="mi">13125</span>
<span class="mi">12</span><span class="p">,</span><span class="mi">5747</span>
<span class="mi">13</span><span class="p">,</span><span class="mi">2518</span>
<span class="mi">14</span><span class="p">,</span><span class="mi">1035</span>
<span class="mi">15</span><span class="p">,</span><span class="mi">403</span>
<span class="mi">16</span><span class="p">,</span><span class="mi">177</span>
<span class="mi">17</span><span class="p">,</span><span class="mi">78</span>
<span class="mi">18</span><span class="p">,</span><span class="mi">34</span>
<span class="mi">19</span><span class="p">,</span><span class="mi">12</span>
<span class="mi">20</span><span class="p">,</span><span class="mi">9</span>
<span class="mi">21</span><span class="p">,</span><span class="mi">1</span>
<span class="mi">22</span><span class="p">,</span><span class="mi">2</span>
<span class="mi">24</span><span class="p">,</span><span class="mi">1</span>
<span class="mi">25</span><span class="p">,</span><span class="mi">1</span>
<span class="mi">75</span><span class="p">,</span><span class="mi">1</span>
</pre></div>
</div>
</div>
<div class="section" id="barcode-statistics-file">
<h3><a class="toc-backref" href="#id9">Barcode statistics file</a><a class="headerlink" href="#barcode-statistics-file" title="Permalink to this headline">¶</a></h3>
<p>This file has the suffix <code class="docutils literal notranslate"><span class="pre">_bcstats.csv</span></code>.
It provides information on how many of the barcodes could be successfully aligned.
For instance:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aligned</span><span class="p">,</span><span class="ow">not</span> <span class="n">alignable</span><span class="p">,</span><span class="n">too</span> <span class="n">few</span> <span class="n">reads</span><span class="p">,</span><span class="n">total</span>
<span class="mi">2129522</span><span class="p">,</span><span class="mi">232269</span><span class="p">,</span><span class="mi">1248294</span><span class="p">,</span><span class="mi">3610085</span>
</pre></div>
</div>
</div>
<div class="section" id="counts-file">
<h3><a class="toc-backref" href="#id10">Counts file</a><a class="headerlink" href="#counts-file" title="Permalink to this headline">¶</a></h3>
<p>This is output file that has the results that you will probably use for subsequent analyses.
It has the suffix <code class="docutils literal notranslate"><span class="pre">_codoncounts.csv</span></code> if you are using <code class="docutils literal notranslate"><span class="pre">--chartype</span> <span class="pre">codon</span></code>.
It gives the number of called identities at each site in the sequence, as well as the wildtype sequence.
For instance, here are the first few lines:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">site</span><span class="p">,</span><span class="n">wildtype</span><span class="p">,</span><span class="n">AAA</span><span class="p">,</span><span class="n">AAC</span><span class="p">,</span><span class="n">AAG</span><span class="p">,</span><span class="n">AAT</span><span class="p">,</span><span class="n">ACA</span><span class="p">,</span><span class="n">ACC</span><span class="p">,</span><span class="n">ACG</span><span class="p">,</span><span class="n">ACT</span><span class="p">,</span><span class="n">AGA</span><span class="p">,</span><span class="n">AGC</span><span class="p">,</span><span class="n">AGG</span><span class="p">,</span><span class="n">AGT</span><span class="p">,</span><span class="n">ATA</span><span class="p">,</span><span class="n">ATC</span><span class="p">,</span><span class="n">ATG</span><span class="p">,</span><span class="n">ATT</span><span class="p">,</span><span class="n">CAA</span><span class="p">,</span><span class="n">CAC</span><span class="p">,</span><span class="n">CAG</span><span class="p">,</span><span class="n">CAT</span><span class="p">,</span><span class="n">CCA</span><span class="p">,</span><span class="n">CCC</span><span class="p">,</span><span class="n">CCG</span><span class="p">,</span><span class="n">CCT</span><span class="p">,</span><span class="n">CGA</span><span class="p">,</span><span class="n">CGC</span><span class="p">,</span><span class="n">CGG</span><span class="p">,</span><span class="n">CGT</span><span class="p">,</span><span class="n">CTA</span><span class="p">,</span><span class="n">CTC</span><span class="p">,</span><span class="n">CTG</span><span class="p">,</span><span class="n">CTT</span><span class="p">,</span><span class="n">GAA</span><span class="p">,</span><span class="n">GAC</span><span class="p">,</span><span class="n">GAG</span><span class="p">,</span><span class="n">GAT</span><span class="p">,</span><span class="n">GCA</span><span class="p">,</span><span class="n">GCC</span><span class="p">,</span><span class="n">GCG</span><span class="p">,</span><span class="n">GCT</span><span class="p">,</span><span class="n">GGA</span><span class="p">,</span><span class="n">GGC</span><span class="p">,</span><span class="n">GGG</span><span class="p">,</span><span class="n">GGT</span><span class="p">,</span><span class="n">GTA</span><span class="p">,</span><span class="n">GTC</span><span class="p">,</span><span class="n">GTG</span><span class="p">,</span><span class="n">GTT</span><span class="p">,</span><span class="n">TAA</span><span class="p">,</span><span class="n">TAC</span><span class="p">,</span><span class="n">TAG</span><span class="p">,</span><span class="n">TAT</span><span class="p">,</span><span class="n">TCA</span><span class="p">,</span><span class="n">TCC</span><span class="p">,</span><span class="n">TCG</span><span class="p">,</span><span class="n">TCT</span><span class="p">,</span><span class="n">TGA</span><span class="p">,</span><span class="n">TGC</span><span class="p">,</span><span class="n">TGG</span><span class="p">,</span><span class="n">TGT</span><span class="p">,</span><span class="n">TTA</span><span class="p">,</span><span class="n">TTC</span><span class="p">,</span><span class="n">TTG</span><span class="p">,</span><span class="n">TTT</span>
<span class="mi">1</span><span class="p">,</span><span class="n">ATG</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">333985</span><span class="p">,</span><span class="mi">14</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span>
<span class="mi">2</span><span class="p">,</span><span class="n">AAG</span><span class="p">,</span><span class="mi">16</span><span class="p">,</span><span class="mi">20</span><span class="p">,</span><span class="mi">333132</span><span class="p">,</span><span class="mi">41</span><span class="p">,</span><span class="mi">13</span><span class="p">,</span><span class="mi">12</span><span class="p">,</span><span class="mi">27</span><span class="p">,</span><span class="mi">14</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">67</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">9</span><span class="p">,</span><span class="mi">13</span><span class="p">,</span><span class="mi">29</span><span class="p">,</span><span class="mi">9</span><span class="p">,</span><span class="mi">10</span><span class="p">,</span><span class="mi">11</span><span class="p">,</span><span class="mi">12</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">10</span><span class="p">,</span><span class="mi">15</span><span class="p">,</span><span class="mi">15</span><span class="p">,</span><span class="mi">11</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">9</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">10</span><span class="p">,</span><span class="mi">17</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">49</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">9</span><span class="p">,</span><span class="mi">14</span><span class="p">,</span><span class="mi">9</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">10</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">9</span><span class="p">,</span><span class="mi">11</span><span class="p">,</span><span class="mi">11</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">14</span><span class="p">,</span><span class="mi">14</span><span class="p">,</span><span class="mi">11</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">13</span><span class="p">,</span><span class="mi">16</span><span class="p">,</span><span class="mi">15</span><span class="p">,</span><span class="mi">14</span><span class="p">,</span><span class="mi">9</span><span class="p">,</span><span class="mi">9</span><span class="p">,</span><span class="mi">15</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">9</span><span class="p">,</span><span class="mi">11</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">15</span>
<span class="mi">3</span><span class="p">,</span><span class="n">GCA</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">34</span><span class="p">,</span><span class="mi">11</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">9</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">14</span><span class="p">,</span><span class="mi">10</span><span class="p">,</span><span class="mi">12</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">10</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">11</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">12</span><span class="p">,</span><span class="mi">19</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">11</span><span class="p">,</span><span class="mi">9</span><span class="p">,</span><span class="mi">333250</span><span class="p">,</span><span class="mi">10</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">9</span><span class="p">,</span><span class="mi">15</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">37</span><span class="p">,</span><span class="mi">9</span><span class="p">,</span><span class="mi">9</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">23</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">11</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">10</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">9</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">6</span>
</pre></div>
</div>
</div>
<div class="section" id="detailed-barcode-information-file">
<h3><a class="toc-backref" href="#id11">Detailed barcode information file</a><a class="headerlink" href="#detailed-barcode-information-file" title="Permalink to this headline">¶</a></h3>
<p>This file has the suffix <code class="docutils literal notranslate"><span class="pre">_bcinfo.txt.gz</span></code>.
It is a very large gzipped text file that contains information on all the reads and barcodes.
The format should be self explanatory.
This file is only created if you use the <code class="docutils literal notranslate"><span class="pre">--bcinfo</span></code> option, and may be helpful for debugging if your reads aren’t aligning as expected.</p>
</div>
</div>
<div class="section" id="memory-usage">
<span id="bcsubamp-memoryusage"></span><h2><a class="toc-backref" href="#id12">Memory usage</a><a class="headerlink" href="#memory-usage" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">dms2_bcsubamp</span></code> stores all of the reads in the FASTQ files in memory. Therefore, it uses a substantial amount of memory, typically around a gigabyte per million paired-end sequencing reads. However, for typical data sets such memory usage is well within the capacity of modern large-memory nodes.</p>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">dms_tools2</h1>
    
  </a>
</p>



<p class="blurb">Tools for analyzing deep mutational scanning data.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=dms_tools2&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a class="badge" href="https://travis-ci.org/jbloomlab/dms_tools2">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="bcsubamp.html">Barcoded-subamplicon sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="prefs.html">Amino-acid preferences</a></li>
<li class="toctree-l1"><a class="reference internal" href="diffsel.html">Differential selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="fracsurvive.html">Fraction surviving</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="programs.html">Executable programs</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#"><code class="docutils literal notranslate"><span class="pre">dms2_bcsubamp</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="dms2_batch_bcsubamp.html"><code class="docutils literal notranslate"><span class="pre">dms2_batch_bcsubamp</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="dms2_prefs.html"><code class="docutils literal notranslate"><span class="pre">dms2_prefs</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="dms2_batch_prefs.html"><code class="docutils literal notranslate"><span class="pre">dms2_batch_prefs</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="dms2_diffsel.html"><code class="docutils literal notranslate"><span class="pre">dms2_diffsel</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="dms2_batch_diffsel.html"><code class="docutils literal notranslate"><span class="pre">dms2_batch_diffsel</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="dms2_fracsurvive.html"><code class="docutils literal notranslate"><span class="pre">dms2_fracsurvive</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="dms2_batch_fracsurvive.html"><code class="docutils literal notranslate"><span class="pre">dms2_batch_fracsurvive</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="dms2_logoplot.html"><code class="docutils literal notranslate"><span class="pre">dms2_logoplot</span></code></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="api.html">Python API</a></li>
<li class="toctree-l1"><a class="reference internal" href="citations.html">Citations</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="programs.html">Executable programs</a><ul>
      <li>Previous: <a href="programs.html" title="previous chapter">Executable programs</a></li>
      <li>Next: <a href="dms2_batch_bcsubamp.html" title="next chapter"><code class="docutils literal notranslate"><span class="pre">dms2_batch_bcsubamp</span></code></a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017--2019, the Bloom lab.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/dms2_bcsubamp.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/dms_tools2" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>