
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>dms2_batch_bcsubamp &#8212; dms_tools2 2.2.5 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '2.2.5',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="dms2_prefs" href="dms2_prefs.html" />
    <link rel="prev" title="dms2_bcsubamp" href="dms2_bcsubamp.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="dms2-batch-bcsubamp">
<span id="id1"></span><h1><code class="docutils literal"><span class="pre">dms2_batch_bcsubamp</span></code><a class="headerlink" href="#dms2-batch-bcsubamp" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="contents">
<ul class="simple">
<li><a class="reference internal" href="#overview" id="id3">Overview</a></li>
<li><a class="reference internal" href="#command-line-usage" id="id4">Command-line usage</a><ul>
<li><a class="reference internal" href="#Named Arguments" id="id5">Named Arguments</a></li>
</ul>
</li>
<li><a class="reference internal" href="#output-files" id="id6">Output files</a><ul>
<li><a class="reference internal" href="#results-for-each-sample" id="id7">Results for each sample</a></li>
<li><a class="reference internal" href="#summary-files" id="id8">Summary files</a></li>
</ul>
</li>
<li><a class="reference internal" href="#memory-usage" id="id9">Memory usage</a></li>
</ul>
</div>
<div class="section" id="overview">
<h2><a class="toc-backref" href="#id3">Overview</a><a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal"><span class="pre">dms2_batch_bcsubamp</span></code> program processes FASTQ files generated by <a class="reference internal" href="bcsubamp.html#bcsubamp"><span class="std std-ref">Barcoded-subamplicon sequencing</span></a> to count the frequencies of mutations at each site for a set of samples, and then summarize the results.</p>
<p>The <code class="docutils literal"><span class="pre">dms2_batch_bcsubamp</span></code> program simply runs <a class="reference internal" href="dms2_bcsubamp.html#dms2-bcsubamp"><span class="std std-ref">dms2_bcsubamp</span></a> for each sample listed in a batch file specified by <code class="docutils literal"><span class="pre">--batchfile</span></code>.
Specifically, as described in <a class="reference internal" href="#batch-bcsubamp-commandlineusage"><span class="std std-ref">Command-line usage</span></a>, you can specify a few sample-specific arguments in the <code class="docutils literal"><span class="pre">--batchfile</span></code>.
All other arguments are specified using the normal option syntax (e.g., <code class="docutils literal"><span class="pre">--bclen</span> <span class="pre">BCLEN</span></code>) and are shared between all samples specified in <code class="docutils literal"><span class="pre">--batchfile</span></code>.
The result is the output for each individual run of <a class="reference internal" href="dms2_bcsubamp.html#dms2-bcsubamp"><span class="std std-ref">dms2_bcsubamp</span></a> plus the summary plots described in <a class="reference internal" href="#output-files">Output files</a>.</p>
<p>The <a class="reference external" href="https://github.com/jbloomlab/dms_tools2/blob/master/examples/Doud2016/analysis_notebook.ipynb">Doud2016 example</a> to illustrates the usage of <code class="docutils literal"><span class="pre">dms2_batch_bcsubamp</span></code> on a real dataset.</p>
<p>Because <code class="docutils literal"><span class="pre">dms2_batch_bcsubamp</span></code> simply runs <code class="docutils literal"><span class="pre">dms2_bcsubamp</span></code> on each sample specfied by the <code class="docutils literal"><span class="pre">--batchfile</span></code> argument described below, see the <code class="docutils literal"><span class="pre">dms2_bcsubamp</span></code> <a class="reference internal" href="dms2_bcsubamp.html#bcsubamp-algorithm"><span class="std std-ref">Algorithm for assembling and aligning subamplicons</span></a> and the <code class="docutils literal"><span class="pre">dms2_bcsubamp</span></code> <a class="reference internal" href="dms2_bcsubamp.html#bcsubamp-commandlineusage"><span class="std std-ref">Command-line usage</span></a> for details that are helpful for understanding many of the arguments in the <code class="docutils literal"><span class="pre">dms2_batch_bcsubamp</span></code> <a class="reference internal" href="#batch-bcsubamp-commandlineusage"><span class="std std-ref">Command-line usage</span></a> below.</p>
</div>
<div class="section" id="command-line-usage">
<span id="batch-bcsubamp-commandlineusage"></span><h2><a class="toc-backref" href="#id4">Command-line usage</a><a class="headerlink" href="#command-line-usage" title="Permalink to this headline">¶</a></h2>
<p>Perform many runs of <code class="docutils literal"><span class="pre">dms2_bcsubamp</span></code> and plot results. Part of <a class="reference external" href="http://jbloomlab.github.io/dms_tools2">dms_tools2</a> (version 2.2.5) written by <a class="reference external" href="https://github.com/jbloomlab/dms_tools2/graphs/contributors">the Bloom Lab</a>.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">dms2_batch_bcsubamp</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">outdir</span> <span class="n">OUTDIR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">ncpus</span> <span class="n">NCPUS</span><span class="p">]</span>
                           <span class="p">[</span><span class="o">--</span><span class="n">use_existing</span> <span class="p">{</span><span class="n">yes</span><span class="p">,</span><span class="n">no</span><span class="p">}]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span> <span class="o">--</span><span class="n">refseq</span> <span class="n">REFSEQ</span>
                           <span class="o">--</span><span class="n">alignspecs</span> <span class="n">ALIGNSPECS</span> <span class="p">[</span><span class="n">ALIGNSPECS</span> <span class="o">...</span><span class="p">]</span>
                           <span class="p">[</span><span class="o">--</span><span class="n">bclen</span> <span class="n">BCLEN</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">fastqdir</span> <span class="n">FASTQDIR</span><span class="p">]</span>
                           <span class="p">[</span><span class="o">--</span><span class="n">R2</span> <span class="n">R2</span> <span class="p">[</span><span class="n">R2</span> <span class="o">...</span><span class="p">]]</span> <span class="p">[</span><span class="o">--</span><span class="n">R1trim</span> <span class="n">R1TRIM</span> <span class="p">[</span><span class="n">R1TRIM</span> <span class="o">...</span><span class="p">]]</span>
                           <span class="p">[</span><span class="o">--</span><span class="n">R2trim</span> <span class="n">R2TRIM</span> <span class="p">[</span><span class="n">R2TRIM</span> <span class="o">...</span><span class="p">]]</span> <span class="p">[</span><span class="o">--</span><span class="n">bclen2</span> <span class="n">BCLEN2</span><span class="p">]</span>
                           <span class="p">[</span><span class="o">--</span><span class="n">chartype</span> <span class="p">{</span><span class="n">codon</span><span class="p">}]</span> <span class="p">[</span><span class="o">--</span><span class="n">maxmuts</span> <span class="n">MAXMUTS</span><span class="p">]</span>
                           <span class="p">[</span><span class="o">--</span><span class="n">minq</span> <span class="n">MINQ</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">minreads</span> <span class="n">MINREADS</span><span class="p">]</span>
                           <span class="p">[</span><span class="o">--</span><span class="n">minfraccall</span> <span class="n">MINFRACCALL</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">minconcur</span> <span class="n">MINCONCUR</span><span class="p">]</span>
                           <span class="p">[</span><span class="o">--</span><span class="n">sitemask</span> <span class="n">SITEMASK</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">purgeread</span> <span class="n">PURGEREAD</span><span class="p">]</span>
                           <span class="p">[</span><span class="o">--</span><span class="n">purgebc</span> <span class="n">PURGEBC</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">bcinfo</span><span class="p">]</span> <span class="o">--</span><span class="n">batchfile</span>
                           <span class="n">BATCHFILE</span> <span class="o">--</span><span class="n">summaryprefix</span> <span class="n">SUMMARYPREFIX</span>
</pre></div>
</div>
<div class="section" id="Named Arguments">
<h3><a class="toc-backref" href="#id5">Named Arguments</a><a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>--outdir</kbd></td>
<td>Output files to this directory (create if needed).</td></tr>
<tr><td class="option-group">
<kbd>--ncpus</kbd></td>
<td><p class="first">Number of CPUs to use, -1 is all available.</p>
<p>Default: -1</p>
<p class="last">Multiple runs of dms2_bcsubamp can be performed in parallel on the different samples specified by –batchfile.
This argument determines how many CPUs are used if running multiple jobs.</p>
</td></tr>
<tr><td class="option-group">
<kbd>--use_existing</kbd></td>
<td><p class="first">Possible choices: yes, no</p>
<p>If files with names of expected output already exist, do not re-run.</p>
<p class="last">Default: “no”</p>
</td></tr>
<tr><td class="option-group">
<kbd>-v, --version</kbd></td>
<td>show program’s version number and exit</td></tr>
<tr><td class="option-group">
<kbd>--refseq</kbd></td>
<td>Align subamplicons to gene in this FASTA file.</td></tr>
<tr><td class="option-group">
<kbd>--alignspecs</kbd></td>
<td>Subamplicon alignment positions as ‘REFSEQSTART,REFSEQEND,R1START,R2START’. REFSEQSTART is nt (1, 2, … numbering) in ‘refseq’ where nt R1START in R1 aligns. REFSEQEND is nt in ‘refseq’ where nt R2START in R2 aligns.’</td></tr>
<tr><td class="option-group">
<kbd>--bclen</kbd></td>
<td><p class="first">Length of NNN… barcode at start of each read. Assumed to be same for R1 and R2, use <cite>–bclen2</cite> if this is not the case.</p>
<p class="last">Default: 8</p>
</td></tr>
<tr><td class="option-group">
<kbd>--fastqdir</kbd></td>
<td>R1 and R2 files in this directory.</td></tr>
<tr><td class="option-group">
<kbd>--R2</kbd></td>
<td>Read 2 (R2) FASTQ files assumed to have same names as R1 but with ‘_R1’ replaced by ‘_R2’. If that is not case, provide names here.</td></tr>
<tr><td class="option-group">
<kbd>--R1trim</kbd></td>
<td>Trim R1 from 3’ end to this length. One value for all reads or values for each subamplicon in <code class="docutils literal"><span class="pre">--alignspecs</span></code>.</td></tr>
<tr><td class="option-group">
<kbd>--R2trim</kbd></td>
<td>Like ‘–R1trim’, but for R2.</td></tr>
<tr><td class="option-group">
<kbd>--bclen2</kbd></td>
<td>If R1 and R2 have different length barcodes, use <cite>–bclen</cite> for R1 length and <cite>–bclen2</cite> for R2 length.</td></tr>
<tr><td class="option-group">
<kbd>--chartype</kbd></td>
<td><p class="first">Possible choices: codon</p>
<p>Character type for which we count mutations.</p>
<p class="last">Default: “codon”</p>
</td></tr>
<tr><td class="option-group">
<kbd>--maxmuts</kbd></td>
<td><p class="first">Max allowed mismatches in alignment of subamplicon; mismatches counted in terms of character ‘–chartype’.</p>
<p class="last">Default: 4</p>
</td></tr>
<tr><td class="option-group">
<kbd>--minq</kbd></td>
<td><p class="first">Only call nucleotides with Q score &gt;= this.</p>
<p class="last">Default: 15</p>
</td></tr>
<tr><td class="option-group">
<kbd>--minreads</kbd></td>
<td><p class="first">Require this many reads in a barcode to agree to call consensus nucleotide identity.</p>
<p class="last">Default: 2</p>
</td></tr>
<tr><td class="option-group">
<kbd>--minfraccall</kbd></td>
<td><p class="first">Retain only barcodes where trimmed consensus sequence for each read has &gt;= this frac sites called.</p>
<p class="last">Default: 0.95</p>
</td></tr>
<tr><td class="option-group">
<kbd>--minconcur</kbd></td>
<td><p class="first">Only call consensus identity for barcode when &gt;= this fraction of reads concur.</p>
<p class="last">Default: 0.75</p>
</td></tr>
<tr><td class="option-group">
<kbd>--sitemask</kbd></td>
<td>Use to only consider mutations at a subset of sites. Should be a CSV file with column named <cite>site</cite> listing all sites to include.</td></tr>
<tr><td class="option-group">
<kbd>--purgeread</kbd></td>
<td><p class="first">Randomly purge read pairs with this probability to subsample data.</p>
<p class="last">Default: 0</p>
</td></tr>
<tr><td class="option-group">
<kbd>--purgebc</kbd></td>
<td><p class="first">Randomly purge barcodes with this probability to subsample data.</p>
<p class="last">Default: 0</p>
</td></tr>
<tr><td class="option-group">
<kbd>--bcinfo</kbd></td>
<td><p class="first">Create file with suffix ‘bcinfo.txt.gz’ with info about each barcode.</p>
<p class="last">Default: False</p>
</td></tr>
<tr><td class="option-group">
<kbd>--batchfile</kbd></td>
<td>CSV file specifying each <code class="docutils literal"><span class="pre">dms2_bcsubamp</span></code> run. Must have these columns: <cite>name</cite>, <cite>R1</cite>. Can optionally have columns <cite>R1trim</cite> and <cite>R2trim</cite> with spaces delimiting subamplicon-specific trimming. If <cite>R1trim</cite> / <cite>R2trim</cite> in batch file, do <strong>not</strong> also give values for <code class="docutils literal"><span class="pre">--R1trim</span></code> and <code class="docutils literal"><span class="pre">--R2trim</span></code>. Other columns are ignored, so other <code class="docutils literal"><span class="pre">dms2_bcsubamp</span></code> args should be passed as separate command line args rather than in <code class="docutils literal"><span class="pre">--batchfile</span></code>.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>--summaryprefix</kbd></td>
</tr>
<tr><td>&#160;</td><td><p class="first">Prefix of output summary plots.</p>
<p class="last">As detailed in Output files below, dms2_batch_bcsubamp creates a variety of plots summarizing the output.
These files are in the directory specified by –outdir, and have the prefix specified here.
This prefix should only contain letters, numbers, dashes, and spaces.
Underscores are not allowed as they are a LaTex special character.</p>
</td></tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="output-files">
<h2><a class="toc-backref" href="#id6">Output files</a><a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h2>
<p>Running <code class="docutils literal"><span class="pre">dms2_batch_bcsubamp</span></code> produces a variety of output files, all of which will be found in the directory specified by <code class="docutils literal"><span class="pre">--outdir</span></code>.</p>
<div class="section" id="results-for-each-sample">
<h3><a class="toc-backref" href="#id7">Results for each sample</a><a class="headerlink" href="#results-for-each-sample" title="Permalink to this headline">¶</a></h3>
<p>The program <code class="docutils literal"><span class="pre">dms2_bcsubamp</span></code> is run on each sample specified by <code class="docutils literal"><span class="pre">--batchfile</span></code>, so you will create all of the <code class="docutils literal"><span class="pre">dms2_bcsubamp</span></code> <a class="reference internal" href="dms2_bcsubamp.html#bcsubamp-outputfiles"><span class="std std-ref">Output files</span></a>.</p>
</div>
<div class="section" id="summary-files">
<h3><a class="toc-backref" href="#id8">Summary files</a><a class="headerlink" href="#summary-files" title="Permalink to this headline">¶</a></h3>
<p>Plots are created that summarize the output for all samples specified by <code class="docutils literal"><span class="pre">--batchfile</span></code>.
These samples have the prefix specified by <code class="docutils literal"><span class="pre">--summaryprefix</span></code>.
So for instance, if you run <code class="docutils literal"><span class="pre">dms2_batch_bcsubamp</span></code> with the arguments <code class="docutils literal"><span class="pre">--outdir</span> <span class="pre">results</span> <span class="pre">--summaryprefix</span> <span class="pre">summary</span></code> then these files will have the prefix <code class="docutils literal"><span class="pre">./results/summary</span></code>.
They will have the suffixes listed below:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">.log</span></code>: a text file that logs the progress of the program.</li>
<li><code class="docutils literal"><span class="pre">_readstats.pdf</span></code>: plot of reads for each sample.</li>
<li><code class="docutils literal"><span class="pre">_bcstats.pdf</span></code>: plot of barcodes for each sample.</li>
<li><code class="docutils literal"><span class="pre">_readsperbc.pdf</span></code>: plot of distribution of the number of reads per-barcode for each sample.</li>
<li><code class="docutils literal"><span class="pre">_depth.pdf</span></code>: plot of number of counts called at each site for each sample.</li>
<li><code class="docutils literal"><span class="pre">_mutfreq.pdf</span></code>: plot of mutation frequency at each site for each sample.</li>
<li><code class="docutils literal"><span class="pre">_codonmuttypes.pdf</span></code>: plot of average frequency of different types of codon mutations.</li>
<li><code class="docutils literal"><span class="pre">_codonmuttypes.csv</span></code>: numerical data in <code class="docutils literal"><span class="pre">_codonmuttypes.pdf</span></code>.</li>
<li><code class="docutils literal"><span class="pre">_codnntchanges.pdf</span></code>: plot of average frequency of codon mutations with different numbers of nucleotide changes.</li>
<li><code class="docutils literal"><span class="pre">_singlentchanges.pdf</span></code>: plot frequencies of different types of nucleotide mutations among codons with just one nucleotide change.</li>
<li><code class="docutils literal"><span class="pre">_cumulmutcounts.pdf</span></code>: plot fraction of mutations that occur <span class="math">\(\leq\)</span> a given number of times.</li>
</ul>
</div></blockquote>
<p>Examples and more detailed explanations of these plots can be found in the <a class="reference external" href="https://github.com/jbloomlab/dms_tools2/blob/master/examples/Doud2016/analysis_notebook.ipynb">Doud2016 example</a>.</p>
</div>
</div>
<div class="section" id="memory-usage">
<h2><a class="toc-backref" href="#id9">Memory usage</a><a class="headerlink" href="#memory-usage" title="Permalink to this headline">¶</a></h2>
<p>As described in the <a class="reference internal" href="dms2_bcsubamp.html#bcsubamp-memoryusage"><span class="std std-ref">Memory usage</span></a> section for <code class="docutils literal"><span class="pre">dms2_bcsubamp</span></code>, each iteration of that program can consume substantial memory.
So obviously running it multiple times in parallel with <code class="docutils literal"><span class="pre">dms2_batch_bcsubamp</span></code> will consume even more memory.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">dms_tools2</h1>
    
  </a>
</p>



<p class="blurb">Tools for analyzing deep mutational scanning data.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=dms_tools2&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a href="https://travis-ci.org/jbloomlab/dms_tools2">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="bcsubamp.html">Barcoded-subamplicon sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="prefs.html">Amino-acid preferences</a></li>
<li class="toctree-l1"><a class="reference internal" href="diffsel.html">Differential selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="fracsurvive.html">Fraction surviving</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="programs.html">Executable programs</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="dms2_bcsubamp.html"><code class="docutils literal"><span class="pre">dms2_bcsubamp</span></code></a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#"><code class="docutils literal"><span class="pre">dms2_batch_bcsubamp</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="dms2_prefs.html"><code class="docutils literal"><span class="pre">dms2_prefs</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="dms2_batch_prefs.html"><code class="docutils literal"><span class="pre">dms2_batch_prefs</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="dms2_diffsel.html"><code class="docutils literal"><span class="pre">dms2_diffsel</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="dms2_batch_diffsel.html"><code class="docutils literal"><span class="pre">dms2_batch_diffsel</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="dms2_fracsurvive.html"><code class="docutils literal"><span class="pre">dms2_fracsurvive</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="dms2_batch_fracsurvive.html"><code class="docutils literal"><span class="pre">dms2_batch_fracsurvive</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="dms2_logoplot.html"><code class="docutils literal"><span class="pre">dms2_logoplot</span></code></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="api.html">Python API</a></li>
<li class="toctree-l1"><a class="reference internal" href="citations.html">Citations</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="programs.html">Executable programs</a><ul>
      <li>Previous: <a href="dms2_bcsubamp.html" title="previous chapter"><code class="docutils literal"><span class="pre">dms2_bcsubamp</span></code></a></li>
      <li>Next: <a href="dms2_prefs.html" title="next chapter"><code class="docutils literal"><span class="pre">dms2_prefs</span></code></a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017, the Bloom lab.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.6.6</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
      |
      <a href="_sources/dms2_batch_bcsubamp.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/dms_tools2" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>