
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>dms2_fracsurvive &#8212; dms_tools2 2.6.3 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="dms2_batch_fracsurvive" href="dms2_batch_fracsurvive.html" />
    <link rel="prev" title="dms2_batch_diffsel" href="dms2_batch_diffsel.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="dms2-fracsurvive">
<span id="id1"></span><h1><code class="docutils literal notranslate"><span class="pre">dms2_fracsurvive</span></code><a class="headerlink" href="#dms2-fracsurvive" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="contents">
<ul class="simple">
<li><p><a class="reference internal" href="#overview" id="id3">Overview</a></p></li>
<li><p><a class="reference internal" href="#command-line-usage" id="id4">Command-line usage</a></p>
<ul>
<li><p><a class="reference internal" href="#Named Arguments" id="id5">Named Arguments</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#output-files" id="id6">Output files</a></p>
<ul>
<li><p><a class="reference internal" href="#log-file" id="id7">Log file</a></p></li>
<li><p><a class="reference internal" href="#mutation-fraction-surviving-file" id="id8">Mutation fraction surviving file</a></p></li>
<li><p><a class="reference internal" href="#site-fraction-surviving-file" id="id9">Site fraction surviving file</a></p></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="overview">
<h2><a class="toc-backref" href="#id3">Overview</a><a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">dms2_fracsurvive</span></code> program processes files giving the number of observed counts of characters in a selected and mock-selected condition along with a measurement of the overall fraction of the library surviving the selection to estimate the <a class="reference internal" href="fracsurvive.html#fracsurvive"><span class="std std-ref">Fraction surviving</span></a> for each mutation.</p>
<p>If you have multiple related replicates or samples (or even if you have just one), you should probably use the <a class="reference internal" href="dms2_batch_fracsurvive.html#dms2-batch-fracsurvive"><span class="std std-ref">dms2_batch_fracsurvive</span></a> program rather than running <code class="docutils literal notranslate"><span class="pre">dms2_fracsurvive</span></code> directly.
This is because <a class="reference internal" href="dms2_batch_fracsurvive.html#dms2-batch-fracsurvive"><span class="std std-ref">dms2_batch_fracsurvive</span></a> runs <code class="docutils literal notranslate"><span class="pre">dms2_fracsurvive</span></code>, but then also makes some nice summary plots.</p>
</div>
<div class="section" id="command-line-usage">
<span id="fracsurvive-commandlineusage"></span><h2><a class="toc-backref" href="#id4">Command-line usage</a><a class="headerlink" href="#command-line-usage" title="Permalink to this headline">¶</a></h2>
<p><p>Estimate fraction surviving for each mutation. Part of <a class="reference external" href="http://jbloomlab.github.io/dms_tools2">dms_tools2</a> (version 2.6.3) written by <a class="reference external" href="https://github.com/jbloomlab/dms_tools2/graphs/contributors">the Bloom Lab</a>.</p>
</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">dms2_fracsurvive</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">outdir</span> <span class="n">OUTDIR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">ncpus</span> <span class="n">NCPUS</span><span class="p">]</span>
                        <span class="p">[</span><span class="o">--</span><span class="n">use_existing</span> <span class="p">{</span><span class="n">yes</span><span class="p">,</span><span class="n">no</span><span class="p">}]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">indir</span> <span class="n">INDIR</span><span class="p">]</span>
                        <span class="p">[</span><span class="o">--</span><span class="n">chartype</span> <span class="p">{</span><span class="n">codon_to_aa</span><span class="p">}]</span> <span class="p">[</span><span class="o">--</span><span class="n">aboveavg</span> <span class="p">{</span><span class="n">yes</span><span class="p">,</span><span class="n">no</span><span class="p">}]</span>
                        <span class="p">[</span><span class="o">--</span><span class="n">excludestop</span> <span class="p">{</span><span class="n">yes</span><span class="p">,</span><span class="n">no</span><span class="p">}]</span> <span class="p">[</span><span class="o">--</span><span class="n">pseudocount</span> <span class="n">PSEUDOCOUNT</span><span class="p">]</span>
                        <span class="p">[</span><span class="o">--</span><span class="n">mincount</span> <span class="n">MINCOUNT</span><span class="p">]</span> <span class="o">--</span><span class="n">name</span> <span class="n">NAME</span> <span class="o">--</span><span class="n">sel</span> <span class="n">SEL</span> <span class="o">--</span><span class="n">mock</span>
                        <span class="n">MOCK</span> <span class="o">--</span><span class="n">libfracsurvive</span> <span class="n">LIBFRACSURVIVE</span> <span class="p">[</span><span class="o">--</span><span class="n">err</span> <span class="n">ERR</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Named Arguments">
<h3><a class="toc-backref" href="#id5">Named Arguments</a><a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h3>
<dl class="option-list">
<dt><kbd>--outdir</kbd></dt>
<dd><p>Output files to this directory (create if needed).</p>
</dd>
<dt><kbd>--ncpus</kbd></dt>
<dd><p>Number of CPUs to use, -1 is all available.</p>
<p>Default: -1</p>
</dd>
<dt><kbd>--use_existing</kbd></dt>
<dd><p>Possible choices: yes, no</p>
<p>If files with names of expected output already exist, do not re-run.</p>
<p>Default: “no”</p>
</dd>
<dt><kbd>-v, --version</kbd></dt>
<dd><p>show program’s version number and exit</p>
</dd>
<dt><kbd>--indir</kbd></dt>
<dd><p>Input counts files in this directory.</p>
<p>This option can be useful if the counts files are found in a common directory.
Instead of repeatedly listing that directory name, you can just provide it here.</p>
</dd>
<dt><kbd>--chartype</kbd></dt>
<dd><p>Possible choices: codon_to_aa</p>
<p>Characters for which fraction surviving selection is estimated. <cite>codon_to_aa</cite> = amino acids from codon counts.</p>
<p>Default: “codon_to_aa”</p>
</dd>
<dt><kbd>--aboveavg</kbd></dt>
<dd><p>Possible choices: yes, no</p>
<p>Report fracsurvive <strong>above</strong> the library average rather than direct fracsurvive values.</p>
<p>Default: “no”</p>
</dd>
<dt><kbd>--excludestop</kbd></dt>
<dd><p>Possible choices: yes, no</p>
<p>Exclude stop codons as a possible amino acid?</p>
<p>Default: “yes”</p>
</dd>
<dt><kbd>--pseudocount</kbd></dt>
<dd><p>Pseudocount added to each count for sample with smaller depth; pseudocount for other sample scaled by relative depth.</p>
<p>Default: 5</p>
</dd>
<dt><kbd>--mincount</kbd></dt>
<dd><p>Report as <cite>NaN</cite> the fracsurvive of mutations for which both selected and mock-selected samples have &lt; this many counts.</p>
<p>Default: 0</p>
</dd>
<dt><kbd>--name</kbd></dt>
<dd><p>Name used for output files.</p>
<p>The <a class="reference internal" href="#output-files">Output files</a> will have a prefix equal to the name specified here.
This name should only contain letters, numbers, dashes, and spaces.
Underscores are <strong>not</strong> allowed as they are a LaTex special character.</p>
</dd>
<dt><kbd>--sel</kbd></dt>
<dd><p>Post-selection counts file or prefix used when creating this file.</p>
<p>The counts files have the format of the files created by programs such as <code class="docutils literal notranslate"><span class="pre">dms2_bcsubamp</span></code>.
Specifically, they must have the following columns: ‘site’, ‘wildtype’, and then a column for each possible character (e.g., codon).</p>
</dd>
<dt><kbd>--mock</kbd></dt>
<dd><p>Like <code class="docutils literal notranslate"><span class="pre">--sel</span></code>, but for mock-selection counts.</p>
</dd>
<dt><kbd>--libfracsurvive</kbd></dt>
<dd><p>Overall fraction of total library surviving selection versus mock condition. Should be between 0 and 1.</p>
</dd>
<dt><kbd>--err</kbd></dt>
<dd><p>Like <code class="docutils literal notranslate"><span class="pre">--sel</span></code> but for error-control to correct mutation counts.</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="output-files">
<span id="fracsurvive-outputfiles"></span><h2><a class="toc-backref" href="#id6">Output files</a><a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h2>
<p>The output files all have the prefix specified by <code class="docutils literal notranslate"><span class="pre">--outdir</span></code> and <code class="docutils literal notranslate"><span class="pre">--name</span></code>.
For instance, if you use <code class="docutils literal notranslate"><span class="pre">--outdir</span> <span class="pre">results</span> <span class="pre">--name</span> <span class="pre">replicate-1</span></code>, then the output files will have the prefix <code class="docutils literal notranslate"><span class="pre">./results/replicate-1</span></code> and the suffixes described below.</p>
<p>Here are the specific output files:</p>
<div class="section" id="log-file">
<h3><a class="toc-backref" href="#id7">Log file</a><a class="headerlink" href="#log-file" title="Permalink to this headline">¶</a></h3>
<p>This file has the suffix <code class="docutils literal notranslate"><span class="pre">.log</span></code>.
It is a text file that logs the progress of the program.</p>
</div>
<div class="section" id="mutation-fraction-surviving-file">
<h3><a class="toc-backref" href="#id8">Mutation fraction surviving file</a><a class="headerlink" href="#mutation-fraction-surviving-file" title="Permalink to this headline">¶</a></h3>
<p>This file has the suffix <code class="docutils literal notranslate"><span class="pre">_mutfracsurvive.csv</span></code>.
It gives the fraction surviving for each mutation at each site, which is the <span class="math notranslate nohighlight">\(F_{r,x}\)</span> value defined in Equation <a class="reference internal" href="fracsurvive.html#equation-fracsurvive">(30)</a> of the <a class="reference internal" href="fracsurvive.html#fracsurvive"><span class="std std-ref">Fraction surviving</span></a> section.
Note that the quantity is calculated for the wildtype as well as the mutant characters at each site.
Note also that if you are using <code class="docutils literal notranslate"><span class="pre">--aboveavg</span> <span class="pre">yes</span></code> then these are the fraction surviving <strong>above the library average</strong>, denoted as <span class="math notranslate nohighlight">\(F_{r,x}^{\rm{aboveavg}}\)</span> in Equation <a class="reference internal" href="fracsurvive.html#equation-fracsurviveaboveavg">(31)</a> of the <a class="reference internal" href="fracsurvive.html#fracsurvive"><span class="std std-ref">Fraction surviving</span></a> section.
If <code class="docutils literal notranslate"><span class="pre">--mincounts</span></code> is greater than zero, the fraction surviving may be undefined for some mutations due to low counts, and any such undefined values are also shown as <cite>NaN</cite>.</p>
<p>Here are the first and last few lines of a <code class="docutils literal notranslate"><span class="pre">_mutfracsurvive.csv</span></code> file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">site</span><span class="p">,</span><span class="n">wildtype</span><span class="p">,</span><span class="n">mutation</span><span class="p">,</span><span class="n">mutfracsurvive</span>
<span class="mi">156</span><span class="p">,</span><span class="n">G</span><span class="p">,</span><span class="n">S</span><span class="p">,</span><span class="mf">0.8189280293912643</span>
<span class="mi">146</span><span class="p">,</span><span class="n">N</span><span class="p">,</span><span class="n">D</span><span class="p">,</span><span class="mf">0.626080490632122</span>
<span class="mi">157</span><span class="p">,</span><span class="n">K</span><span class="p">,</span><span class="n">S</span><span class="p">,</span><span class="mf">0.5933429890043687</span>
<span class="mi">158</span><span class="p">,</span><span class="n">S</span><span class="p">,</span><span class="n">A</span><span class="p">,</span><span class="mf">0.5723610875357631</span>
<span class="o">...</span>
<span class="mi">540</span><span class="p">,</span><span class="n">L</span><span class="p">,</span><span class="n">C</span><span class="p">,</span><span class="mf">0.0016521655105078295</span>
<span class="mi">175</span><span class="p">,</span><span class="n">P</span><span class="p">,</span><span class="n">G</span><span class="p">,</span><span class="mf">0.0013556328151850907</span>
<span class="mi">545</span><span class="p">,</span><span class="n">S</span><span class="p">,</span><span class="n">T</span><span class="p">,</span><span class="mf">0.001310686952133144</span>
<span class="mi">490</span><span class="p">,</span><span class="n">E</span><span class="p">,</span><span class="n">D</span><span class="p">,</span><span class="mf">0.001016678753248357</span>
</pre></div>
</div>
<p>Note that the file is sorted from largest to smallest fraction surviving.</p>
</div>
<div class="section" id="site-fraction-surviving-file">
<h3><a class="toc-backref" href="#id9">Site fraction surviving file</a><a class="headerlink" href="#site-fraction-surviving-file" title="Permalink to this headline">¶</a></h3>
<p>This file has the suffix <code class="docutils literal notranslate"><span class="pre">_sitefracsurvive.csv</span></code>
It gives several measures that summarize the fraction surviving each site.
All values in the <code class="docutils literal notranslate"><span class="pre">_sitefracsurvive.csv</span></code> file can be calculated from the values in the <code class="docutils literal notranslate"><span class="pre">_mutfracsurvive.csv</span></code> file, but the program outputs both files to make things simpler for the user.</p>
<p>Specifically, it gives the following quantities:</p>
<ul class="simple">
<li><p><cite>avgfracsurvive</cite> is the average of the mutation fraction surviving values. If any of the mutation fraction surviving values are <cite>NaN</cite> (which can happen if you use <code class="docutils literal notranslate"><span class="pre">--mincounts</span></code>), they are <strong>not</strong> included in this average.</p></li>
<li><p><cite>maxfracsurvive</cite> is the <strong>maximum</strong> mutation fraction surviving taken over all <strong>non</strong>-wildtype characters for each site.</p></li>
</ul>
<p>Here are the first and last lines of a <code class="docutils literal notranslate"><span class="pre">_sitefracsurvive.csv</span></code> file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">site</span><span class="p">,</span><span class="n">avgfracsurvive</span><span class="p">,</span><span class="n">maxfracsurvive</span>
<span class="mi">153</span><span class="p">,</span><span class="mf">0.2841412228950011</span><span class="p">,</span><span class="mf">0.54440652255242</span>
<span class="mi">157</span><span class="p">,</span><span class="mf">0.25281575693873276</span><span class="p">,</span><span class="mf">0.5933429890043687</span>
<span class="mi">136</span><span class="p">,</span><span class="mf">0.16447395413487326</span><span class="p">,</span><span class="mf">0.33671209426557835</span>
<span class="mi">156</span><span class="p">,</span><span class="mf">0.13827268994538547</span><span class="p">,</span><span class="mf">0.8189280293912643</span>
<span class="o">...</span>
<span class="mi">210</span><span class="p">,</span><span class="mf">0.013137991003787609</span><span class="p">,</span><span class="mf">0.022298531591000946</span>
<span class="mi">170</span><span class="p">,</span><span class="mf">0.011505865316217256</span><span class="p">,</span><span class="mf">0.029289469175867944</span>
<span class="mi">176</span><span class="p">,</span><span class="mf">0.010814303017871948</span><span class="p">,</span><span class="mf">0.02157678717020969</span>
<span class="mi">175</span><span class="p">,</span><span class="mf">0.008361156202363286</span><span class="p">,</span><span class="mf">0.027496593984557578</span>
</pre></div>
</div>
<p>If all mutations at a site have a mutation fraction surviving of <cite>NaN</cite> (which can be the case if <code class="docutils literal notranslate"><span class="pre">--mincounts</span></code> is &gt; 0), then the site values are reported as <cite>NaN</cite>.</p>
</div>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">dms_tools2</h1>
    
  </a>
</p>



<p class="blurb">Tools for analyzing deep mutational scanning data.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=dms_tools2&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a class="badge" href="https://travis-ci.org/jbloomlab/dms_tools2">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="bcsubamp.html">Barcoded-subamplicon sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="prefs.html">Amino-acid preferences</a></li>
<li class="toctree-l1"><a class="reference internal" href="diffsel.html">Differential selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="fracsurvive.html">Fraction surviving</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="programs.html">Executable programs</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="dms2_bcsubamp.html"><code class="docutils literal notranslate"><span class="pre">dms2_bcsubamp</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="dms2_batch_bcsubamp.html"><code class="docutils literal notranslate"><span class="pre">dms2_batch_bcsubamp</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="dms2_prefs.html"><code class="docutils literal notranslate"><span class="pre">dms2_prefs</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="dms2_batch_prefs.html"><code class="docutils literal notranslate"><span class="pre">dms2_batch_prefs</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="dms2_diffsel.html"><code class="docutils literal notranslate"><span class="pre">dms2_diffsel</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="dms2_batch_diffsel.html"><code class="docutils literal notranslate"><span class="pre">dms2_batch_diffsel</span></code></a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#"><code class="docutils literal notranslate"><span class="pre">dms2_fracsurvive</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="dms2_batch_fracsurvive.html"><code class="docutils literal notranslate"><span class="pre">dms2_batch_fracsurvive</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="dms2_logoplot.html"><code class="docutils literal notranslate"><span class="pre">dms2_logoplot</span></code></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="api.html">Python API</a></li>
<li class="toctree-l1"><a class="reference internal" href="citations.html">Citations</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="programs.html">Executable programs</a><ul>
      <li>Previous: <a href="dms2_batch_diffsel.html" title="previous chapter"><code class="docutils literal notranslate"><span class="pre">dms2_batch_diffsel</span></code></a></li>
      <li>Next: <a href="dms2_batch_fracsurvive.html" title="next chapter"><code class="docutils literal notranslate"><span class="pre">dms2_batch_fracsurvive</span></code></a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017--2019, the Bloom lab.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/dms2_fracsurvive.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/dms_tools2" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>