
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>plot &#8212; dms_tools2 2.4.9 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="prefs" href="dms_tools2.prefs.html" />
    <link rel="prev" title="parseargs" href="dms_tools2.parseargs.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <span class="target" id="module-dms_tools2.plot"></span><div class="section" id="plot">
<h1>plot<a class="headerlink" href="#plot" title="Permalink to this headline">¶</a></h1>
<p>Plotting related functions for <code class="docutils literal notranslate"><span class="pre">dms_tools2</span></code>.</p>
<p>Uses <a class="reference external" href="https://plotnine.readthedocs.io/en/stable">plotnine</a>
and <a class="reference external" href="https://seaborn.pydata.org/index.html">seaborn</a>.</p>
<dl class="class">
<dt id="dms_tools2.plot.AugmentedPairGrid">
<em class="property">class </em><code class="descclassname">dms_tools2.plot.</code><code class="descname">AugmentedPairGrid</code><span class="sig-paren">(</span><em>data</em>, <em>hue=None</em>, <em>hue_order=None</em>, <em>palette=None</em>, <em>hue_kws=None</em>, <em>vars=None</em>, <em>x_vars=None</em>, <em>y_vars=None</em>, <em>diag_sharey=True</em>, <em>height=2.5</em>, <em>aspect=1</em>, <em>despine=True</em>, <em>dropna=True</em>, <em>size=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/plot.html#AugmentedPairGrid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.plot.AugmentedPairGrid" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">seaborn.axisgrid.PairGrid</span></code></p>
<p>Augmented version of <cite>seaborn.PairGrid</cite>.</p>
<dl class="method">
<dt id="dms_tools2.plot.AugmentedPairGrid.ax_lims_clip_outliers">
<code class="descname">ax_lims_clip_outliers</code><span class="sig-paren">(</span><em>frac_clip=0.001</em>, <em>extend=0.03</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/plot.html#AugmentedPairGrid.ax_lims_clip_outliers"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.plot.AugmentedPairGrid.ax_lims_clip_outliers" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets axis limits to clip outliers in data.</p>
<p>Useful if there are a few data points far outside the range
of most of the data.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>frac_clip</cite> (float)</dt>
<dd>Set upper and lower limits so that this fraction of
data is outside limits at both ends. Done <strong>before</strong>
adding <cite>extend</cite>.</dd>
<dt><cite>extend</cite> (float)</dt>
<dd>Extend the limits determined by <cite>frac_clip</cite> by this
fraction of the data range.</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="data">
<dt id="dms_tools2.plot.COLOR_BLIND_PALETTE">
<code class="descclassname">dms_tools2.plot.</code><code class="descname">COLOR_BLIND_PALETTE</code><em class="property"> = ['#000000', '#E69F00', '#56B4E9', '#009E73', '#F0E442', '#0072B2', '#D55E00', '#CC79A7']</em><a class="headerlink" href="#dms_tools2.plot.COLOR_BLIND_PALETTE" title="Permalink to this definition">¶</a></dt>
<dd><p><a class="reference external" href="http://bconnelly.net/2013/10/creating-colorblind-friendly-figures/">color-blind safe palette</a>
use by adding to your plots the following
<cite>scale_fill_manual(COLOR_BLIND_PALETTE)</cite> or
<cite>scale_color_manual(COLOR_BLIND_PALETTE)</cite>.</p>
</dd></dl>

<dl class="data">
<dt id="dms_tools2.plot.COLOR_BLIND_PALETTE_GRAY">
<code class="descclassname">dms_tools2.plot.</code><code class="descname">COLOR_BLIND_PALETTE_GRAY</code><em class="property"> = ['#999999', '#E69F00', '#56B4E9', '#009E73', '#F0E442', '#0072B2', '#D55E00', '#CC79A7']</em><a class="headerlink" href="#dms_tools2.plot.COLOR_BLIND_PALETTE_GRAY" title="Permalink to this definition">¶</a></dt>
<dd><p><a class="reference external" href="http://bconnelly.net/2013/10/creating-colorblind-friendly-figures/">color-blind safe palette</a>
that differs from <cite>COLOR_BLIND_PALETTE</cite> in that first 
color is gray rather than black.</p>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.plot.breaksAndLabels">
<code class="descclassname">dms_tools2.plot.</code><code class="descname">breaksAndLabels</code><span class="sig-paren">(</span><em>xi</em>, <em>x</em>, <em>n</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/plot.html#breaksAndLabels"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.plot.breaksAndLabels" title="Permalink to this definition">¶</a></dt>
<dd><p>Get breaks and labels for an axis.</p>
<p>Useful when you would like to re-label a numeric x-axis
with string labels.</p>
<p>Uses <cite>matplotlib.ticker.MaxNLocator</cite> to choose pretty breaks.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>xi</cite> (list or array)</dt>
<dd>Integer values actually assigned to axis points.</dd>
<dt><cite>x</cite> (list)</dt>
<dd>Strings corresponding to each numeric value in <cite>xi</cite>.</dd>
<dt><cite>n</cite> (int)</dt>
<dd>Approximate number of ticks to use.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>The tuple <cite>(breaks, labels)</cite> where <cite>breaks</cite> gives the
locations of breaks taken from <cite>xi</cite>, and <cite>labels</cite> is
the label for each break.</dd>
</dl>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">xi</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">213</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">xi</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="n">breaks</span><span class="p">,</span> <span class="n">labels</span><span class="p">)</span> <span class="o">=</span> <span class="n">breaksAndLabels</span><span class="p">(</span><span class="n">xi</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="mi">5</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">breaks</span>
<span class="go">[0, 50, 100, 150, 200]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">labels</span>
<span class="go">[&#39;1&#39;, &#39;51&#39;, &#39;101&#39;, &#39;151&#39;, &#39;201&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.plot.findSigSel">
<code class="descclassname">dms_tools2.plot.</code><code class="descname">findSigSel</code><span class="sig-paren">(</span><em>df</em>, <em>valcol</em>, <em>plotfile</em>, <em>fdr=0.05</em>, <em>title=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/plot.html#findSigSel"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.plot.findSigSel" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds “significant” selection at sites / mutations.</p>
<p>Designed for the case where most sites / mutations are not
under selection, but a few may be. It tries to find those
few that are under selection.</p>
<p>It does not use a mechanistic statistical model, but rather uses
<a class="reference external" href="http://scipy-cookbook.readthedocs.io/items/robust_regression.html">robust regression</a>
(soft L1 loss) to fit a gamma distribution. The rationale for
a gamma distribution is that it is negative binomial’s continuous
<a class="reference external" href="http://www.nehalemlabs.net/prototype/blog/2013/12/01/gamma-distribution-approximation-to-the-negative-binomial-distribution/">analog</a>.
It then identifies sites that clearly have <strong>larger</strong> values than
expected under this distribution. It currently does not
identify sites with <strong>smaller</strong> (or more negative) than expected
values.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>df</cite> (pandas DataFrame)</dt>
<dd>Contains data to analyze</dd>
<dt><cite>valcol</cite> (string)</dt>
<dd>Column in <cite>df</cite> with values (e.g., <cite>fracsurvive</cite>)</dd>
<dt><cite>plotfile</cite> (string)</dt>
<dd>Name of file to which we plot fit.</dd>
<dt><cite>fdr</cite> (float)</dt>
<dd>Find sites that are significant at this <cite>fdr</cite>
given fitted distribution.</dd>
<dt><cite>title</cite> (string or <cite>None</cite>)</dt>
<dd>Title for plot.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd><p class="first">Creates the plot in <cite>plotfile</cite>. Also returns
the 3-tuple <cite>(df_sigsel, cutoff, gamma_fit)</cite> where:</p>
<blockquote class="last">
<div><ul class="simple">
<li><cite>df_sigsel</cite> is copy of <cite>df</cite> with new columns
<cite>P</cite>, <cite>Q</cite>, and <cite>sig</cite>. These give P value, Q
value, and whether site meets <cite>fdr</cite> cutoff
for significance.</li>
<li><cite>cutoff</cite> is the maximum value that is <strong>not</strong>
called significant. Because FDR is a property
of a distribution, this value cannot be 
interpreted as meaning a new data point would
be called based on this cutoff, as the cutoff
would change. But <cite>cutoff</cite> is useful for 
plotting to see significant / non-significant.</li>
<li><cite>gamma_params</cite> is a <cite>numpy.ndarray</cite> that of 
length 3 that gives the shape, scale, and location
parameter of the fit gamma distribution.</li>
</ul>
</div></blockquote>
</dd>
</dl>
<p>An example: First, simulate points from a gamma distribution:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">shape_sim</span> <span class="o">=</span> <span class="mf">1.5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">scale_sim</span> <span class="o">=</span> <span class="mf">0.005</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">loc_sim</span> <span class="o">=</span> <span class="mf">0.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gamma_sim</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">gamma</span><span class="p">(</span><span class="n">shape_sim</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="n">scale_sim</span><span class="p">,</span>
<span class="gp">... </span>        <span class="n">loc</span><span class="o">=</span><span class="n">loc_sim</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nsites</span> <span class="o">=</span> <span class="mi">1000</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">scipy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span>
<span class="gp">... </span>        <span class="s1">&#39;site&#39;</span><span class="p">:[</span><span class="n">r</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nsites</span><span class="p">)],</span>
<span class="gp">... </span>        <span class="s1">&#39;fracsurvive&#39;</span><span class="p">:</span><span class="n">gamma_sim</span><span class="o">.</span><span class="n">rvs</span><span class="p">(</span><span class="n">nsites</span><span class="p">)})</span>
</pre></div>
</div>
<p>Now make two sites have “significantly” higher values:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sigsites</span> <span class="o">=</span> <span class="p">[</span><span class="mi">100</span><span class="p">,</span> <span class="mi">200</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">sigsites</span><span class="p">,</span> <span class="s1">&#39;fracsurvive&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.08</span>
</pre></div>
</div>
<p>Now plot and find the significant sites:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">plotfile</span> <span class="o">=</span> <span class="s1">&#39;_findSigSel.png&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="n">df_sigsel</span><span class="p">,</span> <span class="n">cutoff</span><span class="p">,</span> <span class="n">gamma_params</span><span class="p">)</span> <span class="o">=</span> <span class="n">findSigSel</span><span class="p">(</span>
<span class="gp">... </span>        <span class="n">df</span><span class="p">,</span> <span class="s1">&#39;fracsurvive&#39;</span><span class="p">,</span> <span class="n">plotfile</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;example&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Here is the resulting plot:</p>
<a class="reference internal image-reference" href="_images/_findSigSel.png"><img alt="_images/_findSigSel.png" class="align-center" src="_images/_findSigSel.png" style="width: 4in;" /></a>
<p>Make sure the fitted params are close to the ones used to
simulate the data:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">shape_sim</span><span class="p">,</span> <span class="n">gamma_params</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">rtol</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">atol</span><span class="o">=</span><span class="mf">1e-3</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">scale_sim</span><span class="p">,</span> <span class="n">gamma_params</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">rtol</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">atol</span><span class="o">=</span><span class="mf">1e-3</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">loc_sim</span><span class="p">,</span> <span class="n">gamma_params</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">rtol</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">atol</span><span class="o">=</span><span class="mf">1e-3</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Check that we find the correct significant sites:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">set</span><span class="p">(</span><span class="n">sigsites</span><span class="p">)</span> <span class="o">==</span> <span class="nb">set</span><span class="p">(</span><span class="n">df_sigsel</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;sig&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">site</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Make sure that sites above cutoff are significant:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">df_sigsel</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;sig&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">equals</span><span class="p">(</span><span class="n">df_sigsel</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;fracsurvive &gt; @cutoff&#39;</span><span class="p">))</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.plot.from_white_cmap">
<code class="descclassname">dms_tools2.plot.</code><code class="descname">from_white_cmap</code><span class="sig-paren">(</span><em>color</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/plot.html#from_white_cmap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.plot.from_white_cmap" title="Permalink to this definition">¶</a></dt>
<dd><p>Get matplotlib color map from white to <cite>color</cite>.</p>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.plot.hist_bins_intsafe">
<code class="descclassname">dms_tools2.plot.</code><code class="descname">hist_bins_intsafe</code><span class="sig-paren">(</span><em>x</em>, <em>method='fd'</em>, <em>shrink_threshold=None</em>, <em>maxbins=100</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/plot.html#hist_bins_intsafe"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.plot.hist_bins_intsafe" title="Permalink to this definition">¶</a></dt>
<dd><p>Histogram bins that work for integer data.</p>
<p>You can auto-choose bins using <cite>numpy.histogram</cite>.
However, if the data are integer, these bins
may be non-integer and so some bins will capture
more integers. This function fixes that.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>x</cite> (numpy array)</dt>
<dd>The data to bin.</dd>
<dt><cite>method</cite> (str)</dt>
<dd>The binning method. Can be anything
acceptable to <cite>numpy.histogram</cite> as
a <cite>bins</cite> argument.</dd>
<dt><cite>shrink_threshold</cite> (<cite>None</cite> or int)</dt>
<dd>If set to a value other than <cite>None</cite>,
apply a heuristic threshold to slow
the growth in number of bins if they
exceed this number.</dd>
<dt><cite>maxbins</cite> (int)</dt>
<dd>Maximum number of bins.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>The bin edges as returned in the second
element of <cite>numpy.histogram</cite>, but adjusted
to be of integer width if the data are all
integers.</dd>
</dl>
<p>Just like <cite>numpy.histogram</cite> for non-int data:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="mi">100</span> <span class="o">*</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">(</span><span class="mi">500</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bin_edges</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">histogram</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="s1">&#39;fd&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bin_edges_intsafe</span> <span class="o">=</span> <span class="n">hist_bins_intsafe</span><span class="p">(</span><span class="n">x</span><span class="p">)[</span> <span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">bin_edges</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">bin_edges</span><span class="p">,</span> <span class="n">bin_edges_intsafe</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">bin_edges</span><span class="p">,</span> <span class="n">bin_edges</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;int&#39;</span><span class="p">))</span>
<span class="go">False</span>
</pre></div>
</div>
<p>But gives integer bins for int data:
&gt;&gt;&gt; x = x.astype(‘int’)
&gt;&gt;&gt; bin_edges = numpy.histogram(x, bins=’fd’)[1]
&gt;&gt;&gt; bin_edges_intsafe = hist_bins_intsafe(x)
&gt;&gt;&gt; numpy.allclose(bin_edges, bin_edges_intsafe)
False
&gt;&gt;&gt; numpy.allclose(bin_edges, bin_edges.astype(‘int’))
False
&gt;&gt;&gt; numpy.allclose(bin_edges_intsafe, bin_edges_intsafe.astype(‘int’))
True</p>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.plot.latexSciNot">
<code class="descclassname">dms_tools2.plot.</code><code class="descname">latexSciNot</code><span class="sig-paren">(</span><em>xlist</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/plot.html#latexSciNot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.plot.latexSciNot" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts list of numbers to LaTex scientific notation.</p>
<p>Useful for nice axis-tick formatting.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>xlist</cite> (list or number)</dt>
<dd>Numbers to format.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>List of latex scientific notation formatted strings,
or single string if <cite>xlist</cite> is a number.</dd>
</dl>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">latexSciNot</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3120</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.0000927</span><span class="p">])</span>
<span class="go">[&#39;$0$&#39;, &#39;$3$&#39;, &#39;$3.1 \\times 10^{3}$&#39;, &#39;$-9.3 \\times 10^{-5}$&#39;]</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">latexSciNot</span><span class="p">([</span><span class="mf">0.001</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1000</span><span class="p">,</span> <span class="mf">1e6</span><span class="p">])</span>
<span class="go">[&#39;$0.001$&#39;, &#39;$1$&#39;, &#39;$10^{3}$&#39;, &#39;$10^{6}$&#39;]</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">latexSciNot</span><span class="p">([</span><span class="o">-</span><span class="mf">0.002</span><span class="p">,</span> <span class="mf">0.003</span><span class="p">,</span> <span class="mf">0.000011</span><span class="p">])</span>
<span class="go">[&#39;$-0.002$&#39;, &#39;$0.003$&#39;, &#39;$1.1 \\times 10^{-5}$&#39;]</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">latexSciNot</span><span class="p">([</span><span class="o">-</span><span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">])</span>
<span class="go">[&#39;$-0.1$&#39;, &#39;$0$&#39;, &#39;$0.1$&#39;, &#39;$0.2$&#39;]</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">latexSciNot</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">])</span>
<span class="go">[&#39;$0$&#39;, &#39;$1$&#39;, &#39;$2$&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.plot.plotBCStats">
<code class="descclassname">dms_tools2.plot.</code><code class="descname">plotBCStats</code><span class="sig-paren">(</span><em>names</em>, <em>bcstatsfiles</em>, <em>plotfile</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/plot.html#plotBCStats"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.plot.plotBCStats" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots <code class="docutils literal notranslate"><span class="pre">dms2_bcsubamp</span></code> barcode statistics for set of samples.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>names</cite> (list or series)</dt>
<dd>Names of the samples for which we are plotting statistics.</dd>
<dt><cite>bcstatsfiles</cite> (list or series)</dt>
<dd>Names of <code class="docutils literal notranslate"><span class="pre">*_bcstats.csv</span></code> files created by <code class="docutils literal notranslate"><span class="pre">dms2_bcsubamp</span></code>.</dd>
<dt><cite>plotfile</cite> (str)</dt>
<dd>Name of PDF plot file to create.</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.plot.plotCodonMutTypes">
<code class="descclassname">dms_tools2.plot.</code><code class="descname">plotCodonMutTypes</code><span class="sig-paren">(</span><em>names</em>, <em>countsfiles</em>, <em>plotfile</em>, <em>classification='aachange'</em>, <em>csvfile=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/plot.html#plotCodonMutTypes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.plot.plotCodonMutTypes" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot average frequency codon mutation types.</p>
<p>The averages are determined by summing counts for all sites.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>names</cite> (list or series)</dt>
<dd>Names of samples for which we plot statistics.</dd>
<dt><cite>countsfiles</cite> (list or series)</dt>
<dd><code class="docutils literal notranslate"><span class="pre">*_codoncounts.csv</span></code> files of type created by <code class="docutils literal notranslate"><span class="pre">dms2_bcsubamp</span></code>.</dd>
<dt><cite>plotfile</cite> (str)</dt>
<dd>Name of created PDF plot file.</dd>
<dt><cite>classification</cite> (str)</dt>
<dd><p class="first">The method used to classify the mutation types. Can be:</p>
<blockquote class="last">
<div><p><cite>aachange</cite> : stop, synonymous, nonsynonymous</p>
<p><cite>n_ntchanges</cite> : number of nucleotide changes per codon</p>
<p><cite>singlentchanges</cite> : nucleotide change in 1-nt mutations</p>
</div></blockquote>
</dd>
<dt><cite>csvfile</cite> (str or <cite>None</cite>)</dt>
<dd><cite>None</cite> or name of CSV file to which numerical data are written.</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.plot.plotColCorrs">
<code class="descclassname">dms_tools2.plot.</code><code class="descname">plotColCorrs</code><span class="sig-paren">(</span><em>df</em>, <em>plotfile</em>, <em>cols</em>, <em>*</em>, <em>lower_filter=None</em>, <em>title=None</em>, <em>shrink_threshold=25</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/plot.html#plotColCorrs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.plot.plotColCorrs" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots correlation among columns in pandas Data Frame.</p>
<p>Plots distribution of each variable and pairwise correlations.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>df</cite> (pandas DataFrame)</dt>
<dd>Data frame with data to plot.</dd>
<dt><cite>plotfile</cite> (str or <cite>None</cite>)</dt>
<dd>Name of created plot, or <cite>None</cite> if you want
plot returned.</dd>
<dt><cite>cols</cite> (list)</dt>
<dd>List of columns in <cite>df</cite> to plot.</dd>
<dt><cite>lower_filter</cite> (<cite>None</cite> or str)</dt>
<dd>Can be any string that can passed to the <cite>query</cite> function
of <cite>df</cite>. In this case, on the lower diagonal only plot
data for which this query is <cite>True</cite>.</dd>
<dt><cite>title</cite> (<cite>None</cite> or str)</dt>
<dd>Title of plot.</dd>
<dt><cite>shrink_threshold</cite> (float)</dt>
<dd>See argument of same name to <a class="reference internal" href="#dms_tools2.plot.hist_bins_intsafe" title="dms_tools2.plot.hist_bins_intsafe"><code class="xref py py-meth docutils literal notranslate"><span class="pre">hist_bins_intsafe()</span></code></a>.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>If <cite>plotfile</cite> is a string, makes the plot and does
not return anything. If <cite>plotfile</cite> is <cite>None</cite>, returns
the plot.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.plot.plotCorrMatrix">
<code class="descclassname">dms_tools2.plot.</code><code class="descname">plotCorrMatrix</code><span class="sig-paren">(</span><em>names</em>, <em>infiles</em>, <em>plotfile</em>, <em>datatype</em>, <em>trim_unshared=True</em>, <em>title=''</em>, <em>colors='black'</em>, <em>contour=False</em>, <em>ncontours=10</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/plot.html#plotCorrMatrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.plot.plotCorrMatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots correlations among replicates.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>names</cite> (list or series)</dt>
<dd>Names of samples for which we plot statistics.</dd>
<dt><cite>infiles</cite> (list or series)</dt>
<dd>CSV files containing data. Format depends on <cite>datatype</cite>.</dd>
<dt><cite>plotfile</cite> (str)</dt>
<dd>Name of created PDF plot file.</dd>
<dt><cite>datatype</cite> (str)</dt>
<dd><dl class="first last docutils">
<dt>Type of data for which we are plotting correlations:</dt>
<dd><ul class="first last simple">
<li><cite>prefs</cite>: in format returned by <code class="docutils literal notranslate"><span class="pre">dms2_prefs</span></code></li>
<li><cite>mutdiffsel</cite>: mutdiffsel from <code class="docutils literal notranslate"><span class="pre">dms2_diffsel</span></code></li>
<li><cite>abs_diffsel</cite>: sitediffsel from <code class="docutils literal notranslate"><span class="pre">dms2_diffsel</span></code></li>
<li><cite>positive_diffsel</cite>: sitediffsel from <code class="docutils literal notranslate"><span class="pre">dms2_diffsel</span></code></li>
<li><cite>max_diffsel</cite>: sitediffsel from <code class="docutils literal notranslate"><span class="pre">dms2_diffsel</span></code></li>
<li><cite>mutfracsurvive</cite>: from <code class="docutils literal notranslate"><span class="pre">dms2_fracsurvive</span></code></li>
</ul>
</dd>
</dl>
</dd>
<dt><cite>trim_unshared</cite> (bool)</dt>
<dd>What if files in <cite>infiles</cite> don’t have same sites / mutations?
If <cite>True</cite>, trim unshared one and just analyze ones
shared among all files. If <cite>False</cite>, raise an error.</dd>
<dt><cite>title</cite> (str)</dt>
<dd>Title to place above plot.</dd>
<dt><cite>colors</cite> (str or list)</dt>
<dd>Color(s) to color scatter points. If a string, should
specify one color for all plots. Otherwise should be
list of length <cite>len(names) * (len(names) - 1) // 2</cite>
giving lists of colors for plots from top to bottom, 
left to right.</dd>
<dt><cite>contour</cite> (bool)</dt>
<dd>Show contour lines from KDE rather than points.</dd>
<dt><cite>ncontours</cite> (int)</dt>
<dd>Number of contour lines if using <cite>contour</cite>.</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.plot.plotCumulMutCounts">
<code class="descclassname">dms_tools2.plot.</code><code class="descname">plotCumulMutCounts</code><span class="sig-paren">(</span><em>names</em>, <em>countsfiles</em>, <em>plotfile</em>, <em>chartype</em>, <em>nmax=15</em>, <em>maxcol=4</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/plot.html#plotCumulMutCounts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.plot.plotCumulMutCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot fraction of mutations seen &lt;= some number of times.</p>
<p>For each set of counts in <cite>countsfiles</cite>, plot the fraction
of mutations seen greater than or equal to some number of
times. This is essentially a cumulative fraction plot.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>names</cite> (list or series)</dt>
<dd>Names of samples for which we plot statistics.</dd>
<dt><cite>countsfiles</cite> (list or series)</dt>
<dd><code class="docutils literal notranslate"><span class="pre">*_codoncounts.csv</span></code> files of type created by <code class="docutils literal notranslate"><span class="pre">dms2_bcsubamp</span></code>.</dd>
<dt><cite>plotfile</cite> (str)</dt>
<dd>Name of created PDF plot file.</dd>
<dt><cite>chartype</cite> (str)</dt>
<dd><p class="first">The type of character in <cite>countsfiles</cite>.</p>
<blockquote class="last">
<div><ul class="simple">
<li><cite>codon</cite></li>
</ul>
</div></blockquote>
</dd>
<dt><cite>nmax</cite> (int)</dt>
<dd>Plot out to this number of mutation occurrences.</dd>
<dt><cite>maxcol</cite> (int)</dt>
<dd>Number of columns in faceted plot.</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.plot.plotDepth">
<code class="descclassname">dms_tools2.plot.</code><code class="descname">plotDepth</code><span class="sig-paren">(</span><em>names, countsfiles, plotfile, maxcol=4, charlist=['AAA', 'AAC', 'AAG', 'AAT', 'ACA', 'ACC', 'ACG', 'ACT', 'AGA', 'AGC', 'AGG', 'AGT', 'ATA', 'ATC', 'ATG', 'ATT', 'CAA', 'CAC', 'CAG', 'CAT', 'CCA', 'CCC', 'CCG', 'CCT', 'CGA', 'CGC', 'CGG', 'CGT', 'CTA', 'CTC', 'CTG', 'CTT', 'GAA', 'GAC', 'GAG', 'GAT', 'GCA', 'GCC', 'GCG', 'GCT', 'GGA', 'GGC', 'GGG', 'GGT', 'GTA', 'GTC', 'GTG', 'GTT', 'TAA', 'TAC', 'TAG', 'TAT', 'TCA', 'TCC', 'TCG', 'TCT', 'TGA', 'TGC', 'TGG', 'TGT', 'TTA', 'TTC', 'TTG', 'TTT']</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/plot.html#plotDepth"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.plot.plotDepth" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot sequencing depth along primary sequence.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>names</cite> (list or series)</dt>
<dd>Names of samples for which we plot statistics.</dd>
<dt><cite>countsfiles</cite> (list or series)</dt>
<dd>Files containing character counts at each site.
Should have column named <cite>site</cite> and a column
for each character in <cite>charlist</cite>.</dd>
<dt><cite>plotfile</cite> (str)</dt>
<dd>Name of created PDF plot file containing count depth.</dd>
<dt><cite>maxcol</cite> (int)</dt>
<dd>Number of columns in faceted plot.</dd>
<dt><cite>charlist</cite> (list)</dt>
<dd>Characters contained in <cite>countsfiles</cite>. For instance,
list of codons or amino acids.</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.plot.plotFacetedNeutCurves">
<code class="descclassname">dms_tools2.plot.</code><code class="descname">plotFacetedNeutCurves</code><span class="sig-paren">(</span><em>neutdata</em>, <em>plotfile</em>, <em>xlabel</em>, <em>ylabel</em>, <em>maxcol=3</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/plot.html#plotFacetedNeutCurves"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.plot.plotFacetedNeutCurves" title="Permalink to this definition">¶</a></dt>
<dd><p>Faceted neutralization curves with points and fit line.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>neutdata</cite> (pandas DataFrame)</dt>
<dd>Should have the following columns:
<cite>concentration</cite>, <cite>sample</cite>, <cite>fit</cite>, <cite>points</cite>.
The plot is faceted on <cite>sample</cite>. The line
smoothly connects all points in column
<cite>fit</cite>, and points are drawn anywhere
that <cite>points</cite> is not <cite>NaN</cite>.</dd>
<dt><cite>plotfile</cite> (str)</dt>
<dd>Name of created plot.</dd>
<dt><cite>xlabel</cite> (str)</dt>
<dd>x-axis label</dd>
<dt><cite>ylabel</cite> (str)</dt>
<dd>y-axis label</dd>
<dt><cite>maxcol</cite> (int)</dt>
<dd>Number of columns in facets.</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.plot.plotMutFreq">
<code class="descclassname">dms_tools2.plot.</code><code class="descname">plotMutFreq</code><span class="sig-paren">(</span><em>names</em>, <em>countsfiles</em>, <em>plotfile</em>, <em>maxcol=4</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/plot.html#plotMutFreq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.plot.plotMutFreq" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot mutation frequency along primary sequence.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>names</cite> (list or series)</dt>
<dd>Names of samples for which we plot statistics.</dd>
<dt><cite>countsfiles</cite> (list or series)</dt>
<dd><code class="docutils literal notranslate"><span class="pre">*_codoncounts.csv</span></code> files of type created by <code class="docutils literal notranslate"><span class="pre">dms2_bcsubamp</span></code>.</dd>
<dt><cite>plotfile</cite> (str)</dt>
<dd>Name of created PDF plot file.</dd>
<dt><cite>maxcol</cite> (int)</dt>
<dd>Number of columns in faceted plot.</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.plot.plotRarefactionCurves">
<code class="descclassname">dms_tools2.plot.</code><code class="descname">plotRarefactionCurves</code><span class="sig-paren">(</span><em>df</em>, <em>rarefy_col</em>, <em>plotfile</em>, <em>*</em>, <em>facet_col=None</em>, <em>nrow=1</em>, <em>xlabel='reads'</em>, <em>ylabel=None</em>, <em>facet_scales='free'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/plot.html#plotRarefactionCurves"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.plot.plotRarefactionCurves" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots rarefaction curves.</p>
<p>The rarefaction curves are calculated analytically using
<a class="reference internal" href="dms_tools2.utils.html#dms_tools2.utils.rarefactionCurve" title="dms_tools2.utils.rarefactionCurve"><code class="xref py py-mod docutils literal notranslate"><span class="pre">dms_tools2.utils.rarefactionCurve</span></code></a>.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>df</cite> (pandas DataFrame)</dt>
<dd>Data frame containing data. In tidy form if
faceting.</dd>
<dt><cite>rarefy_col</cite> (str)</dt>
<dd>Name of column in <cite>df</cite> that contains the variable
that we rarify. For instance, these might be strings
giving barcodes.</dd>
<dt><cite>plotfile</cite> (str)</dt>
<dd>Name of created plot.</dd>
<dt><cite>facet_col</cite> (str or <cite>None</cite>)</dt>
<dd>If not <cite>None</cite>, should be name of a column in <cite>df</cite>
that contains a variable we facet in the plot.</dd>
<dt><cite>nrow</cite> (int)</dt>
<dd>If faceting, the number of rows.</dd>
<dt><cite>xlabel</cite> (str)</dt>
<dd>X-axis label.</dd>
<dt><cite>ylabel</cite> (str or <cite>None</cite>)</dt>
<dd>Y-axis label. If <cite>None</cite>, defaults to value of <cite>rarefy_col</cite>.</dd>
<dt><cite>facet_scales</cite> (str`)</dt>
<dd>Scales for faceting. Can be “free”, “free_x”, “free_y”, or
“fixed”</dd>
</dl>
</dd>
</dl>
<p>Here is an example. First, we simulate two sets of barcodes.
For ease of fast simulation, the barcodes are just numbers here.
One samples a large set, the other a set a quarter that size with
half as many reads:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">nbc</span> <span class="o">=</span> <span class="mi">40000</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bclen</span> <span class="o">=</span> <span class="mi">10</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nreads</span> <span class="o">=</span> <span class="mi">200000</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">barcodes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">nbc</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">large_set</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">barcodes</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">nreads</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">small_set</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">barcodes</span><span class="p">[</span> <span class="p">:</span> <span class="n">nbc</span> <span class="o">//</span> <span class="mi">4</span><span class="p">],</span> <span class="n">size</span><span class="o">=</span><span class="n">nreads</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
<p>Now we put these in a tidy data frame where one column is named
“barcodes” and the other is named “sample”:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
<span class="gp">... </span>        <span class="s2">&quot;barcodes&quot;</span><span class="p">:</span><span class="nb">list</span><span class="p">(</span><span class="n">small_set</span><span class="p">)</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">large_set</span><span class="p">),</span>
<span class="gp">... </span>        <span class="s2">&quot;sample&quot;</span><span class="p">:[</span><span class="s1">&#39;small_set&#39;</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">small_set</span><span class="p">)</span> <span class="o">+</span>
<span class="gp">... </span>                 <span class="p">[</span><span class="s1">&#39;large_set&#39;</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">large_set</span><span class="p">)})</span>
</pre></div>
</div>
<p>Finally, plot the rarefaction curves:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">plotfile</span> <span class="o">=</span> <span class="s1">&#39;_plotRarefactionCurves.png&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">plotRarefactionCurves</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="s1">&#39;barcodes&#39;</span><span class="p">,</span> <span class="n">plotfile</span><span class="p">,</span> <span class="n">facet_col</span><span class="o">=</span><span class="s1">&#39;sample&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Here is the resulting plot:</p>
<a class="reference internal image-reference" href="_images/_plotRarefactionCurves.png"><img alt="_images/_plotRarefactionCurves.png" class="align-center" src="_images/_plotRarefactionCurves.png" style="width: 6in;" /></a>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.plot.plotReadStats">
<code class="descclassname">dms_tools2.plot.</code><code class="descname">plotReadStats</code><span class="sig-paren">(</span><em>names</em>, <em>readstatfiles</em>, <em>plotfile</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/plot.html#plotReadStats"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.plot.plotReadStats" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots <code class="docutils literal notranslate"><span class="pre">dms2_bcsubamp</span></code> read statistics for a set of samples.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>names</cite> (list or series)</dt>
<dd>Names of the samples for which we are plotting statistics.</dd>
<dt><cite>readstatfiles</cite> (list or series)</dt>
<dd>Names of <code class="docutils literal notranslate"><span class="pre">*_readstats.csv</span></code> files created by <code class="docutils literal notranslate"><span class="pre">dms2_bcsubamp</span></code>.</dd>
<dt><cite>plotfile</cite> (str)</dt>
<dd>Name of PDF plot file to create.</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.plot.plotReadsPerBC">
<code class="descclassname">dms_tools2.plot.</code><code class="descname">plotReadsPerBC</code><span class="sig-paren">(</span><em>names</em>, <em>readsperbcfiles</em>, <em>plotfile</em>, <em>maxreads=10</em>, <em>maxcol=6</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/plot.html#plotReadsPerBC"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.plot.plotReadsPerBC" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots <code class="docutils literal notranslate"><span class="pre">dms2_bcsubamp</span></code> reads-per-barcode stats for set of samples.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>names</cite> (list or series)</dt>
<dd>Names of samples for which we plot statistics.</dd>
<dt><cite>readsperbcfiles</cite> (list or series)</dt>
<dd>Names of <code class="docutils literal notranslate"><span class="pre">*_readsperbc.csv</span></code> files created by <code class="docutils literal notranslate"><span class="pre">dms2_bcsubamp</span></code>.</dd>
<dt><cite>plotfile</cite> (str)</dt>
<dd>Name of PDF plot file to create.</dd>
<dt><cite>maxreads</cite> (int)</dt>
<dd>For any barcodes with &gt; this many reads, just make a category
of &gt;= this.</dd>
<dt><cite>maxcol</cite> (int)</dt>
<dd>Number of columns in faceted plot.</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.plot.plotSiteDiffSel">
<code class="descclassname">dms_tools2.plot.</code><code class="descname">plotSiteDiffSel</code><span class="sig-paren">(</span><em>names</em>, <em>diffselfiles</em>, <em>plotfile</em>, <em>diffseltype</em>, <em>maxcol=2</em>, <em>white_bg=False</em>, <em>highlighted_sites=[]</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/plot.html#plotSiteDiffSel"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.plot.plotSiteDiffSel" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot site diffsel or fracsurvive along sequence.</p>
<p>Despite the function name, this function can be used to
plot either differential selection or fraction surviving.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>names</cite> (list or series)</dt>
<dd>Names of samples for which we plot statistics.</dd>
<dt><cite>diffselfiles</cite> (list or series)</dt>
<dd><code class="docutils literal notranslate"><span class="pre">*sitediffsel.csv</span></code> files from <code class="docutils literal notranslate"><span class="pre">dms2_diffsel</span></code> or
<code class="docutils literal notranslate"><span class="pre">*sitefracsurvive.csv</span></code> files from <code class="docutils literal notranslate"><span class="pre">dms2_fracsurvive</span></code>.</dd>
<dt><cite>plotfile</cite> (str)</dt>
<dd>Name of created PDF plot file.</dd>
<dt><cite>diffseltype</cite> (str)</dt>
<dd><dl class="first last docutils">
<dt>Type of diffsel or fracsurvive to plot:</dt>
<dd><ul class="first last simple">
<li><cite>positive</cite>: positive sitediffsel</li>
<li><cite>total</cite>: positive and negative sitediffsel</li>
<li><cite>max</cite>: maximum mutdiffsel</li>
<li><cite>minmax</cite>: minimum and maximum mutdiffsel</li>
<li><cite>avgfracsurvive</cite>: total site fracsurvive</li>
<li><cite>maxfracsurvive</cite>: max mutfracsurvive at site</li>
</ul>
</dd>
</dl>
</dd>
<dt><cite>maxcol</cite> (int)</dt>
<dd>Number of columns in faceted plot.</dd>
<dt><cite>white_bg</cite> (bool)</dt>
<dd>Plots will have a white background with limited other formatting.</dd>
<dt><cite>highlighted_sites</cite> (list)</dt>
<dd>Highlight sites of interest (passed in string format) in grey.</dd>
</dl>
</dd>
</dl>
</dd></dl>

</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">dms_tools2</h1>
    
  </a>
</p>



<p class="blurb">Tools for analyzing deep mutational scanning data.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=dms_tools2&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a class="badge" href="https://travis-ci.org/jbloomlab/dms_tools2">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="bcsubamp.html">Barcoded-subamplicon sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="prefs.html">Amino-acid preferences</a></li>
<li class="toctree-l1"><a class="reference internal" href="diffsel.html">Differential selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="fracsurvive.html">Fraction surviving</a></li>
<li class="toctree-l1"><a class="reference internal" href="programs.html">Executable programs</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="api.html">Python API</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="dms_tools2.html">dms_tools2 package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="citations.html">Citations</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="api.html">Python API</a><ul>
  <li><a href="dms_tools2.html">dms_tools2 package</a><ul>
      <li>Previous: <a href="dms_tools2.parseargs.html" title="previous chapter">parseargs</a></li>
      <li>Next: <a href="dms_tools2.prefs.html" title="next chapter">prefs</a></li>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017--2019, the Bloom lab.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/dms_tools2.plot.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/dms_tools2" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>