
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>prefs &#8212; dms_tools2 2.6.3 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="protstruct" href="dms_tools2.protstruct.html" />
    <link rel="prev" title="plot" href="dms_tools2.plot.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <span class="target" id="module-dms_tools2.prefs"></span><div class="section" id="prefs">
<h1>prefs<a class="headerlink" href="#prefs" title="Permalink to this headline">¶</a></h1>
<p>Performs operations related to estimating site-specific amino-acid
preferences.</p>
<p>Uses <a class="reference external" href="https://pystan.readthedocs.io/en/latest">pystan</a>
to perform MCMC for Bayesian inferences.</p>
<dl class="data">
<dt id="dms_tools2.prefs.PRIOR_MIN_VALUE">
<code class="sig-prename descclassname">dms_tools2.prefs.</code><code class="sig-name descname">PRIOR_MIN_VALUE</code><em class="property"> = 1e-07</em><a class="headerlink" href="#dms_tools2.prefs.PRIOR_MIN_VALUE" title="Permalink to this definition">¶</a></dt>
<dd><p>minimum value for Dirichlet prior elements</p>
</dd></dl>

<dl class="class">
<dt id="dms_tools2.prefs.StanModelDifferentErr">
<em class="property">class </em><code class="sig-prename descclassname">dms_tools2.prefs.</code><code class="sig-name descname">StanModelDifferentErr</code><span class="sig-paren">(</span><em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/prefs.html#StanModelDifferentErr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.prefs.StanModelDifferentErr" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p><code class="docutils literal notranslate"><span class="pre">pystan</span></code> model when <cite>error_model</cite> is <cite>different</cite>.</p>
<p>For use by inferSitePrefs`.</p>
</dd></dl>

<dl class="class">
<dt id="dms_tools2.prefs.StanModelNoneErr">
<em class="property">class </em><code class="sig-prename descclassname">dms_tools2.prefs.</code><code class="sig-name descname">StanModelNoneErr</code><span class="sig-paren">(</span><em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/prefs.html#StanModelNoneErr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.prefs.StanModelNoneErr" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p><code class="docutils literal notranslate"><span class="pre">pystan</span></code> model when <cite>error_model</cite> is <cite>none</cite>.</p>
<p>For use by inferSitePrefs`.</p>
</dd></dl>

<dl class="class">
<dt id="dms_tools2.prefs.StanModelSameErr">
<em class="property">class </em><code class="sig-prename descclassname">dms_tools2.prefs.</code><code class="sig-name descname">StanModelSameErr</code><span class="sig-paren">(</span><em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/prefs.html#StanModelSameErr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.prefs.StanModelSameErr" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p><code class="docutils literal notranslate"><span class="pre">pystan</span></code> model when <cite>error_model</cite> is <cite>same</cite>.</p>
<p>For use by inferSitePrefs`.</p>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.prefs.aafreqsFromAlignment">
<code class="sig-prename descclassname">dms_tools2.prefs.</code><code class="sig-name descname">aafreqsFromAlignment</code><span class="sig-paren">(</span><em class="sig-param">alignmentfile</em>, <em class="sig-param">codon_to_aa</em>, <em class="sig-param">ignore_gaps=True</em>, <em class="sig-param">ignore_stop=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/prefs.html#aafreqsFromAlignment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.prefs.aafreqsFromAlignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Get amino-acid frequencies at each site in alignment.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>alignmentfile</cite> (str)</dt><dd><p>FASTA file with alignment of proteins or coding sequences.</p>
</dd>
<dt><cite>codon_to_aa</cite> (bool)</dt><dd><p>If <cite>True</cite>, translate codon alignment to amino acids.</p>
</dd>
<dt><cite>ignore_gaps</cite> (bool)</dt><dd><p>Ignore gaps when calculating frequencies.</p>
</dd>
<dt><cite>ignore_stop</cite> (bool)</dt><dd><p>Ignore stop codons when calculating frequencies.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>A <cite>pandas.DataFrame</cite> with columns being <cite>site</cite> (1, 2, …
numbering) and other columns being amino acids and values
giving frequencies in alignment.</p>
</dd>
</dl>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">x</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;&gt;seq1</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>                <span class="s1">&#39;ATGGGGCAG</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>                <span class="s1">&#39;&gt;seq2</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>                <span class="s1">&#39;---AGGCAG</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>                <span class="s1">&#39;&gt;seq3</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>                <span class="s1">&#39;ATGTGACAG&#39;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">f</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
<span class="gp">... </span>    <span class="n">aafreqs</span> <span class="o">=</span> <span class="n">aafreqsFromAlignment</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">codon_to_aa</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aas_counts</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;M&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;R&#39;</span><span class="p">,</span> <span class="s1">&#39;Q&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aas_nocounts</span> <span class="o">=</span> <span class="p">[</span><span class="n">a</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">dms_tools2</span><span class="o">.</span><span class="n">AAS</span> <span class="k">if</span> <span class="n">a</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">aas_counts</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="mi">0</span> <span class="o">==</span> <span class="n">aafreqs</span><span class="p">[</span><span class="n">aas_nocounts</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">()</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">expected_counts</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span>
<span class="gp">... </span>        <span class="n">collections</span><span class="o">.</span><span class="n">OrderedDict</span><span class="p">([</span>
<span class="gp">... </span>        <span class="p">(</span><span class="s1">&#39;site&#39;</span><span class="p">,</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">]),</span> <span class="p">(</span><span class="s1">&#39;M&#39;</span><span class="p">,</span> <span class="p">[</span><span class="mf">1.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">]),</span>
<span class="gp">... </span>        <span class="p">(</span><span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="p">[</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="mi">0</span><span class="p">]),</span> <span class="p">(</span><span class="s1">&#39;R&#39;</span><span class="p">,</span> <span class="p">[</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">]),</span>
<span class="gp">... </span>        <span class="p">(</span><span class="s1">&#39;Q&#39;</span><span class="p">,</span> <span class="p">[</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">])]))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">expected_counts</span><span class="o">.</span><span class="n">equals</span><span class="p">(</span><span class="n">aafreqs</span><span class="p">[[</span><span class="s1">&#39;site&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">aas_counts</span><span class="p">])</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.prefs.avgPrefs">
<code class="sig-prename descclassname">dms_tools2.prefs.</code><code class="sig-name descname">avgPrefs</code><span class="sig-paren">(</span><em class="sig-param">prefsfiles</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/prefs.html#avgPrefs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.prefs.avgPrefs" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets average of site-specific preferences.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>prefsfiles</cite> (list)</dt><dd><p>List of CSV files containing preferences, must all be
for same sites and characters.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>A <cite>pandas.DataFrame</cite> containing the average of the
preferences in <cite>prefsfiles</cite>. In this returned
data frame, <cite>site</cite> is the index</p>
</dd>
</dl>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tf1</span> <span class="o">=</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tf2</span> <span class="o">=</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">tf1</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file1</span><span class="p">,</span> <span class="n">tf2</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file2</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">x</span> <span class="o">=</span> <span class="n">file1</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;site,A,C,G,T</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>                <span class="s1">&#39;10,0.2,0.2,0.5,0.1</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>                <span class="s1">&#39;2a,0.3,0.3,0.3,0.1&#39;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">file1</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
<span class="gp">... </span>    <span class="n">x</span> <span class="o">=</span> <span class="n">file2</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;site,A,C,G,T</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>                <span class="s1">&#39;10,0.4,0.1,0.1,0.4</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>                <span class="s1">&#39;2a,0.3,0.4,0.1,0.2&#39;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">file2</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
<span class="gp">... </span>    <span class="n">avg</span> <span class="o">=</span> <span class="n">avgPrefs</span><span class="p">([</span><span class="n">file1</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">file2</span><span class="o">.</span><span class="n">name</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="n">avg</span><span class="p">[</span><span class="s1">&#39;site&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="s1">&#39;2a&#39;</span><span class="p">,</span> <span class="s1">&#39;10&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">all</span><span class="p">()</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">avg</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">avg</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.35</span><span class="p">,</span> <span class="mf">0.15</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">avg</span><span class="p">[</span><span class="s1">&#39;G&#39;</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.2</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">avg</span><span class="p">[</span><span class="s1">&#39;T&#39;</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.15</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">])</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.prefs.inferPrefsByRatio">
<code class="sig-prename descclassname">dms_tools2.prefs.</code><code class="sig-name descname">inferPrefsByRatio</code><span class="sig-paren">(</span><em class="sig-param">charlist</em>, <em class="sig-param">sites</em>, <em class="sig-param">wts</em>, <em class="sig-param">pre</em>, <em class="sig-param">post</em>, <em class="sig-param">errpre</em>, <em class="sig-param">errpost</em>, <em class="sig-param">pseudocount</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/prefs.html#inferPrefsByRatio"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.prefs.inferPrefsByRatio" title="Permalink to this definition">¶</a></dt>
<dd><p>Site-specific preferences from normalized enrichment ratios.</p>
<p>Calculates site-specific preference <span class="math notranslate nohighlight">\(\pi_{r,a}\)</span> of each
site <span class="math notranslate nohighlight">\(r\)</span> for each character <span class="math notranslate nohighlight">\(a\)</span> using re-normalized
enrichment ratios. This function accomplishes the same goal as
<cite>inferSitePrefs</cite>, but is much simpler and faster, although less
statistically rigorous in principle.</p>
<p>Specifically, this function calculates the preferences as</p>
<div class="math notranslate nohighlight">
\[\pi_{r,a} = \frac{\phi_{r,a}}{\sum_{a'} \phi_{r,a'}}\]</div>
<p>where <span class="math notranslate nohighlight">\(\phi_{r,a}\)</span> is the enrichment ratio of character
<span class="math notranslate nohighlight">\(a\)</span> relative to wildtype after selection.</p>
<p>The actual definition of these enrichment ratios is somewhat
complex due to the need to correct for errors and add a
pseudocount. We have 4 sets of counts:</p>
<blockquote>
<div><ul class="simple">
<li><p><cite>pre</cite>: pre-selection counts</p></li>
<li><p><cite>post</cite>: post-selection counts</p></li>
<li><p><cite>errpre</cite>: error-control for pre-selection counts</p></li>
<li><p><cite>errpost</cite>: error-control for post-selection counts</p></li>
</ul>
</div></blockquote>
<p>For each set of counts <span class="math notranslate nohighlight">\(s\)</span>, let <span class="math notranslate nohighlight">\(N_r^s\)</span> (e.g., 
<span class="math notranslate nohighlight">\(N_r^{pre}\)</span>) denote the total counts for this sample at
site <span class="math notranslate nohighlight">\(r\)</span>, and let <span class="math notranslate nohighlight">\(n_{r,a}^{s}\)</span> (e.g., 
<span class="math notranslate nohighlight">\(n_{r,a}^{pre}\)</span>) denote the counts at that site for
character <span class="math notranslate nohighlight">\(a\)</span>.</p>
<p>Because some of the counts are low, we add a pseudocount 
<span class="math notranslate nohighlight">\(P\)</span> to each observation. Importantly, we need to scale
this pseudocount by the total depth for that sample at that site
in order to avoid systematically estimating different frequencies
purely as a function of sequencing depth, which will bias
the preference estimates.
Therefore, given the pseudocount value <span class="math notranslate nohighlight">\(P\)</span> defined via the
<code class="docutils literal notranslate"><span class="pre">--pseudocount</span></code> argument to <a class="reference internal" href="dms2_prefs.html#dms2-prefs"><span class="std std-ref">dms2_prefs</span></a>, we calculate the
scaled pseudocount for sample <span class="math notranslate nohighlight">\(s\)</span> and site <span class="math notranslate nohighlight">\(r\)</span> as</p>
<div class="math notranslate nohighlight">
\[P_r^s = P \times \frac{N_r^s}{\min\limits_{s'} N_r^{s'}}.\]</div>
<p>This equation makes the pseudocount <span class="math notranslate nohighlight">\(P\)</span> for the lowest-depth
sample, and scales up the pseodocounts for all other samples
proportional to their depth.</p>
<p>With this pseudocount, the
estimated frequency of character <span class="math notranslate nohighlight">\(a\)</span> at site <span class="math notranslate nohighlight">\(r\)</span>
in sample <span class="math notranslate nohighlight">\(s\)</span> is</p>
<div class="math notranslate nohighlight">
\[f_{r,a}^s = \frac{n_{r,a}^s + P_r^s}{N_{r}^s + A \times P_r^s}\]</div>
<p>where <span class="math notranslate nohighlight">\(A\)</span> is the number of characters in the alphabet (e.g.,
20 for amino acids without stop codons).</p>
<p>The <strong>error-corrected</strong> estimates of the frequency of character
<span class="math notranslate nohighlight">\(a\)</span> before and after selection are then</p>
<div class="math notranslate nohighlight">
\[ \begin{align}\begin{aligned}f_{r,a}^{before} &amp;= \max\left(\frac{P_r^{pre}}{N_{r}^{pre} + A \times P_r^{pre}},
\; f_{r,a}^{pre} + \delta_{a,\rm{wt}\left(r\right)}
- f_{r,a}^{errpre}\right)\\f_{r,a}^{after} &amp;= \max\left(\frac{P_r^{post}}{N_{r}^{post} + A \times P_r^{post}},
\; f_{r,a}^{post} + \delta_{a,\rm{wt}\left(r\right)}
- f_{r,a}^{errpost}\right)\end{aligned}\end{align} \]</div>
<p>where <span class="math notranslate nohighlight">\(\delta_{a,\rm{wt}\left(r\right)}\)</span> is the 
Kronecker-delta, equal to 1 if <span class="math notranslate nohighlight">\(a\)</span> is the same as the
wildtype character <span class="math notranslate nohighlight">\(\rm{wt}\left(r\right)\)</span> at site
<span class="math notranslate nohighlight">\(r\)</span>, and 0 otherwise. We use the <span class="math notranslate nohighlight">\(\max\)</span> operator
to ensure that even if the error-control frequency exceeds
the actual estimate, our estimated frequency never drops
below the pseudocount over the depth of the uncorrected sample.</p>
<p>Given these definitions, we then simply calculate the enrichment 
ratios as</p>
<div class="math notranslate nohighlight">
\[\phi_{r,a} = \frac{\left(f_{r,a}^{after}\right) / 
\left(f_{r,\rm{wt}\left(r\right)}^{after}\right)}
{\left(f_{r,a}^{before}\right) / 
\left(f_{r,\rm{wt}\left(r\right)}^{before}\right)}\]</div>
<p>In the case where we are <strong>not</strong> using any error-controls, then
we simply set 
<span class="math notranslate nohighlight">\(f_{r,\rm{wt}}^{errpre} = f_{r,\rm{wt}}^{errpost} 
= \delta_{a,\rm{wt}\left(r\right)}\)</span>.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>charlist</cite> (list)</dt><dd><p>Valid characters (e.g., codons, amino acids, nts).</p>
</dd>
<dt><cite>sites</cite> (list)</dt><dd><p>Sites to analyze.</p>
</dd>
<dt><cite>wts</cite> (list)</dt><dd><p><cite>wts[r]</cite> is the wildtype character at site <cite>sites[r]</cite>.</p>
</dd>
<dt><cite>pre</cite> (pandas.DataFrame)</dt><dd><p>Gives pre-selection counts. Should have columns
with names of ‘site’ and all characters in <cite>charlist</cite>.
The rows give the counts of each character at that site.</p>
</dd>
<dt><cite>post</cite> (pandas.DataFrame)</dt><dd><p>Like <cite>pre</cite> but for post-selection counts.</p>
</dd>
<dt><cite>errpre</cite> (<cite>None</cite> or pandas.DataFrame)</dt><dd><p>Like <cite>pre</cite> but for pre-selection error-control counts,
or <cite>None</cite> if there is no such control.</p>
</dd>
<dt><cite>errpost</cite> (<cite>None</cite> or pandas.DataFrame)</dt><dd><p>Like <cite>pre</cite> but for post-selection error-control counts,
or <cite>None</cite> if there is no such control.</p>
</dd>
<dt><cite>pseudocount</cite> (float or int)</dt><dd><p>The pseudocount to add to each observation.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>A pandas.DataFrame holding the preferences. The columns of
this dataframe are ‘site’ and all characters in <cite>charlist</cite>.
For each site in <cite>sites</cite>, the rows give the preference
for that character.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.prefs.inferSitePrefs">
<code class="sig-prename descclassname">dms_tools2.prefs.</code><code class="sig-name descname">inferSitePrefs</code><span class="sig-paren">(</span><em class="sig-param">charlist</em>, <em class="sig-param">wtchar</em>, <em class="sig-param">error_model</em>, <em class="sig-param">counts</em>, <em class="sig-param">priors</em>, <em class="sig-param">seed=1</em>, <em class="sig-param">niter=10000</em>, <em class="sig-param">increasetries=5</em>, <em class="sig-param">n_jobs=1</em>, <em class="sig-param">r_max=1.1</em>, <em class="sig-param">neff_min=100</em>, <em class="sig-param">nchains=4</em>, <em class="sig-param">increasefac=2</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/prefs.html#inferSitePrefs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.prefs.inferSitePrefs" title="Permalink to this definition">¶</a></dt>
<dd><p>Infers site-specific preferences by MCMC for a specific site.</p>
<p>Infer the site-specific preferences <span class="math notranslate nohighlight">\(\pi_{r,a}\)</span> for some site
<span class="math notranslate nohighlight">\(r\)</span> for each character <span class="math notranslate nohighlight">\(a\)</span> by integrating over the 
posterior defined by the product of the following priors and 
likelihoods, where <span class="math notranslate nohighlight">\(\boldsymbol{\mathbf{\pi_r}}\)</span> indicates
the vector of <span class="math notranslate nohighlight">\(\pi_{r,a}\)</span> values for all characters:</p>
<div class="math notranslate nohighlight">
\[ \begin{align}\begin{aligned}\Pr\left(\boldsymbol{\mathbf{\pi_r}}\right) 
&amp;= \mathrm{Dirichlet}\left(\boldsymbol{\mathbf{a_{\pi,r}}}\right)\\\Pr\left(\boldsymbol{\mathbf{\mu_r}}\right) 
&amp;= \mathrm{Dirichlet}\left(\boldsymbol{\mathbf{a_{\mu,r}}}\right)\\\Pr\left(\boldsymbol{\mathbf{\epsilon_r}}\right) 
&amp;= \mathrm{Dirichlet}\left(\boldsymbol{\mathbf{a_{\epsilon,r}}}\right)\\\Pr\left(\boldsymbol{\mathbf{\rho_r}}\right) 
&amp;= \mathrm{Dirichlet}\left(\boldsymbol{\mathbf{a_{\rho,r}}}\right)\\\Pr\left(\boldsymbol{\mathbf{n_r^{\rm{pre}}}} \mid \boldsymbol{\mathbf{\mu_r}}, \boldsymbol{\mathbf{\epsilon_r}}\right) 
&amp;= \mathrm{Multinomial}\left(\boldsymbol{\mathbf{n_r^{\rm{pre}}}}; \boldsymbol{\mathbf{\mu_r}} + \boldsymbol{\mathbf{\epsilon_r}} - \boldsymbol{\mathbf{\delta_r}}\right)\\\Pr\left(\boldsymbol{\mathbf{n_r^{\rm{post}}}} \mid \boldsymbol{\mathbf{\mu_r}}, \boldsymbol{\mathbf{\epsilon_r}}\right) 
&amp;= \mathrm{Multinomial}\left(\boldsymbol{\mathbf{n_r^{\rm{post}}}}; \frac{\boldsymbol{\mathbf{\mu_r}} \circ \boldsymbol{\mathbf{\pi_r}}}{\boldsymbol{\mathbf{\mu_r}} \cdot \boldsymbol{\mathbf{\pi_r}}} + \boldsymbol{\mathbf{\rho_r}} - \boldsymbol{\mathbf{\delta_r}}\right)\\\Pr\left(\boldsymbol{\mathbf{n_r^{\rm{err,pre}}}} \mid \boldsymbol{\mathbf{\epsilon_r}}\right) 
&amp;= \mathrm{Multinomial}\left(\boldsymbol{\mathbf{n_r^{\rm{err,pre}}}}; \boldsymbol{\mathbf{\epsilon_r}}\right)\\\Pr\left(\boldsymbol{\mathbf{n_r^{\rm{err,post}}}} \mid \boldsymbol{\mathbf{\rho_r}}\right) 
&amp;= \mathrm{Multinomial}\left(\boldsymbol{\mathbf{n_r^{\rm{err,post}}}}; \boldsymbol{\mathbf{\rho_r}}\right)\end{aligned}\end{align} \]</div>
<p>where <span class="math notranslate nohighlight">\(\boldsymbol{\mathbf{\delta_r}}\)</span> is a vector that is 
zero except for a one at the element corresponding to the wildtype.</p>
<p>The MCMC tries to guarantee convergence via the parameters specified
by <cite>r_max</cite>, <cite>neff_min</cite>, <cite>nchains</cite>, <cite>niter</cite>, <cite>increasefac</cite>, 
and <cite>increasetries</cite>.</p>
<dl>
<dt>Args:</dt><dd><dl>
<dt><cite>charlist</cite> (list)</dt><dd><p>List of valid characters (e.g., codons, amino acids, nts).</p>
</dd>
<dt><cite>wtchar</cite> (str)</dt><dd><p>Wildtype character at the site.</p>
</dd>
<dt><cite>error_model</cite> (str or object)</dt><dd><p>Specifies how errors are estimated. Passing error model
objects is faster if calling this function repeatedly
as they will not need to be compiled. Can be:</p>
<blockquote>
<div><ul class="simple">
<li><p>The str <cite>none</cite> or instance of <cite>StanModelNoneErr</cite>:
no errors (<span class="math notranslate nohighlight">\(\boldsymbol{\mathbf{\epsilon_r}} = \boldsymbol{\mathbf{\rho_r}} = \mathbf{0}\)</span>)</p></li>
<li><p>The str <cite>same</cite> or instance of <cite>StanModelSameErr</cite>: 
same error rates  pre and post-selection 
(<span class="math notranslate nohighlight">\(\boldsymbol{\mathbf{\epsilon_r}} = \boldsymbol{\mathbf{\rho_r}}\)</span>).</p></li>
<li><p>The str <cite>different</cite> or instance of <cite>StanModelDifferentErr</cite>: 
different error rates pre- and post-selection
(<span class="math notranslate nohighlight">\(\boldsymbol{\mathbf{\epsilon_r}} \ne \boldsymbol{\mathbf{\rho_r}}\)</span>).</p></li>
</ul>
</div></blockquote>
</dd>
<dt><cite>counts</cite> (dict)</dt><dd><p>Deep sequencing counts. Each string key should specify
a dict keyed by all characters in <cite>charlist</cite> with
the values giving integer counts for that character. The keys:</p>
<blockquote>
<div><ul class="simple">
<li><p><cite>pre</cite>: <span class="math notranslate nohighlight">\(\boldsymbol{\mathbf{n_r^{\rm{pre}}}}\)</span></p></li>
<li><p><cite>post</cite>: <span class="math notranslate nohighlight">\(\boldsymbol{\mathbf{n_r^{\rm{post}}}}\)</span></p></li>
<li><p><em>err</em>: required if <cite>error_model</cite> is <cite>same</cite>, specifies
<span class="math notranslate nohighlight">\(\boldsymbol{\mathbf{n_r^{\rm{err,pre}}}} = \boldsymbol{\mathbf{n_r^{\rm{err,post}}}}\)</span></p></li>
<li><p><cite>errpre</cite> and <cite>errpost</cite>: required if <cite>error_model</cite> is
<cite>different</cite>, specify 
<span class="math notranslate nohighlight">\(\boldsymbol{\mathbf{n_r^{\rm{err,pre}}}}\)</span> and
<span class="math notranslate nohighlight">\(\boldsymbol{\mathbf{n_r^{\rm{err,post}}}}\)</span>.</p></li>
</ul>
</div></blockquote>
</dd>
<dt><cite>priors</cite> (dict)</dt><dd><p>Specifies parameter vectors for Dirichlet priors. Each string
key should specify a dict keyed by all characters and values
giving the prior for that character. Values less than 
<cite>PRIOR_MIN_VALUE</cite> are set to <cite>PRIOR_MIN_VALUE</cite>. Keys are</p>
<blockquote>
<div><ul class="simple">
<li><p><cite>pir_prior_params</cite>: <span class="math notranslate nohighlight">\(\boldsymbol{\mathbf{a_{\pi,r}}}\)</span></p></li>
<li><p><cite>mur_prior_params</cite>: <span class="math notranslate nohighlight">\(\boldsymbol{\mathbf{a_{\mu,r}}}\)</span></p></li>
<li><p><cite>epsilonr_prior_params</cite>: only required if <cite>error_model</cite> is 
<cite>same</cite> or <cite>different</cite>, specifies 
<span class="math notranslate nohighlight">\(\boldsymbol{\mathbf{a_{\epsilon,r}}}\)</span></p></li>
<li><p><cite>rhor_prior_params</cite> : only required if <cite>error_model</cite> is 
<cite>different</cite>, specifies
<span class="math notranslate nohighlight">\(\boldsymbol{\mathbf{a_{\rho,r}}}\)</span></p></li>
</ul>
</div></blockquote>
</dd>
<dt><cite>seed</cite> (int)</dt><dd><p>Random number seed for MCMC.</p>
</dd>
<dt><cite>n_jobs</cite> (int)</dt><dd><p>Number of CPUs to use, -1 means all available.</p>
</dd>
<dt><cite>niter</cite>, <cite>increasetries</cite>, <cite>r_max</cite>, <cite>neff_min</cite>, <cite>nchains</cite>, <cite>increasefac</cite></dt><dd><p>Specify MCMC convergence. They all have reasonable defaults.
The MCMC is considered to have converged if the mean Gelman-Rubin R
statistic (<a class="reference external" href="http://www.jstor.org/stable/2246093">http://www.jstor.org/stable/2246093</a>) over all 
<span class="math notranslate nohighlight">\(\pi_{r,x}\)</span> values &lt;= <cite>r_max</cite> and the mean effective sample
size is &gt;= <cite>neff_min</cite>. The MCMC first runs with <cite>nchains</cite> chains
and <cite>niter</cite> iterations. If it fails to converge, it increases
the iterations by a factor of <cite>increasefac</cite> and tries again,
and repeats until it converges or has tried <cite>increasetries</cite>
times to increase the number of iterations. If the effective
sample size exceeds 3 times <cite>neff_min</cite> then we allow
R to be <cite>1 + 1.5 (r_max - 1)</cite>.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>The tuple <cite>(converged, pi_means, pi_95credint, logstring)</cite> where:</dt><dd><ul class="simple">
<li><p><cite>converged</cite> is <cite>True</cite> if the MCMC converged, <cite>False</cite> otherwise.</p></li>
<li><dl class="simple">
<dt><cite>pi_means</cite> is dict keyed by characters in <cite>charlist</cite></dt><dd><p>with the value giving the <span class="math notranslate nohighlight">\(\pi_{r,a}\)</span>.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>pi_95credint</cite> is dict keyed by characters, values are 2-tuples</dt><dd><p>giving median-centered credible interval for <span class="math notranslate nohighlight">\(\pi_{r,a}\)</span>.</p>
</dd>
</dl>
</li>
<li><p><cite>logstring</cite> is a string describing MCMC run and convergence.</p></li>
</ul>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.prefs.prefsEntropy">
<code class="sig-prename descclassname">dms_tools2.prefs.</code><code class="sig-name descname">prefsEntropy</code><span class="sig-paren">(</span><em class="sig-param">prefs</em>, <em class="sig-param">charlist</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/prefs.html#prefsEntropy"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.prefs.prefsEntropy" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate site entropy and number of effective characters.</p>
<p>The site entropy <span class="math notranslate nohighlight">\(h_r\)</span> for site <span class="math notranslate nohighlight">\(r\)</span> is defined as
<span class="math notranslate nohighlight">\(h_r = -\sum_{x} \pi_{r,x} \log\left(\pi_{r,x}\right)\)</span>
where <span class="math notranslate nohighlight">\(\pi_{r,x}\)</span> is the preference for character (e.g.,
amino acid) <span class="math notranslate nohighlight">\(x\)</span> at site <span class="math notranslate nohighlight">\(r\)</span>, and the log is
the natural logarithm.</p>
<p>The number of effective characters at site <span class="math notranslate nohighlight">\(r\)</span>
is <span class="math notranslate nohighlight">\(N_{\rm{eff}, r} = \exp\left(h_r\right)\)</span>.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>prefs</cite> (pandas DataFrame)</dt><dd><p>Data frame with the preferences. Should have a column
named <cite>site</cite> and a column for each character in
<cite>charlist</cite>.</p>
</dd>
<dt><cite>charlist</cite> (list)</dt><dd><p>List of the characters of interest, for example
the amino acids.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>A copy of <cite>prefs</cite> that has additional columns named
<cite>entropy</cite> and <cite>neffective</cite>, giving the site entropy and
number of effective characters for each site. For
the entropies, log is taken to base <span class="math notranslate nohighlight">\(e\)</span>.</p>
</dd>
</dl>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">charlist</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prefs</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
<span class="gp">... </span>        <span class="s1">&#39;site&#39;</span><span class="p">:[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;A&#39;</span><span class="p">:[</span><span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">,</span> <span class="mi">1</span> <span class="o">/</span> <span class="mf">3.</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;C&#39;</span><span class="p">:[</span><span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">,</span> <span class="mi">1</span> <span class="o">/</span> <span class="mf">3.</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;G&#39;</span><span class="p">:[</span><span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mi">1</span> <span class="o">/</span> <span class="mf">3.</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;T&#39;</span><span class="p">:[</span><span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">],</span>
<span class="gp">... </span>        <span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prefs_entropy</span> <span class="o">=</span> <span class="n">prefsEntropy</span><span class="p">(</span><span class="n">prefs</span><span class="p">,</span> <span class="n">charlist</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="nb">set</span><span class="p">([</span><span class="s1">&#39;entropy&#39;</span><span class="p">,</span> <span class="s1">&#39;neffective&#39;</span><span class="p">,</span> <span class="s1">&#39;site&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">charlist</span><span class="p">)</span> <span class="o">==</span> 
<span class="gp">... </span>        <span class="nb">set</span><span class="p">(</span><span class="n">prefs_entropy</span><span class="o">.</span><span class="n">columns</span><span class="p">))</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">h2</span> <span class="o">=</span> <span class="o">-</span><span class="mf">0.6</span> <span class="o">*</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mf">0.6</span><span class="p">)</span> <span class="o">-</span> <span class="mf">0.3</span> <span class="o">*</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mf">0.3</span><span class="p">)</span> <span class="o">-</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mf">0.1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">prefs_entropy</span><span class="p">[</span><span class="s1">&#39;entropy&#39;</span><span class="p">],</span> <span class="p">[</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mi">4</span><span class="p">),</span> <span class="n">h2</span><span class="p">,</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mi">3</span><span class="p">)])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">prefs_entropy</span><span class="p">[</span><span class="s1">&#39;neffective&#39;</span><span class="p">],</span> <span class="p">[</span><span class="mi">4</span><span class="p">,</span> <span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="n">h2</span><span class="p">),</span> <span class="mi">3</span><span class="p">])</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.prefs.prefsToMutEffects">
<code class="sig-prename descclassname">dms_tools2.prefs.</code><code class="sig-name descname">prefsToMutEffects</code><span class="sig-paren">(</span><em class="sig-param">prefs</em>, <em class="sig-param">charlist</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/prefs.html#prefsToMutEffects"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.prefs.prefsToMutEffects" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts amino acid preferences to effects of specific mutations.</p>
<p>If the preference of site <span class="math notranslate nohighlight">\(r\)</span> for amino acid <span class="math notranslate nohighlight">\(a\)</span> is
<span class="math notranslate nohighlight">\(\pi_{r,a}\)</span>, then the estimated effect (e.g., ratio of
enrichment ratios) for mutating that site from <span class="math notranslate nohighlight">\(x\)</span> to
<span class="math notranslate nohighlight">\(y\)</span> is <span class="math notranslate nohighlight">\(\frac{\pi_{r,y}}{\pi_{r,x}}\)</span>. Very small
values indicate disfavored mutations, and very large values
indicate highly favored mutations. The logarithm base 2 of the
expected effects is also a useful measure – negative values
are disfavored mutations, positive values are favored ones.</p>
<dl>
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>prefs</cite> (pandas DataFrame)</dt><dd><p>Preferences to analyze. The columns should be <cite>site</cite> and
a column for every character in <cite>charlist</cite>.</p>
</dd>
<dt><cite>charlist</cite> (list)</dt><dd><p>The list of characters that we are analyzing. For instance,
<cite>dms_tools2.AAS</cite> for amino acids.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>A pandas Data Frame where the columns are:</p>
<blockquote>
<div><ul class="simple">
<li><p><cite>site</cite>: the site</p></li>
<li><p><cite>initial</cite>: the initial character (e.g., amino acid)</p></li>
<li><p><cite>final</cite>: the final character after the mutation</p></li>
<li><p><cite>mutation</cite>: mutation in the string form <cite>A1G</cite></p></li>
<li><p><cite>effect</cite>: the effect of the mutation</p></li>
<li><p><cite>log2effect</cite>: the log2 of the effect of the mutation.</p></li>
</ul>
</div></blockquote>
</dd>
</dl>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">charlist</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prefs</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
<span class="gp">... </span>        <span class="s1">&#39;site&#39;</span><span class="p">:[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;A&#39;</span><span class="p">:[</span><span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;C&#39;</span><span class="p">:[</span><span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;G&#39;</span><span class="p">:[</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">],</span>
<span class="gp">... </span>        <span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">effects</span> <span class="o">=</span> <span class="n">prefsToMutEffects</span><span class="p">(</span><span class="n">prefs</span><span class="p">,</span> <span class="n">charlist</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">set</span><span class="p">(</span><span class="n">effects</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">==</span> <span class="p">{</span><span class="s1">&#39;site&#39;</span><span class="p">,</span> <span class="s1">&#39;initial&#39;</span><span class="p">,</span> <span class="s1">&#39;final&#39;</span><span class="p">,</span>
<span class="gp">... </span>        <span class="s1">&#39;mutation&#39;</span><span class="p">,</span> <span class="s1">&#39;effect&#39;</span><span class="p">,</span> <span class="s1">&#39;log2effect&#39;</span><span class="p">}</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">effects</span><span class="p">[</span><span class="n">effects</span><span class="p">[</span><span class="s1">&#39;initial&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;A&#39;</span><span class="p">][</span><span class="s1">&#39;effect&#39;</span><span class="p">],</span>
<span class="gp">... </span>        <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">effects</span><span class="p">[</span><span class="n">effects</span><span class="p">[</span><span class="s1">&#39;initial&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;C&#39;</span><span class="p">][</span><span class="s1">&#39;effect&#39;</span><span class="p">],</span>
<span class="gp">... </span>        <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">effects</span><span class="p">[</span><span class="s1">&#39;effect&#39;</span><span class="p">],</span> <span class="mi">2</span><span class="o">**</span><span class="n">effects</span><span class="p">[</span><span class="s1">&#39;log2effect&#39;</span><span class="p">])</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.prefs.prefsToMutFromWtEffects">
<code class="sig-prename descclassname">dms_tools2.prefs.</code><code class="sig-name descname">prefsToMutFromWtEffects</code><span class="sig-paren">(</span><em class="sig-param">prefs</em>, <em class="sig-param">charlist</em>, <em class="sig-param">wts</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/prefs.html#prefsToMutFromWtEffects"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.prefs.prefsToMutFromWtEffects" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts preferences effects of mutations away from wildtype.</p>
<p>This function is similar to <a class="reference internal" href="#dms_tools2.prefs.prefsToMutEffects" title="dms_tools2.prefs.prefsToMutEffects"><code class="xref py py-func docutils literal notranslate"><span class="pre">prefsToMutEffects()</span></code></a> but
computes the effects of mutations away from a given wildtype
sequence.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>prefs</cite></dt><dd><p>Same meaning as for <a class="reference internal" href="#dms_tools2.prefs.prefsToMutEffects" title="dms_tools2.prefs.prefsToMutEffects"><code class="xref py py-func docutils literal notranslate"><span class="pre">prefsToMutEffects()</span></code></a>.</p>
</dd>
<dt><cite>charlist</cite></dt><dd><p>Same meaning as for <a class="reference internal" href="#dms_tools2.prefs.prefsToMutEffects" title="dms_tools2.prefs.prefsToMutEffects"><code class="xref py py-func docutils literal notranslate"><span class="pre">prefsToMutEffects()</span></code></a>.</p>
</dd>
<dt><cite>wts</cite> (pandas DataFrame)</dt><dd><p>Has columns named “site” and “wildtype” giving the
wildtype amino acid for each site. The sites must
match those in <cite>prefs</cite>.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>A pandas DataFrame similar to that returned by
<a class="reference internal" href="#dms_tools2.prefs.prefsToMutEffects" title="dms_tools2.prefs.prefsToMutEffects"><code class="xref py py-func docutils literal notranslate"><span class="pre">prefsToMutEffects()</span></code></a> except that it only
contains mutations away from the wildtype character
at each site, and the columns specifying the wildtype
and mutant characters are named “wildtype” and “mutant”
rather than “initial” and “final”.</p>
</dd>
</dl>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">charlist</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prefs</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
<span class="gp">... </span>        <span class="s1">&#39;site&#39;</span><span class="p">:[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;A&#39;</span><span class="p">:[</span><span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;C&#39;</span><span class="p">:[</span><span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;G&#39;</span><span class="p">:[</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">],</span>
<span class="gp">... </span>        <span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">wts</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
<span class="gp">... </span>        <span class="s1">&#39;site&#39;</span><span class="p">:[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;wildtype&#39;</span><span class="p">:[</span><span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">]</span>
<span class="gp">... </span>        <span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prefsToMutFromWtEffects</span><span class="p">(</span><span class="n">prefs</span><span class="p">,</span> <span class="n">charlist</span><span class="p">,</span> <span class="n">wts</span><span class="p">)</span>
<span class="go">   site wildtype mutant mutation  effect  log2effect</span>
<span class="go">0     1        G      A      G1A     0.5        -1.0</span>
<span class="go">1     1        G      C      G1C     0.5        -1.0</span>
<span class="go">2     1        G      G      G1G     1.0         0.0</span>
<span class="go">3     2        C      A      C2A     0.5        -1.0</span>
<span class="go">4     2        C      C      C2C     1.0         0.0</span>
<span class="go">5     2        C      G      C2G     0.5        -1.0</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.prefs.rescalePrefs">
<code class="sig-prename descclassname">dms_tools2.prefs.</code><code class="sig-name descname">rescalePrefs</code><span class="sig-paren">(</span><em class="sig-param">prefs</em>, <em class="sig-param">stringency</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/prefs.html#rescalePrefs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.prefs.rescalePrefs" title="Permalink to this definition">¶</a></dt>
<dd><p>Re-scale amino acid preferences by stringency parameter.</p>
<p>If the initial preference of site <span class="math notranslate nohighlight">\(r\)</span> for amino-acid 
<span class="math notranslate nohighlight">\(a\)</span> is <span class="math notranslate nohighlight">\(\pi_{r,a}\)</span>, then the re-scaled preference is</p>
<div class="math notranslate nohighlight">
\[\frac{\left(\pi_{r,a}\right)^{\beta}}{\sum_{a'} \left(\pi_{r,a'}\right)}\]</div>
<p>where <span class="math notranslate nohighlight">\(\beta\)</span> is the stringency parameter.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>prefs</cite> (pandas.DataFrame)</dt><dd><p>Columns are ‘site’ and then every character (e.g.,
amino acid).</p>
</dd>
<dt><cite>stringency</cite> (float &gt;= 0)</dt><dd><p>The stringency parameter.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>A data frame in the same format as <cite>prefs</cite> but with the
re-scaled preferences.</p>
</dd>
</dl>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">prefs</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="nb">dict</span><span class="p">(</span>
<span class="gp">... </span>        <span class="p">[(</span><span class="s1">&#39;site&#39;</span><span class="p">,</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">]),</span> <span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="p">[</span><span class="mf">0.24</span><span class="p">,</span> <span class="mf">0.03</span><span class="p">]),</span> <span class="p">(</span><span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="p">[</span><span class="mf">0.04</span><span class="p">,</span> <span class="mf">0.43</span><span class="p">])]</span>
<span class="gp">... </span>        <span class="o">+</span> <span class="p">[(</span><span class="n">aa</span><span class="p">,</span> <span class="p">[</span><span class="mf">0.04</span><span class="p">,</span> <span class="mf">0.03</span><span class="p">])</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">dms_tools2</span><span class="o">.</span><span class="n">AAS</span> <span class="k">if</span>
<span class="gp">... </span>        <span class="n">aa</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">]]))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">prefs</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s1">&#39;site&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rescaled</span> <span class="o">=</span> <span class="n">rescalePrefs</span><span class="p">(</span><span class="n">prefs</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">([</span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">prefs</span><span class="p">[</span><span class="n">c</span><span class="p">],</span> <span class="n">rescaled</span><span class="p">[</span><span class="n">c</span><span class="p">])</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">prefs</span><span class="o">.</span><span class="n">columns</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rescaled2</span> <span class="o">=</span> <span class="n">rescalePrefs</span><span class="p">(</span><span class="n">prefs</span><span class="p">,</span> <span class="mf">2.0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">rescaled2</span><span class="p">[</span><span class="s1">&#39;site&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">prefs</span><span class="p">[</span><span class="s1">&#39;site&#39;</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">rescaled2</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.6545</span><span class="p">,</span> <span class="mf">0.0045</span><span class="p">],</span> <span class="n">atol</span><span class="o">=</span><span class="mf">1e-3</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">rescaled2</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.0182</span><span class="p">,</span> <span class="mf">0.9153</span><span class="p">],</span> <span class="n">atol</span><span class="o">=</span><span class="mf">1e-3</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">dms_tools2</h1>
    
  </a>
</p>



<p class="blurb">Tools for analyzing deep mutational scanning data.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=dms_tools2&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a class="badge" href="https://travis-ci.org/jbloomlab/dms_tools2">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="bcsubamp.html">Barcoded-subamplicon sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="prefs.html">Amino-acid preferences</a></li>
<li class="toctree-l1"><a class="reference internal" href="diffsel.html">Differential selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="fracsurvive.html">Fraction surviving</a></li>
<li class="toctree-l1"><a class="reference internal" href="programs.html">Executable programs</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="api.html">Python API</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="dms_tools2.html">dms_tools2 package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="citations.html">Citations</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="api.html">Python API</a><ul>
  <li><a href="dms_tools2.html">dms_tools2 package</a><ul>
      <li>Previous: <a href="dms_tools2.plot.html" title="previous chapter">plot</a></li>
      <li>Next: <a href="dms_tools2.protstruct.html" title="next chapter">protstruct</a></li>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017--2019, the Bloom lab.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/dms_tools2.prefs.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/dms_tools2" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>