
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>barcodes &#8212; dms_tools2 2.4.4 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="codonvarianttable" href="dms_tools2.codonvarianttable.html" />
    <link rel="prev" title="dms_tools2 package" href="dms_tools2.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <span class="target" id="module-dms_tools2.barcodes"></span><div class="section" id="barcodes">
<h1>barcodes<a class="headerlink" href="#barcodes" title="Permalink to this headline">¶</a></h1>
<p>Operations for sequence barcodes and UMIs.</p>
<dl class="class">
<dt id="dms_tools2.barcodes.IlluminaBarcodeParser">
<em class="property">class </em><code class="descclassname">dms_tools2.barcodes.</code><code class="descname">IlluminaBarcodeParser</code><span class="sig-paren">(</span><em>*</em>, <em>bclen=None</em>, <em>upstream=''</em>, <em>downstream=''</em>, <em>upstream_mismatch=0</em>, <em>downstream_mismatch=0</em>, <em>valid_barcodes=None</em>, <em>rc_barcode=True</em>, <em>minq=20</em>, <em>chastity_filter=True</em>, <em>list_all_valid_barcodes=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/barcodes.html#IlluminaBarcodeParser"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.barcodes.IlluminaBarcodeParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Parser for Illumina barcodes.</p>
<p>The barcodes should be read by R1 and optionally R2.
The arrangement of elements is shown below:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">5</span><span class="s1">&#39;-[R2_start]-upstream-barcode-downstream-[R1_start]-3&#39;</span>
</pre></div>
</div>
<p>R1 anneals downstream of the barcode and reads backwards. If
R2 is used, it anneals upstream of the barcode and reads forward.
There can be sequences (<cite>upstream</cite> and <cite>downstream</cite>) on either
side of the barcode: <cite>downstream</cite> must fully cover the
region between where R1 starts and the barcode, and if you are
using R2 then <cite>upstream</cite> must fully cover the region between
where R2 starts and the barcode. However, it is fine if R1
reads backwards past <cite>upstream</cite>, and if <cite>R2</cite> reads forward
past <cite>downstream</cite>.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>bclen</cite> (int or <cite>None</cite>)</dt>
<dd>Length of the barcode, or <cite>None</cite> if length is to be
determined from <cite>valid_barcodes</cite> argument.</dd>
<dt><cite>upstream</cite> (str)</dt>
<dd>Sequence upstream of the barcode.</dd>
<dt><cite>downstream</cite> (str)</dt>
<dd>Sequence downstream of barcode.</dd>
<dt><cite>upstream_mismatch</cite> (int)</dt>
<dd>Max number of mismatches allowed in <cite>upstream</cite>.</dd>
<dt><cite>downstream_mismatch</cite> (int)</dt>
<dd>Like <cite>upstream_mismatches</cite> but for <cite>downstream</cite>.</dd>
<dt><cite>valid_barcodes</cite> (<cite>None</cite> or iterable such as list, Series)</dt>
<dd>If not <cite>None</cite>, only retain barcodes listed here.
Use if you know the set of possible valid barcodes.</dd>
<dt><cite>rc_barcode</cite> (bool)</dt>
<dd>Parse the reverse complement of the barcode (the
orientation read by R1).</dd>
<dt><cite>minq</cite> (int)</dt>
<dd>Require at least this quality score for all bases
in barcode.</dd>
<dt><cite>chastity_filter</cite> (bool)</dt>
<dd>Drop any reads that fail Illumina chastity filter.</dd>
<dt><cite>list_all_valid_barcodes</cite> (bool)</dt>
<dd>If using <cite>valid_barcodes</cite>, then barcode sets returned
by <a class="reference internal" href="#dms_tools2.barcodes.IlluminaBarcodeParser.parse" title="dms_tools2.barcodes.IlluminaBarcodeParser.parse"><code class="xref py py-class docutils literal notranslate"><span class="pre">IlluminaBarcodeParser.parse</span></code></a> includes all
valid barcodes even if no counts.</dd>
</dl>
</dd>
</dl>
<p>To use, first initialize a <a class="reference internal" href="#dms_tools2.barcodes.IlluminaBarcodeParser" title="dms_tools2.barcodes.IlluminaBarcodeParser"><code class="xref py py-class docutils literal notranslate"><span class="pre">IlluminaBarcodeParser</span></code></a>, then
parse barcodes using <a class="reference internal" href="#dms_tools2.barcodes.IlluminaBarcodeParser.parse" title="dms_tools2.barcodes.IlluminaBarcodeParser.parse"><code class="xref py py-class docutils literal notranslate"><span class="pre">IlluminaBarcodeParser.parse</span></code></a>.
Barcodes are retained as valid only if R1 and R2 agree at every
nucleotide in barcode and if at each site at least one read has
a quality of at least <cite>minq</cite>.</p>
<p>Here is an example. Imagine we are parsing 4 nucleotide barcodes
that have the following construction:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">5</span><span class="s1">&#39;-[R2 binding site]-ACATGA-NNNN-GACT-[R1 binding site]-3&#39;</span>
</pre></div>
</div>
<p>First, we initialize an appropriate <a class="reference internal" href="#dms_tools2.barcodes.IlluminaBarcodeParser" title="dms_tools2.barcodes.IlluminaBarcodeParser"><code class="xref py py-class docutils literal notranslate"><span class="pre">IlluminaBarcodeParser</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">parser</span> <span class="o">=</span> <span class="n">IlluminaBarcodeParser</span><span class="p">(</span>
<span class="gp">... </span>             <span class="n">bclen</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span>
<span class="gp">... </span>             <span class="n">upstream</span><span class="o">=</span><span class="s1">&#39;ACATGA&#39;</span><span class="p">,</span>
<span class="gp">... </span>             <span class="n">downstream</span><span class="o">=</span><span class="s1">&#39;GACT&#39;</span>
<span class="gp">... </span>             <span class="p">)</span>
</pre></div>
</div>
<p>Now we write some test FASTQ files. We write valid test
reads and some invalid reads. The header for each read
explains why it is valid / invalid. We use quality scores
of <code class="docutils literal notranslate"><span class="pre">?</span></code> (30) or <code class="docutils literal notranslate"><span class="pre">+</span></code> (10) for high- and low-quality bases:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">r1file</span> <span class="o">=</span> <span class="s1">&#39;_temp_R1.fastq&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">r2file</span> <span class="o">=</span> <span class="s1">&#39;_temp_R2.fastq&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">r1file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f1</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="n">r2file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f2</span><span class="p">:</span>
<span class="gp">...</span>
<span class="gp">... </span>    <span class="c1"># valid TACG barcode, full flanking regions</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">f1</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
<span class="gp">... </span>        <span class="s1">&#39;@valid_CGTA_barcode_full_flanking_region</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;AGTCCGTATCATGT</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;+</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;??????????????</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">f2</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
<span class="gp">... </span>        <span class="s1">&#39;@valid_CGTA_barcode_full_flanking_region</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;ACATGATACGGACT</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;+</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;??????????????</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="gp">...</span>
<span class="gp">... </span>    <span class="c1"># valid CGTA barcode, partial flanking regions</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">f1</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
<span class="gp">... </span>        <span class="s1">&#39;@valid_CGTA_barcode_partial_flanking_region</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;AGTCCGTATCAT</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;+</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;??????????????</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">f2</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
<span class="gp">... </span>        <span class="s1">&#39;@valid_CGTA_barcode_partial_flanking_region</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;ACATGATACG</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;+</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;??????????????</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="gp">...</span>
<span class="gp">... </span>    <span class="c1"># valid GCCG barcode, extended flanking regions</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">f1</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
<span class="gp">... </span>        <span class="s1">&#39;@valid_GCCG_barcode_extended_flanking_region</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;AGTCGCCGTCATGTTAC</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;+</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;??????????????</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">f2</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
<span class="gp">... </span>        <span class="s1">&#39;@valid_GCCG_barcode_extended_flanking_region</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;ACATGACGGCGACTGAC</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;+</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;??????????????</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="gp">...</span>
<span class="gp">... </span>    <span class="c1"># AAGT barcode in R1 but R2 differs</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">f1</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
<span class="gp">... </span>        <span class="s1">&#39;@AAGT_R1_barcode_but_R2_differs</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;AGTCAAGTTCATGT</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;+</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;??????????????</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">f2</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
<span class="gp">... </span>        <span class="s1">&#39;@AAGT_R1_barcode_but_R2_differs</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;ACATGAACTAGACT</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;+</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;??????????????</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="gp">...</span>
<span class="gp">... </span>    <span class="c1"># same site low quality in R1 and R2</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">f1</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
<span class="gp">... </span>        <span class="s1">&#39;@low_quality_site_in_R1_and_R2</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;AGTCCGTATCATGT</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;+</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;?????+????????</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">f2</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
<span class="gp">... </span>        <span class="s1">&#39;@low_quality_site_in_R1_and_R2</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;ACATGATACGGACT</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;+</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;????????+?????</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="gp">...</span>
<span class="gp">... </span>    <span class="c1"># different site low quality in R1 and R2</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">f1</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
<span class="gp">... </span>        <span class="s1">&#39;@AGTA_with_low_quality_site_in_R1</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;AGTCAGTATCATGT</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;+</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;?????+????????</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">f2</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
<span class="gp">... </span>        <span class="s1">&#39;@AGTA_with_low_quality_site_in_R1</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;ACATGATACTGACT</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;+</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;?????????+????</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="gp">...</span>
<span class="gp">... </span>    <span class="c1"># N in barcode</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">f1</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
<span class="gp">... </span>        <span class="s1">&#39;@N_in_barcode</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;AGTCCGTNTCATGT</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;+</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;??????????????</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">f2</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
<span class="gp">... </span>        <span class="s1">&#39;@N_in_barcode</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;ACATGATACGGACT</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;+</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;??????????????</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="gp">...</span>
<span class="gp">... </span>    <span class="c1"># GGAG barcode, one mismatch in each flanking region</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">f1</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
<span class="gp">... </span>        <span class="s1">&#39;@GGAG_barcode_one_mismatch_per_flank</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;GGTCGGAGTCATGA</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;+</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;??????????????</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">f2</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
<span class="gp">... </span>        <span class="s1">&#39;@GGAG_barcode_one_mismatch_per_flank</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;TCATGACTCCGACG</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;+</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;??????????????</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="gp">...</span>
<span class="gp">... </span>    <span class="c1"># GGAG barcode, two mismatch in a flanking region</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">f1</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
<span class="gp">... </span>        <span class="s1">&#39;@GGAG_barcode_two_mismatch_in_a_flank</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;GGTCGGAGTCATAA</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;+</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;??????????????</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">f2</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
<span class="gp">... </span>        <span class="s1">&#39;@GGAG_barcode_two_mismatch_in_a_flank</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;TCATGACTCCGACG</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;+</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>        <span class="s1">&#39;??????????????</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Now parse the barcodes using both the R1 and R2 files:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">barcodes</span><span class="p">,</span> <span class="n">fates</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">r1file</span><span class="p">,</span> <span class="n">r2file</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">barcodes</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
<span class="go">barcode count</span>
<span class="go">CGTA 2</span>
<span class="go">AGTA 1</span>
<span class="go">GCCG 1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">fates</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
<span class="go">fate count</span>
<span class="go">&quot;valid barcode&quot; 4</span>
<span class="go">&quot;unparseable barcode&quot; 3</span>
<span class="go">&quot;R1 / R2 disagree&quot; 1</span>
<span class="go">&quot;low quality barcode&quot; 1</span>
</pre></div>
</div>
<p>Now we parse just using R1. We gain the barcode where R1 and
R2 disagree, but lose the one where R1 is low quality at a
position where R2 is OK:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">barcodes</span><span class="p">,</span> <span class="n">fates</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">r1file</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">barcodes</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
<span class="go">barcode count</span>
<span class="go">CGTA 2</span>
<span class="go">AAGT 1</span>
<span class="go">GCCG 1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">fates</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
<span class="go">fate count</span>
<span class="go">&quot;valid barcode&quot; 4</span>
<span class="go">&quot;unparseable barcode&quot; 3</span>
<span class="go">&quot;low quality barcode&quot; 2</span>
</pre></div>
</div>
<p>Now create a parser that allows a mismatch in each flanking
region, and check that we recover a “GGAG” barcode:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">parser_mismatch</span> <span class="o">=</span> <span class="n">IlluminaBarcodeParser</span><span class="p">(</span>
<span class="gp">... </span>             <span class="n">bclen</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span>
<span class="gp">... </span>             <span class="n">upstream</span><span class="o">=</span><span class="s1">&#39;ACATGA&#39;</span><span class="p">,</span>
<span class="gp">... </span>             <span class="n">downstream</span><span class="o">=</span><span class="s1">&#39;GACT&#39;</span><span class="p">,</span>
<span class="gp">... </span>             <span class="n">upstream_mismatch</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
<span class="gp">... </span>             <span class="n">downstream_mismatch</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
<span class="gp">... </span>             <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">barcodes_mismatch</span><span class="p">,</span> <span class="n">fates_mismatch</span> <span class="o">=</span> <span class="n">parser_mismatch</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">r1file</span><span class="p">,</span> <span class="n">r2file</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">barcodes_mismatch</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
<span class="go">barcode count</span>
<span class="go">CGTA 2</span>
<span class="go">AGTA 1</span>
<span class="go">GCCG 1</span>
<span class="go">GGAG 1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">fates_mismatch</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
<span class="go">fate count</span>
<span class="go">&quot;valid barcode&quot; 5</span>
<span class="go">&quot;unparseable barcode&quot; 2</span>
<span class="go">&quot;R1 / R2 disagree&quot; 1</span>
<span class="go">&quot;low quality barcode&quot; 1</span>
</pre></div>
</div>
<p>Now parse the barcodes using <cite>valid_barcodes</cite> to set a
barcode whitelist:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">parser_wl</span> <span class="o">=</span> <span class="n">IlluminaBarcodeParser</span><span class="p">(</span>
<span class="gp">... </span>             <span class="n">upstream</span><span class="o">=</span><span class="s1">&#39;ACATGA&#39;</span><span class="p">,</span>
<span class="gp">... </span>             <span class="n">downstream</span><span class="o">=</span><span class="s1">&#39;GACT&#39;</span><span class="p">,</span>
<span class="gp">... </span>             <span class="n">valid_barcodes</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;CGTA&#39;</span><span class="p">,</span> <span class="s1">&#39;AGTA&#39;</span><span class="p">,</span> <span class="s1">&#39;TAAT&#39;</span><span class="p">}</span>
<span class="gp">... </span>             <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">barcodes_wl</span><span class="p">,</span> <span class="n">fates_wl</span> <span class="o">=</span> <span class="n">parser_wl</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">r1file</span><span class="p">,</span> <span class="n">r2file</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">barcodes_wl</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
<span class="go">barcode count</span>
<span class="go">CGTA 2</span>
<span class="go">AGTA 1</span>
<span class="go">TAAT 0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">fates_wl</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
<span class="go">fate count</span>
<span class="go">&quot;unparseable barcode&quot; 3</span>
<span class="go">&quot;valid barcode&quot; 3</span>
<span class="go">&quot;R1 / R2 disagree&quot; 1</span>
<span class="go">&quot;invalid barcode&quot; 1</span>
<span class="go">&quot;low quality barcode&quot; 1</span>
</pre></div>
</div>
<p>Remove the test FASTQ files:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">r1file</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">r2file</span><span class="p">)</span>
</pre></div>
</div>
<dl class="attribute">
<dt id="dms_tools2.barcodes.IlluminaBarcodeParser.VALID_NTS">
<code class="descname">VALID_NTS</code><em class="property"> = 'ACGTN'</em><a class="headerlink" href="#dms_tools2.barcodes.IlluminaBarcodeParser.VALID_NTS" title="Permalink to this definition">¶</a></dt>
<dd><p>valid nucleotide characters</p>
</dd></dl>

<dl class="method">
<dt id="dms_tools2.barcodes.IlluminaBarcodeParser.parse">
<code class="descname">parse</code><span class="sig-paren">(</span><em>r1files</em>, <em>r2files=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/barcodes.html#IlluminaBarcodeParser.parse"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.barcodes.IlluminaBarcodeParser.parse" title="Permalink to this definition">¶</a></dt>
<dd><p>Parses barcodes from files.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>r1files</cite> (str or list)</dt>
<dd>Name of R1 FASTQ file, or list of such files
Can optionally be gzipped.</dd>
<dt><cite>r2files</cite> (<cite>None</cite>, str, or list)</dt>
<dd><cite>None</cite> or empty list if not using R2, otherwise like R1.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd><p class="first">The 2-tuple <cite>(barcodes, fates)</cite>. In this 2-tuple:</p>
<blockquote class="last">
<div><ul class="simple">
<li><cite>barcodes</cite> is a pandas DataFrame giving the
number of observations of each barcode. The
columns are named “barcode” and “count”.</li>
<li><cite>fates</cite> is a pandas DataFrame giving the
total number of reads with each fate. The
columns are named “fate” and “count”.<ul>
<li>“valid barcode”</li>
<li>“invalid barcode”: not in our barcode whitelist</li>
<li>“R1 / R2 disagree”</li>
<li>“low quality barcode”: sequencing quality low</li>
<li>“unparseable barcode”: invalid flanking sequences
or N in barcode.</li>
</ul>
</li>
</ul>
</div></blockquote>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="dms_tools2.barcodes.almost_duplicated">
<code class="descclassname">dms_tools2.barcodes.</code><code class="descname">almost_duplicated</code><span class="sig-paren">(</span><em>barcodes</em>, <em>threshold=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/barcodes.html#almost_duplicated"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.barcodes.almost_duplicated" title="Permalink to this definition">¶</a></dt>
<dd><p>Identifies nearly identical barcodes.</p>
<p>This function mimics the pandas <cite>duplicated</cite>
function except it can also mark as duplicated almost
but not exactly identical sequences.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>barcodes</cite> (list or pandas Series)</dt>
<dd>Lists all the barcodes, which should be strings
of same length.</dd>
<dt><cite>threshold</cite> (int)</dt>
<dd>Max number of mismatches for barcodes to be
considered almost identical.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>A pandas Series of the same length as <cite>barcodes</cite>
filled with bools that indicates whether a barcode
is a near duplicate. For each group of barcodes within
<cite>threshold</cite> of each other, only one barcode will have
an entry of <cite>False</cite> (not duplicated) and the rest will
be <cite>True</cite> (duplicated). The barcode listed as not duplicated
is chosen as follows: (1) it has the most abundant barcode
in the group, (2) among barcodes that are equivalent abundant
we take the one listed first in <cite>barcodes</cite>.
Groups are computed using the directional method of
<a class="reference external" href="https://github.com/CGATOxford/UMI-tools">umi_tools</a>.</dd>
</dl>
<p>When <cite>threshold</cite> is zero, just like <cite>pandas.duplicated</cite>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">barcodes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CTC&#39;</span><span class="p">,</span> <span class="s1">&#39;ATG&#39;</span><span class="p">,</span> <span class="s1">&#39;ATG&#39;</span><span class="p">,</span> <span class="s1">&#39;ATA&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">almost_duplicated</span><span class="p">(</span><span class="n">barcodes</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">equals</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">([</span><span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">]))</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">barcodes</span><span class="p">)</span><span class="o">.</span><span class="n">duplicated</span><span class="p">()</span><span class="o">.</span><span class="n">equals</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">([</span><span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">]))</span>
<span class="go">True</span>
</pre></div>
</div>
<p>But when <cite>threshold</cite> is 1, also identifies <strong>almost</strong> equal
barcodes (in this case, ones that differ by &lt;= 1 mutation:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">almost_duplicated</span><span class="p">(</span><span class="n">barcodes</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">equals</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">([</span><span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">True</span><span class="p">]))</span>
<span class="go">True</span>
</pre></div>
</div>
<p>All barcodes are within a distance of two, so all are
near duplicates. Note how we mark as not duplicated
(<cite>False</cite>) the first one listed among the most abundant
barcode (‘ATG’):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">almost_duplicated</span><span class="p">(</span><span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">barcodes</span><span class="p">),</span> <span class="n">threshold</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">equals</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">([</span><span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">True</span><span class="p">]))</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.barcodes.fracIdentWithinBarcode">
<code class="descclassname">dms_tools2.barcodes.</code><code class="descname">fracIdentWithinBarcode</code><span class="sig-paren">(</span><em>df</em>, <em>*</em>, <em>barcode_col='barcode'</em>, <em>variant_col='variant'</em>, <em>library_col=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/barcodes.html#fracIdentWithinBarcode"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.barcodes.fracIdentWithinBarcode" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets fraction of identical variants within barcodes.</p>
<p>This function is designed for the case when you have barcoded
variants, and you want to determine how often the variants
with the same barcode are identical. To calculate this, all
variants with the same barcode are grouped, and for every pair
of variants within a barcode we compute the fraction that
are identical (note that this fraction is the
<a class="reference external" href="https://en.wikipedia.org/wiki/Diversity_index">Simpson diversity index without replacement</a>.
We then compute the average of this fraction over all pairs.</p>
<p>In the case where we are sequencing barcoded genes, the square
root of this fraction can be interpreted as the sequencing accuracy
per variant.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>df</cite> (pandas Data Frame)</dt>
<dd>Holds barcodes, variants, and (optionally) library names.</dd>
<dt><cite>barcode_col</cite> (str)</dt>
<dd>Name of column holding barcodes.</dd>
<dt><cite>variant_col</cite> (str)</dt>
<dd>Name of column holding variants (these are what we
compare to see if they are identical).</dd>
<dt><cite>library_col</cite> (str or None)</dt>
<dd>Name of libary. We do the computation separately
for each library. Set to <cite>None</cite> if only one library.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>A pandas data frame with columns named <cite>fraction_identical</cite>
and <cite>accuracy</cite> that contain the computed fraction identical
and accuracy for each library.</dd>
</dl>
<p>Here is an example. In this data frame, we see that
of the four pairs (one for barcode <em>AT</em> and three for
barcode <em>TA</em>), two are identical. As you can see
the function correctly calculates that 50% are identical,
and computes an accuracy that is the square root of 0.5:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
<span class="gp">... </span>    <span class="s1">&#39;barcode&#39;</span><span class="p">:[</span><span class="s1">&#39;AT&#39;</span><span class="p">,</span> <span class="s1">&#39;AT&#39;</span><span class="p">,</span> <span class="s1">&#39;TG&#39;</span><span class="p">,</span> <span class="s1">&#39;TA&#39;</span><span class="p">,</span> <span class="s1">&#39;TA&#39;</span><span class="p">,</span> <span class="s1">&#39;TA&#39;</span><span class="p">],</span>
<span class="gp">... </span>    <span class="s1">&#39;variant&#39;</span><span class="p">:[</span><span class="s1">&#39;v1&#39;</span><span class="p">,</span> <span class="s1">&#39;v1&#39;</span><span class="p">,</span> <span class="s1">&#39;v2&#39;</span><span class="p">,</span> <span class="s1">&#39;v3&#39;</span><span class="p">,</span> <span class="s1">&#39;v4&#39;</span><span class="p">,</span> <span class="s1">&#39;v3&#39;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fracIdentWithinBarcode</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
<span class="go">   fraction_identical  accuracy</span>
<span class="go">0                 0.5  0.707107</span>
</pre></div>
</div>
<p>Now another example with two libraries and non-standard column
names. Note that we only get results for the two libraries
with barcodes found multiple times:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
<span class="gp">... </span>    <span class="s1">&#39;bc&#39;</span>    <span class="p">:[</span><span class="s1">&#39;AT&#39;</span><span class="p">,</span> <span class="s1">&#39;AT&#39;</span><span class="p">,</span> <span class="s1">&#39;TG&#39;</span><span class="p">,</span> <span class="s1">&#39;TA&#39;</span><span class="p">,</span> <span class="s1">&#39;TA&#39;</span><span class="p">,</span> <span class="s1">&#39;TA&#39;</span><span class="p">,</span> <span class="s1">&#39;TA&#39;</span><span class="p">],</span>
<span class="gp">... </span>    <span class="s1">&#39;var&#39;</span>   <span class="p">:[</span><span class="s1">&#39;v1&#39;</span><span class="p">,</span> <span class="s1">&#39;v1&#39;</span><span class="p">,</span> <span class="s1">&#39;v2&#39;</span><span class="p">,</span> <span class="s1">&#39;v3&#39;</span><span class="p">,</span> <span class="s1">&#39;v4&#39;</span><span class="p">,</span> <span class="s1">&#39;v3&#39;</span><span class="p">,</span> <span class="s1">&#39;v3&#39;</span><span class="p">],</span>
<span class="gp">... </span>    <span class="s1">&#39;library&#39;</span><span class="p">:[</span><span class="s1">&#39;s1&#39;</span><span class="p">,</span> <span class="s1">&#39;s1&#39;</span><span class="p">,</span> <span class="s1">&#39;s2&#39;</span><span class="p">,</span> <span class="s1">&#39;s3&#39;</span><span class="p">,</span> <span class="s1">&#39;s3&#39;</span><span class="p">,</span> <span class="s1">&#39;s3&#39;</span><span class="p">,</span> <span class="s1">&#39;s4&#39;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fracIdentWithinBarcode</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">library_col</span><span class="o">=</span><span class="s1">&#39;library&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">barcode_col</span><span class="o">=</span><span class="s1">&#39;bc&#39;</span><span class="p">,</span> <span class="n">variant_col</span><span class="o">=</span><span class="s1">&#39;var&#39;</span><span class="p">)</span>
<span class="go">  library  fraction_identical  accuracy</span>
<span class="go">0      s1            1.000000   1.00000</span>
<span class="go">1      s3            0.333333   0.57735</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.barcodes.simpleConsensus">
<code class="descclassname">dms_tools2.barcodes.</code><code class="descname">simpleConsensus</code><span class="sig-paren">(</span><em>df</em>, <em>*</em>, <em>barcode_col='barcode'</em>, <em>substitution_col='substitutions'</em>, <em>insertion_col='insertions'</em>, <em>deletion_col='deletions'</em>, <em>library_col=None</em>, <em>max_sub_diffs=1</em>, <em>max_indel_diffs=2</em>, <em>max_minor_muts=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/barcodes.html#simpleConsensus"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.barcodes.simpleConsensus" title="Permalink to this definition">¶</a></dt>
<dd><p>Simple method to get consensus of mutations within barcode.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>df</cite> (pandas Data Frame)</dt>
<dd>Holds variants and their barcodes. Each row gives a
sequence variant and its barcode. There need to be
columns with the names given by the next four arguments
described below.</dd>
<dt><cite>barcode_col</cite> (str)</dt>
<dd>Name of column holding barcodes.</dd>
<dt><cite>substitution_col</cite> (str)</dt>
<dd>Name of column holding substitutions as list of strings.</dd>
<dt><cite>insertion_col</cite> (str)</dt>
<dd>Name of column holding insertions as list of strings.</dd>
<dt><cite>deletion_col</cite> (str)</dt>
<dd>Name of column holding insertions as list of strings.</dd>
<dt><cite>library_col</cite> (<cite>None</cite> or str)</dt>
<dd>If we have multiple libraries, analyze each barcode only
within its library. In that case, <cite>library_col</cite> should be
name of column giving library name.</dd>
<dt><cite>max_sub_diffs</cite> (int)</dt>
<dd>Drop any barcode where any variant differs from all other
variants for that barcode by more than this many substitution
(point mutation) differences.</dd>
<dt><cite>max_indel_diffs</cite> (int)</dt>
<dd>Drop any barcode where any variant differs from all other
variants for that barcode by more than this many indel
(insertion or deletion) differences.</dd>
<dt><cite>max_minor_muts</cite> (int)</dt>
<dd>Drop any barcode where there is a minor (non-consensus)
mutation found more than this many times.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd><p class="first">The 2-tuple <cite>(consensus, dropped)</cite>. These are each data frames:</p>
<blockquote class="last">
<div><ul class="simple">
<li><cite>consensus</cite> is a new data frame with a row for each
barcode for which we could call the consensus. The
columns have the same names as <cite>barcode_col</cite>,
<cite>substitution_col</cite>, <cite>insertion_col</cite>, <cite>deletion_col</cite>,
and (optionally) <cite>library_col</cite>–but the the three
columns for the mutations now just list the <strong>consensus</strong>
mutations of that type. In addition, there is a new
column called “variant_call_support” that gives the number
of sequences supporting the call of that barcode.</li>
<li><cite>dropped</cite> simply contains all rows in the original <cite>df</cite>
that correspond to sequences that were dropped due to
<cite>max_diffs</cite> or <cite>max_minor_muts</cite> not being satisfied.
There is also a column called “drop_reason” that gives
the reason that the barcode was dropped.</li>
</ul>
</div></blockquote>
</dd>
</dl>
<p>The approach is as follows:</p>
<blockquote>
<div><ol class="arabic simple">
<li>Group all variants by library and barcode.</li>
<li>If there are multiple sequences, check if any of them differ
from all the others by more than <cite>max_diffs</cite> mutations total
(taking substitutions, insertions, and deletions together).
If so, drop the entire barcode. The reason is that if there
are variants that are very different, it becomes likely that
it isn’t just sequencing error, but rather something is wrong
with that barcode (multiple variants with same barcode or
strand exchange).</li>
<li>Take the consensus of the sequences, which means keeping
mutations that are present in <strong>greater</strong> than half of the
variants. Note that this calling scheme means that the
consensus being called is dependent on the reference used
to call the mutations, which is an important caveat if you
are calling variants relative to multiple different parent
sequences.</li>
<li>If there are any minor mutations (mutations not in consensus)
that are present in more than <cite>max_minor_muts</cite> variants or missing
from more than <cite>max_minor_muts</cite> variants, then
drop that barcode. The reason is that recurring minor mutations
also suggest some problem more complex than sequencing error
that may render the whole barcode family invalid.</li>
</ol>
</div></blockquote>
<p>Note that this method returns a consensus even if there is just
one sequence for the barcode (in that case, this sequence is
the consensus). This is fine–if you want to get consensus calls
that are more strongly supported, simply filter the returned
<cite>consensus</cite> data frame for larger values of <cite>variant_call_support</cite>,
as the more sequences that support a call the more accurate it
is expected to be.</p>
<p>Here is an example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([</span>
<span class="gp">... </span>    <span class="p">(</span><span class="s1">&#39;s1&#39;</span><span class="p">,</span> <span class="s1">&#39;AG&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;A2C&#39;</span><span class="p">],</span> <span class="p">[],</span> <span class="p">[</span><span class="s1">&#39;del5to7&#39;</span><span class="p">]),</span>
<span class="gp">... </span>    <span class="p">(</span><span class="s1">&#39;s1&#39;</span><span class="p">,</span> <span class="s1">&#39;AG&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;A2C&#39;</span><span class="p">],</span> <span class="p">[],</span> <span class="p">[]),</span>
<span class="gp">... </span>    <span class="p">(</span><span class="s1">&#39;s1&#39;</span><span class="p">,</span> <span class="s1">&#39;TA&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;G3A&#39;</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;ins4len3&#39;</span><span class="p">],</span> <span class="p">[]),</span>
<span class="gp">... </span>    <span class="p">(</span><span class="s1">&#39;s2&#39;</span><span class="p">,</span> <span class="s1">&#39;TA&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;C5A&#39;</span><span class="p">,</span> <span class="s1">&#39;T6C&#39;</span><span class="p">],</span> <span class="p">[],</span> <span class="p">[]),</span>
<span class="gp">... </span>    <span class="p">(</span><span class="s1">&#39;s2&#39;</span><span class="p">,</span> <span class="s1">&#39;TA&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;T6C&#39;</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;ins5len1&#39;</span><span class="p">],</span> <span class="p">[]),</span>
<span class="gp">... </span>    <span class="p">(</span><span class="s1">&#39;s2&#39;</span><span class="p">,</span> <span class="s1">&#39;TA&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;T6C&#39;</span><span class="p">],</span> <span class="p">[],</span> <span class="p">[]),</span>
<span class="gp">... </span>    <span class="p">(</span><span class="s1">&#39;s2&#39;</span><span class="p">,</span> <span class="s1">&#39;TG&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;T6A&#39;</span><span class="p">],</span> <span class="p">[],</span> <span class="p">[]),</span>
<span class="gp">... </span>    <span class="p">(</span><span class="s1">&#39;s2&#39;</span><span class="p">,</span> <span class="s1">&#39;TG&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;A2G&#39;</span><span class="p">],</span> <span class="p">[],</span> <span class="p">[]),</span>
<span class="gp">... </span>    <span class="p">(</span><span class="s1">&#39;s2&#39;</span><span class="p">,</span> <span class="s1">&#39;GG&#39;</span><span class="p">,</span> <span class="p">[],</span> <span class="p">[],</span> <span class="p">[</span><span class="s1">&#39;del1to4&#39;</span><span class="p">]),</span>
<span class="gp">... </span>    <span class="p">(</span><span class="s1">&#39;s2&#39;</span><span class="p">,</span> <span class="s1">&#39;GG&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;A1C&#39;</span><span class="p">],</span> <span class="p">[],</span> <span class="p">[]),</span>
<span class="gp">... </span>    <span class="p">(</span><span class="s1">&#39;s2&#39;</span><span class="p">,</span> <span class="s1">&#39;AA&#39;</span><span class="p">,</span> <span class="p">[],</span> <span class="p">[],</span> <span class="p">[]),</span>
<span class="gp">... </span>    <span class="p">(</span><span class="s1">&#39;s2&#39;</span><span class="p">,</span> <span class="s1">&#39;AA&#39;</span><span class="p">,</span> <span class="p">[],</span> <span class="p">[],</span> <span class="p">[]),</span>
<span class="gp">... </span>    <span class="p">(</span><span class="s1">&#39;s2&#39;</span><span class="p">,</span> <span class="s1">&#39;AA&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;T6C&#39;</span><span class="p">],</span> <span class="p">[],</span> <span class="p">[]),</span>
<span class="gp">... </span>    <span class="p">(</span><span class="s1">&#39;s2&#39;</span><span class="p">,</span> <span class="s1">&#39;AA&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;T6C&#39;</span><span class="p">],</span> <span class="p">[],</span> <span class="p">[]),</span>
<span class="gp">... </span>    <span class="p">(</span><span class="s1">&#39;s3&#39;</span><span class="p">,</span> <span class="s1">&#39;AA&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;T6G&#39;</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;ins1len1&#39;</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;del1to2&#39;</span><span class="p">]),</span>
<span class="gp">... </span>    <span class="p">(</span><span class="s1">&#39;s3&#39;</span><span class="p">,</span> <span class="s1">&#39;AA&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;T6G&#39;</span><span class="p">],</span> <span class="p">[],</span> <span class="p">[</span><span class="s1">&#39;del5to7&#39;</span><span class="p">])],</span>
<span class="gp">... </span>    <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;substitutions&#39;</span><span class="p">,</span>
<span class="gp">... </span>             <span class="s1">&#39;insertions&#39;</span><span class="p">,</span> <span class="s1">&#39;deletions&#39;</span><span class="p">]</span>
<span class="gp">... </span>    <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">consensus</span><span class="p">,</span> <span class="n">dropped</span> <span class="o">=</span> <span class="n">simpleConsensus</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">library_col</span><span class="o">=</span><span class="s1">&#39;library&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pandas</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s1">&#39;display.max_columns&#39;</span><span class="p">,</span> <span class="mi">10</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pandas</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s1">&#39;display.width&#39;</span><span class="p">,</span> <span class="mi">500</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">consensus</span>
<span class="go">  library barcode substitutions  insertions deletions  variant_call_support</span>
<span class="go">0      s1      AG         [A2C]          []        []                     2</span>
<span class="go">1      s1      TA         [G3A]  [ins4len3]        []                     1</span>
<span class="go">2      s2      GG            []          []        []                     2</span>
<span class="go">3      s2      TA         [T6C]          []        []                     3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dropped</span>
<span class="go">  library barcode substitutions  insertions  deletions           drop_reason</span>
<span class="go">0      s2      TG         [T6A]          []         []  excess substitutions</span>
<span class="go">1      s2      TG         [A2G]          []         []  excess substitutions</span>
<span class="go">2      s2      AA            []          []         []     excess minor muts</span>
<span class="go">3      s2      AA            []          []         []     excess minor muts</span>
<span class="go">4      s2      AA         [T6C]          []         []     excess minor muts</span>
<span class="go">5      s2      AA         [T6C]          []         []     excess minor muts</span>
<span class="go">6      s3      AA         [T6G]  [ins1len1]  [del1to2]         excess indels</span>
<span class="go">7      s3      AA         [T6G]          []  [del5to7]         excess indels</span>
</pre></div>
</div>
</dd></dl>

</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">dms_tools2</h1>
    
  </a>
</p>



<p class="blurb">Tools for analyzing deep mutational scanning data.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=dms_tools2&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a class="badge" href="https://travis-ci.org/jbloomlab/dms_tools2">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="bcsubamp.html">Barcoded-subamplicon sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="prefs.html">Amino-acid preferences</a></li>
<li class="toctree-l1"><a class="reference internal" href="diffsel.html">Differential selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="fracsurvive.html">Fraction surviving</a></li>
<li class="toctree-l1"><a class="reference internal" href="programs.html">Executable programs</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="api.html">Python API</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="dms_tools2.html">dms_tools2 package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="citations.html">Citations</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="api.html">Python API</a><ul>
  <li><a href="dms_tools2.html">dms_tools2 package</a><ul>
      <li>Previous: <a href="dms_tools2.html" title="previous chapter">dms_tools2 package</a></li>
      <li>Next: <a href="dms_tools2.codonvarianttable.html" title="next chapter">codonvarianttable</a></li>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017--2019, the Bloom lab.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/dms_tools2.barcodes.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/dms_tools2" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>