
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>prefs &#8212; dms_tools2 2.0a0 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '2.0a0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="sra" href="dms_tools2.sra.html" />
    <link rel="prev" title="plot" href="dms_tools2.plot.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-dms_tools2.prefs"></span><div class="section" id="prefs">
<h1>prefs<a class="headerlink" href="#prefs" title="Permalink to this headline">¶</a></h1>
<p>Performs operations related to estimating site-specific amino-acid
preferences.</p>
<p>Uses <a class="reference external" href="https://pystan.readthedocs.io/en/latest">pystan</a>
to perform MCMC for Bayesian inferences.</p>
<dl class="data">
<dt id="dms_tools2.prefs.PRIOR_MIN_VALUE">
<code class="descclassname">dms_tools2.prefs.</code><code class="descname">PRIOR_MIN_VALUE</code><em class="property"> = 1e-07</em><a class="headerlink" href="#dms_tools2.prefs.PRIOR_MIN_VALUE" title="Permalink to this definition">¶</a></dt>
<dd><p>minimum value for Dirichlet prior elements</p>
</dd></dl>

<dl class="class">
<dt id="dms_tools2.prefs.StanModelDifferentErr">
<em class="property">class </em><code class="descclassname">dms_tools2.prefs.</code><code class="descname">StanModelDifferentErr</code><span class="sig-paren">(</span><em>verbose=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/prefs.html#StanModelDifferentErr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.prefs.StanModelDifferentErr" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p><code class="docutils literal"><span class="pre">pystan</span></code> model when <cite>error_model</cite> is <cite>different</cite>.</p>
<p>For use by inferSitePrefs`.</p>
</dd></dl>

<dl class="class">
<dt id="dms_tools2.prefs.StanModelNoneErr">
<em class="property">class </em><code class="descclassname">dms_tools2.prefs.</code><code class="descname">StanModelNoneErr</code><span class="sig-paren">(</span><em>verbose=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/prefs.html#StanModelNoneErr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.prefs.StanModelNoneErr" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p><code class="docutils literal"><span class="pre">pystan</span></code> model when <cite>error_model</cite> is <cite>none</cite>.</p>
<p>For use by inferSitePrefs`.</p>
</dd></dl>

<dl class="class">
<dt id="dms_tools2.prefs.StanModelSameErr">
<em class="property">class </em><code class="descclassname">dms_tools2.prefs.</code><code class="descname">StanModelSameErr</code><span class="sig-paren">(</span><em>verbose=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/prefs.html#StanModelSameErr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.prefs.StanModelSameErr" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p><code class="docutils literal"><span class="pre">pystan</span></code> model when <cite>error_model</cite> is <cite>same</cite>.</p>
<p>For use by inferSitePrefs`.</p>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.prefs.avgPrefs">
<code class="descclassname">dms_tools2.prefs.</code><code class="descname">avgPrefs</code><span class="sig-paren">(</span><em>prefsfiles</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/prefs.html#avgPrefs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.prefs.avgPrefs" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets average of site-specific preferences.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>prefsfiles</cite> (list)</dt>
<dd>List of CSV files containing preferences, must all be
for same sites and characters.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>A <cite>pandas.DataFrame</cite> containing the average of the
preferences in <cite>prefsfiles</cite>. In this returned
data frame, <cite>site</cite> is the index</dd>
</dl>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tf1</span> <span class="o">=</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tf2</span> <span class="o">=</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">tf1</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file1</span><span class="p">,</span> <span class="n">tf2</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file2</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">x</span> <span class="o">=</span> <span class="n">file1</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;site,A,C,G,T</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>                <span class="s1">&#39;1,0.2,0.2,0.5,0.1</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>                <span class="s1">&#39;2,0.3,0.3,0.3,0.1&#39;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">file1</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
<span class="gp">... </span>    <span class="n">x</span> <span class="o">=</span> <span class="n">file2</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;site,A,C,G,T</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>                <span class="s1">&#39;1,0.4,0.1,0.1,0.4</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>                <span class="s1">&#39;2,0.3,0.4,0.1,0.2&#39;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">file2</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
<span class="gp">... </span>    <span class="n">avg</span> <span class="o">=</span> <span class="n">avgPrefs</span><span class="p">([</span><span class="n">file1</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">file2</span><span class="o">.</span><span class="n">name</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="n">avg</span><span class="p">[</span><span class="s1">&#39;site&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">])</span><span class="o">.</span><span class="n">all</span><span class="p">()</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">avg</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">avg</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.15</span><span class="p">,</span> <span class="mf">0.35</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">avg</span><span class="p">[</span><span class="s1">&#39;G&#39;</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">avg</span><span class="p">[</span><span class="s1">&#39;T&#39;</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.15</span><span class="p">])</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.prefs.inferPrefsByRatio">
<code class="descclassname">dms_tools2.prefs.</code><code class="descname">inferPrefsByRatio</code><span class="sig-paren">(</span><em>charlist</em>, <em>sites</em>, <em>wts</em>, <em>pre</em>, <em>post</em>, <em>errpre</em>, <em>errpost</em>, <em>pseudocount</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/prefs.html#inferPrefsByRatio"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.prefs.inferPrefsByRatio" title="Permalink to this definition">¶</a></dt>
<dd><p>Site-specific preferences from normalized enrichment ratios.</p>
<p>Calculates site-specific preference <span class="math">\(\pi_{r,a}\)</span> of each
site <span class="math">\(r\)</span> for each character <span class="math">\(a\)</span> using re-normalized
enrichment ratios. This function accomplishes the same goal as
<cite>inferSitePrefs</cite>, but is much simpler and faster, although less
statistically rigorous in principle.</p>
<p>Specifically, this function calculates the preferences as</p>
<div class="math">
\[\pi_{r,a} = \frac{\phi_{r,a}}{\sum_{a'} \phi_{r,a'}}\]</div>
<p>where <span class="math">\(\phi_{r,a}\)</span> is the enrichment ratio of character
<span class="math">\(a\)</span> relative to wildtype after selection.</p>
<p>The actual definition of these enrichment ratios is somewhat
complex due to the need to correct for errors and add a
pseudocount. We have 4 sets of counts:</p>
<blockquote>
<div><ul class="simple">
<li><cite>pre</cite>: pre-selection counts</li>
<li><cite>post</cite>: post-selection counts</li>
<li><cite>errpre</cite>: error-control for pre-selection counts</li>
<li><cite>errpost</cite>: error-control for post-selection counts</li>
</ul>
</div></blockquote>
<p>For each set of counts <span class="math">\(s\)</span>, let <span class="math">\(N_r^s\)</span> (e.g., 
<span class="math">\(N_r^{pre}\)</span>) denote the total counts for this sample at
site <span class="math">\(r\)</span>, and let <span class="math">\(n_{r,a}^{s}\)</span> (e.g., 
<span class="math">\(n_{r,a}^{pre}\)</span>) denote the counts at that site for
character <span class="math">\(a\)</span>.</p>
<p>Because some of the counts are low, we add a pseudocount 
<span class="math">\(P\)</span> to each observation. With this pseudocount, the 
estimated frequency of character <span class="math">\(a\)</span> at site <span class="math">\(r\)</span>
in sample <span class="math">\(s\)</span> is</p>
<div class="math">
\[f_{r,a}^s = \frac{n_{r,a}^s + P}{N_{r,a}^s + A \times P}\]</div>
<p>where <span class="math">\(A\)</span> is the number of characters in the alphabet (e.g.,
20 for amino acids without stop codons).</p>
<p>The <strong>error-corrected</strong> estimates of the frequency of character
<span class="math">\(a\)</span> before and after selection are then</p>
<div class="math">
\[ \begin{align}\begin{aligned}f_{r,a}^{before} &amp;= \max\left(\frac{P}{N_{r,a}^{pre} + A \times P},
\; f_{r,a}^{pre} + \delta_{a,\rm{wt}\left(r\right)}
- f_{r,a}^{errpre}\right)\\f_{r,a}^{after} &amp;= \max\left(\frac{P}{N_{r,a}^{post} + A \times P},
\; f_{r,a}^{post} + \delta_{a,\rm{wt}\left(r\right)}
- f_{r,a}^{errpost}\right)\end{aligned}\end{align} \]</div>
<p>where <span class="math">\(\delta_{a,\rm{wt}\left(r\right)}\)</span> is the 
Kronecker-delta, equal to 1 if <span class="math">\(a\)</span> is the same as the
wildtype character <span class="math">\(\rm{wt}\left(r\right)\)</span> at site
<span class="math">\(r\)</span>, and 0 otherwise. We use the <span class="math">\(\max\)</span> operator
to ensure that even if the error-control frequency exceeds
the actual estimate, our estimated frequency never drops
below the pseudocount over the depth of the uncorrected sample.</p>
<p>Given these definitions, we then simply calculate the enrichment 
ratios as</p>
<div class="math">
\[\phi_{r,a} = \frac{\left(f_{r,a}^{after}\right) / 
\left(f_{r,\rm{wt}\left(r\right)}^{after}\right)}
{\left(f_{r,a}^{before}\right) / 
\left(f_{r,\rm{wt}\left(r\right)}^{before}\right)}\]</div>
<p>In the case where we are <strong>not</strong> using any error-controls, then
we simply set 
<span class="math">\(f_{r,\rm{wt}}^{errpre} = f_{r,\rm{wt}}^{errpost} 
= \delta_{a,\rm{wt}\left(r\right)}\)</span>.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>charlist</cite> (list)</dt>
<dd>Valid characters (e.g., codons, amino acids, nts).</dd>
<dt><cite>sites</cite> (list)</dt>
<dd>Sites to analyze.</dd>
<dt><cite>wts</cite> (list)</dt>
<dd><cite>wts[r]</cite> is the wildtype character at site <cite>sites[r]</cite>.</dd>
<dt><cite>pre</cite> (pandas.DataFrame)</dt>
<dd>Gives pre-selection counts. Should have columns
with names of ‘site’ and all characters in <cite>charlist</cite>.
The rows give the counts of each character at that site.</dd>
<dt><cite>post</cite> (pandas.DataFrame)</dt>
<dd>Like <cite>pre</cite> but for post-selection counts.</dd>
<dt><cite>errpre</cite> (<cite>None</cite> or pandas.DataFrame)</dt>
<dd>Like <cite>pre</cite> but for pre-selection error-control counts,
or <cite>None</cite> if there is no such control.</dd>
<dt><cite>errpost</cite> (<cite>None</cite> or pandas.DataFrame)</dt>
<dd>Like <cite>pre</cite> but for post-selection error-control counts,
or <cite>None</cite> if there is no such control.</dd>
<dt><cite>pseudocount</cite> (float or int)</dt>
<dd>The pseudocount to add to each observation.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>A pandas.DataFrame holding the preferences. The columns of
this dataframe are ‘site’ and all characters in <cite>charlist</cite>.
For each site in <cite>sites</cite>, the rows give the preference
for that character.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.prefs.inferSitePrefs">
<code class="descclassname">dms_tools2.prefs.</code><code class="descname">inferSitePrefs</code><span class="sig-paren">(</span><em>charlist</em>, <em>wtchar</em>, <em>error_model</em>, <em>counts</em>, <em>priors</em>, <em>seed=1</em>, <em>niter=10000</em>, <em>increasetries=5</em>, <em>n_jobs=1</em>, <em>r_max=1.1</em>, <em>neff_min=100</em>, <em>nchains=4</em>, <em>increasefac=2</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/prefs.html#inferSitePrefs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.prefs.inferSitePrefs" title="Permalink to this definition">¶</a></dt>
<dd><p>Infers site-specific preferences by MCMC for a specific site.</p>
<p>Infer the site-specific preferences <span class="math">\(\pi_{r,a}\)</span> for some site
<span class="math">\(r\)</span> for each character <span class="math">\(a\)</span> by integrating over the 
posterior defined by the product of the following priors and 
likelihoods, where <span class="math">\(\boldsymbol{\mathbf{\pi_r}}\)</span> indicates
the vector of <span class="math">\(\pi_{r,a}\)</span> values for all characters:</p>
<div class="math">
\[ \begin{align}\begin{aligned}\Pr\left(\boldsymbol{\mathbf{\pi_r}}\right) 
&amp;= \mathrm{Dirichlet}\left(\boldsymbol{\mathbf{a_{\pi,r}}}\right)\\\Pr\left(\boldsymbol{\mathbf{\mu_r}}\right) 
&amp;= \mathrm{Dirichlet}\left(\boldsymbol{\mathbf{a_{\mu,r}}}\right)\\\Pr\left(\boldsymbol{\mathbf{\epsilon_r}}\right) 
&amp;= \mathrm{Dirichlet}\left(\boldsymbol{\mathbf{a_{\epsilon,r}}}\right)\\\Pr\left(\boldsymbol{\mathbf{\rho_r}}\right) 
&amp;= \mathrm{Dirichlet}\left(\boldsymbol{\mathbf{a_{\rho,r}}}\right)\\\Pr\left(\boldsymbol{\mathbf{n_r^{\rm{pre}}}} \mid \boldsymbol{\mathbf{\mu_r}}, \boldsymbol{\mathbf{\epsilon_r}}\right) 
&amp;= \mathrm{Multinomial}\left(\boldsymbol{\mathbf{n_r^{\rm{pre}}}}; \boldsymbol{\mathbf{\mu_r}} + \boldsymbol{\mathbf{\epsilon_r}} - \boldsymbol{\mathbf{\delta_r}}\right)\\\Pr\left(\boldsymbol{\mathbf{n_r^{\rm{post}}}} \mid \boldsymbol{\mathbf{\mu_r}}, \boldsymbol{\mathbf{\epsilon_r}}\right) 
&amp;= \mathrm{Multinomial}\left(\boldsymbol{\mathbf{n_r^{\rm{post}}}}; \frac{\boldsymbol{\mathbf{\mu_r}} \circ \boldsymbol{\mathbf{\pi_r}}}{\boldsymbol{\mathbf{\mu_r}} \cdot \boldsymbol{\mathbf{\pi_r}}} + \boldsymbol{\mathbf{\rho_r}} - \boldsymbol{\mathbf{\delta_r}}\right)\\\Pr\left(\boldsymbol{\mathbf{n_r^{\rm{err,pre}}}} \mid \boldsymbol{\mathbf{\epsilon_r}}\right) 
&amp;= \mathrm{Multinomial}\left(\boldsymbol{\mathbf{n_r^{\rm{err,pre}}}}; \boldsymbol{\mathbf{\epsilon_r}}\right)\\\Pr\left(\boldsymbol{\mathbf{n_r^{\rm{err,post}}}} \mid \boldsymbol{\mathbf{\rho_r}}\right) 
&amp;= \mathrm{Multinomial}\left(\boldsymbol{\mathbf{n_r^{\rm{err,post}}}}; \boldsymbol{\mathbf{\rho_r}}\right)\end{aligned}\end{align} \]</div>
<p>where <span class="math">\(\boldsymbol{\mathbf{\delta_r}}\)</span> is a vector that is 
zero except for a one at the element corresponding to the wildtype.</p>
<p>The MCMC tries to guarantee convergence via the parameters specified
by <cite>r_max</cite>, <cite>neff_min</cite>, <cite>nchains</cite>, <cite>niter</cite>, <cite>increasefac</cite>, 
and <cite>increasetries</cite>.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>charlist</cite> (list)</dt>
<dd>List of valid characters (e.g., codons, amino acids, nts).</dd>
<dt><cite>wtchar</cite> (str)</dt>
<dd>Wildtype character at the site.</dd>
<dt><cite>error_model</cite> (str or object)</dt>
<dd><p class="first">Specifies how errors are estimated. Passing error model
objects is faster if calling this function repeatedly
as they will not need to be compiled. Can be:</p>
<blockquote class="last">
<div><ul class="simple">
<li>The str <cite>none</cite> or instance of <cite>StanModelNoneErr</cite>:
no errors (<span class="math">\(\boldsymbol{\mathbf{\epsilon_r}} = \boldsymbol{\mathbf{\rho_r}} = \mathbf{0}\)</span>)</li>
<li>The str <cite>same</cite> or instance of <cite>StanModelSameErr</cite>: 
same error rates  pre and post-selection 
(<span class="math">\(\boldsymbol{\mathbf{\epsilon_r}} = \boldsymbol{\mathbf{\rho_r}}\)</span>).</li>
<li>The str <cite>different</cite> or instance of <cite>StanModelDifferentErr</cite>: 
different error rates pre- and post-selection
(<span class="math">\(\boldsymbol{\mathbf{\epsilon_r}} \ne \boldsymbol{\mathbf{\rho_r}}\)</span>).</li>
</ul>
</div></blockquote>
</dd>
<dt><cite>counts</cite> (dict)</dt>
<dd><p class="first">Deep sequencing counts. Each string key should specify
a dict keyed by all characters in <cite>charlist</cite> with
the values giving integer counts for that character. The keys:</p>
<blockquote class="last">
<div><ul class="simple">
<li><cite>pre</cite>: <span class="math">\(\boldsymbol{\mathbf{n_r^{\rm{pre}}}}\)</span></li>
<li><cite>post</cite>: <span class="math">\(\boldsymbol{\mathbf{n_r^{\rm{post}}}}\)</span></li>
<li><em>err</em>: required if <cite>error_model</cite> is <cite>same</cite>, specifies
<span class="math">\(\boldsymbol{\mathbf{n_r^{\rm{err,pre}}}} = \boldsymbol{\mathbf{n_r^{\rm{err,post}}}}\)</span></li>
<li><cite>errpre</cite> and <cite>errpost</cite>: required if <cite>error_model</cite> is
<cite>different</cite>, specify 
<span class="math">\(\boldsymbol{\mathbf{n_r^{\rm{err,pre}}}}\)</span> and
<span class="math">\(\boldsymbol{\mathbf{n_r^{\rm{err,post}}}}\)</span>.</li>
</ul>
</div></blockquote>
</dd>
<dt><cite>priors</cite> (dict)</dt>
<dd><p class="first">Specifies parameter vectors for Dirichlet priors. Each string
key should specify a dict keyed by all characters and values
giving the prior for that character. Values less than 
<cite>PRIOR_MIN_VALUE</cite> are set to <cite>PRIOR_MIN_VALUE</cite>. Keys are</p>
<blockquote class="last">
<div><ul class="simple">
<li><cite>pir_prior_params</cite>: <span class="math">\(\boldsymbol{\mathbf{a_{\pi,r}}}\)</span></li>
<li><cite>mur_prior_params</cite>: <span class="math">\(\boldsymbol{\mathbf{a_{\mu,r}}}\)</span></li>
<li><cite>epsilonr_prior_params</cite>: only required if <cite>error_model</cite> is 
<cite>same</cite> or <cite>different</cite>, specifies 
<span class="math">\(\boldsymbol{\mathbf{a_{\epsilon,r}}}\)</span></li>
<li><cite>rhor_prior_params</cite> : only required if <cite>error_model</cite> is 
<cite>different</cite>, specifies
<span class="math">\(\boldsymbol{\mathbf{a_{\rho,r}}}\)</span></li>
</ul>
</div></blockquote>
</dd>
<dt><cite>seed</cite> (int)</dt>
<dd>Random number seed for MCMC.</dd>
<dt><cite>n_jobs</cite> (int)</dt>
<dd>Number of CPUs to use, -1 means all available.</dd>
<dt><cite>niter</cite>, <cite>increasetries</cite>, <cite>r_max</cite>, <cite>neff_min</cite>, <cite>nchains</cite>, <cite>increasefac</cite></dt>
<dd>Specify MCMC convergence. They all have reasonable defaults.
The MCMC is considered to have converged if the mean Gelman-Rubin R
statistic (<a class="reference external" href="http://www.jstor.org/stable/2246093">http://www.jstor.org/stable/2246093</a>) over all 
<span class="math">\(\pi_{r,x}\)</span> values &lt;= <cite>r_max</cite> and the mean effective sample
size is &gt;= <cite>neff_min</cite>. The MCMC first runs with <cite>nchains</cite> chains
and <cite>niter</cite> iterations. If it fails to converge, it increases
the iterations by a factor of <cite>increasefac</cite> and tries again,
and repeats until it converges or has tried <cite>increasetries</cite>
times to increase the number of iterations. If the effective
sample size exceeds 3 times <cite>neff_min</cite> then we allow
R to be <cite>1 + 1.5 (r_max - 1)</cite>.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd><dl class="first last docutils">
<dt>The tuple <cite>(converged, pi_means, pi_95credint, logstring)</cite> where:</dt>
<dd><ul class="first last simple">
<li><cite>converged</cite> is <cite>True</cite> if the MCMC converged, <cite>False</cite> otherwise.</li>
<li><dl class="first docutils">
<dt><cite>pi_means</cite> is dict keyed by characters in <cite>charlist</cite></dt>
<dd>with the value giving the <span class="math">\(\pi_{r,a}\)</span>.</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><cite>pi_95credint</cite> is dict keyed by characters, values are 2-tuples</dt>
<dd>giving median-centered credible interval for <span class="math">\(\pi_{r,a}\)</span>.</dd>
</dl>
</li>
<li><cite>logstring</cite> is a string describing MCMC run and convergence.</li>
</ul>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.prefs.prefsEntropy">
<code class="descclassname">dms_tools2.prefs.</code><code class="descname">prefsEntropy</code><span class="sig-paren">(</span><em>prefs</em>, <em>charlist</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/prefs.html#prefsEntropy"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.prefs.prefsEntropy" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate site entropy and number of effective characters.</p>
<p>The site entropy <span class="math">\(h_r\)</span> for site <span class="math">\(r\)</span> is defined as
<span class="math">\(h_r = -\sum_{x} \pi_{r,x} \log\left(\pi_{r,x}\right)\)</span>
where <span class="math">\(\pi_{r,x}\)</span> is the preference for character (e.g.,
amino acid) <span class="math">\(x\)</span> at site <span class="math">\(r\)</span>, and the log is
the natural logarithm.</p>
<p>The number of effective characters at site <span class="math">\(r\)</span>
is <span class="math">\(N_{\rm{eff}, r} = \exp\left(h_r\right)\)</span>.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>prefs</cite> (pandas DataFrame)</dt>
<dd>Data frame with the preferences. Should have a column
named <cite>site</cite> and a column for each character in
<cite>charlist</cite>.</dd>
<dt><cite>charlist</cite> (list)</dt>
<dd>List of the characters of interest, for example
the amino acids.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>A copy of <cite>prefs</cite> that has additional columns named
<cite>entropy</cite> and <cite>neffective</cite>, giving the site entropy and
number of effective characters for each site. For
the entropies, log is taken to base <span class="math">\(e\)</span>.</dd>
</dl>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">charlist</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prefs</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
<span class="gp">... </span>        <span class="s1">&#39;site&#39;</span><span class="p">:[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;A&#39;</span><span class="p">:[</span><span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">,</span> <span class="mi">1</span> <span class="o">/</span> <span class="mf">3.</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;C&#39;</span><span class="p">:[</span><span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">,</span> <span class="mi">1</span> <span class="o">/</span> <span class="mf">3.</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;G&#39;</span><span class="p">:[</span><span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mi">1</span> <span class="o">/</span> <span class="mf">3.</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;T&#39;</span><span class="p">:[</span><span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">],</span>
<span class="gp">... </span>        <span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prefs_entropy</span> <span class="o">=</span> <span class="n">prefsEntropy</span><span class="p">(</span><span class="n">prefs</span><span class="p">,</span> <span class="n">charlist</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="nb">set</span><span class="p">([</span><span class="s1">&#39;entropy&#39;</span><span class="p">,</span> <span class="s1">&#39;neffective&#39;</span><span class="p">,</span> <span class="s1">&#39;site&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">charlist</span><span class="p">)</span> <span class="o">==</span> 
<span class="gp">... </span>        <span class="nb">set</span><span class="p">(</span><span class="n">prefs_entropy</span><span class="o">.</span><span class="n">columns</span><span class="p">))</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">h2</span> <span class="o">=</span> <span class="o">-</span><span class="mf">0.6</span> <span class="o">*</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mf">0.6</span><span class="p">)</span> <span class="o">-</span> <span class="mf">0.3</span> <span class="o">*</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mf">0.3</span><span class="p">)</span> <span class="o">-</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mf">0.1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">prefs_entropy</span><span class="p">[</span><span class="s1">&#39;entropy&#39;</span><span class="p">],</span> <span class="p">[</span><span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mi">4</span><span class="p">),</span> <span class="n">h2</span><span class="p">,</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mi">3</span><span class="p">)])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">prefs_entropy</span><span class="p">[</span><span class="s1">&#39;neffective&#39;</span><span class="p">],</span> <span class="p">[</span><span class="mi">4</span><span class="p">,</span> <span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="n">h2</span><span class="p">),</span> <span class="mi">3</span><span class="p">])</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.prefs.prefsToMutEffects">
<code class="descclassname">dms_tools2.prefs.</code><code class="descname">prefsToMutEffects</code><span class="sig-paren">(</span><em>prefs</em>, <em>charlist</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/prefs.html#prefsToMutEffects"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.prefs.prefsToMutEffects" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts amino acid preferences to effects of specific mutations.</p>
<p>If the preference of site <span class="math">\(r\)</span> for amino acid <span class="math">\(a\)</span> is
<span class="math">\(\pi_{r,a}\)</span>, then the estimated effect (e.g., ratio of
enrichment ratios) for mutating that site from <span class="math">\(x\)</span> to
<span class="math">\(y\)</span> is <span class="math">\(\frac{\pi_{r,y}}{\pi_{r,x}}\)</span>. Very small
values indicate disfavored mutations, and very large values
indicate highly favored mutations. The logarithm base 2 of the
expected effects is also a useful measure – negative values
are disfavored mutations, positive values are favored ones.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>prefs</cite> (pandas DataFrame)</dt>
<dd>Preferences to analyze. The columns should be <cite>site</cite> and
a column for every character in <cite>charlist</cite>.</dd>
<dt><cite>charlist</cite> (list)</dt>
<dd>The list of characters that we are analyzing. For instance,
<cite>dms_tools2.AAS</cite> for amino acids.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd><p class="first">A pandas Data Frame where the columns are:</p>
<blockquote class="last">
<div><ul class="simple">
<li><cite>site</cite>: the site</li>
<li><cite>initial</cite>: the initial character (e.g., amino acid)</li>
<li><cite>final</cite>: the final character after the mutation</li>
<li><cite>mutation</cite>: mutation in the string form <cite>A1G</cite></li>
<li><cite>effect</cite>: the effect of the mutation</li>
<li><cite>log2effect</cite>: the log2 of the effect of the mutation.</li>
</ul>
</div></blockquote>
</dd>
</dl>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">charlist</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prefs</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
<span class="gp">... </span>        <span class="s1">&#39;site&#39;</span><span class="p">:[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;A&#39;</span><span class="p">:[</span><span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;C&#39;</span><span class="p">:[</span><span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;G&#39;</span><span class="p">:[</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">],</span>
<span class="gp">... </span>        <span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">effects</span> <span class="o">=</span> <span class="n">prefsToMutEffects</span><span class="p">(</span><span class="n">prefs</span><span class="p">,</span> <span class="n">charlist</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">set</span><span class="p">(</span><span class="n">effects</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">==</span> <span class="p">{</span><span class="s1">&#39;site&#39;</span><span class="p">,</span> <span class="s1">&#39;initial&#39;</span><span class="p">,</span> <span class="s1">&#39;final&#39;</span><span class="p">,</span>
<span class="gp">... </span>        <span class="s1">&#39;mutation&#39;</span><span class="p">,</span> <span class="s1">&#39;effect&#39;</span><span class="p">,</span> <span class="s1">&#39;log2effect&#39;</span><span class="p">}</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">effects</span><span class="p">[</span><span class="n">effects</span><span class="p">[</span><span class="s1">&#39;initial&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;A&#39;</span><span class="p">][</span><span class="s1">&#39;effect&#39;</span><span class="p">],</span>
<span class="gp">... </span>        <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">effects</span><span class="p">[</span><span class="n">effects</span><span class="p">[</span><span class="s1">&#39;initial&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;C&#39;</span><span class="p">][</span><span class="s1">&#39;effect&#39;</span><span class="p">],</span>
<span class="gp">... </span>        <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">effects</span><span class="p">[</span><span class="s1">&#39;effect&#39;</span><span class="p">],</span> <span class="mi">2</span><span class="o">**</span><span class="n">effects</span><span class="p">[</span><span class="s1">&#39;log2effect&#39;</span><span class="p">])</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.prefs.rescalePrefs">
<code class="descclassname">dms_tools2.prefs.</code><code class="descname">rescalePrefs</code><span class="sig-paren">(</span><em>prefs</em>, <em>stringency</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/prefs.html#rescalePrefs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.prefs.rescalePrefs" title="Permalink to this definition">¶</a></dt>
<dd><p>Re-scale amino acid preferences by stringency parameter.</p>
<p>If the initial preference of site <span class="math">\(r\)</span> for amino-acid 
<span class="math">\(a\)</span> is <span class="math">\(\pi_{r,a}\)</span>, then the re-scaled preference is</p>
<div class="math">
\[\frac{\left(\pi_{r,a}\right)^{\beta}}{\sum_{a'} \left(\pi_{r,a'}\right)}\]</div>
<p>where <span class="math">\(\beta\)</span> is the stringency parameter.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>prefs</cite> (pandas.DataFrame)</dt>
<dd>Columns are ‘site’ and then every character (e.g.,
amino acid).</dd>
<dt><cite>stringency</cite> (float &gt;= 0)</dt>
<dd>The stringency parameter.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>A data frame in the same format as <cite>prefs</cite> but with the
re-scaled preferences.</dd>
</dl>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">prefs</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="nb">dict</span><span class="p">(</span>
<span class="gp">... </span>        <span class="p">[(</span><span class="s1">&#39;site&#39;</span><span class="p">,</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">]),</span> <span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="p">[</span><span class="mf">0.24</span><span class="p">,</span> <span class="mf">0.03</span><span class="p">]),</span> <span class="p">(</span><span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="p">[</span><span class="mf">0.04</span><span class="p">,</span> <span class="mf">0.43</span><span class="p">])]</span>
<span class="gp">... </span>        <span class="o">+</span> <span class="p">[(</span><span class="n">aa</span><span class="p">,</span> <span class="p">[</span><span class="mf">0.04</span><span class="p">,</span> <span class="mf">0.03</span><span class="p">])</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">dms_tools2</span><span class="o">.</span><span class="n">AAS</span> <span class="k">if</span>
<span class="gp">... </span>        <span class="n">aa</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">]]))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">prefs</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s1">&#39;site&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rescaled</span> <span class="o">=</span> <span class="n">rescalePrefs</span><span class="p">(</span><span class="n">prefs</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">([</span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">prefs</span><span class="p">[</span><span class="n">c</span><span class="p">],</span> <span class="n">rescaled</span><span class="p">[</span><span class="n">c</span><span class="p">])</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">prefs</span><span class="o">.</span><span class="n">columns</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rescaled2</span> <span class="o">=</span> <span class="n">rescalePrefs</span><span class="p">(</span><span class="n">prefs</span><span class="p">,</span> <span class="mf">2.0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">rescaled2</span><span class="p">[</span><span class="s1">&#39;site&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">prefs</span><span class="p">[</span><span class="s1">&#39;site&#39;</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">rescaled2</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.6545</span><span class="p">,</span> <span class="mf">0.0045</span><span class="p">],</span> <span class="n">atol</span><span class="o">=</span><span class="mf">1e-3</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">rescaled2</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.0182</span><span class="p">,</span> <span class="mf">0.9153</span><span class="p">],</span> <span class="n">atol</span><span class="o">=</span><span class="mf">1e-3</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">dms_tools2</h1>
    
  </a>
</p>



<p class="blurb">Tools for analyzing deep mutational scanning data.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=dms_tools2&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a href="https://travis-ci.org/jbloomlab/dms_tools2">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="bcsubamp.html">Barcoded-subamplicon sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="prefs.html">Amino-acid preferences</a></li>
<li class="toctree-l1"><a class="reference internal" href="diffsel.html">Differential selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="programs.html">Executable programs</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="api.html">Python API</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="dms_tools2.html">dms_tools2 package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="citations.html">Citations</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="api.html">Python API</a><ul>
  <li><a href="dms_tools2.html">dms_tools2 package</a><ul>
      <li>Previous: <a href="dms_tools2.plot.html" title="previous chapter">plot</a></li>
      <li>Next: <a href="dms_tools2.sra.html" title="next chapter">sra</a></li>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017, the Bloom lab.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.6.3</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
      |
      <a href="_sources/dms_tools2.prefs.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/dms_tools2" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>