
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>utils &#8212; dms_tools2 2.6.5 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Citations" href="citations.html" />
    <link rel="prev" title="syn_selection" href="dms_tools2.syn_selection.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <span class="target" id="module-dms_tools2.utils"></span><div class="section" id="utils">
<h1>utils<a class="headerlink" href="#utils" title="Permalink to this headline">¶</a></h1>
<p>Miscellaneous utilities for <code class="docutils literal notranslate"><span class="pre">dms_tools2</span></code>.</p>
<dl class="function">
<dt id="dms_tools2.utils.adjustErrorCounts">
<code class="sig-prename descclassname">dms_tools2.utils.</code><code class="sig-name descname">adjustErrorCounts</code><span class="sig-paren">(</span><em class="sig-param">errcounts</em>, <em class="sig-param">counts</em>, <em class="sig-param">charlist</em>, <em class="sig-param">maxexcess</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/utils.html#adjustErrorCounts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.utils.adjustErrorCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Adjust error counts to not greatly exceed counts of interest.</p>
<p>This function is useful when estimating preferences. Under the
model, the error-control should not have a higher rate of error
than the actual sample. However, this could happen if the experimental
data don’t fully meet the assumptions. So this function scales
down the error counts in that case.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>errcounts</cite> (pandas.DataFrame)</dt><dd><p>Holds counts for error control.</p>
</dd>
<dt><cite>counts</cite> (pandas.DataFrame)</dt><dd><p>Holds counts for which we are correcting errors.</p>
</dd>
<dt><cite>charlist</cite> (list)</dt><dd><p>Characters for which we have counts.</p>
</dd>
<dt><cite>maxexcess</cite> (int)</dt><dd><p>Only let error-control counts exceed actual by this much.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>A copy of <cite>errcounts</cite> except for any non-wildtype character,
the maximum frequency of that character is adjusted to be
at most the number predicted by the frequency in <cite>counts</cite>
plus <cite>maxexcess</cite>.</p>
</dd>
</dl>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;site&#39;</span><span class="p">:[</span><span class="mi">1</span><span class="p">],</span> <span class="s1">&#39;wildtype&#39;</span><span class="p">:[</span><span class="s1">&#39;A&#39;</span><span class="p">],</span> 
<span class="gp">... </span>        <span class="s1">&#39;A&#39;</span><span class="p">:</span><span class="mi">500</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">:</span><span class="mi">10</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">:</span><span class="mi">40</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">:</span><span class="mi">20</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">errcounts</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;site&#39;</span><span class="p">:[</span><span class="mi">1</span><span class="p">],</span> <span class="s1">&#39;wildtype&#39;</span><span class="p">:[</span><span class="s1">&#39;A&#39;</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;A&#39;</span><span class="p">:</span><span class="mi">250</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">:</span><span class="mi">30</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">:</span><span class="mi">10</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">charlist</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">errcounts</span> <span class="o">=</span> <span class="n">errcounts</span><span class="p">[[</span><span class="s1">&#39;site&#39;</span><span class="p">,</span> <span class="s1">&#39;wildtype&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">charlist</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adj_errcounts</span> <span class="o">=</span> <span class="n">adjustErrorCounts</span><span class="p">(</span><span class="n">errcounts</span><span class="p">,</span> <span class="n">counts</span><span class="p">,</span> <span class="n">charlist</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">set</span><span class="p">(</span><span class="n">adj_errcounts</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">==</span> <span class="nb">set</span><span class="p">(</span><span class="n">errcounts</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">adj_errcounts</span><span class="p">[</span><span class="s1">&#39;site&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">errcounts</span><span class="p">[</span><span class="s1">&#39;site&#39;</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">adj_errcounts</span><span class="p">[</span><span class="s1">&#39;wildtype&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">errcounts</span><span class="p">[</span><span class="s1">&#39;wildtype&#39;</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="n">adj_errcounts</span><span class="p">[</span><span class="n">adj_errcounts</span><span class="p">[</span><span class="s1">&#39;site&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">][</span><span class="n">charlist</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">... </span>        <span class="o">==</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">250</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">21</span><span class="p">,</span> <span class="mi">10</span><span class="p">]))</span><span class="o">.</span><span class="n">all</span><span class="p">()</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.alignSubamplicon">
<code class="sig-prename descclassname">dms_tools2.utils.</code><code class="sig-name descname">alignSubamplicon</code><span class="sig-paren">(</span><em class="sig-param">refseq</em>, <em class="sig-param">r1</em>, <em class="sig-param">r2</em>, <em class="sig-param">refseqstart</em>, <em class="sig-param">refseqend</em>, <em class="sig-param">maxmuts</em>, <em class="sig-param">maxN</em>, <em class="sig-param">chartype</em>, <em class="sig-param">use_cutils=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/utils.html#alignSubamplicon"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.utils.alignSubamplicon" title="Permalink to this definition">¶</a></dt>
<dd><p>Try to align subamplicon to reference sequence at defined location.</p>
<p>Tries to align reads <cite>r1</cite> and <cite>r2</cite> to <cite>refseq</cite> at location
specified by <cite>refseqstart</cite> and <cite>refseqend</cite>. Determines how many
sites of type <cite>chartype</cite> have mutations, and if &lt;= <cite>maxmuts</cite> conside
the subamplicon to align if fraction of ambiguous nucleotides &lt;= <cite>maxN</cite>.
In <cite>r1</cite> and <cite>r2</cite>, an <code class="docutils literal notranslate"><span class="pre">N</span></code> indicates a non-called ambiguous identity.
If the reads disagree in a region of overlap that is set to <code class="docutils literal notranslate"><span class="pre">N</span></code> in
the final subamplicon, but if one read has <code class="docutils literal notranslate"><span class="pre">N</span></code> and the other a called
identity, then the called identity is used in the final subamplicon.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>refseq</cite> (str)</dt><dd><p>Sequence to which we align. if <cite>chartype</cite> is ‘codon’,
must be a valid coding (length multiple of 3).</p>
</dd>
<dt><cite>r1</cite> (str)</dt><dd><p>The forward sequence to align.</p>
</dd>
<dt><cite>r2</cite> (str)</dt><dd><p>The reverse sequence to align. When reverse complemented,
should read backwards in <cite>refseq</cite>.</p>
</dd>
<dt><cite>refseqstart</cite> (int)</dt><dd><p>The nucleotide in <cite>refseq</cite> (1, 2, … numbering) where the
first nucleotide in <cite>r1</cite> aligns.</p>
</dd>
<dt><cite>refseqend</cite> (int)</dt><dd><p>The nucleotide in <cite>refseq</cite> (1, 2, … numbering) where the
first nucleotide in <cite>r2</cite> aligns (note that <cite>r2</cite> then reads
backwards towards the 5’ end of <cite>refseq</cite>).</p>
</dd>
<dt><cite>maxmuts</cite> (int or float)</dt><dd><p>Maximum number of mutations of character <cite>chartype</cite> that
are allowed in the aligned subamplicons from the two reads.</p>
</dd>
<dt><cite>maxN</cite> (int or float)</dt><dd><p>Maximum number of nucleotides for which we allow
ambiguous (<code class="docutils literal notranslate"><span class="pre">N</span></code>) identities in final subamplicon.</p>
</dd>
<dt><cite>chartype</cite> (str)</dt><dd><p>Character type for which we count mutations.
Currently, the only allowable value is ‘codon’.</p>
</dd>
<dt><cite>use_cutils</cite> (bool)</dt><dd><p>Use the faster implementation in the <cite>_cutils</cite> module.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>If reads align, return aligned subamplicon as string (of length
<cite>refseqend - refseqstart + 1</cite>). Otherwise return <cite>None</cite>.</p>
</dd>
</dl>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">refseq</span> <span class="o">=</span> <span class="s1">&#39;ATGGGGAAA&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">alignSubamplicon</span><span class="p">(</span><span class="n">refseq</span><span class="p">,</span> <span class="s1">&#39;GGGGAA&#39;</span><span class="p">,</span> <span class="s1">&#39;TTTCCC&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">==</span> <span class="s1">&#39;GGGGAAA&#39;</span> 
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">alignSubamplicon</span><span class="p">(</span><span class="n">refseq</span><span class="p">,</span> <span class="s1">&#39;GGGGAA&#39;</span><span class="p">,</span> <span class="s1">&#39;TTTCCC&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">==</span> <span class="kc">False</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">alignSubamplicon</span><span class="p">(</span><span class="n">refseq</span><span class="p">,</span> <span class="s1">&#39;GGGGAT&#39;</span><span class="p">,</span> <span class="s1">&#39;TTTCCC&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">==</span> <span class="kc">False</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">alignSubamplicon</span><span class="p">(</span><span class="n">refseq</span><span class="p">,</span> <span class="s1">&#39;GGGGAT&#39;</span><span class="p">,</span> <span class="s1">&#39;TTTCCC&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">==</span> <span class="s1">&#39;GGGGANA&#39;</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">alignSubamplicon</span><span class="p">(</span><span class="n">refseq</span><span class="p">,</span> <span class="s1">&#39;GGGGAT&#39;</span><span class="p">,</span> <span class="s1">&#39;TATCCC&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">==</span> <span class="s1">&#39;GGGGATA&#39;</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">alignSubamplicon</span><span class="p">(</span><span class="n">refseq</span><span class="p">,</span> <span class="s1">&#39;GGGGAT&#39;</span><span class="p">,</span> <span class="s1">&#39;TATCCC&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">==</span> <span class="kc">False</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">alignSubamplicon</span><span class="p">(</span><span class="n">refseq</span><span class="p">,</span> <span class="s1">&#39;GGGNAA&#39;</span><span class="p">,</span> <span class="s1">&#39;TTTCCC&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">==</span> <span class="s1">&#39;GGGGAAA&#39;</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">alignSubamplicon</span><span class="p">(</span><span class="n">refseq</span><span class="p">,</span> <span class="s1">&#39;GGGNAA&#39;</span><span class="p">,</span> <span class="s1">&#39;TTNCCC&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">==</span> <span class="s1">&#39;GGGGAAA&#39;</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">alignSubamplicon</span><span class="p">(</span><span class="n">refseq</span><span class="p">,</span> <span class="s1">&#39;GTTTAA&#39;</span><span class="p">,</span> <span class="s1">&#39;TTTAAA&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">==</span> <span class="s1">&#39;GTTTAAA&#39;</span> 
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">alignSubamplicon</span><span class="p">(</span><span class="n">refseq</span><span class="p">,</span> <span class="s1">&#39;GGGGTA&#39;</span><span class="p">,</span> <span class="s1">&#39;TTACCC&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">==</span> <span class="s1">&#39;GGGGTAA&#39;</span> 
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">alignSubamplicon</span><span class="p">(</span><span class="n">refseq</span><span class="p">,</span> <span class="s1">&#39;GGGCTA&#39;</span><span class="p">,</span> <span class="s1">&#39;TTAGCC&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">==</span> <span class="kc">False</span> 
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.annotateCodonCounts">
<code class="sig-prename descclassname">dms_tools2.utils.</code><code class="sig-name descname">annotateCodonCounts</code><span class="sig-paren">(</span><em class="sig-param">counts</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/utils.html#annotateCodonCounts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.utils.annotateCodonCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets annotated <cite>pandas.DataFrame</cite> from codon counts.</p>
<p>Some of the programs (e.g., <cite>dms2_bcsubamplicons</cite>) create 
<code class="docutils literal notranslate"><span class="pre">*_codoncounts.csv</span></code> files when run with <code class="docutils literal notranslate"><span class="pre">--chartype</span> <span class="pre">codon</span></code>.
These CSV files have columns indicating the <cite>site</cite> and <cite>wildtype</cite>
codon, as well as a column for each codon giving the counts for that 
codon. This function reads that file (or a <cite>pandas.DataFrame</cite> read
from it) to return a <cite>pandas.DataFrame</cite> where a variety of additional
useful annotations have been added.</p>
<dl>
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>counts</cite> (str)</dt><dd><p>Name of existing codon counts CSV file, or <cite>pandas.DataFrame</cite>
holding counts.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl>
<dt><cite>df</cite> (<cite>pandas.DataFrame</cite>)</dt><dd><p>The DataFrame with the information in <cite>counts</cite> plus
the following added columns for each site:</p>
<blockquote>
<div><p><cite>ncounts</cite> : number of counts at site</p>
<p><cite>mutfreq</cite> : mutation frequency at site</p>
<p><cite>nstop</cite> : number of stop-codon mutations</p>
<p><cite>nsyn</cite> : number of synonymous mutations</p>
<p><cite>nnonsyn</cite> : number of nonsynonymous mutations</p>
<p><cite>n1nt</cite> : number of 1-nucleotide codon mutations</p>
<p><cite>n2nt</cite> : number of 2-nucleotide codon mutations</p>
<p><cite>n3nt</cite> : number of 3-nucleotide codon mutations</p>
<p><cite>AtoC</cite>, <cite>AtoG</cite>, etc : number of each nucleotide mutation
type among codon mutations with <strong>one</strong> nucleotide change.</p>
<p><cite>mutfreq1nt</cite>, <cite>mutfreq2nt</cite>, <cite>mutfreq3nt</cite> : frequency
of 1-, 2-, and 3-nucleotide codon mutations at site.</p>
</div></blockquote>
</dd>
</dl>
</dd>
</dl>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;site&#39;</span><span class="p">:[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="s1">&#39;wildtype&#39;</span><span class="p">:[</span><span class="s1">&#39;ATG&#39;</span><span class="p">,</span> <span class="s1">&#39;GGG&#39;</span><span class="p">],</span> <span class="s1">&#39;ATG&#39;</span><span class="p">:[</span><span class="mi">105</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;GGG&#39;</span><span class="p">:[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">117</span><span class="p">],</span> <span class="s1">&#39;GGA&#39;</span><span class="p">:[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">20</span><span class="p">],</span> <span class="s1">&#39;TGA&#39;</span><span class="p">:[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">codon</span> <span class="ow">in</span> <span class="n">CODONS</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="n">codon</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">d</span><span class="p">:</span>
<span class="gp">... </span>        <span class="n">d</span><span class="p">[</span><span class="n">codon</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">counts</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">f</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
<span class="gp">... </span>    <span class="n">df</span> <span class="o">=</span> <span class="n">annotateCodonCounts</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">([</span><span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">==</span> <span class="n">counts</span><span class="p">[</span><span class="n">col</span><span class="p">])</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">counts</span><span class="o">.</span><span class="n">columns</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;ncounts&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">110</span><span class="p">,</span> <span class="mi">139</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;mutfreq&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">5</span> <span class="o">/</span> <span class="mf">110.</span><span class="p">,</span> <span class="mi">22</span> <span class="o">/</span> <span class="mf">139.</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;nstop&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;nsyn&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">20</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;nnonsyn&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;n1nt&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">20</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;n2nt&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;n3nt&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;GtoA&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">20</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;AtoC&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;mutfreq1nt&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">20</span> <span class="o">/</span> <span class="mf">139.</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;mutfreq3nt&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">2</span> <span class="o">/</span> <span class="mf">110.</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.barcodeInfoToCodonVariantTable">
<code class="sig-prename descclassname">dms_tools2.utils.</code><code class="sig-name descname">barcodeInfoToCodonVariantTable</code><span class="sig-paren">(</span><em class="sig-param">samples</em>, <em class="sig-param">geneseq</em>, <em class="sig-param">path=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/utils.html#barcodeInfoToCodonVariantTable"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.utils.barcodeInfoToCodonVariantTable" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert barcode info files into a CodonVariantTable</p>
<p>Convert barcode info files output from <cite>dms2_bcsubamp</cite> into a
<cite>CodonVariantTable</cite>. Barcode info files contain reads and barcodes from
barcoded subamplicon sequencing, described 
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/bcsubamp.html">here</a>.
This function takes consensus reads retained by <cite>dms2_bcsubamp</cite>, 
gives each unique sequence a numerical barcode (since the barcodes from 
<cite>dms2_bcsubamp</cite> could come from the same variant), and counts the number
of retained consensus reads corresponding to each sequence. Then, a
<cite>CodonVariantTable</cite> is made using the sequences and their numerical 
barcodes, and counts are added based on the number of retained consensus
reads of those sequences. Therefore, the <cite>CodonVariantTable</cite> will only 
contain one ‘variant’ for each unique sequence with the total count for all
the unbarcoded variants in the experiment which had the same sequence.</p>
<dl>
<dt>Args:</dt><dd><dl>
<dt><cite>samples</cite> (dict):</dt><dd><p>Dictionary with libraries as keys and lists of info file prefixes
(file names without the ‘_bcinfo.txt.gz’) for files corresponding
to those libraries as values.</p>
<dl class="simple">
<dt>Example: {‘library-1’:[‘condition-1-library-1’],</dt><dd><p>‘library-2’:[‘condition-1-library-2’]}</p>
</dd>
</dl>
</dd>
<dt><cite>geneseq</cite> (str):</dt><dd><p>The wildtype gene sequence</p>
</dd>
<dt><cite>path</cite> (str)</dt><dd><p>Directory in which barcode info files are located</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>A <cite>dms_variants.codonvarianttable.CodonVariantTable</cite> with ‘counts’
generated from the barcode info files</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.buildReadConsensus">
<code class="sig-prename descclassname">dms_tools2.utils.</code><code class="sig-name descname">buildReadConsensus</code><span class="sig-paren">(</span><em class="sig-param">reads</em>, <em class="sig-param">minreads</em>, <em class="sig-param">minconcur</em>, <em class="sig-param">use_cutils=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/utils.html#buildReadConsensus"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.utils.buildReadConsensus" title="Permalink to this definition">¶</a></dt>
<dd><p>Builds consensus sequence of some reads.</p>
<p>You may want to pre-fill low-quality sites with <code class="docutils literal notranslate"><span class="pre">N</span></code>
using <cite>lowQtoN</cite>. An <code class="docutils literal notranslate"><span class="pre">N</span></code> is considered a non-called identity.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>reads</cite> (list)</dt><dd><p>List of reads as strings. If reads are not all same
length, shorter ones are extended from 3’ end with <code class="docutils literal notranslate"><span class="pre">N</span></code>
to match maximal length.</p>
</dd>
<dt><cite>minreads</cite> (int)</dt><dd><p>Only call consensus at a site if at least this many reads 
have called identity.</p>
</dd>
<dt><cite>minconcur</cite> (float)</dt><dd><p>Only call consensus at site if &gt;= this fraction of called
identities agree.</p>
</dd>
<dt><cite>use_cutils</cite> (bool)</dt><dd><p>Use the faster implementation in the <cite>_cutils</cite> module.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>A string giving the consensus sequence. Non-called 
sites are returned as <code class="docutils literal notranslate"><span class="pre">N`</span></code>.</p>
</dd>
</dl>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">reads</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ATGCAT&#39;</span><span class="p">,</span>
<span class="gp">... </span>         <span class="s1">&#39;NTGNANA&#39;</span><span class="p">,</span>
<span class="gp">... </span>         <span class="s1">&#39;ACGNNTAT&#39;</span><span class="p">,</span>
<span class="gp">... </span>         <span class="s1">&#39;NTGNTA&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">buildReadConsensus</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mf">0.75</span><span class="p">)</span> <span class="o">==</span> <span class="s1">&#39;ATGNNNAN&#39;</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">reads</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;CTGCATAT&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">buildReadConsensus</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mf">0.75</span><span class="p">)</span> <span class="o">==</span> <span class="s1">&#39;NTGCATAT&#39;</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.codonEvolAccessibility">
<code class="sig-prename descclassname">dms_tools2.utils.</code><code class="sig-name descname">codonEvolAccessibility</code><span class="sig-paren">(</span><em class="sig-param">seqs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/utils.html#codonEvolAccessibility"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.utils.codonEvolAccessibility" title="Permalink to this definition">¶</a></dt>
<dd><p>Accessibility of amino acids by nucleotide mutations.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>seqs</cite> (str or list)</dt><dd><p>A single coding sequence or a list of such sequences.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>A pandas DataFrame listing all sites in the sequence(s)
numbered 1, 2, …, with columns giving the accessibility
of each amino acid by single nucleotide mutations.</p>
</dd>
</dl>
<p>The accessibility of codon <span class="math notranslate nohighlight">\(c\)</span> to amino-acid <span class="math notranslate nohighlight">\(a\)</span>
by single-nucleotide mutations is defined as the minimum
number of nucleotide mutations needed to generate that
amino-acid.</p>
<p>For a collection of sequences, we calculate the 
accessibility as the weighted average of the accessibilities
of all codons observed at that site in the collection of
sequences.</p>
<p>As an example, compute accessibility for one sequence:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="s2">&quot;ATGGGA&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">acc</span> <span class="o">=</span> <span class="n">codonEvolAccessibility</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
</pre></div>
</div>
<p>The returned pandas DataFrame <cite>acc</cite> is has a column named
<cite>site</cite> plus columns for all amino acids:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">acc</span><span class="o">.</span><span class="n">columns</span> <span class="o">==</span> <span class="p">[</span><span class="s1">&#39;site&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">AAS_WITHSTOP</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<p>We look at entries for a few amino acids. At the first 
site, the wildtype entry in the sequence <cite>s</cite> is the codon
for <em>M</em> (<code class="docutils literal notranslate"><span class="pre">ATG</span></code>). So at this site, the distance to <em>M</em>
is 0. The distance to <em>I</em> (which has codon <code class="docutils literal notranslate"><span class="pre">ATA</span></code> as a
codon) is 1, and the distance to <em>W</em> (which has only <code class="docutils literal notranslate"><span class="pre">TGG</span></code>
as a codon) is 2.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">acc</span><span class="p">[[</span><span class="s1">&#39;site&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;I&#39;</span><span class="p">,</span> <span class="s1">&#39;M&#39;</span><span class="p">,</span> <span class="s1">&#39;W&#39;</span><span class="p">]]</span>
<span class="go">   site    G    I    M    W</span>
<span class="go">0     1  2.0  1.0  0.0  2.0</span>
<span class="go">1     2  0.0  2.0  3.0  2.0</span>
</pre></div>
</div>
<p>If we pass the function a list of multiple sequences,
then the accessibilities are averaged over the sequences:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">acc2</span> <span class="o">=</span> <span class="n">codonEvolAccessibility</span><span class="p">([</span><span class="s1">&#39;ATGGGA&#39;</span><span class="p">,</span> <span class="s1">&#39;ATAGGA&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">acc2</span><span class="p">[[</span><span class="s1">&#39;site&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;I&#39;</span><span class="p">,</span> <span class="s1">&#39;M&#39;</span><span class="p">,</span> <span class="s1">&#39;W&#39;</span><span class="p">]]</span>
<span class="go">   site    G    I    M    W</span>
<span class="go">0     1  2.0  0.5  0.5  2.5</span>
<span class="go">1     2  0.0  2.0  3.0  2.0</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.codonToAACounts">
<code class="sig-prename descclassname">dms_tools2.utils.</code><code class="sig-name descname">codonToAACounts</code><span class="sig-paren">(</span><em class="sig-param">counts</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/utils.html#codonToAACounts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.utils.codonToAACounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Makes amino-acid counts <cite>pandas.DataFrame</cite> from codon counts.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>counts</cite> (<cite>pandas.DataFrame</cite>)</dt><dd><p>Columns are the string <cite>site</cite> <cite>wildtype</cite> and all codons
in <cite>CODONS</cite>. Additional columns are allowed
but ignored.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt><cite>aacounts</cite> (<cite>pandas.DataFrame</cite>)</dt><dd><p>Columns are the string <cite>site</cite> and all amino acids
in <cite>AAS_WITHSTOP</cite> with counts for each
amino acid made by summing counts for encoding codons.</p>
</dd>
</dl>
</dd>
</dl>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;site&#39;</span><span class="p">:[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="s1">&#39;othercol&#39;</span><span class="p">:[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="s1">&#39;ATG&#39;</span><span class="p">:[</span><span class="mi">105</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;GGG&#39;</span><span class="p">:[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">117</span><span class="p">],</span> <span class="s1">&#39;GGA&#39;</span><span class="p">:[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">20</span><span class="p">],</span> <span class="s1">&#39;TGA&#39;</span><span class="p">:[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;wildtype&#39;</span><span class="p">:[</span><span class="s1">&#39;ATG&#39;</span><span class="p">,</span> <span class="s1">&#39;GGG&#39;</span><span class="p">]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">codon</span> <span class="ow">in</span> <span class="n">CODONS</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="n">codon</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">d</span><span class="p">:</span>
<span class="gp">... </span>        <span class="n">d</span><span class="p">[</span><span class="n">codon</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aacounts</span> <span class="o">=</span> <span class="n">codonToAACounts</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s1">&#39;othercol&#39;</span> <span class="ow">in</span> <span class="n">aacounts</span><span class="o">.</span><span class="n">columns</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">aacounts</span><span class="p">[</span><span class="s1">&#39;site&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">aacounts</span><span class="p">[</span><span class="s1">&#39;wildtype&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="s1">&#39;M&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">aacounts</span><span class="p">[</span><span class="s1">&#39;M&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">105</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">aacounts</span><span class="p">[</span><span class="s1">&#39;G&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">137</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">aacounts</span><span class="p">[</span><span class="s1">&#39;*&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">aacounts</span><span class="p">[</span><span class="s1">&#39;V&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.codon_to_nt_counts">
<code class="sig-prename descclassname">dms_tools2.utils.</code><code class="sig-name descname">codon_to_nt_counts</code><span class="sig-paren">(</span><em class="sig-param">codoncounts</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/utils.html#codon_to_nt_counts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.utils.codon_to_nt_counts" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert codon counts file to nucleotide counts.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>codoncounts</cite> (str or pandas.DataFrame)</dt><dd><p>Codon counts in format produced by <code class="docutils literal notranslate"><span class="pre">dms2_bcsubamp</span></code>,
either as CSV file or data frame holding CSV.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>pandas.DataFrame with nucleotide counts.</p>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span><span class="p">(</span><span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">textwrap</span><span class="o">.</span><span class="n">dedent</span><span class="p">(</span><span class="s1">&#39;&#39;&#39;</span>
<span class="gp">... </span><span class="s1">            site,wildtype,AAA,AAC,AAG,AAT,ACA,ACC,ACG,ACT,AGA,AGC,AGG,AGT,ATA,ATC,ATG,ATT,CAA,CAC,CAG,CAT,CCA,CCC,CCG,CCT,CGA,CGC,CGG,CGT,CTA,CTC,CTG,CTT,GAA,GAC,GAG,GAT,GCA,GCC,GCG,GCT,GGA,GGC,GGG,GGT,GTA,GTC,GTG,GTT,TAA,TAC,TAG,TAT,TCA,TCC,TCG,TCT,TGA,TGC,TGG,TGT,TTA,TTC,TTG,TTT</span>
<span class="gp">... </span><span class="s1">            1,ATG,0,0,0,0,0,0,2,0,0,0,0,0,8,0,333985,14,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,6,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0</span>
<span class="gp">... </span><span class="s1">            2,AAG,16,20,333132,41,13,12,27,14,8,6,67,8,9,13,29,9,10,11,12,8,10,15,15,11,6,9,3,7,8,10,17,4,3,7,49,7,9,14,9,4,10,7,7,7,9,11,11,5,14,14,11,6,13,16,15,14,9,9,15,8,9,11,8,15</span>
<span class="gp">... </span><span class="s1">            3,GCA,2,3,8,3,34,11,7,6,7,6,9,8,4,3,5,0,6,14,10,12,6,8,7,10,5,11,7,6,6,1,3,12,19,6,11,9,333250,10,6,9,15,3,5,5,37,9,9,7,8,4,8,3,23,5,7,8,6,11,7,10,7,9,3,6 </span>
<span class="gp">... </span><span class="s1">            &#39;&#39;&#39;</span><span class="o">.</span><span class="n">strip</span><span class="p">()))</span>
<span class="gp">... </span>    <span class="n">f</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
<span class="gp">... </span>    <span class="n">nt_counts</span> <span class="o">=</span> <span class="n">codon_to_nt_counts</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nt_counts</span>
<span class="go">   site wildtype       A       C       G       T</span>
<span class="go">0     1        A  334009       0       6       0</span>
<span class="go">1     2        T       0       2       0  334013</span>
<span class="go">2     3        G       8       0  333993      14</span>
<span class="go">3     4        A  333424     156     169     187</span>
<span class="go">4     5        A  333361     211     186     178</span>
<span class="go">5     6        G     156     185  333427     168</span>
<span class="go">6     7        G     116     124  333410     125</span>
<span class="go">7     8        C     126  333407     121     121</span>
<span class="go">8     9        A  333435     114     112     114</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.convertCountsFormat">
<code class="sig-prename descclassname">dms_tools2.utils.</code><code class="sig-name descname">convertCountsFormat</code><span class="sig-paren">(</span><em class="sig-param">oldfile</em>, <em class="sig-param">newfile</em>, <em class="sig-param">charlist</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/utils.html#convertCountsFormat"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.utils.convertCountsFormat" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert counts file from <code class="docutils literal notranslate"><span class="pre">dms_tools</span></code> to <code class="docutils literal notranslate"><span class="pre">dms_tools2</span></code> format.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>oldfile</cite> (str)</dt><dd><p>Name of counts file in the old <code class="docutils literal notranslate"><span class="pre">dms_tools</span></code> format:
<a class="reference external" href="http://jbloomlab.github.io/dms_tools/fileformats.html">http://jbloomlab.github.io/dms_tools/fileformats.html</a></p>
</dd>
<dt><cite>newfile</cite> (str)</dt><dd><p>Name of created counts file in the <code class="docutils literal notranslate"><span class="pre">dms_tools2</span></code> format:
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms2_bcsubamp.html">https://jbloomlab.github.io/dms_tools2/dms2_bcsubamp.html</a></p>
</dd>
<dt><cite>charlist</cite> (list)</dt><dd><p>List of characters that we expect in the counts files.
For instance, could be <cite>CODONS</cite>.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.getSubstitutions">
<code class="sig-prename descclassname">dms_tools2.utils.</code><code class="sig-name descname">getSubstitutions</code><span class="sig-paren">(</span><em class="sig-param">wildtype</em>, <em class="sig-param">mutant</em>, <em class="sig-param">amino_acid=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/utils.html#getSubstitutions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.utils.getSubstitutions" title="Permalink to this definition">¶</a></dt>
<dd><p>Get space delimited string of substitutions</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>wildtype</cite> (str):</dt><dd><p>The wildtype sequence</p>
</dd>
<dt><cite>mutant</cite> (str):</dt><dd><p>The mutant sequence</p>
</dd>
<dt><cite>amino_acid</cite> (bool)</dt><dd><p>Specify whether the sequence is amino acid.
Default is False</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>A space delimited string of substitutions present in the
mutant sequence</p>
</dd>
</dl>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">getSubstitutions</span><span class="p">(</span><span class="s1">&#39;AGT&#39;</span><span class="p">,</span> <span class="s1">&#39;TGT&#39;</span><span class="p">)</span>
<span class="go">&#39;A1T&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">getSubstitutions</span><span class="p">(</span><span class="s1">&#39;AAGTAACGA&#39;</span><span class="p">,</span> <span class="s1">&#39;ATCTAACGA&#39;</span><span class="p">)</span>
<span class="go">&#39;A2T G3C&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">getSubstitutions</span><span class="p">(</span><span class="s1">&#39;TYARV&#39;</span><span class="p">,</span> <span class="s1">&#39;GYAGV&#39;</span><span class="p">,</span> <span class="n">amino_acid</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">&#39;T1G R4G&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.incrementCounts">
<code class="sig-prename descclassname">dms_tools2.utils.</code><code class="sig-name descname">incrementCounts</code><span class="sig-paren">(</span><em class="sig-param">refseqstart</em>, <em class="sig-param">subamplicon</em>, <em class="sig-param">chartype</em>, <em class="sig-param">counts</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/utils.html#incrementCounts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.utils.incrementCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Increment counts dict based on an aligned subamplicon.</p>
<p>This is designed for keeping track of counts of different
mutations / identities when aligning many subamplicons to
a sequence.</p>
<p>Any positions where <cite>subamplicon</cite> has an <code class="docutils literal notranslate"><span class="pre">N</span></code> are ignored,
and not added to <cite>counts</cite>.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>refseqstart</cite> (int)</dt><dd><p>First nucleotide position in 1, 2, … numbering 
where <cite>subamplicon</cite> aligns.</p>
</dd>
<dt><cite>subamplicon</cite> (str)</dt><dd><p>The subamplicon.</p>
</dd>
<dt><cite>chartype</cite> (str)</dt><dd><p>Character type for which we are counting mutations.
Currently, only allowable value is ‘codon’.</p>
</dd>
<dt><cite>counts</cite> (dict)</dt><dd><p>Stores counts of identities, and is incremented by
this function. Is a dict keyed by every possible
character (e.g., codon), with values lists with
element <cite>i</cite> holding the counts for position <cite>i</cite>
in 0, 1, … numbering.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>On completion, <cite>counts</cite> has been incremented.</p>
</dd>
</dl>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">codonlen</span> <span class="o">=</span> <span class="mi">10</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">([(</span><span class="n">codon</span><span class="p">,</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">codonlen</span><span class="p">)</span> <span class="k">for</span> <span class="n">codon</span> 
<span class="gp">... </span>        <span class="ow">in</span> <span class="n">CODONS</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">subamplicon1</span> <span class="o">=</span> <span class="s1">&#39;ATGGACTTTC&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">incrementCounts</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">subamplicon1</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">,</span> <span class="n">counts</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">subamplicon2</span> <span class="o">=</span> <span class="s1">&#39;GGTCTTTCCCGGN&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">incrementCounts</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="n">subamplicon2</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">,</span> <span class="n">counts</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s1">&#39;ATG&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s1">&#39;GAC&#39;</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s1">&#39;GTC&#39;</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s1">&#39;TTT&#39;</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="mi">2</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s1">&#39;CCC&#39;</span><span class="p">][</span><span class="mi">3</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">sum</span><span class="p">([</span><span class="nb">sum</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">counts</span><span class="o">.</span><span class="n">values</span><span class="p">()])</span> <span class="o">==</span> <span class="mi">6</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.initLogger">
<code class="sig-prename descclassname">dms_tools2.utils.</code><code class="sig-name descname">initLogger</code><span class="sig-paren">(</span><em class="sig-param">logfile</em>, <em class="sig-param">prog</em>, <em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/utils.html#initLogger"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.utils.initLogger" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize output logging for scripts.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>logfile</cite> (str or <cite>sys.stdout</cite>)</dt><dd><p>Name of file to which log is written, or 
<cite>sys.stdout</cite> if you just want to write information
to standard output.</p>
</dd>
<dt><cite>prog</cite> (str)</dt><dd><p>Name of program for which we are logging.</p>
</dd>
<dt><cite>args</cite> (dict)</dt><dd><p>Program arguments as arg / value pairs.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>If <cite>logfile</cite> is a string giving a file name, returns
an opened and initialized <cite>logging.Logger</cite>. If <cite>logfile</cite>
is <cite>sys.stdout</cite>, then writes information to <cite>sys.stdout</cite>.
In either case, basic information is written about the program 
and args.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.iteratePairedFASTQ">
<code class="sig-prename descclassname">dms_tools2.utils.</code><code class="sig-name descname">iteratePairedFASTQ</code><span class="sig-paren">(</span><em class="sig-param">r1files</em>, <em class="sig-param">r2files</em>, <em class="sig-param">r1trim=None</em>, <em class="sig-param">r2trim=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/utils.html#iteratePairedFASTQ"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.utils.iteratePairedFASTQ" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterates over FASTQ files for single or paired-end sequencing.</p>
<dl>
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>r1files</cite> (list or str)</dt><dd><p>Name of R1 FASTQ file or list of such files. Can optionally
be gzipped.</p>
</dd>
<dt><cite>r2files</cite> (list or str or <cite>None</cite>)</dt><dd><p>Like <cite>r1files</cite> but for R2 files, or <cite>None</cite> if no R2.</p>
</dd>
<dt><cite>r1trim</cite> (int or <cite>None</cite>)</dt><dd><p>If not <cite>None</cite>, trim <cite>r1</cite> and <cite>q1</cite> to be no longer than this.</p>
</dd>
<dt><cite>r2trim</cite> (int or <cite>None</cite>)</dt><dd><p>Like <cite>r1trim</cite> but for R2.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>Each iteration returns <cite>(name, r1, r2, q1, q2, fail)</cite> where:</p>
<blockquote>
<div><ul class="simple">
<li><p><cite>name</cite> is a string giving the read name</p></li>
<li><p><cite>r1</cite> and <cite>r2</cite> are strings giving the reads; <cite>r2</cite>
is <cite>None</cite> if no R2.</p></li>
<li><p><cite>q1</cite> and <cite>q2</cite> are strings giving the PHRED Q scores;
<cite>q2</cite> is none if no R2.</p></li>
<li><p><cite>fail</cite> is <cite>True</cite> if either read failed Illumina chastity
filter, <cite>False</cite> if both passed, <cite>None</cite> if info not present.</p></li>
</ul>
</div></blockquote>
</dd>
</dl>
<p>We run a simple test by first writing an example FASTQ file and
then testing on it.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">n1_1</span> <span class="o">=</span> <span class="s1">&#39;@DH1DQQN1:933:HMLH5BCXY:1:1101:2165:1984 1:N:0:CGATGT&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">r1_1</span> <span class="o">=</span> <span class="s1">&#39;ATGCAATTG&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">q1_1</span> <span class="o">=</span> <span class="s1">&#39;GGGGGIIII&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">n2_1</span> <span class="o">=</span> <span class="s1">&#39;@DH1DQQN1:933:HMLH5BCXY:1:1101:2165:1984 2:N:0:CGATGT&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">r2_1</span> <span class="o">=</span> <span class="s1">&#39;CATGCATA&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">q2_1</span> <span class="o">=</span> <span class="s1">&#39;G&lt;GGGIII&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tf</span> <span class="o">=</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">tf</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">r1file</span><span class="p">,</span> <span class="n">tf</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">r2file</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">r1file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span>
<span class="gp">... </span>            <span class="n">n1_1</span><span class="p">,</span> <span class="n">r1_1</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="n">q1_1</span><span class="p">,</span>
<span class="gp">... </span>            <span class="n">n1_1</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;:N:&#39;</span><span class="p">,</span> <span class="s1">&#39;:Y:&#39;</span><span class="p">),</span> <span class="n">r1_1</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="n">q1_1</span><span class="p">,</span>
<span class="gp">... </span>            <span class="n">n1_1</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">],</span> <span class="n">r1_1</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="n">q1_1</span><span class="p">,</span>
<span class="gp">... </span>            <span class="p">]))</span>
<span class="gp">... </span>    <span class="n">r1file</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">r2file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span>
<span class="gp">... </span>            <span class="n">n2_1</span><span class="p">,</span> <span class="n">r2_1</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="n">q2_1</span><span class="p">,</span>
<span class="gp">... </span>            <span class="n">n2_1</span><span class="p">,</span> <span class="n">r2_1</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="n">q2_1</span><span class="p">,</span>
<span class="gp">... </span>            <span class="n">n2_1</span><span class="p">,</span> <span class="n">r2_1</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="n">q2_1</span><span class="p">,</span>
<span class="gp">... </span>            <span class="p">]))</span>
<span class="gp">... </span>    <span class="n">r2file</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
<span class="gp">... </span>    <span class="n">itr</span> <span class="o">=</span> <span class="n">iteratePairedFASTQ</span><span class="p">(</span><span class="n">r1file</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">r2file</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">r1trim</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">r2trim</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="gp">... </span>    <span class="nb">next</span><span class="p">(</span><span class="n">itr</span><span class="p">)</span> <span class="o">==</span> <span class="p">(</span><span class="n">n1_1</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span> <span class="p">:</span> <span class="p">],</span> <span class="n">r1_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">4</span><span class="p">],</span>
<span class="gp">... </span>            <span class="n">r2_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">5</span><span class="p">],</span> <span class="n">q1_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">4</span><span class="p">],</span> <span class="n">q2_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">5</span><span class="p">],</span> <span class="kc">False</span><span class="p">)</span>
<span class="gp">... </span>    <span class="nb">next</span><span class="p">(</span><span class="n">itr</span><span class="p">)</span> <span class="o">==</span> <span class="p">(</span><span class="n">n1_1</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span> <span class="p">:</span> <span class="p">],</span> <span class="n">r1_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">4</span><span class="p">],</span>
<span class="gp">... </span>            <span class="n">r2_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">5</span><span class="p">],</span> <span class="n">q1_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">4</span><span class="p">],</span> <span class="n">q2_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">5</span><span class="p">],</span> <span class="kc">True</span><span class="p">)</span>
<span class="gp">... </span>    <span class="nb">next</span><span class="p">(</span><span class="n">itr</span><span class="p">)</span> <span class="o">==</span> <span class="p">(</span><span class="n">n1_1</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span> <span class="p">:</span> <span class="p">],</span> <span class="n">r1_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">4</span><span class="p">],</span>
<span class="gp">... </span>            <span class="n">r2_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">5</span><span class="p">],</span> <span class="n">q1_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">4</span><span class="p">],</span> <span class="n">q2_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">5</span><span class="p">],</span> <span class="kc">None</span><span class="p">)</span>
<span class="go">True</span>
<span class="go">True</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Now do the same test but for just R1:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">tf</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">r1file</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">r1file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span>
<span class="gp">... </span>            <span class="n">n1_1</span><span class="p">,</span> <span class="n">r1_1</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="n">q1_1</span><span class="p">,</span>
<span class="gp">... </span>            <span class="n">n1_1</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;:N:&#39;</span><span class="p">,</span> <span class="s1">&#39;:Y:&#39;</span><span class="p">),</span> <span class="n">r1_1</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="n">q1_1</span><span class="p">,</span>
<span class="gp">... </span>            <span class="n">n1_1</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">],</span> <span class="n">r1_1</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="n">q1_1</span><span class="p">,</span>
<span class="gp">... </span>            <span class="p">]))</span>
<span class="gp">... </span>    <span class="n">r1file</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
<span class="gp">... </span>    <span class="n">itr_R1</span> <span class="o">=</span> <span class="n">iteratePairedFASTQ</span><span class="p">(</span><span class="n">r1file</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="n">r1trim</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">... </span>    <span class="nb">next</span><span class="p">(</span><span class="n">itr_R1</span><span class="p">)</span> <span class="o">==</span> <span class="p">(</span><span class="n">n1_1</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span> <span class="p">:</span> <span class="p">],</span> <span class="n">r1_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">4</span><span class="p">],</span>
<span class="gp">... </span>            <span class="kc">None</span><span class="p">,</span> <span class="n">q1_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">4</span><span class="p">],</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
<span class="gp">... </span>    <span class="nb">next</span><span class="p">(</span><span class="n">itr_R1</span><span class="p">)</span> <span class="o">==</span> <span class="p">(</span><span class="n">n1_1</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span> <span class="p">:</span> <span class="p">],</span> <span class="n">r1_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">4</span><span class="p">],</span>
<span class="gp">... </span>            <span class="kc">None</span><span class="p">,</span> <span class="n">q1_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">4</span><span class="p">],</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
<span class="gp">... </span>    <span class="nb">next</span><span class="p">(</span><span class="n">itr_R1</span><span class="p">)</span> <span class="o">==</span> <span class="p">(</span><span class="n">n1_1</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span> <span class="p">:</span> <span class="p">],</span> <span class="n">r1_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">4</span><span class="p">],</span>
<span class="gp">... </span>            <span class="kc">None</span><span class="p">,</span> <span class="n">q1_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">4</span><span class="p">],</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
<span class="go">True</span>
<span class="go">True</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.lowQtoN">
<code class="sig-prename descclassname">dms_tools2.utils.</code><code class="sig-name descname">lowQtoN</code><span class="sig-paren">(</span><em class="sig-param">r</em>, <em class="sig-param">q</em>, <em class="sig-param">minq</em>, <em class="sig-param">use_cutils=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/utils.html#lowQtoN"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.utils.lowQtoN" title="Permalink to this definition">¶</a></dt>
<dd><p>Replaces low quality nucleotides with <code class="docutils literal notranslate"><span class="pre">N</span></code> characters.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>r</cite> (str)</dt><dd><p>A string representing a sequencing read.</p>
</dd>
<dt><cite>q</cite> (str)</dt><dd><p>String of same length as <cite>r</cite> holding Q scores
in Sanger ASCII encoding.</p>
</dd>
<dt><cite>minq</cite> (length-one string)</dt><dd><p>Replace all positions in <cite>r</cite> where <cite>q</cite> is &lt; this.</p>
</dd>
<dt><cite>use_cutils</cite> (bool)</dt><dd><p>Use the faster implementation in the <cite>_cutils</cite> module.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>A version of <cite>r</cite> where all positions <cite>i</cite> where 
<cite>q[i] &lt; minq</cite> have been replaced with <code class="docutils literal notranslate"><span class="pre">N</span></code>.</p>
</dd>
</dl>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">r</span> <span class="o">=</span> <span class="s1">&#39;ATGCAT&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">q</span> <span class="o">=</span> <span class="s1">&#39;GB&lt;.0+&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">minq</span> <span class="o">=</span> <span class="s1">&#39;0&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lowQtoN</span><span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="n">q</span><span class="p">,</span> <span class="n">minq</span><span class="p">)</span> <span class="o">==</span> <span class="s1">&#39;ATGNAN&#39;</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.rarefactionCurve">
<code class="sig-prename descclassname">dms_tools2.utils.</code><code class="sig-name descname">rarefactionCurve</code><span class="sig-paren">(</span><em class="sig-param">barcodes</em>, <em class="sig-param">*</em>, <em class="sig-param">maxpoints=100000.0</em>, <em class="sig-param">logspace=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/utils.html#rarefactionCurve"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.utils.rarefactionCurve" title="Permalink to this definition">¶</a></dt>
<dd><p>Rarefaction curve from list of barcodes.</p>
<p>Uses the analytical formula for the rarefaction curve defined
<a class="reference external" href="https://en.wikipedia.org/wiki/Rarefaction_(ecology)#Derivation">on Wikipedia</a>.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>barcodes</cite> (list or pandas Series)</dt><dd><p>Holds the list of unique barcodes for which we calculate
the rarefaction curve. It is expected that some of these
barcodes will be repeated multiple times in the list if
the sampling is approaching saturation.</p>
</dd>
<dt><cite>maxpoints</cite> (int)</dt><dd><p>Only calculate values at this many points. The benefit
of this is that it can become very costly to calculate
the curve at every point when there are many points.</p>
</dd>
<dt><cite>logspace</cite> (True)</dt><dd><p>Logarithmically space the <cite>maxpoints</cite> points for
the calculation. This will give better results if
we are subsampling and the curve saturates. Only
done if we have to subsample.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>The 2-tuple <cite>(nreads, nbarcodes)</cite>, where both <cite>nreads</cite> and
<cite>nbarcodes</cite> are lists of the same length, and <cite>nbarcodes[i]</cite>
is the expected number of barcodes observed when there are
<cite>nreads[i]</cite> reads.</p>
</dd>
</dl>
<p>Here we take a very small list and show that the results given
by the function are equivalent to those obtained by random
subsampling:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">barcodes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="n">nreads</span><span class="p">,</span> <span class="n">nbarcodes</span><span class="p">)</span> <span class="o">=</span> <span class="n">rarefactionCurve</span><span class="p">(</span><span class="n">barcodes</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nrand</span> <span class="o">=</span> <span class="mi">100000</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim_equal_calc</span> <span class="o">=</span> <span class="p">[]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">barcodes</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">nbarcodes_sim</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">([</span><span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="n">barcodes</span><span class="p">,</span> <span class="n">n</span><span class="p">)))</span>
<span class="gp">... </span>            <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nrand</span><span class="p">)])</span> <span class="o">/</span> <span class="n">nrand</span>
<span class="gp">... </span>    <span class="n">sim_equal_calc</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">nbarcodes_sim</span><span class="p">,</span>
<span class="gp">... </span>            <span class="n">nbarcodes</span><span class="p">[</span><span class="n">nreads</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">n</span><span class="p">)],</span> <span class="n">atol</span><span class="o">=</span><span class="mf">1e-2</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">sim_equal_calc</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.renumberSites">
<code class="sig-prename descclassname">dms_tools2.utils.</code><code class="sig-name descname">renumberSites</code><span class="sig-paren">(</span><em class="sig-param">renumbfile</em>, <em class="sig-param">infiles</em>, <em class="sig-param">missing='error'</em>, <em class="sig-param">outfiles=None</em>, <em class="sig-param">outprefix=None</em>, <em class="sig-param">outdir=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/utils.html#renumberSites"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.utils.renumberSites" title="Permalink to this definition">¶</a></dt>
<dd><p>Renumber sites in CSV files.</p>
<p>Switch numbering scheme in files with a column named <cite>site</cite>.</p>
<p>You must specify <strong>exactly one</strong> of <cite>outfiles</cite>,
<cite>outprefix</cite>, and <cite>outdir</cite> as something other than <cite>None</cite>.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>renumbfile</cite> (str)</dt><dd><p>Name of existing CSV file with the re-numbering scheme.
Should have columns with name <cite>original</cite> and <cite>new</cite>.
Each entry in <cite>original</cite> should refer to a site in 
the input files, and each entry in <cite>new</cite> should be
the new number for this site. If an entry in <cite>new</cite>
is <cite>None</cite> or <cite>nan</cite> then it is dropped from the newly
numbered files regardless of <cite>missing</cite>.</p>
</dd>
<dt><cite>infiles</cite> (list)</dt><dd><p>List of existing CSV files that we are re-numbering.
Each file must have an entry of <cite>site</cite>.</p>
</dd>
<dt><cite>missing</cite> (str)</dt><dd><dl class="simple">
<dt>How to handle sites in <cite>infiles</cite> but not <cite>renumbfile</cite>.</dt><dd><ul class="simple">
<li><p><cite>error</cite>: raise an error</p></li>
<li><p><cite>skip</cite>: skip renumbering, leave with original number</p></li>
<li><p><cite>drop</cite>: drop any sites not in <cite>renumbfile</cite></p></li>
</ul>
</dd>
</dl>
</dd>
<dt><cite>outfiles</cite> (list)</dt><dd><p>List of output files of the same length as <cite>infiles</cite>.
The numbered version of <cite>infiles</cite> is named as the
corresponding entry in <cite>outfiles</cite>.</p>
</dd>
<dt><cite>outdir</cite> (str)</dt><dd><p>A directory name. The renumbered files have the same
names as in <cite>infile</cite>, but are now placed in <cite>outdir</cite>.</p>
</dd>
<dt><cite>outprefix</cite> (str)</dt><dd><p>The renumbered files have the same names and locations
as <cite>infiles</cite>, but have the pre-pended filename extension
<cite>outprefix</cite>.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.reverseComplement">
<code class="sig-prename descclassname">dms_tools2.utils.</code><code class="sig-name descname">reverseComplement</code><span class="sig-paren">(</span><em class="sig-param">s</em>, <em class="sig-param">use_cutils=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/utils.html#reverseComplement"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.utils.reverseComplement" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets reverse complement of DNA sequence <cite>s</cite>.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>s</cite> (str)</dt><dd><p>Sequence to reverse complement.</p>
</dd>
<dt><cite>use_cutils</cite> (bool)</dt><dd><p>Use the faster implementation in the <cite>_cutils</cite> module.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>Reverse complement of <cite>s</cite> as a str.</p>
</dd>
</dl>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="s1">&#39;ATGCAAN&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">reverseComplement</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="o">==</span> <span class="s1">&#39;NTTGCAT&#39;</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.sessionInfo">
<code class="sig-prename descclassname">dms_tools2.utils.</code><code class="sig-name descname">sessionInfo</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/utils.html#sessionInfo"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.utils.sessionInfo" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns string with information about session / packages.</p>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.sigFigStr">
<code class="sig-prename descclassname">dms_tools2.utils.</code><code class="sig-name descname">sigFigStr</code><span class="sig-paren">(</span><em class="sig-param">x</em>, <em class="sig-param">nsig</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/utils.html#sigFigStr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.utils.sigFigStr" title="Permalink to this definition">¶</a></dt>
<dd><p>Get str of <cite>x</cite> with <cite>nsig</cite> significant figures.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sigFigStr</span><span class="p">(</span><span class="mi">11190</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="go">&#39;11000&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sigFigStr</span><span class="p">(</span><span class="mi">117</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="go">&#39;120&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sigFigStr</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="go">&#39;6.0&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sigFigStr</span><span class="p">(</span><span class="mf">0.213</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="go">&#39;0.21&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sigFigStr</span><span class="p">(</span><span class="mf">0.007517</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
<span class="go">&#39;0.00752&#39;</span>
</pre></div>
</div>
</dd></dl>

</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">dms_tools2</h1>
    
  </a>
</p>



<p class="blurb">Tools for analyzing deep mutational scanning data.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=dms_tools2&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a class="badge" href="https://travis-ci.org/jbloomlab/dms_tools2">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="bcsubamp.html">Barcoded-subamplicon sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="prefs.html">Amino-acid preferences</a></li>
<li class="toctree-l1"><a class="reference internal" href="diffsel.html">Differential selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="fracsurvive.html">Fraction surviving</a></li>
<li class="toctree-l1"><a class="reference internal" href="programs.html">Executable programs</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="api.html">Python API</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="dms_tools2.html">dms_tools2 package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="citations.html">Citations</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="api.html">Python API</a><ul>
  <li><a href="dms_tools2.html">dms_tools2 package</a><ul>
      <li>Previous: <a href="dms_tools2.syn_selection.html" title="previous chapter">syn_selection</a></li>
      <li>Next: <a href="citations.html" title="next chapter">Citations</a></li>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017--2020, the Bloom lab.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/dms_tools2.utils.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/dms_tools2" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>