
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>dms_tools2 package &#8212; dms_tools2 2.0a0 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '2.0a0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Citations" href="citations.html" />
    <link rel="prev" title="Python API" href="api.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="dms-tools2-package">
<h1>dms_tools2 package<a class="headerlink" href="#dms-tools2-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="module-dms_tools2.diffsel">
<span id="submodules"></span><h2>Submodules<a class="headerlink" href="#module-dms_tools2.diffsel" title="Permalink to this headline">¶</a></h2>
<div class="section" id="diffsel">
<h3>diffsel<a class="headerlink" href="#diffsel" title="Permalink to this headline">¶</a></h3>
<p>Performs operations related to estimating differential selection.</p>
<dl class="function">
<dt id="dms_tools2.diffsel.computeMutDiffSel">
<code class="descclassname">dms_tools2.diffsel.</code><code class="descname">computeMutDiffSel</code><span class="sig-paren">(</span><em>sel</em>, <em>mock</em>, <em>countcharacters</em>, <em>pseudocount</em>, <em>translate_to_aa</em>, <em>err=None</em>, <em>mincount=0</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.diffsel.computeMutDiffSel" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute mutation differential selection.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>sel</cite> (pandas.DataFrame)</dt>
<dd>Counts for selected sample. Columns should be
<cite>site</cite>, <cite>wildtype</cite>, and every character in
<cite>countcharacters</cite>.</dd>
<dt><cite>mock</cite> (pandas.DataFrame)</dt>
<dd>Like <cite>sel</cite> but counts for mock-selected sample.</dd>
<dt><cite>countcharacters</cite> (list)</dt>
<dd>List of all characters (e.g., codons).</dd>
<dt><cite>pseudocount</cite> (int or float &gt; 0)</dt>
<dd>Pseudocount to add to counts.</dd>
<dt><cite>translate_to_aa</cite> (bool)</dt>
<dd>Should be <cite>True</cite> if counts are for codons and we are
estimating diffsel for amino acids, <cite>False</cite> otherwise.</dd>
<dt><cite>err</cite> (pandas.DataFrame or <cite>None</cite>)</dt>
<dd>Optional error-control counts, in same format as <cite>sel</cite>.</dd>
<dt><cite>mincount</cite> (int &gt;= 0)</dt>
<dd>Report as <cite>NaN</cite> the mutdiffsel for any mutation in which
neither <cite>sel</cite> nor <cite>mock</cite> has at least this many counts.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>A <cite>pandas.DataFrame</cite> with the mutation differential selection.
Columns are <cite>site</cite>, <cite>wildtype</cite>, <cite>mutation</cite>, <cite>mutdiffsel</cite>.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.diffsel.mutToSiteDiffSel">
<code class="descclassname">dms_tools2.diffsel.</code><code class="descname">mutToSiteDiffSel</code><span class="sig-paren">(</span><em>mutdiffsel</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.diffsel.mutToSiteDiffSel" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes sitediffsel from mutdiffsel.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>mutdiffsel</cite> (pandas.DataFrame)</dt>
<dd>Dataframe with mutdiffsel as returned by <cite>computeMutDiffSel</cite></dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd><dl class="first last docutils">
<dt>The dataframe <cite>sitediffsel</cite>, which has the following columns:</dt>
<dd><ul class="first last simple">
<li><cite>site</cite></li>
<li><cite>abs_diffsel</cite>: sum of absolute values of mutdiffsel at site</li>
<li><cite>positive_diffsel</cite>: sum of positive mutdiffsel at site</li>
<li><cite>negative_diffsel</cite>: sum of negative mutdiffsel at site</li>
<li><cite>max_diffsel</cite>: maximum mutdiffsel at site</li>
<li><cite>min_diffsel</cite>: minimum mutdiffsel at site</li>
</ul>
</dd>
</dl>
</dd>
</dl>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">mutdiffsel</span> <span class="o">=</span> <span class="p">(</span><span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
<span class="gp">... </span>        <span class="s1">&#39;site&#39;</span><span class="p">:[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;wildtype&#39;</span><span class="p">:[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;A&#39;</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;A&#39;</span><span class="p">:[</span><span class="n">numpy</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="mf">4.1</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">nan</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;C&#39;</span><span class="p">:[</span><span class="o">-</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;G&#39;</span><span class="p">:[</span><span class="mf">3.2</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.2</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;T&#39;</span><span class="p">:[</span><span class="o">-</span><span class="mf">0.2</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.4</span><span class="p">],</span>
<span class="gp">... </span>        <span class="p">})</span>
<span class="gp">... </span>        <span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;site&#39;</span><span class="p">,</span> <span class="s1">&#39;wildtype&#39;</span><span class="p">],</span>
<span class="gp">... </span>              <span class="n">var_name</span><span class="o">=</span><span class="s1">&#39;mutation&#39;</span><span class="p">,</span> <span class="n">value_name</span><span class="o">=</span><span class="s1">&#39;mutdiffsel&#39;</span><span class="p">)</span>
<span class="gp">... </span>        <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">... </span>        <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sitediffsel</span> <span class="o">=</span> <span class="n">mutToSiteDiffSel</span><span class="p">(</span><span class="n">mutdiffsel</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">sitediffsel</span><span class="o">.</span><span class="n">columns</span> <span class="o">==</span> <span class="p">[</span><span class="s1">&#39;site&#39;</span><span class="p">,</span> <span class="s1">&#39;abs_diffsel&#39;</span><span class="p">,</span> 
<span class="gp">... </span>        <span class="s1">&#39;positive_diffsel&#39;</span><span class="p">,</span> <span class="s1">&#39;negative_diffsel&#39;</span><span class="p">,</span> 
<span class="gp">... </span>        <span class="s1">&#39;max_diffsel&#39;</span><span class="p">,</span> <span class="s1">&#39;min_diffsel&#39;</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">sitediffsel</span><span class="p">[</span><span class="s1">&#39;site&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">sitediffsel</span><span class="p">[</span><span class="s1">&#39;abs_diffsel&#39;</span><span class="p">],</span> <span class="p">[</span><span class="mf">3.6</span><span class="p">,</span> <span class="mf">4.2</span><span class="p">,</span> <span class="mf">0.4</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">sitediffsel</span><span class="p">[</span><span class="s1">&#39;positive_diffsel&#39;</span><span class="p">],</span> <span class="p">[</span><span class="mf">3.2</span><span class="p">,</span> <span class="mf">4.2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">sitediffsel</span><span class="p">[</span><span class="s1">&#39;negative_diffsel&#39;</span><span class="p">],</span> <span class="p">[</span><span class="o">-</span><span class="mf">0.4</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.4</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">sitediffsel</span><span class="p">[</span><span class="s1">&#39;max_diffsel&#39;</span><span class="p">],</span> <span class="p">[</span><span class="mf">3.2</span><span class="p">,</span> <span class="mf">4.1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mf">0.4</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">sitediffsel</span><span class="p">[</span><span class="s1">&#39;min_diffsel&#39;</span><span class="p">],</span> <span class="p">[</span><span class="o">-</span><span class="mf">0.2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.2</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

</div>
<span class="target" id="module-dms_tools2.dssp"></span><div class="section" id="dssp">
<h3>dssp<a class="headerlink" href="#dssp" title="Permalink to this headline">¶</a></h3>
<p>Process output from <a class="reference external" href="http://swift.cmbi.ru.nl/gv/dssp/">dssp</a>.</p>
<p><a class="reference external" href="http://swift.cmbi.ru.nl/gv/dssp/">dssp</a> can be used to calculate
secondary structure and solvent accessibility information from a 
PDB structure. This module can process that output.</p>
<dl class="data">
<dt id="dms_tools2.dssp.MAX_ASA_TIEN">
<code class="descclassname">dms_tools2.dssp.</code><code class="descname">MAX_ASA_TIEN</code><em class="property"> = {'A': 129.0, 'C': 167.0, 'Q': 225.0, 'F': 240.0, 'N': 195.0, 'L': 201.0, 'H': 224.0, 'K': 236.0, 'Y': 263.0, 'P': 159.0, 'W': 285.0, 'I': 197.0, 'R': 274.0, 'M': 224.0, 'G': 104.0, 'E': 223.0, 'D': 193.0, 'V': 174.0, 'T': 172.0, 'S': 155.0}</em><a class="headerlink" href="#dms_tools2.dssp.MAX_ASA_TIEN" title="Permalink to this definition">¶</a></dt>
<dd><p>max accessible surface area (square angstroms) for amino acids, from
<a class="reference external" href="https://doi.org/10.1371/journal.pone.0080635">Tien et al (2013)</a>.
<cite>MAX_ASA_TIEN[a]</cite> is the max surface area for amino acid <cite>a</cite>.</p>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.dssp.processDSSP">
<code class="descclassname">dms_tools2.dssp.</code><code class="descname">processDSSP</code><span class="sig-paren">(</span><em>dsspfile</em>, <em>chain=None</em>, <em>max_asa={'A': 129.0</em>, <em>'C': 167.0</em>, <em>'Q': 225.0</em>, <em>'F': 240.0</em>, <em>'N': 195.0</em>, <em>'L': 201.0</em>, <em>'H': 224.0</em>, <em>'K': 236.0</em>, <em>'Y': 263.0</em>, <em>'P': 159.0</em>, <em>'W': 285.0</em>, <em>'I': 197.0</em>, <em>'R': 274.0</em>, <em>'M': 224.0</em>, <em>'G': 104.0</em>, <em>'E': 223.0</em>, <em>'D': 193.0</em>, <em>'V': 174.0</em>, <em>'T': 172.0</em>, <em>'S': 155.0}</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.dssp.processDSSP" title="Permalink to this definition">¶</a></dt>
<dd><p>Get secondary structure and solvent accessibility from <code class="docutils literal"><span class="pre">dssp</span></code>.</p>
<p><a class="reference external" href="http://swift.cmbi.ru.nl/gv/dssp/">dssp</a> is a program
that calculates secondary structure and absolute solvent
accessibility from a PDB file.</p>
<p>This function processes the text output provided by the
<a class="reference external" href="http://swift.cmbi.ru.nl/gv/dssp/">dssp webserver</a>, at
least given the format of that output as of Sept-4-2017.</p>
<p>It returns a <cite>pandas.DataFrame</cite> that gives the secondary
structure and solvent accessibility for each residue in the
<code class="docutils literal"><span class="pre">dssp</span></code> output.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>dsspfile</cite> (str)</dt>
<dd>Name of text file containing <code class="docutils literal"><span class="pre">dssp</span></code> output.</dd>
<dt><cite>chain</cite> (str or <cite>None</cite>)</dt>
<dd>If the PDB file analyzed by <code class="docutils literal"><span class="pre">dssp</span></code> to create <cite>dsspfile</cite>
has more than one chain, specify the letter code for one
of those chains with this argument.</dd>
<dt><cite>max_asa</cite> (dict)</dt>
<dd>Max surface area for each amino acid in square angstroms.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd><p class="first">A <cite>pandas.DataFrame</cite> with the following columns:</p>
<blockquote class="last">
<div><ul>
<li><p class="first"><cite>site</cite>: residue number for all sites in <cite>dsspfile</cite>.</p>
</li>
<li><p class="first"><cite>amino_acid</cite>: amino acid identity of site in <cite>dsspfile</cite>.</p>
</li>
<li><p class="first"><cite>ASA</cite>: absolute solvent accessibility of the residue.</p>
</li>
<li><p class="first"><cite>RSA</cite>: relative solvent accessibility of the residue.</p>
</li>
<li><p class="first"><cite>SS</cite>: <code class="docutils literal"><span class="pre">dssp</span></code> secondary structure code, one of:</p>
<blockquote>
<div><ul class="simple">
<li><em>G</em>: 3-10 helix</li>
<li><em>H</em>: alpha helix</li>
<li><em>I</em>: pi helix</li>
<li><em>B</em>: beta bridge</li>
<li><em>E</em>: beta bulge</li>
<li><em>T</em>: turn</li>
<li><em>S</em>: high curvature</li>
<li><em>-</em>: loop</li>
</ul>
</div></blockquote>
</li>
<li><p class="first"><cite>SS_class</cite>: broader secondary structure class:</p>
<blockquote>
<div><ul class="simple">
<li><em>helix</em>: <cite>SS</cite> value of <em>G</em>, <em>H</em>, or <em>I</em></li>
<li><em>strand</em>: <cite>SS</cite> value of <em>B</em> or <em>E</em></li>
<li><em>loop</em>: any of the other <cite>SS</cite> values.</li>
</ul>
</div></blockquote>
</li>
</ul>
</div></blockquote>
</dd>
</dl>
</dd></dl>

</div>
<span class="target" id="module-dms_tools2.ipython_utils"></span><div class="section" id="ipython-utils">
<h3>ipython_utils<a class="headerlink" href="#ipython-utils" title="Permalink to this headline">¶</a></h3>
<p>Utilities for using <code class="docutils literal"><span class="pre">dms_tools2</span></code> in an <cite>IPython</cite> (<cite>Jupyter</cite>) notebook.</p>
<dl class="function">
<dt id="dms_tools2.ipython_utils.showPDF">
<code class="descclassname">dms_tools2.ipython_utils.</code><code class="descname">showPDF</code><span class="sig-paren">(</span><em>pdfs</em>, <em>width=None</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.ipython_utils.showPDF" title="Permalink to this definition">¶</a></dt>
<dd><p>Displays PDF images side-by-side in <cite>IPython</cite> notebook.</p>
<p>This function is useful over just directly displaying the PDF
as many web browsers do not show inline PDFs. This function
converts the PDF to PNG and displays that. It requires that
ImageMagick <code class="docutils literal"><span class="pre">convert</span></code> be available for the conversion.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>pdfs</cite> (str or list)</dt>
<dd>Filename of a PDF, or a list of such filenames.
Multiple images are displayed side-by-side.</dd>
<dt><cite>width</cite> (float or int)</dt>
<dd>Width of the displayed PDF.</dd>
</dl>
</dd>
</dl>
</dd></dl>

</div>
<span class="target" id="module-dms_tools2.parseargs"></span><div class="section" id="parseargs">
<h3>parseargs<a class="headerlink" href="#parseargs" title="Permalink to this headline">¶</a></h3>
<p>Argument parsing for the executable scripts in <code class="docutils literal"><span class="pre">dms_tools2</span></code>.</p>
<dl class="function">
<dt id="dms_tools2.parseargs.batch_bcsubampParser">
<code class="descclassname">dms_tools2.parseargs.</code><code class="descname">batch_bcsubampParser</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.parseargs.batch_bcsubampParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <cite>argparse.ArgumentParser</cite> for <code class="docutils literal"><span class="pre">dms2_batch_bcsubamp</span></code>.</p>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.parseargs.batch_diffselParser">
<code class="descclassname">dms_tools2.parseargs.</code><code class="descname">batch_diffselParser</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.parseargs.batch_diffselParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <cite>argparse.ArgumentParser</cite> for <code class="docutils literal"><span class="pre">dms2_batch_diffsel</span></code>.</p>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.parseargs.batch_prefsParser">
<code class="descclassname">dms_tools2.parseargs.</code><code class="descname">batch_prefsParser</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.parseargs.batch_prefsParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <cite>argparse.ArgumentParser</cite> for <code class="docutils literal"><span class="pre">dms2_batch_prefs</span></code>.</p>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.parseargs.bcsubampParentParser">
<code class="descclassname">dms_tools2.parseargs.</code><code class="descname">bcsubampParentParser</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.parseargs.bcsubampParentParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Parent parser for <code class="docutils literal"><span class="pre">dms2_bcsubamp</span></code> / <code class="docutils literal"><span class="pre">dms2_batch_bcsubamp</span></code>.</p>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.parseargs.bcsubampParser">
<code class="descclassname">dms_tools2.parseargs.</code><code class="descname">bcsubampParser</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.parseargs.bcsubampParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <cite>argparse.ArgumentParser</cite> for <code class="docutils literal"><span class="pre">dms2_bcsubamp</span></code>.</p>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.parseargs.checkName">
<code class="descclassname">dms_tools2.parseargs.</code><code class="descname">checkName</code><span class="sig-paren">(</span><em>name</em>, <em>nametype</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.parseargs.checkName" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if <cite>name</cite> is an allowable name.</p>
<p>Allowed names can contain most characters but <strong>not</strong>
LaTex special characters.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>name</cite> (str)</dt>
<dd>Name to check</dd>
<dt><cite>nametype</cite> (str)</dt>
<dd>If we print an exception what do we call the parameter
that failed? For instance, <cite>name</cite> or <cite>group</cite>.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>Returns <cite>True</cite> if <cite>name</cite> is an allowable name.
Otherwise raises a <cite>ValueError</cite> explaining why the
<cite>name</cite> is invalid.</dd>
</dl>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">checkName</span><span class="p">(</span><span class="s1">&#39;sample-1&#39;</span><span class="p">,</span> <span class="s1">&#39;name&#39;</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">checkName</span><span class="p">(</span><span class="s1">&#39;sample 1&#39;</span><span class="p">,</span> <span class="s1">&#39;name&#39;</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">checkName</span><span class="p">(</span><span class="s1">&#39;PGT151 - 5 nM&#39;</span><span class="p">,</span> <span class="s1">&#39;name&#39;</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">checkName</span><span class="p">(</span><span class="s1">&#39;sample_1&#39;</span><span class="p">,</span> <span class="s1">&#39;name&#39;</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">ValueError</span>: <span class="n">name sample_1 contains following illegal characters: _</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">checkName</span><span class="p">(</span><span class="s1">&#39;sample_1&#39;</span><span class="p">,</span> <span class="s1">&#39;group&#39;</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">ValueError</span>: <span class="n">group sample_1 contains following illegal characters: _</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.parseargs.diffselParentParser">
<code class="descclassname">dms_tools2.parseargs.</code><code class="descname">diffselParentParser</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.parseargs.diffselParentParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Parent parser for <code class="docutils literal"><span class="pre">dms2_diffsel</span></code> / <code class="docutils literal"><span class="pre">dms2_batch_diffsel</span></code>.</p>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.parseargs.diffselParser">
<code class="descclassname">dms_tools2.parseargs.</code><code class="descname">diffselParser</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.parseargs.diffselParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <cite>argparse.ArgumentParser</cite> for <code class="docutils literal"><span class="pre">dms2_diffsel</span></code>.</p>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.parseargs.logoplotParser">
<code class="descclassname">dms_tools2.parseargs.</code><code class="descname">logoplotParser</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.parseargs.logoplotParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <cite>argparse.ArgumentParser</cite> for <code class="docutils literal"><span class="pre">dms2_logoplot</span></code>.</p>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.parseargs.parentParser">
<code class="descclassname">dms_tools2.parseargs.</code><code class="descname">parentParser</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.parseargs.parentParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns parent parser with some common options added.</p>
<dl class="docutils">
<dt>Returns:</dt>
<dd><p class="first"><cite>argparse.ArgumentParser</cite> with the following arguments
already added:</p>
<blockquote class="last">
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">--outdir</span></code></li>
<li><code class="docutils literal"><span class="pre">--ncpus</span></code></li>
<li><code class="docutils literal"><span class="pre">--use_existing</span></code></li>
<li><code class="docutils literal"><span class="pre">-v</span> <span class="pre">/</span> <span class="pre">--version</span></code></li>
</ul>
</div></blockquote>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.parseargs.parserDescription">
<code class="descclassname">dms_tools2.parseargs.</code><code class="descname">parserDescription</code><span class="sig-paren">(</span><em>description</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.parseargs.parserDescription" title="Permalink to this definition">¶</a></dt>
<dd><p>Augments description with program information.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>description</cite> (str)</dt>
<dd>Description of program</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>A string with <cite>description</cite> augmented with information
on the <cite>dms_tools2</cite> package / version.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.parseargs.prefsParentParser">
<code class="descclassname">dms_tools2.parseargs.</code><code class="descname">prefsParentParser</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.parseargs.prefsParentParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Parent parser for <code class="docutils literal"><span class="pre">dms2_prefs</span></code> / <code class="docutils literal"><span class="pre">dms2_batch_prefs</span></code>.</p>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.parseargs.prefsParser">
<code class="descclassname">dms_tools2.parseargs.</code><code class="descname">prefsParser</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.parseargs.prefsParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <cite>argparse.ArgumentParser</cite> for <code class="docutils literal"><span class="pre">dms2_prefs</span></code>.</p>
</dd></dl>

</div>
<span class="target" id="module-dms_tools2.plot"></span><div class="section" id="plot">
<h3>plot<a class="headerlink" href="#plot" title="Permalink to this headline">¶</a></h3>
<p>Plotting functions for <code class="docutils literal"><span class="pre">dms_tools2</span></code>.</p>
<p>Uses <a class="reference external" href="https://plotnine.readthedocs.io/en/stable">plotnine</a>
and <a class="reference external" href="https://seaborn.pydata.org/index.html">seaborn</a>.</p>
<dl class="data">
<dt id="dms_tools2.plot.COLOR_BLIND_PALETTE">
<code class="descclassname">dms_tools2.plot.</code><code class="descname">COLOR_BLIND_PALETTE</code><em class="property"> = ['#000000', '#E69F00', '#56B4E9', '#009E73', '#F0E442', '#0072B2', '#D55E00', '#CC79A7']</em><a class="headerlink" href="#dms_tools2.plot.COLOR_BLIND_PALETTE" title="Permalink to this definition">¶</a></dt>
<dd><p><a class="reference external" href="http://bconnelly.net/2013/10/creating-colorblind-friendly-figures/">color-blind safe palette</a>
use by adding to your plots the following
<cite>scale_fill_manual(COLOR_BLIND_PALETTE)</cite> or
<cite>scale_color_manual(COLOR_BLIND_PALETTE)</cite>.</p>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.plot.latexSciNot">
<code class="descclassname">dms_tools2.plot.</code><code class="descname">latexSciNot</code><span class="sig-paren">(</span><em>xlist</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.plot.latexSciNot" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts list of numbers to LaTex scientific notation.</p>
<p>Useful for nice axis-tick formatting.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>xlist</cite> (list)</dt>
<dd>Numbers to format.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>List of latex scientific notation formatted strings.</dd>
</dl>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">latexSciNot</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3120</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.0000927</span><span class="p">])</span>
<span class="go">[&#39;$0$&#39;, &#39;$3$&#39;, &#39;$3.1 \\times 10^{3}$&#39;, &#39;$-9.3 \\times 10^{-5}$&#39;]</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">latexSciNot</span><span class="p">([</span><span class="mf">0.001</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1000</span><span class="p">,</span> <span class="mf">1e6</span><span class="p">])</span>
<span class="go">[&#39;$0.001$&#39;, &#39;$1$&#39;, &#39;$10^{3}$&#39;, &#39;$10^{6}$&#39;]</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">latexSciNot</span><span class="p">([</span><span class="o">-</span><span class="mf">0.002</span><span class="p">,</span> <span class="mf">0.003</span><span class="p">,</span> <span class="mf">0.000011</span><span class="p">])</span>
<span class="go">[&#39;$-0.002$&#39;, &#39;$0.003$&#39;, &#39;$1.1 \\times 10^{-5}$&#39;]</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">latexSciNot</span><span class="p">([</span><span class="o">-</span><span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">])</span>
<span class="go">[&#39;$-0.1$&#39;, &#39;$0$&#39;, &#39;$0.1$&#39;, &#39;$0.2$&#39;]</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">latexSciNot</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">])</span>
<span class="go">[&#39;$0$&#39;, &#39;$1$&#39;, &#39;$2$&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.plot.plotBCStats">
<code class="descclassname">dms_tools2.plot.</code><code class="descname">plotBCStats</code><span class="sig-paren">(</span><em>names</em>, <em>bcstatsfiles</em>, <em>plotfile</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.plot.plotBCStats" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots <code class="docutils literal"><span class="pre">dms2_bcsubamp</span></code> barcode statistics for set of samples.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>names</cite> (list or series)</dt>
<dd>Names of the samples for which we are plotting statistics.</dd>
<dt><cite>bcstatsfiles</cite> (list or series)</dt>
<dd>Names of <code class="docutils literal"><span class="pre">*_bcstats.csv</span></code> files created by <code class="docutils literal"><span class="pre">dms2_bcsubamp</span></code>.</dd>
<dt><cite>plotfile</cite> (str)</dt>
<dd>Name of PDF plot file to create.</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.plot.plotCodonMutTypes">
<code class="descclassname">dms_tools2.plot.</code><code class="descname">plotCodonMutTypes</code><span class="sig-paren">(</span><em>names</em>, <em>countsfiles</em>, <em>plotfile</em>, <em>classification='aachange'</em>, <em>csvfile=None</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.plot.plotCodonMutTypes" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot average frequency codon mutation types.</p>
<p>The averages are determined by summing counts for all sites.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>names</cite> (list or series)</dt>
<dd>Names of samples for which we plot statistics.</dd>
<dt><cite>countsfiles</cite> (list or series)</dt>
<dd><code class="docutils literal"><span class="pre">*_codoncounts.csv</span></code> files of type created by <code class="docutils literal"><span class="pre">dms2_bcsubamp</span></code>.</dd>
<dt><cite>plotfile</cite> (str)</dt>
<dd>Name of created PDF plot file.</dd>
<dt><cite>classification</cite> (str)</dt>
<dd><p class="first">The method used to classify the mutation types. Can be:</p>
<blockquote class="last">
<div><p><cite>aachange</cite> : stop, synonymous, nonsynonymous</p>
<p><cite>n_ntchanges</cite> : number of nucleotide changes per codon</p>
<p><cite>singlentchanges</cite> : nucleotide change in 1-nt mutations</p>
</div></blockquote>
</dd>
<dt><cite>csvfile</cite> (str or <cite>None</cite>)</dt>
<dd><cite>None</cite> or name of CSV file to which numerical data are written.</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.plot.plotCorrMatrix">
<code class="descclassname">dms_tools2.plot.</code><code class="descname">plotCorrMatrix</code><span class="sig-paren">(</span><em>names</em>, <em>infiles</em>, <em>plotfile</em>, <em>datatype</em>, <em>trim_unshared_sites=True</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.plot.plotCorrMatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots correlations among replicates.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>names</cite> (list or series)</dt>
<dd>Names of samples for which we plot statistics.</dd>
<dt><cite>infiles</cite> (list or series)</dt>
<dd>CSV files containing data. Format depends on <cite>datatype</cite>.</dd>
<dt><cite>plotfile</cite> (str)</dt>
<dd>Name of created PDF plot file.</dd>
<dt><cite>datatype</cite> (str)</dt>
<dd><dl class="first last docutils">
<dt>Type of data for which we are plotting correlations:</dt>
<dd><ul class="first last simple">
<li><cite>prefs</cite>: in format returned by <code class="docutils literal"><span class="pre">dms2_prefs</span></code></li>
</ul>
</dd>
</dl>
</dd>
<dt><cite>trim_unshared_sites</cite> (bool)</dt>
<dd>What if files in <cite>infiles</cite> don’t all have same sites?
If <cite>True</cite>, trim unshared sites and just analyze ones
shared among all files. If <cite>False</cite>, raise an error
if unshared sites.</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.plot.plotCumulMutCounts">
<code class="descclassname">dms_tools2.plot.</code><code class="descname">plotCumulMutCounts</code><span class="sig-paren">(</span><em>names</em>, <em>countsfiles</em>, <em>plotfile</em>, <em>chartype</em>, <em>nmax=15</em>, <em>maxcol=6</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.plot.plotCumulMutCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot fraction of mutations seen &lt;= some number of times.</p>
<p>For each set of counts in <cite>countsfiles</cite>, plot the fraction
of mutations seen greater than or equal to some number of
times. This is essentially a cumulative fraction plot.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>names</cite> (list or series)</dt>
<dd>Names of samples for which we plot statistics.</dd>
<dt><cite>countsfiles</cite> (list or series)</dt>
<dd><code class="docutils literal"><span class="pre">*_codoncounts.csv</span></code> files of type created by <code class="docutils literal"><span class="pre">dms2_bcsubamp</span></code>.</dd>
<dt><cite>plotfile</cite> (str)</dt>
<dd>Name of created PDF plot file.</dd>
<dt><cite>chartype</cite> (str)</dt>
<dd><p class="first">The type of character in <cite>countsfiles</cite>.</p>
<blockquote class="last">
<div><ul class="simple">
<li><cite>codon</cite></li>
</ul>
</div></blockquote>
</dd>
<dt><cite>nmax</cite> (int)</dt>
<dd>Plot out to this number of mutation occurrences.</dd>
<dt><cite>maxcol</cite> (int)</dt>
<dd>Number of columns in faceted plot.</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.plot.plotDepth">
<code class="descclassname">dms_tools2.plot.</code><code class="descname">plotDepth</code><span class="sig-paren">(</span><em>names</em>, <em>countsfiles</em>, <em>plotfile</em>, <em>maxcol=4</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.plot.plotDepth" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot sequencing depth along primary sequence.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>names</cite> (list or series)</dt>
<dd>Names of samples for which we plot statistics.</dd>
<dt><cite>countsfiles</cite> (list or series)</dt>
<dd><code class="docutils literal"><span class="pre">*_codoncounts.csv</span></code> files of type created by <code class="docutils literal"><span class="pre">dms2_bcsubamp</span></code>.</dd>
<dt><cite>plotfile</cite> (str)</dt>
<dd>Name of created PDF plot file containing count depth.</dd>
<dt><cite>maxcol</cite> (int)</dt>
<dd>Number of columns in faceted plot.</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.plot.plotMutFreq">
<code class="descclassname">dms_tools2.plot.</code><code class="descname">plotMutFreq</code><span class="sig-paren">(</span><em>names</em>, <em>countsfiles</em>, <em>plotfile</em>, <em>maxcol=4</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.plot.plotMutFreq" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot mutation frequency along primary sequence.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>names</cite> (list or series)</dt>
<dd>Names of samples for which we plot statistics.</dd>
<dt><cite>countsfiles</cite> (list or series)</dt>
<dd><code class="docutils literal"><span class="pre">*_codoncounts.csv</span></code> files of type created by <code class="docutils literal"><span class="pre">dms2_bcsubamp</span></code>.</dd>
<dt><cite>plotfile</cite> (str)</dt>
<dd>Name of created PDF plot file.</dd>
<dt><cite>maxcol</cite> (int)</dt>
<dd>Number of columns in faceted plot.</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.plot.plotReadStats">
<code class="descclassname">dms_tools2.plot.</code><code class="descname">plotReadStats</code><span class="sig-paren">(</span><em>names</em>, <em>readstatfiles</em>, <em>plotfile</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.plot.plotReadStats" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots <code class="docutils literal"><span class="pre">dms2_bcsubamp</span></code> read statistics for a set of samples.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>names</cite> (list or series)</dt>
<dd>Names of the samples for which we are plotting statistics.</dd>
<dt><cite>readstatfiles</cite> (list or series)</dt>
<dd>Names of <code class="docutils literal"><span class="pre">*_readstats.csv</span></code> files created by <code class="docutils literal"><span class="pre">dms2_bcsubamp</span></code>.</dd>
<dt><cite>plotfile</cite> (str)</dt>
<dd>Name of PDF plot file to create.</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.plot.plotReadsPerBC">
<code class="descclassname">dms_tools2.plot.</code><code class="descname">plotReadsPerBC</code><span class="sig-paren">(</span><em>names</em>, <em>readsperbcfiles</em>, <em>plotfile</em>, <em>maxreads=10</em>, <em>maxcol=6</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.plot.plotReadsPerBC" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots <code class="docutils literal"><span class="pre">dms2_bcsubamp</span></code> reads-per-barcode stats for set of samples.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>names</cite> (list or series)</dt>
<dd>Names of samples for which we plot statistics.</dd>
<dt><cite>readsperbcfiles</cite> (list or series)</dt>
<dd>Names of <code class="docutils literal"><span class="pre">*_readsperbc.csv</span></code> files created by <code class="docutils literal"><span class="pre">dms2_bcsubamp</span></code>.</dd>
<dt><cite>plotfile</cite> (str)</dt>
<dd>Name of PDF plot file to create.</dd>
<dt><cite>maxreads</cite> (int)</dt>
<dd>For any barcodes with &gt; this many reads, just make a category
of &gt;= this.</dd>
<dt><cite>maxcol</cite> (int)</dt>
<dd>Number of columns in faceted plot.</dd>
</dl>
</dd>
</dl>
</dd></dl>

</div>
<span class="target" id="module-dms_tools2.prefs"></span><div class="section" id="prefs">
<h3>prefs<a class="headerlink" href="#prefs" title="Permalink to this headline">¶</a></h3>
<p>Performs operations related to estimating site-specific amino-acid
preferences.</p>
<p>Uses <a class="reference external" href="https://pystan.readthedocs.io/en/latest">pystan</a>
to perform MCMC for Bayesian inferences.</p>
<dl class="data">
<dt id="dms_tools2.prefs.PRIOR_MIN_VALUE">
<code class="descclassname">dms_tools2.prefs.</code><code class="descname">PRIOR_MIN_VALUE</code><em class="property"> = 1e-07</em><a class="headerlink" href="#dms_tools2.prefs.PRIOR_MIN_VALUE" title="Permalink to this definition">¶</a></dt>
<dd><p>minimum value for Dirichlet prior elements</p>
</dd></dl>

<dl class="class">
<dt id="dms_tools2.prefs.StanModelDifferentErr">
<em class="property">class </em><code class="descclassname">dms_tools2.prefs.</code><code class="descname">StanModelDifferentErr</code><span class="sig-paren">(</span><em>verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.prefs.StanModelDifferentErr" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p><code class="docutils literal"><span class="pre">pystan</span></code> model when <cite>error_model</cite> is <cite>different</cite>.</p>
<p>For use by inferSitePrefs`.</p>
</dd></dl>

<dl class="class">
<dt id="dms_tools2.prefs.StanModelNoneErr">
<em class="property">class </em><code class="descclassname">dms_tools2.prefs.</code><code class="descname">StanModelNoneErr</code><span class="sig-paren">(</span><em>verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.prefs.StanModelNoneErr" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p><code class="docutils literal"><span class="pre">pystan</span></code> model when <cite>error_model</cite> is <cite>none</cite>.</p>
<p>For use by inferSitePrefs`.</p>
</dd></dl>

<dl class="class">
<dt id="dms_tools2.prefs.StanModelSameErr">
<em class="property">class </em><code class="descclassname">dms_tools2.prefs.</code><code class="descname">StanModelSameErr</code><span class="sig-paren">(</span><em>verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.prefs.StanModelSameErr" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p><code class="docutils literal"><span class="pre">pystan</span></code> model when <cite>error_model</cite> is <cite>same</cite>.</p>
<p>For use by inferSitePrefs`.</p>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.prefs.inferPrefsByRatio">
<code class="descclassname">dms_tools2.prefs.</code><code class="descname">inferPrefsByRatio</code><span class="sig-paren">(</span><em>charlist</em>, <em>sites</em>, <em>wts</em>, <em>pre</em>, <em>post</em>, <em>errpre</em>, <em>errpost</em>, <em>pseudocount</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.prefs.inferPrefsByRatio" title="Permalink to this definition">¶</a></dt>
<dd><p>Site-specific preferences from normalized enrichment ratios.</p>
<p>Calculates site-specific preference <span class="math">\(\pi_{r,a}\)</span> of each
site <span class="math">\(r\)</span> for each character <span class="math">\(a\)</span> using re-normalized
enrichment ratios. This function accomplishes the same goal as
<cite>inferSitePrefs</cite>, but is much simpler and faster, although less
statistically rigorous in principle.</p>
<p>Specifically, this function calculates the preferences as</p>
<div class="math">
\[\pi_{r,a} = \frac{\phi_{r,a}}{\sum_{a'} \phi_{r,a'}}\]</div>
<p>where <span class="math">\(\phi_{r,a}\)</span> is the enrichment ratio of character
<span class="math">\(a\)</span> relative to wildtype after selection.</p>
<p>The actual definition of these enrichment ratios is somewhat
complex due to the need to correct for errors and add a
pseudocount. We have 4 sets of counts:</p>
<blockquote>
<div><ul class="simple">
<li><cite>pre</cite>: pre-selection counts</li>
<li><cite>post</cite>: post-selection counts</li>
<li><cite>errpre</cite>: error-control for pre-selection counts</li>
<li><cite>errpost</cite>: error-control for post-selection counts</li>
</ul>
</div></blockquote>
<p>For each set of counts <span class="math">\(s\)</span>, let <span class="math">\(N_r^s\)</span> (e.g., 
<span class="math">\(N_r^{pre}\)</span>) denote the total counts for this sample at
site <span class="math">\(r\)</span>, and let <span class="math">\(n_{r,a}^{s}\)</span> (e.g., 
<span class="math">\(n_{r,a}^{pre}\)</span>) denote the counts at that site for
character <span class="math">\(a\)</span>.</p>
<p>Because some of the counts are low, we add a pseudocount 
<span class="math">\(P\)</span> to each observation. With this pseudocount, the 
estimated frequency of character <span class="math">\(a\)</span> at site <span class="math">\(r\)</span>
in sample <span class="math">\(s\)</span> is</p>
<div class="math">
\[f_{r,a}^s = \frac{n_{r,a}^s + P}{N_{r,a}^s + A \times P}\]</div>
<p>where <span class="math">\(A\)</span> is the number of characters in the alphabet (e.g.,
20 for amino acids without stop codons).</p>
<p>The <strong>error-corrected</strong> estimates of the frequency of character
<span class="math">\(a\)</span> before and after selection are then</p>
<div class="math">
\[ \begin{align}\begin{aligned}f_{r,a}^{before} &amp;= \max\left(\frac{P}{N_{r,a}^{pre} + A \times P},
\; f_{r,a}^{pre} + \delta_{a,\rm{wt}\left(r\right)}
- f_{r,a}^{errpre}\right)\\f_{r,a}^{after} &amp;= \max\left(\frac{P}{N_{r,a}^{post} + A \times P},
\; f_{r,a}^{post} + \delta_{a,\rm{wt}\left(r\right)}
- f_{r,a}^{errpost}\right)\end{aligned}\end{align} \]</div>
<p>where <span class="math">\(\delta_{a,\rm{wt}\left(r\right)}\)</span> is the 
Kronecker-delta, equal to 1 if <span class="math">\(a\)</span> is the same as the
wildtype character <span class="math">\(\rm{wt}\left(r\right)\)</span> at site
<span class="math">\(r\)</span>, and 0 otherwise. We use the <span class="math">\(\max\)</span> operator
to ensure that even if the error-control frequency exceeds
the actual estimate, our estimated frequency never drops
below the pseudocount over the depth of the uncorrected sample.</p>
<p>Given these definitions, we then simply calculate the enrichment 
ratios as</p>
<div class="math">
\[\phi_{r,a} = \frac{\left(f_{r,a}^{after}\right) / 
\left(f_{r,\rm{wt}\left(r\right)}^{after}\right)}
{\left(f_{r,a}^{before}\right) / 
\left(f_{r,\rm{wt}\left(r\right)}^{before}\right)}\]</div>
<p>In the case where we are <strong>not</strong> using any error-controls, then
we simply set 
<span class="math">\(f_{r,\rm{wt}}^{errpre} = f_{r,\rm{wt}}^{errpost} 
= \delta_{a,\rm{wt}\left(r\right)}\)</span>.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>charlist</cite> (list)</dt>
<dd>Valid characters (e.g., codons, amino acids, nts).</dd>
<dt><cite>sites</cite> (list)</dt>
<dd>Sites to analyze.</dd>
<dt><cite>wts</cite> (list)</dt>
<dd><cite>wts[r]</cite> is the wildtype character at site <cite>sites[r]</cite>.</dd>
<dt><cite>pre</cite> (pandas.DataFrame)</dt>
<dd>Gives pre-selection counts. Should have columns
with names of ‘site’ and all characters in <cite>charlist</cite>.
The rows give the counts of each character at that site.</dd>
<dt><cite>post</cite> (pandas.DataFrame)</dt>
<dd>Like <cite>pre</cite> but for post-selection counts.</dd>
<dt><cite>errpre</cite> (<cite>None</cite> or pandas.DataFrame)</dt>
<dd>Like <cite>pre</cite> but for pre-selection error-control counts,
or <cite>None</cite> if there is no such control.</dd>
<dt><cite>errpost</cite> (<cite>None</cite> or pandas.DataFrame)</dt>
<dd>Like <cite>pre</cite> but for post-selection error-control counts,
or <cite>None</cite> if there is no such control.</dd>
<dt><cite>pseudocount</cite> (float or int)</dt>
<dd>The pseudocount to add to each observation.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>A pandas.DataFrame holding the preferences. The columns of
this dataframe are ‘site’ and all characters in <cite>charlist</cite>.
For each site in <cite>sites</cite>, the rows give the preference
for that character.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.prefs.inferSitePrefs">
<code class="descclassname">dms_tools2.prefs.</code><code class="descname">inferSitePrefs</code><span class="sig-paren">(</span><em>charlist</em>, <em>wtchar</em>, <em>error_model</em>, <em>counts</em>, <em>priors</em>, <em>seed=1</em>, <em>niter=10000</em>, <em>increasetries=5</em>, <em>n_jobs=1</em>, <em>r_max=1.1</em>, <em>neff_min=100</em>, <em>nchains=4</em>, <em>increasefac=2</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.prefs.inferSitePrefs" title="Permalink to this definition">¶</a></dt>
<dd><p>Infers site-specific preferences by MCMC for a specific site.</p>
<p>Infer the site-specific preferences <span class="math">\(\pi_{r,a}\)</span> for some site
<span class="math">\(r\)</span> for each character <span class="math">\(a\)</span> by integrating over the 
posterior defined by the product of the following priors and 
likelihoods, where <span class="math">\(\boldsymbol{\mathbf{\pi_r}}\)</span> indicates
the vector of <span class="math">\(\pi_{r,a}\)</span> values for all characters:</p>
<div class="math">
\[ \begin{align}\begin{aligned}\Pr\left(\boldsymbol{\mathbf{\pi_r}}\right) 
&amp;= \mathrm{Dirichlet}\left(\boldsymbol{\mathbf{a_{\pi,r}}}\right)\\\Pr\left(\boldsymbol{\mathbf{\mu_r}}\right) 
&amp;= \mathrm{Dirichlet}\left(\boldsymbol{\mathbf{a_{\mu,r}}}\right)\\\Pr\left(\boldsymbol{\mathbf{\epsilon_r}}\right) 
&amp;= \mathrm{Dirichlet}\left(\boldsymbol{\mathbf{a_{\epsilon,r}}}\right)\\\Pr\left(\boldsymbol{\mathbf{\rho_r}}\right) 
&amp;= \mathrm{Dirichlet}\left(\boldsymbol{\mathbf{a_{\rho,r}}}\right)\\\Pr\left(\boldsymbol{\mathbf{n_r^{\rm{pre}}}} \mid \boldsymbol{\mathbf{\mu_r}}, \boldsymbol{\mathbf{\epsilon_r}}\right) 
&amp;= \mathrm{Multinomial}\left(\boldsymbol{\mathbf{n_r^{\rm{pre}}}}; \boldsymbol{\mathbf{\mu_r}} + \boldsymbol{\mathbf{\epsilon_r}} - \boldsymbol{\mathbf{\delta_r}}\right)\\\Pr\left(\boldsymbol{\mathbf{n_r^{\rm{post}}}} \mid \boldsymbol{\mathbf{\mu_r}}, \boldsymbol{\mathbf{\epsilon_r}}\right) 
&amp;= \mathrm{Multinomial}\left(\boldsymbol{\mathbf{n_r^{\rm{post}}}}; \frac{\boldsymbol{\mathbf{\mu_r}} \circ \boldsymbol{\mathbf{\pi_r}}}{\boldsymbol{\mathbf{\mu_r}} \cdot \boldsymbol{\mathbf{\pi_r}}} + \boldsymbol{\mathbf{\rho_r}} - \boldsymbol{\mathbf{\delta_r}}\right)\\\Pr\left(\boldsymbol{\mathbf{n_r^{\rm{err,pre}}}} \mid \boldsymbol{\mathbf{\epsilon_r}}\right) 
&amp;= \mathrm{Multinomial}\left(\boldsymbol{\mathbf{n_r^{\rm{err,pre}}}}; \boldsymbol{\mathbf{\epsilon_r}}\right)\\\Pr\left(\boldsymbol{\mathbf{n_r^{\rm{err,post}}}} \mid \boldsymbol{\mathbf{\rho_r}}\right) 
&amp;= \mathrm{Multinomial}\left(\boldsymbol{\mathbf{n_r^{\rm{err,post}}}}; \boldsymbol{\mathbf{\rho_r}}\right)\end{aligned}\end{align} \]</div>
<p>where <span class="math">\(\boldsymbol{\mathbf{\delta_r}}\)</span> is a vector that is 
zero except for a one at the element corresponding to the wildtype.</p>
<p>The MCMC tries to guarantee convergence via the parameters specified
by <cite>r_max</cite>, <cite>neff_min</cite>, <cite>nchains</cite>, <cite>niter</cite>, <cite>increasefac</cite>, 
and <cite>increasetries</cite>.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>charlist</cite> (list)</dt>
<dd>List of valid characters (e.g., codons, amino acids, nts).</dd>
<dt><cite>wtchar</cite> (str)</dt>
<dd>Wildtype character at the site.</dd>
<dt><cite>error_model</cite> (str or object)</dt>
<dd><p class="first">Specifies how errors are estimated. Passing error model
objects is faster if calling this function repeatedly
as they will not need to be compiled. Can be:</p>
<blockquote class="last">
<div><ul class="simple">
<li>The str <cite>none</cite> or instance of <cite>StanModelNoneErr</cite>:
no errors (<span class="math">\(\boldsymbol{\mathbf{\epsilon_r}} = \boldsymbol{\mathbf{\rho_r}} = \mathbf{0}\)</span>)</li>
<li>The str <cite>same</cite> or instance of <cite>StanModelSameErr</cite>: 
same error rates  pre and post-selection 
(<span class="math">\(\boldsymbol{\mathbf{\epsilon_r}} = \boldsymbol{\mathbf{\rho_r}}\)</span>).</li>
<li>The str <cite>different</cite> or instance of <cite>StanModelDifferentErr</cite>: 
different error rates pre- and post-selection
(<span class="math">\(\boldsymbol{\mathbf{\epsilon_r}} \ne \boldsymbol{\mathbf{\rho_r}}\)</span>).</li>
</ul>
</div></blockquote>
</dd>
<dt><cite>counts</cite> (dict)</dt>
<dd><p class="first">Deep sequencing counts. Each string key should specify
a dict keyed by all characters in <cite>charlist</cite> with
the values giving integer counts for that character. The keys:</p>
<blockquote class="last">
<div><ul class="simple">
<li><cite>pre</cite>: <span class="math">\(\boldsymbol{\mathbf{n_r^{\rm{pre}}}}\)</span></li>
<li><cite>post</cite>: <span class="math">\(\boldsymbol{\mathbf{n_r^{\rm{post}}}}\)</span></li>
<li><em>err</em>: required if <cite>error_model</cite> is <cite>same</cite>, specifies
<span class="math">\(\boldsymbol{\mathbf{n_r^{\rm{err,pre}}}} = \boldsymbol{\mathbf{n_r^{\rm{err,post}}}}\)</span></li>
<li><cite>errpre</cite> and <cite>errpost</cite>: required if <cite>error_model</cite> is
<cite>different</cite>, specify 
<span class="math">\(\boldsymbol{\mathbf{n_r^{\rm{err,pre}}}}\)</span> and
<span class="math">\(\boldsymbol{\mathbf{n_r^{\rm{err,post}}}}\)</span>.</li>
</ul>
</div></blockquote>
</dd>
<dt><cite>priors</cite> (dict)</dt>
<dd><p class="first">Specifies parameter vectors for Dirichlet priors. Each string
key should specify a dict keyed by all characters and values
giving the prior for that character. Values less than 
<cite>PRIOR_MIN_VALUE</cite> are set to <cite>PRIOR_MIN_VALUE</cite>. Keys are</p>
<blockquote class="last">
<div><ul class="simple">
<li><cite>pir_prior_params</cite>: <span class="math">\(\boldsymbol{\mathbf{a_{\pi,r}}}\)</span></li>
<li><cite>mur_prior_params</cite>: <span class="math">\(\boldsymbol{\mathbf{a_{\mu,r}}}\)</span></li>
<li><cite>epsilonr_prior_params</cite>: only required if <cite>error_model</cite> is 
<cite>same</cite> or <cite>different</cite>, specifies 
<span class="math">\(\boldsymbol{\mathbf{a_{\epsilon,r}}}\)</span></li>
<li><cite>rhor_prior_params</cite> : only required if <cite>error_model</cite> is 
<cite>different</cite>, specifies
<span class="math">\(\boldsymbol{\mathbf{a_{\rho,r}}}\)</span></li>
</ul>
</div></blockquote>
</dd>
<dt><cite>seed</cite> (int)</dt>
<dd>Random number seed for MCMC.</dd>
<dt><cite>n_jobs</cite> (int)</dt>
<dd>Number of CPUs to use, -1 means all available.</dd>
<dt><cite>niter</cite>, <cite>increasetries</cite>, <cite>r_max</cite>, <cite>neff_min</cite>, <cite>nchains</cite>, <cite>increasefac</cite></dt>
<dd>Specify MCMC convergence. They all have reasonable defaults.
The MCMC is considered to have converged if the mean Gelman-Rubin R
statistic (<a class="reference external" href="http://www.jstor.org/stable/2246093">http://www.jstor.org/stable/2246093</a>) over all 
<span class="math">\(\pi_{r,x}\)</span> values &lt;= <cite>r_max</cite> and the mean effective sample
size is &gt;= <cite>neff_min</cite>. The MCMC first runs with <cite>nchains</cite> chains
and <cite>niter</cite> iterations. If it fails to converge, it increases
the iterations by a factor of <cite>increasefac</cite> and tries again,
and repeats until it converges or has tried <cite>increasetries</cite>
times to increase the number of iterations. If the effective
sample size exceeds 3 times <cite>neff_min</cite> then we allow
R to be <cite>1 + 1.5 (r_max - 1)</cite>.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd><dl class="first last docutils">
<dt>The tuple <cite>(converged, pi_means, pi_95credint, logstring)</cite> where:</dt>
<dd><ul class="first last simple">
<li><cite>converged</cite> is <cite>True</cite> if the MCMC converged, <cite>False</cite> otherwise.</li>
<li><dl class="first docutils">
<dt><cite>pi_means</cite> is dict keyed by characters in <cite>charlist</cite></dt>
<dd>with the value giving the <span class="math">\(\pi_{r,a}\)</span>.</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><cite>pi_95credint</cite> is dict keyed by characters, values are 2-tuples</dt>
<dd>giving median-centered credible interval for <span class="math">\(\pi_{r,a}\)</span>.</dd>
</dl>
</li>
<li><cite>logstring</cite> is a string describing MCMC run and convergence.</li>
</ul>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.prefs.prefsToMutEffects">
<code class="descclassname">dms_tools2.prefs.</code><code class="descname">prefsToMutEffects</code><span class="sig-paren">(</span><em>prefs</em>, <em>charlist</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.prefs.prefsToMutEffects" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts amino acid preferences to effects of specific mutations.</p>
<p>If the preference of site <span class="math">\(r\)</span> for amino acid <span class="math">\(a\)</span> is
<span class="math">\(\pi_{r,a}\)</span>, then the estimated effect (e.g., ratio of
enrichment ratios) for mutating that site from <span class="math">\(x\)</span> to
<span class="math">\(y\)</span> is <span class="math">\(\frac{\pi_{r,y}}{\pi_{r,x}}\)</span>. Very small
values indicate disfavored mutations, and very large values
indicate highly favored mutations. The logarithm base 2 of the
expected effects is also a useful measure – negative values
are disfavored mutations, positive values are favored ones.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>prefs</cite> (pandas DataFrame)</dt>
<dd>Preferences to analyze. The columns should be <cite>site</cite> and
a column for every character in <cite>charlist</cite>.</dd>
<dt><cite>charlist</cite> (list)</dt>
<dd>The list of characters that we are analyzing. For instance,
<cite>dms_tools2.AAS</cite> for amino acids.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd><p class="first">A pandas Data Frame where the columns are:</p>
<blockquote class="last">
<div><ul class="simple">
<li><cite>site</cite>: the site</li>
<li><cite>initial</cite>: the initial character (e.g., amino acid)</li>
<li><cite>final</cite>: the final character after the mutation</li>
<li><cite>mutation</cite>: mutation in the string form <cite>A1G</cite></li>
<li><cite>effect</cite>: the effect of the mutation</li>
<li><cite>log2effect</cite>: the log2 of the effect of the mutation.</li>
</ul>
</div></blockquote>
</dd>
</dl>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">charlist</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prefs</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
<span class="gp">... </span>        <span class="s1">&#39;site&#39;</span><span class="p">:[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;A&#39;</span><span class="p">:[</span><span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;C&#39;</span><span class="p">:[</span><span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;G&#39;</span><span class="p">:[</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">],</span>
<span class="gp">... </span>        <span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">effects</span> <span class="o">=</span> <span class="n">prefsToMutEffects</span><span class="p">(</span><span class="n">prefs</span><span class="p">,</span> <span class="n">charlist</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">set</span><span class="p">(</span><span class="n">effects</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">==</span> <span class="p">{</span><span class="s1">&#39;site&#39;</span><span class="p">,</span> <span class="s1">&#39;initial&#39;</span><span class="p">,</span> <span class="s1">&#39;final&#39;</span><span class="p">,</span>
<span class="gp">... </span>        <span class="s1">&#39;mutation&#39;</span><span class="p">,</span> <span class="s1">&#39;effect&#39;</span><span class="p">,</span> <span class="s1">&#39;log2effect&#39;</span><span class="p">}</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">effects</span><span class="p">[</span><span class="n">effects</span><span class="p">[</span><span class="s1">&#39;initial&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;A&#39;</span><span class="p">][</span><span class="s1">&#39;effect&#39;</span><span class="p">],</span>
<span class="gp">... </span>        <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">effects</span><span class="p">[</span><span class="n">effects</span><span class="p">[</span><span class="s1">&#39;initial&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;C&#39;</span><span class="p">][</span><span class="s1">&#39;effect&#39;</span><span class="p">],</span>
<span class="gp">... </span>        <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">effects</span><span class="p">[</span><span class="s1">&#39;effect&#39;</span><span class="p">],</span> <span class="mi">2</span><span class="o">**</span><span class="n">effects</span><span class="p">[</span><span class="s1">&#39;log2effect&#39;</span><span class="p">])</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

</div>
<span class="target" id="module-dms_tools2.utils"></span><div class="section" id="utils">
<h3>utils<a class="headerlink" href="#utils" title="Permalink to this headline">¶</a></h3>
<p>Miscellaneous utilities for <code class="docutils literal"><span class="pre">dms_tools2</span></code>.</p>
<dl class="function">
<dt id="dms_tools2.utils.adjustErrorCounts">
<code class="descclassname">dms_tools2.utils.</code><code class="descname">adjustErrorCounts</code><span class="sig-paren">(</span><em>errcounts</em>, <em>counts</em>, <em>charlist</em>, <em>maxexcess</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.utils.adjustErrorCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Adjust error counts to not greatly exceed counts of interest.</p>
<p>This function is useful when estimating preferences. Under the
model, the error-control should not have a higher rate of error
than the actual sample. However, this could happen if the experimental
data don’t fully meet the assumptions. So this function scales
down the error counts in that case.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>errcounts</cite> (pandas.DataFrame)</dt>
<dd>Holds counts for error control.</dd>
<dt><cite>counts</cite> (pandas.DataFrame)</dt>
<dd>Holds counts for which we are correcting errors.</dd>
<dt><cite>charlist</cite> (list)</dt>
<dd>Characters for which we have counts.</dd>
<dt><cite>maxexcess</cite> (int)</dt>
<dd>Only let error-control counts exceed actual by this much.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>A copy of <cite>errcounts</cite> except for any non-wildtype character,
the maximum frequency of that character is adjusted to be
at most the number predicted by the frequency in <cite>counts</cite>
plus <cite>maxexcess</cite>.</dd>
</dl>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;site&#39;</span><span class="p">:[</span><span class="mi">1</span><span class="p">],</span> <span class="s1">&#39;wildtype&#39;</span><span class="p">:[</span><span class="s1">&#39;A&#39;</span><span class="p">],</span> 
<span class="gp">... </span>        <span class="s1">&#39;A&#39;</span><span class="p">:</span><span class="mi">500</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">:</span><span class="mi">10</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">:</span><span class="mi">40</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">:</span><span class="mi">20</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">errcounts</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;site&#39;</span><span class="p">:[</span><span class="mi">1</span><span class="p">],</span> <span class="s1">&#39;wildtype&#39;</span><span class="p">:[</span><span class="s1">&#39;A&#39;</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;A&#39;</span><span class="p">:</span><span class="mi">250</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">:</span><span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">:</span><span class="mi">30</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">:</span><span class="mi">10</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">charlist</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">errcounts</span> <span class="o">=</span> <span class="n">errcounts</span><span class="p">[[</span><span class="s1">&#39;site&#39;</span><span class="p">,</span> <span class="s1">&#39;wildtype&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">charlist</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adj_errcounts</span> <span class="o">=</span> <span class="n">adjustErrorCounts</span><span class="p">(</span><span class="n">errcounts</span><span class="p">,</span> <span class="n">counts</span><span class="p">,</span> <span class="n">charlist</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">set</span><span class="p">(</span><span class="n">adj_errcounts</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">==</span> <span class="nb">set</span><span class="p">(</span><span class="n">errcounts</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">adj_errcounts</span><span class="p">[</span><span class="s1">&#39;site&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">errcounts</span><span class="p">[</span><span class="s1">&#39;site&#39;</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">adj_errcounts</span><span class="p">[</span><span class="s1">&#39;wildtype&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">errcounts</span><span class="p">[</span><span class="s1">&#39;wildtype&#39;</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="n">adj_errcounts</span><span class="p">[</span><span class="n">adj_errcounts</span><span class="p">[</span><span class="s1">&#39;site&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">][</span><span class="n">charlist</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">... </span>        <span class="o">==</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">250</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">21</span><span class="p">,</span> <span class="mi">10</span><span class="p">]))</span><span class="o">.</span><span class="n">all</span><span class="p">()</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.alignSubamplicon">
<code class="descclassname">dms_tools2.utils.</code><code class="descname">alignSubamplicon</code><span class="sig-paren">(</span><em>refseq</em>, <em>r1</em>, <em>r2</em>, <em>refseqstart</em>, <em>refseqend</em>, <em>maxmuts</em>, <em>maxN</em>, <em>chartype</em>, <em>use_cutils=True</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.utils.alignSubamplicon" title="Permalink to this definition">¶</a></dt>
<dd><p>Try to align subamplicon to reference sequence at defined location.</p>
<p>Tries to align reads <cite>r1</cite> and <cite>r2</cite> to <cite>refseq</cite> at location
specified by <cite>refseqstart</cite> and <cite>refseqend</cite>. Determines how many
sites of type <cite>chartype</cite> have mutations, and if &lt;= <cite>maxmuts</cite> conside
the subamplicon to align if fraction of ambiguous nucleotides &lt;= <cite>maxN</cite>.
In <cite>r1</cite> and <cite>r2</cite>, an <code class="docutils literal"><span class="pre">N</span></code> indicates a non-called ambiguous identity.
If the reads disagree in a region of overlap that is set to <code class="docutils literal"><span class="pre">N</span></code> in
the final subamplicon, but if one read has <code class="docutils literal"><span class="pre">N</span></code> and the other a called
identity, then the called identity is used in the final subamplicon.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>refseq</cite> (str)</dt>
<dd>Sequence to which we align. if <cite>chartype</cite> is ‘codon’,
must be a valid coding (length multiple of 3).</dd>
<dt><cite>r1</cite> (str)</dt>
<dd>The forward sequence to align.</dd>
<dt><cite>r2</cite> (str)</dt>
<dd>The reverse sequence to align. When reverse complemented,
should read backwards in <cite>refseq</cite>.</dd>
<dt><cite>refseqstart</cite> (int)</dt>
<dd>The nucleotide in <cite>refseq</cite> (1, 2, … numbering) where the
first nucleotide in <cite>r1</cite> aligns.</dd>
<dt><cite>refseqend</cite> (int)</dt>
<dd>The nucleotide in <cite>refseq</cite> (1, 2, … numbering) where the
first nucleotide in <cite>r2</cite> aligns (note that <cite>r2</cite> then reads
backwards towards the 5’ end of <cite>refseq</cite>).</dd>
<dt><cite>maxmuts</cite> (int or float)</dt>
<dd>Maximum number of mutations of character <cite>chartype</cite> that
are allowed in the aligned subamplicons from the two reads.</dd>
<dt><cite>maxN</cite> (int or float)</dt>
<dd>Maximum number of nucleotides for which we allow
ambiguous (<code class="docutils literal"><span class="pre">N</span></code>) identities in final subamplicon.</dd>
<dt><cite>chartype</cite> (str)</dt>
<dd>Character type for which we count mutations.
Currently, the only allowable value is ‘codon’.</dd>
<dt><cite>use_cutils</cite> (bool)</dt>
<dd>Use the faster implementation in the <cite>_cutils</cite> module.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>If reads align, return aligned subamplicon as string (of length
<cite>refseqend - refseqstart + 1</cite>). Otherwise return <cite>None</cite>.</dd>
</dl>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">refseq</span> <span class="o">=</span> <span class="s1">&#39;ATGGGGAAA&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">alignSubamplicon</span><span class="p">(</span><span class="n">refseq</span><span class="p">,</span> <span class="s1">&#39;GGGGAA&#39;</span><span class="p">,</span> <span class="s1">&#39;TTTCCC&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">==</span> <span class="s1">&#39;GGGGAAA&#39;</span> 
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">alignSubamplicon</span><span class="p">(</span><span class="n">refseq</span><span class="p">,</span> <span class="s1">&#39;GGGGAA&#39;</span><span class="p">,</span> <span class="s1">&#39;TTTCCC&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">==</span> <span class="kc">False</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">alignSubamplicon</span><span class="p">(</span><span class="n">refseq</span><span class="p">,</span> <span class="s1">&#39;GGGGAT&#39;</span><span class="p">,</span> <span class="s1">&#39;TTTCCC&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">==</span> <span class="kc">False</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">alignSubamplicon</span><span class="p">(</span><span class="n">refseq</span><span class="p">,</span> <span class="s1">&#39;GGGGAT&#39;</span><span class="p">,</span> <span class="s1">&#39;TTTCCC&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">==</span> <span class="s1">&#39;GGGGANA&#39;</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">alignSubamplicon</span><span class="p">(</span><span class="n">refseq</span><span class="p">,</span> <span class="s1">&#39;GGGGAT&#39;</span><span class="p">,</span> <span class="s1">&#39;TATCCC&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">==</span> <span class="s1">&#39;GGGGATA&#39;</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">alignSubamplicon</span><span class="p">(</span><span class="n">refseq</span><span class="p">,</span> <span class="s1">&#39;GGGGAT&#39;</span><span class="p">,</span> <span class="s1">&#39;TATCCC&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">==</span> <span class="kc">False</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">alignSubamplicon</span><span class="p">(</span><span class="n">refseq</span><span class="p">,</span> <span class="s1">&#39;GGGNAA&#39;</span><span class="p">,</span> <span class="s1">&#39;TTTCCC&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">==</span> <span class="s1">&#39;GGGGAAA&#39;</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">alignSubamplicon</span><span class="p">(</span><span class="n">refseq</span><span class="p">,</span> <span class="s1">&#39;GGGNAA&#39;</span><span class="p">,</span> <span class="s1">&#39;TTNCCC&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">==</span> <span class="s1">&#39;GGGGAAA&#39;</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">alignSubamplicon</span><span class="p">(</span><span class="n">refseq</span><span class="p">,</span> <span class="s1">&#39;GTTTAA&#39;</span><span class="p">,</span> <span class="s1">&#39;TTTAAA&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">==</span> <span class="s1">&#39;GTTTAAA&#39;</span> 
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">alignSubamplicon</span><span class="p">(</span><span class="n">refseq</span><span class="p">,</span> <span class="s1">&#39;GGGGTA&#39;</span><span class="p">,</span> <span class="s1">&#39;TTACCC&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">==</span> <span class="s1">&#39;GGGGTAA&#39;</span> 
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">alignSubamplicon</span><span class="p">(</span><span class="n">refseq</span><span class="p">,</span> <span class="s1">&#39;GGGCTA&#39;</span><span class="p">,</span> <span class="s1">&#39;TTAGCC&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">==</span> <span class="kc">False</span> 
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.annotateCodonCounts">
<code class="descclassname">dms_tools2.utils.</code><code class="descname">annotateCodonCounts</code><span class="sig-paren">(</span><em>counts</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.utils.annotateCodonCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets annotated <cite>pandas.DataFrame</cite> from codon counts.</p>
<p>Some of the programs (e.g., <cite>dms2_bcsubamplicons</cite>) create 
<code class="docutils literal"><span class="pre">*_codoncounts.csv</span></code> files when run with <code class="docutils literal"><span class="pre">--chartype</span> <span class="pre">codon</span></code>.
These CSV files have columns indicating the <cite>site</cite> and <cite>wildtype</cite>
codon, as well as a column for each codon giving the counts for that 
codon. This function reads that file (or a <cite>pandas.DataFrame</cite> read
from it) to return a <cite>pandas.DataFrame</cite> where a variety of additional
useful annotations have been added.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>counts</cite> (str)</dt>
<dd>Name of existing codon counts CSV file, or <cite>pandas.DataFrame</cite>
holding counts.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd><dl class="first last docutils">
<dt><cite>df</cite> (<cite>pandas.DataFrame</cite>)</dt>
<dd><p class="first">The DataFrame with the information in <cite>counts</cite> plus
the following added columns for each site:</p>
<blockquote class="last">
<div><p><cite>ncounts</cite> : number of counts at site</p>
<p><cite>mutfreq</cite> : mutation frequency at site</p>
<p><cite>nstop</cite> : number of stop-codon mutations</p>
<p><cite>nsyn</cite> : number of synonymous mutations</p>
<p><cite>nnonsyn</cite> : number of nonsynonymous mutations</p>
<p><cite>n1nt</cite> : number of 1-nucleotide codon mutations</p>
<p><cite>n2nt</cite> : number of 2-nucleotide codon mutations</p>
<p><cite>n3nt</cite> : number of 3-nucleotide codon mutations</p>
<p><cite>AtoC</cite>, <cite>AtoG</cite>, etc : number of each nucleotide mutation
type among codon mutations with <strong>one</strong> nucleotide change.</p>
<p><cite>mutfreq1nt</cite>, <cite>mutfreq2nt</cite>, <cite>mutfreq3nt</cite> : frequency
of 1-, 2-, and 3-nucleotide codon mutations at site.</p>
</div></blockquote>
</dd>
</dl>
</dd>
</dl>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;site&#39;</span><span class="p">:[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="s1">&#39;wildtype&#39;</span><span class="p">:[</span><span class="s1">&#39;ATG&#39;</span><span class="p">,</span> <span class="s1">&#39;GGG&#39;</span><span class="p">],</span> <span class="s1">&#39;ATG&#39;</span><span class="p">:[</span><span class="mi">105</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;GGG&#39;</span><span class="p">:[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">117</span><span class="p">],</span> <span class="s1">&#39;GGA&#39;</span><span class="p">:[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">20</span><span class="p">],</span> <span class="s1">&#39;TGA&#39;</span><span class="p">:[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">codon</span> <span class="ow">in</span> <span class="n">dms_tools2</span><span class="o">.</span><span class="n">CODONS</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="n">codon</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">d</span><span class="p">:</span>
<span class="gp">... </span>        <span class="n">d</span><span class="p">[</span><span class="n">codon</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">counts</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">f</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
<span class="gp">... </span>    <span class="n">df</span> <span class="o">=</span> <span class="n">annotateCodonCounts</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">([</span><span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">==</span> <span class="n">counts</span><span class="p">[</span><span class="n">col</span><span class="p">])</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">counts</span><span class="o">.</span><span class="n">columns</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;ncounts&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">110</span><span class="p">,</span> <span class="mi">139</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;mutfreq&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">5</span> <span class="o">/</span> <span class="mf">110.</span><span class="p">,</span> <span class="mi">22</span> <span class="o">/</span> <span class="mf">139.</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;nstop&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;nsyn&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">20</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;nnonsyn&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;n1nt&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">20</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;n2nt&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;n3nt&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;GtoA&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">20</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;AtoC&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;mutfreq1nt&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">20</span> <span class="o">/</span> <span class="mf">139.</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;mutfreq3nt&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">2</span> <span class="o">/</span> <span class="mf">110.</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.avgPrefs">
<code class="descclassname">dms_tools2.utils.</code><code class="descname">avgPrefs</code><span class="sig-paren">(</span><em>prefsfiles</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.utils.avgPrefs" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets average of site-specific preferences.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>prefsfiles</cite> (list)</dt>
<dd>List of CSV files containing preferences, must all be
for same sites and characters.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>A <cite>pandas.DataFrame</cite> containing the average of the
preferences in <cite>prefsfiles</cite>. In this returned
data frame, <cite>site</cite> is the index</dd>
</dl>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tf1</span> <span class="o">=</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tf2</span> <span class="o">=</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">tf1</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file1</span><span class="p">,</span> <span class="n">tf2</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file2</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">x</span> <span class="o">=</span> <span class="n">file1</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;site,A,C,G,T</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>                <span class="s1">&#39;1,0.2,0.2,0.5,0.1</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>                <span class="s1">&#39;2,0.3,0.3,0.3,0.1&#39;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">file1</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
<span class="gp">... </span>    <span class="n">x</span> <span class="o">=</span> <span class="n">file2</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;site,A,C,G,T</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>                <span class="s1">&#39;1,0.4,0.1,0.1,0.4</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>                <span class="s1">&#39;2,0.3,0.4,0.1,0.2&#39;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">file2</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
<span class="gp">... </span>    <span class="n">avg</span> <span class="o">=</span> <span class="n">avgPrefs</span><span class="p">([</span><span class="n">file1</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">file2</span><span class="o">.</span><span class="n">name</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="n">avg</span><span class="p">[</span><span class="s1">&#39;site&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">])</span><span class="o">.</span><span class="n">all</span><span class="p">()</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">avg</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">avg</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.15</span><span class="p">,</span> <span class="mf">0.35</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">avg</span><span class="p">[</span><span class="s1">&#39;G&#39;</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">avg</span><span class="p">[</span><span class="s1">&#39;T&#39;</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.15</span><span class="p">])</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.buildReadConsensus">
<code class="descclassname">dms_tools2.utils.</code><code class="descname">buildReadConsensus</code><span class="sig-paren">(</span><em>reads</em>, <em>minreads</em>, <em>minconcur</em>, <em>use_cutils=True</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.utils.buildReadConsensus" title="Permalink to this definition">¶</a></dt>
<dd><p>Builds consensus sequence of some reads.</p>
<p>You may want to pre-fill low-quality sites with <code class="docutils literal"><span class="pre">N</span></code>
using <cite>lowQtoN</cite>. An <code class="docutils literal"><span class="pre">N</span></code> is considered a non-called identity.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>reads</cite> (list)</dt>
<dd>List of reads as strings. If reads are not all same
length, shorter ones are extended from 3’ end with <code class="docutils literal"><span class="pre">N</span></code>
to match maximal length.</dd>
<dt><cite>minreads</cite> (int)</dt>
<dd>Only call consensus at a site if at least this many reads 
have called identity.</dd>
<dt><cite>minconcur</cite> (float)</dt>
<dd>Only call consensus at site if &gt;= this fraction of called
identities agree.</dd>
<dt><cite>use_cutils</cite> (bool)</dt>
<dd>Use the faster implementation in the <cite>_cutils</cite> module.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>A string giving the consensus sequence. Non-called 
sites are returned as <code class="docutils literal"><span class="pre">N`</span></code>.</dd>
</dl>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">reads</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ATGCAT&#39;</span><span class="p">,</span>
<span class="gp">... </span>         <span class="s1">&#39;NTGNANA&#39;</span><span class="p">,</span>
<span class="gp">... </span>         <span class="s1">&#39;ACGNNTAT&#39;</span><span class="p">,</span>
<span class="gp">... </span>         <span class="s1">&#39;NTGNTA&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">buildReadConsensus</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mf">0.75</span><span class="p">)</span> <span class="o">==</span> <span class="s1">&#39;ATGNNNAN&#39;</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">reads</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;CTGCATAT&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">buildReadConsensus</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mf">0.75</span><span class="p">)</span> <span class="o">==</span> <span class="s1">&#39;NTGCATAT&#39;</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.codonToAACounts">
<code class="descclassname">dms_tools2.utils.</code><code class="descname">codonToAACounts</code><span class="sig-paren">(</span><em>counts</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.utils.codonToAACounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Makes amino-acid counts <cite>pandas.DataFrame</cite> from codon counts.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>counts</cite> (<cite>pandas.DataFrame</cite>)</dt>
<dd>Columns are the string <cite>site</cite> <cite>wildtype</cite> and all codons
in <cite>dms_tools2.CODONS</cite>. Additional columns are allowed
but ignored.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd><dl class="first last docutils">
<dt><cite>aacounts</cite> (<cite>pandas.DataFrame</cite>)</dt>
<dd>Columns are the string <cite>site</cite> and all amino acids
in <cite>dms_tools.AAS_WITHSTOP</cite> with counts for each
amino acid made by summing counts for encoding codons.</dd>
</dl>
</dd>
</dl>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;site&#39;</span><span class="p">:[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="s1">&#39;othercol&#39;</span><span class="p">:[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="s1">&#39;ATG&#39;</span><span class="p">:[</span><span class="mi">105</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;GGG&#39;</span><span class="p">:[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">117</span><span class="p">],</span> <span class="s1">&#39;GGA&#39;</span><span class="p">:[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">20</span><span class="p">],</span> <span class="s1">&#39;TGA&#39;</span><span class="p">:[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;wildtype&#39;</span><span class="p">:[</span><span class="s1">&#39;ATG&#39;</span><span class="p">,</span> <span class="s1">&#39;GGG&#39;</span><span class="p">]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">codon</span> <span class="ow">in</span> <span class="n">dms_tools2</span><span class="o">.</span><span class="n">CODONS</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="n">codon</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">d</span><span class="p">:</span>
<span class="gp">... </span>        <span class="n">d</span><span class="p">[</span><span class="n">codon</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aacounts</span> <span class="o">=</span> <span class="n">codonToAACounts</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s1">&#39;othercol&#39;</span> <span class="ow">in</span> <span class="n">aacounts</span><span class="o">.</span><span class="n">columns</span>
<span class="go">False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">aacounts</span><span class="p">[</span><span class="s1">&#39;site&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">aacounts</span><span class="p">[</span><span class="s1">&#39;wildtype&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="s1">&#39;M&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">aacounts</span><span class="p">[</span><span class="s1">&#39;M&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">105</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">aacounts</span><span class="p">[</span><span class="s1">&#39;G&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">137</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">aacounts</span><span class="p">[</span><span class="s1">&#39;*&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">aacounts</span><span class="p">[</span><span class="s1">&#39;V&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.convertCountsFormat">
<code class="descclassname">dms_tools2.utils.</code><code class="descname">convertCountsFormat</code><span class="sig-paren">(</span><em>oldfile</em>, <em>newfile</em>, <em>charlist</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.utils.convertCountsFormat" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert counts file from <code class="docutils literal"><span class="pre">dms_tools</span></code> to <code class="docutils literal"><span class="pre">dms_tools2</span></code> format.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>oldfile</cite> (str)</dt>
<dd>Name of counts file in the old <code class="docutils literal"><span class="pre">dms_tools</span></code> format:
<a class="reference external" href="http://jbloomlab.github.io/dms_tools/fileformats.html">http://jbloomlab.github.io/dms_tools/fileformats.html</a></dd>
<dt><cite>newfile</cite> (str)</dt>
<dd>Name of created counts file in the <code class="docutils literal"><span class="pre">dms_tools2</span></code> format:
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms2_bcsubamp.html">https://jbloomlab.github.io/dms_tools2/dms2_bcsubamp.html</a></dd>
<dt><cite>charlist</cite> (list)</dt>
<dd>List of characters that we expect in the counts files.
For instance, could be <cite>dms_tools2.CODONS</cite>.</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.incrementCounts">
<code class="descclassname">dms_tools2.utils.</code><code class="descname">incrementCounts</code><span class="sig-paren">(</span><em>refseqstart</em>, <em>subamplicon</em>, <em>chartype</em>, <em>counts</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.utils.incrementCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Increment counts dict based on an aligned subamplicon.</p>
<p>This is designed for keeping track of counts of different
mutations / identities when aligning many subamplicons to
a sequence.</p>
<p>Any positions where <cite>subamplicon</cite> has an <code class="docutils literal"><span class="pre">N</span></code> are ignored,
and not added to <cite>counts</cite>.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>refseqstart</cite> (int)</dt>
<dd>First nucleotide position in 1, 2, … numbering 
where <cite>subamplicon</cite> aligns.</dd>
<dt><cite>subamplicon</cite> (str)</dt>
<dd>The subamplicon.</dd>
<dt><cite>chartype</cite> (str)</dt>
<dd>Character type for which we are counting mutations.
Currently, only allowable value is ‘codon’.</dd>
<dt><cite>counts</cite> (dict)</dt>
<dd>Stores counts of identities, and is incremented by
this function. Is a dict keyed by every possible
character (e.g., codon), with values lists with
element <cite>i</cite> holding the counts for position <cite>i</cite>
in 0, 1, … numbering.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>On completion, <cite>counts</cite> has been incremented.</dd>
</dl>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">codonlen</span> <span class="o">=</span> <span class="mi">10</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">([(</span><span class="n">codon</span><span class="p">,</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">codonlen</span><span class="p">)</span> <span class="k">for</span> <span class="n">codon</span> 
<span class="gp">... </span>        <span class="ow">in</span> <span class="n">dms_tools2</span><span class="o">.</span><span class="n">CODONS</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">subamplicon1</span> <span class="o">=</span> <span class="s1">&#39;ATGGACTTTC&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">incrementCounts</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">subamplicon1</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">,</span> <span class="n">counts</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">subamplicon2</span> <span class="o">=</span> <span class="s1">&#39;GGTCTTTCCCGGN&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">incrementCounts</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="n">subamplicon2</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">,</span> <span class="n">counts</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s1">&#39;ATG&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s1">&#39;GAC&#39;</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s1">&#39;GTC&#39;</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s1">&#39;TTT&#39;</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="mi">2</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="p">[</span><span class="s1">&#39;CCC&#39;</span><span class="p">][</span><span class="mi">3</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">sum</span><span class="p">([</span><span class="nb">sum</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">counts</span><span class="o">.</span><span class="n">values</span><span class="p">()])</span> <span class="o">==</span> <span class="mi">6</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.initLogger">
<code class="descclassname">dms_tools2.utils.</code><code class="descname">initLogger</code><span class="sig-paren">(</span><em>logfile</em>, <em>prog</em>, <em>args</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.utils.initLogger" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize output logging for scripts.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>logfile</cite> (str or <cite>sys.stdout</cite>)</dt>
<dd>Name of file to which log is written, or 
<cite>sys.stdout</cite> if you just want to write information
to standard output.</dd>
<dt><cite>prog</cite> (str)</dt>
<dd>Name of program for which we are logging.</dd>
<dt><cite>args</cite> (dict)</dt>
<dd>Program arguments as arg / value pairs.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>If <cite>logfile</cite> is a string giving a file name, returns
an opened and initialized <cite>logging.Logger</cite>. If <cite>logfile</cite>
is <cite>sys.stdout</cite>, then writes information to <cite>sys.stdout</cite>.
In either case, basic information is written about the program 
and args.</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.iteratePairedFASTQ">
<code class="descclassname">dms_tools2.utils.</code><code class="descname">iteratePairedFASTQ</code><span class="sig-paren">(</span><em>r1files</em>, <em>r2files</em>, <em>r1trim=None</em>, <em>r2trim=None</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.utils.iteratePairedFASTQ" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterates over FASTQ file pairs for paired-end sequencing reads.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>r1files</cite> (list or str)</dt>
<dd>Name of R1 FASTQ file or list of such files. Can optionally
be gzipped.</dd>
<dt><cite>r2files</cite> (list or str)</dt>
<dd>Like <cite>r1files</cite> but for R2 files.</dd>
<dt><cite>r1trim</cite> (int or <cite>None</cite>)</dt>
<dd>If not <cite>None</cite>, trim <cite>r1</cite> and <cite>q1</cite> to be no longer than this.</dd>
<dt><cite>r2trim</cite> (int or <cite>None</cite>)</dt>
<dd>Like <cite>r1trim</cite> but for R2.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd><p class="first">Each iteration returns <cite>(name, r1, r2, q1, q2, fail)</cite> where:</p>
<blockquote class="last">
<div><ul class="simple">
<li><cite>name</cite> is a string giving the read name</li>
<li><cite>r1</cite> and <cite>r2</cite> are strings giving the reads</li>
<li><cite>q1</cite> and <cite>q2</cite> are strings giving the PHRED Q scores</li>
<li><cite>fail</cite> is <cite>True</cite> if either read failed Illumina chastity
filter, <cite>False</cite> if both passed, <cite>None</cite> if info not present.</li>
</ul>
</div></blockquote>
</dd>
</dl>
<p>We run a simple test by first writing an example FASTQ file and
then testing on it.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">n1_1</span> <span class="o">=</span> <span class="s1">&#39;@DH1DQQN1:933:HMLH5BCXY:1:1101:2165:1984 1:N:0:CGATGT&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">r1_1</span> <span class="o">=</span> <span class="s1">&#39;ATGCAATTG&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">q1_1</span> <span class="o">=</span> <span class="s1">&#39;GGGGGIIII&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">n2_1</span> <span class="o">=</span> <span class="s1">&#39;@DH1DQQN1:933:HMLH5BCXY:1:1101:2165:1984 2:N:0:CGATGT&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">r2_1</span> <span class="o">=</span> <span class="s1">&#39;CATGCATA&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">q2_1</span> <span class="o">=</span> <span class="s1">&#39;G&lt;GGGIII&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tf</span> <span class="o">=</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">tf</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">r1file</span><span class="p">,</span> <span class="n">tf</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">r2file</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">dummyvar</span> <span class="o">=</span> <span class="n">r1file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span>
<span class="gp">... </span>            <span class="n">n1_1</span><span class="p">,</span> <span class="n">r1_1</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="n">q1_1</span><span class="p">,</span>
<span class="gp">... </span>            <span class="n">n1_1</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;:N:&#39;</span><span class="p">,</span> <span class="s1">&#39;:Y:&#39;</span><span class="p">),</span> <span class="n">r1_1</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="n">q1_1</span><span class="p">,</span>
<span class="gp">... </span>            <span class="n">n1_1</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">],</span> <span class="n">r1_1</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="n">q1_1</span><span class="p">,</span>
<span class="gp">... </span>            <span class="p">]))</span>
<span class="gp">... </span>    <span class="n">r1file</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
<span class="gp">... </span>    <span class="n">dummyvar</span> <span class="o">=</span> <span class="n">r2file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span>
<span class="gp">... </span>            <span class="n">n2_1</span><span class="p">,</span> <span class="n">r2_1</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="n">q2_1</span><span class="p">,</span>
<span class="gp">... </span>            <span class="n">n2_1</span><span class="p">,</span> <span class="n">r2_1</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="n">q2_1</span><span class="p">,</span>
<span class="gp">... </span>            <span class="n">n2_1</span><span class="p">,</span> <span class="n">r2_1</span><span class="p">,</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="n">q2_1</span><span class="p">,</span>
<span class="gp">... </span>            <span class="p">]))</span>
<span class="gp">... </span>    <span class="n">r2file</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
<span class="gp">... </span>    <span class="n">itr</span> <span class="o">=</span> <span class="n">iteratePairedFASTQ</span><span class="p">(</span><span class="n">r1file</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">r2file</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">r1trim</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">r2trim</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="gp">... </span>    <span class="nb">next</span><span class="p">(</span><span class="n">itr</span><span class="p">)</span> <span class="o">==</span> <span class="p">(</span><span class="n">n1_1</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span> <span class="p">:</span> <span class="p">],</span> <span class="n">r1_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">4</span><span class="p">],</span> 
<span class="gp">... </span>            <span class="n">r2_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">5</span><span class="p">],</span> <span class="n">q1_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">4</span><span class="p">],</span> <span class="n">q2_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">5</span><span class="p">],</span> <span class="kc">False</span><span class="p">)</span>
<span class="gp">... </span>    <span class="nb">next</span><span class="p">(</span><span class="n">itr</span><span class="p">)</span> <span class="o">==</span> <span class="p">(</span><span class="n">n1_1</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span> <span class="p">:</span> <span class="p">],</span> <span class="n">r1_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">4</span><span class="p">],</span> 
<span class="gp">... </span>            <span class="n">r2_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">5</span><span class="p">],</span> <span class="n">q1_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">4</span><span class="p">],</span> <span class="n">q2_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">5</span><span class="p">],</span> <span class="kc">True</span><span class="p">)</span>
<span class="gp">... </span>    <span class="nb">next</span><span class="p">(</span><span class="n">itr</span><span class="p">)</span> <span class="o">==</span> <span class="p">(</span><span class="n">n1_1</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span> <span class="p">:</span> <span class="p">],</span> <span class="n">r1_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">4</span><span class="p">],</span> 
<span class="gp">... </span>            <span class="n">r2_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">5</span><span class="p">],</span> <span class="n">q1_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">4</span><span class="p">],</span> <span class="n">q2_1</span><span class="p">[</span> <span class="p">:</span> <span class="mi">5</span><span class="p">],</span> <span class="kc">None</span><span class="p">)</span>
<span class="go">True</span>
<span class="go">True</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.lowQtoN">
<code class="descclassname">dms_tools2.utils.</code><code class="descname">lowQtoN</code><span class="sig-paren">(</span><em>r</em>, <em>q</em>, <em>minq</em>, <em>use_cutils=True</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.utils.lowQtoN" title="Permalink to this definition">¶</a></dt>
<dd><p>Replaces low quality nucleotides with <code class="docutils literal"><span class="pre">N</span></code> characters.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>r</cite> (str)</dt>
<dd>A string representing a sequencing read.</dd>
<dt><cite>q</cite> (str)</dt>
<dd>String of same length as <cite>r</cite> holding Q scores
in Sanger ASCII encoding.</dd>
<dt><cite>minq</cite> (length-one string)</dt>
<dd>Replace all positions in <cite>r</cite> where <cite>q</cite> is &lt; this.</dd>
<dt><cite>use_cutils</cite> (bool)</dt>
<dd>Use the faster implementation in the <cite>_cutils</cite> module.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>A version of <cite>r</cite> where all positions <cite>i</cite> where 
<cite>q[i] &lt; minq</cite> have been replaced with <code class="docutils literal"><span class="pre">N</span></code>.</dd>
</dl>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">r</span> <span class="o">=</span> <span class="s1">&#39;ATGCAT&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">q</span> <span class="o">=</span> <span class="s1">&#39;GB&lt;.0+&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">minq</span> <span class="o">=</span> <span class="s1">&#39;0&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lowQtoN</span><span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="n">q</span><span class="p">,</span> <span class="n">minq</span><span class="p">)</span> <span class="o">==</span> <span class="s1">&#39;ATGNAN&#39;</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.rescalePrefs">
<code class="descclassname">dms_tools2.utils.</code><code class="descname">rescalePrefs</code><span class="sig-paren">(</span><em>prefs</em>, <em>stringency</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.utils.rescalePrefs" title="Permalink to this definition">¶</a></dt>
<dd><p>Re-scale amino acid preferences by stringency parameter.</p>
<p>If the initial preference of site <span class="math">\(r\)</span> for amino-acid 
<span class="math">\(a\)</span> is <span class="math">\(\pi_{r,a}\)</span>, then the re-scaled preference is</p>
<div class="math">
\[\frac{\left(\pi_{r,a}\right)^{\beta}}{\sum_{a'} \left(\pi_{r,a'}\right)}\]</div>
<p>where <span class="math">\(\beta\)</span> is the stringency parameter.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>prefs</cite> (pandas.DataFrame)</dt>
<dd>Columns are ‘site’ and then every character (e.g.,
amino acid).</dd>
<dt><cite>stringency</cite> (float &gt;= 0)</dt>
<dd>The stringency parameter.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>A data frame in the same format as <cite>prefs</cite> but with the
re-scaled preferences.</dd>
</dl>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">prefs</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="nb">dict</span><span class="p">(</span>
<span class="gp">... </span>        <span class="p">[(</span><span class="s1">&#39;site&#39;</span><span class="p">,</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">]),</span> <span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="p">[</span><span class="mf">0.24</span><span class="p">,</span> <span class="mf">0.03</span><span class="p">]),</span> <span class="p">(</span><span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="p">[</span><span class="mf">0.04</span><span class="p">,</span> <span class="mf">0.43</span><span class="p">])]</span>
<span class="gp">... </span>        <span class="o">+</span> <span class="p">[(</span><span class="n">aa</span><span class="p">,</span> <span class="p">[</span><span class="mf">0.04</span><span class="p">,</span> <span class="mf">0.03</span><span class="p">])</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">dms_tools2</span><span class="o">.</span><span class="n">AAS</span> <span class="k">if</span>
<span class="gp">... </span>        <span class="n">aa</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">]]))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">prefs</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s1">&#39;site&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rescaled</span> <span class="o">=</span> <span class="n">rescalePrefs</span><span class="p">(</span><span class="n">prefs</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">([</span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">prefs</span><span class="p">[</span><span class="n">c</span><span class="p">],</span> <span class="n">rescaled</span><span class="p">[</span><span class="n">c</span><span class="p">])</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">prefs</span><span class="o">.</span><span class="n">columns</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rescaled2</span> <span class="o">=</span> <span class="n">rescalePrefs</span><span class="p">(</span><span class="n">prefs</span><span class="p">,</span> <span class="mf">2.0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">rescaled2</span><span class="p">[</span><span class="s1">&#39;site&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">prefs</span><span class="p">[</span><span class="s1">&#39;site&#39;</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">rescaled2</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.6545</span><span class="p">,</span> <span class="mf">0.0045</span><span class="p">],</span> <span class="n">atol</span><span class="o">=</span><span class="mf">1e-3</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">numpy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">rescaled2</span><span class="p">[</span><span class="s1">&#39;C&#39;</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.0182</span><span class="p">,</span> <span class="mf">0.9153</span><span class="p">],</span> <span class="n">atol</span><span class="o">=</span><span class="mf">1e-3</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.reverseComplement">
<code class="descclassname">dms_tools2.utils.</code><code class="descname">reverseComplement</code><span class="sig-paren">(</span><em>s</em>, <em>use_cutils=True</em><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.utils.reverseComplement" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets reverse complement of DNA sequence <cite>s</cite>.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>s</cite> (str)</dt>
<dd>Sequence to reverse complement.</dd>
<dt><cite>use_cutils</cite> (bool)</dt>
<dd>Use the faster implementation in the <cite>_cutils</cite> module.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>Reverse complement of <cite>s</cite> as a str.</dd>
</dl>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="s1">&#39;ATGCAAN&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">reverseComplement</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="o">==</span> <span class="s1">&#39;NTTGCAT&#39;</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.utils.sessionInfo">
<code class="descclassname">dms_tools2.utils.</code><code class="descname">sessionInfo</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#dms_tools2.utils.sessionInfo" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns string with information about session / packages.</p>
</dd></dl>

</div>
</div>
<div class="section" id="module-dms_tools2">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-dms_tools2" title="Permalink to this headline">¶</a></h2>
<div class="section" id="dms-tools2">
<h3>dms_tools2<a class="headerlink" href="#dms-tools2" title="Permalink to this headline">¶</a></h3>
<p>Python package for analyzing deep mutational scanning (DMS) data.</p>
<p>See <a class="reference external" href="http://jbloomlab.github.io/dms_tools2">http://jbloomlab.github.io/dms_tools2</a> for documentation.</p>
<dl class="data">
<dt id="dms_tools2.AAS">
<code class="descclassname">dms_tools2.</code><code class="descname">AAS</code><em class="property"> = ['A', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'K', 'L', 'M', 'N', 'P', 'Q', 'R', 'S', 'T', 'V', 'W', 'Y']</em><a class="headerlink" href="#dms_tools2.AAS" title="Permalink to this definition">¶</a></dt>
<dd><p>alphabetized list of all 20 amino acids</p>
</dd></dl>

<dl class="data">
<dt id="dms_tools2.AAS_WITHSTOP">
<code class="descclassname">dms_tools2.</code><code class="descname">AAS_WITHSTOP</code><em class="property"> = ['A', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'K', 'L', 'M', 'N', 'P', 'Q', 'R', 'S', 'T', 'V', 'W', 'Y', '*']</em><a class="headerlink" href="#dms_tools2.AAS_WITHSTOP" title="Permalink to this definition">¶</a></dt>
<dd><p>like <cite>AAS</cite> but includes stop codon (<code class="docutils literal"><span class="pre">*</span></code>) at end</p>
</dd></dl>

<dl class="data">
<dt id="dms_tools2.AA_TO_CODONS">
<code class="descclassname">dms_tools2.</code><code class="descname">AA_TO_CODONS</code><em class="property"> = {'A': ['GCA', 'GCC', 'GCG', 'GCT'], 'C': ['TGC', 'TGT'], '*': ['TAA', 'TAG', 'TGA'], 'I': ['ATA', 'ATC', 'ATT'], 'N': ['AAC', 'AAT'], 'F': ['TTC', 'TTT'], 'L': ['CTA', 'CTC', 'CTG', 'CTT', 'TTA', 'TTG'], 'D': ['GAC', 'GAT'], 'K': ['AAA', 'AAG'], 'Y': ['TAC', 'TAT'], 'Q': ['CAA', 'CAG'], 'P': ['CCA', 'CCC', 'CCG', 'CCT'], 'W': ['TGG'], 'M': ['ATG'], 'R': ['AGA', 'AGG', 'CGA', 'CGC', 'CGG', 'CGT'], 'G': ['GGA', 'GGC', 'GGG', 'GGT'], 'E': ['GAA', 'GAG'], 'H': ['CAC', 'CAT'], 'V': ['GTA', 'GTC', 'GTG', 'GTT'], 'T': ['ACA', 'ACC', 'ACG', 'ACT'], 'S': ['AGC', 'AGT', 'TCA', 'TCC', 'TCG', 'TCT']}</em><a class="headerlink" href="#dms_tools2.AA_TO_CODONS" title="Permalink to this definition">¶</a></dt>
<dd><p>dict back-translating amino acid to list of codons.</p>
</dd></dl>

<dl class="data">
<dt id="dms_tools2.CODONS">
<code class="descclassname">dms_tools2.</code><code class="descname">CODONS</code><em class="property"> = ['AAA', 'AAC', 'AAG', 'AAT', 'ACA', 'ACC', 'ACG', 'ACT', 'AGA', 'AGC', 'AGG', 'AGT', 'ATA', 'ATC', 'ATG', 'ATT', 'CAA', 'CAC', 'CAG', 'CAT', 'CCA', 'CCC', 'CCG', 'CCT', 'CGA', 'CGC', 'CGG', 'CGT', 'CTA', 'CTC', 'CTG', 'CTT', 'GAA', 'GAC', 'GAG', 'GAT', 'GCA', 'GCC', 'GCG', 'GCT', 'GGA', 'GGC', 'GGG', 'GGT', 'GTA', 'GTC', 'GTG', 'GTT', 'TAA', 'TAC', 'TAG', 'TAT', 'TCA', 'TCC', 'TCG', 'TCT', 'TGA', 'TGC', 'TGG', 'TGT', 'TTA', 'TTC', 'TTG', 'TTT']</em><a class="headerlink" href="#dms_tools2.CODONS" title="Permalink to this definition">¶</a></dt>
<dd><p>alphabetized list of all codons</p>
</dd></dl>

<dl class="data">
<dt id="dms_tools2.CODON_TO_AA">
<code class="descclassname">dms_tools2.</code><code class="descname">CODON_TO_AA</code><em class="property"> = {'GAA': 'E', 'TGA': '*', 'GCT': 'A', 'TCT': 'S', 'TTG': 'L', 'TCA': 'S', 'CTA': 'L', 'CTG': 'L', 'CAC': 'H', 'AGG': 'R', 'TTA': 'L', 'GTT': 'V', 'TGG': 'W', 'CCG': 'P', 'GAG': 'E', 'TAA': '*', 'TTT': 'F', 'GGT': 'G', 'TAC': 'Y', 'AGA': 'R', 'AAA': 'K', 'CAT': 'H', 'TGT': 'C', 'TGC': 'C', 'TCG': 'S', 'GTG': 'V', 'AGT': 'S', 'CGT': 'R', 'CGC': 'R', 'TCC': 'S', 'GCC': 'A', 'AAC': 'N', 'GCA': 'A', 'TTC': 'F', 'CTC': 'L', 'GAT': 'D', 'GGC': 'G', 'AGC': 'S', 'CGA': 'R', 'AAG': 'K', 'ATG': 'M', 'ACG': 'T', 'TAG': '*', 'ACA': 'T', 'ACC': 'T', 'GGA': 'G', 'CCA': 'P', 'GCG': 'A', 'ACT': 'T', 'AAT': 'N', 'ATC': 'I', 'CTT': 'L', 'ATT': 'I', 'TAT': 'Y', 'CAG': 'Q', 'GGG': 'G', 'CCC': 'P', 'GTC': 'V', 'ATA': 'I', 'CCT': 'P', 'GTA': 'V', 'CGG': 'R', 'CAA': 'Q', 'GAC': 'D'}</em><a class="headerlink" href="#dms_tools2.CODON_TO_AA" title="Permalink to this definition">¶</a></dt>
<dd><p>dict translating codons to amino acids</p>
</dd></dl>

<dl class="data">
<dt id="dms_tools2.NTCOMPLEMENT">
<code class="descclassname">dms_tools2.</code><code class="descname">NTCOMPLEMENT</code><em class="property"> = {'A': 'T', 'C': 'G', 'T': 'A', 'G': 'C', 'N': 'N'}</em><a class="headerlink" href="#dms_tools2.NTCOMPLEMENT" title="Permalink to this definition">¶</a></dt>
<dd><p>dict mapping each nucleotide to its complement.</p>
</dd></dl>

<dl class="data">
<dt id="dms_tools2.NTS">
<code class="descclassname">dms_tools2.</code><code class="descname">NTS</code><em class="property"> = ['A', 'C', 'G', 'T']</em><a class="headerlink" href="#dms_tools2.NTS" title="Permalink to this definition">¶</a></dt>
<dd><p>alphabetized list of all nucleotides</p>
</dd></dl>

</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">dms_tools2</h1>
    
  </a>
</p>



<p class="blurb">Tools for analyzing deep mutational scanning data.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=dms_tools2&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a href="https://travis-ci.org/jbloomlab/dms_tools2">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="bcsubamp.html">Barcoded-subamplicon sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="prefs.html">Amino-acid preferences</a></li>
<li class="toctree-l1"><a class="reference internal" href="diffsel.html">Differential selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="programs.html">Executable programs</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="api.html">Python API</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">dms_tools2 package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="citations.html">Citations</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="api.html">Python API</a><ul>
      <li>Previous: <a href="api.html" title="previous chapter">Python API</a></li>
      <li>Next: <a href="citations.html" title="next chapter">Citations</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017, the Bloom lab.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.6.3</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
      |
      <a href="_sources/dms_tools2.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/dms_tools2" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>