
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>codonvarianttable &#8212; dms_tools2 2.4.4 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="compareprefs" href="dms_tools2.compareprefs.html" />
    <link rel="prev" title="barcodes" href="dms_tools2.barcodes.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <span class="target" id="module-dms_tools2.codonvarianttable"></span><div class="section" id="codonvarianttable">
<h1>codonvarianttable<a class="headerlink" href="#codonvarianttable" title="Permalink to this headline">¶</a></h1>
<p>This module defines <a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable" title="dms_tools2.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a> objects
for storing and handling codon variants of a gene. They are
useful when you are performing deep mutational scanning
on barcoded codon-mutant libraries of a gene.</p>
<dl class="data">
<dt id="dms_tools2.codonvarianttable.CBPALETTE">
<code class="descclassname">dms_tools2.codonvarianttable.</code><code class="descname">CBPALETTE</code><em class="property"> = ['#999999', '#E69F00', '#56B4E9', '#009E73', '#F0E442', '#0072B2', '#D55E00', '#CC79A7']</em><a class="headerlink" href="#dms_tools2.codonvarianttable.CBPALETTE" title="Permalink to this definition">¶</a></dt>
<dd><p><a class="reference external" href="http://bconnelly.net/2013/10/creating-colorblind-friendly-figures/">color-blind safe palette</a></p>
</dd></dl>

<dl class="class">
<dt id="dms_tools2.codonvarianttable.CodonVariantTable">
<em class="property">class </em><code class="descclassname">dms_tools2.codonvarianttable.</code><code class="descname">CodonVariantTable</code><span class="sig-paren">(</span><em>*</em>, <em>barcode_variant_file</em>, <em>geneseq</em>, <em>substitutions_are_codon=False</em>, <em>extra_cols=[]</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/codonvarianttable.html#CodonVariantTable"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.codonvarianttable.CodonVariantTable" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Associates barcodes with codon mutants of gene.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>barcode_variant_file</cite> (str)</dt>
<dd>CSV file giving barcodes and variants. Must have
columns named “library”, “barcode”, “substitutions”,
(nucleotide mutations in 1, … numbering in a format
like “G301A A302T G856C”), and “variant_call_support”
(sequences supporting barcode-variant call). Any
additional columns are removed unless they are specified
in <cite>extra_cols</cite>.</dd>
<dt><cite>geneseq</cite> (str)</dt>
<dd>Sequence of wildtype protein-coding gene.</dd>
<dt><cite>substitutions_are_codon</cite> (bool)</dt>
<dd>If <cite>True</cite>, then the “substitutions” column in
<cite>barcode_variant_file</cite> gives the substitutions
as codon rather than nucleotide mutations (e.g.,
“ATG1ATA GTA5CCC” for substitutions at codons
1 and 5 in 1, 2, … numbering).</dd>
<dt><cite>extra_cols</cite> (list)</dt>
<dd>Additional columns in <cite>barcode_variant_file</cite> to
retain when creating <cite>barcode_variant_df</cite> and
<cite>variant_count_df</cite> attributes.</dd>
</dl>
</dd>
<dt>Attributes:</dt>
<dd><dl class="first last docutils">
<dt><cite>geneseq</cite> (str)</dt>
<dd><cite>geneseq</cite> passed at initialization.</dd>
<dt><cite>sites</cite> (list)</dt>
<dd>List of all codon sites in 1, 2, … numbering.</dd>
<dt><cite>codons</cite> (OrderedDict)</dt>
<dd><cite>codons[r]</cite> is wildtype codon at site <cite>r</cite>, ordered
by sites.</dd>
<dt><cite>aas</cite> (OrderedDict)</dt>
<dd><cite>aas[r]</cite> is wildtype amino acid at site <cite>r</cite>,
ordered by sites.</dd>
<dt><cite>libraries</cite> (list)</dt>
<dd>List of libraries in <cite>barcode_variantfile</cite>.</dd>
<dt><cite>barcode_variant_df</cite> (pandas DataFrame)</dt>
<dd>Info about codon mutations parsed from <cite>barcode_variantfile</cite>.</dd>
<dt><cite>variant_count_df</cite> (pandas DataFrame or None)</dt>
<dd>Initially <cite>None</cite>, but after data are added with
<a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable.addSampleCounts" title="dms_tools2.codonvarianttable.CodonVariantTable.addSampleCounts"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.addSampleCounts</span></code></a>, holds
counts of each variant for each sample. Differs from
<cite>barcode_variant_df</cite> in that the former just holds
initial barcode-variant data, whereas <cite>variant_count_df</cite>
is updated with variant counts for samples.</dd>
</dl>
</dd>
</dl>
<p>Here is an example.</p>
<p>First, initialize a <a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable" title="dms_tools2.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">geneseq</span> <span class="o">=</span> <span class="s1">&#39;ATGGGATGA&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variantfile</span> <span class="o">=</span> <span class="s1">&#39;_variantfile.csv&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">variantfile</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
<span class="gp">... </span>          <span class="s1">&#39;library,barcode,substitutions,variant_call_support</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>          <span class="s1">&#39;lib_1,AAC,,2</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>          <span class="s1">&#39;lib_1,GAT,G4C A6C,1</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>          <span class="s1">&#39;lib_2,AAC,T2A G4T,2</span><span class="se">\n</span><span class="s1">&#39;</span>
<span class="gp">... </span>          <span class="s1">&#39;lib_2,CAT,A6C,3&#39;</span>
<span class="gp">... </span>          <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span> <span class="o">=</span> <span class="n">CodonVariantTable</span><span class="p">(</span>
<span class="gp">... </span>            <span class="n">barcode_variant_file</span><span class="o">=</span><span class="n">variantfile</span><span class="p">,</span>
<span class="gp">... </span>            <span class="n">geneseq</span><span class="o">=</span><span class="n">geneseq</span>
<span class="gp">... </span>            <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">variantfile</span><span class="p">)</span>
</pre></div>
</div>
<p>Check attributes of the <a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable" title="dms_tools2.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">sites</span>
<span class="go">[1, 2, 3]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">codons</span> <span class="o">==</span> <span class="p">{</span><span class="mi">1</span><span class="p">:</span><span class="s1">&#39;ATG&#39;</span><span class="p">,</span> <span class="mi">2</span><span class="p">:</span><span class="s1">&#39;GGA&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">:</span><span class="s1">&#39;TGA&#39;</span><span class="p">}</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">aas</span> <span class="o">==</span> <span class="p">{</span><span class="mi">1</span><span class="p">:</span><span class="s1">&#39;M&#39;</span><span class="p">,</span> <span class="mi">2</span><span class="p">:</span><span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">:</span><span class="s1">&#39;*&#39;</span><span class="p">}</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">libraries</span>
<span class="go">[&#39;lib_1&#39;, &#39;lib_2&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">valid_barcodes</span><span class="p">(</span><span class="s1">&#39;lib_1&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="p">{</span><span class="s1">&#39;AAC&#39;</span><span class="p">,</span> <span class="s1">&#39;GAT&#39;</span><span class="p">}</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">valid_barcodes</span><span class="p">(</span><span class="s1">&#39;lib_2&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="p">{</span><span class="s1">&#39;AAC&#39;</span><span class="p">,</span> <span class="s1">&#39;CAT&#39;</span><span class="p">}</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s1">&#39;display.max_columns&#39;</span><span class="p">,</span> <span class="mi">15</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s1">&#39;display.width&#39;</span><span class="p">,</span> <span class="mi">500</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">barcode_variant_df</span>
<span class="go">  library barcode  variant_call_support codon_substitutions aa_substitutions  n_codon_substitutions  n_aa_substitutions</span>
<span class="go">0   lib_1     AAC                     2                                                           0                   0</span>
<span class="go">1   lib_1     GAT                     1             GGA2CGC              G2R                      1                   1</span>
<span class="go">2   lib_2     AAC                     2     ATG1AAG GGA2TGA          M1K G2*                      2                   2</span>
<span class="go">3   lib_2     CAT                     3             GGA2GGC                                       1                   0</span>
</pre></div>
</div>
<p>We can also look at the number of variants; we get this
by calling <a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable.n_variants_df" title="dms_tools2.codonvarianttable.CodonVariantTable.n_variants_df"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.n_variants_df</span></code></a> with
<cite>samples=None</cite> since we don’t have samples, and just want
stats across barcoded variants:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">n_variants_df</span><span class="p">(</span><span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="go">         library             sample  count</span>
<span class="go">0          lib_1  barcoded variants      2</span>
<span class="go">1          lib_2  barcoded variants      2</span>
<span class="go">2  all libraries  barcoded variants      4</span>
</pre></div>
</div>
<p>Here is the number of variants if we look at just single amino-acid
mutants (and wildtype):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">n_variants_df</span><span class="p">(</span><span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">variant_type</span><span class="o">=</span><span class="s1">&#39;single&#39;</span><span class="p">,</span> <span class="n">mut_type</span><span class="o">=</span><span class="s1">&#39;aa&#39;</span><span class="p">)</span>
<span class="go">         library             sample  count</span>
<span class="go">0          lib_1  barcoded variants      2</span>
<span class="go">1          lib_2  barcoded variants      1</span>
<span class="go">2  all libraries  barcoded variants      3</span>
</pre></div>
</div>
<p>We can also see how these numbers change if we require a
variant call support of at least 2:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">n_variants_df</span><span class="p">(</span><span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">min_support</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="go">         library             sample  count</span>
<span class="go">0          lib_1  barcoded variants      1</span>
<span class="go">1          lib_2  barcoded variants      2</span>
<span class="go">2  all libraries  barcoded variants      3</span>
</pre></div>
</div>
<p>If we want to combine the data for the two libraries, we can use
<a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable.addMergedLibraries" title="dms_tools2.codonvarianttable.CodonVariantTable.addMergedLibraries"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.addMergedLibraries</span></code></a>, which creates a
new combined library called “all libraries”:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">addMergedLibraries</span><span class="p">(</span><span class="n">variants</span><span class="o">.</span><span class="n">barcode_variant_df</span><span class="p">)</span>
<span class="go">         library    barcode  variant_call_support codon_substitutions aa_substitutions  n_codon_substitutions  n_aa_substitutions</span>
<span class="go">0          lib_1        AAC                     2                                                           0                   0</span>
<span class="go">1          lib_1        GAT                     1             GGA2CGC              G2R                      1                   1</span>
<span class="go">2          lib_2        AAC                     2     ATG1AAG GGA2TGA          M1K G2*                      2                   2</span>
<span class="go">3          lib_2        CAT                     3             GGA2GGC                                       1                   0</span>
<span class="go">4  all libraries  lib_1-AAC                     2                                                           0                   0</span>
<span class="go">5  all libraries  lib_1-GAT                     1             GGA2CGC              G2R                      1                   1</span>
<span class="go">6  all libraries  lib_2-AAC                     2     ATG1AAG GGA2TGA          M1K G2*                      2                   2</span>
<span class="go">7  all libraries  lib_2-CAT                     3             GGA2GGC                                       1                   0</span>
</pre></div>
</div>
<p>Note however that <a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable.addMergedLibraries" title="dms_tools2.codonvarianttable.CodonVariantTable.addMergedLibraries"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.addMergedLibraries</span></code></a>
doesn’t do anything if there is only one library:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">addMergedLibraries</span><span class="p">(</span><span class="n">variants</span><span class="o">.</span><span class="n">barcode_variant_df</span>
<span class="gp">... </span>                            <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;library == &quot;lib_1&quot;&#39;</span><span class="p">))</span>
<span class="go">  library barcode  variant_call_support codon_substitutions aa_substitutions  n_codon_substitutions  n_aa_substitutions</span>
<span class="go">0   lib_1     AAC                     2                                                           0                   0</span>
<span class="go">1   lib_1     GAT                     1             GGA2CGC              G2R                      1                   1</span>
</pre></div>
</div>
<p>Count number of barcoded variants with each mutation:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">mutCounts</span><span class="p">(</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="s1">&#39;aa&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)[</span> <span class="p">:</span> <span class="mi">2</span><span class="p">]</span>
<span class="go">  library             sample mutation  count  mutation_type  site</span>
<span class="go">0   lib_1  barcoded variants      G2R      1  nonsynonymous     2</span>
<span class="go">1   lib_1  barcoded variants      *3A      0  nonsynonymous     3</span>
</pre></div>
</div>
<p>We can do the same for codon mutations (here for only a
single library), first for all variants:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">mutCounts</span><span class="p">(</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="gp">... </span>        <span class="n">libraries</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;lib_2&#39;</span><span class="p">])[</span> <span class="p">:</span> <span class="mi">4</span><span class="p">]</span>
<span class="go">  library             sample mutation  count  mutation_type  site</span>
<span class="go">0   lib_2  barcoded variants  ATG1AAG      1  nonsynonymous     1</span>
<span class="go">1   lib_2  barcoded variants  GGA2GGC      1     synonymous     2</span>
<span class="go">2   lib_2  barcoded variants  GGA2TGA      1           stop     2</span>
<span class="go">3   lib_2  barcoded variants  ATG1AAA      0  nonsynonymous     1</span>
</pre></div>
</div>
<p>Then for just single-mutant variants:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">mutCounts</span><span class="p">(</span><span class="s1">&#39;single&#39;</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="gp">... </span>        <span class="n">libraries</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;lib_2&#39;</span><span class="p">])[</span> <span class="p">:</span> <span class="mi">3</span><span class="p">]</span>
<span class="go">  library             sample mutation  count  mutation_type  site</span>
<span class="go">0   lib_2  barcoded variants  GGA2GGC      1     synonymous     2</span>
<span class="go">1   lib_2  barcoded variants  ATG1AAA      0  nonsynonymous     1</span>
<span class="go">2   lib_2  barcoded variants  ATG1AAC      0  nonsynonymous     1</span>
</pre></div>
</div>
<p>Initially we haven’t added any barcode count information
for any samples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">variants</span><span class="o">.</span><span class="n">samples</span><span class="p">(</span><span class="n">lib</span><span class="p">)</span> <span class="o">==</span> <span class="p">[]</span> <span class="k">for</span> <span class="n">lib</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">libraries</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">variant_count_df</span> <span class="ow">is</span> <span class="kc">None</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Now we add barcode count information for sample “input”
from library 1 using <a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable.addSampleCounts" title="dms_tools2.codonvarianttable.CodonVariantTable.addSampleCounts"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.addSampleCounts</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts_lib1_input</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
<span class="gp">... </span>        <span class="p">{</span><span class="s1">&#39;barcode&#39;</span><span class="p">:[</span><span class="s1">&#39;AAC&#39;</span><span class="p">,</span> <span class="s1">&#39;GAT&#39;</span><span class="p">],</span>
<span class="gp">... </span>         <span class="s1">&#39;count&#39;</span>  <span class="p">:[</span>  <span class="mi">253</span><span class="p">,</span>  <span class="mi">1101</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">addSampleCounts</span><span class="p">(</span><span class="s1">&#39;lib_1&#39;</span><span class="p">,</span> <span class="s1">&#39;input&#39;</span><span class="p">,</span> <span class="n">counts_lib1_input</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">variant_count_df</span>
<span class="go">  barcode  count library sample  variant_call_support codon_substitutions aa_substitutions  n_codon_substitutions  n_aa_substitutions</span>
<span class="go">0     GAT   1101   lib_1  input                     1             GGA2CGC              G2R                      1                   1</span>
<span class="go">1     AAC    253   lib_1  input                     2                                                           0                   0</span>
</pre></div>
</div>
<p>We get an error if we try to add these same data again,
as they are already added for that sample to that library:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">addSampleCounts</span><span class="p">(</span><span class="s1">&#39;lib_1&#39;</span><span class="p">,</span> <span class="s1">&#39;input&#39;</span><span class="p">,</span> <span class="n">counts_lib1_input</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="c">...</span>
<span class="gr">ValueError</span>: <span class="n">`library` lib_1 already has `sample` input</span>
</pre></div>
</div>
<p>But, we can add barcode counts for another
sample (named “selected” in this case) to library 1:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts_lib1_selected</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
<span class="gp">... </span>        <span class="p">{</span><span class="s1">&#39;barcode&#39;</span><span class="p">:[</span><span class="s1">&#39;AAC&#39;</span><span class="p">,</span> <span class="s1">&#39;GAT&#39;</span><span class="p">],</span>
<span class="gp">... </span>         <span class="s1">&#39;count&#39;</span>  <span class="p">:[</span>  <span class="mi">513</span><span class="p">,</span>  <span class="mi">401</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">addSampleCounts</span><span class="p">(</span><span class="s1">&#39;lib_1&#39;</span><span class="p">,</span> <span class="s1">&#39;selected&#39;</span><span class="p">,</span> <span class="n">counts_lib1_selected</span><span class="p">)</span>
</pre></div>
</div>
<p>As well as barcode counts for the same two samples
(“input” and “selected”) to our other library (library 2):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts_lib2_input</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
<span class="gp">... </span>        <span class="p">{</span><span class="s1">&#39;barcode&#39;</span><span class="p">:[</span><span class="s1">&#39;AAC&#39;</span><span class="p">,</span> <span class="s1">&#39;CAT&#39;</span><span class="p">],</span>
<span class="gp">... </span>         <span class="s1">&#39;count&#39;</span>  <span class="p">:[</span> <span class="mi">1253</span><span class="p">,</span>  <span class="mi">923</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">addSampleCounts</span><span class="p">(</span><span class="s1">&#39;lib_2&#39;</span><span class="p">,</span> <span class="s1">&#39;input&#39;</span><span class="p">,</span> <span class="n">counts_lib2_input</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts_lib2_selected</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
<span class="gp">... </span>        <span class="p">{</span><span class="s1">&#39;barcode&#39;</span><span class="p">:[</span><span class="s1">&#39;AAC&#39;</span><span class="p">,</span> <span class="s1">&#39;CAT&#39;</span><span class="p">],</span>
<span class="gp">... </span>         <span class="s1">&#39;count&#39;</span>  <span class="p">:[</span>  <span class="mi">113</span><span class="p">,</span>  <span class="mi">1200</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">addSampleCounts</span><span class="p">(</span><span class="s1">&#39;lib_2&#39;</span><span class="p">,</span> <span class="s1">&#39;selected&#39;</span><span class="p">,</span> <span class="n">counts_lib2_selected</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">variant_count_df</span>
<span class="go">  barcode  count library    sample  variant_call_support codon_substitutions aa_substitutions  n_codon_substitutions  n_aa_substitutions</span>
<span class="go">0     GAT   1101   lib_1     input                     1             GGA2CGC              G2R                      1                   1</span>
<span class="go">1     AAC    253   lib_1     input                     2                                                           0                   0</span>
<span class="go">2     AAC    513   lib_1  selected                     2                                                           0                   0</span>
<span class="go">3     GAT    401   lib_1  selected                     1             GGA2CGC              G2R                      1                   1</span>
<span class="go">4     AAC   1253   lib_2     input                     2     ATG1AAG GGA2TGA          M1K G2*                      2                   2</span>
<span class="go">5     CAT    923   lib_2     input                     3             GGA2GGC                                       1                   0</span>
<span class="go">6     CAT   1200   lib_2  selected                     3             GGA2GGC                                       1                   0</span>
<span class="go">7     AAC    113   lib_2  selected                     2     ATG1AAG GGA2TGA          M1K G2*                      2                   2</span>
</pre></div>
</div>
<p>We can also use <a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable.mutCounts" title="dms_tools2.codonvarianttable.CodonVariantTable.mutCounts"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.mutCounts</span></code></a>
to look at total counts of each mutation:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">mutCounts</span><span class="p">(</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="s1">&#39;aa&#39;</span><span class="p">)[</span> <span class="p">:</span> <span class="mi">2</span><span class="p">]</span>
<span class="go">  library sample mutation  count  mutation_type  site</span>
<span class="go">0   lib_1  input      G2R   1101  nonsynonymous     2</span>
<span class="go">1   lib_1  input      *3A      0  nonsynonymous     3</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">mutCounts</span><span class="p">(</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="s1">&#39;aa&#39;</span><span class="p">,</span> <span class="n">libraries</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;lib_2&#39;</span><span class="p">])[</span> <span class="p">:</span> <span class="mi">3</span><span class="p">]</span>
<span class="go">  library sample mutation  count  mutation_type  site</span>
<span class="go">0   lib_2  input      G2*   1253           stop     2</span>
<span class="go">1   lib_2  input      M1K   1253  nonsynonymous     1</span>
<span class="go">2   lib_2  input      *3A      0  nonsynonymous     3</span>
</pre></div>
</div>
<p>We can use <a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable.writeCodonCounts" title="dms_tools2.codonvarianttable.CodonVariantTable.writeCodonCounts"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.writeCodonCounts</span></code></a> to
write codon count files. First for only <strong>single</strong> mutants:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">TemporaryDirectory</span><span class="p">()</span> <span class="k">as</span> <span class="n">tmpdir</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">countfiles</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">writeCodonCounts</span><span class="p">(</span><span class="s2">&quot;single&quot;</span><span class="p">,</span> <span class="n">outdir</span><span class="o">=</span><span class="n">tmpdir</span><span class="p">,</span> <span class="n">include_all_libs</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">lib1_input</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{tmpdir}</span><span class="s1">/lib_1_input_codoncounts.csv&#39;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">all_sel</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{tmpdir}</span><span class="s1">/all-libraries_selected_codoncounts.csv&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Make sure we created expected countfiles:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">countfiles</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">countfile</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">countfile</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">))</span>
<span class="go">         library    sample                               countfile</span>
<span class="go">0          lib_1     input             lib_1_input_codoncounts.csv</span>
<span class="go">1          lib_1  selected          lib_1_selected_codoncounts.csv</span>
<span class="go">2          lib_2     input             lib_2_input_codoncounts.csv</span>
<span class="go">3          lib_2  selected          lib_2_selected_codoncounts.csv</span>
<span class="go">4  all-libraries     input     all-libraries_input_codoncounts.csv</span>
<span class="go">5  all-libraries  selected  all-libraries_selected_codoncounts.csv</span>
</pre></div>
</div>
<p>Check for expected values in a few of these counts files, only
showing columns with non-zero entries:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">lib1_input</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="p">(</span><span class="n">lib1_input</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;rows&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="p">]</span>
<span class="go">   site wildtype  ATG   CGC  GGA  TGA</span>
<span class="go">0     1      ATG  253     0    0    0</span>
<span class="go">1     2      GGA    0  1101  253    0</span>
<span class="go">2     3      TGA    0     0    0  253</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">all_sel</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="p">(</span><span class="n">all_sel</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;rows&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="p">]</span>
<span class="go">   site wildtype  ATG  CGC  GGA   GGC  TGA</span>
<span class="go">0     1      ATG  513    0    0     0    0</span>
<span class="go">1     2      GGA    0  401  513  1200    0</span>
<span class="go">2     3      TGA    0    0    0     0  513</span>
</pre></div>
</div>
<p>Now write codon counts files for <strong>all</strong> mutants:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">TemporaryDirectory</span><span class="p">()</span> <span class="k">as</span> <span class="n">tmpdir</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">writeCodonCounts</span><span class="p">(</span><span class="s2">&quot;all&quot;</span><span class="p">,</span> <span class="n">outdir</span><span class="o">=</span><span class="n">tmpdir</span><span class="p">,</span> <span class="n">include_all_libs</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">lib1_input_all</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{tmpdir}</span><span class="s1">/lib_1_input_codoncounts.csv&#39;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">all_sel_all</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{tmpdir}</span><span class="s1">/all-libraries_selected_codoncounts.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">lib1_input_all</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="p">(</span><span class="n">lib1_input_all</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;rows&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="p">]</span>
<span class="go">   site wildtype   ATG   CGC  GGA   TGA</span>
<span class="go">0     1      ATG  1354     0    0     0</span>
<span class="go">1     2      GGA     0  1101  253     0</span>
<span class="go">2     3      TGA     0     0    0  1354</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">all_sel_all</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="p">(</span><span class="n">all_sel_all</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;rows&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="p">]</span>
<span class="go">   site wildtype  AAG   ATG  CGC  GGA   GGC   TGA</span>
<span class="go">0     1      ATG  113  2114    0    0     0     0</span>
<span class="go">1     2      GGA    0     0  401  513  1200   113</span>
<span class="go">2     3      TGA    0     0    0    0     0  2227</span>
</pre></div>
</div>
<p>We can also initialize a <a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable" title="dms_tools2.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a> from the
<cite>variant_count_df</cite> if we have written that to a CSV file.
We do this using <a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable.from_variant_count_df" title="dms_tools2.codonvarianttable.CodonVariantTable.from_variant_count_df"><code class="xref py py-meth docutils literal notranslate"><span class="pre">CodonVariantTable.from_variant_count_df()</span></code></a>.
The example below shows how this newly initialized variant table
is equal to the original one used to write the CSV file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">variants</span><span class="o">.</span><span class="n">variant_count_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">f</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
<span class="gp">... </span>    <span class="n">variants_eq</span> <span class="o">=</span> <span class="n">CodonVariantTable</span><span class="o">.</span><span class="n">from_variant_count_df</span><span class="p">(</span>
<span class="gp">... </span>                    <span class="n">variant_count_df_file</span><span class="o">=</span><span class="n">f</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
<span class="gp">... </span>                    <span class="n">geneseq</span><span class="o">=</span><span class="n">geneseq</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span> <span class="o">==</span> <span class="n">variants_eq</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Of course, the initialized variant table is <strong>not</strong> equal
to original one if we don’t write the full <cite>variant_count_df</cite>
to the CSV file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
<span class="gp">... </span>    <span class="p">(</span><span class="n">variants</span>
<span class="gp">... </span>        <span class="o">.</span><span class="n">variant_count_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;sample == &quot;input&quot;&#39;</span><span class="p">)</span>
<span class="gp">... </span>        <span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">... </span>        <span class="p">)</span>
<span class="gp">... </span>    <span class="n">f</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
<span class="gp">... </span>    <span class="n">variants_ne</span> <span class="o">=</span> <span class="n">CodonVariantTable</span><span class="o">.</span><span class="n">from_variant_count_df</span><span class="p">(</span>
<span class="gp">... </span>                    <span class="n">variant_count_df_file</span><span class="o">=</span><span class="n">f</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
<span class="gp">... </span>                    <span class="n">geneseq</span><span class="o">=</span><span class="n">geneseq</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span> <span class="o">==</span> <span class="n">variants_ne</span>
<span class="go">False</span>
</pre></div>
</div>
<p>We can use <a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable.func_scores" title="dms_tools2.codonvarianttable.CodonVariantTable.func_scores"><code class="xref py py-meth docutils literal notranslate"><span class="pre">CodonVariantTable.func_scores()</span></code></a> to compute
the functional effects of mutations. We cannot use this method
with default options as we have no wildtype counts (needed for
normalization) for <cite>lib_2</cite> so we get an error:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">func_scores</span><span class="p">(</span><span class="s1">&#39;input&#39;</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="c">...</span>
<span class="gr">ValueError</span>: <span class="n">no wildtype counts:</span>
<span class="go">  library    sample  count</span>
<span class="go">0   lib_1     input    253</span>
<span class="go">1   lib_1  selected    513</span>
<span class="go">2   lib_2     input      0</span>
<span class="go">3   lib_2  selected      0</span>
</pre></div>
</div>
<p>However, we can use the method with the <cite>permit_zero_wt</cite> option:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">func_scores</span><span class="p">(</span><span class="s1">&#39;input&#39;</span><span class="p">,</span> <span class="n">permit_zero_wt</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">  library pre_sample post_sample barcode  func_score  func_score_var  pre_count  post_count  pre_count_wt  post_count_wt  pseudocount codon_substitutions  n_codon_substitutions aa_substitutions  n_aa_substitutions</span>
<span class="go">0   lib_1      input    selected     AAC    0.000000        0.024528        253         513           253            513          0.5                                          0                                    0</span>
<span class="go">1   lib_1      input    selected     GAT   -2.474376        0.019337       1101         401           253            513          0.5             GGA2CGC                      1              G2R                   1</span>
<span class="go">2   lib_2      input    selected     AAC   -3.465198        8.345474       1253         113             0              0          0.5     ATG1AAG GGA2TGA                      2          M1K G2*                   2</span>
<span class="go">3   lib_2      input    selected     CAT    0.378452        8.329463        923        1200             0              0          0.5             GGA2GGC                      1                                    0</span>
</pre></div>
</div>
<dl class="staticmethod">
<dt id="dms_tools2.codonvarianttable.CodonVariantTable.addMergedLibraries">
<em class="property">static </em><code class="descname">addMergedLibraries</code><span class="sig-paren">(</span><em>df</em>, <em>*</em>, <em>all_lib='all libraries'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/codonvarianttable.html#CodonVariantTable.addMergedLibraries"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.codonvarianttable.CodonVariantTable.addMergedLibraries" title="Permalink to this definition">¶</a></dt>
<dd><p>Add data to <cite>df</cite> for all libraries merged.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>df</cite> (pandas DataFrame)</dt>
<dd>DataFrame that includes columns named
“library” and “barcode”.</dd>
<dt><cite>all_lib</cite> (str)</dt>
<dd>Name given to library that is merge of all
other libraries.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>If <cite>df</cite> only has data for one library, just returns
<cite>df</cite>. Otherwise returns a copy of <cite>df</cite> that has a
new library with the name given by <cite>all_lib</cite>
and contains the data for all individual libraries,
with the “barcode” column giving the original
library name followed by a hyphen and the barcode.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_tools2.codonvarianttable.CodonVariantTable.addSampleCounts">
<code class="descname">addSampleCounts</code><span class="sig-paren">(</span><em>library</em>, <em>sample</em>, <em>barcodecounts</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/codonvarianttable.html#CodonVariantTable.addSampleCounts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.codonvarianttable.CodonVariantTable.addSampleCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Add variant counts for a sample to <cite>variant_count_df</cite>.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>library</cite> (str)</dt>
<dd>Valid <cite>library</cite> for the <a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable" title="dms_tools2.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a>.</dd>
<dt><cite>sample</cite> (str)</dt>
<dd>Sample name, must <strong>not</strong> already be in
<a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable.samples" title="dms_tools2.codonvarianttable.CodonVariantTable.samples"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.samples</span></code></a> for <cite>library</cite>.</dd>
<dt><cite>barcodecounts</cite> (pandas DataFrame)</dt>
<dd>Gives counts for each variant by barcode. Must
have columns named “barcode” and “count”. The
“barcode” column must contain all the barcodes
in <a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable.valid_barcodes" title="dms_tools2.codonvarianttable.CodonVariantTable.valid_barcodes"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.valid_barcodes</span></code></a> for
<cite>library</cite>. Such data frames are returned
by <a class="reference internal" href="dms_tools2.barcodes.html#dms_tools2.barcodes.IlluminaBarcodeParser.parse" title="dms_tools2.barcodes.IlluminaBarcodeParser.parse"><code class="xref py py-class docutils literal notranslate"><span class="pre">dms_tools2.barcodes.IlluminaBarcodeParser.parse</span></code></a>.</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="classmethod">
<dt id="dms_tools2.codonvarianttable.CodonVariantTable.codonToAAMuts">
<em class="property">classmethod </em><code class="descname">codonToAAMuts</code><span class="sig-paren">(</span><em>codon_mut_str</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/codonvarianttable.html#CodonVariantTable.codonToAAMuts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.codonvarianttable.CodonVariantTable.codonToAAMuts" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts string of codon mutations to amino-acid mutations.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>codon_mut_str</cite> (str)</dt>
<dd>Codon mutations, delimited by a space and in
1, 2, … numbering.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>String with amino acid mutations in 1, 2, … numbering.</dd>
</dl>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">CodonVariantTable</span><span class="o">.</span><span class="n">codonToAAMuts</span><span class="p">(</span><span class="s1">&#39;ATG1GTG GGA2GGC TGA3AGA&#39;</span><span class="p">)</span>
<span class="go">&#39;M1V *3R&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="classmethod">
<dt id="dms_tools2.codonvarianttable.CodonVariantTable.from_variant_count_df">
<em class="property">classmethod </em><code class="descname">from_variant_count_df</code><span class="sig-paren">(</span><em>*</em>, <em>variant_count_df_file</em>, <em>geneseq</em>, <em>drop_all_libs=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/codonvarianttable.html#CodonVariantTable.from_variant_count_df"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.codonvarianttable.CodonVariantTable.from_variant_count_df" title="Permalink to this definition">¶</a></dt>
<dd><p><a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable" title="dms_tools2.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a> from CSV of <cite>variant_count_df</cite>.</p>
<p>Use this method when you have written a CSV file of the
<cite>variant_count_df</cite> attribute of a <a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable" title="dms_tools2.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a>,
and now wish to re-initialize that <a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable" title="dms_tools2.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a>.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>variant_count_df_file</cite> (str)</dt>
<dd>Name of CSV file containing the <cite>variant_count_df</cite>.
Must have following columns: “barcode”, “library”,
“variant_call_support”, “codon_substitutions”,
“sample”, and “count”.</dd>
<dt><cite>geneseq</cite> (str)</dt>
<dd>Sequence of wildtype protein-coding gene.</dd>
<dt><cite>drop_all_libs</cite> (bool)</dt>
<dd>If there is a library named “all libraries”,
drop it from the list of libraries in the created
<a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable" title="dms_tools2.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a> as it probably was
added by <a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable.addMergedLibraries" title="dms_tools2.codonvarianttable.CodonVariantTable.addMergedLibraries"><code class="xref py py-meth docutils literal notranslate"><span class="pre">CodonVariantTable.addMergedLibraries()</span></code></a>
and duplicates information for the individual libraries.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>The <a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable" title="dms_tools2.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a> used to write
<cite>variant_count_df_file</cite>.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_tools2.codonvarianttable.CodonVariantTable.func_scores">
<code class="descname">func_scores</code><span class="sig-paren">(</span><em>preselection</em>, <em>*</em>, <em>pseudocount=0.5</em>, <em>by='barcode'</em>, <em>combine_libs=False</em>, <em>syn_as_wt=False</em>, <em>logbase=2</em>, <em>permit_zero_wt=False</em>, <em>permit_self_comp=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/codonvarianttable.html#CodonVariantTable.func_scores"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.codonvarianttable.CodonVariantTable.func_scores" title="Permalink to this definition">¶</a></dt>
<dd><p>Get data frame with functional scores for variants.</p>
<p>The functional score is calculated from the change in counts
for a variant pre- and post-selection using the formula in
<a class="reference external" href="https://doi.org/10.1073/pnas.1804015115">Otwinoski et al (2018)</a>.
Specifically, if <span class="math notranslate nohighlight">\(n^v_{pre}\)</span> and <span class="math notranslate nohighlight">\(n^v_{post}\)</span> are
the counts of variant <span class="math notranslate nohighlight">\(v\)</span> pre- and post-selection, and
if <span class="math notranslate nohighlight">\(n^{wt}_{pre}\)</span> and <span class="math notranslate nohighlight">\(n^{wt}_{post}\)</span> are
the summed counts of <strong>all</strong> wildtype variants pre- and post-
selection, then the functional score of the variant is
<span class="math notranslate nohighlight">\(f_v = \log_b\left(\frac{n^v_{post} / n^{wt}_{post}}{n^v_{pre} / n^{wt}_{pre}}\right)\)</span>
and the variance due to Poisson sampling statistics is
<span class="math notranslate nohighlight">\(\frac{1}{\left(\ln b\right)^2}\left(\sigma^2_v = \frac{1}{n^v_{post}} + \frac{1}{n^{wt}_{post}} + \frac{1}{n^v_{pre}} + \frac{1}{n^{wt}_{pre}}\right)\)</span>
where <span class="math notranslate nohighlight">\(b\)</span> is logarithm base (see <cite>logbase</cite> argument).
For both calculations, a pseudocount (see <cite>pseudocount</cite> argument)
is added to each count first. The wildtype counts are computed
across all <strong>fully wildtype</strong> variants (see <cite>syn_as_wt</cite> for
how this is defined).</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>preselection</cite> (str or dict)</dt>
<dd>The pre-selection sample. If the same for all post-
selection samples, then provide the name of the
pre-selection sample. If it differs among post-selection
samples, then provide a dict keyed by each post-selection
sample with the pre-selection sample being the value.</dd>
<dt><cite>pseudocount</cite> (float)</dt>
<dd>The pseudocount added to each count.</dd>
<dt><cite>by</cite> (str)</dt>
<dd>Compute effects for each “barcode”, set of
“aa_substitutions”, or set of “codon_substitutions”.
In the latter two cases, all barcodes with each
set of substitutions are combined (see <cite>combine_libs</cite>).
Note that if you use “aa_substitutions” then it may
be more sensible to set <cite>syn_as_wt</cite> to <cite>True</cite>.</dd>
<dt><cite>syn_as_wt</cite> (bool)</dt>
<dd>In the formula for the functional scores, consider
variants with only synonymous mutations when determining
wildtype counts? If <cite>False</cite>, only variants with <strong>no</strong>
mutations of any type contribute to wildtype counts.</dd>
<dt><cite>combine_libs</cite> (bool)</dt>
<dd>If <cite>by</cite> is “aa_substitutions” or “codon_substitutions”,
do we combine across libraries as well as barcodes?</dd>
<dt><cite>logbase</cite> (float)</dt>
<dd>Base for logarithm when calculating functional score.</dd>
<dt><cite>permit_zero_wt</cite> (bool)</dt>
<dd>If the wildtype counts are zero for any sample, do
we raise an error or permit the calculation to proceed
just using the pseudocount?</dd>
<dt><cite>permit_self_comp</cite> (bool)</dt>
<dd>Permit comparisons where the same sample is used as
the pre- and post-selection?</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd><dl class="first last docutils">
<dt>A pandas Data Frame with the following columns:</dt>
<dd><ul class="first last simple">
<li>“library”: the library (“all libraries” if <cite>combine_libs</cite>)</li>
<li>“pre_sample”: the pre-selection sample</li>
<li>“post_sample”: the post-selection sample</li>
<li>the value corresponding to the grouping we are using
to compute effects (the value of <cite>by</cite>)</li>
<li>“func_score”: the functional score</li>
<li>“func_score_var”: variance on the functional score</li>
<li>“pre_count”: pre-selection counts</li>
<li>“post_count: post-selection counts</li>
<li>“pre_count_wt”: pre-selection counts for all wildtype</li>
<li>“post_count_wt”: post-selection counts for all wildtype</li>
<li>“pseudocount”: the pseudocount value</li>
<li>as many of “aa_substitutions”, “n_aa_substitutions”,
“codon_substitutions”, and “n_codon_substitutions”
as can be retained given the value of <cite>by</cite>.</li>
</ul>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_tools2.codonvarianttable.CodonVariantTable.mutCounts">
<code class="descname">mutCounts</code><span class="sig-paren">(</span><em>variant_type</em>, <em>mut_type</em>, <em>*</em>, <em>libraries='all'</em>, <em>samples='all'</em>, <em>min_support=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/codonvarianttable.html#CodonVariantTable.mutCounts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.codonvarianttable.CodonVariantTable.mutCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Get counts of each individual mutation.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>variant_type</cite> (“single” or “all”)</dt>
<dd>Include just single mutants, or all mutants?</dd>
<dt>Other args:</dt>
<dd>Same meaning as for
<a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable.plotNumMutsHistogram" title="dms_tools2.codonvarianttable.CodonVariantTable.plotNumMutsHistogram"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.plotNumMutsHistogram</span></code></a></dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>A tidy data frame with columns named “library”,
“sample”, “mutation”, “count”, “mutation_type”,
and “site”.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_tools2.codonvarianttable.CodonVariantTable.n_variants_df">
<code class="descname">n_variants_df</code><span class="sig-paren">(</span><em>*</em>, <em>libraries='all'</em>, <em>samples='all'</em>, <em>min_support=1</em>, <em>variant_type='all'</em>, <em>mut_type=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/codonvarianttable.html#CodonVariantTable.n_variants_df"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.codonvarianttable.CodonVariantTable.n_variants_df" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of variants per library / sample.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>variant_type</cite> (“single” or “all”)</dt>
<dd>Include just variants with &lt;= 1 <cite>mut_type</cite> mutation.</dd>
<dt><cite>mut_type</cite> (“aa”, “codon”, or <cite>None</cite>)</dt>
<dd>If <cite>variant_type</cite> is single, set to “codon” or “aa”
to indicate how to filter for single mutants.</dd>
<dt>All other args:</dt>
<dd>Same meaning as for
<a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable.plotNumMutsHistogram" title="dms_tools2.codonvarianttable.CodonVariantTable.plotNumMutsHistogram"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.plotNumMutsHistogram</span></code></a>.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>DataFrame giving number of variants per library /
sample.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_tools2.codonvarianttable.CodonVariantTable.plotCumulMutCoverage">
<code class="descname">plotCumulMutCoverage</code><span class="sig-paren">(</span><em>variant_type</em>, <em>mut_type</em>, <em>*</em>, <em>libraries='all'</em>, <em>samples='all'</em>, <em>plotfile=None</em>, <em>orientation='h'</em>, <em>widthscale=1</em>, <em>heightscale=1</em>, <em>min_support=1</em>, <em>max_count=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/codonvarianttable.html#CodonVariantTable.plotCumulMutCoverage"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.codonvarianttable.CodonVariantTable.plotCumulMutCoverage" title="Permalink to this definition">¶</a></dt>
<dd><p>Fraction of mutation seen &lt;= some number of times.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>variant_type</cite> (“single” or “all”)</dt>
<dd>Include just single mutants, or all mutants?
Mutations are counted relative to <cite>mut_type</cite>.</dd>
<dt><cite>max_count</cite> (<cite>None</cite> or int)</dt>
<dd>Plot cumulative fraction plot out to this
number of observations of mutation. If <cite>None</cite>,
a reasonable value is automatically determined.</dd>
<dt>Other args:</dt>
<dd>Same meaning as for
<a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable.plotNumMutsHistogram" title="dms_tools2.codonvarianttable.CodonVariantTable.plotNumMutsHistogram"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.plotNumMutsHistogram</span></code></a></dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>A <a class="reference external" href="https://plotnine.readthedocs.io">plotnine</a>
plot; can be displayed in a Jupyter notebook with <cite>p.draw()</cite>.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_tools2.codonvarianttable.CodonVariantTable.plotCumulVariantCounts">
<code class="descname">plotCumulVariantCounts</code><span class="sig-paren">(</span><em>*</em>, <em>variant_type='all'</em>, <em>libraries='all'</em>, <em>samples='all'</em>, <em>plotfile=None</em>, <em>orientation='h'</em>, <em>widthscale=1</em>, <em>heightscale=1</em>, <em>min_support=1</em>, <em>mut_type='aa'</em>, <em>tot_variants_hline=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/codonvarianttable.html#CodonVariantTable.plotCumulVariantCounts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.codonvarianttable.CodonVariantTable.plotCumulVariantCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot number variants with &gt;= that each number of counts.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>variant_type</cite> (“single” or “all”)</dt>
<dd>Include just variants with &lt;= one mutation of type,
<cite>mut_type</cite> or all mutants?</dd>
<dt><cite>tot_variants_hline</cite> (bool)</dt>
<dd>Include a dotted horizontal line indicating the total
number of variants (useful since ones not observed
at all will not apparent on plot).</dd>
<dt>Other args:</dt>
<dd>Same meaning as for
<a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable.plotNumMutsHistogram" title="dms_tools2.codonvarianttable.CodonVariantTable.plotNumMutsHistogram"><code class="xref py py-meth docutils literal notranslate"><span class="pre">CodonVariantTable.plotNumMutsHistogram()</span></code></a>.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd><a class="reference external" href="https://plotnine.readthedocs.io">plotnine</a>
plot; can be displayed in a Jupyter notebook with <cite>p.draw()</cite>.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_tools2.codonvarianttable.CodonVariantTable.plotMutFreqs">
<code class="descname">plotMutFreqs</code><span class="sig-paren">(</span><em>variant_type</em>, <em>mut_type</em>, <em>*</em>, <em>libraries='all'</em>, <em>samples='all'</em>, <em>plotfile=None</em>, <em>orientation='h'</em>, <em>widthscale=1</em>, <em>heightscale=1</em>, <em>min_support=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/codonvarianttable.html#CodonVariantTable.plotMutFreqs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.codonvarianttable.CodonVariantTable.plotMutFreqs" title="Permalink to this definition">¶</a></dt>
<dd><p>Mutation frequency along length of gene.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd>Args have same meaning as for
<a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable.plotCumulMutCoverage" title="dms_tools2.codonvarianttable.CodonVariantTable.plotCumulMutCoverage"><code class="xref py py-meth docutils literal notranslate"><span class="pre">CodonVariantTable.plotCumulMutCoverage()</span></code></a>.</dd>
<dt>Returns:</dt>
<dd>A <a class="reference external" href="https://plotnine.readthedocs.io">plotnine</a>
plot; can be displayed in a Jupyter notebook with <cite>p.draw()</cite>.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_tools2.codonvarianttable.CodonVariantTable.plotMutHeatmap">
<code class="descname">plotMutHeatmap</code><span class="sig-paren">(</span><em>variant_type</em>, <em>mut_type</em>, <em>*</em>, <em>count_or_frequency='frequency'</em>, <em>libraries='all'</em>, <em>samples='all'</em>, <em>plotfile=None</em>, <em>orientation='h'</em>, <em>widthscale=1</em>, <em>heightscale=1</em>, <em>min_support=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/codonvarianttable.html#CodonVariantTable.plotMutHeatmap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.codonvarianttable.CodonVariantTable.plotMutHeatmap" title="Permalink to this definition">¶</a></dt>
<dd><p>Heatmap of mutation counts or frequencies.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>count_or_frequency</cite> (“count” or “frequency”)</dt>
<dd>Plot mutation counts or frequencies?</dd>
<dt>All other args</dt>
<dd>Same meaning as for
<a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable.plotCumulMutCoverage" title="dms_tools2.codonvarianttable.CodonVariantTable.plotCumulMutCoverage"><code class="xref py py-meth docutils literal notranslate"><span class="pre">CodonVariantTable.plotCumulMutCoverage()</span></code></a>.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>A <a class="reference external" href="https://plotnine.readthedocs.io">plotnine</a>
plot; can be displayed in a Jupyter notebook with <cite>p.draw()</cite>.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_tools2.codonvarianttable.CodonVariantTable.plotNumCodonMutsByType">
<code class="descname">plotNumCodonMutsByType</code><span class="sig-paren">(</span><em>variant_type</em>, <em>*</em>, <em>libraries='all'</em>, <em>samples='all'</em>, <em>plotfile=None</em>, <em>orientation='h'</em>, <em>widthscale=1</em>, <em>heightscale=1</em>, <em>min_support=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/codonvarianttable.html#CodonVariantTable.plotNumCodonMutsByType"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.codonvarianttable.CodonVariantTable.plotNumCodonMutsByType" title="Permalink to this definition">¶</a></dt>
<dd><p>Nonsynonymous, synonymous, stop mutations per variant.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>variant_type</cite> (“single” or “all”)</dt>
<dd>Include just single-codon mutants and wildtype,
or include all mutants.</dd>
<dt>Other args:</dt>
<dd>Same meaning as for
<a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable.plotNumMutsHistogram" title="dms_tools2.codonvarianttable.CodonVariantTable.plotNumMutsHistogram"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.plotNumMutsHistogram</span></code></a></dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>A <a class="reference external" href="https://plotnine.readthedocs.io">plotnine</a>
plot; can be displayed in a Jupyter notebook with <cite>p.draw()</cite>.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_tools2.codonvarianttable.CodonVariantTable.plotNumMutsHistogram">
<code class="descname">plotNumMutsHistogram</code><span class="sig-paren">(</span><em>mut_type</em>, <em>*</em>, <em>libraries='all'</em>, <em>samples='all'</em>, <em>plotfile=None</em>, <em>orientation='h'</em>, <em>widthscale=1</em>, <em>heightscale=1</em>, <em>min_support=1</em>, <em>max_muts=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/codonvarianttable.html#CodonVariantTable.plotNumMutsHistogram"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.codonvarianttable.CodonVariantTable.plotNumMutsHistogram" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot histograms of number of mutations per variant.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>mut_type</cite> (str)</dt>
<dd>Type of mutation: “codon” or “aa”.</dd>
<dt><cite>libraries</cite> (“all”, “all_only”, or list)</dt>
<dd>Set to “all” to include all libraries including
a merge of the libraries, “all_only” to only
include the merge of the libraries, or a list
of libraries.</dd>
<dt><cite>samples</cite> (the str “all”, <cite>None</cite>, or list)</dt>
<dd>Set to “all” to include all samples, <cite>None</cite> to
just count each barcoded variant once, or specify
a list of samples.</dd>
<dt><cite>plotfile</cite> (<cite>None</cite> or str)</dt>
<dd>Name of file to which we save plot.</dd>
<dt><cite>orientation</cite> (the str ‘h’ or ‘v’)</dt>
<dd>Do we facet libraries horizontally or vertically?</dd>
<dt><cite>widthscale</cite> (float or int)</dt>
<dd>Expand width of plot by this much.</dd>
<dt><cite>heightscale</cite> (float or int)</dt>
<dd>Expand height of plot by this much.</dd>
<dt><cite>min_support</cite> (int)</dt>
<dd>Only include variants with at least this large
of a variant call support.</dd>
<dt><cite>max_muts</cite> (int or <cite>None</cite>)</dt>
<dd>In histogram, group together all variants with
&gt;= this many mutations; set to <cite>None</cite> for no
cutoff.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd><a class="reference external" href="https://plotnine.readthedocs.io">plotnine</a>
plot; can be displayed in a Jupyter notebook with <cite>p.draw()</cite>.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_tools2.codonvarianttable.CodonVariantTable.samples">
<code class="descname">samples</code><span class="sig-paren">(</span><em>library</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/codonvarianttable.html#CodonVariantTable.samples"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.codonvarianttable.CodonVariantTable.samples" title="Permalink to this definition">¶</a></dt>
<dd><p>List of all samples for <cite>library</cite>.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>library</cite> (str)</dt>
<dd>Valid <cite>library</cite> for the <a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable" title="dms_tools2.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a>.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>List of all samples for which barcode counts have
been added.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_tools2.codonvarianttable.CodonVariantTable.valid_barcodes">
<code class="descname">valid_barcodes</code><span class="sig-paren">(</span><em>library</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/codonvarianttable.html#CodonVariantTable.valid_barcodes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.codonvarianttable.CodonVariantTable.valid_barcodes" title="Permalink to this definition">¶</a></dt>
<dd><p>Set of valid barcodes for <cite>library</cite>.</p>
</dd></dl>

<dl class="method">
<dt id="dms_tools2.codonvarianttable.CodonVariantTable.writeCodonCounts">
<code class="descname">writeCodonCounts</code><span class="sig-paren">(</span><em>single_or_all</em>, <em>*</em>, <em>outdir=None</em>, <em>include_all_libs=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/codonvarianttable.html#CodonVariantTable.writeCodonCounts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.codonvarianttable.CodonVariantTable.writeCodonCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes codon counts files for all libraries and samples.</p>
<p>Files are written in
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms2_bcsubamp.html#id12">format</a>
produced by
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms2_bcsubamp.html">dms2_bcsubamp</a>.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>single_or_all</cite> (“single” or “all”)</dt>
<dd>If “single”, then counts are just from single-codon
mutants and wildtype, and we count all wildtype codons
and just mutated codon for single-codon mutants. If
“all”, we count all codons for all variants at all
sites. This is appropriate if enrichment of each mutation
is supposed to represent its effect for “single”, and if
enrichment of mutation is supposed to represent its
average effect across genetic backgrounds in the library
for “all” provided mutations are Poisson distributed.</dd>
<dt><cite>outdir</cite> (<cite>None</cite> or str)</dt>
<dd>Name of directory into which we write counts,
created if it does not exist. Use <cite>None</cite> to
write to current directory.</dd>
<dt><cite>include_all_libs</cite> (bool)</dt>
<dd>Include data for a library (named “all-libraries”)
that has the summmed data for all individual libraries
if there are multiple libraries?</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>Pandas data frame with columns “library”, “sample”,
and “countfile”. The “countfile” columns gives name of the
createad CSV file, <code class="docutils literal notranslate"><span class="pre">&lt;library&gt;_&lt;sample&gt;_codoncounts.csv</span></code>.</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="dms_tools2.codonvarianttable.PhenotypeSimulator">
<em class="property">class </em><code class="descclassname">dms_tools2.codonvarianttable.</code><code class="descname">PhenotypeSimulator</code><span class="sig-paren">(</span><em>geneseq, *, seed=1, wt_latent=1, norm_weights=[(0.4, -0.5, 1), (0.6, -5, 2.5)], stop_effect=-10</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/codonvarianttable.html#PhenotypeSimulator"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.codonvarianttable.PhenotypeSimulator" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Simulates phenotypes of variants under plausible model.</p>
<p>Mutational effects on latent phenotype are simulated to follow
compound normal distribution; latent phenotype maps to observed
phenotype via sigmoid. This distinction between latent and
observed phenotype parallel the “global epistasis” models of
<a class="reference external" href="https://doi.org/10.1073/pnas.1804015115">Otwinoski et al</a> and
<a class="reference external" href="http://www.genetics.org/content/205/3/1079">Sailer and Harms</a>.</p>
<p>Initialize with a codon sequence and random number seed.
The effects of mutations on the latent phenotype are then drawn
from a compound normal distribution biased to negative values.</p>
<p>To calculate latent and observed phenotypes, pass rows of a
<code class="xref py py-meth docutils literal notranslate"><span class="pre">CodonVariantTable.barcode_variant_df()</span></code> to
<a class="reference internal" href="#dms_tools2.codonvarianttable.PhenotypeSimulator.latentPhenotype" title="dms_tools2.codonvarianttable.PhenotypeSimulator.latentPhenotype"><code class="xref py py-meth docutils literal notranslate"><span class="pre">PhenotypeSimulator.latentPhenotype()</span></code></a> or
<a class="reference internal" href="#dms_tools2.codonvarianttable.PhenotypeSimulator.observedPhenotype" title="dms_tools2.codonvarianttable.PhenotypeSimulator.observedPhenotype"><code class="xref py py-meth docutils literal notranslate"><span class="pre">PhenotypeSimulator.observedPhenotype()</span></code></a>.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>geneseq</cite> (str)</dt>
<dd>Codon sequence of wild-type gene.</dd>
<dt><cite>seed</cite> (int)</dt>
<dd>Random number seed.</dd>
<dt><cite>wt_latent</cite> (float)</dt>
<dd>Latent phenotype of wildtype.</dd>
<dt><cite>norm_weights</cite> (list of tuples)</dt>
<dd>Specifies compound normal distribution of mutational
effects on latent phenotype. Each tuple is
<cite>(weight, mean, sd)</cite>, giving weight, mean, and standard
deviation of each Gaussian in compound normal.</dd>
<dt><cite>stop_effect</cite> (float)</dt>
<dd>Effect of stop codon at any position.</dd>
</dl>
</dd>
<dt>Attributes:</dt>
<dd><dl class="first last docutils">
<dt><cite>wt_latent</cite> (float)</dt>
<dd>Value for wildtype latent phenotype.</dd>
<dt><cite>muteffects</cite> (dict)</dt>
<dd>Effects on latent phenotype of each amino-acid mutation.</dd>
</dl>
</dd>
</dl>
<dl class="method">
<dt id="dms_tools2.codonvarianttable.PhenotypeSimulator.latentPhenotype">
<code class="descname">latentPhenotype</code><span class="sig-paren">(</span><em>v</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/codonvarianttable.html#PhenotypeSimulator.latentPhenotype"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.codonvarianttable.PhenotypeSimulator.latentPhenotype" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns latent phenotype of a variant.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>v</cite> (dict or row of pandas DataFrame)</dt>
<dd>Must have key ‘aa_substitutions’ that gives
space de-limited list of amino-acid mutations.</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="staticmethod">
<dt id="dms_tools2.codonvarianttable.PhenotypeSimulator.latentToObservedPhenotype">
<em class="property">static </em><code class="descname">latentToObservedPhenotype</code><span class="sig-paren">(</span><em>latent</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/codonvarianttable.html#PhenotypeSimulator.latentToObservedPhenotype"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.codonvarianttable.PhenotypeSimulator.latentToObservedPhenotype" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns observed phenotype from latent phenotype.</p>
</dd></dl>

<dl class="method">
<dt id="dms_tools2.codonvarianttable.PhenotypeSimulator.observedPhenotype">
<code class="descname">observedPhenotype</code><span class="sig-paren">(</span><em>v</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/codonvarianttable.html#PhenotypeSimulator.observedPhenotype"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.codonvarianttable.PhenotypeSimulator.observedPhenotype" title="Permalink to this definition">¶</a></dt>
<dd><p>Like <cite>latentPhenotype</cite> but returns observed phenotype.</p>
</dd></dl>

<dl class="method">
<dt id="dms_tools2.codonvarianttable.PhenotypeSimulator.plotLatentVersusObservedPhenotype">
<code class="descname">plotLatentVersusObservedPhenotype</code><span class="sig-paren">(</span><em>*</em>, <em>latent_min=-15</em>, <em>latent_max=5</em>, <em>npoints=200</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/codonvarianttable.html#PhenotypeSimulator.plotLatentVersusObservedPhenotype"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.codonvarianttable.PhenotypeSimulator.plotLatentVersusObservedPhenotype" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots observed phenotype as function of latent phenotype.</p>
<p>Plot includes a vertical line at wildtype latent phenotype.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>latent_min</cite> (float)</dt>
<dd>Smallest value of latent phenotype on plot.</dd>
<dt><cite>latent_max</cite> (float)</dt>
<dd>Largest value of latent phenotype on plot.</dd>
<dt><cite>npoints</cite> (int)</dt>
<dd>Plot a line fit to this many points.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>A <a class="reference external" href="https://plotnine.readthedocs.io">plotnine</a>
plot; can be displayed in a Jupyter notebook with <cite>p.draw()</cite>.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_tools2.codonvarianttable.PhenotypeSimulator.plotMutsHistogram">
<code class="descname">plotMutsHistogram</code><span class="sig-paren">(</span><em>latent_or_observed</em>, <em>*</em>, <em>mutant_order=1</em>, <em>bins=30</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/codonvarianttable.html#PhenotypeSimulator.plotMutsHistogram"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.codonvarianttable.PhenotypeSimulator.plotMutsHistogram" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots distribution of phenotype for all mutants.</p>
<p>Plot includes a vertical line at wildtype phenotype.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>latent_or_observed</cite> (“latent” or “observed”)</dt>
<dd>Which type of phenotype to plot.</dd>
<dt><cite>mutant_order</cite> (int)</dt>
<dd>Plot mutations of this order. Currently only works
for 1 (single mutants).</dd>
<dt><cite>bins</cite> (int)</dt>
<dd>Number of bins in histogram.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>A <a class="reference external" href="https://plotnine.readthedocs.io">plotnine</a>
plot; can be displayed in a Jupyter notebook with <cite>p.draw()</cite>.</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="dms_tools2.codonvarianttable.getCumulVariantsByCount">
<code class="descclassname">dms_tools2.codonvarianttable.</code><code class="descname">getCumulVariantsByCount</code><span class="sig-paren">(</span><em>df</em>, <em>*</em>, <em>group_cols=None</em>, <em>group_cols_as_str=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/codonvarianttable.html#getCumulVariantsByCount"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.codonvarianttable.getCumulVariantsByCount" title="Permalink to this definition">¶</a></dt>
<dd><p>Get number of variants with observed &gt;= each number of times.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>df</cite> (pandas DataFrame)</dt>
<dd>A row for each variant, and a column named “count” giving
how many times variant observed. Can have other columns.</dd>
<dt><cite>group_cols</cite> (None or list)</dt>
<dd>Group by these columns and analyze each group separately.</dd>
<dt><cite>group_cols_as_str</cite> (bool)</dt>
<dd>Explicitly convert any <cite>group_cols</cite> columns to str type.
For some reason, this is needed if you are calling in <code class="docutils literal notranslate"><span class="pre">R</span></code>
using <a class="reference external" href="https://rstudio.github.io/reticulate/">reticulate</a>.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>A pandas Data Frame with columns named “count”, “nvariants”,
and “total_variants” (plus any columns in <cite>group_cols</cite>).
For each value of “count”, “nvariants” gives the number of
variants with &gt;= that many counts. The “total_variants”
column gives the total number of variants.</dd>
</dl>
<p>Here is an example. First, create input Data Frame:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="nb">dict</span><span class="p">(</span>
<span class="gp">... </span>         <span class="n">sample</span><span class="o">=</span> <span class="p">[</span><span class="s1">&#39;a&#39;</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">,</span> <span class="s1">&#39;b&#39;</span><span class="p">,</span> <span class="s1">&#39;b&#39;</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">],</span>
<span class="gp">... </span>         <span class="n">count</span><span class="o">=</span>  <span class="p">[</span>  <span class="mi">9</span><span class="p">,</span>   <span class="mi">0</span><span class="p">,</span>   <span class="mi">1</span><span class="p">,</span>   <span class="mi">4</span><span class="p">,</span>   <span class="mi">3</span><span class="p">,</span>   <span class="mi">3</span><span class="p">]))</span>
</pre></div>
</div>
<p>Now apply function, first <strong>not</strong> grouping by sample. Note
how the sample column is therefore simply ignored and dropped:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">getCumulVariantsByCount</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
<span class="go">   count  nvariants  total_variants</span>
<span class="go">0      9          1               6</span>
<span class="go">1      4          2               6</span>
<span class="go">2      3          4               6</span>
<span class="go">3      1          5               6</span>
<span class="go">4      0          6               6</span>
</pre></div>
</div>
<p>Now apply function, grouping by sample. Note how each sample
is now analyzed separately:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">getCumulVariantsByCount</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">group_cols</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">])</span>
<span class="go">  sample  count  nvariants  total_variants</span>
<span class="go">0      a      9          1               4</span>
<span class="go">1      a      3          3               4</span>
<span class="go">2      a      0          4               4</span>
<span class="go">3      b      4          1               2</span>
<span class="go">4      b      1          2               2</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.codonvarianttable.rarefyBarcodes">
<code class="descclassname">dms_tools2.codonvarianttable.</code><code class="descname">rarefyBarcodes</code><span class="sig-paren">(</span><em>barcodecounts</em>, <em>*</em>, <em>barcodecol='barcode'</em>, <em>countcol='count'</em>, <em>maxpoints=100000.0</em>, <em>logspace=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/codonvarianttable.html#rarefyBarcodes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.codonvarianttable.rarefyBarcodes" title="Permalink to this definition">¶</a></dt>
<dd><p>Rarefaction curve of barcode observations.</p>
<p>Uses analytical formula for rarefaction defined
<a class="reference external" href="https://en.wikipedia.org/wiki/Rarefaction_(ecology)#Derivation">here</a>.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>barcodecounts</cite> (pandas DataFrame)</dt>
<dd>Has columns with names matching <cite>barcodecol</cite> and <cite>countcol</cite>.</dd>
<dt><cite>barcodecol</cite> (str)</dt>
<dd>Name of column listing all unique barcodes.</dd>
<dt><cite>countcol</cite> (str)</dt>
<dd>Name of column with observed counts of each barcode.</dd>
<dt><cite>maxpoints</cite> (int)</dt>
<dd>Only calculate rarefaction curve at this many points.
The benefit is that it can become very costly to calculate
the curve for many points.</dd>
<dt><cite>logspace</cite> (bool)</dt>
<dd>Logarithmically space the <cite>maxpoint</cite> points. If False,
space them linearly.</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd>A pandas DataFrame with the columns <cite>ncounts</cite> and
<cite>nbarcodes</cite>, giving the number of unique barcodes
observed for each total number observed counts.</dd>
</dl>
<p>Here is a small example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">barcodecounts</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;barcode&#39;</span><span class="p">:[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">],</span>
<span class="gp">... </span>                              <span class="s1">&#39;count&#39;</span><span class="p">:</span>  <span class="p">[</span>  <span class="mi">4</span><span class="p">,</span>   <span class="mi">2</span><span class="p">,</span>   <span class="mi">1</span><span class="p">,</span>   <span class="mi">0</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rarefaction_curve</span> <span class="o">=</span> <span class="n">rarefyBarcodes</span><span class="p">(</span><span class="n">barcodecounts</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rarefaction_curve</span>
<span class="go">   ncounts  nbarcodes</span>
<span class="go">0        1   1.000000</span>
<span class="go">1        2   1.666667</span>
<span class="go">2        3   2.114286</span>
<span class="go">3        4   2.428571</span>
<span class="go">4        5   2.666667</span>
<span class="go">5        6   2.857143</span>
<span class="go">6        7   3.000000</span>
</pre></div>
</div>
<p>Verify that the result from this example matches what is
obtained by random sampling:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">barcodelist</span> <span class="o">=</span> <span class="p">[]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">tup</span> <span class="ow">in</span> <span class="n">barcodecounts</span><span class="o">.</span><span class="n">itertuples</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">barcodelist</span> <span class="o">+=</span> <span class="p">[</span><span class="n">tup</span><span class="o">.</span><span class="n">barcode</span><span class="p">]</span> <span class="o">*</span> <span class="n">tup</span><span class="o">.</span><span class="n">count</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nrand</span> <span class="o">=</span> <span class="mi">10000</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ncounts</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">barcodecounts</span><span class="p">[</span><span class="s1">&#39;count&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">+</span> <span class="mi">1</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nbarcodes</span> <span class="o">=</span> <span class="p">[]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">ncount</span> <span class="ow">in</span> <span class="n">ncounts</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">nbarcodes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="n">barcodelist</span><span class="p">,</span> <span class="n">ncount</span><span class="p">)))</span>
<span class="gp">... </span>                     <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nrand</span><span class="p">))</span> <span class="o">/</span> <span class="n">nrand</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sim_rarefaction_curve</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="nb">dict</span><span class="p">(</span><span class="n">ncounts</span><span class="o">=</span><span class="n">ncounts</span><span class="p">,</span>
<span class="gp">... </span>                                          <span class="n">nbarcodes</span><span class="o">=</span><span class="n">nbarcodes</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">scipy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">rarefaction_curve</span><span class="p">,</span> <span class="n">sim_rarefaction_curve</span><span class="p">,</span> <span class="n">atol</span><span class="o">=</span><span class="mf">1e-2</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.codonvarianttable.simulateSampleCounts">
<code class="descclassname">dms_tools2.codonvarianttable.</code><code class="descname">simulateSampleCounts</code><span class="sig-paren">(</span><em>*</em>, <em>variants</em>, <em>phenotype_func</em>, <em>variant_error_rate</em>, <em>pre_sample</em>, <em>post_samples</em>, <em>pre_sample_name='pre-selection'</em>, <em>seed=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/codonvarianttable.html#simulateSampleCounts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.codonvarianttable.simulateSampleCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Simulate pre- and post-selection variant counts.</p>
<p>Simulate variant counts for experiment where barcodes
are sequenced pre- and post-selection.</p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt><cite>variants</cite> (<a class="reference internal" href="#dms_tools2.codonvarianttable.CodonVariantTable" title="dms_tools2.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a>)</dt>
<dd>Holds variants used in simulation.</dd>
<dt><cite>phenotype_func</cite> (function)</dt>
<dd>Takes a row from <cite>variants.barcode_variant_df</cite>
and returns the phenotype. Typically this is
calculated from the “aa_substitutions” or
“codon_substitutions” column. The phenotype is a
number &gt;= 0, and represents the expected enrichment of
the variant relative to wildtype post-selection (values
&gt; 1 indicate beneficial). For instance, you could pass
<a class="reference internal" href="#dms_tools2.codonvarianttable.PhenotypeSimulator.observedPhenotype" title="dms_tools2.codonvarianttable.PhenotypeSimulator.observedPhenotype"><code class="xref py py-meth docutils literal notranslate"><span class="pre">PhenotypeSimulator.observedPhenotype()</span></code></a>.</dd>
<dt><cite>variant_error_rate</cite> (float)</dt>
<dd>Rate at which variants in <cite>variants</cite> are
mis-called. Provide the probability that a
variant has a spuriously called (or missing)
codon mutation: with this probability, each
variant then has a random codon mutation added
or taken away before being passed to
<cite>phenotype_func</cite>. Designed to simulate the fact
that the method used to link barcodes to variants
is probably not perfect.</dd>
<dt><cite>pre_sample</cite> (pandas DataFrame or dict)</dt>
<dd><p class="first">The counts of each variant pre-selection. You can
specify counts or have them simulated:</p>
<blockquote class="last">
<div><ul class="simple">
<li>To specify counts, provide a data frame with
columns “library”, “barcode”, and “count”
where “count” is the pre-selection counts.</li>
<li>To simulate, provide dict with keys “total_count”
and “uniformity”. For each library, we simulate
pre-selection counts as a draw of “total_count”
counts from a multinomial parameterized by
pre-selection frequences drawn from a Dirichlet
distribution with a concentration parameter of
of “uniformity” (larger uniformity values mean
more even libraries; 5 is reasonable value).</li>
</ul>
</div></blockquote>
</dd>
<dt><cite>post_samples</cite> (dict)</dt>
<dd><p class="first">Nested dict indicating post-selection samples.
Keyed by name of each post-selection sample, with
value being another dict. The post-selection counts
are drawn from a multinomial parameterized
by the pre-selection frequencies (frequencies calculated from
counts if <cite>pre_sample</cite> gives counts, Dirichlet frequencies
if <cite>pre_sample</cite> is simulated). Add noise by the
following parameters provided in the dict for each sample:</p>
<blockquote class="last">
<div><ul class="simple">
<li>“total_count”: total overall counts per library.</li>
<li>“noise”: add additional noise to selection by
multiplying phenotype times a random variable
drawn from a normal distribution with mean 1
and this standard deviation (truncated at lower
end to zero). Set noise to 0 for no noise.</li>
<li>“bottleneck”: put the pre-selection frequencies
through a bottleneck of this size, then re-calcuate
initial frequencies that selection acts upon.</li>
</ul>
</div></blockquote>
</dd>
<dt><cite>pre_sample_name</cite> (str)</dt>
<dd>Name used for the pre-selection sample.</dd>
<dt><cite>seed</cite> (None or int)</dt>
<dd>If not <cite>None</cite>, random number seed set before executing
function (sets <cite>scipy.random.seed</cite>).</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd><p class="first">A pandas DataFrame with the following columns:</p>
<blockquote>
<div><ul class="simple">
<li>“library”</li>
<li>“barcode”</li>
<li>“sample”</li>
<li>“count”</li>
</ul>
</div></blockquote>
<p class="last">The first two columns indicate the library and barcode
for each variant as in the <cite>barcode_variant_df</cite> attribute
of <cite>variants</cite>, the “sample” and “count” columns give counts
for each sample.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_tools2.codonvarianttable.tidy_split">
<code class="descclassname">dms_tools2.codonvarianttable.</code><code class="descname">tidy_split</code><span class="sig-paren">(</span><em>df</em>, <em>column</em>, <em>sep=' '</em>, <em>keep=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_tools2/codonvarianttable.html#tidy_split"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_tools2.codonvarianttable.tidy_split" title="Permalink to this definition">¶</a></dt>
<dd><p>Split values of a column and expand so new DataFrame has one split
value per row. Filters rows where the column is missing.</p>
<p>Taken from <a class="reference external" href="https://stackoverflow.com/a/39946744">https://stackoverflow.com/a/39946744</a></p>
<dl class="docutils">
<dt>Args:</dt>
<dd><dl class="first last docutils">
<dt>df <span class="classifier-delimiter">:</span> <span class="classifier">pandas DataFrame</span></dt>
<dd>dataframe with the column to split and expand</dd>
<dt>column <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>the column to split and expand</dd>
<dt>sep <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>the string used to split the column’s values</dd>
<dt>keep <span class="classifier-delimiter">:</span> <span class="classifier">bool</span></dt>
<dd>whether to retain the presplit value as it’s own row</dd>
</dl>
</dd>
<dt>Returns:</dt>
<dd><dl class="first last docutils">
<dt>pandas DataFrame</dt>
<dd>Returns a dataframe with the same columns as <cite>df</cite>.</dd>
</dl>
</dd>
</dl>
</dd></dl>

</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">dms_tools2</h1>
    
  </a>
</p>



<p class="blurb">Tools for analyzing deep mutational scanning data.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=dms_tools2&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a class="badge" href="https://travis-ci.org/jbloomlab/dms_tools2">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="bcsubamp.html">Barcoded-subamplicon sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="prefs.html">Amino-acid preferences</a></li>
<li class="toctree-l1"><a class="reference internal" href="diffsel.html">Differential selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="fracsurvive.html">Fraction surviving</a></li>
<li class="toctree-l1"><a class="reference internal" href="programs.html">Executable programs</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="api.html">Python API</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="dms_tools2.html">dms_tools2 package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="citations.html">Citations</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="api.html">Python API</a><ul>
  <li><a href="dms_tools2.html">dms_tools2 package</a><ul>
      <li>Previous: <a href="dms_tools2.barcodes.html" title="previous chapter">barcodes</a></li>
      <li>Next: <a href="dms_tools2.compareprefs.html" title="next chapter">compareprefs</a></li>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017--2019, the Bloom lab.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/dms_tools2.codonvarianttable.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/dms_tools2" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>