
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>Fraction surviving &#8212; dms_tools2 2.6.6 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Executable programs" href="programs.html" />
    <link rel="prev" title="Differential selection" href="diffsel.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="fraction-surviving">
<span id="fracsurvive"></span><h1>Fraction surviving<a class="headerlink" href="#fraction-surviving" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="contents">
<ul class="simple">
<li><p><a class="reference internal" href="#overview" id="id2">Overview</a></p></li>
<li><p><a class="reference internal" href="#definition-of-fraction-surviving" id="id3">Definition of fraction surviving</a></p>
<ul>
<li><p><a class="reference internal" href="#fraction-surviving-above-average" id="id4">Fraction surviving above average</a></p></li>
<li><p><a class="reference internal" href="#error-correction" id="id5">Error correction</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#programs-and-examples" id="id6">Programs and examples</a></p></li>
</ul>
</div>
<div class="section" id="overview">
<h2><a class="toc-backref" href="#id2">Overview</a><a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>In some cases, you might subject a deep mutational scanning library to a harsh
selection that purges most mutations.
An example of such an experiment is selecting a virus library with a monoclonal
antibody – most variants will be inactivated by the antibody, but those with mutations
in the epitope will survive.
If the experiment measures the fraction of the entire library
that survives the selection as well as using sequencing to identify the frequencies
of each mutation among the surviving variants, then it is possible to compute the <em>fraction of variants with each mutation that survive the selection.</em>
This quantity is expected to range from zero (if all variants with a specific mutation are inactivated) to one (if all variants with a specific mutation survive the selection).
An example of hypothetical fraction surviving data for such an experiment are in <a class="reference internal" href="#fracsurviveexamplefig"><span class="std std-numref">Fig. 7</span></a>.</p>
<div class="figure align-center" id="id1" style="width: 95%">
<span id="fracsurviveexamplefig"></span><img alt="Illustration of `fracsurvive` definition on hypothetical antibody-escape mutations." src="_images/fracsurvive_fig.jpg" />
<p class="caption"><span class="caption-number">Fig. 7 </span><span class="caption-text">Quantifying the fraction of virions with each mutation that escape antibody neutralization.
This figure shows hypothetical data for four viral variants: wildtype and three mutants.
<strong>(A)</strong> Virions with the V1K mutation (orange) completely survive an antibody concentration where most other virions are neutralized.
<strong>(B)</strong> This resistance is manifested by a large shift in V1K’s neutralization curve.
<strong>(C)</strong> For each dotted vertical line drawn through the neutralization curves in (B), we calculate the fraction of virions with that mutation that survive the antibody, and indicate this fraction by the height of the letter corresponding to that amino acid at that site.
<strong>(D-F)</strong> Similar data to the first three panels, but now V1K has only a small antigenic effect, and so only modestly increases the fraction of virions that survive antibody treatment.</span><a class="headerlink" href="#id1" title="Permalink to this image">¶</a></p>
</div>
<p>The <a class="reference external" href="http://jbloomlab.github.io/dms_tools2">dms_tools2</a> software contains programs to estimate the fraction of each mutation surviving from the overall fraction of the library that survives the selection and the deep sequencing data.
Specifically, you can make these estimates using <a class="reference internal" href="dms2_batch_fracsurvive.html#dms2-batch-fracsurvive"><span class="std std-ref">dms2_batch_fracsurvive</span></a>, which in turn calls <a class="reference internal" href="dms2_fracsurvive.html#dms2-fracsurvive"><span class="std std-ref">dms2_fracsurvive</span></a>.
These programs create visualizations of the results, and you can also visualize the fraction surviving with <a class="reference internal" href="dms2_logoplot.html#dms2-logoplot"><span class="std std-ref">dms2_logoplot</span></a>.</p>
</div>
<div class="section" id="definition-of-fraction-surviving">
<h2><a class="toc-backref" href="#id3">Definition of fraction surviving</a><a class="headerlink" href="#definition-of-fraction-surviving" title="Permalink to this headline">¶</a></h2>
<p>This definition of the fraction surviving was originally introduced in <a class="reference external" href="https://www.nature.com/articles/s41467-018-03665-3">Doud et al (2018)</a>.</p>
<p>We use <span class="math notranslate nohighlight">\(F_{r,x}\)</span> to denote the fraction of variants with character <span class="math notranslate nohighlight">\(x\)</span> (e.g., an amino acid) at site <span class="math notranslate nohighlight">\(r\)</span> that survive the selection.</p>
<p>In order to calculate <span class="math notranslate nohighlight">\(F_{r,x}\)</span>, we need three pieces of information:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>The <strong>overall</strong> fraction of the entire library that survives the selection. We will denote this quantity as <span class="math notranslate nohighlight">\(\gamma\)</span>, and it should range from zero to one. For instance, <span class="math notranslate nohighlight">\(\gamma\)</span> might be calculated by using qPCR to determine the number of virions that infect cells in the presence of antibody versus the number that infect cells in the absence of antibody – the ratio of these two numbers would be the overall fraction of the entire library the survives the selection. You could also compute <span class="math notranslate nohighlight">\(\gamma\)</span> directly from deep-sequencing if your library contains some internal control variant that you know will not be affected by the selection (e.g., a variant with a totally orthogonal envelope protein that is not recognized by the antibody).</p></li>
<li><p>The number of deep sequencing counts of each character <span class="math notranslate nohighlight">\(x\)</span> at each site <span class="math notranslate nohighlight">\(r\)</span> in both the selected (e.g., antibody neutralized) and mock-selected samples. Denote these quantities as <span class="math notranslate nohighlight">\(n_{r,x}^{\rm{selected}}\)</span> and <span class="math notranslate nohighlight">\(n_{r,x}^{\rm{mock}}\)</span>, respectively. Also define variables that give the total number of counts at each site for each condition as <span class="math notranslate nohighlight">\(N_r^{\rm{selected}} = \sum_x n_{r,x}^{\rm{selected}}\)</span> and <span class="math notranslate nohighlight">\(N_r^{\rm{mock}} = \sum_x n_{r,x}^{\rm{mock}}\)</span>.</p></li>
</ol>
</div></blockquote>
<p>Given these pieces of information, the estimated fraction of variants in the selected and mock-selected libraries that contain character <span class="math notranslate nohighlight">\(x\)</span> at site <span class="math notranslate nohighlight">\(r\)</span> are calculated as:</p>
<div class="math notranslate nohighlight" id="equation-fracsurvive-rho">
<span class="eqno">(29)<a class="headerlink" href="#equation-fracsurvive-rho" title="Permalink to this equation">¶</a></span>\[\begin{split}\rho_{r,x}^{\rm{selected}} &amp;=&amp; \frac{n_{r,x}^{\rm{selected}} + f_{r, \rm{selected}} \times P}{N_r^{\rm{selected}} + f_{r, \rm{selected}} \times P \times A} \\
\rho_{r,x}^{\rm{mock}} &amp;=&amp; \frac{n_{r,x}^{\rm{mock}} + f_{r, \rm{mock}} \times P}{N_r^{\rm{mock}} + f_{r, \rm{mock}} \times P \times A} \\\end{split}\]</div>
<p>where <span class="math notranslate nohighlight">\(A\)</span> is the number of characters (e.g., 20 for amino acids), <span class="math notranslate nohighlight">\(P &gt; 0\)</span> is a pseudocount that is added to each observed count (specified by <code class="docutils literal notranslate"><span class="pre">--pseudocount</span></code> option to <a class="reference internal" href="dms2_fracsurvive.html#dms2-fracsurvive"><span class="std std-ref">dms2_fracsurvive</span></a> / <a class="reference internal" href="dms2_batch_fracsurvive.html#dms2-batch-fracsurvive"><span class="std std-ref">dms2_batch_fracsurvive</span></a>), and <span class="math notranslate nohighlight">\(f_{r, \rm{selected}}\)</span> and <span class="math notranslate nohighlight">\(f_{r, \rm{mock}}\)</span> are defined so that the pseudocount is scaled up for the library (<em>mock</em> or <em>selected</em>) with higher depth at site <span class="math notranslate nohighlight">\(r\)</span>:</p>
<div class="math notranslate nohighlight">
\[f_{r, \rm{selected}} = \max\left[1, \left(\sum_x n_{r,x}^{\rm{selected}}\right) / \left(\sum_x n_{r,x}^{\rm{mock}}\right)\right]\]</div>
<div class="math notranslate nohighlight">
\[f_{r, \rm{mock}} = \max\left[1, \left(\sum_x n_{r,x}^{\rm{mock}}\right) / \left(\sum_x n_{r,x}^{\rm{selected}}\right)\right].\]</div>
<p>The reason for scaling the pseudocount by library depth is that the <em>mock</em> and <em>selected</em> libraries may often be sequenced at different depths.
If the same pseudocount is added to both, then the relative frequencies of mutations will be systematically different than one even if the relative counts for the wildtype and mutant amino acid are the same in both two conditions.
Scaling the pseudocounts by the ratio of depths fixes this problem.
If you are using <code class="docutils literal notranslate"><span class="pre">--chartype</span></code> of <code class="docutils literal notranslate"><span class="pre">codon_to_aa</span></code>, the counts for amino acids are aggregated <strong>before</strong> doing the calculations above.</p>
<p>If the mutation has no effect on whether or not a variant survives the selection, then we expect <span class="math notranslate nohighlight">\(\rho_{r,x}^{\rm{selected}} = \rho_{r,x}^{\rm{mock}}\)</span>.
On the other hand, if the mutation makes the variant more likely to survive the selection, then we expected <span class="math notranslate nohighlight">\(\rho_{r,x}^{\rm{selected}} &gt; \rho_{r,x}^{\rm{mock}}\)</span>.
If the <strong>only</strong> variants that survive have the mutation, then we expect <span class="math notranslate nohighlight">\(\rho_{r,x}^{\rm{selected}} = 1\)</span> and the total fraction of the library <span class="math notranslate nohighlight">\(\gamma\)</span> that survives to just be equal to the unselected frequency of this mutation, <span class="math notranslate nohighlight">\(\gamma = \rho_{r,x}^{\rm{mock}}\)</span>.
More generally, the fraction of variants with mutation of <span class="math notranslate nohighlight">\(r\)</span> to <span class="math notranslate nohighlight">\(x\)</span> that survive the selection is</p>
<div class="math notranslate nohighlight" id="equation-fracsurvive">
<span class="eqno">(30)<a class="headerlink" href="#equation-fracsurvive" title="Permalink to this equation">¶</a></span>\[F_{r,x} = \frac{\gamma \times \rho_{r,x}^{\rm{selected}}}{\rho_{r,x}^{\rm{mock}}}.\]</div>
<p>It is this quantity <span class="math notranslate nohighlight">\(F_{r,x}\)</span> that is reported by <a class="reference internal" href="dms2_fracsurvive.html#dms2-fracsurvive"><span class="std std-ref">dms2_fracsurvive</span></a> / <a class="reference internal" href="dms2_batch_fracsurvive.html#dms2-batch-fracsurvive"><span class="std std-ref">dms2_batch_fracsurvive</span></a> as the fraction of variants with the mutation surviving the selection (denoted as <cite>mutfracsurvive</cite>).
Note that this quantity is calculated for the wildtype character at each site as well as all mutant characters.</p>
<p>In addition, these programs report two site-wise measures.
These measures are calculated over all <strong>non</strong>-wildtype characters at each site:</p>
<blockquote>
<div><ul class="simple">
<li><p>The <cite>avgfracsurvive</cite> which is simply the average over all non-wildtype characters of the <cite>mutfracsurvive</cite> values at a site. If some <cite>mutfracsurvive</cite> values are <cite>NaN</cite> (which can happen if you use the <code class="docutils literal notranslate"><span class="pre">--mincounts</span></code> option), they are <strong>not</strong> included in the average.</p></li>
<li><p>The <cite>maxfracsurvive</cite> which is simply the <strong>maximum</strong> <cite>mutfracsurvive</cite> over all non-wildtype characters at a site.</p></li>
</ul>
</div></blockquote>
<div class="section" id="fraction-surviving-above-average">
<h3><a class="toc-backref" href="#id4">Fraction surviving above average</a><a class="headerlink" href="#fraction-surviving-above-average" title="Permalink to this headline">¶</a></h3>
<p>In the section above, <span class="math notranslate nohighlight">\(F_{r,x}\)</span> is defined in Equation <a class="reference internal" href="#equation-fracsurvive">(30)</a> as the fraction of variants that survive given that they have the mutation of site <span class="math notranslate nohighlight">\(r\)</span> to <span class="math notranslate nohighlight">\(x\)</span>.
Note also that <span class="math notranslate nohighlight">\(\gamma\)</span> is the total fraction of the library that survives selection, or equivalently the average fraction that survives weighted over the variant frequencies.</p>
<p>In some cases, it is more useful to show that average fraction of variants that survive <strong>above this library average.</strong>
To do this, we calculate the average fraction that survives above average as</p>
<div class="math notranslate nohighlight" id="equation-fracsurviveaboveavg">
<span class="eqno">(31)<a class="headerlink" href="#equation-fracsurviveaboveavg" title="Permalink to this equation">¶</a></span>\[F_{r,x}^{\rm{aboveavg}} = \max\left(0, F_{r,x} - \gamma\right).\]</div>
<p>If you call <a class="reference internal" href="dms2_fracsurvive.html#dms2-fracsurvive"><span class="std std-ref">dms2_fracsurvive</span></a> or <a class="reference internal" href="dms2_batch_fracsurvive.html#dms2-batch-fracsurvive"><span class="std std-ref">dms2_batch_fracsurvive</span></a> with the <code class="docutils literal notranslate"><span class="pre">--aboveavg</span> <span class="pre">yes</span></code> option then all results use these <span class="math notranslate nohighlight">\(F_{r,x}^{\rm{aboveavg}}\)</span> in place of <span class="math notranslate nohighlight">\(F_{r,x}\)</span> values.</p>
<p>The distinction is as follows.
If you want your results to reflect the average fraction of variants with each mutation that survive, then use the default <code class="docutils literal notranslate"><span class="pre">--aboveavg</span> <span class="pre">no</span></code> option.
In this case, you expect the fraction surviving to be &gt; 0 for most variants, since a bit of your library probably gets through the selection even without an advantageous mutation.
On the other hand, if you just want to look for mutations that are enhancing how variants perform above the average, then use the <code class="docutils literal notranslate"><span class="pre">--aboveavg</span> <span class="pre">yes</span></code> option.
Then you only see fraction surviving &gt; 0 for variants that are favored during the selection – but the quantitative values of this fraction surviving are no longer directly interpretable as the true fraction surviving, since they are in reference to the library average baseline.</p>
</div>
<div class="section" id="error-correction">
<h3><a class="toc-backref" href="#id5">Error correction</a><a class="headerlink" href="#error-correction" title="Permalink to this headline">¶</a></h3>
<p>You can optionally correct for the potential inflation of some counts by sequencing errors by using the <code class="docutils literal notranslate"><span class="pre">--err</span></code> option to <a class="reference internal" href="dms2_fracsurvive.html#dms2-fracsurvive"><span class="std std-ref">dms2_fracsurvive</span></a> / <a class="reference internal" href="dms2_batch_fracsurvive.html#dms2-batch-fracsurvive"><span class="std std-ref">dms2_batch_fracsurvive</span></a>.
To do this, you should use some control where the <strong>only</strong> mutations will come from sequencing / library-prep-induced errors that are presumed to occur at the same rate as the actual samples being analyzed.
Typically, this would be deep sequencing of a wildtype control.</p>
<p>With an error control, the counts are adjusted as follows.
Let <span class="math notranslate nohighlight">\(n_{r,x}\)</span> be the counts for character <span class="math notranslate nohighlight">\(x\)</span> at site <span class="math notranslate nohighlight">\(r\)</span> in the <em>mock</em> or <em>selected</em> sample.
Let <span class="math notranslate nohighlight">\(n^{\rm{err}}_{r,x}\)</span> be the number of counts of <span class="math notranslate nohighlight">\(x\)</span> at site <span class="math notranslate nohighlight">\(r\)</span> in the error control.
Define</p>
<div class="math notranslate nohighlight">
\[\epsilon_{r,x} = \left(n^{\rm{err}}_{r,x}\right) / \left(\sum_y n^{\rm{err}}_{r,y}\right).\]</div>
<p>When <span class="math notranslate nohighlight">\(x \ne \operatorname{wt}\left(r\right)\)</span> then <span class="math notranslate nohighlight">\(\epsilon_{r,x}\)</span> is the rate of errors to <span class="math notranslate nohighlight">\(x\)</span> at site <span class="math notranslate nohighlight">\(r\)</span>, and when <span class="math notranslate nohighlight">\(x = \operatorname{wt}\left(r\right)\)</span> then <span class="math notranslate nohighlight">\(\epsilon_{r,x}\)</span> is one minus the rate of errors away from the wildtype at site <span class="math notranslate nohighlight">\(r\)</span>.</p>
<p>We then adjust observed counts <span class="math notranslate nohighlight">\(n_{r,x}\)</span> to the error-corrected counts <span class="math notranslate nohighlight">\(\hat{n}_{r,x}\)</span> by</p>
<div class="math notranslate nohighlight">
\[\begin{split}\hat{n}_{r,x} = \begin{cases}
\max\left[\left(\sum_y n_{r,y}\right) \left(\frac{n_{r,x}}{\sum_y n_{r,y}} - \epsilon_{r,x}\right), 0\right] &amp; \mbox{if } x \ne \operatorname{wt}\left(r\right) \\
n_{r,x} / \epsilon_{r,x} &amp; \mbox{if } x = \operatorname{wt}\left(r\right) \\
\end{cases}\end{split}\]</div>
<p>These adjusted counts are then used in place of the un-adjusted counts in Equation <a class="reference internal" href="#equation-fracsurvive-rho">(29)</a> above.</p>
</div>
</div>
<div class="section" id="programs-and-examples">
<h2><a class="toc-backref" href="#id6">Programs and examples</a><a class="headerlink" href="#programs-and-examples" title="Permalink to this headline">¶</a></h2>
<p>The <a class="reference external" href="http://jbloomlab.github.io/dms_tools2">dms_tools2</a> software contains programs to calculate the fraction of each mutation surviving:</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="dms2_fracsurvive.html#dms2-fracsurvive"><span class="std std-ref">dms2_fracsurvive</span></a> estimates the fraction surviving from counts files and the overall library fractiont that survives.</p></li>
<li><p><a class="reference internal" href="dms2_batch_fracsurvive.html#dms2-batch-fracsurvive"><span class="std std-ref">dms2_batch_fracsurvive</span></a> runs <a class="reference internal" href="dms2_fracsurvive.html#dms2-fracsurvive"><span class="std std-ref">dms2_fracsurvive</span></a> on a set of samples and summarizes the results. It is generally easier to use <a class="reference internal" href="dms2_batch_fracsurvive.html#dms2-batch-fracsurvive"><span class="std std-ref">dms2_batch_fracsurvive</span></a> rather than runnning <a class="reference internal" href="dms2_fracsurvive.html#dms2-fracsurvive"><span class="std std-ref">dms2_fracsurvive</span></a> directly.</p></li>
<li><p><a class="reference internal" href="dms2_logoplot.html#dms2-logoplot"><span class="std std-ref">dms2_logoplot</span></a> can be used to create logo plots that visualize the fraction surviving.</p></li>
</ul>
</div></blockquote>
<p>Here are some example analyses:</p>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference external" href="https://github.com/jbloomlab/dms_tools2/blob/master/examples/Doud2018/analysis_notebook.ipynb">Doud2018 example</a></p></li>
</ul>
</div></blockquote>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">dms_tools2</h1>
    
  </a>
</p>



<p class="blurb">Tools for analyzing deep mutational scanning data.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=dms_tools2&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a class="badge" href="https://travis-ci.org/jbloomlab/dms_tools2">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="bcsubamp.html">Barcoded-subamplicon sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="prefs.html">Amino-acid preferences</a></li>
<li class="toctree-l1"><a class="reference internal" href="diffsel.html">Differential selection</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Fraction surviving</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="#definition-of-fraction-surviving">Definition of fraction surviving</a></li>
<li class="toctree-l2"><a class="reference internal" href="#programs-and-examples">Programs and examples</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="programs.html">Executable programs</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">Python API</a></li>
<li class="toctree-l1"><a class="reference internal" href="citations.html">Citations</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="diffsel.html" title="previous chapter">Differential selection</a></li>
      <li>Next: <a href="programs.html" title="next chapter">Executable programs</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017--2020, the Bloom lab.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/fracsurvive.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/dms_tools2" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>