{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis of deep mutational scanning of the hemagglutinin of WSN influenza virus by Doud and Bloom (2016)\n",
    "\n",
    "## Overview\n",
    "[Doud and Bloom (2016)](http://www.mdpi.com/1999-4915/8/6/155) performed deep mutational scanning on the hemagglutinin (HA) from A/WSN/1933 (H1N1) influenza virus.\n",
    "In that original paper, the data was analyzed using the older [dms_tools](https://github.com/jbloomlab/dms_tools) software package.\n",
    "Here we re-analyze the data using the newer [dms_tools2](https://github.com/jbloomlab/dms_tools2) software.\n",
    "\n",
    "## Experimental summary\n",
    "The goal of the experiment was to quantify the effect of all amino-acid mutations to HA on viral replication in cell culture.\n",
    "To do this, [Doud and Bloom (2016)](http://www.mdpi.com/1999-4915/8/6/155) created three independent plasmid mutant libraries carrying nearly all codon mutations of HA. \n",
    "These plasmid mutant libraries are referred to as *mutDNA-1*, *mutDNA-2*, and *mutDNA-3*.\n",
    "They then used these libraries to generate three libraries of mutant viruses, which they passaged at low MOI to create a genotype-phenotype link.\n",
    "These three virus mutant libraries are referred to as *mutvirus-1*, *mutvirus-2*, and *mutvirus-3*.\n",
    "In addition, they performed a single replicate of the some procedure using **wildtype** (unmutated plasmid) to create non-mutagenized virus -- these plasmid and virus samples are referred to as *wtDNA* and *wtvirus*.\n",
    "\n",
    "All of these plasmid and virus samples were deep sequenced using barcoded-subamplicon sequencing to obtain high accuracy.\n",
    "This notebook analyzes those data to determine the \"preference\" of each site in HA for each possible amino acid."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import modules and define general functions / variables\n",
    "Import the Python modules used by this notebook, print the version of [dms_tools2](https://github.com/jbloomlab/dms_tools2) being used, and define some general functions.\n",
    "\n",
    "Some key variables defined below:\n",
    "\n",
    "* **ncpus** is the number of CPUs to use. This number should not exceed the number of CPUs that you specify if you are submitting your job using some queue-ing system such as `slurm` / `sbatch`.\n",
    "\n",
    "* **use_existing** specifies whether we use existing results if they exist, or specify everything new. Note that `dms_tools2` programs just check for existing output but do not do true dependency tracking (as would be done by `make`, for instance) -- so if you have changed key inputs then set this option to *False* to re-run everything fresh."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using dms_tools2 version 2.0.dev0\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "from collections import OrderedDict\n",
    "import pandas\n",
    "from IPython.display import display, HTML\n",
    "import dms_tools2\n",
    "\n",
    "print(\"Using dms_tools2 version {0}\".format(dms_tools2.__version__))\n",
    "\n",
    "# results will go in this directory\n",
    "resultsdir = './results/' \n",
    "if not os.path.isdir(resultsdir):\n",
    "    os.mkdir(resultsdir)\n",
    "    \n",
    "# CPUs to use, should not exceed the number you request with slurm\n",
    "ncpus = 14\n",
    "\n",
    "# do we use existing results or generate everything new?\n",
    "use_existing = True"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download the FASTQ files from the SRA\n",
    "The first step is to obtain the FASTQ files with the deep sequencing data.\n",
    "[Doud and Bloom (2016)](http://www.mdpi.com/1999-4915/8/6/155) have already transferred these FASTQ files from their local server to the [Sequence Read Archive](https://www.ncbi.nlm.nih.gov/sra), so we simply download those files.\n",
    "\n",
    "We download the FASQ files using the [fastq-dump](https://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=toolkit_doc&f=fastq-dump) program from the [SRA Toolkit](https://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=toolkit_doc).\n",
    "If you do not already have this toolkit installed, you will need to install it by [following these instructions](https://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=toolkit_doc&f=std).\n",
    "\n",
    "The code cell below does the following:\n",
    "1. Defines the SRA accession number for each sample.\n",
    "2. Downloads each sample using `fastq-dump`. To understand the commands used by `fastq-dump` below, [look here](https://edwards.sdsu.edu/research/fastq-dump/) since the [NCBI's instructions](https://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=toolkit_doc&f=fastq-dump) are sparse.\n",
    "3. Renames each set of FASTQ files to have nice names like `mutDNA-1_R1.fastq.gz`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading FASTQ files from the SRA using:\n",
      "\r\n",
      "fastq-dump : 2.8.2\r\n",
      "\r\n",
      "FASTQ files for mutDNA-1 already exist\n",
      "FASTQ files for mutDNA-2 already exist\n",
      "FASTQ files for mutDNA-3 already exist\n",
      "FASTQ files for mutvirus-1 already exist\n",
      "FASTQ files for mutvirus-2 already exist\n",
      "FASTQ files for mutvirus-3 already exist\n",
      "FASTQ files for wtDNA already exist\n",
      "FASTQ files for wtvirus already exist\n"
     ]
    }
   ],
   "source": [
    "# directory for FASTQ files\n",
    "fastqdir = './FASTQ_files/' \n",
    "if not os.path.isdir(fastqdir):\n",
    "    os.mkdir(fastqdir)\n",
    "\n",
    "# define SRA accessions for each sample\n",
    "sample_to_acc = OrderedDict([\n",
    "        ('mutDNA-1', 'SRR3113656'),\n",
    "        ('mutDNA-2', 'SRR3113657'),\n",
    "        ('mutDNA-3', 'SRR3113658'),\n",
    "        ('mutvirus-1', 'SRR3113660'),\n",
    "        ('mutvirus-2', 'SRR3113661'),\n",
    "        ('mutvirus-3', 'SRR3113662'),\n",
    "        ('wtDNA', 'SRR3113655'),\n",
    "        ('wtvirus', 'SRR3113659'),\n",
    "        ])\n",
    "\n",
    "print(\"Downloading FASTQ files from the SRA using:\")\n",
    "!fastq-dump --version\n",
    "\n",
    "for (sample, acc) in sample_to_acc.items():\n",
    "    # r1file / r2file are final names we will assign files\n",
    "    r1file = '{0}/{1}_R1.fastq.gz'.format(fastqdir, sample)\n",
    "    r2file = r1file.replace('R1', 'R2')\n",
    "    # r1filedownload / r2filedownload are initial names after fastq-dump download\n",
    "    r1filedownload = '{0}/{1}_pass_1.fastq.gz'.format(fastqdir, acc)\n",
    "    r2filedownload = '{0}/{1}_pass_2.fastq.gz'.format(fastqdir, acc)\n",
    "    if all(map(os.path.isfile, [r1file, r2file])):\n",
    "        print(\"FASTQ files for {0} already exist\".format(sample))\n",
    "    else:\n",
    "        print(\"Downloading FASTQ files for {0}...\".format(sample))\n",
    "        !fastq-dump --outdir {fastqdir} --gzip --readids --skip-technical \\\n",
    "                --dumpbase --clip --split-files --read-filter pass {acc}\n",
    "        os.rename(r1filedownload, r1file)\n",
    "        os.rename(r2filedownload, r2file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Align the deep sequencing data and count mutations\n",
    "The alignments are made to the wildtype WSN HA coding sequence, which is available at [./data/WSN-HA.fasta](./data/WSN-HA.fasta)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We will process the following samples:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>name</th>\n",
       "      <th>R1</th>\n",
       "      <th>plotgroup</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>mutDNA-1</td>\n",
       "      <td>mutDNA-1_R1.fastq.gz</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mutDNA-2</td>\n",
       "      <td>mutDNA-2_R1.fastq.gz</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mutDNA-3</td>\n",
       "      <td>mutDNA-3_R1.fastq.gz</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mutvirus-1</td>\n",
       "      <td>mutvirus-1_R1.fastq.gz</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mutvirus-2</td>\n",
       "      <td>mutvirus-2_R1.fastq.gz</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mutvirus-3</td>\n",
       "      <td>mutvirus-3_R1.fastq.gz</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>wtDNA</td>\n",
       "      <td>wtDNA_R1.fastq.gz</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>wtvirus</td>\n",
       "      <td>wtvirus_R1.fastq.gz</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now running dms2_batch_bcsubamp...\n",
      "2017-08-09 08:14:11,247 - INFO - Beginning execution of dms2_batch_bcsubamp in directory /home/jbloom/dms_tools2/examples/Doud2016\r\n",
      "\r\n",
      "2017-08-09 08:14:11,247 - INFO - Progress is being logged to ./results/codoncounts/summary.log\r\n",
      "2017-08-09 08:14:11,247 - INFO - Version information:\r\n",
      "\tTime and date: Wed Aug  9 08:14:11 2017\r\n",
      "\tPlatform: Linux-3.13.0-68-generic-x86_64-with-Ubuntu-14.04-trusty\r\n",
      "\tPython version: 3.4.3 (default, Oct 14 2015, 20:28:29)  [GCC 4.8.4]\r\n",
      "\tdms_tools2 version: 2.0.dev0\r\n",
      "\tBio version: 1.68\r\n",
      "\tHTSeq version: 0.9.1\r\n",
      "\tpandas version: 0.20.3\r\n",
      "\tnumpy version: 1.13.1\r\n",
      "\r\n",
      "2017-08-09 08:14:11,247 - INFO - Parsed the following arguments:\r\n",
      "\tR1trim = [200]\r\n",
      "\tR2trim = [170]\r\n",
      "\tbclen = 8\r\n",
      "\tminreads = 2\r\n",
      "\toutdir = ./results/codoncounts\r\n",
      "\tminfraccall = 0.9\r\n",
      "\tseed = 1\r\n",
      "\tsummaryprefix = summary\r\n",
      "\tuse_existing = True\r\n",
      "\tbcinfo = False\r\n",
      "\tpurgeread = 0\r\n",
      "\talignspecs = ['1,285,36,37', '286,570,31,32', '571,855,37,32', '856,1140,31,36', '1141,1425,29,33', '1426,1698,40,43']\r\n",
      "\tmaxmuts = 4\r\n",
      "\tR2 = None\r\n",
      "\tncpus = 14\r\n",
      "\trefseq = ./data/WSN-HA.fasta\r\n",
      "\tpurgebc = 0\r\n",
      "\tbatchfile = ./results/codoncounts/batch.csv\r\n",
      "\tminq = 15\r\n",
      "\tchartype = codon\r\n",
      "\tminconcur = 0.75\r\n",
      "\tfastqdir = ./FASTQ_files/\r\n",
      "\r\n",
      "2017-08-09 08:14:11,247 - INFO - Parsing sample info from ./results/codoncounts/batch.csv\r\n",
      "2017-08-09 08:14:11,255 - INFO - Read the following sample information:\r\n",
      "name,R1,plotgroup\r\n",
      "mutDNA-1,mutDNA-1_R1.fastq.gz,\r\n",
      "mutDNA-2,mutDNA-2_R1.fastq.gz,\r\n",
      "mutDNA-3,mutDNA-3_R1.fastq.gz,\r\n",
      "mutvirus-1,mutvirus-1_R1.fastq.gz,\r\n",
      "mutvirus-2,mutvirus-2_R1.fastq.gz,\r\n",
      "mutvirus-3,mutvirus-3_R1.fastq.gz,\r\n",
      "wtDNA,wtDNA_R1.fastq.gz,\r\n",
      "wtvirus,wtvirus_R1.fastq.gz,\r\n",
      "\r\n",
      "\r\n",
      "2017-08-09 08:14:11,255 - INFO - Running dms2_bcsubamp on all samples using 14 CPUs...\r\n",
      "2017-08-09 08:14:13,469 - INFO - Completed runs of dms2_bcsubamp.\r\n",
      "\r\n",
      "2017-08-09 08:14:13,473 - INFO - Making summary plots for the 1 plotgroups.\r\n",
      "         name                      R1 plotgroup  \\\r\n",
      "0    mutDNA-1    mutDNA-1_R1.fastq.gz             \r\n",
      "1    mutDNA-2    mutDNA-2_R1.fastq.gz             \r\n",
      "2    mutDNA-3    mutDNA-3_R1.fastq.gz             \r\n",
      "3  mutvirus-1  mutvirus-1_R1.fastq.gz             \r\n",
      "4  mutvirus-2  mutvirus-2_R1.fastq.gz             \r\n",
      "5  mutvirus-3  mutvirus-3_R1.fastq.gz             \r\n",
      "6       wtDNA       wtDNA_R1.fastq.gz             \r\n",
      "7     wtvirus     wtvirus_R1.fastq.gz             \r\n",
      "\r\n",
      "                                             counts  \\\r\n",
      "0    ./results/codoncounts/mutDNA-1_codoncounts.csv   \r\n",
      "1    ./results/codoncounts/mutDNA-2_codoncounts.csv   \r\n",
      "2    ./results/codoncounts/mutDNA-3_codoncounts.csv   \r\n",
      "3  ./results/codoncounts/mutvirus-1_codoncounts.csv   \r\n",
      "4  ./results/codoncounts/mutvirus-2_codoncounts.csv   \r\n",
      "5  ./results/codoncounts/mutvirus-3_codoncounts.csv   \r\n",
      "6       ./results/codoncounts/wtDNA_codoncounts.csv   \r\n",
      "7     ./results/codoncounts/wtvirus_codoncounts.csv   \r\n",
      "\r\n",
      "                                        readstats  \\\r\n",
      "0    ./results/codoncounts/mutDNA-1_readstats.csv   \r\n",
      "1    ./results/codoncounts/mutDNA-2_readstats.csv   \r\n",
      "2    ./results/codoncounts/mutDNA-3_readstats.csv   \r\n",
      "3  ./results/codoncounts/mutvirus-1_readstats.csv   \r\n",
      "4  ./results/codoncounts/mutvirus-2_readstats.csv   \r\n",
      "5  ./results/codoncounts/mutvirus-3_readstats.csv   \r\n",
      "6       ./results/codoncounts/wtDNA_readstats.csv   \r\n",
      "7     ./results/codoncounts/wtvirus_readstats.csv   \r\n",
      "\r\n",
      "                                        readsperbc  \\\r\n",
      "0    ./results/codoncounts/mutDNA-1_readsperbc.csv   \r\n",
      "1    ./results/codoncounts/mutDNA-2_readsperbc.csv   \r\n",
      "2    ./results/codoncounts/mutDNA-3_readsperbc.csv   \r\n",
      "3  ./results/codoncounts/mutvirus-1_readsperbc.csv   \r\n",
      "4  ./results/codoncounts/mutvirus-2_readsperbc.csv   \r\n",
      "5  ./results/codoncounts/mutvirus-3_readsperbc.csv   \r\n",
      "6       ./results/codoncounts/wtDNA_readsperbc.csv   \r\n",
      "7     ./results/codoncounts/wtvirus_readsperbc.csv   \r\n",
      "\r\n",
      "                                        bcstats  \r\n",
      "0    ./results/codoncounts/mutDNA-1_bcstats.csv  \r\n",
      "1    ./results/codoncounts/mutDNA-2_bcstats.csv  \r\n",
      "2    ./results/codoncounts/mutDNA-3_bcstats.csv  \r\n",
      "3  ./results/codoncounts/mutvirus-1_bcstats.csv  \r\n",
      "4  ./results/codoncounts/mutvirus-2_bcstats.csv  \r\n",
      "5  ./results/codoncounts/mutvirus-3_bcstats.csv  \r\n",
      "6       ./results/codoncounts/wtDNA_bcstats.csv  \r\n",
      "7     ./results/codoncounts/wtvirus_bcstats.csv  \r\n",
      "2017-08-09 08:14:13,479 - INFO - Successful completion of dms2_batch_bcsubamp\r\n"
     ]
    }
   ],
   "source": [
    "# file containing wildtype WSN HA sequence\n",
    "refseq = './data/WSN-HA.fasta'\n",
    "\n",
    "# define subamplicon alignment specifications\n",
    "alignspecs = ' '.join(['1,285,36,37', \n",
    "                       '286,570,31,32',\n",
    "                       '571,855,37,32',\n",
    "                       '856,1140,31,36',\n",
    "                       '1141,1425,29,33',\n",
    "                       '1426,1698,40,43'])\n",
    "\n",
    "# counts and alignments placed in this directory\n",
    "countsdir = os.path.join(resultsdir, 'codoncounts')\n",
    "if not os.path.isdir(countsdir):\n",
    "    os.mkdir(countsdir)\n",
    "\n",
    "# countsbatch is a dataframe that defines batch input to dms2_batch_bcsubamp\n",
    "countsbatch = pandas.DataFrame({'name':list(sample_to_acc.keys())})\n",
    "countsbatch['R1'] = countsbatch['name'].astype('str') + '_R1.fastq.gz'\n",
    "countsbatch['plotgroup'] = '' # all samples are in the same plotgroup\n",
    "print(\"We will process the following samples:\")\n",
    "display(HTML(countsbatch.to_html(index=False)))\n",
    "countsbatchfile = os.path.join(countsdir, 'batch.csv')\n",
    "countsbatch.to_csv(countsbatchfile, index=False)\n",
    "\n",
    "print('\\nNow running dms2_batch_bcsubamp...')\n",
    "!dms2_batch_bcsubamp \\\n",
    "        --batchfile {countsbatchfile} \\\n",
    "        --refseq {refseq} \\\n",
    "        --alignspecs {alignspecs} \\\n",
    "        --outdir {countsdir} \\\n",
    "        --summaryprefix summary \\\n",
    "        --R1trim 200 \\\n",
    "        --R2trim 170 \\\n",
    "        --fastqdir {fastqdir} \\\n",
    "        --ncpus {ncpus} \\\n",
    "        --use_existing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
