
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>dms_tools2.barcodes &#8212; dms_tools2 2.6.3 documentation</title>
    <link rel="stylesheet" href="../../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <script type="text/javascript" src="../../_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
   
  <link rel="stylesheet" href="../../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <h1>Source code for dms_tools2.barcodes</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">===============</span>
<span class="sd">barcodes</span>
<span class="sd">===============</span>

<span class="sd">Operations for sequence barcodes and UMIs.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">import</span> <span class="nn">tempfile</span>

<span class="kn">import</span> <span class="nn">numpy</span>
<span class="kn">import</span> <span class="nn">pandas</span>
<span class="kn">import</span> <span class="nn">regex</span>
<span class="kn">import</span> <span class="nn">umi_tools.network</span>
<span class="kn">import</span> <span class="nn">Bio.SeqUtils.ProtParamData</span>

<span class="c1"># use plotnine for plotting</span>
<span class="kn">from</span> <span class="nn">plotnine</span> <span class="k">import</span> <span class="o">*</span>

<span class="kn">import</span> <span class="nn">dms_tools2.plot</span>
<span class="kn">from</span> <span class="nn">dms_tools2.plot</span> <span class="k">import</span> <span class="n">COLOR_BLIND_PALETTE_GRAY</span>
<span class="kn">import</span> <span class="nn">dms_tools2.utils</span>
<span class="kn">import</span> <span class="nn">dms_tools2.pacbio</span>
<span class="kn">from</span> <span class="nn">dms_tools2</span> <span class="k">import</span> <span class="n">CODON_TO_AA</span><span class="p">,</span> <span class="n">CODONS</span><span class="p">,</span> <span class="n">AAS_WITHSTOP</span><span class="p">,</span> <span class="n">AA_TO_CODONS</span>

<span class="n">_umi_clusterer</span> <span class="o">=</span> <span class="n">umi_tools</span><span class="o">.</span><span class="n">network</span><span class="o">.</span><span class="n">UMIClusterer</span><span class="p">()</span>



<div class="viewcode-block" id="almost_duplicated"><a class="viewcode-back" href="../../dms_tools2.barcodes.html#dms_tools2.barcodes.almost_duplicated">[docs]</a><span class="k">def</span> <span class="nf">almost_duplicated</span><span class="p">(</span><span class="n">barcodes</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Identifies nearly identical barcodes.</span>

<span class="sd">    This function mimics the pandas `duplicated`</span>
<span class="sd">    function except it can also mark as duplicated almost</span>
<span class="sd">    but not exactly identical sequences.</span>

<span class="sd">    Args:</span>
<span class="sd">        `barcodes` (list or pandas Series)</span>
<span class="sd">            Lists all the barcodes, which should be strings</span>
<span class="sd">            of same length.</span>
<span class="sd">        `threshold` (int)</span>
<span class="sd">            Max number of mismatches for barcodes to be</span>
<span class="sd">            considered almost identical.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A pandas Series of the same length as `barcodes`</span>
<span class="sd">        filled with bools that indicates whether a barcode</span>
<span class="sd">        is a near duplicate. For each group of barcodes within</span>
<span class="sd">        `threshold` of each other, only one barcode will have</span>
<span class="sd">        an entry of `False` (not duplicated) and the rest will</span>
<span class="sd">        be `True` (duplicated). The barcode listed as not duplicated</span>
<span class="sd">        is chosen as follows: (1) it has the most abundant barcode</span>
<span class="sd">        in the group, (2) among barcodes that are equivalent abundant</span>
<span class="sd">        we take the one listed first in `barcodes`.</span>
<span class="sd">        Groups are computed using the directional method of</span>
<span class="sd">        `umi_tools &lt;https://github.com/CGATOxford/UMI-tools&gt;`_.</span>

<span class="sd">    When `threshold` is zero, just like `pandas.duplicated`:</span>

<span class="sd">    &gt;&gt;&gt; barcodes = [&#39;CTC&#39;, &#39;ATG&#39;, &#39;ATG&#39;, &#39;ATA&#39;]</span>
<span class="sd">    &gt;&gt;&gt; almost_duplicated(barcodes, threshold=0).equals(</span>
<span class="sd">    ...     pandas.Series([False, False, True, False]))</span>
<span class="sd">    True</span>
<span class="sd">    &gt;&gt;&gt; pandas.Series(barcodes).duplicated().equals(</span>
<span class="sd">    ...     pandas.Series([False, False, True, False]))</span>
<span class="sd">    True</span>

<span class="sd">    But when `threshold` is 1, also identifies **almost** equal</span>
<span class="sd">    barcodes (in this case, ones that differ by &lt;= 1 mutation:</span>

<span class="sd">    &gt;&gt;&gt; almost_duplicated(barcodes, threshold=1).equals(</span>
<span class="sd">    ...     pandas.Series([False, False, True, True]))</span>
<span class="sd">    True</span>

<span class="sd">    All barcodes are within a distance of two, so all are</span>
<span class="sd">    near duplicates. Note how we mark as not duplicated</span>
<span class="sd">    (`False`) the first one listed among the most abundant</span>
<span class="sd">    barcode (&#39;ATG&#39;):</span>

<span class="sd">    &gt;&gt;&gt; almost_duplicated(pandas.Series(barcodes), threshold=2).equals(</span>
<span class="sd">    ...     pandas.Series([True, False, True, True]))</span>
<span class="sd">    True</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">threshold</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;`threshold` must be &gt;= 0&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">barcodes</span><span class="p">,</span> <span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">):</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">barcodes</span><span class="p">,</span> <span class="n">collections</span><span class="o">.</span><span class="n">Iterable</span><span class="p">):</span>
            <span class="n">barcodes</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">barcodes</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;`barcodes` invalid type {type(barcodes)}&quot;</span><span class="p">)</span>
    <span class="n">barcodes</span> <span class="o">=</span> <span class="n">barcodes</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span>

    <span class="n">counts</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">Counter</span><span class="p">(</span><span class="n">barcodes</span><span class="p">)</span>

    <span class="n">groups</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">_umi_clusterer</span><span class="p">(</span><span class="n">counts</span><span class="p">,</span> <span class="n">threshold</span><span class="p">):</span>
        <span class="c1"># keep only barcode(s) most abundant in group</span>
        <span class="n">max_count</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">counts</span><span class="p">[</span><span class="n">barcode</span><span class="p">]</span> <span class="k">for</span> <span class="n">barcode</span> <span class="ow">in</span> <span class="n">group</span><span class="p">)</span>
        <span class="n">groups</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">frozenset</span><span class="p">(</span><span class="n">barcode</span> <span class="k">for</span> <span class="n">barcode</span> <span class="ow">in</span> <span class="n">group</span>
                      <span class="k">if</span> <span class="n">counts</span><span class="p">[</span><span class="n">barcode</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">max_count</span><span class="p">))</span>

    <span class="n">dups</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">barcode</span> <span class="ow">in</span> <span class="n">barcodes</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">groups</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">barcode</span> <span class="ow">in</span> <span class="n">g</span><span class="p">:</span>
                <span class="n">dups</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">groups</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">g</span><span class="p">)</span>
                <span class="k">break</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">dups</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">assert</span> <span class="ow">not</span> <span class="n">groups</span>

    <span class="k">return</span> <span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">dups</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">barcodes</span><span class="o">.</span><span class="n">index</span><span class="p">)</span></div>


<div class="viewcode-block" id="simpleConsensus"><a class="viewcode-back" href="../../dms_tools2.barcodes.html#dms_tools2.barcodes.simpleConsensus">[docs]</a><span class="k">def</span> <span class="nf">simpleConsensus</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span>
        <span class="n">barcode_col</span><span class="o">=</span><span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="n">substitution_col</span><span class="o">=</span><span class="s1">&#39;substitutions&#39;</span><span class="p">,</span>
        <span class="n">insertion_col</span><span class="o">=</span><span class="s1">&#39;insertions&#39;</span><span class="p">,</span> <span class="n">deletion_col</span><span class="o">=</span><span class="s1">&#39;deletions&#39;</span><span class="p">,</span>
        <span class="n">library_col</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">max_sub_diffs</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">max_indel_diffs</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
        <span class="n">max_minor_muts</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Simple method to get consensus of mutations within barcode.</span>

<span class="sd">    Args:</span>
<span class="sd">        `df` (pandas Data Frame)</span>
<span class="sd">            Holds variants and their barcodes. Each row gives a</span>
<span class="sd">            sequence variant and its barcode. There need to be</span>
<span class="sd">            columns with the names given by the next four arguments</span>
<span class="sd">            described below.</span>
<span class="sd">        `barcode_col` (str)</span>
<span class="sd">            Name of column holding barcodes.</span>
<span class="sd">        `substitution_col` (str)</span>
<span class="sd">            Name of column holding substitutions as list of strings.</span>
<span class="sd">        `insertion_col` (str)</span>
<span class="sd">            Name of column holding insertions as list of strings.</span>
<span class="sd">        `deletion_col` (str)</span>
<span class="sd">            Name of column holding insertions as list of strings.</span>
<span class="sd">        `library_col` (`None` or str)</span>
<span class="sd">            If we have multiple libraries, analyze each barcode only</span>
<span class="sd">            within its library. In that case, `library_col` should be</span>
<span class="sd">            name of column giving library name.</span>
<span class="sd">        `max_sub_diffs` (int)</span>
<span class="sd">            Drop any barcode where any variant differs from all other</span>
<span class="sd">            variants for that barcode by more than this many substitution</span>
<span class="sd">            (point mutation) differences.</span>
<span class="sd">        `max_indel_diffs` (int)</span>
<span class="sd">            Drop any barcode where any variant differs from all other</span>
<span class="sd">            variants for that barcode by more than this many indel</span>
<span class="sd">            (insertion or deletion) differences.</span>
<span class="sd">        `max_minor_muts` (int)</span>
<span class="sd">            Drop any barcode where there is a minor (non-consensus)</span>
<span class="sd">            mutation found more than this many times.</span>

<span class="sd">    Returns:</span>
<span class="sd">        The 2-tuple `(consensus, dropped)`. These are each data frames:</span>

<span class="sd">            - `consensus` is a new data frame with a row for each</span>
<span class="sd">              barcode for which we could call the consensus. The</span>
<span class="sd">              columns have the same names as `barcode_col`,</span>
<span class="sd">              `substitution_col`, `insertion_col`, `deletion_col`,</span>
<span class="sd">              and (optionally) `library_col`--but the the three</span>
<span class="sd">              columns for the mutations now just list the **consensus**</span>
<span class="sd">              mutations of that type. In addition, there is a new</span>
<span class="sd">              column called &quot;variant_call_support&quot; that gives the number</span>
<span class="sd">              of sequences supporting the call of that barcode.</span>

<span class="sd">            - `dropped` simply contains all rows in the original `df`</span>
<span class="sd">              that correspond to sequences that were dropped due to</span>
<span class="sd">              `max_diffs` or `max_minor_muts` not being satisfied.</span>
<span class="sd">              There is also a column called &quot;drop_reason&quot; that gives</span>
<span class="sd">              the reason that the barcode was dropped.</span>

<span class="sd">    The approach is as follows:</span>

<span class="sd">      1. Group all variants by library and barcode.</span>

<span class="sd">      2. If there are multiple sequences, check if any of them differ</span>
<span class="sd">         from all the others by more than `max_diffs` mutations total</span>
<span class="sd">         (taking substitutions, insertions, and deletions together).</span>
<span class="sd">         If so, drop the entire barcode. The reason is that if there</span>
<span class="sd">         are variants that are very different, it becomes likely that</span>
<span class="sd">         it isn&#39;t just sequencing error, but rather something is wrong</span>
<span class="sd">         with that barcode (multiple variants with same barcode or</span>
<span class="sd">         strand exchange).</span>

<span class="sd">      3. Take the consensus of the sequences, which means keeping</span>
<span class="sd">         mutations that are present in **greater** than half of the</span>
<span class="sd">         variants. Note that this calling scheme means that the</span>
<span class="sd">         consensus being called is dependent on the reference used</span>
<span class="sd">         to call the mutations, which is an important caveat if you</span>
<span class="sd">         are calling variants relative to multiple different parent</span>
<span class="sd">         sequences.</span>

<span class="sd">      4. If there are any minor mutations (mutations not in consensus)</span>
<span class="sd">         that are present in more than `max_minor_muts` variants or missing</span>
<span class="sd">         from more than `max_minor_muts` variants, then</span>
<span class="sd">         drop that barcode. The reason is that recurring minor mutations</span>
<span class="sd">         also suggest some problem more complex than sequencing error</span>
<span class="sd">         that may render the whole barcode family invalid.</span>

<span class="sd">    Note that this method returns a consensus even if there is just</span>
<span class="sd">    one sequence for the barcode (in that case, this sequence is</span>
<span class="sd">    the consensus). This is fine--if you want to get consensus calls</span>
<span class="sd">    that are more strongly supported, simply filter the returned</span>
<span class="sd">    `consensus` data frame for larger values of `variant_call_support`,</span>
<span class="sd">    as the more sequences that support a call the more accurate it</span>
<span class="sd">    is expected to be.</span>

<span class="sd">    Here is an example:</span>

<span class="sd">    &gt;&gt;&gt; df = pandas.DataFrame([</span>
<span class="sd">    ...     (&#39;s1&#39;, &#39;AG&#39;, [&#39;A2C&#39;], [], [&#39;del5to7&#39;]),</span>
<span class="sd">    ...     (&#39;s1&#39;, &#39;AG&#39;, [&#39;A2C&#39;], [], []),</span>
<span class="sd">    ...     (&#39;s1&#39;, &#39;TA&#39;, [&#39;G3A&#39;], [&#39;ins4len3&#39;], []),</span>
<span class="sd">    ...     (&#39;s2&#39;, &#39;TA&#39;, [&#39;C5A&#39;, &#39;T6C&#39;], [], []),</span>
<span class="sd">    ...     (&#39;s2&#39;, &#39;TA&#39;, [&#39;T6C&#39;], [&#39;ins5len1&#39;], []),</span>
<span class="sd">    ...     (&#39;s2&#39;, &#39;TA&#39;, [&#39;T6C&#39;], [], []),</span>
<span class="sd">    ...     (&#39;s2&#39;, &#39;TG&#39;, [&#39;T6A&#39;], [], []),</span>
<span class="sd">    ...     (&#39;s2&#39;, &#39;TG&#39;, [&#39;A2G&#39;], [], []),</span>
<span class="sd">    ...     (&#39;s2&#39;, &#39;GG&#39;, [], [], [&#39;del1to4&#39;]),</span>
<span class="sd">    ...     (&#39;s2&#39;, &#39;GG&#39;, [&#39;A1C&#39;], [], []),</span>
<span class="sd">    ...     (&#39;s2&#39;, &#39;AA&#39;, [], [], []),</span>
<span class="sd">    ...     (&#39;s2&#39;, &#39;AA&#39;, [], [], []),</span>
<span class="sd">    ...     (&#39;s2&#39;, &#39;AA&#39;, [&#39;T6C&#39;], [], []),</span>
<span class="sd">    ...     (&#39;s2&#39;, &#39;AA&#39;, [&#39;T6C&#39;], [], []),</span>
<span class="sd">    ...     (&#39;s3&#39;, &#39;AA&#39;, [&#39;T6G&#39;], [&#39;ins1len1&#39;], [&#39;del1to2&#39;]),</span>
<span class="sd">    ...     (&#39;s3&#39;, &#39;AA&#39;, [&#39;T6G&#39;], [], [&#39;del5to7&#39;])],</span>
<span class="sd">    ...     columns=[&#39;library&#39;, &#39;barcode&#39;, &#39;substitutions&#39;,</span>
<span class="sd">    ...              &#39;insertions&#39;, &#39;deletions&#39;]</span>
<span class="sd">    ...     )</span>
<span class="sd">    &gt;&gt;&gt; consensus, dropped = simpleConsensus(df, library_col=&#39;library&#39;)</span>
<span class="sd">    &gt;&gt;&gt; pandas.set_option(&#39;display.max_columns&#39;, 10)</span>
<span class="sd">    &gt;&gt;&gt; pandas.set_option(&#39;display.width&#39;, 500)</span>
<span class="sd">    &gt;&gt;&gt; consensus</span>
<span class="sd">      library barcode substitutions  insertions deletions  variant_call_support</span>
<span class="sd">    0      s1      AG         [A2C]          []        []                     2</span>
<span class="sd">    1      s1      TA         [G3A]  [ins4len3]        []                     1</span>
<span class="sd">    2      s2      GG            []          []        []                     2</span>
<span class="sd">    3      s2      TA         [T6C]          []        []                     3</span>
<span class="sd">    &gt;&gt;&gt; dropped</span>
<span class="sd">      library barcode substitutions  insertions  deletions           drop_reason</span>
<span class="sd">    0      s2      TG         [T6A]          []         []  excess substitutions</span>
<span class="sd">    1      s2      TG         [A2G]          []         []  excess substitutions</span>
<span class="sd">    2      s2      AA            []          []         []     excess minor muts</span>
<span class="sd">    3      s2      AA            []          []         []     excess minor muts</span>
<span class="sd">    4      s2      AA         [T6C]          []         []     excess minor muts</span>
<span class="sd">    5      s2      AA         [T6C]          []         []     excess minor muts</span>
<span class="sd">    6      s3      AA         [T6G]  [ins1len1]  [del1to2]         excess indels</span>
<span class="sd">    7      s3      AA         [T6G]          []  [del5to7]         excess indels</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">library_col</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">library_col</span> <span class="o">=</span> <span class="s1">&#39;library&#39;</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">library_col</span><span class="o">=</span><span class="s1">&#39;dummy&#39;</span><span class="p">)</span>
        <span class="n">drop_library_col</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">drop_library_col</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="n">mut_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">substitution_col</span><span class="p">,</span> <span class="n">insertion_col</span><span class="p">,</span> <span class="n">deletion_col</span><span class="p">]</span>
    <span class="n">all_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">library_col</span><span class="p">,</span> <span class="n">barcode_col</span><span class="p">]</span> <span class="o">+</span> <span class="n">mut_cols</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">([</span><span class="n">col</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">all_cols</span><span class="p">]):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Cannot find column </span><span class="si">{col}</span><span class="s2">&quot;</span><span class="p">)</span>


    <span class="c1"># make sure no mutations duplicated, otherwise approach below fails</span>
    <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">mut_cols</span><span class="p">:</span>
        <span class="n">duplicated</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="nb">len</span><span class="p">)</span> <span class="o">-</span> <span class="n">df</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="nb">len</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">duplicated</span><span class="o">.</span><span class="n">any</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;duplicated </span><span class="si">{col}</span><span class="s2">:</span><span class="se">\n</span><span class="s2">&quot;</span>
                             <span class="n">f</span><span class="s2">&quot;</span><span class="si">{df[col][duplicated &gt; 0]}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">dropped</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">consensus</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="p">(</span><span class="n">library</span><span class="p">,</span> <span class="n">barcode</span><span class="p">),</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">df</span><span class="p">[</span><span class="n">all_cols</span><span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span>
            <span class="p">[</span><span class="n">library_col</span><span class="p">,</span> <span class="n">barcode_col</span><span class="p">]):</span>

        <span class="n">nseqs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">g</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">nseqs</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">consensus</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span> <span class="o">+</span> <span class="p">[</span><span class="n">nseqs</span><span class="p">])</span>
            <span class="k">continue</span>

        <span class="n">consensus_failed</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="c1"># are max_sub_diffs and max_indel_diffs satisfied?</span>
        <span class="k">for</span> <span class="n">difftype</span><span class="p">,</span> <span class="n">diff_cols</span><span class="p">,</span> <span class="n">max_diffs</span> <span class="ow">in</span> <span class="p">[</span>
                <span class="p">(</span><span class="s1">&#39;substitutions&#39;</span><span class="p">,</span> <span class="p">[</span><span class="n">substitution_col</span><span class="p">],</span> <span class="n">max_sub_diffs</span><span class="p">),</span>
                <span class="p">(</span><span class="s1">&#39;indels&#39;</span><span class="p">,</span> <span class="p">[</span><span class="n">insertion_col</span><span class="p">,</span> <span class="n">deletion_col</span><span class="p">],</span> <span class="n">max_indel_diffs</span><span class="p">)]:</span>
            <span class="n">min_variant_diffs</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="k">lambda</span><span class="p">:</span> <span class="n">max_diffs</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">v1</span><span class="p">,</span> <span class="n">v2</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">combinations</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">itertuples</span><span class="p">(),</span> <span class="mi">2</span><span class="p">):</span>
                <span class="n">i1</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">v1</span><span class="p">,</span> <span class="s1">&#39;Index&#39;</span><span class="p">)</span>
                <span class="n">i2</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">v2</span><span class="p">,</span> <span class="s1">&#39;Index&#39;</span><span class="p">)</span>
                <span class="n">n_diffs</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">([</span><span class="nb">len</span><span class="p">(</span>
                        <span class="nb">set</span><span class="p">(</span><span class="nb">getattr</span><span class="p">(</span><span class="n">v1</span><span class="p">,</span> <span class="n">col</span><span class="p">))</span><span class="o">.</span><span class="n">symmetric_difference</span><span class="p">(</span>
                        <span class="nb">set</span><span class="p">(</span><span class="nb">getattr</span><span class="p">(</span><span class="n">v2</span><span class="p">,</span> <span class="n">col</span><span class="p">))))</span>
                        <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">diff_cols</span><span class="p">])</span>
                <span class="n">min_variant_diffs</span><span class="p">[</span><span class="n">i1</span><span class="p">]</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span>
                        <span class="n">min_variant_diffs</span><span class="p">[</span><span class="n">i1</span><span class="p">],</span> <span class="n">n_diffs</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">nseqs</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="ow">and</span> <span class="nb">any</span><span class="p">(</span>
                    <span class="p">[</span><span class="n">d</span> <span class="o">&gt;</span> <span class="n">max_diffs</span> <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">min_variant_diffs</span><span class="o">.</span><span class="n">values</span><span class="p">()]):</span>
                <span class="c1"># need to add to `dropped` because of max_diffs failing</span>
                <span class="n">dropped</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="n">g</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">drop_reason</span><span class="o">=</span><span class="n">f</span><span class="s2">&quot;excess </span><span class="si">{difftype}</span><span class="s2">&quot;</span><span class="p">)</span>
                        <span class="p">)</span>
                <span class="n">consensus_failed</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">break</span>
        <span class="k">if</span> <span class="n">consensus_failed</span><span class="p">:</span>
            <span class="k">continue</span>

        <span class="c1"># get consensus and see if `max_minor_muts` is satisfied</span>
        <span class="n">g_consensus</span> <span class="o">=</span> <span class="p">[</span><span class="n">library</span><span class="p">,</span> <span class="n">barcode</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">mut_cols</span><span class="p">:</span>
            <span class="n">counts</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">Counter</span><span class="p">(</span>
                    <span class="n">itertools</span><span class="o">.</span><span class="n">chain</span><span class="o">.</span><span class="n">from_iterable</span><span class="p">(</span><span class="n">g</span><span class="p">[</span><span class="n">col</span><span class="p">]))</span>
            <span class="k">if</span> <span class="nb">any</span><span class="p">([</span><span class="n">max_minor_muts</span> <span class="o">&lt;</span> <span class="n">count</span> <span class="o">&lt;</span> <span class="p">(</span><span class="n">nseqs</span> <span class="o">-</span> <span class="n">max_minor_muts</span><span class="p">)</span>
                   <span class="k">for</span> <span class="n">count</span> <span class="ow">in</span> <span class="n">counts</span><span class="o">.</span><span class="n">values</span><span class="p">()]):</span>
                <span class="n">consensus_failed</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">break</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">col_consensus</span> <span class="o">=</span> <span class="p">[</span><span class="n">mut</span> <span class="k">for</span> <span class="n">mut</span><span class="p">,</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">counts</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
                        <span class="k">if</span> <span class="n">c</span> <span class="o">&gt;</span> <span class="mf">0.5</span> <span class="o">*</span> <span class="n">nseqs</span><span class="p">]</span>
                <span class="c1"># order mutations based on first number in string</span>
                <span class="n">n_col_consensus</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">col_consensus</span><span class="p">:</span>
                    <span class="n">m</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;(\-{0,1}\d+)&#39;</span><span class="p">,</span> <span class="n">mut</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">m</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                        <span class="n">n_col_consensus</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">math</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">mut</span><span class="p">))</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">n_col_consensus</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="nb">int</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">()),</span> <span class="n">mut</span><span class="p">))</span>
                <span class="n">g_consensus</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">mut</span> <span class="k">for</span> <span class="n">n</span><span class="p">,</span> <span class="n">mut</span> <span class="ow">in</span>
                        <span class="nb">sorted</span><span class="p">(</span><span class="n">n_col_consensus</span><span class="p">)])</span>
        <span class="k">if</span> <span class="n">consensus_failed</span><span class="p">:</span>
            <span class="c1"># need to add to dropped</span>
            <span class="n">dropped</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="n">g</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">drop_reason</span><span class="o">=</span><span class="s2">&quot;excess minor muts&quot;</span><span class="p">)</span>
                    <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">consensus</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g_consensus</span> <span class="o">+</span> <span class="p">[</span><span class="n">nseqs</span><span class="p">])</span>

    <span class="n">consensus</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">consensus</span><span class="p">,</span>
            <span class="n">columns</span><span class="o">=</span><span class="n">all_cols</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;variant_call_support&#39;</span><span class="p">])</span>
    <span class="k">if</span> <span class="n">dropped</span><span class="p">:</span>
        <span class="n">dropped</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">dropped</span><span class="p">)</span><span class="o">.</span><span class="n">sort_index</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">dropped</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">drop_library_col</span><span class="p">:</span>
        <span class="n">dropped</span> <span class="o">=</span> <span class="n">dropped</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">library_col</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="s1">&#39;columns&#39;</span><span class="p">)</span>
        <span class="n">consensus</span> <span class="o">=</span> <span class="n">consensus</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">library_col</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="s1">&#39;columns&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="p">(</span><span class="n">consensus</span><span class="p">,</span> <span class="n">dropped</span><span class="p">)</span></div>


<div class="viewcode-block" id="IlluminaBarcodeParser"><a class="viewcode-back" href="../../dms_tools2.barcodes.html#dms_tools2.barcodes.IlluminaBarcodeParser">[docs]</a><span class="k">class</span> <span class="nc">IlluminaBarcodeParser</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;Parser for Illumina barcodes.</span>

<span class="sd">    The barcodes should be read by R1 and optionally R2.</span>
<span class="sd">    The arrangement of elements is shown below::</span>

<span class="sd">        5&#39;-[R2_start]-upstream-barcode-downstream-[R1_start]-3&#39;</span>

<span class="sd">    R1 anneals downstream of the barcode and reads backwards. If</span>
<span class="sd">    R2 is used, it anneals upstream of the barcode and reads forward.</span>
<span class="sd">    There can be sequences (`upstream` and `downstream`) on either</span>
<span class="sd">    side of the barcode: `downstream` must fully cover the</span>
<span class="sd">    region between where R1 starts and the barcode, and if you are</span>
<span class="sd">    using R2 then `upstream` must fully cover the region between</span>
<span class="sd">    where R2 starts and the barcode. However, it is fine if R1</span>
<span class="sd">    reads backwards past `upstream`, and if `R2` reads forward</span>
<span class="sd">    past `downstream`.</span>

<span class="sd">    Args:</span>
<span class="sd">        `bclen` (int or `None`)</span>
<span class="sd">            Length of the barcode, or `None` if length is to be</span>
<span class="sd">            determined from `valid_barcodes` argument.</span>
<span class="sd">        `upstream` (str)</span>
<span class="sd">            Sequence upstream of the barcode.</span>
<span class="sd">        `downstream` (str)</span>
<span class="sd">            Sequence downstream of barcode.</span>
<span class="sd">        `upstream_mismatch` (int)</span>
<span class="sd">            Max number of mismatches allowed in `upstream`.</span>
<span class="sd">        `downstream_mismatch` (int)</span>
<span class="sd">            Like `upstream_mismatches` but for `downstream`.</span>
<span class="sd">        `valid_barcodes` (`None` or iterable such as list, Series)</span>
<span class="sd">            If not `None`, only retain barcodes listed here.</span>
<span class="sd">            Use if you know the set of possible valid barcodes.</span>
<span class="sd">        `rc_barcode` (bool)</span>
<span class="sd">            Parse the reverse complement of the barcode (the</span>
<span class="sd">            orientation read by R1).</span>
<span class="sd">        `minq` (int)</span>
<span class="sd">            Require at least this quality score for all bases</span>
<span class="sd">            in barcode.</span>
<span class="sd">        `chastity_filter` (bool)</span>
<span class="sd">            Drop any reads that fail Illumina chastity filter.</span>
<span class="sd">        `list_all_valid_barcodes` (bool)</span>
<span class="sd">            If using `valid_barcodes`, then barcode sets returned</span>
<span class="sd">            by :class:`IlluminaBarcodeParser.parse` includes all</span>
<span class="sd">            valid barcodes even if no counts.</span>

<span class="sd">    To use, first initialize a :class:`IlluminaBarcodeParser`, then</span>
<span class="sd">    parse barcodes using :class:`IlluminaBarcodeParser.parse`.</span>
<span class="sd">    Barcodes are retained as valid only if R1 and R2 agree at every</span>
<span class="sd">    nucleotide in barcode and if at each site at least one read has</span>
<span class="sd">    a quality of at least `minq`.</span>

<span class="sd">    Here is an example. Imagine we are parsing 4 nucleotide barcodes</span>
<span class="sd">    that have the following construction::</span>

<span class="sd">        5&#39;-[R2 binding site]-ACATGA-NNNN-GACT-[R1 binding site]-3&#39;</span>

<span class="sd">    First, we initialize an appropriate :class:`IlluminaBarcodeParser`:</span>

<span class="sd">    &gt;&gt;&gt; parser = IlluminaBarcodeParser(</span>
<span class="sd">    ...              bclen=4,</span>
<span class="sd">    ...              upstream=&#39;ACATGA&#39;,</span>
<span class="sd">    ...              downstream=&#39;GACT&#39;</span>
<span class="sd">    ...              )</span>

<span class="sd">    Now we write some test FASTQ files. We write valid test</span>
<span class="sd">    reads and some invalid reads. The header for each read</span>
<span class="sd">    explains why it is valid / invalid. We use quality scores</span>
<span class="sd">    of ``?`` (30) or ``+`` (10) for high- and low-quality bases:</span>

<span class="sd">    &gt;&gt;&gt; r1file = &#39;_temp_R1.fastq&#39;</span>
<span class="sd">    &gt;&gt;&gt; r2file = &#39;_temp_R2.fastq&#39;</span>
<span class="sd">    &gt;&gt;&gt; with open(r1file, &#39;w&#39;) as f1, open(r2file, &#39;w&#39;) as f2:</span>
<span class="sd">    ...</span>
<span class="sd">    ...     # valid TACG barcode, full flanking regions</span>
<span class="sd">    ...     _ = f1.write(</span>
<span class="sd">    ...         &#39;@valid_CGTA_barcode_full_flanking_region\\n&#39;</span>
<span class="sd">    ...         &#39;AGTCCGTATCATGT\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;??????????????\\n&#39;)</span>
<span class="sd">    ...     _ = f2.write(</span>
<span class="sd">    ...         &#39;@valid_CGTA_barcode_full_flanking_region\\n&#39;</span>
<span class="sd">    ...         &#39;ACATGATACGGACT\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;??????????????\\n&#39;)</span>
<span class="sd">    ...</span>
<span class="sd">    ...     # valid CGTA barcode, partial flanking regions</span>
<span class="sd">    ...     _ = f1.write(</span>
<span class="sd">    ...         &#39;@valid_CGTA_barcode_partial_flanking_region\\n&#39;</span>
<span class="sd">    ...         &#39;AGTCCGTATCAT\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;????????????\\n&#39;)</span>
<span class="sd">    ...     _ = f2.write(</span>
<span class="sd">    ...         &#39;@valid_CGTA_barcode_partial_flanking_region\\n&#39;</span>
<span class="sd">    ...         &#39;ACATGATACG\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;??????????\\n&#39;)</span>
<span class="sd">    ...</span>
<span class="sd">    ...     # valid GCCG barcode, extended flanking regions</span>
<span class="sd">    ...     _ = f1.write(</span>
<span class="sd">    ...         &#39;@valid_GCCG_barcode_extended_flanking_region\\n&#39;</span>
<span class="sd">    ...         &#39;AGTCGCCGTCATGTTAC\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;?????????????????\\n&#39;)</span>
<span class="sd">    ...     _ = f2.write(</span>
<span class="sd">    ...         &#39;@valid_GCCG_barcode_extended_flanking_region\\n&#39;</span>
<span class="sd">    ...         &#39;ACATGACGGCGACTGAC\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;?????????????????\\n&#39;)</span>
<span class="sd">    ...</span>
<span class="sd">    ...     # AAGT barcode in R1 but R2 differs</span>
<span class="sd">    ...     _ = f1.write(</span>
<span class="sd">    ...         &#39;@AAGT_R1_barcode_but_R2_differs\\n&#39;</span>
<span class="sd">    ...         &#39;AGTCAAGTTCATGT\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;??????????????\\n&#39;)</span>
<span class="sd">    ...     _ = f2.write(</span>
<span class="sd">    ...         &#39;@AAGT_R1_barcode_but_R2_differs\\n&#39;</span>
<span class="sd">    ...         &#39;ACATGAACTAGACT\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;??????????????\\n&#39;)</span>
<span class="sd">    ...</span>
<span class="sd">    ...     # same site low quality in R1 and R2</span>
<span class="sd">    ...     _ = f1.write(</span>
<span class="sd">    ...         &#39;@low_quality_site_in_R1_and_R2\\n&#39;</span>
<span class="sd">    ...         &#39;AGTCCGTATCATGT\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;?????+????????\\n&#39;)</span>
<span class="sd">    ...     _ = f2.write(</span>
<span class="sd">    ...         &#39;@low_quality_site_in_R1_and_R2\\n&#39;</span>
<span class="sd">    ...         &#39;ACATGATACGGACT\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;????????+?????\\n&#39;)</span>
<span class="sd">    ...</span>
<span class="sd">    ...     # different site low quality in R1 and R2</span>
<span class="sd">    ...     _ = f1.write(</span>
<span class="sd">    ...         &#39;@AGTA_with_low_quality_site_in_R1\\n&#39;</span>
<span class="sd">    ...         &#39;AGTCAGTATCATGT\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;?????+????????\\n&#39;)</span>
<span class="sd">    ...     _ = f2.write(</span>
<span class="sd">    ...         &#39;@AGTA_with_low_quality_site_in_R1\\n&#39;</span>
<span class="sd">    ...         &#39;ACATGATACTGACT\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;?????????+????\\n&#39;)</span>
<span class="sd">    ...</span>
<span class="sd">    ...     # N in barcode</span>
<span class="sd">    ...     _ = f1.write(</span>
<span class="sd">    ...         &#39;@N_in_barcode\\n&#39;</span>
<span class="sd">    ...         &#39;AGTCCGTNTCATGT\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;??????????????\\n&#39;)</span>
<span class="sd">    ...     _ = f2.write(</span>
<span class="sd">    ...         &#39;@N_in_barcode\\n&#39;</span>
<span class="sd">    ...         &#39;ACATGATACGGACT\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;??????????????\\n&#39;)</span>
<span class="sd">    ...</span>
<span class="sd">    ...     # GGAG barcode, one mismatch in each flanking region</span>
<span class="sd">    ...     _ = f1.write(</span>
<span class="sd">    ...         &#39;@GGAG_barcode_one_mismatch_per_flank\\n&#39;</span>
<span class="sd">    ...         &#39;GGTCGGAGTCATGA\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;??????????????\\n&#39;)</span>
<span class="sd">    ...     _ = f2.write(</span>
<span class="sd">    ...         &#39;@GGAG_barcode_one_mismatch_per_flank\\n&#39;</span>
<span class="sd">    ...         &#39;TCATGACTCCGACG\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;??????????????\\n&#39;)</span>
<span class="sd">    ...</span>
<span class="sd">    ...     # GGAG barcode, two mismatch in a flanking region</span>
<span class="sd">    ...     _ = f1.write(</span>
<span class="sd">    ...         &#39;@GGAG_barcode_two_mismatch_in_a_flank\\n&#39;</span>
<span class="sd">    ...         &#39;GGTCGGAGTCATAA\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;??????????????\\n&#39;)</span>
<span class="sd">    ...     _ = f2.write(</span>
<span class="sd">    ...         &#39;@GGAG_barcode_two_mismatch_in_a_flank\\n&#39;</span>
<span class="sd">    ...         &#39;TCATGACTCCGACG\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;??????????????\\n&#39;)</span>


<span class="sd">    Now parse the barcodes using both the R1 and R2 files:</span>

<span class="sd">    &gt;&gt;&gt; barcodes, fates = parser.parse(r1file, r2file)</span>
<span class="sd">    &gt;&gt;&gt; print(barcodes.to_csv(sep=&#39; &#39;, index=False).strip())</span>
<span class="sd">    barcode count</span>
<span class="sd">    CGTA 2</span>
<span class="sd">    AGTA 1</span>
<span class="sd">    GCCG 1</span>
<span class="sd">    &gt;&gt;&gt; print(fates.to_csv(sep=&#39; &#39;, index=False).strip())</span>
<span class="sd">    fate count</span>
<span class="sd">    &quot;valid barcode&quot; 4</span>
<span class="sd">    &quot;unparseable barcode&quot; 3</span>
<span class="sd">    &quot;R1 / R2 disagree&quot; 1</span>
<span class="sd">    &quot;low quality barcode&quot; 1</span>

<span class="sd">    Now we parse just using R1. We gain the barcode where R1 and</span>
<span class="sd">    R2 disagree, but lose the one where R1 is low quality at a</span>
<span class="sd">    position where R2 is OK:</span>

<span class="sd">    &gt;&gt;&gt; barcodes, fates = parser.parse(r1file)</span>
<span class="sd">    &gt;&gt;&gt; print(barcodes.to_csv(sep=&#39; &#39;, index=False).strip())</span>
<span class="sd">    barcode count</span>
<span class="sd">    CGTA 2</span>
<span class="sd">    AAGT 1</span>
<span class="sd">    GCCG 1</span>
<span class="sd">    &gt;&gt;&gt; print(fates.to_csv(sep=&#39; &#39;, index=False).strip())</span>
<span class="sd">    fate count</span>
<span class="sd">    &quot;valid barcode&quot; 4</span>
<span class="sd">    &quot;unparseable barcode&quot; 3</span>
<span class="sd">    &quot;low quality barcode&quot; 2</span>

<span class="sd">    Now create a parser that allows a mismatch in each flanking</span>
<span class="sd">    region, and check that we recover a &quot;GGAG&quot; barcode:</span>

<span class="sd">    &gt;&gt;&gt; parser_mismatch = IlluminaBarcodeParser(</span>
<span class="sd">    ...              bclen=4,</span>
<span class="sd">    ...              upstream=&#39;ACATGA&#39;,</span>
<span class="sd">    ...              downstream=&#39;GACT&#39;,</span>
<span class="sd">    ...              upstream_mismatch=1,</span>
<span class="sd">    ...              downstream_mismatch=1,</span>
<span class="sd">    ...              )</span>
<span class="sd">    &gt;&gt;&gt; barcodes_mismatch, fates_mismatch = parser_mismatch.parse(r1file, r2file)</span>
<span class="sd">    &gt;&gt;&gt; print(barcodes_mismatch.to_csv(sep=&#39; &#39;, index=False).strip())</span>
<span class="sd">    barcode count</span>
<span class="sd">    CGTA 2</span>
<span class="sd">    AGTA 1</span>
<span class="sd">    GCCG 1</span>
<span class="sd">    GGAG 1</span>
<span class="sd">    &gt;&gt;&gt; print(fates_mismatch.to_csv(sep=&#39; &#39;, index=False).strip())</span>
<span class="sd">    fate count</span>
<span class="sd">    &quot;valid barcode&quot; 5</span>
<span class="sd">    &quot;unparseable barcode&quot; 2</span>
<span class="sd">    &quot;R1 / R2 disagree&quot; 1</span>
<span class="sd">    &quot;low quality barcode&quot; 1</span>

<span class="sd">    Now parse the barcodes using `valid_barcodes` to set a</span>
<span class="sd">    barcode whitelist:</span>

<span class="sd">    &gt;&gt;&gt; parser_wl = IlluminaBarcodeParser(</span>
<span class="sd">    ...              upstream=&#39;ACATGA&#39;,</span>
<span class="sd">    ...              downstream=&#39;GACT&#39;,</span>
<span class="sd">    ...              valid_barcodes={&#39;CGTA&#39;, &#39;AGTA&#39;, &#39;TAAT&#39;}</span>
<span class="sd">    ...              )</span>
<span class="sd">    &gt;&gt;&gt; barcodes_wl, fates_wl = parser_wl.parse(r1file, r2file)</span>
<span class="sd">    &gt;&gt;&gt; print(barcodes_wl.to_csv(sep=&#39; &#39;, index=False).strip())</span>
<span class="sd">    barcode count</span>
<span class="sd">    CGTA 2</span>
<span class="sd">    AGTA 1</span>
<span class="sd">    TAAT 0</span>
<span class="sd">    &gt;&gt;&gt; print(fates_wl.to_csv(sep=&#39; &#39;, index=False).strip())</span>
<span class="sd">    fate count</span>
<span class="sd">    &quot;unparseable barcode&quot; 3</span>
<span class="sd">    &quot;valid barcode&quot; 3</span>
<span class="sd">    &quot;R1 / R2 disagree&quot; 1</span>
<span class="sd">    &quot;invalid barcode&quot; 1</span>
<span class="sd">    &quot;low quality barcode&quot; 1</span>

<span class="sd">    Remove the test FASTQ files:</span>

<span class="sd">    &gt;&gt;&gt; os.remove(r1file)</span>
<span class="sd">    &gt;&gt;&gt; os.remove(r2file)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1">#: valid nucleotide characters</span>
    <span class="n">VALID_NTS</span> <span class="o">=</span> <span class="s1">&#39;ACGTN&#39;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">bclen</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">upstream</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">downstream</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span>
            <span class="n">upstream_mismatch</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">downstream_mismatch</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
            <span class="n">valid_barcodes</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">rc_barcode</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">minq</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>
            <span class="n">chastity_filter</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">list_all_valid_barcodes</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;See main class doc string.&quot;&quot;&quot;</span>

        <span class="c1"># first make all arguments into attributes</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">bclen</span> <span class="o">=</span> <span class="n">bclen</span>
        <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;^[</span><span class="si">{self.VALID_NTS}</span><span class="s2">]*$&quot;</span><span class="p">,</span> <span class="n">upstream</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">upstream</span> <span class="o">=</span> <span class="n">upstream</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid chars in upstream </span><span class="si">{upstream}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;^[</span><span class="si">{self.VALID_NTS}</span><span class="s2">]*$&quot;</span><span class="p">,</span> <span class="n">downstream</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">downstream</span> <span class="o">=</span> <span class="n">downstream</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid chars in downstream </span><span class="si">{downstream}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">upstream_mismatch</span> <span class="o">=</span> <span class="n">upstream_mismatch</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">downstream_mismatch</span> <span class="o">=</span> <span class="n">downstream_mismatch</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">valid_barcodes</span> <span class="o">=</span> <span class="n">valid_barcodes</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_barcodes</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">valid_barcodes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">valid_barcodes</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">valid_barcodes</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;empty list for `valid_barcodes`&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">bclen</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">bclen</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">valid_barcodes</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
            <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">bc</span><span class="p">)</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">bclen</span> <span class="k">for</span> <span class="n">bc</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_barcodes</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;`valid_barcodes` not all valid length&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">bclen</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;must specify `bclen` or `valid_barcodes`&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">minq</span> <span class="o">=</span> <span class="n">minq</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">rc_barcode</span> <span class="o">=</span> <span class="n">rc_barcode</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">chastity_filter</span> <span class="o">=</span> <span class="n">chastity_filter</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">list_all_valid_barcodes</span> <span class="o">=</span> <span class="n">list_all_valid_barcodes</span>

        <span class="c1"># specify information about R1 / R2 matches</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_bcend</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s1">&#39;R1&#39;</span><span class="p">:</span><span class="bp">self</span><span class="o">.</span><span class="n">bclen</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">downstream</span><span class="p">),</span>
                <span class="s1">&#39;R2&#39;</span><span class="p">:</span><span class="bp">self</span><span class="o">.</span><span class="n">bclen</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">upstream</span><span class="p">)</span>
                <span class="p">}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_rcdownstream</span> <span class="o">=</span> <span class="n">dms_tools2</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">reverseComplement</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">downstream</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_rcupstream</span> <span class="o">=</span> <span class="n">dms_tools2</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">reverseComplement</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">upstream</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_matches</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;R1&#39;</span><span class="p">:{},</span> <span class="s1">&#39;R2&#39;</span><span class="p">:{}}</span> <span class="c1"># saves match object by read length</span>


<div class="viewcode-block" id="IlluminaBarcodeParser.parse"><a class="viewcode-back" href="../../dms_tools2.barcodes.html#dms_tools2.barcodes.IlluminaBarcodeParser.parse">[docs]</a>    <span class="k">def</span> <span class="nf">parse</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">r1files</span><span class="p">,</span> <span class="n">r2files</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Parses barcodes from files.</span>

<span class="sd">        Args:</span>
<span class="sd">            `r1files` (str or list)</span>
<span class="sd">                Name of R1 FASTQ file, or list of such files</span>
<span class="sd">                Can optionally be gzipped.</span>
<span class="sd">            `r2files` (`None`, str, or list)</span>
<span class="sd">                `None` or empty list if not using R2, otherwise like R1.</span>

<span class="sd">        Returns:</span>
<span class="sd">            The 2-tuple `(barcodes, fates)`. In this 2-tuple:</span>

<span class="sd">                - `barcodes` is a pandas DataFrame giving the</span>
<span class="sd">                  number of observations of each barcode. The</span>
<span class="sd">                  columns are named &quot;barcode&quot; and &quot;count&quot;.</span>

<span class="sd">                - `fates` is a pandas DataFrame giving the</span>
<span class="sd">                  total number of reads with each fate. The</span>
<span class="sd">                  columns are named &quot;fate&quot; and &quot;count&quot;.</span>

<span class="sd">                  - &quot;valid barcode&quot;</span>

<span class="sd">                  - &quot;invalid barcode&quot;: not in our barcode whitelist</span>

<span class="sd">                  - &quot;R1 / R2 disagree&quot;</span>

<span class="sd">                  - &quot;low quality barcode&quot;: sequencing quality low</span>

<span class="sd">                  - &quot;unparseable barcode&quot;: invalid flanking sequences</span>
<span class="sd">                    or N in barcode.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">r2files</span><span class="p">:</span>
            <span class="n">reads</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;R1&#39;</span><span class="p">]</span>
            <span class="n">r2files</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">reads</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;R1&#39;</span><span class="p">,</span> <span class="s1">&#39;R2&#39;</span><span class="p">]</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_barcodes</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">list_all_valid_barcodes</span><span class="p">:</span>
            <span class="n">barcodes</span> <span class="o">=</span> <span class="p">{</span><span class="n">bc</span><span class="p">:</span><span class="mi">0</span> <span class="k">for</span> <span class="n">bc</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_barcodes</span><span class="p">}</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">barcodes</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>

        <span class="n">fates</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">r1</span><span class="p">,</span> <span class="n">r2</span><span class="p">,</span> <span class="n">q1</span><span class="p">,</span> <span class="n">q2</span><span class="p">,</span> <span class="n">fail</span> <span class="ow">in</span> \
                <span class="n">dms_tools2</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">iteratePairedFASTQ</span><span class="p">(</span><span class="n">r1files</span><span class="p">,</span> <span class="n">r2files</span><span class="p">):</span>

            <span class="k">if</span> <span class="n">fail</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">chastity_filter</span><span class="p">:</span>
                <span class="n">fates</span><span class="p">[</span><span class="s1">&#39;failed chastity filter&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">continue</span>

            <span class="n">matches</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">read</span><span class="p">,</span> <span class="n">r</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="p">[</span><span class="n">r1</span><span class="p">,</span> <span class="n">r2</span><span class="p">]):</span>
                <span class="n">rlen</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>

                <span class="c1"># get or build matcher for read of this length</span>
                <span class="n">len_past_bc</span> <span class="o">=</span> <span class="n">rlen</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">_bcend</span><span class="p">[</span><span class="n">read</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">len_past_bc</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;</span><span class="si">{read}</span><span class="s2"> too short: </span><span class="si">{rlen}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">rlen</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_matches</span><span class="p">[</span><span class="n">read</span><span class="p">]:</span>
                    <span class="n">matcher</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_matches</span><span class="p">[</span><span class="n">read</span><span class="p">][</span><span class="n">rlen</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">read</span> <span class="o">==</span> <span class="s1">&#39;R1&#39;</span><span class="p">:</span>
                        <span class="n">match_str</span> <span class="o">=</span> <span class="p">(</span>
                                <span class="n">f</span><span class="s1">&#39;^(</span><span class="si">{self._rcdownstream}</span><span class="s1">)&#39;</span>
                                <span class="n">f</span><span class="s1">&#39;{{s&lt;=</span><span class="si">{self.downstream_mismatch}</span><span class="s1">}}&#39;</span> <span class="o">+</span>
                                <span class="n">f</span><span class="s1">&#39;(?P&lt;bc&gt;N{{</span><span class="si">{self.bclen}</span><span class="s1">}})&#39;</span> <span class="o">+</span>
                                <span class="n">f</span><span class="s1">&#39;(</span><span class="si">{self._rcupstream[ : len_past_bc]}</span><span class="s1">)&#39;</span> <span class="o">+</span>
                                <span class="n">f</span><span class="s1">&#39;{{s&lt;=</span><span class="si">{self.upstream_mismatch}</span><span class="s1">}}&#39;</span>
                                <span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="k">assert</span> <span class="n">read</span> <span class="o">==</span> <span class="s1">&#39;R2&#39;</span>
                        <span class="n">match_str</span> <span class="o">=</span> <span class="p">(</span>
                                <span class="n">f</span><span class="s1">&#39;^(</span><span class="si">{self.upstream}</span><span class="s1">)&#39;</span> <span class="o">+</span>
                                <span class="n">f</span><span class="s1">&#39;{{s&lt;=</span><span class="si">{self.upstream_mismatch}</span><span class="s1">}}&#39;</span> <span class="o">+</span>
                                <span class="n">f</span><span class="s1">&#39;(?P&lt;bc&gt;N{{</span><span class="si">{self.bclen}</span><span class="s1">}})&#39;</span> <span class="o">+</span>
                                <span class="n">f</span><span class="s1">&#39;(</span><span class="si">{self.downstream[ : len_past_bc]}</span><span class="s1">)&#39;</span> <span class="o">+</span>
                                <span class="n">f</span><span class="s1">&#39;{{s&lt;=</span><span class="si">{self.downstream_mismatch}</span><span class="s1">}}&#39;</span>
                                <span class="p">)</span>
                    <span class="n">matcher</span> <span class="o">=</span> <span class="n">regex</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span>
                            <span class="n">dms_tools2</span><span class="o">.</span><span class="n">pacbio</span><span class="o">.</span><span class="n">re_expandIUPAC</span><span class="p">(</span><span class="n">match_str</span><span class="p">),</span>
                            <span class="n">flags</span><span class="o">=</span><span class="n">regex</span><span class="o">.</span><span class="n">BESTMATCH</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_matches</span><span class="p">[</span><span class="n">read</span><span class="p">][</span><span class="n">rlen</span><span class="p">]</span> <span class="o">=</span> <span class="n">matcher</span>

                <span class="n">m</span> <span class="o">=</span> <span class="n">matcher</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">m</span><span class="p">:</span>
                    <span class="n">matches</span><span class="p">[</span><span class="n">read</span><span class="p">]</span> <span class="o">=</span> <span class="n">m</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">break</span>

            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">matches</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">reads</span><span class="p">):</span>
                <span class="n">bc</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="n">bc_q</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="k">for</span> <span class="n">read</span><span class="p">,</span> <span class="n">q</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="p">[</span><span class="n">q1</span><span class="p">,</span> <span class="n">q2</span><span class="p">]):</span>
                    <span class="n">bc</span><span class="p">[</span><span class="n">read</span><span class="p">]</span> <span class="o">=</span> <span class="n">matches</span><span class="p">[</span><span class="n">read</span><span class="p">]</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;bc&#39;</span><span class="p">)</span>
                    <span class="n">bc_q</span><span class="p">[</span><span class="n">read</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span>
                                 <span class="nb">ord</span><span class="p">(</span><span class="n">qi</span><span class="p">)</span> <span class="o">-</span> <span class="mi">33</span> <span class="k">for</span> <span class="n">qi</span> <span class="ow">in</span>
                                 <span class="n">q</span><span class="p">[</span><span class="n">matches</span><span class="p">[</span><span class="n">read</span><span class="p">]</span><span class="o">.</span><span class="n">start</span><span class="p">(</span><span class="s1">&#39;bc&#39;</span><span class="p">)</span> <span class="p">:</span>
                                   <span class="n">matches</span><span class="p">[</span><span class="n">read</span><span class="p">]</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;bc&#39;</span><span class="p">)]],</span>
                                 <span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;int&#39;</span><span class="p">)</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">rc_barcode</span> <span class="ow">and</span> <span class="s1">&#39;R2&#39;</span> <span class="ow">in</span> <span class="n">reads</span><span class="p">:</span>
                    <span class="n">bc</span><span class="p">[</span><span class="s1">&#39;R2&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dms_tools2</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">reverseComplement</span><span class="p">(</span><span class="n">bc</span><span class="p">[</span><span class="s1">&#39;R2&#39;</span><span class="p">])</span>
                    <span class="n">bc_q</span><span class="p">[</span><span class="s1">&#39;R2&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">flip</span><span class="p">(</span><span class="n">bc_q</span><span class="p">[</span><span class="s1">&#39;R2&#39;</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
                <span class="k">elif</span> <span class="s1">&#39;R2&#39;</span> <span class="ow">in</span> <span class="n">reads</span><span class="p">:</span>
                    <span class="n">bc</span><span class="p">[</span><span class="s1">&#39;R1&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dms_tools2</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">reverseComplement</span><span class="p">(</span><span class="n">bc</span><span class="p">[</span><span class="s1">&#39;R1&#39;</span><span class="p">])</span>
                    <span class="n">bc_q</span><span class="p">[</span><span class="s1">&#39;R1&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">flip</span><span class="p">(</span><span class="n">bc_q</span><span class="p">[</span><span class="s1">&#39;R1&#39;</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">reads</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="k">if</span> <span class="p">(</span><span class="n">bc_q</span><span class="p">[</span><span class="s1">&#39;R1&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">minq</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
                        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_barcodes</span> <span class="ow">and</span> <span class="p">(</span>
                                <span class="n">bc</span><span class="p">[</span><span class="s1">&#39;R1&#39;</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_barcodes</span><span class="p">):</span>
                            <span class="n">fates</span><span class="p">[</span><span class="s1">&#39;invalid barcode&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">barcodes</span><span class="p">[</span><span class="n">bc</span><span class="p">[</span><span class="s1">&#39;R1&#39;</span><span class="p">]]</span> <span class="o">+=</span> <span class="mi">1</span>
                            <span class="n">fates</span><span class="p">[</span><span class="s1">&#39;valid barcode&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">fates</span><span class="p">[</span><span class="s1">&#39;low quality barcode&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">bc</span><span class="p">[</span><span class="s1">&#39;R1&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">bc</span><span class="p">[</span><span class="s1">&#39;R2&#39;</span><span class="p">]:</span>
                        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_barcodes</span> <span class="ow">and</span> <span class="p">(</span>
                                <span class="n">bc</span><span class="p">[</span><span class="s1">&#39;R1&#39;</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_barcodes</span><span class="p">):</span>
                            <span class="n">fates</span><span class="p">[</span><span class="s1">&#39;invalid barcode&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="k">elif</span> <span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">maximum</span><span class="p">(</span><span class="n">bc_q</span><span class="p">[</span><span class="s1">&#39;R1&#39;</span><span class="p">],</span> <span class="n">bc_q</span><span class="p">[</span><span class="s1">&#39;R2&#39;</span><span class="p">])</span>
                                <span class="o">&gt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">minq</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
                            <span class="n">barcodes</span><span class="p">[</span><span class="n">bc</span><span class="p">[</span><span class="s1">&#39;R1&#39;</span><span class="p">]]</span> <span class="o">+=</span> <span class="mi">1</span>
                            <span class="n">fates</span><span class="p">[</span><span class="s1">&#39;valid barcode&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">fates</span><span class="p">[</span><span class="s1">&#39;low quality barcode&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">fates</span><span class="p">[</span><span class="s1">&#39;R1 / R2 disagree&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># invalid flanking sequence or N in barcode</span>
                <span class="n">fates</span><span class="p">[</span><span class="s1">&#39;unparseable barcode&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="n">barcodes</span> <span class="o">=</span> <span class="p">(</span><span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
                        <span class="nb">list</span><span class="p">(</span><span class="n">barcodes</span><span class="o">.</span><span class="n">items</span><span class="p">()),</span>
                        <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">])</span>
                    <span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s1">&#39;count&#39;</span><span class="p">,</span> <span class="s1">&#39;barcode&#39;</span><span class="p">],</span>
                                 <span class="n">ascending</span><span class="o">=</span><span class="p">[</span><span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">])</span>
                    <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                    <span class="p">)</span>

        <span class="n">fates</span> <span class="o">=</span> <span class="p">(</span><span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
                    <span class="nb">list</span><span class="p">(</span><span class="n">fates</span><span class="o">.</span><span class="n">items</span><span class="p">()),</span>
                    <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;fate&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">])</span>
                 <span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s1">&#39;count&#39;</span><span class="p">,</span> <span class="s1">&#39;fate&#39;</span><span class="p">],</span>
                              <span class="n">ascending</span><span class="o">=</span><span class="p">[</span><span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">])</span>
                 <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                 <span class="p">)</span>

        <span class="k">return</span> <span class="p">(</span><span class="n">barcodes</span><span class="p">,</span> <span class="n">fates</span><span class="p">)</span></div></div>


<div class="viewcode-block" id="tidy_split"><a class="viewcode-back" href="../../dms_tools2.barcodes.html#dms_tools2.barcodes.tidy_split">[docs]</a><span class="k">def</span> <span class="nf">tidy_split</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">column</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="n">keep</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Split values of a column and expand so new DataFrame has one split</span>
<span class="sd">    value per row. Filters rows where the column is missing.</span>

<span class="sd">    Taken from https://stackoverflow.com/a/39946744</span>

<span class="sd">    Args:</span>
<span class="sd">        df : pandas DataFrame</span>
<span class="sd">            dataframe with the column to split and expand</span>
<span class="sd">        column : str</span>
<span class="sd">            the column to split and expand</span>
<span class="sd">        sep : str</span>
<span class="sd">            the string used to split the column&#39;s values</span>
<span class="sd">        keep : bool</span>
<span class="sd">            whether to retain the presplit value as it&#39;s own row</span>

<span class="sd">    Returns:</span>
<span class="sd">        pandas DataFrame</span>
<span class="sd">            Returns a dataframe with the same columns as `df`.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">indexes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
    <span class="n">new_values</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="n">column</span><span class="p">])</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">presplit</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">column</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)):</span>
        <span class="n">values</span> <span class="o">=</span> <span class="n">presplit</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">sep</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">keep</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">values</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">indexes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
            <span class="n">new_values</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">presplit</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">values</span><span class="p">:</span>
            <span class="n">indexes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
            <span class="n">new_values</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
    <span class="n">new_df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">indexes</span><span class="p">,</span> <span class="p">:]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">new_df</span><span class="p">[</span><span class="n">column</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_values</span>
    <span class="k">return</span> <span class="n">new_df</span></div>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">doctest</span>
    <span class="n">doctest</span><span class="o">.</span><span class="n">testmod</span><span class="p">()</span>
</pre></div>

          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="../../index.html">
    <img class="logo" src="../../_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">dms_tools2</h1>
    
  </a>
</p>



<p class="blurb">Tools for analyzing deep mutational scanning data.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=dms_tools2&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a class="badge" href="https://travis-ci.org/jbloomlab/dms_tools2">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../examples.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../bcsubamp.html">Barcoded-subamplicon sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../prefs.html">Amino-acid preferences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../diffsel.html">Differential selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../fracsurvive.html">Fraction surviving</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../programs.html">Executable programs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api.html">Python API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../citations.html">Citations</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../index.html">Documentation overview</a><ul>
  <li><a href="../index.html">Module code</a><ul>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017--2019, the Bloom lab.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
    </div>

    
    <a href="https://github.com/jbloomlab/dms_tools2" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>