
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>dms_tools2.barcodes &#8212; dms_tools2 2.4.0 documentation</title>
    <link rel="stylesheet" href="../../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
   
  <link rel="stylesheet" href="../../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <h1>Source code for dms_tools2.barcodes</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">===============</span>
<span class="sd">barcodes</span>
<span class="sd">===============</span>

<span class="sd">Operations for sequence barcodes and UMIs.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">import</span> <span class="nn">tempfile</span>

<span class="kn">import</span> <span class="nn">numpy</span>
<span class="kn">import</span> <span class="nn">pandas</span>
<span class="kn">import</span> <span class="nn">regex</span>
<span class="kn">import</span> <span class="nn">umi_tools.network</span>
<span class="kn">import</span> <span class="nn">Bio.SeqUtils.ProtParamData</span>

<span class="c1"># use plotnine for plotting</span>
<span class="kn">from</span> <span class="nn">plotnine</span> <span class="k">import</span> <span class="o">*</span>

<span class="kn">import</span> <span class="nn">dms_tools2.plot</span>
<span class="kn">from</span> <span class="nn">dms_tools2.plot</span> <span class="k">import</span> <span class="n">COLOR_BLIND_PALETTE_GRAY</span>
<span class="kn">import</span> <span class="nn">dms_tools2.utils</span>
<span class="kn">import</span> <span class="nn">dms_tools2.pacbio</span>
<span class="kn">from</span> <span class="nn">dms_tools2</span> <span class="k">import</span> <span class="n">CODON_TO_AA</span><span class="p">,</span> <span class="n">CODONS</span><span class="p">,</span> <span class="n">AAS_WITHSTOP</span><span class="p">,</span> <span class="n">AA_TO_CODONS</span>

<span class="n">_umi_clusterer</span> <span class="o">=</span> <span class="n">umi_tools</span><span class="o">.</span><span class="n">network</span><span class="o">.</span><span class="n">UMIClusterer</span><span class="p">()</span>


<div class="viewcode-block" id="almost_duplicated"><a class="viewcode-back" href="../../dms_tools2.barcodes.html#dms_tools2.barcodes.almost_duplicated">[docs]</a><span class="k">def</span> <span class="nf">almost_duplicated</span><span class="p">(</span><span class="n">barcodes</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Identifies nearly identical barcodes.</span>

<span class="sd">    This function mimics the pandas `duplicated`</span>
<span class="sd">    function except it can also mark as duplicated almost</span>
<span class="sd">    but not exactly identical sequences.</span>

<span class="sd">    Args:</span>
<span class="sd">        `barcodes` (list or pandas Series)</span>
<span class="sd">            Lists all the barcodes, which should be strings</span>
<span class="sd">            of same length.</span>
<span class="sd">        `threshold` (int)</span>
<span class="sd">            Max number of mismatches for barcodes to be</span>
<span class="sd">            considered almost identical.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A pandas Series of the same length as `barcodes`</span>
<span class="sd">        filled with bools that indicates whether a barcode</span>
<span class="sd">        is a near duplicate. For each group of barcodes within</span>
<span class="sd">        `threshold` of each other, only one barcode will have</span>
<span class="sd">        an entry of `False` (not duplicated) and the rest will</span>
<span class="sd">        be `True` (duplicated). The barcode listed as not duplicated</span>
<span class="sd">        is chosen as follows: (1) it has the most abundant barcode</span>
<span class="sd">        in the group, (2) among barcodes that are equivalent abundant</span>
<span class="sd">        we take the one listed first in `barcodes`.</span>
<span class="sd">        Groups are computed using the directional method of</span>
<span class="sd">        `umi_tools &lt;https://github.com/CGATOxford/UMI-tools&gt;`_.</span>

<span class="sd">    When `threshold` is zero, just like `pandas.duplicated`:</span>

<span class="sd">    &gt;&gt;&gt; barcodes = [&#39;CTC&#39;, &#39;ATG&#39;, &#39;ATG&#39;, &#39;ATA&#39;]</span>
<span class="sd">    &gt;&gt;&gt; almost_duplicated(barcodes, threshold=0).equals(</span>
<span class="sd">    ...     pandas.Series([False, False, True, False]))</span>
<span class="sd">    True</span>
<span class="sd">    &gt;&gt;&gt; pandas.Series(barcodes).duplicated().equals(</span>
<span class="sd">    ...     pandas.Series([False, False, True, False]))</span>
<span class="sd">    True</span>

<span class="sd">    But when `threshold` is 1, also identifies **almost** equal</span>
<span class="sd">    barcodes (in this case, ones that differ by &lt;= 1 mutation:</span>

<span class="sd">    &gt;&gt;&gt; almost_duplicated(barcodes, threshold=1).equals(</span>
<span class="sd">    ...     pandas.Series([False, False, True, True]))</span>
<span class="sd">    True</span>

<span class="sd">    All barcodes are within a distance of two, so all are</span>
<span class="sd">    near duplicates. Note how we mark as not duplicated</span>
<span class="sd">    (`False`) the first one listed among the most abundant</span>
<span class="sd">    barcode (&#39;ATG&#39;):</span>

<span class="sd">    &gt;&gt;&gt; almost_duplicated(pandas.Series(barcodes), threshold=2).equals(</span>
<span class="sd">    ...     pandas.Series([True, False, True, True]))</span>
<span class="sd">    True</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">threshold</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;`threshold` must be &gt;= 0&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">barcodes</span><span class="p">,</span> <span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">):</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">barcodes</span><span class="p">,</span> <span class="n">collections</span><span class="o">.</span><span class="n">Iterable</span><span class="p">):</span>
            <span class="n">barcodes</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">barcodes</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;`barcodes` invalid type {type(barcodes)}&quot;</span><span class="p">)</span>
    <span class="n">barcodes</span> <span class="o">=</span> <span class="n">barcodes</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">encode</span><span class="p">(</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span>

    <span class="n">counts</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">Counter</span><span class="p">(</span><span class="n">barcodes</span><span class="p">)</span>

    <span class="n">groups</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">_umi_clusterer</span><span class="p">(</span><span class="n">barcodes</span><span class="p">,</span> <span class="n">counts</span><span class="p">,</span> <span class="n">threshold</span><span class="p">):</span>
        <span class="c1"># keep only barcode(s) most abundant in group</span>
        <span class="n">max_count</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">counts</span><span class="p">[</span><span class="n">barcode</span><span class="p">]</span> <span class="k">for</span> <span class="n">barcode</span> <span class="ow">in</span> <span class="n">group</span><span class="p">)</span>
        <span class="n">groups</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">frozenset</span><span class="p">(</span><span class="n">barcode</span> <span class="k">for</span> <span class="n">barcode</span> <span class="ow">in</span> <span class="n">group</span>
                      <span class="k">if</span> <span class="n">counts</span><span class="p">[</span><span class="n">barcode</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">max_count</span><span class="p">))</span>

    <span class="n">dups</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">barcode</span> <span class="ow">in</span> <span class="n">barcodes</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">groups</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">barcode</span> <span class="ow">in</span> <span class="n">g</span><span class="p">:</span>
                <span class="n">dups</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">groups</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">g</span><span class="p">)</span>
                <span class="k">break</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">dups</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">assert</span> <span class="ow">not</span> <span class="n">groups</span>

    <span class="k">return</span> <span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">dups</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">barcodes</span><span class="o">.</span><span class="n">index</span><span class="p">)</span></div>


<div class="viewcode-block" id="fracIdentWithinBarcode"><a class="viewcode-back" href="../../dms_tools2.barcodes.html#dms_tools2.barcodes.fracIdentWithinBarcode">[docs]</a><span class="k">def</span> <span class="nf">fracIdentWithinBarcode</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">barcode_col</span><span class="o">=</span><span class="s1">&#39;barcode&#39;</span><span class="p">,</span>
        <span class="n">variant_col</span><span class="o">=</span><span class="s1">&#39;variant&#39;</span><span class="p">,</span> <span class="n">library_col</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Gets fraction of identical variants within barcodes.</span>

<span class="sd">    This function is designed for the case when you have barcoded</span>
<span class="sd">    variants, and you want to determine how often the variants</span>
<span class="sd">    with the same barcode are identical. To calculate this, all</span>
<span class="sd">    variants with the same barcode are grouped, and for every pair</span>
<span class="sd">    of variants within a barcode we compute the fraction that</span>
<span class="sd">    are identical (note that this fraction is the</span>
<span class="sd">    `Simpson diversity index without replacement &lt;https://en.wikipedia.org/wiki/Diversity_index&gt;`_.</span>
<span class="sd">    We then compute the average of this fraction over all pairs.</span>

<span class="sd">    In the case where we are sequencing barcoded genes, the square</span>
<span class="sd">    root of this fraction can be interpreted as the sequencing accuracy</span>
<span class="sd">    per variant.</span>

<span class="sd">    Args:</span>
<span class="sd">        `df` (pandas Data Frame)</span>
<span class="sd">            Holds barcodes, variants, and (optionally) library names.</span>
<span class="sd">        `barcode_col` (str)</span>
<span class="sd">            Name of column holding barcodes.</span>
<span class="sd">        `variant_col` (str)</span>
<span class="sd">            Name of column holding variants (these are what we</span>
<span class="sd">            compare to see if they are identical).</span>
<span class="sd">        `library_col` (str or None)</span>
<span class="sd">            Name of libary. We do the computation separately</span>
<span class="sd">            for each library. Set to `None` if only one library.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A pandas data frame with columns named `fraction_identical`</span>
<span class="sd">        and `accuracy` that contain the computed fraction identical</span>
<span class="sd">        and accuracy for each library.</span>

<span class="sd">    Here is an example. In this data frame, we see that</span>
<span class="sd">    of the four pairs (one for barcode *AT* and three for</span>
<span class="sd">    barcode *TA*), two are identical. As you can see</span>
<span class="sd">    the function correctly calculates that 50% are identical,</span>
<span class="sd">    and computes an accuracy that is the square root of 0.5:</span>

<span class="sd">    &gt;&gt;&gt; df = pandas.DataFrame({</span>
<span class="sd">    ...     &#39;barcode&#39;:[&#39;AT&#39;, &#39;AT&#39;, &#39;TG&#39;, &#39;TA&#39;, &#39;TA&#39;, &#39;TA&#39;],</span>
<span class="sd">    ...     &#39;variant&#39;:[&#39;v1&#39;, &#39;v1&#39;, &#39;v2&#39;, &#39;v3&#39;, &#39;v4&#39;, &#39;v3&#39;]})</span>
<span class="sd">    &gt;&gt;&gt; fracIdentWithinBarcode(df)</span>
<span class="sd">       fraction_identical  accuracy</span>
<span class="sd">    0                 0.5  0.707107</span>

<span class="sd">    Now another example with two libraries and non-standard column</span>
<span class="sd">    names. Note that we only get results for the two libraries</span>
<span class="sd">    with barcodes found multiple times:</span>

<span class="sd">    &gt;&gt;&gt; df = pandas.DataFrame({</span>
<span class="sd">    ...     &#39;bc&#39;    :[&#39;AT&#39;, &#39;AT&#39;, &#39;TG&#39;, &#39;TA&#39;, &#39;TA&#39;, &#39;TA&#39;, &#39;TA&#39;],</span>
<span class="sd">    ...     &#39;var&#39;   :[&#39;v1&#39;, &#39;v1&#39;, &#39;v2&#39;, &#39;v3&#39;, &#39;v4&#39;, &#39;v3&#39;, &#39;v3&#39;],</span>
<span class="sd">    ...     &#39;library&#39;:[&#39;s1&#39;, &#39;s1&#39;, &#39;s2&#39;, &#39;s3&#39;, &#39;s3&#39;, &#39;s3&#39;, &#39;s4&#39;]})</span>
<span class="sd">    &gt;&gt;&gt; fracIdentWithinBarcode(df, library_col=&#39;library&#39;,</span>
<span class="sd">    ...     barcode_col=&#39;bc&#39;, variant_col=&#39;var&#39;)</span>
<span class="sd">      library  fraction_identical  accuracy</span>
<span class="sd">    0      s1            1.000000   1.00000</span>
<span class="sd">    1      s3            0.333333   0.57735</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">library_col</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">library_col</span> <span class="o">=</span> <span class="s1">&#39;library&#39;</span>
        <span class="n">drop_library_col</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">library</span><span class="o">=</span><span class="s1">&#39;dummy&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">drop_library_col</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="p">[</span><span class="n">barcode_col</span><span class="p">,</span> <span class="n">variant_col</span><span class="p">,</span> <span class="n">library_col</span><span class="p">]:</span>
        <span class="k">if</span> <span class="n">col</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;No columns </span><span class="si">{col}</span><span class="s2"> in df&quot;</span><span class="p">)</span>

    <span class="n">result</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">df</span>
        <span class="c1"># get just sequences that have a barcode found multiple times</span>
        <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">barcode_counts</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">barcode_counts</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
            <span class="n">x</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="n">library_col</span><span class="p">,</span> <span class="n">barcode_col</span><span class="p">])</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="s1">&#39;count&#39;</span><span class="p">))</span>
        <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;barcode_counts &gt; 1&#39;</span><span class="p">)</span>
        <span class="c1"># within each barcode, count number of sequences of each mutation combo</span>
        <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">dummy</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="n">library_col</span><span class="p">,</span> <span class="n">barcode_col</span><span class="p">,</span> <span class="s1">&#39;barcode_counts&#39;</span><span class="p">,</span> <span class="n">variant_col</span><span class="p">])</span>
        <span class="o">.</span><span class="n">dummy</span>
        <span class="o">.</span><span class="n">count</span><span class="p">()</span>
        <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;sequence_counts&#39;</span><span class="p">)</span>
        <span class="c1"># compute Simpson diversity without replacement for each barcode</span>
        <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="n">library_col</span><span class="p">,</span> <span class="n">barcode_col</span><span class="p">,</span> <span class="s1">&#39;barcode_counts&#39;</span><span class="p">])</span>
        <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">sequence_counts</span> <span class="o">*</span> <span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">sequence_counts</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">/</span>
                      <span class="n">x</span><span class="o">.</span><span class="n">barcode_counts</span> <span class="o">/</span> <span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">barcode_counts</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">sum</span><span class="p">())</span>
        <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;simpson_diversity&#39;</span><span class="p">)</span>
        <span class="c1"># compute weighted average of fraction identical across all pairs</span>
        <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">npairs</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">barcode_counts</span> <span class="o">*</span> <span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">barcode_counts</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span>
            <span class="n">weighted_diversity</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">npairs</span> <span class="o">*</span> <span class="n">x</span><span class="o">.</span><span class="n">simpson_diversity</span><span class="p">)</span>
        <span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">library_col</span><span class="p">)</span>
        <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">weighted_diversity</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">/</span> <span class="n">x</span><span class="o">.</span><span class="n">npairs</span><span class="o">.</span><span class="n">sum</span><span class="p">())</span>
        <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;fraction_identical&#39;</span><span class="p">)</span>
        <span class="c1"># estimate accuracy as square root of fraction identical</span>
        <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">accuracy</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">fraction_identical</span><span class="p">))</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">drop_library_col</span><span class="p">:</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">library_col</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="s1">&#39;columns&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">result</span></div>


<div class="viewcode-block" id="simpleConsensus"><a class="viewcode-back" href="../../dms_tools2.barcodes.html#dms_tools2.barcodes.simpleConsensus">[docs]</a><span class="k">def</span> <span class="nf">simpleConsensus</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span>
        <span class="n">barcode_col</span><span class="o">=</span><span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="n">substitution_col</span><span class="o">=</span><span class="s1">&#39;substitutions&#39;</span><span class="p">,</span>
        <span class="n">insertion_col</span><span class="o">=</span><span class="s1">&#39;insertions&#39;</span><span class="p">,</span> <span class="n">deletion_col</span><span class="o">=</span><span class="s1">&#39;deletions&#39;</span><span class="p">,</span>
        <span class="n">library_col</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">max_sub_diffs</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">max_indel_diffs</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
        <span class="n">max_minor_muts</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Simple method to get consensus of mutations within barcode.</span>

<span class="sd">    Args:</span>
<span class="sd">        `df` (pandas Data Frame)</span>
<span class="sd">            Holds variants and their barcodes. Each row gives a</span>
<span class="sd">            sequence variant and its barcode. There need to be</span>
<span class="sd">            columns with the names given by the next four arguments</span>
<span class="sd">            described below.</span>
<span class="sd">        `barcode_col` (str)</span>
<span class="sd">            Name of column holding barcodes.</span>
<span class="sd">        `substitution_col` (str)</span>
<span class="sd">            Name of column holding substitutions as list of strings.</span>
<span class="sd">        `insertion_col` (str)</span>
<span class="sd">            Name of column holding insertions as list of strings.</span>
<span class="sd">        `deletion_col` (str)</span>
<span class="sd">            Name of column holding insertions as list of strings.</span>
<span class="sd">        `library_col` (`None` or str)</span>
<span class="sd">            If we have multiple libraries, analyze each barcode only</span>
<span class="sd">            within its library. In that case, `library_col` should be</span>
<span class="sd">            name of column giving library name.</span>
<span class="sd">        `max_sub_diffs` (int)</span>
<span class="sd">            Drop any barcode where any variant differs from all other</span>
<span class="sd">            variants for that barcode by more than this many substitution</span>
<span class="sd">            (point mutation) differences.</span>
<span class="sd">        `max_indel_diffs` (int)</span>
<span class="sd">            Drop any barcode where any variant differs from all other</span>
<span class="sd">            variants for that barcode by more than this many indel</span>
<span class="sd">            (insertion or deletion) differences.</span>
<span class="sd">        `max_minor_muts` (int)</span>
<span class="sd">            Drop any barcode where there is a minor (non-consensus)</span>
<span class="sd">            mutation found more than this many times.</span>

<span class="sd">    Returns:</span>
<span class="sd">        The 2-tuple `(consensus, dropped)`. These are each data frames:</span>

<span class="sd">            - `consensus` is a new data frame with a row for each</span>
<span class="sd">              barcode for which we could call the consensus. The</span>
<span class="sd">              columns have the same names as `barcode_col`,</span>
<span class="sd">              `substitution_col`, `insertion_col`, `deletion_col`,</span>
<span class="sd">              and (optionally) `library_col`--but the the three</span>
<span class="sd">              columns for the mutations now just list the **consensus**</span>
<span class="sd">              mutations of that type. In addition, there is a new</span>
<span class="sd">              column called &quot;variant_call_support&quot; that gives the number</span>
<span class="sd">              of sequences supporting the call of that barcode.</span>

<span class="sd">            - `dropped` simply contains all rows in the original `df`</span>
<span class="sd">              that correspond to sequences that were dropped due to</span>
<span class="sd">              `max_diffs` or `max_minor_muts` not being satisfied.</span>
<span class="sd">              There is also a column called &quot;drop_reason&quot; that gives</span>
<span class="sd">              the reason that the barcode was dropped.</span>

<span class="sd">    The approach is as follows:</span>

<span class="sd">      1. Group all variants by library and barcode.</span>

<span class="sd">      2. If there are multiple sequences, check if any of them differ</span>
<span class="sd">         from all the others by more than `max_diffs` mutations total</span>
<span class="sd">         (taking substitutions, insertions, and deletions together).</span>
<span class="sd">         If so, drop the entire barcode. The reason is that if there</span>
<span class="sd">         are variants that are very different, it becomes likely that</span>
<span class="sd">         it isn&#39;t just sequencing error, but rather something is wrong</span>
<span class="sd">         with that barcode (multiple variants with same barcode or</span>
<span class="sd">         strand exchange).</span>

<span class="sd">      3. Take the consensus of the sequences, which means keeping</span>
<span class="sd">         mutations that are present in **greater** than half of the</span>
<span class="sd">         variants. Note that this calling scheme means that the</span>
<span class="sd">         consensus being called is dependent on the reference used</span>
<span class="sd">         to call the mutations, which is an important caveat if you</span>
<span class="sd">         are calling variants relative to multiple different parent</span>
<span class="sd">         sequences.</span>

<span class="sd">      4. If there are any minor mutations (mutations not in consensus)</span>
<span class="sd">         that are present in more than `max_minor_muts` variants or missing</span>
<span class="sd">         from more than `max_minor_muts` variants, then</span>
<span class="sd">         drop that barcode. The reason is that recurring minor mutations</span>
<span class="sd">         also suggest some problem more complex than sequencing error</span>
<span class="sd">         that may render the whole barcode family invalid.</span>

<span class="sd">    Note that this method returns a consensus even if there is just</span>
<span class="sd">    one sequence for the barcode (in that case, this sequence is</span>
<span class="sd">    the consensus). This is fine--if you want to get consensus calls</span>
<span class="sd">    that are more strongly supported, simply filter the returned</span>
<span class="sd">    `consensus` data frame for larger values of `variant_call_support`,</span>
<span class="sd">    as the more sequences that support a call the more accurate it</span>
<span class="sd">    is expected to be.</span>

<span class="sd">    Here is an example:</span>

<span class="sd">    &gt;&gt;&gt; df = pandas.DataFrame([</span>
<span class="sd">    ...     (&#39;s1&#39;, &#39;AG&#39;, [&#39;A2C&#39;], [], [&#39;del5to7&#39;]),</span>
<span class="sd">    ...     (&#39;s1&#39;, &#39;AG&#39;, [&#39;A2C&#39;], [], []),</span>
<span class="sd">    ...     (&#39;s1&#39;, &#39;TA&#39;, [&#39;G3A&#39;], [&#39;ins4len3&#39;], []),</span>
<span class="sd">    ...     (&#39;s2&#39;, &#39;TA&#39;, [&#39;C5A&#39;, &#39;T6C&#39;], [], []),</span>
<span class="sd">    ...     (&#39;s2&#39;, &#39;TA&#39;, [&#39;T6C&#39;], [&#39;ins5len1&#39;], []),</span>
<span class="sd">    ...     (&#39;s2&#39;, &#39;TA&#39;, [&#39;T6C&#39;], [], []),</span>
<span class="sd">    ...     (&#39;s2&#39;, &#39;TG&#39;, [&#39;T6A&#39;], [], []),</span>
<span class="sd">    ...     (&#39;s2&#39;, &#39;TG&#39;, [&#39;A2G&#39;], [], []),</span>
<span class="sd">    ...     (&#39;s2&#39;, &#39;GG&#39;, [], [], [&#39;del1to4&#39;]),</span>
<span class="sd">    ...     (&#39;s2&#39;, &#39;GG&#39;, [&#39;A1C&#39;], [], []),</span>
<span class="sd">    ...     (&#39;s2&#39;, &#39;AA&#39;, [], [], []),</span>
<span class="sd">    ...     (&#39;s2&#39;, &#39;AA&#39;, [], [], []),</span>
<span class="sd">    ...     (&#39;s2&#39;, &#39;AA&#39;, [&#39;T6C&#39;], [], []),</span>
<span class="sd">    ...     (&#39;s2&#39;, &#39;AA&#39;, [&#39;T6C&#39;], [], []),</span>
<span class="sd">    ...     (&#39;s3&#39;, &#39;AA&#39;, [&#39;T6G&#39;], [&#39;ins1len1&#39;], [&#39;del1to2&#39;]),</span>
<span class="sd">    ...     (&#39;s3&#39;, &#39;AA&#39;, [&#39;T6G&#39;], [], [&#39;del5to7&#39;])],</span>
<span class="sd">    ...     columns=[&#39;library&#39;, &#39;barcode&#39;, &#39;substitutions&#39;,</span>
<span class="sd">    ...              &#39;insertions&#39;, &#39;deletions&#39;]</span>
<span class="sd">    ...     )</span>
<span class="sd">    &gt;&gt;&gt; consensus, dropped = simpleConsensus(df, library_col=&#39;library&#39;)</span>
<span class="sd">    &gt;&gt;&gt; consensus</span>
<span class="sd">      library barcode substitutions  insertions deletions  variant_call_support</span>
<span class="sd">    0      s1      AG         [A2C]          []        []                     2</span>
<span class="sd">    1      s1      TA         [G3A]  [ins4len3]        []                     1</span>
<span class="sd">    2      s2      GG            []          []        []                     2</span>
<span class="sd">    3      s2      TA         [T6C]          []        []                     3</span>
<span class="sd">    &gt;&gt;&gt; pandas.set_option(&#39;display.max_columns&#39;, 10)</span>
<span class="sd">    &gt;&gt;&gt; pandas.set_option(&#39;display.width&#39;, 500)</span>
<span class="sd">    &gt;&gt;&gt; dropped</span>
<span class="sd">      library barcode substitutions  insertions  deletions           drop_reason</span>
<span class="sd">    0      s2      TG         [T6A]          []         []  excess substitutions</span>
<span class="sd">    1      s2      TG         [A2G]          []         []  excess substitutions</span>
<span class="sd">    2      s2      AA            []          []         []     excess minor muts</span>
<span class="sd">    3      s2      AA            []          []         []     excess minor muts</span>
<span class="sd">    4      s2      AA         [T6C]          []         []     excess minor muts</span>
<span class="sd">    5      s2      AA         [T6C]          []         []     excess minor muts</span>
<span class="sd">    6      s3      AA         [T6G]  [ins1len1]  [del1to2]         excess indels</span>
<span class="sd">    7      s3      AA         [T6G]          []  [del5to7]         excess indels</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">library_col</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">library_col</span> <span class="o">=</span> <span class="s1">&#39;library&#39;</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">library_col</span><span class="o">=</span><span class="s1">&#39;dummy&#39;</span><span class="p">)</span>
        <span class="n">drop_library_col</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">drop_library_col</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="n">mut_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">substitution_col</span><span class="p">,</span> <span class="n">insertion_col</span><span class="p">,</span> <span class="n">deletion_col</span><span class="p">]</span>
    <span class="n">all_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">library_col</span><span class="p">,</span> <span class="n">barcode_col</span><span class="p">]</span> <span class="o">+</span> <span class="n">mut_cols</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">([</span><span class="n">col</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">all_cols</span><span class="p">]):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Cannot find column </span><span class="si">{col}</span><span class="s2">&quot;</span><span class="p">)</span>


    <span class="c1"># make sure no mutations duplicated, otherwise approach below fails</span>
    <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">mut_cols</span><span class="p">:</span>
        <span class="n">duplicated</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="nb">len</span><span class="p">)</span> <span class="o">-</span> <span class="n">df</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="nb">len</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">duplicated</span><span class="o">.</span><span class="n">any</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;duplicated </span><span class="si">{col}</span><span class="s2">:</span><span class="se">\n</span><span class="s2">&quot;</span>
                             <span class="n">f</span><span class="s2">&quot;</span><span class="si">{df[col][duplicated &gt; 0]}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">dropped</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">consensus</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="p">(</span><span class="n">library</span><span class="p">,</span> <span class="n">barcode</span><span class="p">),</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">df</span><span class="p">[</span><span class="n">all_cols</span><span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span>
            <span class="p">[</span><span class="n">library_col</span><span class="p">,</span> <span class="n">barcode_col</span><span class="p">]):</span>

        <span class="n">nseqs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">g</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">nseqs</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">consensus</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span> <span class="o">+</span> <span class="p">[</span><span class="n">nseqs</span><span class="p">])</span>
            <span class="k">continue</span>

        <span class="n">consensus_failed</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="c1"># are max_sub_diffs and max_indel_diffs satisfied?</span>
        <span class="k">for</span> <span class="n">difftype</span><span class="p">,</span> <span class="n">diff_cols</span><span class="p">,</span> <span class="n">max_diffs</span> <span class="ow">in</span> <span class="p">[</span>
                <span class="p">(</span><span class="s1">&#39;substitutions&#39;</span><span class="p">,</span> <span class="p">[</span><span class="n">substitution_col</span><span class="p">],</span> <span class="n">max_sub_diffs</span><span class="p">),</span>
                <span class="p">(</span><span class="s1">&#39;indels&#39;</span><span class="p">,</span> <span class="p">[</span><span class="n">insertion_col</span><span class="p">,</span> <span class="n">deletion_col</span><span class="p">],</span> <span class="n">max_indel_diffs</span><span class="p">)]:</span>
            <span class="n">min_variant_diffs</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="k">lambda</span><span class="p">:</span> <span class="n">max_diffs</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">v1</span><span class="p">,</span> <span class="n">v2</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">combinations</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">itertuples</span><span class="p">(),</span> <span class="mi">2</span><span class="p">):</span>
                <span class="n">i1</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">v1</span><span class="p">,</span> <span class="s1">&#39;Index&#39;</span><span class="p">)</span>
                <span class="n">i2</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">v2</span><span class="p">,</span> <span class="s1">&#39;Index&#39;</span><span class="p">)</span>
                <span class="n">n_diffs</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">([</span><span class="nb">len</span><span class="p">(</span>
                        <span class="nb">set</span><span class="p">(</span><span class="nb">getattr</span><span class="p">(</span><span class="n">v1</span><span class="p">,</span> <span class="n">col</span><span class="p">))</span><span class="o">.</span><span class="n">symmetric_difference</span><span class="p">(</span>
                        <span class="nb">set</span><span class="p">(</span><span class="nb">getattr</span><span class="p">(</span><span class="n">v2</span><span class="p">,</span> <span class="n">col</span><span class="p">))))</span>
                        <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">diff_cols</span><span class="p">])</span>
                <span class="n">min_variant_diffs</span><span class="p">[</span><span class="n">i1</span><span class="p">]</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span>
                        <span class="n">min_variant_diffs</span><span class="p">[</span><span class="n">i1</span><span class="p">],</span> <span class="n">n_diffs</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">nseqs</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="ow">and</span> <span class="nb">any</span><span class="p">(</span>
                    <span class="p">[</span><span class="n">d</span> <span class="o">&gt;</span> <span class="n">max_diffs</span> <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">min_variant_diffs</span><span class="o">.</span><span class="n">values</span><span class="p">()]):</span>
                <span class="c1"># need to add to `dropped` because of max_diffs failing</span>
                <span class="n">dropped</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="n">g</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">drop_reason</span><span class="o">=</span><span class="n">f</span><span class="s2">&quot;excess </span><span class="si">{difftype}</span><span class="s2">&quot;</span><span class="p">)</span>
                        <span class="p">)</span>
                <span class="n">consensus_failed</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">break</span>
        <span class="k">if</span> <span class="n">consensus_failed</span><span class="p">:</span>
            <span class="k">continue</span>

        <span class="c1"># get consensus and see if `max_minor_muts` is satisfied</span>
        <span class="n">g_consensus</span> <span class="o">=</span> <span class="p">[</span><span class="n">library</span><span class="p">,</span> <span class="n">barcode</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">mut_cols</span><span class="p">:</span>
            <span class="n">counts</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">Counter</span><span class="p">(</span>
                    <span class="n">itertools</span><span class="o">.</span><span class="n">chain</span><span class="o">.</span><span class="n">from_iterable</span><span class="p">(</span><span class="n">g</span><span class="p">[</span><span class="n">col</span><span class="p">]))</span>
            <span class="k">if</span> <span class="nb">any</span><span class="p">([</span><span class="n">max_minor_muts</span> <span class="o">&lt;</span> <span class="n">count</span> <span class="o">&lt;</span> <span class="p">(</span><span class="n">nseqs</span> <span class="o">-</span> <span class="n">max_minor_muts</span><span class="p">)</span>
                   <span class="k">for</span> <span class="n">count</span> <span class="ow">in</span> <span class="n">counts</span><span class="o">.</span><span class="n">values</span><span class="p">()]):</span>
                <span class="n">consensus_failed</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">break</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">col_consensus</span> <span class="o">=</span> <span class="p">[</span><span class="n">mut</span> <span class="k">for</span> <span class="n">mut</span><span class="p">,</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">counts</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
                        <span class="k">if</span> <span class="n">c</span> <span class="o">&gt;</span> <span class="mf">0.5</span> <span class="o">*</span> <span class="n">nseqs</span><span class="p">]</span>
                <span class="c1"># order mutations based on first number in string</span>
                <span class="n">n_col_consensus</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">col_consensus</span><span class="p">:</span>
                    <span class="n">m</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s1">&#39;(\-{0,1}\d+)&#39;</span><span class="p">,</span> <span class="n">mut</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">m</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                        <span class="n">n_col_consensus</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">math</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">mut</span><span class="p">))</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">n_col_consensus</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="nb">int</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">()),</span> <span class="n">mut</span><span class="p">))</span>
                <span class="n">g_consensus</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">mut</span> <span class="k">for</span> <span class="n">n</span><span class="p">,</span> <span class="n">mut</span> <span class="ow">in</span>
                        <span class="nb">sorted</span><span class="p">(</span><span class="n">n_col_consensus</span><span class="p">)])</span>
        <span class="k">if</span> <span class="n">consensus_failed</span><span class="p">:</span>
            <span class="c1"># need to add to dropped</span>
            <span class="n">dropped</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="n">g</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">drop_reason</span><span class="o">=</span><span class="s2">&quot;excess minor muts&quot;</span><span class="p">)</span>
                    <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">consensus</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g_consensus</span> <span class="o">+</span> <span class="p">[</span><span class="n">nseqs</span><span class="p">])</span>

    <span class="n">consensus</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">consensus</span><span class="p">,</span>
            <span class="n">columns</span><span class="o">=</span><span class="n">all_cols</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;variant_call_support&#39;</span><span class="p">])</span>
    <span class="k">if</span> <span class="n">dropped</span><span class="p">:</span>
        <span class="n">dropped</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">dropped</span><span class="p">)</span><span class="o">.</span><span class="n">sort_index</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">dropped</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">drop_library_col</span><span class="p">:</span>
        <span class="n">dropped</span> <span class="o">=</span> <span class="n">dropped</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">library_col</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="s1">&#39;columns&#39;</span><span class="p">)</span>
        <span class="n">consensus</span> <span class="o">=</span> <span class="n">consensus</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">library_col</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="s1">&#39;columns&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="p">(</span><span class="n">consensus</span><span class="p">,</span> <span class="n">dropped</span><span class="p">)</span></div>


<div class="viewcode-block" id="IlluminaBarcodeParser"><a class="viewcode-back" href="../../dms_tools2.barcodes.html#dms_tools2.barcodes.IlluminaBarcodeParser">[docs]</a><span class="k">class</span> <span class="nc">IlluminaBarcodeParser</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;Parser for Illumina barcodes.</span>

<span class="sd">    The barcodes should be read by R1 and optionally R2.</span>
<span class="sd">    The arrangement of elements is shown below::</span>

<span class="sd">        5&#39;-[R2_start]-upstream-barcode-downstream-[R1_start]-3&#39;</span>

<span class="sd">    R1 anneals downstream of the barcode and reads backwards. If</span>
<span class="sd">    R2 is used, it anneals upstream of the barcode and reads forward.</span>
<span class="sd">    There can be sequences (`upstream` and `downstream`) on either</span>
<span class="sd">    side of the barcode: `downstream` must fully cover the</span>
<span class="sd">    region between where R1 starts and the barcode, and if you are</span>
<span class="sd">    using R2 then `upstream` must fully cover the region between</span>
<span class="sd">    where R2 starts and the barcode. However, it is fine if R1</span>
<span class="sd">    reads backwards past `upstream`, and if `R2` reads forward</span>
<span class="sd">    past `downstream`.</span>

<span class="sd">    Args:</span>
<span class="sd">        `bclen` (int)</span>
<span class="sd">            Length of the barcode.</span>
<span class="sd">        `upstream` (str)</span>
<span class="sd">            Sequence upstream of the barcode.</span>
<span class="sd">        `downstream` (str)</span>
<span class="sd">            Sequence downstream of barcode.</span>
<span class="sd">        `upstream_mismatch` (int)</span>
<span class="sd">            Max number of mismatches allowed in `upstream`.</span>
<span class="sd">        `downstream_mismatch` (int)</span>
<span class="sd">            Like `upstream_mismatches` but for `downstream`.</span>
<span class="sd">        `valid_barcodes` (`None` or iterable such as list, Series)</span>
<span class="sd">            If not `None`, only retain barcodes listed here.</span>
<span class="sd">            Use if you know the set of possible valid barcodes.</span>
<span class="sd">        `rc_barcode` (bool)</span>
<span class="sd">            Parse the reverse complement of the barcode (the</span>
<span class="sd">            orientation read by R1).</span>
<span class="sd">        `minq` (int)</span>
<span class="sd">            Require at least this quality score for all bases</span>
<span class="sd">            in barcode.</span>
<span class="sd">        `chastity_filter` (bool)</span>
<span class="sd">            Drop any reads that fail Illumina chastity filter.</span>
<span class="sd">        `list_all_valid_barcodes` (bool)</span>
<span class="sd">            If using `valid_barcodes`, then barcode sets returned</span>
<span class="sd">            by :class:`IlluminaBarcodeParser.parse` includes all</span>
<span class="sd">            valid barcodes even if no counts.</span>

<span class="sd">    To use, first initialize a :class:`IlluminaBarcodeParser`, then</span>
<span class="sd">    parse barcodes using :class:`IlluminaBarcodeParser.parse`.</span>
<span class="sd">    Barcodes are retained as valid only if R1 and R2 agree at every</span>
<span class="sd">    nucleotide in barcode and if at each site at least one read has</span>
<span class="sd">    a quality of at least `minq`.</span>

<span class="sd">    Here is an example. Imagine we are parsing 4 nucleotide barcodes</span>
<span class="sd">    that have the following construction::</span>

<span class="sd">        5&#39;-[R2 binding site]-ACATGA-NNNN-GACT-[R1 binding site]-3&#39;</span>

<span class="sd">    First, we initialize an appropriate :class:`IlluminaBarcodeParser`:</span>

<span class="sd">    &gt;&gt;&gt; parser = IlluminaBarcodeParser(</span>
<span class="sd">    ...              bclen=4,</span>
<span class="sd">    ...              upstream=&#39;ACATGA&#39;,</span>
<span class="sd">    ...              downstream=&#39;GACT&#39;</span>
<span class="sd">    ...              )</span>

<span class="sd">    Now we write some test FASTQ files. We write valid test</span>
<span class="sd">    reads and some invalid reads. The header for each read</span>
<span class="sd">    explains why it is valid / invalid. We use quality scores</span>
<span class="sd">    of ``?`` (30) or ``+`` (10) for high- and low-quality bases:</span>

<span class="sd">    &gt;&gt;&gt; r1file = &#39;_temp_R1.fastq&#39;</span>
<span class="sd">    &gt;&gt;&gt; r2file = &#39;_temp_R2.fastq&#39;</span>
<span class="sd">    &gt;&gt;&gt; with open(r1file, &#39;w&#39;) as f1, open(r2file, &#39;w&#39;) as f2:</span>
<span class="sd">    ...</span>
<span class="sd">    ...     # valid TACG barcode, full flanking regions</span>
<span class="sd">    ...     _ = f1.write(</span>
<span class="sd">    ...         &#39;@valid_CGTA_barcode_full_flanking_region\\n&#39;</span>
<span class="sd">    ...         &#39;AGTCCGTATCATGT\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;??????????????\\n&#39;)</span>
<span class="sd">    ...     _ = f2.write(</span>
<span class="sd">    ...         &#39;@valid_CGTA_barcode_full_flanking_region\\n&#39;</span>
<span class="sd">    ...         &#39;ACATGATACGGACT\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;??????????????\\n&#39;)</span>
<span class="sd">    ...</span>
<span class="sd">    ...     # valid CGTA barcode, partial flanking regions</span>
<span class="sd">    ...     _ = f1.write(</span>
<span class="sd">    ...         &#39;@valid_CGTA_barcode_partial_flanking_region\\n&#39;</span>
<span class="sd">    ...         &#39;AGTCCGTATCAT\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;??????????????\\n&#39;)</span>
<span class="sd">    ...     _ = f2.write(</span>
<span class="sd">    ...         &#39;@valid_CGTA_barcode_partial_flanking_region\\n&#39;</span>
<span class="sd">    ...         &#39;ACATGATACG\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;??????????????\\n&#39;)</span>
<span class="sd">    ...</span>
<span class="sd">    ...     # valid GCCG barcode, extended flanking regions</span>
<span class="sd">    ...     _ = f1.write(</span>
<span class="sd">    ...         &#39;@valid_GCCG_barcode_extended_flanking_region\\n&#39;</span>
<span class="sd">    ...         &#39;AGTCGCCGTCATGTTAC\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;??????????????\\n&#39;)</span>
<span class="sd">    ...     _ = f2.write(</span>
<span class="sd">    ...         &#39;@valid_GCCG_barcode_extended_flanking_region\\n&#39;</span>
<span class="sd">    ...         &#39;ACATGACGGCGACTGAC\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;??????????????\\n&#39;)</span>
<span class="sd">    ...</span>
<span class="sd">    ...     # AAGT barcode in R1 but R2 differs</span>
<span class="sd">    ...     _ = f1.write(</span>
<span class="sd">    ...         &#39;@AAGT_R1_barcode_but_R2_differs\\n&#39;</span>
<span class="sd">    ...         &#39;AGTCAAGTTCATGT\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;??????????????\\n&#39;)</span>
<span class="sd">    ...     _ = f2.write(</span>
<span class="sd">    ...         &#39;@AAGT_R1_barcode_but_R2_differs\\n&#39;</span>
<span class="sd">    ...         &#39;ACATGAACTAGACT\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;??????????????\\n&#39;)</span>
<span class="sd">    ...</span>
<span class="sd">    ...     # same site low quality in R1 and R2</span>
<span class="sd">    ...     _ = f1.write(</span>
<span class="sd">    ...         &#39;@low_quality_site_in_R1_and_R2\\n&#39;</span>
<span class="sd">    ...         &#39;AGTCCGTATCATGT\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;?????+????????\\n&#39;)</span>
<span class="sd">    ...     _ = f2.write(</span>
<span class="sd">    ...         &#39;@low_quality_site_in_R1_and_R2\\n&#39;</span>
<span class="sd">    ...         &#39;ACATGATACGGACT\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;????????+?????\\n&#39;)</span>
<span class="sd">    ...</span>
<span class="sd">    ...     # different site low quality in R1 and R2</span>
<span class="sd">    ...     _ = f1.write(</span>
<span class="sd">    ...         &#39;@AGTA_with_low_quality_site_in_R1\\n&#39;</span>
<span class="sd">    ...         &#39;AGTCAGTATCATGT\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;?????+????????\\n&#39;)</span>
<span class="sd">    ...     _ = f2.write(</span>
<span class="sd">    ...         &#39;@AGTA_with_low_quality_site_in_R1\\n&#39;</span>
<span class="sd">    ...         &#39;ACATGATACTGACT\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;?????????+????\\n&#39;)</span>
<span class="sd">    ...</span>
<span class="sd">    ...     # N in barcode</span>
<span class="sd">    ...     _ = f1.write(</span>
<span class="sd">    ...         &#39;@N_in_barcode\\n&#39;</span>
<span class="sd">    ...         &#39;AGTCCGTNTCATGT\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;??????????????\\n&#39;)</span>
<span class="sd">    ...     _ = f2.write(</span>
<span class="sd">    ...         &#39;@N_in_barcode\\n&#39;</span>
<span class="sd">    ...         &#39;ACATGATACGGACT\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;??????????????\\n&#39;)</span>
<span class="sd">    ...</span>
<span class="sd">    ...     # GGAG barcode, one mismatch in each flanking region</span>
<span class="sd">    ...     _ = f1.write(</span>
<span class="sd">    ...         &#39;@GGAG_barcode_one_mismatch_per_flank\\n&#39;</span>
<span class="sd">    ...         &#39;GGTCGGAGTCATGA\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;??????????????\\n&#39;)</span>
<span class="sd">    ...     _ = f2.write(</span>
<span class="sd">    ...         &#39;@GGAG_barcode_one_mismatch_per_flank\\n&#39;</span>
<span class="sd">    ...         &#39;TCATGACTCCGACG\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;??????????????\\n&#39;)</span>
<span class="sd">    ...</span>
<span class="sd">    ...     # GGAG barcode, two mismatch in a flanking region</span>
<span class="sd">    ...     _ = f1.write(</span>
<span class="sd">    ...         &#39;@GGAG_barcode_two_mismatch_in_a_flank\\n&#39;</span>
<span class="sd">    ...         &#39;GGTCGGAGTCATAA\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;??????????????\\n&#39;)</span>
<span class="sd">    ...     _ = f2.write(</span>
<span class="sd">    ...         &#39;@GGAG_barcode_two_mismatch_in_a_flank\\n&#39;</span>
<span class="sd">    ...         &#39;TCATGACTCCGACG\\n&#39;</span>
<span class="sd">    ...         &#39;+\\n&#39;</span>
<span class="sd">    ...         &#39;??????????????\\n&#39;)</span>


<span class="sd">    Now parse the barcodes using both the R1 and R2 files:</span>

<span class="sd">    &gt;&gt;&gt; barcodes, fates = parser.parse(r1file, r2file)</span>
<span class="sd">    &gt;&gt;&gt; print(barcodes.to_csv(sep=&#39; &#39;, index=False).strip())</span>
<span class="sd">    barcode count</span>
<span class="sd">    CGTA 2</span>
<span class="sd">    AGTA 1</span>
<span class="sd">    GCCG 1</span>
<span class="sd">    &gt;&gt;&gt; print(fates.to_csv(sep=&#39; &#39;, index=False).strip())</span>
<span class="sd">    fate count</span>
<span class="sd">    &quot;valid barcode&quot; 4</span>
<span class="sd">    &quot;unparseable barcode&quot; 3</span>
<span class="sd">    &quot;R1 / R2 disagree&quot; 1</span>
<span class="sd">    &quot;low quality barcode&quot; 1</span>

<span class="sd">    Now we parse just using R1. We gain the barcode where R1 and</span>
<span class="sd">    R2 disagree, but lose the one where R1 is low quality at a</span>
<span class="sd">    position where R2 is OK:</span>

<span class="sd">    &gt;&gt;&gt; barcodes, fates = parser.parse(r1file)</span>
<span class="sd">    &gt;&gt;&gt; print(barcodes.to_csv(sep=&#39; &#39;, index=False).strip())</span>
<span class="sd">    barcode count</span>
<span class="sd">    CGTA 2</span>
<span class="sd">    AAGT 1</span>
<span class="sd">    GCCG 1</span>
<span class="sd">    &gt;&gt;&gt; print(fates.to_csv(sep=&#39; &#39;, index=False).strip())</span>
<span class="sd">    fate count</span>
<span class="sd">    &quot;valid barcode&quot; 4</span>
<span class="sd">    &quot;unparseable barcode&quot; 3</span>
<span class="sd">    &quot;low quality barcode&quot; 2</span>

<span class="sd">    Now create a parser that allows a mismatch in each flanking</span>
<span class="sd">    region, and check that we recover a &quot;GGAG&quot; barcode:</span>

<span class="sd">    &gt;&gt;&gt; parser_mismatch = IlluminaBarcodeParser(</span>
<span class="sd">    ...              bclen=4,</span>
<span class="sd">    ...              upstream=&#39;ACATGA&#39;,</span>
<span class="sd">    ...              downstream=&#39;GACT&#39;,</span>
<span class="sd">    ...              upstream_mismatch=1,</span>
<span class="sd">    ...              downstream_mismatch=1,</span>
<span class="sd">    ...              )</span>
<span class="sd">    &gt;&gt;&gt; barcodes_mismatch, fates_mismatch = parser_mismatch.parse(r1file, r2file)</span>
<span class="sd">    &gt;&gt;&gt; print(barcodes_mismatch.to_csv(sep=&#39; &#39;, index=False).strip())</span>
<span class="sd">    barcode count</span>
<span class="sd">    CGTA 2</span>
<span class="sd">    AGTA 1</span>
<span class="sd">    GCCG 1</span>
<span class="sd">    GGAG 1</span>
<span class="sd">    &gt;&gt;&gt; print(fates_mismatch.to_csv(sep=&#39; &#39;, index=False).strip())</span>
<span class="sd">    fate count</span>
<span class="sd">    &quot;valid barcode&quot; 5</span>
<span class="sd">    &quot;unparseable barcode&quot; 2</span>
<span class="sd">    &quot;R1 / R2 disagree&quot; 1</span>
<span class="sd">    &quot;low quality barcode&quot; 1</span>

<span class="sd">    Now parse the barcodes using `valid_barcodes` to set a</span>
<span class="sd">    barcode whitelist:</span>

<span class="sd">    &gt;&gt;&gt; parser_wl = IlluminaBarcodeParser(</span>
<span class="sd">    ...              bclen=4,</span>
<span class="sd">    ...              upstream=&#39;ACATGA&#39;,</span>
<span class="sd">    ...              downstream=&#39;GACT&#39;,</span>
<span class="sd">    ...              valid_barcodes={&#39;CGTA&#39;, &#39;AGTA&#39;, &#39;TAAT&#39;}</span>
<span class="sd">    ...              )</span>
<span class="sd">    &gt;&gt;&gt; barcodes_wl, fates_wl = parser_wl.parse(r1file, r2file)</span>
<span class="sd">    &gt;&gt;&gt; print(barcodes_wl.to_csv(sep=&#39; &#39;, index=False).strip())</span>
<span class="sd">    barcode count</span>
<span class="sd">    CGTA 2</span>
<span class="sd">    AGTA 1</span>
<span class="sd">    TAAT 0</span>
<span class="sd">    &gt;&gt;&gt; print(fates_wl.to_csv(sep=&#39; &#39;, index=False).strip())</span>
<span class="sd">    fate count</span>
<span class="sd">    &quot;unparseable barcode&quot; 3</span>
<span class="sd">    &quot;valid barcode&quot; 3</span>
<span class="sd">    &quot;R1 / R2 disagree&quot; 1</span>
<span class="sd">    &quot;invalid barcode&quot; 1</span>
<span class="sd">    &quot;low quality barcode&quot; 1</span>

<span class="sd">    Remove the test FASTQ files:</span>

<span class="sd">    &gt;&gt;&gt; os.remove(r1file)</span>
<span class="sd">    &gt;&gt;&gt; os.remove(r2file)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1">#: valid nucleotide characters</span>
    <span class="n">VALID_NTS</span> <span class="o">=</span> <span class="s1">&#39;ACGTN&#39;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">bclen</span><span class="p">,</span>
            <span class="n">upstream</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">downstream</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span>
            <span class="n">upstream_mismatch</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">downstream_mismatch</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
            <span class="n">valid_barcodes</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">rc_barcode</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">minq</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>
            <span class="n">chastity_filter</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">list_all_valid_barcodes</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;See main class doc string.&quot;&quot;&quot;</span>

        <span class="c1"># first make all arguments into attributes</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">bclen</span> <span class="o">=</span> <span class="n">bclen</span>
        <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;^[</span><span class="si">{self.VALID_NTS}</span><span class="s2">]*$&quot;</span><span class="p">,</span> <span class="n">upstream</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">upstream</span> <span class="o">=</span> <span class="n">upstream</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid chars in upstream </span><span class="si">{upstream}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;^[</span><span class="si">{self.VALID_NTS}</span><span class="s2">]*$&quot;</span><span class="p">,</span> <span class="n">downstream</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">downstream</span> <span class="o">=</span> <span class="n">downstream</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid chars in downstream </span><span class="si">{downstream}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">upstream_mismatch</span> <span class="o">=</span> <span class="n">upstream_mismatch</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">downstream_mismatch</span> <span class="o">=</span> <span class="n">downstream_mismatch</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">valid_barcodes</span> <span class="o">=</span> <span class="n">valid_barcodes</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_barcodes</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">valid_barcodes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">valid_barcodes</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">minq</span> <span class="o">=</span> <span class="n">minq</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">rc_barcode</span> <span class="o">=</span> <span class="n">rc_barcode</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">chastity_filter</span> <span class="o">=</span> <span class="n">chastity_filter</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">list_all_valid_barcodes</span> <span class="o">=</span> <span class="n">list_all_valid_barcodes</span>

        <span class="c1"># specify information about R1 / R2 matches</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_bcend</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s1">&#39;R1&#39;</span><span class="p">:</span><span class="bp">self</span><span class="o">.</span><span class="n">bclen</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">downstream</span><span class="p">),</span>
                <span class="s1">&#39;R2&#39;</span><span class="p">:</span><span class="bp">self</span><span class="o">.</span><span class="n">bclen</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">upstream</span><span class="p">)</span>
                <span class="p">}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_rcdownstream</span> <span class="o">=</span> <span class="n">dms_tools2</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">reverseComplement</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">downstream</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_rcupstream</span> <span class="o">=</span> <span class="n">dms_tools2</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">reverseComplement</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">upstream</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_matches</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;R1&#39;</span><span class="p">:{},</span> <span class="s1">&#39;R2&#39;</span><span class="p">:{}}</span> <span class="c1"># saves match object by read length</span>


<div class="viewcode-block" id="IlluminaBarcodeParser.parse"><a class="viewcode-back" href="../../dms_tools2.barcodes.html#dms_tools2.barcodes.IlluminaBarcodeParser.parse">[docs]</a>    <span class="k">def</span> <span class="nf">parse</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">r1files</span><span class="p">,</span> <span class="n">r2files</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Parses barcodes from files.</span>

<span class="sd">        Args:</span>
<span class="sd">            `r1file` (str or list)</span>
<span class="sd">                Name of R1 FASTQ file, or list of such files</span>
<span class="sd">                Can optionally be gzipped.</span>
<span class="sd">            `r2file` (`None`, str, or list)</span>
<span class="sd">                `None` of not using R2, otherwise like R1.</span>

<span class="sd">        Returns:</span>
<span class="sd">            The 2-tuple `(barcodes, fates)`. In this 2-tuple:</span>

<span class="sd">                - `barcodes` is a pandas DataFrame giving the</span>
<span class="sd">                  number of observations of each barcode. The</span>
<span class="sd">                  columns are named &quot;barcode&quot; and &quot;count&quot;.</span>

<span class="sd">                - `fates` is a pandas DataFrame giving the</span>
<span class="sd">                  total number of reads with each fate. The</span>
<span class="sd">                  columns are named &quot;fate&quot; and &quot;count&quot;.</span>

<span class="sd">                  - &quot;valid barcode&quot;</span>

<span class="sd">                  - &quot;invalid barcode&quot;: not in our barcode whitelist</span>

<span class="sd">                  - &quot;R1 / R2 disagree&quot;</span>

<span class="sd">                  - &quot;low quality barcode&quot;: sequencing quality low</span>

<span class="sd">                  - &quot;unparseable barcode&quot;: invalid flanking sequences</span>
<span class="sd">                    or N in barcode.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">r2files</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">reads</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;R1&#39;</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">reads</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;R1&#39;</span><span class="p">,</span> <span class="s1">&#39;R2&#39;</span><span class="p">]</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_barcodes</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">list_all_valid_barcodes</span><span class="p">:</span>
            <span class="n">barcodes</span> <span class="o">=</span> <span class="p">{</span><span class="n">bc</span><span class="p">:</span><span class="mi">0</span> <span class="k">for</span> <span class="n">bc</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_barcodes</span><span class="p">}</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">barcodes</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>

        <span class="n">fates</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">r1</span><span class="p">,</span> <span class="n">r2</span><span class="p">,</span> <span class="n">q1</span><span class="p">,</span> <span class="n">q2</span><span class="p">,</span> <span class="n">fail</span> <span class="ow">in</span> \
                <span class="n">dms_tools2</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">iteratePairedFASTQ</span><span class="p">(</span><span class="n">r1files</span><span class="p">,</span> <span class="n">r2files</span><span class="p">):</span>

            <span class="k">if</span> <span class="n">fail</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">chastity_filter</span><span class="p">:</span>
                <span class="n">fates</span><span class="p">[</span><span class="s1">&#39;failed chastity filter&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">continue</span>

            <span class="n">matches</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">read</span><span class="p">,</span> <span class="n">r</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="p">[</span><span class="n">r1</span><span class="p">,</span> <span class="n">r2</span><span class="p">]):</span>
                <span class="n">rlen</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>

                <span class="c1"># get or build matcher for read of this length</span>
                <span class="n">len_past_bc</span> <span class="o">=</span> <span class="n">rlen</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">_bcend</span><span class="p">[</span><span class="n">read</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">len_past_bc</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;</span><span class="si">{read}</span><span class="s2"> too short: </span><span class="si">{rlen}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">rlen</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_matches</span><span class="p">[</span><span class="n">read</span><span class="p">]:</span>
                    <span class="n">matcher</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_matches</span><span class="p">[</span><span class="n">read</span><span class="p">][</span><span class="n">rlen</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">read</span> <span class="o">==</span> <span class="s1">&#39;R1&#39;</span><span class="p">:</span>
                        <span class="n">match_str</span> <span class="o">=</span> <span class="p">(</span>
                                <span class="n">f</span><span class="s1">&#39;^(</span><span class="si">{self._rcdownstream}</span><span class="s1">)&#39;</span>
                                <span class="n">f</span><span class="s1">&#39;{{s&lt;=</span><span class="si">{self.downstream_mismatch}</span><span class="s1">}}&#39;</span> <span class="o">+</span>
                                <span class="n">f</span><span class="s1">&#39;(?P&lt;bc&gt;N{{</span><span class="si">{self.bclen}</span><span class="s1">}})&#39;</span> <span class="o">+</span>
                                <span class="n">f</span><span class="s1">&#39;(</span><span class="si">{self._rcupstream[ : len_past_bc]}</span><span class="s1">)&#39;</span> <span class="o">+</span>
                                <span class="n">f</span><span class="s1">&#39;{{s&lt;=</span><span class="si">{self.upstream_mismatch}</span><span class="s1">}}&#39;</span>
                                <span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="k">assert</span> <span class="n">read</span> <span class="o">==</span> <span class="s1">&#39;R2&#39;</span>
                        <span class="n">match_str</span> <span class="o">=</span> <span class="p">(</span>
                                <span class="n">f</span><span class="s1">&#39;^(</span><span class="si">{self.upstream}</span><span class="s1">)&#39;</span> <span class="o">+</span>
                                <span class="n">f</span><span class="s1">&#39;{{s&lt;=</span><span class="si">{self.upstream_mismatch}</span><span class="s1">}}&#39;</span> <span class="o">+</span>
                                <span class="n">f</span><span class="s1">&#39;(?P&lt;bc&gt;N{{</span><span class="si">{self.bclen}</span><span class="s1">}})&#39;</span> <span class="o">+</span>
                                <span class="n">f</span><span class="s1">&#39;(</span><span class="si">{self.downstream[ : len_past_bc]}</span><span class="s1">)&#39;</span> <span class="o">+</span>
                                <span class="n">f</span><span class="s1">&#39;{{s&lt;=</span><span class="si">{self.downstream_mismatch}</span><span class="s1">}}&#39;</span>
                                <span class="p">)</span>
                    <span class="n">matcher</span> <span class="o">=</span> <span class="n">regex</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span>
                            <span class="n">dms_tools2</span><span class="o">.</span><span class="n">pacbio</span><span class="o">.</span><span class="n">re_expandIUPAC</span><span class="p">(</span><span class="n">match_str</span><span class="p">),</span>
                            <span class="n">flags</span><span class="o">=</span><span class="n">regex</span><span class="o">.</span><span class="n">BESTMATCH</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_matches</span><span class="p">[</span><span class="n">read</span><span class="p">][</span><span class="n">rlen</span><span class="p">]</span> <span class="o">=</span> <span class="n">matcher</span>

                <span class="n">m</span> <span class="o">=</span> <span class="n">matcher</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">m</span><span class="p">:</span>
                    <span class="n">matches</span><span class="p">[</span><span class="n">read</span><span class="p">]</span> <span class="o">=</span> <span class="n">m</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">break</span>

            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">matches</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">reads</span><span class="p">):</span>
                <span class="n">bc</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="n">bc_q</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="k">for</span> <span class="n">read</span><span class="p">,</span> <span class="n">q</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">reads</span><span class="p">,</span> <span class="p">[</span><span class="n">q1</span><span class="p">,</span> <span class="n">q2</span><span class="p">]):</span>
                    <span class="n">bc</span><span class="p">[</span><span class="n">read</span><span class="p">]</span> <span class="o">=</span> <span class="n">matches</span><span class="p">[</span><span class="n">read</span><span class="p">]</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;bc&#39;</span><span class="p">)</span>
                    <span class="n">bc_q</span><span class="p">[</span><span class="n">read</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span>
                                 <span class="nb">ord</span><span class="p">(</span><span class="n">qi</span><span class="p">)</span> <span class="o">-</span> <span class="mi">33</span> <span class="k">for</span> <span class="n">qi</span> <span class="ow">in</span>
                                 <span class="n">q</span><span class="p">[</span><span class="n">matches</span><span class="p">[</span><span class="n">read</span><span class="p">]</span><span class="o">.</span><span class="n">start</span><span class="p">(</span><span class="s1">&#39;bc&#39;</span><span class="p">)</span> <span class="p">:</span>
                                   <span class="n">matches</span><span class="p">[</span><span class="n">read</span><span class="p">]</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;bc&#39;</span><span class="p">)]],</span>
                                 <span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;int&#39;</span><span class="p">)</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">rc_barcode</span> <span class="ow">and</span> <span class="s1">&#39;R2&#39;</span> <span class="ow">in</span> <span class="n">reads</span><span class="p">:</span>
                    <span class="n">bc</span><span class="p">[</span><span class="s1">&#39;R2&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dms_tools2</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">reverseComplement</span><span class="p">(</span><span class="n">bc</span><span class="p">[</span><span class="s1">&#39;R2&#39;</span><span class="p">])</span>
                    <span class="n">bc_q</span><span class="p">[</span><span class="s1">&#39;R2&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">flip</span><span class="p">(</span><span class="n">bc_q</span><span class="p">[</span><span class="s1">&#39;R2&#39;</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
                <span class="k">elif</span> <span class="s1">&#39;R2&#39;</span> <span class="ow">in</span> <span class="n">reads</span><span class="p">:</span>
                    <span class="n">bc</span><span class="p">[</span><span class="s1">&#39;R1&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dms_tools2</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">reverseComplement</span><span class="p">(</span><span class="n">bc</span><span class="p">[</span><span class="s1">&#39;R1&#39;</span><span class="p">])</span>
                    <span class="n">bc_q</span><span class="p">[</span><span class="s1">&#39;R1&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">flip</span><span class="p">(</span><span class="n">bc_q</span><span class="p">[</span><span class="s1">&#39;R1&#39;</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">reads</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="k">if</span> <span class="p">(</span><span class="n">bc_q</span><span class="p">[</span><span class="s1">&#39;R1&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">minq</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
                        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_barcodes</span> <span class="ow">and</span> <span class="p">(</span>
                                <span class="n">bc</span><span class="p">[</span><span class="s1">&#39;R1&#39;</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_barcodes</span><span class="p">):</span>
                            <span class="n">fates</span><span class="p">[</span><span class="s1">&#39;invalid barcode&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">barcodes</span><span class="p">[</span><span class="n">bc</span><span class="p">[</span><span class="s1">&#39;R1&#39;</span><span class="p">]]</span> <span class="o">+=</span> <span class="mi">1</span>
                            <span class="n">fates</span><span class="p">[</span><span class="s1">&#39;valid barcode&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">fates</span><span class="p">[</span><span class="s1">&#39;low quality barcode&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">bc</span><span class="p">[</span><span class="s1">&#39;R1&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">bc</span><span class="p">[</span><span class="s1">&#39;R2&#39;</span><span class="p">]:</span>
                        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_barcodes</span> <span class="ow">and</span> <span class="p">(</span>
                                <span class="n">bc</span><span class="p">[</span><span class="s1">&#39;R1&#39;</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_barcodes</span><span class="p">):</span>
                            <span class="n">fates</span><span class="p">[</span><span class="s1">&#39;invalid barcode&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="k">elif</span> <span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">maximum</span><span class="p">(</span><span class="n">bc_q</span><span class="p">[</span><span class="s1">&#39;R1&#39;</span><span class="p">],</span> <span class="n">bc_q</span><span class="p">[</span><span class="s1">&#39;R2&#39;</span><span class="p">])</span>
                                <span class="o">&gt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">minq</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
                            <span class="n">barcodes</span><span class="p">[</span><span class="n">bc</span><span class="p">[</span><span class="s1">&#39;R1&#39;</span><span class="p">]]</span> <span class="o">+=</span> <span class="mi">1</span>
                            <span class="n">fates</span><span class="p">[</span><span class="s1">&#39;valid barcode&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">fates</span><span class="p">[</span><span class="s1">&#39;low quality barcode&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">fates</span><span class="p">[</span><span class="s1">&#39;R1 / R2 disagree&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># invalid flanking sequence or N in barcode</span>
                <span class="n">fates</span><span class="p">[</span><span class="s1">&#39;unparseable barcode&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="n">barcodes</span> <span class="o">=</span> <span class="p">(</span><span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
                        <span class="nb">list</span><span class="p">(</span><span class="n">barcodes</span><span class="o">.</span><span class="n">items</span><span class="p">()),</span>
                        <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">])</span>
                    <span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s1">&#39;count&#39;</span><span class="p">,</span> <span class="s1">&#39;barcode&#39;</span><span class="p">],</span>
                                 <span class="n">ascending</span><span class="o">=</span><span class="p">[</span><span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">])</span>
                    <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                    <span class="p">)</span>

        <span class="n">fates</span> <span class="o">=</span> <span class="p">(</span><span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
                    <span class="nb">list</span><span class="p">(</span><span class="n">fates</span><span class="o">.</span><span class="n">items</span><span class="p">()),</span>
                    <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;fate&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">])</span>
                 <span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s1">&#39;count&#39;</span><span class="p">,</span> <span class="s1">&#39;fate&#39;</span><span class="p">],</span>
                              <span class="n">ascending</span><span class="o">=</span><span class="p">[</span><span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">])</span>
                 <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                 <span class="p">)</span>

        <span class="k">return</span> <span class="p">(</span><span class="n">barcodes</span><span class="p">,</span> <span class="n">fates</span><span class="p">)</span></div></div>


<div class="viewcode-block" id="CodonVariantTable"><a class="viewcode-back" href="../../dms_tools2.barcodes.html#dms_tools2.barcodes.CodonVariantTable">[docs]</a><span class="k">class</span> <span class="nc">CodonVariantTable</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;Associates barcodes with codon mutants of gene.</span>

<span class="sd">    Args:</span>
<span class="sd">        `barcode_variant_file` (str)</span>
<span class="sd">            CSV file giving barcodes and variants. Must have</span>
<span class="sd">            columns named &quot;library&quot;, &quot;barcode&quot;, &quot;substitutions&quot;,</span>
<span class="sd">            (nucleotide mutations in 1, ... numbering in a format</span>
<span class="sd">            like &quot;G301A A302T G856C&quot;), and &quot;variant_call_support&quot;</span>
<span class="sd">            (sequences supporting barcode-variant call).</span>
<span class="sd">        `geneseq` (str)</span>
<span class="sd">            Sequence of protein-coding gene.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        `sites` (list)</span>
<span class="sd">            List of all codon sites in 1, 2, ... numbering.</span>
<span class="sd">        `codons` (dict)</span>
<span class="sd">            `codons[r]` is wildtype codon at site `r`.</span>
<span class="sd">        `aas` (dict)</span>
<span class="sd">            `aas[r]` is wildtype amino acid at site `r`.</span>
<span class="sd">        `libraries` (list)</span>
<span class="sd">            List of libraries in `barcode_variantfile`.</span>
<span class="sd">        `barcode_variant_df` (pandas DataFrame)</span>
<span class="sd">            Info about codon mutations parsed from `barcode_variantfile`.</span>
<span class="sd">        `variant_count_df` (pandas DataFrame or None)</span>
<span class="sd">            Initially `None`, but after data are added with</span>
<span class="sd">            :class:`CodonVariantTable.addSampleCounts`, holds</span>
<span class="sd">            counts of each variant for each sample. Differs from</span>
<span class="sd">            `barcode_variant_df` in that the former just holds</span>
<span class="sd">            initial barcode-variant data, whereas `variant_count_df`</span>
<span class="sd">            is updated with variant counts for samples.</span>

<span class="sd">    Initialize a :class:`CodonVariantTable`:</span>

<span class="sd">    &gt;&gt;&gt; geneseq = &#39;ATGGGATGA&#39;</span>
<span class="sd">    &gt;&gt;&gt; variantfile = &#39;_variantfile.csv&#39;</span>
<span class="sd">    &gt;&gt;&gt; with open(variantfile, &#39;w&#39;) as f:</span>
<span class="sd">    ...     _ = f.write(</span>
<span class="sd">    ...           &#39;library,barcode,substitutions,variant_call_support\\n&#39;</span>
<span class="sd">    ...           &#39;lib_1,AAC,,2\\n&#39;</span>
<span class="sd">    ...           &#39;lib_1,GAT,G4C A6C,1\\n&#39;</span>
<span class="sd">    ...           &#39;lib_2,AAC,T2A G4T,2\\n&#39;</span>
<span class="sd">    ...           &#39;lib_2,CAT,A6C,3&#39;</span>
<span class="sd">    ...           )</span>
<span class="sd">    &gt;&gt;&gt; variants = CodonVariantTable(</span>
<span class="sd">    ...             barcode_variant_file=variantfile,</span>
<span class="sd">    ...             geneseq=geneseq</span>
<span class="sd">    ...             )</span>
<span class="sd">    &gt;&gt;&gt; os.remove(variantfile)</span>

<span class="sd">    Check attributes of the :class:`CodonVariantTable`:</span>

<span class="sd">    &gt;&gt;&gt; variants.sites</span>
<span class="sd">    [1, 2, 3]</span>
<span class="sd">    &gt;&gt;&gt; variants.codons == {1:&#39;ATG&#39;, 2:&#39;GGA&#39;, 3:&#39;TGA&#39;}</span>
<span class="sd">    True</span>
<span class="sd">    &gt;&gt;&gt; variants.aas == {1:&#39;M&#39;, 2:&#39;G&#39;, 3:&#39;*&#39;}</span>
<span class="sd">    True</span>
<span class="sd">    &gt;&gt;&gt; variants.libraries</span>
<span class="sd">    [&#39;lib_1&#39;, &#39;lib_2&#39;]</span>
<span class="sd">    &gt;&gt;&gt; variants.valid_barcodes(&#39;lib_1&#39;) == {&#39;AAC&#39;, &#39;GAT&#39;}</span>
<span class="sd">    True</span>
<span class="sd">    &gt;&gt;&gt; variants.valid_barcodes(&#39;lib_2&#39;) == {&#39;AAC&#39;, &#39;CAT&#39;}</span>
<span class="sd">    True</span>
<span class="sd">    &gt;&gt;&gt; pandas.set_option(&#39;display.max_columns&#39;, 10)</span>
<span class="sd">    &gt;&gt;&gt; pandas.set_option(&#39;display.width&#39;, 500)</span>
<span class="sd">    &gt;&gt;&gt; variants.barcode_variant_df</span>
<span class="sd">      library barcode  variant_call_support codon_substitutions aa_substitutions  n_codon_substitutions  n_aa_substitutions</span>
<span class="sd">    0   lib_1     AAC                     2                                                           0                   0</span>
<span class="sd">    1   lib_1     GAT                     1             GGA2CGC              G2R                      1                   1</span>
<span class="sd">    2   lib_2     AAC                     2     ATG1AAG GGA2TGA          M1K G2*                      2                   2</span>
<span class="sd">    3   lib_2     CAT                     3             GGA2GGC                                       1                   0</span>

<span class="sd">    We can also look at the number of variants; we get this</span>
<span class="sd">    by calling :class:`CodonVariantTable.n_variants_df` with</span>
<span class="sd">    `samples=None` since we don&#39;t have samples, and just want</span>
<span class="sd">    stats across barcoded variants:</span>

<span class="sd">    &gt;&gt;&gt; variants.n_variants_df(samples=None)</span>
<span class="sd">             library             sample  count</span>
<span class="sd">    0          lib_1  barcoded variants      2</span>
<span class="sd">    1          lib_2  barcoded variants      2</span>
<span class="sd">    2  all libraries  barcoded variants      4</span>

<span class="sd">    We can also see how these numbers change if we require a</span>
<span class="sd">    variant call support of at least 2:</span>

<span class="sd">    &gt;&gt;&gt; variants.n_variants_df(samples=None, min_support=2)</span>
<span class="sd">             library             sample  count</span>
<span class="sd">    0          lib_1  barcoded variants      1</span>
<span class="sd">    1          lib_2  barcoded variants      2</span>
<span class="sd">    2  all libraries  barcoded variants      3</span>

<span class="sd">    If we want to combine the data for the two libraries, we can use</span>
<span class="sd">    :class:`CodonVariantTable.addMergedLibraries`, which creates a</span>
<span class="sd">    new combined library called &quot;all libraries&quot;:</span>

<span class="sd">    &gt;&gt;&gt; variants.addMergedLibraries(variants.barcode_variant_df)</span>
<span class="sd">             library    barcode  variant_call_support codon_substitutions aa_substitutions  n_codon_substitutions  n_aa_substitutions</span>
<span class="sd">    0          lib_1        AAC                     2                                                           0                   0</span>
<span class="sd">    1          lib_1        GAT                     1             GGA2CGC              G2R                      1                   1</span>
<span class="sd">    2          lib_2        AAC                     2     ATG1AAG GGA2TGA          M1K G2*                      2                   2</span>
<span class="sd">    3          lib_2        CAT                     3             GGA2GGC                                       1                   0</span>
<span class="sd">    4  all libraries  lib_1-AAC                     2                                                           0                   0</span>
<span class="sd">    5  all libraries  lib_1-GAT                     1             GGA2CGC              G2R                      1                   1</span>
<span class="sd">    6  all libraries  lib_2-AAC                     2     ATG1AAG GGA2TGA          M1K G2*                      2                   2</span>
<span class="sd">    7  all libraries  lib_2-CAT                     3             GGA2GGC                                       1                   0</span>

<span class="sd">    Note however that :class:`CodonVariantTable.addMergedLibraries`</span>
<span class="sd">    doesn&#39;t do anything if there is only one library:</span>

<span class="sd">    &gt;&gt;&gt; variants.addMergedLibraries(variants.barcode_variant_df</span>
<span class="sd">    ...                             .query(&#39;library == &quot;lib_1&quot;&#39;))</span>
<span class="sd">      library barcode  variant_call_support codon_substitutions aa_substitutions  n_codon_substitutions  n_aa_substitutions</span>
<span class="sd">    0   lib_1     AAC                     2                                                           0                   0</span>
<span class="sd">    1   lib_1     GAT                     1             GGA2CGC              G2R                      1                   1</span>

<span class="sd">    Count number of barcoded variants with each mutation:</span>

<span class="sd">    &gt;&gt;&gt; variants.mutCounts(&#39;all&#39;, &#39;aa&#39;, samples=None)[ : 2]</span>
<span class="sd">      library             sample mutation  count  mutation_type  site</span>
<span class="sd">    0   lib_1  barcoded variants      G2R      1  nonsynonymous     2</span>
<span class="sd">    1   lib_1  barcoded variants      *3A      0  nonsynonymous     3</span>

<span class="sd">    We can do the same for codon mutations (here for only a</span>
<span class="sd">    single library), first for all variants:</span>

<span class="sd">    &gt;&gt;&gt; variants.mutCounts(&#39;all&#39;, &#39;codon&#39;, samples=None,</span>
<span class="sd">    ...         libraries=[&#39;lib_2&#39;])[ : 4]</span>
<span class="sd">      library             sample mutation  count  mutation_type  site</span>
<span class="sd">    0   lib_2  barcoded variants  ATG1AAG      1  nonsynonymous     1</span>
<span class="sd">    1   lib_2  barcoded variants  GGA2GGC      1     synonymous     2</span>
<span class="sd">    2   lib_2  barcoded variants  GGA2TGA      1           stop     2</span>
<span class="sd">    3   lib_2  barcoded variants  ATG1AAA      0  nonsynonymous     1</span>

<span class="sd">    Then for just single-mutant variants:</span>

<span class="sd">    &gt;&gt;&gt; variants.mutCounts(&#39;single&#39;, &#39;codon&#39;, samples=None,</span>
<span class="sd">    ...         libraries=[&#39;lib_2&#39;])[ : 3]</span>
<span class="sd">      library             sample mutation  count  mutation_type  site</span>
<span class="sd">    0   lib_2  barcoded variants  GGA2GGC      1     synonymous     2</span>
<span class="sd">    1   lib_2  barcoded variants  ATG1AAA      0  nonsynonymous     1</span>
<span class="sd">    2   lib_2  barcoded variants  ATG1AAC      0  nonsynonymous     1</span>

<span class="sd">    Initially we haven&#39;t added any barcode count information</span>
<span class="sd">    for any samples:</span>

<span class="sd">    &gt;&gt;&gt; all(variants.samples(lib) == [] for lib in variants.libraries)</span>
<span class="sd">    True</span>
<span class="sd">    &gt;&gt;&gt; variants.variant_count_df is None</span>
<span class="sd">    True</span>

<span class="sd">    Now we add barcode count information for a sample named</span>
<span class="sd">    &quot;input&quot; from library 1:</span>

<span class="sd">    &gt;&gt;&gt; counts_lib1_input = pandas.DataFrame(</span>
<span class="sd">    ...         {&#39;barcode&#39;:[&#39;AAC&#39;, &#39;GAT&#39;],</span>
<span class="sd">    ...          &#39;count&#39;  :[  253,  1101]})</span>
<span class="sd">    &gt;&gt;&gt; variants.addSampleCounts(&#39;lib_1&#39;, &#39;input&#39;, counts_lib1_input)</span>
<span class="sd">    &gt;&gt;&gt; variants.variant_count_df</span>
<span class="sd">      barcode  count library sample  variant_call_support codon_substitutions aa_substitutions  n_codon_substitutions  n_aa_substitutions</span>
<span class="sd">    0     GAT   1101   lib_1  input                     1             GGA2CGC              G2R                      1                   1</span>
<span class="sd">    1     AAC    253   lib_1  input                     2                                                           0                   0</span>

<span class="sd">    We get an error if we try to add these same data again,</span>
<span class="sd">    as they are already added for that sample to that library:</span>

<span class="sd">    &gt;&gt;&gt; variants.addSampleCounts(&#39;lib_1&#39;, &#39;input&#39;, counts_lib1_input)</span>
<span class="sd">    Traceback (most recent call last):</span>
<span class="sd">    ...</span>
<span class="sd">    ValueError: `library` lib_1 already has `sample` input</span>

<span class="sd">    But, we can add barcode counts for another</span>
<span class="sd">    sample (named &quot;selected&quot; in this case) to library 1:</span>

<span class="sd">    &gt;&gt;&gt; counts_lib1_selected = pandas.DataFrame(</span>
<span class="sd">    ...         {&#39;barcode&#39;:[&#39;AAC&#39;, &#39;GAT&#39;],</span>
<span class="sd">    ...          &#39;count&#39;  :[  513,  401]})</span>
<span class="sd">    &gt;&gt;&gt; variants.addSampleCounts(&#39;lib_1&#39;, &#39;selected&#39;, counts_lib1_selected)</span>

<span class="sd">    As well as barcode counts for the same two samples</span>
<span class="sd">    (&quot;input&quot; and &quot;selected&quot;) to our other library (library 2):</span>

<span class="sd">    &gt;&gt;&gt; counts_lib2_input = pandas.DataFrame(</span>
<span class="sd">    ...         {&#39;barcode&#39;:[&#39;AAC&#39;, &#39;CAT&#39;],</span>
<span class="sd">    ...          &#39;count&#39;  :[ 1253,  923]})</span>
<span class="sd">    &gt;&gt;&gt; variants.addSampleCounts(&#39;lib_2&#39;, &#39;input&#39;, counts_lib2_input)</span>
<span class="sd">    &gt;&gt;&gt; counts_lib2_selected = pandas.DataFrame(</span>
<span class="sd">    ...         {&#39;barcode&#39;:[&#39;AAC&#39;, &#39;CAT&#39;],</span>
<span class="sd">    ...          &#39;count&#39;  :[  113,  1200]})</span>
<span class="sd">    &gt;&gt;&gt; variants.addSampleCounts(&#39;lib_2&#39;, &#39;selected&#39;, counts_lib2_selected)</span>
<span class="sd">    &gt;&gt;&gt; variants.variant_count_df</span>
<span class="sd">      barcode  count library    sample  variant_call_support codon_substitutions aa_substitutions  n_codon_substitutions  n_aa_substitutions</span>
<span class="sd">    0     GAT   1101   lib_1     input                     1             GGA2CGC              G2R                      1                   1</span>
<span class="sd">    1     AAC    253   lib_1     input                     2                                                           0                   0</span>
<span class="sd">    2     AAC    513   lib_1  selected                     2                                                           0                   0</span>
<span class="sd">    3     GAT    401   lib_1  selected                     1             GGA2CGC              G2R                      1                   1</span>
<span class="sd">    4     AAC   1253   lib_2     input                     2     ATG1AAG GGA2TGA          M1K G2*                      2                   2</span>
<span class="sd">    5     CAT    923   lib_2     input                     3             GGA2GGC                                       1                   0</span>
<span class="sd">    6     CAT   1200   lib_2  selected                     3             GGA2GGC                                       1                   0</span>
<span class="sd">    7     AAC    113   lib_2  selected                     2     ATG1AAG GGA2TGA          M1K G2*                      2                   2</span>

<span class="sd">    We can also use :class:`CodonVariantTable.mutCounts`</span>
<span class="sd">    to look at total counts of each mutation:</span>

<span class="sd">    &gt;&gt;&gt; variants.mutCounts(&#39;all&#39;, &#39;aa&#39;)[ : 2]</span>
<span class="sd">      library sample mutation  count  mutation_type  site</span>
<span class="sd">    0   lib_1  input      G2R   1101  nonsynonymous     2</span>
<span class="sd">    1   lib_1  input      *3A      0  nonsynonymous     3</span>

<span class="sd">    &gt;&gt;&gt; variants.mutCounts(&#39;all&#39;, &#39;aa&#39;, libraries=[&#39;lib_2&#39;])[ : 3]</span>
<span class="sd">      library sample mutation  count  mutation_type  site</span>
<span class="sd">    0   lib_2  input      G2*   1253           stop     2</span>
<span class="sd">    1   lib_2  input      M1K   1253  nonsynonymous     1</span>
<span class="sd">    2   lib_2  input      *3A      0  nonsynonymous     3</span>

<span class="sd">    We can use :class:`CodonVariantTable.writeCodonCounts` to</span>
<span class="sd">    write codon count files. First for only **single** mutants:</span>

<span class="sd">    &gt;&gt;&gt; with tempfile.TemporaryDirectory() as tmpdir:</span>
<span class="sd">    ...     countfiles = variants.writeCodonCounts(&quot;single&quot;, outdir=tmpdir)</span>
<span class="sd">    ...     lib1_input = pandas.read_csv(f&#39;{tmpdir}/lib_1_input_codoncounts.csv&#39;)</span>
<span class="sd">    ...     all_sel = pandas.read_csv(f&#39;{tmpdir}/all-libraries_selected_codoncounts.csv&#39;)</span>

<span class="sd">    Make sure we created expected countfiles:</span>

<span class="sd">    &gt;&gt;&gt; countfiles.assign(countfile=lambda x: x.countfile.apply(os.path.basename))</span>
<span class="sd">             library    sample                               countfile</span>
<span class="sd">    0          lib_1     input             lib_1_input_codoncounts.csv</span>
<span class="sd">    1          lib_1  selected          lib_1_selected_codoncounts.csv</span>
<span class="sd">    2          lib_2     input             lib_2_input_codoncounts.csv</span>
<span class="sd">    3          lib_2  selected          lib_2_selected_codoncounts.csv</span>
<span class="sd">    4  all-libraries     input     all-libraries_input_codoncounts.csv</span>
<span class="sd">    5  all-libraries  selected  all-libraries_selected_codoncounts.csv</span>

<span class="sd">    Check for expected values in a few of these counts files, only</span>
<span class="sd">    showing columns with non-zero entries:</span>

<span class="sd">    &gt;&gt;&gt; lib1_input.iloc[:, (lib1_input != 0).any(axis=&#39;rows&#39;).values]</span>
<span class="sd">       site wildtype  ATG   CGC  GGA  TGA</span>
<span class="sd">    0     1      ATG  253     0    0    0</span>
<span class="sd">    1     2      GGA    0  1101  253    0</span>
<span class="sd">    2     3      TGA    0     0    0  253</span>
<span class="sd">    &gt;&gt;&gt; all_sel.iloc[:, (all_sel != 0).any(axis=&#39;rows&#39;).values]</span>
<span class="sd">       site wildtype  ATG  CGC  GGA   GGC  TGA</span>
<span class="sd">    0     1      ATG  513    0    0     0    0</span>
<span class="sd">    1     2      GGA    0  401  513  1200    0</span>
<span class="sd">    2     3      TGA    0    0    0     0  513</span>

<span class="sd">    Now write codon counts files for **all** mutants:</span>

<span class="sd">    &gt;&gt;&gt; with tempfile.TemporaryDirectory() as tmpdir:</span>
<span class="sd">    ...     _ = variants.writeCodonCounts(&quot;all&quot;, outdir=tmpdir)</span>
<span class="sd">    ...     lib1_input_all = pandas.read_csv(f&#39;{tmpdir}/lib_1_input_codoncounts.csv&#39;)</span>
<span class="sd">    ...     all_sel_all = pandas.read_csv(f&#39;{tmpdir}/all-libraries_selected_codoncounts.csv&#39;)</span>
<span class="sd">    &gt;&gt;&gt; lib1_input_all.iloc[:, (lib1_input_all != 0).any(axis=&#39;rows&#39;).values]</span>
<span class="sd">       site wildtype   ATG   CGC  GGA   TGA</span>
<span class="sd">    0     1      ATG  1354     0    0     0</span>
<span class="sd">    1     2      GGA     0  1101  253     0</span>
<span class="sd">    2     3      TGA     0     0    0  1354</span>
<span class="sd">    &gt;&gt;&gt; all_sel_all.iloc[:, (all_sel_all != 0).any(axis=&#39;rows&#39;).values]</span>
<span class="sd">       site wildtype  AAG   ATG  CGC  GGA   GGC   TGA</span>
<span class="sd">    0     1      ATG  113  2114    0    0     0     0</span>
<span class="sd">    1     2      GGA    0     0  401  513  1200   113</span>
<span class="sd">    2     3      TGA    0     0    0    0     0  2227</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">barcode_variant_file</span><span class="p">,</span> <span class="n">geneseq</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;See main class doc string.&quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">geneseq</span> <span class="o">=</span> <span class="n">geneseq</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s1">&#39;^[ATGC]+$&#39;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">geneseq</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid nucleotides in </span><span class="si">{self.geneseq}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">geneseq</span><span class="p">)</span> <span class="o">%</span> <span class="mi">3</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">or</span> <span class="nb">len</span><span class="p">(</span><span class="n">geneseq</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;`geneseq` of invalid length {len(self.geneseq)}&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sites</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">geneseq</span><span class="p">)</span> <span class="o">//</span> <span class="mi">3</span> <span class="o">+</span> <span class="mi">1</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">codons</span> <span class="o">=</span> <span class="p">{</span><span class="n">r</span><span class="p">:</span><span class="bp">self</span><span class="o">.</span><span class="n">geneseq</span><span class="p">[</span><span class="mi">3</span> <span class="o">*</span> <span class="p">(</span><span class="n">r</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="p">:</span> <span class="mi">3</span> <span class="o">*</span> <span class="n">r</span><span class="p">]</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sites</span><span class="p">}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">aas</span> <span class="o">=</span> <span class="p">{</span><span class="n">r</span><span class="p">:</span><span class="n">CODON_TO_AA</span><span class="p">[</span><span class="n">codon</span><span class="p">]</span> <span class="k">for</span> <span class="n">r</span><span class="p">,</span> <span class="n">codon</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">codons</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>

        <span class="n">df</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">barcode_variant_file</span><span class="p">)</span>
        <span class="n">required_cols</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;barcode&#39;</span><span class="p">,</span>
                         <span class="s1">&#39;substitutions&#39;</span><span class="p">,</span> <span class="s1">&#39;variant_call_support&#39;</span><span class="p">}</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">set</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span><span class="o">.</span><span class="n">issuperset</span><span class="p">(</span><span class="n">required_cols</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;`variantfile` does not have &quot;</span>
                             <span class="n">f</span><span class="s2">&quot;required columns </span><span class="si">{required_cols}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">libraries</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">library</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_valid_barcodes</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">lib</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">libraries</span><span class="p">:</span>
            <span class="n">barcodes</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;library == @lib&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">barcode</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">barcodes</span><span class="p">))</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">barcodes</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;duplicated barcodes for </span><span class="si">{lib}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_valid_barcodes</span><span class="p">[</span><span class="n">lib</span><span class="p">]</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">barcodes</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_samples</span> <span class="o">=</span> <span class="p">{</span><span class="n">lib</span><span class="p">:[]</span> <span class="k">for</span> <span class="n">lib</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">libraries</span><span class="p">}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">variant_count_df</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">barcode_variant_df</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">df</span>
                <span class="c1"># info about codon and amino-acid substitutions</span>
                <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">codon_substitutions</span><span class="o">=</span>
                            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">substitutions</span>
                                       <span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
                                       <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_ntToCodonMuts</span><span class="p">),</span>
                        <span class="n">aa_substitutions</span><span class="o">=</span>
                            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">codon_substitutions</span>
                                       <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_codonToAAMuts</span><span class="p">),</span>
                        <span class="n">n_codon_substitutions</span><span class="o">=</span>
                            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">codon_substitutions</span>
                                       <span class="o">.</span><span class="n">str</span>
                                       <span class="o">.</span><span class="n">split</span><span class="p">()</span>
                                       <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="nb">len</span><span class="p">),</span>
                        <span class="n">n_aa_substitutions</span><span class="o">=</span>
                            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">aa_substitutions</span>
                                       <span class="o">.</span><span class="n">str</span>
                                       <span class="o">.</span><span class="n">split</span><span class="p">()</span>
                                       <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="nb">len</span><span class="p">)</span>
                        <span class="p">)</span>
                <span class="c1"># we no longer need initial `substitutions` column</span>
                <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s1">&#39;substitutions&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="s1">&#39;columns&#39;</span><span class="p">)</span>
                <span class="c1"># sort to ensure consistent order</span>
                <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">library</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                                <span class="n">pandas</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
                                    <span class="n">x</span><span class="o">.</span><span class="n">library</span><span class="p">,</span>
                                    <span class="n">categories</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">libraries</span><span class="p">,</span>
                                    <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span>
                                    <span class="p">)</span>
                        <span class="p">)</span>
                <span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;barcode&#39;</span><span class="p">])</span>
                <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                <span class="p">)</span>

        <span class="c1"># define some colors for plotting</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_mutation_type_colors</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s1">&#39;nonsynonymous&#39;</span><span class="p">:</span><span class="n">COLOR_BLIND_PALETTE_GRAY</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                <span class="s1">&#39;synonymous&#39;</span><span class="p">:</span><span class="n">COLOR_BLIND_PALETTE_GRAY</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span>
                <span class="s1">&#39;stop&#39;</span><span class="p">:</span><span class="n">COLOR_BLIND_PALETTE_GRAY</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
                <span class="p">}</span>


<div class="viewcode-block" id="CodonVariantTable.samples"><a class="viewcode-back" href="../../dms_tools2.barcodes.html#dms_tools2.barcodes.CodonVariantTable.samples">[docs]</a>    <span class="k">def</span> <span class="nf">samples</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">library</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;List of all samples for `library`.</span>

<span class="sd">        Args:</span>
<span class="sd">            `library` (str)</span>
<span class="sd">                Valid `library` for the :class:`CodonVariantTable`.</span>

<span class="sd">        Returns:</span>
<span class="sd">            List of all samples for which barcode counts have</span>
<span class="sd">            been added.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_samples</span><span class="p">[</span><span class="n">library</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid `library` </span><span class="si">{library}</span><span class="s2">&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="CodonVariantTable.addSampleCounts"><a class="viewcode-back" href="../../dms_tools2.barcodes.html#dms_tools2.barcodes.CodonVariantTable.addSampleCounts">[docs]</a>    <span class="k">def</span> <span class="nf">addSampleCounts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">library</span><span class="p">,</span> <span class="n">sample</span><span class="p">,</span> <span class="n">barcodecounts</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Add variant counts for a sample to `variant_count_df`.</span>

<span class="sd">        Args:</span>
<span class="sd">            `library` (str)</span>
<span class="sd">                Valid `library` for the :class:`CodonVariantTable`.</span>
<span class="sd">            `sample` (str)</span>
<span class="sd">                Sample name, must **not** already be in</span>
<span class="sd">                :class:`CodonVariantTable.samples` for `library`.</span>
<span class="sd">            `barcodecounts` (pandas DataFrame)</span>
<span class="sd">                Gives counts for each variant by barcode. Must</span>
<span class="sd">                have columns named &quot;barcode&quot; and &quot;count&quot;. The</span>
<span class="sd">                &quot;barcode&quot; column must contain all the barcodes</span>
<span class="sd">                in :class:`CodonVariantTable.valid_barcodes` for</span>
<span class="sd">                `library`. Such data frames are returned</span>
<span class="sd">                by :class:`IlluminaBarcodeParser.parse`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">library</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">libraries</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid library </span><span class="si">{library}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">sample</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">samples</span><span class="p">(</span><span class="n">library</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;`library` </span><span class="si">{library}</span><span class="s2"> already &quot;</span>
                             <span class="n">f</span><span class="s2">&quot;has `sample` </span><span class="si">{sample}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">req_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">set</span><span class="p">(</span><span class="n">barcodecounts</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span><span class="o">.</span><span class="n">issuperset</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">req_cols</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;`barcodecounts` lacks columns </span><span class="si">{req_cols}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">barcodecounts</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">barcodecounts</span><span class="o">.</span><span class="n">barcode</span><span class="o">.</span><span class="n">unique</span><span class="p">())):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;`barcodecounts` has non-unique barcodes&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">set</span><span class="p">(</span><span class="n">barcodecounts</span><span class="o">.</span><span class="n">barcode</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_barcodes</span><span class="p">(</span><span class="n">library</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;barcodes in `barcodecounts` do not match &quot;</span>
                             <span class="n">f</span><span class="s2">&quot;those expected for `library` </span><span class="si">{library}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_samples</span><span class="p">[</span><span class="n">library</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>

        <span class="n">df</span> <span class="o">=</span> <span class="p">(</span><span class="n">barcodecounts</span>
              <span class="p">[</span><span class="n">req_cols</span><span class="p">]</span>
              <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">library</span><span class="o">=</span><span class="n">library</span><span class="p">,</span> <span class="n">sample</span><span class="o">=</span><span class="n">sample</span><span class="p">)</span>
              <span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">barcode_variant_df</span><span class="p">,</span>
                     <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span><span class="p">,</span>
                     <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;barcode&#39;</span><span class="p">],</span>
                     <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                     <span class="n">validate</span><span class="o">=</span><span class="s1">&#39;one_to_one&#39;</span><span class="p">)</span>
              <span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">variant_count_df</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">variant_count_df</span> <span class="o">=</span> <span class="n">df</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">variant_count_df</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
                              <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">variant_count_df</span><span class="p">,</span> <span class="n">df</span><span class="p">],</span>
                              <span class="n">axis</span><span class="o">=</span><span class="s1">&#39;index&#39;</span><span class="p">,</span>
                              <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                              <span class="n">sort</span><span class="o">=</span><span class="kc">False</span>
                              <span class="p">)</span>

        <span class="c1"># samples in order added after ordering by library, getting</span>
        <span class="c1"># unique ones as here: https://stackoverflow.com/a/39835527</span>
        <span class="n">unique_samples</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">collections</span><span class="o">.</span><span class="n">OrderedDict</span><span class="o">.</span><span class="n">fromkeys</span><span class="p">(</span>
                <span class="n">itertools</span><span class="o">.</span><span class="n">chain</span><span class="o">.</span><span class="n">from_iterable</span><span class="p">(</span>
                    <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">samples</span><span class="p">(</span><span class="n">lib</span><span class="p">)</span> <span class="k">for</span> <span class="n">lib</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">libraries</span><span class="p">])</span>
                <span class="p">))</span>

        <span class="c1"># make library and sample categorical and sort</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">variant_count_df</span> <span class="o">=</span> <span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">variant_count_df</span>
                <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">library</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                                <span class="n">pandas</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
                                    <span class="n">x</span><span class="o">.</span><span class="n">library</span><span class="p">,</span>
                                    <span class="n">categories</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">libraries</span><span class="p">,</span>
                                    <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span>
                                    <span class="p">),</span>
                        <span class="n">sample</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                               <span class="n">pandas</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
                                    <span class="n">x</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">],</span>
                                    <span class="n">categories</span><span class="o">=</span><span class="n">unique_samples</span><span class="p">,</span>
                                    <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span>
                                    <span class="p">)</span>
                         <span class="p">)</span>
                <span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">],</span>
                             <span class="n">ascending</span><span class="o">=</span><span class="p">[</span><span class="kc">True</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">])</span>
                <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                <span class="p">)</span></div>


<div class="viewcode-block" id="CodonVariantTable.valid_barcodes"><a class="viewcode-back" href="../../dms_tools2.barcodes.html#dms_tools2.barcodes.CodonVariantTable.valid_barcodes">[docs]</a>    <span class="k">def</span> <span class="nf">valid_barcodes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">library</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Set of valid barcodes for `library`.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">library</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">libraries</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid `library` </span><span class="si">{library}</span><span class="s2">; &quot;</span>
                             <span class="n">f</span><span class="s2">&quot;valid libraries are </span><span class="si">{self.libraries}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_valid_barcodes</span><span class="p">[</span><span class="n">library</span><span class="p">]</span></div>


<div class="viewcode-block" id="CodonVariantTable.n_variants_df"><a class="viewcode-back" href="../../dms_tools2.barcodes.html#dms_tools2.barcodes.CodonVariantTable.n_variants_df">[docs]</a>    <span class="k">def</span> <span class="nf">n_variants_df</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">libraries</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span>
                      <span class="n">min_support</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Number of variants per library / sample.</span>

<span class="sd">        Args:</span>
<span class="sd">            Same meaning as for</span>
<span class="sd">            :class:`CodonVariantTable.plotNumMutsHistogram`.</span>

<span class="sd">        Returns:</span>
<span class="sd">            DataFrame giving number of variants per library /</span>
<span class="sd">            sample.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span><span class="p">,</span> <span class="n">nlibraries</span><span class="p">,</span> <span class="n">nsamples</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_getPlotData</span><span class="p">(</span><span class="n">libraries</span><span class="p">,</span>
                                                     <span class="n">samples</span><span class="p">,</span>
                                                     <span class="n">min_support</span><span class="p">)</span>

        <span class="k">return</span> <span class="p">(</span><span class="n">df</span>
                <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">])</span>
                <span class="o">.</span><span class="n">aggregate</span><span class="p">({</span><span class="s1">&#39;count&#39;</span><span class="p">:</span><span class="s1">&#39;sum&#39;</span><span class="p">})</span>
                <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
                <span class="p">)</span></div>


<div class="viewcode-block" id="CodonVariantTable.mutCounts"><a class="viewcode-back" href="../../dms_tools2.barcodes.html#dms_tools2.barcodes.CodonVariantTable.mutCounts">[docs]</a>    <span class="k">def</span> <span class="nf">mutCounts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">variant_type</span><span class="p">,</span> <span class="n">mut_type</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span>
            <span class="n">libraries</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">min_support</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get counts of each individual mutation.</span>

<span class="sd">        Args:</span>
<span class="sd">            `variant_type` (&quot;single&quot; or &quot;all&quot;)</span>
<span class="sd">                Include just single mutants, or all mutants?</span>
<span class="sd">            Other args:</span>
<span class="sd">                Same meaning as for</span>
<span class="sd">                :class:`CodonVariantTable.plotNumMutsHistogram`</span>

<span class="sd">        Returns:</span>
<span class="sd">            A tidy data frame with columns named &quot;library&quot;,</span>
<span class="sd">            &quot;sample&quot;, &quot;mutation&quot;, &quot;count&quot;, &quot;mutation_type&quot;,</span>
<span class="sd">            and &quot;site&quot;.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span><span class="p">,</span> <span class="n">nlibraries</span><span class="p">,</span> <span class="n">nsamples</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_getPlotData</span><span class="p">(</span><span class="n">libraries</span><span class="p">,</span>
                                                     <span class="n">samples</span><span class="p">,</span>
                                                     <span class="n">min_support</span><span class="p">)</span>

        <span class="n">samplelist</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="n">librarylist</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">mut_type</span> <span class="o">==</span> <span class="s1">&#39;codon&#39;</span><span class="p">:</span>
            <span class="n">wts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">codons</span>
            <span class="n">chars</span> <span class="o">=</span> <span class="n">CODONS</span>
            <span class="n">mutation_types</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;nonsynonymous&#39;</span><span class="p">,</span> <span class="s1">&#39;synonymous&#39;</span><span class="p">,</span> <span class="s1">&#39;stop&#39;</span><span class="p">]</span>
        <span class="k">elif</span> <span class="n">mut_type</span> <span class="o">==</span> <span class="s1">&#39;aa&#39;</span><span class="p">:</span>
            <span class="n">wts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">aas</span>
            <span class="n">chars</span> <span class="o">=</span> <span class="n">AAS_WITHSTOP</span>
            <span class="n">mutation_types</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;nonsynonymous&#39;</span><span class="p">,</span> <span class="s1">&#39;stop&#39;</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid mut_type </span><span class="si">{mut_type}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># data frame listing all mutations with count 0</span>
        <span class="n">mut_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">r</span><span class="p">,</span> <span class="n">wt</span> <span class="ow">in</span> <span class="n">wts</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">chars</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">mut</span> <span class="o">!=</span> <span class="n">wt</span><span class="p">:</span>
                    <span class="n">mut_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{wt}{r}{mut}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">all_muts</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span>
                    <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;mutation&#39;</span><span class="p">:</span><span class="n">mut_list</span><span class="p">,</span>
                                     <span class="s1">&#39;library&#39;</span><span class="p">:</span><span class="n">library</span><span class="p">,</span>
                                     <span class="s1">&#39;sample&#39;</span><span class="p">:</span><span class="n">sample</span><span class="p">,</span>
                                     <span class="s1">&#39;count&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">})</span>
                    <span class="k">for</span> <span class="n">library</span><span class="p">,</span> <span class="n">sample</span> <span class="ow">in</span>
                    <span class="n">itertools</span><span class="o">.</span><span class="n">product</span><span class="p">(</span><span class="n">librarylist</span><span class="p">,</span> <span class="n">samplelist</span><span class="p">)])</span>

        <span class="k">if</span> <span class="n">variant_type</span> <span class="o">==</span> <span class="s1">&#39;single&#39;</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;n_</span><span class="si">{mut_type}</span><span class="s1">_substitutions == 1&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">variant_type</span> <span class="o">==</span> <span class="s1">&#39;all&#39;</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;n_</span><span class="si">{mut_type}</span><span class="s1">_substitutions &gt;= 1&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid variant_type </span><span class="si">{variant_type}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">def</span> <span class="nf">_classify_mutation</span><span class="p">(</span><span class="n">mut_str</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">mut_type</span> <span class="o">==</span> <span class="s1">&#39;aa&#39;</span><span class="p">:</span>
                <span class="n">m</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s1">&#39;^(?P&lt;wt&gt;[A-Z\*])\d+(?P&lt;mut&gt;[A-Z\*])$&#39;</span><span class="p">,</span>
                             <span class="n">mut_str</span><span class="p">)</span>
                <span class="n">wt_aa</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;wt&#39;</span><span class="p">)</span>
                <span class="n">mut_aa</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;mut&#39;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">m</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s1">&#39;^(?P&lt;wt&gt;[ACTG]</span><span class="si">{3}</span><span class="s1">)\d+(?P&lt;mut&gt;[ACTG]</span><span class="si">{3}</span><span class="s1">)$&#39;</span><span class="p">,</span>
                             <span class="n">mut_str</span><span class="p">)</span>
                <span class="n">wt_aa</span> <span class="o">=</span> <span class="n">CODON_TO_AA</span><span class="p">[</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;wt&#39;</span><span class="p">)]</span>
                <span class="n">mut_aa</span> <span class="o">=</span> <span class="n">CODON_TO_AA</span><span class="p">[</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;mut&#39;</span><span class="p">)]</span>
            <span class="k">if</span> <span class="n">wt_aa</span> <span class="o">==</span> <span class="n">mut_aa</span><span class="p">:</span>
                <span class="k">return</span> <span class="s1">&#39;synonymous&#39;</span>
            <span class="k">elif</span> <span class="n">mut_aa</span> <span class="o">==</span> <span class="s1">&#39;*&#39;</span><span class="p">:</span>
                <span class="k">return</span> <span class="s1">&#39;stop&#39;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">return</span> <span class="s1">&#39;nonsynonymous&#39;</span>

        <span class="k">def</span> <span class="nf">_get_site</span><span class="p">(</span><span class="n">mut_str</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">mut_type</span> <span class="o">==</span> <span class="s1">&#39;aa&#39;</span><span class="p">:</span>
                <span class="n">m</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s1">&#39;^[A-Z\*](?P&lt;site&gt;\d+)[A-Z\*]$&#39;</span><span class="p">,</span> <span class="n">mut_str</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">m</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s1">&#39;^[ACTG]</span><span class="si">{3}</span><span class="s1">(?P&lt;site&gt;\d+)[ACTG]</span><span class="si">{3}</span><span class="s1">$&#39;</span><span class="p">,</span> <span class="n">mut_str</span><span class="p">)</span>
            <span class="n">site</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;site&#39;</span><span class="p">))</span>
            <span class="k">assert</span> <span class="n">site</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sites</span>
            <span class="k">return</span> <span class="n">site</span>

        <span class="n">df</span> <span class="o">=</span> <span class="p">(</span><span class="n">df</span>
              <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span>
                  <span class="p">{</span><span class="n">f</span><span class="s2">&quot;</span><span class="si">{mut_type}</span><span class="s2">_substitutions&quot;</span><span class="p">:</span><span class="s2">&quot;mutation&quot;</span><span class="p">}</span>
                  <span class="p">)</span>
              <span class="p">[[</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;mutation&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">]]</span>
              <span class="o">.</span><span class="n">pipe</span><span class="p">(</span><span class="n">tidy_split</span><span class="p">,</span> <span class="n">column</span><span class="o">=</span><span class="s1">&#39;mutation&#39;</span><span class="p">)</span>
              <span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">all_muts</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;outer&#39;</span><span class="p">)</span>
              <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;mutation&#39;</span><span class="p">])</span>
              <span class="o">.</span><span class="n">aggregate</span><span class="p">({</span><span class="s1">&#39;count&#39;</span><span class="p">:</span><span class="s1">&#39;sum&#39;</span><span class="p">})</span>
              <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
              <span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                <span class="n">library</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                         <span class="n">pandas</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
                          <span class="n">x</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">],</span>
                          <span class="n">librarylist</span><span class="p">,</span>
                          <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
                <span class="n">sample</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                         <span class="n">pandas</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
                          <span class="n">x</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">],</span>
                          <span class="n">samplelist</span><span class="p">,</span>
                          <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
                <span class="n">mutation_type</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                         <span class="n">pandas</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
                          <span class="n">x</span><span class="p">[</span><span class="s1">&#39;mutation&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">_classify_mutation</span><span class="p">),</span>
                          <span class="n">mutation_types</span><span class="p">,</span>
                          <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
                <span class="n">site</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;mutation&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">_get_site</span><span class="p">),</span>
                <span class="p">)</span>
              <span class="o">.</span><span class="n">sort_values</span><span class="p">(</span>
                <span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">,</span> <span class="s1">&#39;mutation&#39;</span><span class="p">],</span>
                <span class="n">ascending</span><span class="o">=</span><span class="p">[</span><span class="kc">True</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">])</span>
              <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
              <span class="p">)</span>

        <span class="k">return</span> <span class="n">df</span></div>


<div class="viewcode-block" id="CodonVariantTable.plotMutHeatmap"><a class="viewcode-back" href="../../dms_tools2.barcodes.html#dms_tools2.barcodes.CodonVariantTable.plotMutHeatmap">[docs]</a>    <span class="k">def</span> <span class="nf">plotMutHeatmap</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">variant_type</span><span class="p">,</span> <span class="n">mut_type</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span>
            <span class="n">count_or_frequency</span><span class="o">=</span><span class="s1">&#39;frequency&#39;</span><span class="p">,</span>
            <span class="n">libraries</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">plotfile</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">orientation</span><span class="o">=</span><span class="s1">&#39;h&#39;</span><span class="p">,</span> <span class="n">widthscale</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">heightscale</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
            <span class="n">min_support</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Heatmap of mutation counts or frequencies.</span>

<span class="sd">        Args:</span>
<span class="sd">            `count_or_frequency` (&quot;count&quot; or &quot;frequency&quot;)</span>
<span class="sd">                Plot mutation counts or frequencies?</span>
<span class="sd">            All other args</span>
<span class="sd">                Same meaning as for</span>
<span class="sd">                :meth:`CodonVariantTable.plotCumulMutCoverage`.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A `plotnine &lt;https://plotnine.readthedocs.io/en/stable/&gt;`_</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutCounts</span><span class="p">(</span><span class="n">variant_type</span><span class="p">,</span> <span class="n">mut_type</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="n">samples</span><span class="p">,</span>
                            <span class="n">libraries</span><span class="o">=</span><span class="n">libraries</span><span class="p">,</span> <span class="n">min_support</span><span class="o">=</span><span class="n">min_support</span><span class="p">)</span>

        <span class="n">n_variants</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">n_variants_df</span><span class="p">(</span><span class="n">libraries</span><span class="o">=</span><span class="n">libraries</span><span class="p">,</span>
                                         <span class="n">samples</span><span class="o">=</span><span class="n">samples</span><span class="p">,</span>
                                         <span class="n">min_support</span><span class="o">=</span><span class="n">min_support</span><span class="p">)</span>
                      <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;count&#39;</span><span class="p">:</span><span class="s1">&#39;nseqs&#39;</span><span class="p">})</span>
                      <span class="p">)</span>

        <span class="c1"># order amino acids by Kyte-Doolittle hydrophobicity,</span>
        <span class="c1"># codons by the amino acid they encode</span>
        <span class="n">aa_order</span> <span class="o">=</span> <span class="p">[</span><span class="n">tup</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">tup</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span>
                    <span class="n">Bio</span><span class="o">.</span><span class="n">SeqUtils</span><span class="o">.</span><span class="n">ProtParamData</span><span class="o">.</span><span class="n">kd</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span>
                    <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">tup</span><span class="p">:</span> <span class="n">tup</span><span class="p">[</span><span class="mi">1</span><span class="p">])]</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;*&#39;</span><span class="p">]</span>
        <span class="n">codon_order</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">itertools</span><span class="o">.</span><span class="n">chain</span><span class="o">.</span><span class="n">from_iterable</span><span class="p">(</span>
                        <span class="p">[</span><span class="n">AA_TO_CODONS</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">aa_order</span><span class="p">]))</span>

        <span class="n">df</span> <span class="o">=</span> <span class="p">(</span><span class="n">df</span>
              <span class="p">[[</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;mutation&#39;</span><span class="p">,</span> <span class="s1">&#39;site&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">]]</span>
              <span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">n_variants</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">])</span>
              <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">frequency</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;count&#39;</span><span class="p">]</span> <span class="o">/</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;nseqs&#39;</span><span class="p">],</span>
                      <span class="n">mut_char</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                        <span class="n">pandas</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
                         <span class="n">x</span><span class="o">.</span><span class="n">mutation</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span>
                            <span class="s1">&#39;^[A-Z\*]+\d+(?P&lt;mut_char&gt;[A-Z\*]+)$&#39;</span><span class="p">)</span>
                            <span class="o">.</span><span class="n">mut_char</span><span class="p">,</span>
                         <span class="p">{</span><span class="s1">&#39;aa&#39;</span><span class="p">:</span><span class="n">aa_order</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">:</span><span class="n">codon_order</span><span class="p">}[</span><span class="n">mut_type</span><span class="p">],</span>
                         <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                      <span class="p">)</span>
              <span class="p">)</span>

        <span class="k">if</span> <span class="n">count_or_frequency</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">{</span><span class="s1">&#39;count&#39;</span><span class="p">,</span> <span class="s1">&#39;frequency&#39;</span><span class="p">}:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid count_or_frequency &quot;</span>
                             <span class="n">f</span><span class="s2">&quot;</span><span class="si">{count_or_frequency}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">nlibraries</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>
        <span class="n">nsamples</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>

        <span class="k">if</span> <span class="n">mut_type</span> <span class="o">==</span> <span class="s1">&#39;codon&#39;</span><span class="p">:</span>
            <span class="n">height_per</span> <span class="o">=</span> <span class="mf">5.5</span>
            <span class="n">mut_desc</span> <span class="o">=</span> <span class="s1">&#39;codon&#39;</span>
        <span class="k">elif</span> <span class="n">mut_type</span> <span class="o">==</span> <span class="s1">&#39;aa&#39;</span><span class="p">:</span>
            <span class="n">height_per</span> <span class="o">=</span> <span class="mf">1.7</span>
            <span class="n">mut_desc</span> <span class="o">=</span> <span class="s1">&#39;amino acid&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid mut_type </span><span class="si">{mut_type}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">orientation</span> <span class="o">==</span> <span class="s1">&#39;h&#39;</span><span class="p">:</span>
            <span class="n">facet_str</span> <span class="o">=</span> <span class="s1">&#39;sample ~ library&#39;</span>
            <span class="n">width</span> <span class="o">=</span> <span class="n">widthscale</span> <span class="o">*</span> <span class="p">(</span><span class="mf">1.6</span> <span class="o">+</span> <span class="mf">3.5</span> <span class="o">*</span> <span class="n">nlibraries</span><span class="p">)</span>
            <span class="n">height</span> <span class="o">=</span> <span class="n">heightscale</span> <span class="o">*</span> <span class="p">(</span><span class="mf">0.8</span> <span class="o">+</span> <span class="n">height_per</span> <span class="o">*</span> <span class="n">nsamples</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">orientation</span> <span class="o">==</span> <span class="s1">&#39;v&#39;</span><span class="p">:</span>
            <span class="n">facet_str</span> <span class="o">=</span> <span class="s1">&#39;library ~ sample&#39;</span>
            <span class="n">width</span> <span class="o">=</span> <span class="n">widthscale</span> <span class="o">*</span> <span class="p">(</span><span class="mf">1.6</span> <span class="o">+</span> <span class="mf">3.5</span> <span class="o">*</span> <span class="n">nsamples</span><span class="p">)</span>
            <span class="n">height</span> <span class="o">=</span> <span class="n">heightscale</span> <span class="o">*</span> <span class="p">(</span><span class="mf">0.8</span> <span class="o">+</span> <span class="n">height_per</span> <span class="o">*</span> <span class="n">nlibraries</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid `orientation` </span><span class="si">{orientation}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">p</span> <span class="o">=</span> <span class="p">(</span><span class="n">ggplot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="s1">&#39;site&#39;</span><span class="p">,</span> <span class="s1">&#39;mut_char&#39;</span><span class="p">,</span>
                            <span class="n">fill</span><span class="o">=</span><span class="n">count_or_frequency</span><span class="p">))</span> <span class="o">+</span>
             <span class="n">geom_tile</span><span class="p">()</span> <span class="o">+</span>
             <span class="n">facet_grid</span><span class="p">(</span><span class="n">facet_str</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="n">width</span><span class="p">,</span> <span class="n">height</span><span class="p">),</span>
                   <span class="n">legend_key</span><span class="o">=</span><span class="n">element_blank</span><span class="p">(),</span>
                   <span class="n">axis_text_y</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>
                   <span class="p">)</span> <span class="o">+</span>
             <span class="n">scale_x_continuous</span><span class="p">(</span>
                <span class="n">name</span><span class="o">=</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{mut_desc}</span><span class="s1"> site&#39;</span><span class="p">,</span>
                <span class="n">limits</span><span class="o">=</span><span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sites</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="nb">max</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sites</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span>
                <span class="n">expand</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
                <span class="p">)</span> <span class="o">+</span>
             <span class="n">ylab</span><span class="p">(</span><span class="n">mut_desc</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">scale_fill_cmap</span><span class="p">(</span><span class="s1">&#39;gnuplot&#39;</span><span class="p">)</span>
             <span class="p">)</span>


        <span class="k">if</span> <span class="n">plotfile</span><span class="p">:</span>
            <span class="n">p</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">plotfile</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="n">height</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="n">width</span><span class="p">,</span>
                   <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">limitsize</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">p</span></div>


<div class="viewcode-block" id="CodonVariantTable.plotMutFreqs"><a class="viewcode-back" href="../../dms_tools2.barcodes.html#dms_tools2.barcodes.CodonVariantTable.plotMutFreqs">[docs]</a>    <span class="k">def</span> <span class="nf">plotMutFreqs</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">variant_type</span><span class="p">,</span> <span class="n">mut_type</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span>
            <span class="n">libraries</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">plotfile</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">orientation</span><span class="o">=</span><span class="s1">&#39;h&#39;</span><span class="p">,</span> <span class="n">widthscale</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">heightscale</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
            <span class="n">min_support</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Mutation frequency along length of gene.</span>

<span class="sd">        Args:</span>
<span class="sd">            Args have same meaning as for</span>
<span class="sd">            :CodonVariantTable.plotCumulMutCoverage`.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A `plotnine &lt;https://plotnine.readthedocs.io/en/stable/&gt;`_</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutCounts</span><span class="p">(</span><span class="n">variant_type</span><span class="p">,</span> <span class="n">mut_type</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="n">samples</span><span class="p">,</span>
                            <span class="n">libraries</span><span class="o">=</span><span class="n">libraries</span><span class="p">,</span> <span class="n">min_support</span><span class="o">=</span><span class="n">min_support</span><span class="p">)</span>

        <span class="n">n_variants</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">n_variants_df</span><span class="p">(</span><span class="n">libraries</span><span class="o">=</span><span class="n">libraries</span><span class="p">,</span>
                                         <span class="n">samples</span><span class="o">=</span><span class="n">samples</span><span class="p">,</span>
                                         <span class="n">min_support</span><span class="o">=</span><span class="n">min_support</span><span class="p">)</span>
                      <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;count&#39;</span><span class="p">:</span><span class="s1">&#39;nseqs&#39;</span><span class="p">})</span>
                      <span class="p">)</span>

        <span class="n">df</span> <span class="o">=</span> <span class="p">(</span><span class="n">df</span>
              <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;mutation_type&#39;</span><span class="p">,</span> <span class="s1">&#39;site&#39;</span><span class="p">])</span>
              <span class="o">.</span><span class="n">aggregate</span><span class="p">({</span><span class="s1">&#39;count&#39;</span><span class="p">:</span><span class="s1">&#39;sum&#39;</span><span class="p">})</span>
              <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
              <span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">n_variants</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">])</span>
              <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">freq</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;count&#39;</span><span class="p">]</span> <span class="o">/</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;nseqs&#39;</span><span class="p">])</span>
              <span class="p">)</span>

        <span class="n">nlibraries</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>
        <span class="n">nsamples</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>

        <span class="k">if</span> <span class="n">orientation</span> <span class="o">==</span> <span class="s1">&#39;h&#39;</span><span class="p">:</span>
            <span class="n">facet_str</span> <span class="o">=</span> <span class="s1">&#39;sample ~ library&#39;</span>
            <span class="n">width</span> <span class="o">=</span> <span class="n">widthscale</span> <span class="o">*</span> <span class="p">(</span><span class="mf">1.6</span> <span class="o">+</span> <span class="mf">1.8</span> <span class="o">*</span> <span class="n">nlibraries</span><span class="p">)</span>
            <span class="n">height</span> <span class="o">=</span> <span class="n">heightscale</span> <span class="o">*</span> <span class="p">(</span><span class="mf">0.8</span> <span class="o">+</span> <span class="mi">1</span> <span class="o">*</span> <span class="n">nsamples</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">orientation</span> <span class="o">==</span> <span class="s1">&#39;v&#39;</span><span class="p">:</span>
            <span class="n">facet_str</span> <span class="o">=</span> <span class="s1">&#39;library ~ sample&#39;</span>
            <span class="n">width</span> <span class="o">=</span> <span class="n">widthscale</span> <span class="o">*</span> <span class="p">(</span><span class="mf">1.6</span> <span class="o">+</span> <span class="mf">1.8</span> <span class="o">*</span> <span class="n">nsamples</span><span class="p">)</span>
            <span class="n">height</span> <span class="o">=</span> <span class="n">heightscale</span> <span class="o">*</span> <span class="p">(</span><span class="mf">0.8</span> <span class="o">+</span> <span class="mi">1</span> <span class="o">*</span> <span class="n">nlibraries</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid `orientation` </span><span class="si">{orientation}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">mut_type</span> <span class="o">==</span> <span class="s1">&#39;aa&#39;</span><span class="p">:</span>
            <span class="n">mut_desc</span> <span class="o">=</span> <span class="s1">&#39;amino-acid&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">mut_desc</span> <span class="o">=</span> <span class="n">mut_type</span>

        <span class="k">if</span> <span class="n">height</span> <span class="o">&lt;</span> <span class="mi">3</span><span class="p">:</span>
            <span class="n">ylabel</span> <span class="o">=</span> <span class="p">(</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{mut_desc}</span><span class="s1"> mutation</span><span class="se">\n</span><span class="s1">frequency &#39;</span>
                      <span class="n">f</span><span class="s1">&#39;(</span><span class="si">{variant_type}</span><span class="s1"> mutants)&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">ylabel</span> <span class="o">=</span> <span class="p">(</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{mut_desc}</span><span class="s1"> mutation frequency &#39;</span>
                      <span class="n">f</span><span class="s1">&#39;(</span><span class="si">{variant_type}</span><span class="s1"> mutants)&#39;</span><span class="p">)</span>

        <span class="n">p</span> <span class="o">=</span> <span class="p">(</span><span class="n">ggplot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="s1">&#39;site&#39;</span><span class="p">,</span> <span class="s1">&#39;freq&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;mutation_type&#39;</span><span class="p">))</span> <span class="o">+</span>
             <span class="n">geom_step</span><span class="p">()</span> <span class="o">+</span>
             <span class="n">scale_color_manual</span><span class="p">(</span>
                <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_mutation_type_colors</span><span class="p">[</span><span class="n">m</span><span class="p">]</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span>
                 <span class="n">df</span><span class="o">.</span><span class="n">mutation_type</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">sort_values</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()],</span>
                <span class="n">name</span><span class="o">=</span><span class="s1">&#39;mutation type&#39;</span>
                <span class="p">)</span> <span class="o">+</span>
             <span class="n">scale_x_continuous</span><span class="p">(</span>
                <span class="n">name</span><span class="o">=</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{mut_desc}</span><span class="s1"> site&#39;</span><span class="p">,</span>
                <span class="n">limits</span><span class="o">=</span><span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sites</span><span class="p">),</span> <span class="nb">max</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sites</span><span class="p">))</span>
                <span class="p">)</span> <span class="o">+</span>
             <span class="n">ylab</span><span class="p">(</span><span class="n">ylabel</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">facet_grid</span><span class="p">(</span><span class="n">facet_str</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="n">width</span><span class="p">,</span> <span class="n">height</span><span class="p">),</span>
                   <span class="n">legend_key</span><span class="o">=</span><span class="n">element_blank</span><span class="p">(),</span>
                   <span class="n">legend_text</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="mi">11</span><span class="p">)</span>
                   <span class="p">)</span>
             <span class="p">)</span>

        <span class="k">if</span> <span class="n">plotfile</span><span class="p">:</span>
            <span class="n">p</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">plotfile</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="n">height</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="n">width</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">p</span></div>


<div class="viewcode-block" id="CodonVariantTable.plotCumulMutCoverage"><a class="viewcode-back" href="../../dms_tools2.barcodes.html#dms_tools2.barcodes.CodonVariantTable.plotCumulMutCoverage">[docs]</a>    <span class="k">def</span> <span class="nf">plotCumulMutCoverage</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">variant_type</span><span class="p">,</span> <span class="n">mut_type</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span>
            <span class="n">libraries</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">plotfile</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">orientation</span><span class="o">=</span><span class="s1">&#39;h&#39;</span><span class="p">,</span> <span class="n">widthscale</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">heightscale</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
            <span class="n">min_support</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">max_count</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Fraction of mutation seen &lt;= some number of times.</span>

<span class="sd">        Args:</span>
<span class="sd">            `variant_type` (&quot;single&quot; or &quot;all&quot;)</span>
<span class="sd">                Include just single mutants, or all mutants?</span>
<span class="sd">                Mutations are counted relative to `mut_type`.</span>
<span class="sd">            `max_count` (`None` or int)</span>
<span class="sd">                Plot cumulative fraction plot out to this</span>
<span class="sd">                number of observations of mutation. If `None`,</span>
<span class="sd">                a reasonable value is automatically determined.</span>
<span class="sd">            Other args:</span>
<span class="sd">                Same meaning as for</span>
<span class="sd">                :class:`CodonVariantTable.plotNumMutsHistogram`</span>

<span class="sd">        Returns:</span>
<span class="sd">            A `plotnine &lt;https://plotnine.readthedocs.io/en/stable/&gt;`_</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutCounts</span><span class="p">(</span><span class="n">variant_type</span><span class="p">,</span> <span class="n">mut_type</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="n">samples</span><span class="p">,</span>
                            <span class="n">libraries</span><span class="o">=</span><span class="n">libraries</span><span class="p">,</span> <span class="n">min_support</span><span class="o">=</span><span class="n">min_support</span><span class="p">)</span>

        <span class="c1"># add one to counts to plot fraction found &lt; this many</span>
        <span class="c1"># as stat_ecdf by default does &lt;=</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">count</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;count&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">max_count</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">max_count</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;count&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">quantile</span><span class="p">(</span><span class="mf">0.75</span><span class="p">)</span>

        <span class="n">nlibraries</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>
        <span class="n">nsamples</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>

        <span class="k">if</span> <span class="n">orientation</span> <span class="o">==</span> <span class="s1">&#39;h&#39;</span><span class="p">:</span>
            <span class="n">facet_str</span> <span class="o">=</span> <span class="s1">&#39;sample ~ library&#39;</span>
            <span class="n">width</span> <span class="o">=</span> <span class="n">widthscale</span> <span class="o">*</span> <span class="p">(</span><span class="mf">1.6</span> <span class="o">+</span> <span class="mf">1.3</span> <span class="o">*</span> <span class="n">nlibraries</span><span class="p">)</span>
            <span class="n">height</span> <span class="o">=</span> <span class="n">heightscale</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="mf">1.2</span> <span class="o">*</span> <span class="n">nsamples</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">orientation</span> <span class="o">==</span> <span class="s1">&#39;v&#39;</span><span class="p">:</span>
            <span class="n">facet_str</span> <span class="o">=</span> <span class="s1">&#39;library ~ sample&#39;</span>
            <span class="n">width</span> <span class="o">=</span> <span class="n">widthscale</span> <span class="o">*</span> <span class="p">(</span><span class="mf">1.6</span> <span class="o">+</span> <span class="mf">1.3</span> <span class="o">*</span> <span class="n">nsamples</span><span class="p">)</span>
            <span class="n">height</span> <span class="o">=</span> <span class="n">heightscale</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="mf">1.2</span> <span class="o">*</span> <span class="n">nlibraries</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid `orientation` </span><span class="si">{orientation}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">width</span> <span class="o">&gt;</span> <span class="mi">4</span><span class="p">:</span>
            <span class="n">xlabel</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;counts among </span><span class="si">{variant_type}</span><span class="s1"> mutants&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">xlabel</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;counts among</span><span class="se">\n</span><span class="si">{variant_type}</span><span class="s1"> mutants&#39;</span>

        <span class="n">mut_desc</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;aa&#39;</span><span class="p">:</span><span class="s1">&#39;amino-acid&#39;</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">:</span><span class="s1">&#39;codon&#39;</span><span class="p">}[</span><span class="n">mut_type</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">height</span> <span class="o">&gt;</span> <span class="mi">3</span><span class="p">:</span>
            <span class="n">ylabel</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;frac </span><span class="si">{mut_desc}</span><span class="s1"> mutations found &lt; this many times&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">ylabel</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;frac </span><span class="si">{mut_desc}</span><span class="s1"> mutations</span><span class="se">\n</span><span class="s1">found &lt; this many times&#39;</span>

        <span class="n">p</span> <span class="o">=</span> <span class="p">(</span><span class="n">ggplot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="s1">&#39;count&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;mutation_type&#39;</span><span class="p">))</span> <span class="o">+</span>
             <span class="n">stat_ecdf</span><span class="p">(</span><span class="n">geom</span><span class="o">=</span><span class="s1">&#39;step&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mf">0.75</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">coord_cartesian</span><span class="p">(</span><span class="n">xlim</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">max_count</span><span class="p">))</span> <span class="o">+</span>
             <span class="n">scale_color_manual</span><span class="p">(</span>
                <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_mutation_type_colors</span><span class="p">[</span><span class="n">m</span><span class="p">]</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span>
                 <span class="n">df</span><span class="o">.</span><span class="n">mutation_type</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">sort_values</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()],</span>
                <span class="n">name</span><span class="o">=</span><span class="s1">&#39;mutation type&#39;</span>
                <span class="p">)</span> <span class="o">+</span>
             <span class="n">xlab</span><span class="p">(</span><span class="n">xlabel</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">ylab</span><span class="p">(</span><span class="n">ylabel</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">facet_grid</span><span class="p">(</span><span class="n">facet_str</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="n">width</span><span class="p">,</span> <span class="n">height</span><span class="p">),</span>
                   <span class="n">legend_key</span><span class="o">=</span><span class="n">element_blank</span><span class="p">(),</span>
                   <span class="n">legend_text</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="mi">11</span><span class="p">)</span>
                   <span class="p">)</span>
             <span class="p">)</span>

        <span class="k">if</span> <span class="n">plotfile</span><span class="p">:</span>
            <span class="n">p</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">plotfile</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="n">height</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="n">width</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">p</span></div>


<div class="viewcode-block" id="CodonVariantTable.plotNumCodonMutsByType"><a class="viewcode-back" href="../../dms_tools2.barcodes.html#dms_tools2.barcodes.CodonVariantTable.plotNumCodonMutsByType">[docs]</a>    <span class="k">def</span> <span class="nf">plotNumCodonMutsByType</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">variant_type</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span>
            <span class="n">libraries</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">plotfile</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">orientation</span><span class="o">=</span><span class="s1">&#39;h&#39;</span><span class="p">,</span> <span class="n">widthscale</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">heightscale</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
            <span class="n">min_support</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Nonsynonymous, synonymous, stop mutations per variant.</span>

<span class="sd">        Args:</span>
<span class="sd">            `variant_type` (&quot;single&quot; or &quot;all&quot;)</span>
<span class="sd">                Include just single-codon mutants and wildtype,</span>
<span class="sd">                or include all mutants.</span>
<span class="sd">            Other args:</span>
<span class="sd">                Same meaning as for</span>
<span class="sd">                :class:`CodonVariantTable.plotNumMutsHistogram`</span>

<span class="sd">        Returns:</span>
<span class="sd">            A `plotnine &lt;https://plotnine.readthedocs.io/en/stable/&gt;`_</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span><span class="p">,</span> <span class="n">nlibraries</span><span class="p">,</span> <span class="n">nsamples</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_getPlotData</span><span class="p">(</span><span class="n">libraries</span><span class="p">,</span>
                                                     <span class="n">samples</span><span class="p">,</span>
                                                     <span class="n">min_support</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">variant_type</span> <span class="o">==</span> <span class="s1">&#39;single&#39;</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;n_codon_substitutions &lt;= 1&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">variant_type</span> <span class="o">!=</span> <span class="s1">&#39;all&#39;</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid variant_type </span><span class="si">{variant_type}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">orientation</span> <span class="o">==</span> <span class="s1">&#39;h&#39;</span><span class="p">:</span>
            <span class="n">facet_str</span> <span class="o">=</span> <span class="s1">&#39;sample ~ library&#39;</span>
            <span class="n">width</span> <span class="o">=</span> <span class="n">widthscale</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="mf">1.4</span> <span class="o">*</span> <span class="n">nlibraries</span><span class="p">)</span>
            <span class="n">height</span> <span class="o">=</span> <span class="n">heightscale</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="mf">1.3</span> <span class="o">*</span> <span class="n">nsamples</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">orientation</span> <span class="o">==</span> <span class="s1">&#39;v&#39;</span><span class="p">:</span>
            <span class="n">facet_str</span> <span class="o">=</span> <span class="s1">&#39;library ~ sample&#39;</span>
            <span class="n">width</span> <span class="o">=</span> <span class="n">widthscale</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="mf">1.4</span> <span class="o">*</span> <span class="n">nsamples</span><span class="p">)</span>
            <span class="n">height</span> <span class="o">=</span> <span class="n">heightscale</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="mf">1.3</span> <span class="o">*</span> <span class="n">nlibraries</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid `orientation` </span><span class="si">{orientation}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">height</span> <span class="o">&gt;</span> <span class="mi">3</span><span class="p">:</span>
            <span class="n">ylabel</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;mutations per variant (</span><span class="si">{variant_type}</span><span class="s1"> mutants)&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">ylabel</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;mutations per variant</span><span class="se">\n</span><span class="s1">(</span><span class="si">{variant_type}</span><span class="s1"> mutants)&#39;</span>

        <span class="n">codon_mut_types</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;nonsynonymous&#39;</span><span class="p">,</span> <span class="s1">&#39;synonymous&#39;</span><span class="p">,</span> <span class="s1">&#39;stop&#39;</span><span class="p">]</span>

        <span class="c1"># mutations from stop to another amino-acid counted as nonsyn</span>
        <span class="n">df</span> <span class="o">=</span> <span class="p">(</span><span class="n">df</span>
              <span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                <span class="n">synonymous</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">n_codon_substitutions</span> <span class="o">-</span>
                                     <span class="n">x</span><span class="o">.</span><span class="n">n_aa_substitutions</span><span class="p">,</span>
                <span class="n">stop</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">aa_substitutions</span><span class="o">.</span><span class="n">str</span>
                               <span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="s1">&#39;[A-Z]\d+\*&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="nb">len</span><span class="p">),</span>
                <span class="n">nonsynonymous</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">n_codon_substitutions</span> <span class="o">-</span>
                                        <span class="n">x</span><span class="o">.</span><span class="n">synonymous</span> <span class="o">-</span> <span class="n">x</span><span class="o">.</span><span class="n">stop</span>
                <span class="p">)</span>
              <span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">],</span>
                    <span class="n">value_vars</span><span class="o">=</span><span class="n">codon_mut_types</span><span class="p">,</span>
                    <span class="n">var_name</span><span class="o">=</span><span class="s1">&#39;mutation_type&#39;</span><span class="p">,</span>
                    <span class="n">value_name</span><span class="o">=</span><span class="s1">&#39;num_muts&#39;</span><span class="p">)</span>
              <span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                  <span class="n">mutation_type</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                                <span class="n">pandas</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
                                 <span class="n">x</span><span class="o">.</span><span class="n">mutation_type</span><span class="p">,</span>
                                 <span class="n">categories</span><span class="o">=</span><span class="n">codon_mut_types</span><span class="p">,</span>
                                 <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
                  <span class="n">num_muts_count</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">num_muts</span> <span class="o">*</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;count&#39;</span><span class="p">]</span>
                  <span class="p">)</span>
              <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;mutation_type&#39;</span><span class="p">])</span>
              <span class="o">.</span><span class="n">aggregate</span><span class="p">({</span><span class="s1">&#39;num_muts_count&#39;</span><span class="p">:</span><span class="s1">&#39;sum&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">:</span><span class="s1">&#39;sum&#39;</span><span class="p">})</span>
              <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
              <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">number</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">num_muts_count</span> <span class="o">/</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;count&#39;</span><span class="p">])</span>
              <span class="p">)</span>

        <span class="n">p</span> <span class="o">=</span> <span class="p">(</span><span class="n">ggplot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="s1">&#39;mutation_type&#39;</span><span class="p">,</span> <span class="s1">&#39;number&#39;</span><span class="p">,</span>
                            <span class="n">fill</span><span class="o">=</span><span class="s1">&#39;mutation_type&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;number&#39;</span><span class="p">))</span> <span class="o">+</span>
             <span class="n">geom_bar</span><span class="p">(</span><span class="n">stat</span><span class="o">=</span><span class="s1">&#39;identity&#39;</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">geom_text</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;bottom&#39;</span><span class="p">,</span> <span class="n">format_string</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">{0:.3f}</span><span class="s1">&#39;</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">facet_grid</span><span class="p">(</span><span class="n">facet_str</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">scale_y_continuous</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">ylabel</span><span class="p">,</span>
                                <span class="n">expand</span><span class="o">=</span><span class="p">(</span><span class="mf">0.03</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mf">0.12</span><span class="p">,</span> <span class="mi">0</span><span class="p">))</span> <span class="o">+</span>
             <span class="n">scale_fill_manual</span><span class="p">(</span>
                <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_mutation_type_colors</span><span class="p">[</span><span class="n">m</span><span class="p">]</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span>
                 <span class="n">df</span><span class="o">.</span><span class="n">mutation_type</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">sort_values</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()]</span>
                <span class="p">)</span> <span class="o">+</span>
             <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="n">width</span><span class="p">,</span> <span class="n">height</span><span class="p">),</span>
                   <span class="n">axis_title_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">(),</span>
                   <span class="n">axis_text_x</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="mi">90</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">11</span><span class="p">),</span>
                   <span class="n">legend_position</span><span class="o">=</span><span class="s1">&#39;none&#39;</span><span class="p">)</span>
             <span class="p">)</span>

        <span class="k">if</span> <span class="n">plotfile</span><span class="p">:</span>
            <span class="n">p</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">plotfile</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="n">height</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="n">width</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">p</span></div>


<div class="viewcode-block" id="CodonVariantTable.plotNumMutsHistogram"><a class="viewcode-back" href="../../dms_tools2.barcodes.html#dms_tools2.barcodes.CodonVariantTable.plotNumMutsHistogram">[docs]</a>    <span class="k">def</span> <span class="nf">plotNumMutsHistogram</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">mut_type</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span>
            <span class="n">libraries</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">plotfile</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">orientation</span><span class="o">=</span><span class="s1">&#39;h&#39;</span><span class="p">,</span> <span class="n">widthscale</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">heightscale</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
            <span class="n">min_support</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">max_muts</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Plot histograms of number of mutations per variant.</span>

<span class="sd">        Args:</span>
<span class="sd">            `mut_type` (str)</span>
<span class="sd">                Type of mutation: &quot;codon&quot; or &quot;aa&quot;.</span>
<span class="sd">            `libraries` (&quot;all&quot;, &quot;all_only&quot;, or list)</span>
<span class="sd">                Set to &quot;all&quot; to include all libraries including</span>
<span class="sd">                a merge of the libraries, &quot;all_only&quot; to only</span>
<span class="sd">                include the merge of the libraries, or a list</span>
<span class="sd">                of libraries.</span>
<span class="sd">            `samples` (the str &quot;all&quot;, `None`, or list)</span>
<span class="sd">                Set to &quot;all&quot; to include all samples, `None` to</span>
<span class="sd">                just count each barcoded variant once, or specify</span>
<span class="sd">                a list of samples.</span>
<span class="sd">            `plotfile` (`None` or str)</span>
<span class="sd">                Name of file to which we save plot.</span>
<span class="sd">            `orientation` (the str &#39;h&#39; or &#39;v&#39;)</span>
<span class="sd">                Do we facet libraries horizontally or vertically?</span>
<span class="sd">            `widthscale` (float or int)</span>
<span class="sd">                Expand width of plot by this much.</span>
<span class="sd">            `heightscale` (float or int)</span>
<span class="sd">                Expand height of plot by this much.</span>
<span class="sd">            `min_support` (int)</span>
<span class="sd">                Only include variants with at least this large</span>
<span class="sd">                of a variant call support.</span>
<span class="sd">            `max_muts` (int or `None`)</span>
<span class="sd">                In histogram, group together all variants with</span>
<span class="sd">                &gt;= this many mutations; set to `None` for no</span>
<span class="sd">                cutoff.</span>

<span class="sd">        Returns:</span>
<span class="sd">            `plotnine &lt;https://plotnine.readthedocs.io/en/stable/&gt;`_</span>
<span class="sd">            plot; can be displayed in a Jupyter notebook with `p.draw()`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span><span class="p">,</span> <span class="n">nlibraries</span><span class="p">,</span> <span class="n">nsamples</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_getPlotData</span><span class="p">(</span><span class="n">libraries</span><span class="p">,</span>
                                                     <span class="n">samples</span><span class="p">,</span>
                                                     <span class="n">min_support</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">mut_type</span> <span class="o">==</span> <span class="s1">&#39;aa&#39;</span><span class="p">:</span>
            <span class="n">mut_col</span> <span class="o">=</span> <span class="s1">&#39;n_aa_substitutions&#39;</span>
            <span class="n">xlabel</span> <span class="o">=</span> <span class="s1">&#39;amino-acid mutations&#39;</span>
        <span class="k">elif</span> <span class="n">mut_type</span> <span class="o">==</span> <span class="s1">&#39;codon&#39;</span><span class="p">:</span>
            <span class="n">mut_col</span> <span class="o">=</span> <span class="s1">&#39;n_codon_substitutions&#39;</span>
            <span class="n">xlabel</span> <span class="o">=</span> <span class="s1">&#39;codon mutations&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid mut_type </span><span class="si">{mut_type}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">orientation</span> <span class="o">==</span> <span class="s1">&#39;h&#39;</span><span class="p">:</span>
            <span class="n">facet_str</span> <span class="o">=</span> <span class="s1">&#39;sample ~ library&#39;</span>
            <span class="n">width</span> <span class="o">=</span> <span class="n">widthscale</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="mf">1.5</span> <span class="o">*</span> <span class="n">nlibraries</span><span class="p">)</span>
            <span class="n">height</span> <span class="o">=</span> <span class="n">heightscale</span> <span class="o">*</span> <span class="p">(</span><span class="mf">0.6</span> <span class="o">+</span> <span class="mf">1.5</span> <span class="o">*</span> <span class="n">nsamples</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">orientation</span> <span class="o">==</span> <span class="s1">&#39;v&#39;</span><span class="p">:</span>
            <span class="n">facet_str</span> <span class="o">=</span> <span class="s1">&#39;library ~ sample&#39;</span>
            <span class="n">width</span> <span class="o">=</span> <span class="n">widthscale</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="mf">1.5</span> <span class="o">*</span> <span class="n">nsamples</span><span class="p">)</span>
            <span class="n">height</span> <span class="o">=</span> <span class="n">heightscale</span> <span class="o">*</span> <span class="p">(</span><span class="mf">0.6</span> <span class="o">+</span> <span class="mf">1.5</span> <span class="o">*</span> <span class="n">nlibraries</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid `orientation` </span><span class="si">{orientation}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">df</span><span class="p">[</span><span class="n">mut_col</span><span class="p">]</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">mut_col</span><span class="p">],</span> <span class="kc">None</span><span class="p">,</span> <span class="n">max_muts</span><span class="p">)</span>

        <span class="n">df</span> <span class="o">=</span> <span class="p">(</span><span class="n">df</span>
              <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="n">mut_col</span><span class="p">])</span>
              <span class="o">.</span><span class="n">aggregate</span><span class="p">({</span><span class="s1">&#39;count&#39;</span><span class="p">:</span><span class="s1">&#39;sum&#39;</span><span class="p">})</span>
              <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
              <span class="p">)</span>

        <span class="n">p</span> <span class="o">=</span> <span class="p">(</span><span class="n">ggplot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="n">mut_col</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">))</span> <span class="o">+</span>
             <span class="n">geom_bar</span><span class="p">(</span><span class="n">stat</span><span class="o">=</span><span class="s1">&#39;identity&#39;</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">facet_grid</span><span class="p">(</span><span class="n">facet_str</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">xlab</span><span class="p">(</span><span class="n">xlabel</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">scale_y_continuous</span><span class="p">(</span><span class="n">labels</span><span class="o">=</span><span class="n">dms_tools2</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">latexSciNot</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="n">width</span><span class="p">,</span> <span class="n">height</span><span class="p">))</span>
             <span class="p">)</span>

        <span class="k">if</span> <span class="n">plotfile</span><span class="p">:</span>
            <span class="n">p</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">plotfile</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="n">height</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="n">width</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">p</span></div>


<div class="viewcode-block" id="CodonVariantTable.writeCodonCounts"><a class="viewcode-back" href="../../dms_tools2.barcodes.html#dms_tools2.barcodes.CodonVariantTable.writeCodonCounts">[docs]</a>    <span class="k">def</span> <span class="nf">writeCodonCounts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">single_or_all</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">outdir</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Writes codon counts files for all libraries and samples.</span>

<span class="sd">        Files are written in</span>
<span class="sd">        `format &lt;https://jbloomlab.github.io/dms_tools2/dms2_bcsubamp.html#id12&gt;`_</span>
<span class="sd">        produced by</span>
<span class="sd">        `dms2_bcsubamp &lt;https://jbloomlab.github.io/dms_tools2/dms2_bcsubamp.html&gt;`_.</span>

<span class="sd">        Args:</span>
<span class="sd">            `single_or_all` (&quot;single&quot; or &quot;all&quot;)</span>
<span class="sd">                If &quot;single&quot;, then counts are just from single-codon</span>
<span class="sd">                mutants and wildtype, and we count all wildtype codons</span>
<span class="sd">                and just mutated codon for single-codon mutants. If</span>
<span class="sd">                &quot;all&quot;, we count all codons for all variants at all</span>
<span class="sd">                sites. This is appropriate if enrichment of each mutation</span>
<span class="sd">                is supposed to represent its effect for &quot;single&quot;, and if</span>
<span class="sd">                enrichment of mutation is supposed to represent its</span>
<span class="sd">                average effect across genetic backgrounds in the library</span>
<span class="sd">                for &quot;all&quot; provided mutations are Poisson distributed.</span>
<span class="sd">            `outdir` (`None` or str)</span>
<span class="sd">                Name of directory into which we write counts,</span>
<span class="sd">                created if it does not exist. Use `None` to</span>
<span class="sd">                write to current directory.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A pandas data frame with the columns &quot;library&quot;, &quot;sample&quot;,</span>
<span class="sd">            (&quot;all-libraries&quot; is one library) and &quot;countfile&quot;.</span>
<span class="sd">            The &quot;countfile&quot; column gives the name of the created</span>
<span class="sd">            file, which is ``&lt;library&gt;_&lt;sample&gt;_codoncounts.csv``.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">codonmatch</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;^(?P&lt;wt&gt;{&quot;|&quot;.join(CODONS)})&#39;</span>
                                 <span class="s1">&#39;(?P&lt;r&gt;\d+)&#39;</span>
                                <span class="n">f</span><span class="s1">&#39;(?P&lt;mut&gt;{&quot;|&quot;.join(CODONS)})$&#39;</span>
                                <span class="p">)</span>
        <span class="k">def</span> <span class="nf">_parseCodonMut</span><span class="p">(</span><span class="n">mutstr</span><span class="p">):</span>
            <span class="n">m</span> <span class="o">=</span> <span class="n">codonmatch</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">mutstr</span><span class="p">)</span>
            <span class="k">return</span> <span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;wt&#39;</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">)),</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;mut&#39;</span><span class="p">))</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">variant_count_df</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;no samples with counts&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">single_or_all</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">{</span><span class="s1">&#39;single&#39;</span><span class="p">,</span> <span class="s1">&#39;all&#39;</span><span class="p">}:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid `single_or_all` </span><span class="si">{single_or_all}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">outdir</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">outdir</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>

        <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">addMergedLibraries</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">variant_count_df</span><span class="p">,</span>
                                     <span class="n">all_lib</span><span class="o">=</span><span class="s1">&#39;all-libraries&#39;</span><span class="p">)</span>

        <span class="n">countfiles</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">liblist</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">samplelist</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">lib</span><span class="p">,</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">product</span><span class="p">(</span>
                            <span class="n">df</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span>
                            <span class="n">df</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
                            <span class="p">):</span>

            <span class="n">countfile</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span>
                            <span class="n">f</span><span class="s1">&#39;</span><span class="si">{lib}</span><span class="s1">_</span><span class="si">{sample}</span><span class="s1">_codoncounts.csv&#39;</span><span class="p">)</span>
            <span class="n">countfiles</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">countfile</span><span class="p">)</span>
            <span class="n">liblist</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lib</span><span class="p">)</span>
            <span class="n">samplelist</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>

            <span class="n">i_df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;library == @lib &amp; sample == @sample&#39;</span><span class="p">)</span>

            <span class="n">codoncounts</span> <span class="o">=</span> <span class="p">{</span><span class="n">codon</span><span class="p">:[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sites</span><span class="p">)</span> <span class="k">for</span> <span class="n">codon</span>
                           <span class="ow">in</span> <span class="n">CODONS</span><span class="p">}</span>

            <span class="k">if</span> <span class="n">single_or_all</span> <span class="o">==</span> <span class="s1">&#39;single&#39;</span><span class="p">:</span>
                <span class="n">n_wt</span> <span class="o">=</span> <span class="p">(</span><span class="n">i_df</span>
                        <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;n_codon_substitutions == 0&#39;</span><span class="p">)</span>
                        <span class="p">[</span><span class="s1">&#39;count&#39;</span><span class="p">]</span>
                        <span class="o">.</span><span class="n">sum</span><span class="p">()</span>
                        <span class="p">)</span>
                <span class="k">for</span> <span class="n">isite</span><span class="p">,</span> <span class="n">site</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sites</span><span class="p">):</span>
                    <span class="n">codoncounts</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">codons</span><span class="p">[</span><span class="n">site</span><span class="p">]][</span><span class="n">isite</span><span class="p">]</span> <span class="o">+=</span> <span class="n">n_wt</span>
                <span class="k">for</span> <span class="n">mut</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="p">(</span><span class="n">i_df</span>
                                   <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;n_codon_substitutions == 1&#39;</span><span class="p">)</span>
                                   <span class="p">[[</span><span class="s1">&#39;codon_substitutions&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">]]</span>
                                   <span class="o">.</span><span class="n">itertuples</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
                                   <span class="p">):</span>
                    <span class="n">wtcodon</span><span class="p">,</span> <span class="n">r</span><span class="p">,</span> <span class="n">mutcodon</span> <span class="o">=</span> <span class="n">_parseCodonMut</span><span class="p">(</span><span class="n">mut</span><span class="p">)</span>
                    <span class="n">codoncounts</span><span class="p">[</span><span class="n">mutcodon</span><span class="p">][</span><span class="n">r</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">+=</span> <span class="n">count</span>

            <span class="k">elif</span> <span class="n">single_or_all</span> <span class="o">==</span> <span class="s1">&#39;all&#39;</span><span class="p">:</span>
                <span class="n">n_wt</span> <span class="o">=</span> <span class="n">i_df</span><span class="p">[</span><span class="s1">&#39;count&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
                <span class="k">for</span> <span class="n">isite</span><span class="p">,</span> <span class="n">site</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sites</span><span class="p">):</span>
                    <span class="n">codoncounts</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">codons</span><span class="p">[</span><span class="n">site</span><span class="p">]][</span><span class="n">isite</span><span class="p">]</span> <span class="o">+=</span> <span class="n">n_wt</span>
                <span class="k">for</span> <span class="n">muts</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="p">(</span><span class="n">i_df</span>
                                   <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;n_codon_substitutions &gt; 0&#39;</span><span class="p">)</span>
                                   <span class="p">[[</span><span class="s1">&#39;codon_substitutions&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">]]</span>
                                   <span class="o">.</span><span class="n">itertuples</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
                                   <span class="p">):</span>
                    <span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">muts</span><span class="o">.</span><span class="n">split</span><span class="p">():</span>
                        <span class="n">wtcodon</span><span class="p">,</span> <span class="n">r</span><span class="p">,</span> <span class="n">mutcodon</span> <span class="o">=</span> <span class="n">_parseCodonMut</span><span class="p">(</span><span class="n">mut</span><span class="p">)</span>
                        <span class="n">codoncounts</span><span class="p">[</span><span class="n">mutcodon</span><span class="p">][</span><span class="n">r</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">+=</span> <span class="n">count</span>
                        <span class="n">codoncounts</span><span class="p">[</span><span class="n">wtcodon</span><span class="p">][</span><span class="n">r</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">-=</span> <span class="n">count</span>

            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid `single_or_all` </span><span class="si">{single_or_all}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="n">counts_df</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">collections</span><span class="o">.</span><span class="n">OrderedDict</span><span class="p">(</span>
                         <span class="p">[(</span><span class="s1">&#39;site&#39;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">sites</span><span class="p">),</span>
                          <span class="p">(</span><span class="s1">&#39;wildtype&#39;</span><span class="p">,</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">codons</span><span class="p">[</span><span class="n">r</span><span class="p">]</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sites</span><span class="p">])]</span> <span class="o">+</span>
                         <span class="p">[(</span><span class="n">codon</span><span class="p">,</span> <span class="n">codoncounts</span><span class="p">[</span><span class="n">codon</span><span class="p">])</span> <span class="k">for</span> <span class="n">codon</span> <span class="ow">in</span> <span class="n">CODONS</span><span class="p">]</span>
                         <span class="p">))</span>
            <span class="n">counts_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">countfile</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="k">assert</span> <span class="nb">all</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">,</span> <span class="n">countfiles</span><span class="p">))</span>

        <span class="k">return</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;library&#39;</span><span class="p">:</span><span class="n">liblist</span><span class="p">,</span>
                                 <span class="s1">&#39;sample&#39;</span><span class="p">:</span><span class="n">samplelist</span><span class="p">,</span>
                                 <span class="s1">&#39;countfile&#39;</span><span class="p">:</span><span class="n">countfiles</span><span class="p">})</span></div>


<div class="viewcode-block" id="CodonVariantTable.addMergedLibraries"><a class="viewcode-back" href="../../dms_tools2.barcodes.html#dms_tools2.barcodes.CodonVariantTable.addMergedLibraries">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">addMergedLibraries</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">all_lib</span><span class="o">=</span><span class="s1">&#39;all libraries&#39;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Add data to `df` for all libraries merged.</span>

<span class="sd">        Args:</span>
<span class="sd">            `df` (pandas DataFrame)</span>
<span class="sd">                DataFrame that includes columns named</span>
<span class="sd">                &quot;library&quot; and &quot;barcode&quot;.</span>
<span class="sd">            `all_lib` (str)</span>
<span class="sd">                Name given to library that is merge of all</span>
<span class="sd">                other libraries.</span>

<span class="sd">        Returns:</span>
<span class="sd">            If `df` only has data for one library, just returns</span>
<span class="sd">            `df`. Otherwise returns a copy of `df` that has a</span>
<span class="sd">            new library with the name given by `all_lib`</span>
<span class="sd">            and contains the data for all individual libraries,</span>
<span class="sd">            with the &quot;barcode&quot; column giving the original</span>
<span class="sd">            library name followed by a hyphen and the barcode.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">libs</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">library</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">libs</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">df</span>

        <span class="k">if</span> <span class="n">all_lib</span> <span class="ow">in</span> <span class="n">libs</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;library </span><span class="si">{all_lib}</span><span class="s2"> already exists&quot;</span><span class="p">)</span>

        <span class="n">df</span> <span class="o">=</span> <span class="p">(</span><span class="n">pandas</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">df</span><span class="p">,</span>
                             <span class="n">df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                                <span class="n">barcode</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                                    <span class="n">x</span><span class="o">.</span><span class="n">library</span><span class="o">.</span><span class="n">str</span>
                                     <span class="o">.</span><span class="n">cat</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">barcode</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">),</span>
                                <span class="n">library</span><span class="o">=</span><span class="n">all_lib</span><span class="p">)</span>
                             <span class="p">],</span>
                            <span class="n">axis</span><span class="o">=</span><span class="s1">&#39;index&#39;</span><span class="p">,</span>
                            <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                            <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
              <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">library</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                              <span class="n">pandas</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
                               <span class="n">x</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">],</span>
                               <span class="n">categories</span><span class="o">=</span><span class="n">libs</span> <span class="o">+</span> <span class="p">[</span><span class="n">all_lib</span><span class="p">],</span>
                               <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                      <span class="p">)</span>
              <span class="p">)</span>

        <span class="k">return</span> <span class="n">df</span></div>


    <span class="k">def</span> <span class="nf">_getPlotData</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">libraries</span><span class="p">,</span> <span class="n">samples</span><span class="p">,</span> <span class="n">min_support</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Gets data to plot from library and sample filters.</span>

<span class="sd">        Args:</span>
<span class="sd">            `libraries`, `samples`, `min_support` have meaning as</span>
<span class="sd">            for :class:`CodonVariantTable.plotNumMutsHistogram`.</span>

<span class="sd">        Returns:</span>
<span class="sd">            The 3-tuple `(df, nlibraries, nsamples)` where:</span>

<span class="sd">                - `df`: DataFrame with data to plot.</span>

<span class="sd">                - `nlibraries`: number of libraries being plotted.</span>

<span class="sd">                - `nsamples`: number of samples being plotted.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">samples</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">barcode_variant_df</span>
                  <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">sample</span><span class="o">=</span><span class="s1">&#39;barcoded variants&#39;</span><span class="p">)</span>
                  <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">count</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
                  <span class="p">)</span>
        <span class="k">elif</span> <span class="n">samples</span> <span class="o">==</span> <span class="s1">&#39;all&#39;</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">variant_count_df</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;no samples have been added&#39;</span><span class="p">)</span>
            <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">variant_count_df</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">samples</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="n">all_samples</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">itertools</span><span class="o">.</span><span class="n">chain</span><span class="o">.</span><span class="n">from_iterable</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">samples</span><span class="p">(</span><span class="n">lib</span><span class="p">)</span> <span class="k">for</span> <span class="n">lib</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">libraries</span><span class="p">))</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">all_samples</span><span class="o">.</span><span class="n">issuperset</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">samples</span><span class="p">)):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid sample(s) in </span><span class="si">{samples}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">samples</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">samples</span><span class="p">)):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;duplicate samples in </span><span class="si">{samples}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">df</span> <span class="o">=</span> <span class="p">(</span><span class="n">df</span>
                  <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;sample in @samples&#39;</span><span class="p">)</span>
                  <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">sample</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                                 <span class="n">pandas</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
                                   <span class="n">x</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">],</span>
                                   <span class="n">categories</span><span class="o">=</span><span class="n">samples</span><span class="p">,</span>
                                   <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                          <span class="p">)</span>
                  <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid `samples` </span><span class="si">{samples}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;variant_call_support &gt;= @min_support&#39;</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;no samples </span><span class="si">{samples}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">nsamples</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>

        <span class="k">if</span> <span class="n">libraries</span> <span class="o">==</span> <span class="s1">&#39;all&#39;</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">addMergedLibraries</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">libraries</span> <span class="o">==</span> <span class="s1">&#39;all_only&#39;</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">addMergedLibraries</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
                  <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;library == &quot;all libraries&quot;&#39;</span><span class="p">)</span>
                  <span class="p">)</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">libraries</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">libraries</span><span class="p">)</span><span class="o">.</span><span class="n">issuperset</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">libraries</span><span class="p">)):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid library in </span><span class="si">{libraries}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">libraries</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">libraries</span><span class="p">)):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;duplicate library in </span><span class="si">{libraries}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">df</span> <span class="o">=</span> <span class="p">(</span><span class="n">df</span>
                  <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;library in @libraries&#39;</span><span class="p">)</span>
                  <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">library</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                                  <span class="n">pandas</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
                                   <span class="n">x</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">],</span>
                                   <span class="n">categories</span><span class="o">=</span><span class="n">libraries</span><span class="p">,</span>
                                   <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                          <span class="p">)</span>
                  <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid `libraries` </span><span class="si">{libraries}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;no libraries </span><span class="si">{libraries}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">nlibraries</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>

        <span class="k">return</span> <span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">nlibraries</span><span class="p">,</span> <span class="n">nsamples</span><span class="p">)</span>


    <span class="k">def</span> <span class="nf">_codonToAAMuts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">codon_mut_str</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Converts string of codon mutations to amino-acid mutations.</span>

<span class="sd">        Args:</span>
<span class="sd">            `codon_mut_str` (str)</span>
<span class="sd">                Codon mutations, delimited by a space and in</span>
<span class="sd">                1, 2, ... numbering.</span>

<span class="sd">        Returns:</span>
<span class="sd">            String with amino acid mutations in 1, 2, ... numbering.</span>

<span class="sd">        &gt;&gt;&gt; geneseq = &#39;ATGGGATGA&#39;</span>
<span class="sd">        &gt;&gt;&gt; with tempfile.NamedTemporaryFile(mode=&#39;w&#39;) as f:</span>
<span class="sd">        ...     _ = f.write(&#39;library,barcode,substitutions,variant_call_support&#39;)</span>
<span class="sd">        ...     f.flush()</span>
<span class="sd">        ...     variants = CodonVariantTable(</span>
<span class="sd">        ...                 barcode_variant_file=f.name,</span>
<span class="sd">        ...                 geneseq=geneseq</span>
<span class="sd">        ...                 )</span>
<span class="sd">        &gt;&gt;&gt; variants._codonToAAMuts(&#39;ATG1GTG GGA2GGC TGA3AGA&#39;)</span>
<span class="sd">        &#39;M1V *3R&#39;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">aa_muts</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">codon_mut_str</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">():</span>
            <span class="n">m</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s1">&#39;^(?P&lt;wt&gt;[ATGC]</span><span class="si">{3}</span><span class="s1">)(?P&lt;r&gt;\d+)(?P&lt;mut&gt;[ATGC]</span><span class="si">{3}</span><span class="s1">)$&#39;</span><span class="p">,</span>
                         <span class="n">mut</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">m</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid codon mutation </span><span class="si">{mut}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">r</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">aa_muts</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;duplicate codon mutation for </span><span class="si">{r}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">wt_codon</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;wt&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">geneseq</span><span class="p">[</span><span class="mi">3</span> <span class="o">*</span> <span class="p">(</span><span class="n">r</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="p">:</span> <span class="mi">3</span> <span class="o">*</span> <span class="n">r</span><span class="p">]</span> <span class="o">!=</span> <span class="n">wt_codon</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid wildtype codon in </span><span class="si">{mut}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">mut_codon</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;mut&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">wt_codon</span> <span class="o">==</span> <span class="n">mut_codon</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid mutation </span><span class="si">{mut}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">wt_aa</span> <span class="o">=</span> <span class="n">dms_tools2</span><span class="o">.</span><span class="n">CODON_TO_AA</span><span class="p">[</span><span class="n">wt_codon</span><span class="p">]</span>
            <span class="k">assert</span> <span class="n">wt_aa</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">aas</span><span class="p">[</span><span class="n">r</span><span class="p">]</span>
            <span class="n">mut_aa</span> <span class="o">=</span> <span class="n">dms_tools2</span><span class="o">.</span><span class="n">CODON_TO_AA</span><span class="p">[</span><span class="n">mut_codon</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">wt_aa</span> <span class="o">!=</span> <span class="n">mut_aa</span><span class="p">:</span>
                <span class="n">aa_muts</span><span class="p">[</span><span class="n">r</span><span class="p">]</span> <span class="o">=</span> <span class="n">f</span><span class="s2">&quot;</span><span class="si">{wt_aa}{r}{mut_aa}</span><span class="s2">&quot;</span>

        <span class="k">return</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">mut_str</span> <span class="k">for</span> <span class="n">r</span><span class="p">,</span> <span class="n">mut_str</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">aa_muts</span><span class="o">.</span><span class="n">items</span><span class="p">())])</span>


    <span class="k">def</span> <span class="nf">_ntToCodonMuts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">nt_mut_str</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Converts string of nucleotide mutations to codon mutations.</span>

<span class="sd">        Args:</span>
<span class="sd">            `nt_mut_str` (str)</span>
<span class="sd">                Nucleotide mutations, delimited by a space and in</span>
<span class="sd">                1, 2, ... numbering.</span>

<span class="sd">        Returns:</span>
<span class="sd">            String with codon mutations in 1, 2, ... numbering of</span>
<span class="sd">            codon sites.</span>

<span class="sd">        &gt;&gt;&gt; geneseq = &#39;ATGGGATGA&#39;</span>
<span class="sd">        &gt;&gt;&gt; with tempfile.NamedTemporaryFile(mode=&#39;w&#39;) as f:</span>
<span class="sd">        ...     _ = f.write(&#39;library,barcode,substitutions,variant_call_support&#39;)</span>
<span class="sd">        ...     f.flush()</span>
<span class="sd">        ...     variants = CodonVariantTable(</span>
<span class="sd">        ...                 barcode_variant_file=f.name,</span>
<span class="sd">        ...                 geneseq=geneseq</span>
<span class="sd">        ...                 )</span>
<span class="sd">        &gt;&gt;&gt; variants._ntToCodonMuts(&#39;A1G G4C A6T&#39;)</span>
<span class="sd">        &#39;ATG1GTG GGA2CGT&#39;</span>
<span class="sd">        &gt;&gt;&gt; variants._ntToCodonMuts(&#39;A1G G4C G6T&#39;)</span>
<span class="sd">        Traceback (most recent call last):</span>
<span class="sd">        ...</span>
<span class="sd">        ValueError: nucleotide 6 should be A not G</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">mut_codons</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">nt_mut_str</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">():</span>
            <span class="n">m</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s1">&#39;^(?P&lt;wt&gt;[ATCG])(?P&lt;i&gt;\d+)(?P&lt;mut&gt;[ATCG])$&#39;</span><span class="p">,</span> <span class="n">mut</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">m</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid mutation </span><span class="si">{mut}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">wt_nt</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;wt&#39;</span><span class="p">)</span>
            <span class="n">i</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;i&#39;</span><span class="p">))</span>
            <span class="n">mut_nt</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;mut&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">wt_nt</span> <span class="o">==</span> <span class="n">mut_nt</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid mutation </span><span class="si">{mut}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">&gt;</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">geneseq</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid nucleotide site </span><span class="si">{i}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">geneseq</span><span class="p">[</span><span class="n">i</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="n">wt_nt</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;nucleotide </span><span class="si">{i}</span><span class="s2"> should be &quot;</span>
                                 <span class="n">f</span><span class="s2">&quot;</span><span class="si">{self.geneseq[i - 1]}</span><span class="s2"> not </span><span class="si">{wt_nt}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">icodon</span> <span class="o">=</span> <span class="p">(</span><span class="n">i</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">//</span> <span class="mi">3</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="n">i_nt</span> <span class="o">=</span> <span class="p">(</span><span class="n">i</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">%</span> <span class="mi">3</span>
            <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">codons</span><span class="p">[</span><span class="n">icodon</span><span class="p">][</span><span class="n">i_nt</span><span class="p">]</span> <span class="o">==</span> <span class="n">wt_nt</span>
            <span class="k">if</span> <span class="n">i_nt</span> <span class="ow">in</span> <span class="n">mut_codons</span><span class="p">[</span><span class="n">icodon</span><span class="p">]:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;duplicate mutations </span><span class="si">{i_nt}</span><span class="s2"> in </span><span class="si">{icodon}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">mut_codons</span><span class="p">[</span><span class="n">icodon</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">((</span><span class="n">i_nt</span><span class="p">,</span> <span class="n">mut_nt</span><span class="p">))</span>

        <span class="n">codon_mut_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">r</span><span class="p">,</span> <span class="n">r_muts</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">mut_codons</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
            <span class="n">wt_codon</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">codons</span><span class="p">[</span><span class="n">r</span><span class="p">]</span>
            <span class="n">mut_codon</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">wt_codon</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">mut_nt</span> <span class="ow">in</span> <span class="n">r_muts</span><span class="p">:</span>
                <span class="n">mut_codon</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">mut_nt</span>
            <span class="n">codon_mut_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;</span><span class="si">{wt_codon}{r}</span><span class="s2">{&#39;&#39;.join(mut_codon)}&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">codon_mut_list</span><span class="p">)</span></div>


<div class="viewcode-block" id="tidy_split"><a class="viewcode-back" href="../../dms_tools2.barcodes.html#dms_tools2.barcodes.tidy_split">[docs]</a><span class="k">def</span> <span class="nf">tidy_split</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">column</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="n">keep</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Split the values of a column and expand so the new DataFrame has one split</span>
<span class="sd">    value per row. Filters rows where the column is missing.</span>

<span class="sd">    Taken from https://stackoverflow.com/a/39946744</span>

<span class="sd">    Args:</span>
<span class="sd">        df : pandas.DataFrame</span>
<span class="sd">            dataframe with the column to split and expand</span>
<span class="sd">        column : str</span>
<span class="sd">            the column to split and expand</span>
<span class="sd">        sep : str</span>
<span class="sd">            the string used to split the column&#39;s values</span>
<span class="sd">        keep : bool</span>
<span class="sd">            whether to retain the presplit value as it&#39;s own row</span>

<span class="sd">    Returns:</span>
<span class="sd">        pandas.DataFrame</span>
<span class="sd">            Returns a dataframe with the same columns as `df`.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">indexes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
    <span class="n">new_values</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="n">column</span><span class="p">])</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">presplit</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">column</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)):</span>
        <span class="n">values</span> <span class="o">=</span> <span class="n">presplit</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">sep</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">keep</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">values</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">indexes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
            <span class="n">new_values</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">presplit</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">values</span><span class="p">:</span>
            <span class="n">indexes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
            <span class="n">new_values</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
    <span class="n">new_df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">indexes</span><span class="p">,</span> <span class="p">:]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">new_df</span><span class="p">[</span><span class="n">column</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_values</span>
    <span class="k">return</span> <span class="n">new_df</span></div>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">doctest</span>
    <span class="n">doctest</span><span class="o">.</span><span class="n">testmod</span><span class="p">()</span>
</pre></div>

          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="../../index.html">
    <img class="logo" src="../../_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">dms_tools2</h1>
    
  </a>
</p>



<p class="blurb">Tools for analyzing deep mutational scanning data.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=dms_tools2&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a class="badge" href="https://travis-ci.org/jbloomlab/dms_tools2">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../examples.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../bcsubamp.html">Barcoded-subamplicon sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../prefs.html">Amino-acid preferences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../diffsel.html">Differential selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../fracsurvive.html">Fraction surviving</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../programs.html">Executable programs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api.html">Python API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../citations.html">Citations</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../index.html">Documentation overview</a><ul>
  <li><a href="../index.html">Module code</a><ul>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017--2018, the Bloom lab.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
    </div>

    
    <a href="https://github.com/jbloomlab/dms_tools2" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>