
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>dms_tools2.codonvarianttable &#8212; dms_tools2 2.6.1 documentation</title>
    <link rel="stylesheet" href="../../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <script type="text/javascript" src="../../_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
   
  <link rel="stylesheet" href="../../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <h1>Source code for dms_tools2.codonvarianttable</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">=================</span>
<span class="sd">codonvarianttable</span>
<span class="sd">=================</span>

<span class="sd">This module defines :class:`CodonVariantTable` objects</span>
<span class="sd">for storing and handling codon variants of a gene. They are</span>
<span class="sd">useful when you are performing deep mutational scanning</span>
<span class="sd">on barcoded codon-mutant libraries of a gene.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">math</span>
<span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">import</span> <span class="nn">tempfile</span>
<span class="kn">import</span> <span class="nn">random</span>

<span class="kn">import</span> <span class="nn">scipy</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">Bio.SeqUtils.ProtParamData</span>
<span class="kn">import</span> <span class="nn">gpmap</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="c1"># use plotnine for plotting</span>
<span class="kn">from</span> <span class="nn">plotnine</span> <span class="k">import</span> <span class="o">*</span>

<span class="kn">from</span> <span class="nn">dms_tools2.plot</span> <span class="k">import</span> <span class="n">latexSciNot</span>
<span class="kn">from</span> <span class="nn">dms_tools2</span> <span class="k">import</span> <span class="p">(</span><span class="n">CODON_TO_AA</span><span class="p">,</span>
                        <span class="n">CODONS</span><span class="p">,</span>
                        <span class="n">AAS_WITHSTOP</span><span class="p">,</span>
                        <span class="n">AA_TO_CODONS</span><span class="p">,</span>
                        <span class="n">NTS</span>
                        <span class="p">)</span>

<span class="c1">#: `color-blind safe palette &lt;http://bconnelly.net/2013/10/creating-colorblind-friendly-figures/&gt;`_</span>
<span class="n">CBPALETTE</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;#999999&quot;</span><span class="p">,</span> <span class="s2">&quot;#E69F00&quot;</span><span class="p">,</span> <span class="s2">&quot;#56B4E9&quot;</span><span class="p">,</span> <span class="s2">&quot;#009E73&quot;</span><span class="p">,</span>
             <span class="s2">&quot;#F0E442&quot;</span><span class="p">,</span> <span class="s2">&quot;#0072B2&quot;</span><span class="p">,</span> <span class="s2">&quot;#D55E00&quot;</span><span class="p">,</span> <span class="s2">&quot;#CC79A7&quot;</span><span class="p">]</span>


<div class="viewcode-block" id="CodonVariantTable"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable">[docs]</a><span class="k">class</span> <span class="nc">CodonVariantTable</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;Associates barcodes with codon mutants of gene.</span>

<span class="sd">    Args:</span>
<span class="sd">        `barcode_variant_file` (str)</span>
<span class="sd">            CSV file giving barcodes and variants. Must have</span>
<span class="sd">            columns named &quot;library&quot;, &quot;barcode&quot;, &quot;substitutions&quot;,</span>
<span class="sd">            (nucleotide mutations in 1, ... numbering in a format</span>
<span class="sd">            like &quot;G301A A302T G856C&quot;), and &quot;variant_call_support&quot;</span>
<span class="sd">            (sequences supporting barcode-variant call). Any</span>
<span class="sd">            additional columns are removed unless they are specified</span>
<span class="sd">            in `extra_cols`.</span>
<span class="sd">        `geneseq` (str)</span>
<span class="sd">            Sequence of wildtype protein-coding gene.</span>
<span class="sd">        `substitutions_are_codon` (bool)</span>
<span class="sd">            If `True`, then the &quot;substitutions&quot; column in</span>
<span class="sd">            `barcode_variant_file` gives the substitutions</span>
<span class="sd">            as codon rather than nucleotide mutations (e.g.,</span>
<span class="sd">            &quot;ATG1ATA GTA5CCC&quot; for substitutions at codons</span>
<span class="sd">            1 and 5 in 1, 2, ... numbering).</span>
<span class="sd">        `extra_cols` (list)</span>
<span class="sd">            Additional columns in `barcode_variant_file` to</span>
<span class="sd">            retain when creating `barcode_variant_df` and</span>
<span class="sd">            `variant_count_df` attributes.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        `geneseq` (str)</span>
<span class="sd">            `geneseq` passed at initialization.</span>
<span class="sd">        `sites` (list)</span>
<span class="sd">            List of all codon sites in 1, 2, ... numbering.</span>
<span class="sd">        `codons` (OrderedDict)</span>
<span class="sd">            `codons[r]` is wildtype codon at site `r`, ordered</span>
<span class="sd">            by sites.</span>
<span class="sd">        `aas` (OrderedDict)</span>
<span class="sd">            `aas[r]` is wildtype amino acid at site `r`,</span>
<span class="sd">            ordered by sites.</span>
<span class="sd">        `libraries` (list)</span>
<span class="sd">            List of libraries in `barcode_variantfile`.</span>
<span class="sd">        `barcode_variant_df` (pandas DataFrame)</span>
<span class="sd">            Info about codon mutations parsed from `barcode_variantfile`.</span>
<span class="sd">        `variant_count_df` (pandas DataFrame or None)</span>
<span class="sd">            Initially `None`, but after data are added with</span>
<span class="sd">            :class:`CodonVariantTable.addSampleCounts`, holds</span>
<span class="sd">            counts of each variant for each sample. Differs from</span>
<span class="sd">            `barcode_variant_df` in that the former just holds</span>
<span class="sd">            initial barcode-variant data, whereas `variant_count_df`</span>
<span class="sd">            is updated with variant counts for samples.</span>

<span class="sd">    Here is an example.</span>

<span class="sd">    First, initialize a :class:`CodonVariantTable`:</span>

<span class="sd">    &gt;&gt;&gt; geneseq = &#39;ATGGGATGA&#39;</span>
<span class="sd">    &gt;&gt;&gt; variantfile = &#39;_variantfile.csv&#39;</span>
<span class="sd">    &gt;&gt;&gt; with open(variantfile, &#39;w&#39;) as f:</span>
<span class="sd">    ...     _ = f.write(</span>
<span class="sd">    ...           &#39;library,barcode,substitutions,variant_call_support\\n&#39;</span>
<span class="sd">    ...           &#39;lib_1,AAC,,2\\n&#39;</span>
<span class="sd">    ...           &#39;lib_1,GAT,G4C A6C,1\\n&#39;</span>
<span class="sd">    ...           &#39;lib_2,AAC,T2A G4T,2\\n&#39;</span>
<span class="sd">    ...           &#39;lib_2,CAT,A6C,3&#39;</span>
<span class="sd">    ...           )</span>
<span class="sd">    &gt;&gt;&gt; variants = CodonVariantTable(</span>
<span class="sd">    ...             barcode_variant_file=variantfile,</span>
<span class="sd">    ...             geneseq=geneseq</span>
<span class="sd">    ...             )</span>
<span class="sd">    &gt;&gt;&gt; os.remove(variantfile)</span>

<span class="sd">    Check attributes of the :class:`CodonVariantTable`:</span>

<span class="sd">    &gt;&gt;&gt; variants.sites</span>
<span class="sd">    [1, 2, 3]</span>
<span class="sd">    &gt;&gt;&gt; variants.codons == {1:&#39;ATG&#39;, 2:&#39;GGA&#39;, 3:&#39;TGA&#39;}</span>
<span class="sd">    True</span>
<span class="sd">    &gt;&gt;&gt; variants.aas == {1:&#39;M&#39;, 2:&#39;G&#39;, 3:&#39;*&#39;}</span>
<span class="sd">    True</span>
<span class="sd">    &gt;&gt;&gt; variants.libraries</span>
<span class="sd">    [&#39;lib_1&#39;, &#39;lib_2&#39;]</span>
<span class="sd">    &gt;&gt;&gt; variants.valid_barcodes(&#39;lib_1&#39;) == {&#39;AAC&#39;, &#39;GAT&#39;}</span>
<span class="sd">    True</span>
<span class="sd">    &gt;&gt;&gt; variants.valid_barcodes(&#39;lib_2&#39;) == {&#39;AAC&#39;, &#39;CAT&#39;}</span>
<span class="sd">    True</span>
<span class="sd">    &gt;&gt;&gt; pd.set_option(&#39;display.max_columns&#39;, 15)</span>
<span class="sd">    &gt;&gt;&gt; pd.set_option(&#39;display.width&#39;, 500)</span>
<span class="sd">    &gt;&gt;&gt; variants.barcode_variant_df</span>
<span class="sd">      library barcode  variant_call_support codon_substitutions aa_substitutions  n_codon_substitutions  n_aa_substitutions</span>
<span class="sd">    0   lib_1     AAC                     2                                                           0                   0</span>
<span class="sd">    1   lib_1     GAT                     1             GGA2CGC              G2R                      1                   1</span>
<span class="sd">    2   lib_2     AAC                     2     ATG1AAG GGA2TGA          M1K G2*                      2                   2</span>
<span class="sd">    3   lib_2     CAT                     3             GGA2GGC                                       1                   0</span>

<span class="sd">    We can also look at the number of variants; we get this</span>
<span class="sd">    by calling :class:`CodonVariantTable.n_variants_df` with</span>
<span class="sd">    `samples=None` since we don&#39;t have samples, and just want</span>
<span class="sd">    stats across barcoded variants:</span>

<span class="sd">    &gt;&gt;&gt; variants.n_variants_df(samples=None)</span>
<span class="sd">             library             sample  count</span>
<span class="sd">    0          lib_1  barcoded variants      2</span>
<span class="sd">    1          lib_2  barcoded variants      2</span>
<span class="sd">    2  all libraries  barcoded variants      4</span>

<span class="sd">    Here is the number of variants if we look at just single amino-acid</span>
<span class="sd">    mutants (and wildtype):</span>

<span class="sd">    &gt;&gt;&gt; variants.n_variants_df(samples=None, variant_type=&#39;single&#39;, mut_type=&#39;aa&#39;)</span>
<span class="sd">             library             sample  count</span>
<span class="sd">    0          lib_1  barcoded variants      2</span>
<span class="sd">    1          lib_2  barcoded variants      1</span>
<span class="sd">    2  all libraries  barcoded variants      3</span>

<span class="sd">    We can also see how these numbers change if we require a</span>
<span class="sd">    variant call support of at least 2:</span>

<span class="sd">    &gt;&gt;&gt; variants.n_variants_df(samples=None, min_support=2)</span>
<span class="sd">             library             sample  count</span>
<span class="sd">    0          lib_1  barcoded variants      1</span>
<span class="sd">    1          lib_2  barcoded variants      2</span>
<span class="sd">    2  all libraries  barcoded variants      3</span>

<span class="sd">    If we want to combine the data for the two libraries, we can use</span>
<span class="sd">    :class:`CodonVariantTable.addMergedLibraries`, which creates a</span>
<span class="sd">    new combined library called &quot;all libraries&quot;:</span>

<span class="sd">    &gt;&gt;&gt; variants.addMergedLibraries(variants.barcode_variant_df)</span>
<span class="sd">             library    barcode  variant_call_support codon_substitutions aa_substitutions  n_codon_substitutions  n_aa_substitutions</span>
<span class="sd">    0          lib_1        AAC                     2                                                           0                   0</span>
<span class="sd">    1          lib_1        GAT                     1             GGA2CGC              G2R                      1                   1</span>
<span class="sd">    2          lib_2        AAC                     2     ATG1AAG GGA2TGA          M1K G2*                      2                   2</span>
<span class="sd">    3          lib_2        CAT                     3             GGA2GGC                                       1                   0</span>
<span class="sd">    4  all libraries  lib_1-AAC                     2                                                           0                   0</span>
<span class="sd">    5  all libraries  lib_1-GAT                     1             GGA2CGC              G2R                      1                   1</span>
<span class="sd">    6  all libraries  lib_2-AAC                     2     ATG1AAG GGA2TGA          M1K G2*                      2                   2</span>
<span class="sd">    7  all libraries  lib_2-CAT                     3             GGA2GGC                                       1                   0</span>

<span class="sd">    Note however that :class:`CodonVariantTable.addMergedLibraries`</span>
<span class="sd">    doesn&#39;t do anything if there is only one library:</span>

<span class="sd">    &gt;&gt;&gt; variants.addMergedLibraries(variants.barcode_variant_df</span>
<span class="sd">    ...                             .query(&#39;library == &quot;lib_1&quot;&#39;))</span>
<span class="sd">      library barcode  variant_call_support codon_substitutions aa_substitutions  n_codon_substitutions  n_aa_substitutions</span>
<span class="sd">    0   lib_1     AAC                     2                                                           0                   0</span>
<span class="sd">    1   lib_1     GAT                     1             GGA2CGC              G2R                      1                   1</span>

<span class="sd">    Count number of barcoded variants with each mutation:</span>

<span class="sd">    &gt;&gt;&gt; variants.mutCounts(&#39;all&#39;, &#39;aa&#39;, samples=None)[ : 2]</span>
<span class="sd">      library             sample mutation  count  mutation_type  site</span>
<span class="sd">    0   lib_1  barcoded variants      G2R      1  nonsynonymous     2</span>
<span class="sd">    1   lib_1  barcoded variants      *3A      0  nonsynonymous     3</span>

<span class="sd">    We can do the same for codon mutations (here for only a</span>
<span class="sd">    single library), first for all variants:</span>

<span class="sd">    &gt;&gt;&gt; variants.mutCounts(&#39;all&#39;, &#39;codon&#39;, samples=None,</span>
<span class="sd">    ...         libraries=[&#39;lib_2&#39;])[ : 4]</span>
<span class="sd">      library             sample mutation  count  mutation_type  site</span>
<span class="sd">    0   lib_2  barcoded variants  ATG1AAG      1  nonsynonymous     1</span>
<span class="sd">    1   lib_2  barcoded variants  GGA2GGC      1     synonymous     2</span>
<span class="sd">    2   lib_2  barcoded variants  GGA2TGA      1           stop     2</span>
<span class="sd">    3   lib_2  barcoded variants  ATG1AAA      0  nonsynonymous     1</span>

<span class="sd">    Then for just single-mutant variants:</span>

<span class="sd">    &gt;&gt;&gt; variants.mutCounts(&#39;single&#39;, &#39;codon&#39;, samples=None,</span>
<span class="sd">    ...         libraries=[&#39;lib_2&#39;])[ : 3]</span>
<span class="sd">      library             sample mutation  count  mutation_type  site</span>
<span class="sd">    0   lib_2  barcoded variants  GGA2GGC      1     synonymous     2</span>
<span class="sd">    1   lib_2  barcoded variants  ATG1AAA      0  nonsynonymous     1</span>
<span class="sd">    2   lib_2  barcoded variants  ATG1AAC      0  nonsynonymous     1</span>

<span class="sd">    Initially we haven&#39;t added any barcode count information</span>
<span class="sd">    for any samples:</span>

<span class="sd">    &gt;&gt;&gt; all(variants.samples(lib) == [] for lib in variants.libraries)</span>
<span class="sd">    True</span>
<span class="sd">    &gt;&gt;&gt; variants.variant_count_df is None</span>
<span class="sd">    True</span>

<span class="sd">    Now we add barcode count information for sample &quot;input&quot;</span>
<span class="sd">    from library 1 using :class:`CodonVariantTable.addSampleCounts`:</span>

<span class="sd">    &gt;&gt;&gt; counts_lib1_input = pd.DataFrame(</span>
<span class="sd">    ...         {&#39;barcode&#39;:[&#39;AAC&#39;, &#39;GAT&#39;],</span>
<span class="sd">    ...          &#39;count&#39;  :[  253,  1101]})</span>
<span class="sd">    &gt;&gt;&gt; variants.addSampleCounts(&#39;lib_1&#39;, &#39;input&#39;, counts_lib1_input)</span>
<span class="sd">    &gt;&gt;&gt; variants.variant_count_df</span>
<span class="sd">      barcode  count library sample  variant_call_support codon_substitutions aa_substitutions  n_codon_substitutions  n_aa_substitutions</span>
<span class="sd">    0     GAT   1101   lib_1  input                     1             GGA2CGC              G2R                      1                   1</span>
<span class="sd">    1     AAC    253   lib_1  input                     2                                                           0                   0</span>

<span class="sd">    We get an error if we try to add these same data again,</span>
<span class="sd">    as they are already added for that sample to that library:</span>

<span class="sd">    &gt;&gt;&gt; variants.addSampleCounts(&#39;lib_1&#39;, &#39;input&#39;, counts_lib1_input)</span>
<span class="sd">    Traceback (most recent call last):</span>
<span class="sd">    ...</span>
<span class="sd">    ValueError: `library` lib_1 already has `sample` input</span>

<span class="sd">    But, we can add barcode counts for another</span>
<span class="sd">    sample (named &quot;selected&quot; in this case) to library 1:</span>

<span class="sd">    &gt;&gt;&gt; counts_lib1_selected = pd.DataFrame(</span>
<span class="sd">    ...         {&#39;barcode&#39;:[&#39;AAC&#39;, &#39;GAT&#39;],</span>
<span class="sd">    ...          &#39;count&#39;  :[  513,  401]})</span>
<span class="sd">    &gt;&gt;&gt; variants.addSampleCounts(&#39;lib_1&#39;, &#39;selected&#39;, counts_lib1_selected)</span>

<span class="sd">    As well as barcode counts for the same two samples</span>
<span class="sd">    (&quot;input&quot; and &quot;selected&quot;) to our other library (library 2):</span>

<span class="sd">    &gt;&gt;&gt; counts_lib2_input = pd.DataFrame(</span>
<span class="sd">    ...         {&#39;barcode&#39;:[&#39;AAC&#39;, &#39;CAT&#39;],</span>
<span class="sd">    ...          &#39;count&#39;  :[ 1253,  923]})</span>
<span class="sd">    &gt;&gt;&gt; variants.addSampleCounts(&#39;lib_2&#39;, &#39;input&#39;, counts_lib2_input)</span>
<span class="sd">    &gt;&gt;&gt; counts_lib2_selected = pd.DataFrame(</span>
<span class="sd">    ...         {&#39;barcode&#39;:[&#39;AAC&#39;, &#39;CAT&#39;],</span>
<span class="sd">    ...          &#39;count&#39;  :[  113,  1200]})</span>
<span class="sd">    &gt;&gt;&gt; variants.addSampleCounts(&#39;lib_2&#39;, &#39;selected&#39;, counts_lib2_selected)</span>
<span class="sd">    &gt;&gt;&gt; variants.variant_count_df</span>
<span class="sd">      barcode  count library    sample  variant_call_support codon_substitutions aa_substitutions  n_codon_substitutions  n_aa_substitutions</span>
<span class="sd">    0     GAT   1101   lib_1     input                     1             GGA2CGC              G2R                      1                   1</span>
<span class="sd">    1     AAC    253   lib_1     input                     2                                                           0                   0</span>
<span class="sd">    2     AAC    513   lib_1  selected                     2                                                           0                   0</span>
<span class="sd">    3     GAT    401   lib_1  selected                     1             GGA2CGC              G2R                      1                   1</span>
<span class="sd">    4     AAC   1253   lib_2     input                     2     ATG1AAG GGA2TGA          M1K G2*                      2                   2</span>
<span class="sd">    5     CAT    923   lib_2     input                     3             GGA2GGC                                       1                   0</span>
<span class="sd">    6     CAT   1200   lib_2  selected                     3             GGA2GGC                                       1                   0</span>
<span class="sd">    7     AAC    113   lib_2  selected                     2     ATG1AAG GGA2TGA          M1K G2*                      2                   2</span>

<span class="sd">    We can also use :class:`CodonVariantTable.mutCounts`</span>
<span class="sd">    to look at total counts of each mutation:</span>

<span class="sd">    &gt;&gt;&gt; variants.mutCounts(&#39;all&#39;, &#39;aa&#39;)[ : 2]</span>
<span class="sd">      library sample mutation  count  mutation_type  site</span>
<span class="sd">    0   lib_1  input      G2R   1101  nonsynonymous     2</span>
<span class="sd">    1   lib_1  input      *3A      0  nonsynonymous     3</span>

<span class="sd">    &gt;&gt;&gt; variants.mutCounts(&#39;all&#39;, &#39;aa&#39;, libraries=[&#39;lib_2&#39;])[ : 3]</span>
<span class="sd">      library sample mutation  count  mutation_type  site</span>
<span class="sd">    0   lib_2  input      G2*   1253           stop     2</span>
<span class="sd">    1   lib_2  input      M1K   1253  nonsynonymous     1</span>
<span class="sd">    2   lib_2  input      *3A      0  nonsynonymous     3</span>

<span class="sd">    We can use :class:`CodonVariantTable.writeCodonCounts` to</span>
<span class="sd">    write codon count files. First for only **single** mutants:</span>

<span class="sd">    &gt;&gt;&gt; with tempfile.TemporaryDirectory() as tmpdir:</span>
<span class="sd">    ...     countfiles = variants.writeCodonCounts(&quot;single&quot;, outdir=tmpdir, include_all_libs=True)</span>
<span class="sd">    ...     lib1_input = pd.read_csv(f&#39;{tmpdir}/lib_1_input_codoncounts.csv&#39;)</span>
<span class="sd">    ...     all_sel = pd.read_csv(f&#39;{tmpdir}/all-libraries_selected_codoncounts.csv&#39;)</span>

<span class="sd">    Make sure we created expected countfiles:</span>

<span class="sd">    &gt;&gt;&gt; countfiles.assign(countfile=lambda x: x.countfile.apply(os.path.basename))</span>
<span class="sd">             library    sample                               countfile</span>
<span class="sd">    0          lib_1     input             lib_1_input_codoncounts.csv</span>
<span class="sd">    1          lib_1  selected          lib_1_selected_codoncounts.csv</span>
<span class="sd">    2          lib_2     input             lib_2_input_codoncounts.csv</span>
<span class="sd">    3          lib_2  selected          lib_2_selected_codoncounts.csv</span>
<span class="sd">    4  all-libraries     input     all-libraries_input_codoncounts.csv</span>
<span class="sd">    5  all-libraries  selected  all-libraries_selected_codoncounts.csv</span>

<span class="sd">    Check for expected values in a few of these counts files, only</span>
<span class="sd">    showing columns with non-zero entries:</span>

<span class="sd">    &gt;&gt;&gt; lib1_input.iloc[:, (lib1_input != 0).any(axis=&#39;rows&#39;).values]</span>
<span class="sd">       site wildtype  ATG   CGC  GGA  TGA</span>
<span class="sd">    0     1      ATG  253     0    0    0</span>
<span class="sd">    1     2      GGA    0  1101  253    0</span>
<span class="sd">    2     3      TGA    0     0    0  253</span>
<span class="sd">    &gt;&gt;&gt; all_sel.iloc[:, (all_sel != 0).any(axis=&#39;rows&#39;).values]</span>
<span class="sd">       site wildtype  ATG  CGC  GGA   GGC  TGA</span>
<span class="sd">    0     1      ATG  513    0    0     0    0</span>
<span class="sd">    1     2      GGA    0  401  513  1200    0</span>
<span class="sd">    2     3      TGA    0    0    0     0  513</span>

<span class="sd">    Now write codon counts files for **all** mutants:</span>

<span class="sd">    &gt;&gt;&gt; with tempfile.TemporaryDirectory() as tmpdir:</span>
<span class="sd">    ...     _ = variants.writeCodonCounts(&quot;all&quot;, outdir=tmpdir, include_all_libs=True)</span>
<span class="sd">    ...     lib1_input_all = pd.read_csv(f&#39;{tmpdir}/lib_1_input_codoncounts.csv&#39;)</span>
<span class="sd">    ...     all_sel_all = pd.read_csv(f&#39;{tmpdir}/all-libraries_selected_codoncounts.csv&#39;)</span>
<span class="sd">    &gt;&gt;&gt; lib1_input_all.iloc[:, (lib1_input_all != 0).any(axis=&#39;rows&#39;).values]</span>
<span class="sd">       site wildtype   ATG   CGC  GGA   TGA</span>
<span class="sd">    0     1      ATG  1354     0    0     0</span>
<span class="sd">    1     2      GGA     0  1101  253     0</span>
<span class="sd">    2     3      TGA     0     0    0  1354</span>
<span class="sd">    &gt;&gt;&gt; all_sel_all.iloc[:, (all_sel_all != 0).any(axis=&#39;rows&#39;).values]</span>
<span class="sd">       site wildtype  AAG   ATG  CGC  GGA   GGC   TGA</span>
<span class="sd">    0     1      ATG  113  2114    0    0     0     0</span>
<span class="sd">    1     2      GGA    0     0  401  513  1200   113</span>
<span class="sd">    2     3      TGA    0     0    0    0     0  2227</span>

<span class="sd">    We can also initialize a :class:`CodonVariantTable` from the</span>
<span class="sd">    `variant_count_df` if we have written that to a CSV file.</span>
<span class="sd">    We do this using :meth:`CodonVariantTable.from_variant_count_df`.</span>
<span class="sd">    The example below shows how this newly initialized variant table</span>
<span class="sd">    is equal to the original one used to write the CSV file:</span>

<span class="sd">    &gt;&gt;&gt; with tempfile.NamedTemporaryFile(mode=&#39;w&#39;) as f:</span>
<span class="sd">    ...     variants.variant_count_df.to_csv(f, index=False)</span>
<span class="sd">    ...     f.flush()</span>
<span class="sd">    ...     variants_eq = CodonVariantTable.from_variant_count_df(</span>
<span class="sd">    ...                     variant_count_df_file=f.name,</span>
<span class="sd">    ...                     geneseq=geneseq)</span>
<span class="sd">    &gt;&gt;&gt; variants == variants_eq</span>
<span class="sd">    True</span>

<span class="sd">    Of course, the initialized variant table is **not** equal</span>
<span class="sd">    to original one if we don&#39;t write the full `variant_count_df`</span>
<span class="sd">    to the CSV file:</span>

<span class="sd">    &gt;&gt;&gt; with tempfile.NamedTemporaryFile(mode=&#39;w&#39;) as f:</span>
<span class="sd">    ...     (variants</span>
<span class="sd">    ...         .variant_count_df.query(&#39;sample == &quot;input&quot;&#39;)</span>
<span class="sd">    ...         .to_csv(f, index=False)</span>
<span class="sd">    ...         )</span>
<span class="sd">    ...     f.flush()</span>
<span class="sd">    ...     variants_ne = CodonVariantTable.from_variant_count_df(</span>
<span class="sd">    ...                     variant_count_df_file=f.name,</span>
<span class="sd">    ...                     geneseq=geneseq)</span>
<span class="sd">    &gt;&gt;&gt; variants == variants_ne</span>
<span class="sd">    False</span>

<span class="sd">    We can use :meth:`CodonVariantTable.func_scores` to compute</span>
<span class="sd">    the functional effects of mutations. We cannot use this method</span>
<span class="sd">    with default options as we have no wildtype counts (needed for</span>
<span class="sd">    normalization) for `lib_2` so we get an error:</span>

<span class="sd">    &gt;&gt;&gt; variants.func_scores(&#39;input&#39;)</span>
<span class="sd">    Traceback (most recent call last):</span>
<span class="sd">    ...</span>
<span class="sd">    ValueError: no wildtype counts:</span>
<span class="sd">      library    sample  count</span>
<span class="sd">    0   lib_1     input    253</span>
<span class="sd">    1   lib_1  selected    513</span>
<span class="sd">    2   lib_2     input      0</span>
<span class="sd">    3   lib_2  selected      0</span>

<span class="sd">    However, we can use the method with the `permit_zero_wt` option:</span>

<span class="sd">    &gt;&gt;&gt; variants.func_scores(&#39;input&#39;, permit_zero_wt=True)</span>
<span class="sd">      library pre_sample post_sample barcode  func_score  func_score_var  pre_count  post_count  pre_count_wt  post_count_wt  pseudocount codon_substitutions  n_codon_substitutions aa_substitutions  n_aa_substitutions</span>
<span class="sd">    0   lib_1      input    selected     AAC    0.000000        0.024528        253         513           253            513          0.5                                          0                                    0</span>
<span class="sd">    1   lib_1      input    selected     GAT   -2.474376        0.019337       1101         401           253            513          0.5             GGA2CGC                      1              G2R                   1</span>
<span class="sd">    2   lib_2      input    selected     AAC   -3.465198        8.345474       1253         113             0              0          0.5     ATG1AAG GGA2TGA                      2          M1K G2*                   2</span>
<span class="sd">    3   lib_2      input    selected     CAT    0.378452        8.329463        923        1200             0              0          0.5             GGA2GGC                      1                                    0</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__eq__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">):</span>
        <span class="c1"># following here: https://stackoverflow.com/a/390640</span>
        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">other</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
            <span class="k">return</span> <span class="kc">False</span>
        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__dict__</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="o">!=</span> <span class="n">other</span><span class="o">.</span><span class="vm">__dict__</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="k">return</span> <span class="kc">False</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">val</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__dict__</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="n">val2</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="n">key</span><span class="p">)</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">val</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">val</span><span class="o">.</span><span class="n">equals</span><span class="p">(</span><span class="n">val2</span><span class="p">):</span>
                        <span class="k">return</span> <span class="kc">False</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">val</span> <span class="o">!=</span> <span class="n">val2</span><span class="p">:</span>
                        <span class="k">return</span> <span class="kc">False</span>
            <span class="k">return</span> <span class="kc">True</span>

<div class="viewcode-block" id="CodonVariantTable.from_simulation"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.from_simulation">[docs]</a>    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">from_simulation</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">geneseq</span><span class="p">,</span> <span class="n">bclen</span><span class="p">,</span> <span class="n">library_specs</span><span class="p">,</span>
                        <span class="n">seed</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">variant_call_support</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Simulate :class:`CodonVariantTable` variants.</span>

<span class="sd">        Use this method to simulate the variants in a</span>
<span class="sd">        :class:`CodonVariantTable`. Note that this only</span>
<span class="sd">        simulates the variants, not counts for samples.</span>
<span class="sd">        To add those, then use :func:`simulateSampleCounts`</span>
<span class="sd">        and :meth:`CodonVariantTable.addSampleCounts`.</span>

<span class="sd">        Args:</span>
<span class="sd">            `geneseq` (str)</span>
<span class="sd">                Sequence of wildtype protein-coding gene.</span>
<span class="sd">            `bclen` (int)</span>
<span class="sd">                Length of the barcodes; must enable complexity</span>
<span class="sd">                at least 10-fold greater than max number of variants.</span>
<span class="sd">            `library_specs` (dict)</span>
<span class="sd">                Specifications for each simulated library. Keys</span>
<span class="sd">                are &#39;avgmuts&#39; and &#39;nvariants&#39;, and values are</span>
<span class="sd">                average-codon mutations per variant and number of</span>
<span class="sd">                variants. Mutations per variant are Poisson distributed.</span>
<span class="sd">            `seed` (int or `None`)</span>
<span class="sd">                Random number seed or `None` to set no seed.</span>
<span class="sd">            `variant_call_support` (int or 2-tuple)</span>
<span class="sd">                If an integer, all variant call supports are set to this.</span>
<span class="sd">                If a 2-tuple, variant call support is drawn as a random</span>
<span class="sd">                integer uniformly from this range (inclusive).</span>

<span class="sd">        Returns:</span>
<span class="sd">            The simulated :class:`CodonVariantTable`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">seed</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">scipy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
            <span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">library_specs</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;empty `library_specs`&#39;</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">variant_call_support</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
            <span class="n">variant_call_support</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">([</span><span class="n">variant_call_support</span><span class="p">]</span> <span class="o">*</span> <span class="mi">2</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">geneseq</span><span class="p">)</span> <span class="o">%</span> <span class="mi">3</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;length of `geneseq` not multiple of 3&#39;</span><span class="p">)</span>
        <span class="n">genelength</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">geneseq</span><span class="p">)</span> <span class="o">//</span> <span class="mi">3</span>

        <span class="n">barcode_variant_dict</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">lib</span><span class="p">,</span> <span class="n">specs_dict</span> <span class="ow">in</span> <span class="n">library_specs</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>

            <span class="n">nvariants</span> <span class="o">=</span> <span class="n">specs_dict</span><span class="p">[</span><span class="s1">&#39;nvariants&#39;</span><span class="p">]</span>
            <span class="n">avgmuts</span> <span class="o">=</span> <span class="n">specs_dict</span><span class="p">[</span><span class="s1">&#39;avgmuts&#39;</span><span class="p">]</span>
            <span class="k">if</span> <span class="mi">10</span> <span class="o">*</span> <span class="n">nvariants</span> <span class="o">&gt;</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">NTS</span><span class="p">))</span><span class="o">**</span><span class="n">bclen</span><span class="p">:</span>  <span class="c1"># safety factor 10</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;barcode too short for nvariants&#39;</span><span class="p">)</span>
            <span class="n">existing_barcodes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([])</span>

            <span class="k">for</span> <span class="n">ivariant</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nvariants</span><span class="p">):</span>

                <span class="n">barcode</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">choices</span><span class="p">(</span><span class="n">NTS</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="n">bclen</span><span class="p">))</span>
                <span class="k">while</span> <span class="n">barcode</span> <span class="ow">in</span> <span class="n">existing_barcodes</span><span class="p">:</span>
                    <span class="n">barcode</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">choices</span><span class="p">(</span><span class="n">NTS</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="n">bclen</span><span class="p">))</span>
                <span class="n">existing_barcodes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">barcode</span><span class="p">)</span>

                <span class="n">support</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="o">*</span><span class="n">variant_call_support</span><span class="p">)</span>

                <span class="c1"># get mutations</span>
                <span class="n">substitutions</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="n">nmuts</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">avgmuts</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">icodon</span> <span class="ow">in</span> <span class="n">random</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">genelength</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">nmuts</span><span class="p">):</span>
                    <span class="n">wtcodon</span> <span class="o">=</span> <span class="n">geneseq</span><span class="p">[</span><span class="mi">3</span> <span class="o">*</span> <span class="p">(</span><span class="n">icodon</span> <span class="o">-</span> <span class="mi">1</span><span class="p">):</span> <span class="mi">3</span> <span class="o">*</span> <span class="n">icodon</span><span class="p">]</span>
                    <span class="n">mutcodon</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">([</span><span class="n">c</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">CODONS</span>
                                              <span class="k">if</span> <span class="n">c</span> <span class="o">!=</span> <span class="n">wtcodon</span><span class="p">])</span>
                    <span class="k">for</span> <span class="n">i_nt</span><span class="p">,</span> <span class="p">(</span><span class="n">wt_nt</span><span class="p">,</span> <span class="n">mut_nt</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">wtcodon</span><span class="p">,</span>
                                                               <span class="n">mutcodon</span><span class="p">)):</span>
                        <span class="k">if</span> <span class="n">wt_nt</span> <span class="o">!=</span> <span class="n">mut_nt</span><span class="p">:</span>
                            <span class="n">igene</span> <span class="o">=</span> <span class="mi">3</span> <span class="o">*</span> <span class="p">(</span><span class="n">icodon</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">+</span> <span class="n">i_nt</span> <span class="o">+</span> <span class="mi">1</span>
                            <span class="n">substitutions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{wt_nt}{igene}{mut_nt}</span><span class="s1">&#39;</span><span class="p">)</span>
                <span class="n">substitutions</span> <span class="o">=</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">substitutions</span><span class="p">)</span>

                <span class="n">barcode_variant_dict</span><span class="p">[</span><span class="s1">&#39;barcode&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">barcode</span><span class="p">)</span>
                <span class="n">barcode_variant_dict</span><span class="p">[</span><span class="s1">&#39;substitutions&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">substitutions</span><span class="p">)</span>
                <span class="n">barcode_variant_dict</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lib</span><span class="p">)</span>
                <span class="n">barcode_variant_dict</span><span class="p">[</span><span class="s1">&#39;variant_call_support&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">support</span><span class="p">)</span>

        <span class="n">barcode_variants</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">barcode_variant_dict</span><span class="p">)</span>

        <span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">barcode_variants</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">f</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
            <span class="n">cvt</span> <span class="o">=</span> <span class="bp">cls</span><span class="p">(</span><span class="n">barcode_variant_file</span><span class="o">=</span><span class="n">f</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">geneseq</span><span class="o">=</span><span class="n">geneseq</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">cvt</span></div>

<div class="viewcode-block" id="CodonVariantTable.from_variant_count_df"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.from_variant_count_df">[docs]</a>    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">from_variant_count_df</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">variant_count_df_file</span><span class="p">,</span> <span class="n">geneseq</span><span class="p">,</span>
            <span class="n">drop_all_libs</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;:class:`CodonVariantTable` from CSV of `variant_count_df`.</span>

<span class="sd">        Use this method when you have written a CSV file of the</span>
<span class="sd">        `variant_count_df` attribute of a :class:`CodonVariantTable`,</span>
<span class="sd">        and now wish to re-initialize that :class:`CodonVariantTable`.</span>

<span class="sd">        Args:</span>
<span class="sd">            `variant_count_df_file` (str)</span>
<span class="sd">                Name of CSV file containing the `variant_count_df`.</span>
<span class="sd">                Must have following columns: &quot;barcode&quot;, &quot;library&quot;,</span>
<span class="sd">                &quot;variant_call_support&quot;, &quot;codon_substitutions&quot;,</span>
<span class="sd">                &quot;sample&quot;, and &quot;count&quot;.</span>
<span class="sd">            `geneseq` (str)</span>
<span class="sd">                Sequence of wildtype protein-coding gene.</span>
<span class="sd">            `drop_all_libs` (bool)</span>
<span class="sd">                If there is a library named &quot;all libraries&quot;,</span>
<span class="sd">                drop it from the list of libraries in the created</span>
<span class="sd">                :class:`CodonVariantTable` as it probably was</span>
<span class="sd">                added by :meth:`CodonVariantTable.addMergedLibraries`</span>
<span class="sd">                and duplicates information for the individual libraries.</span>

<span class="sd">        Returns:</span>
<span class="sd">            The :class:`CodonVariantTable` used to write</span>
<span class="sd">            `variant_count_df_file`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">variant_count_df_file</span><span class="p">)</span>

        <span class="n">req_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;variant_call_support&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;codon_substitutions&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">req_cols</span><span class="p">)</span> <span class="o">&lt;</span> <span class="nb">set</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;</span><span class="si">{variant_count_df}</span><span class="s2"> lacks required &quot;</span>
                             <span class="n">f</span><span class="s2">&quot;columns </span><span class="si">{req_cols}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">req_cols</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">drop_all_libs</span><span class="p">:</span>
            <span class="n">dropcol</span> <span class="o">=</span> <span class="s2">&quot;all libraries&quot;</span>
            <span class="k">if</span> <span class="n">dropcol</span> <span class="ow">in</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
                <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;library != @dropcol&#39;</span><span class="p">)</span>

        <span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="p">(</span><span class="n">df</span>
                <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">])</span>
                <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;codon_substitutions&#39;</span><span class="p">:</span><span class="s1">&#39;substitutions&#39;</span><span class="p">})</span>
                <span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
                <span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
                <span class="p">)</span>
            <span class="n">f</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
            <span class="n">cvt</span> <span class="o">=</span> <span class="bp">cls</span><span class="p">(</span><span class="n">barcode_variant_file</span><span class="o">=</span><span class="n">f</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
                      <span class="n">geneseq</span><span class="o">=</span><span class="n">geneseq</span><span class="p">,</span>
                      <span class="n">substitutions_are_codon</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">lib</span> <span class="ow">in</span> <span class="n">cvt</span><span class="o">.</span><span class="n">libraries</span><span class="p">:</span>
                <span class="n">idf</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;sample == @sample &amp; library == @lib&#39;</span><span class="p">)</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">idf</span><span class="p">):</span>
                    <span class="n">cvt</span><span class="o">.</span><span class="n">addSampleCounts</span><span class="p">(</span><span class="n">lib</span><span class="p">,</span>
                                        <span class="n">sample</span><span class="p">,</span>
                                        <span class="n">idf</span><span class="p">[[</span><span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">]]</span>
                                        <span class="p">)</span>

        <span class="k">return</span> <span class="n">cvt</span></div>


    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">barcode_variant_file</span><span class="p">,</span> <span class="n">geneseq</span><span class="p">,</span>
                 <span class="n">substitutions_are_codon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">extra_cols</span><span class="o">=</span><span class="p">[]):</span>
        <span class="sd">&quot;&quot;&quot;See main class doc string.&quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">geneseq</span> <span class="o">=</span> <span class="n">geneseq</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s1">&#39;^[ATGC]+$&#39;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">geneseq</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid nucleotides in </span><span class="si">{self.geneseq}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">geneseq</span><span class="p">)</span> <span class="o">%</span> <span class="mi">3</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">or</span> <span class="nb">len</span><span class="p">(</span><span class="n">geneseq</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;`geneseq` of invalid length {len(self.geneseq)}&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sites</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">geneseq</span><span class="p">)</span> <span class="o">//</span> <span class="mi">3</span> <span class="o">+</span> <span class="mi">1</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">codons</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">OrderedDict</span><span class="p">([</span>
                <span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">geneseq</span><span class="p">[</span><span class="mi">3</span> <span class="o">*</span> <span class="p">(</span><span class="n">r</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="p">:</span> <span class="mi">3</span> <span class="o">*</span> <span class="n">r</span><span class="p">])</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sites</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">aas</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">OrderedDict</span><span class="p">([</span>
                <span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="n">CODON_TO_AA</span><span class="p">[</span><span class="n">codon</span><span class="p">])</span> <span class="k">for</span> <span class="n">r</span><span class="p">,</span> <span class="n">codon</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">codons</span><span class="o">.</span><span class="n">items</span><span class="p">()])</span>

        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">barcode_variant_file</span><span class="p">)</span>
        <span class="n">required_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;barcode&#39;</span><span class="p">,</span>
                         <span class="s1">&#39;substitutions&#39;</span><span class="p">,</span> <span class="s1">&#39;variant_call_support&#39;</span><span class="p">]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">set</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span><span class="o">.</span><span class="n">issuperset</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">required_cols</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;`variantfile` does not have &quot;</span>
                             <span class="n">f</span><span class="s2">&quot;required columns </span><span class="si">{required_cols}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">extra_cols</span> <span class="ow">and</span> <span class="ow">not</span> <span class="nb">set</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span><span class="o">.</span><span class="n">issuperset</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">extra_cols</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;`variantfile` does not have &quot;</span>
                             <span class="n">f</span><span class="s2">&quot;`extra_cols` </span><span class="si">{extra_cols}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">required_cols</span> <span class="o">+</span> <span class="n">extra_cols</span><span class="p">]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">libraries</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">library</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_valid_barcodes</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">lib</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">libraries</span><span class="p">:</span>
            <span class="n">barcodes</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;library == @lib&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">barcode</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">barcodes</span><span class="p">))</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">barcodes</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;duplicated barcodes for </span><span class="si">{lib}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_valid_barcodes</span><span class="p">[</span><span class="n">lib</span><span class="p">]</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">barcodes</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_samples</span> <span class="o">=</span> <span class="p">{</span><span class="n">lib</span><span class="p">:[]</span> <span class="k">for</span> <span class="n">lib</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">libraries</span><span class="p">}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">variant_count_df</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="k">if</span> <span class="n">substitutions_are_codon</span><span class="p">:</span>
            <span class="n">codonSubsFunc</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sortCodonMuts</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">codonSubsFunc</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_ntToCodonMuts</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">barcode_variant_df</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">df</span>
                <span class="c1"># info about codon and amino-acid substitutions</span>
                <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">codon_substitutions</span><span class="o">=</span>
                            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">substitutions</span>
                                       <span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
                                       <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">codonSubsFunc</span><span class="p">),</span>
                        <span class="n">aa_substitutions</span><span class="o">=</span>
                            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">codon_substitutions</span>
                                       <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">codonToAAMuts</span><span class="p">),</span>
                        <span class="n">n_codon_substitutions</span><span class="o">=</span>
                            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">codon_substitutions</span>
                                       <span class="o">.</span><span class="n">str</span>
                                       <span class="o">.</span><span class="n">split</span><span class="p">()</span>
                                       <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="nb">len</span><span class="p">),</span>
                        <span class="n">n_aa_substitutions</span><span class="o">=</span>
                            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">aa_substitutions</span>
                                       <span class="o">.</span><span class="n">str</span>
                                       <span class="o">.</span><span class="n">split</span><span class="p">()</span>
                                       <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="nb">len</span><span class="p">)</span>
                        <span class="p">)</span>
                <span class="c1"># we no longer need initial `substitutions` column</span>
                <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s1">&#39;substitutions&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="s1">&#39;columns&#39;</span><span class="p">)</span>
                <span class="c1"># sort to ensure consistent order</span>
                <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">library</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                                <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
                                    <span class="n">x</span><span class="o">.</span><span class="n">library</span><span class="p">,</span>
                                    <span class="n">categories</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">libraries</span><span class="p">,</span>
                                    <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span>
                                    <span class="p">)</span>
                        <span class="p">)</span>
                <span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;barcode&#39;</span><span class="p">])</span>
                <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                <span class="p">)</span>

        <span class="c1"># check validity of codon substitutions given `geneseq`</span>
        <span class="k">for</span> <span class="n">codonmut</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">chain</span><span class="o">.</span><span class="n">from_iterable</span><span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">barcode_variant_df</span>
                          <span class="o">.</span><span class="n">codon_substitutions</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">()):</span>
            <span class="n">m</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s1">&#39;^(?P&lt;wt&gt;[ATGC]</span><span class="si">{3}</span><span class="s1">)&#39;</span>
                          <span class="s1">&#39;(?P&lt;r&gt;\d+)&#39;</span>
                          <span class="s1">&#39;(?P&lt;mut&gt;[ATGC]</span><span class="si">{3}</span><span class="s1">)$&#39;</span><span class="p">,</span>
                         <span class="n">codonmut</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">m</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid mutation </span><span class="si">{codonmut}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">wt</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;wt&#39;</span><span class="p">)</span>
            <span class="n">r</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">))</span>
            <span class="n">mut</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;mut&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">r</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sites</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid site </span><span class="si">{r}</span><span class="s2"> in codon mutation </span><span class="si">{codonmut}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">codons</span><span class="p">[</span><span class="n">r</span><span class="p">]</span> <span class="o">!=</span> <span class="n">wt</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Wrong wildtype codon in </span><span class="si">{codonmut}</span><span class="s2">. &quot;</span>
                                 <span class="n">f</span><span class="s2">&quot;Expected wildtype of </span><span class="si">{codons[r]}</span><span class="s2">.&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">wt</span> <span class="o">==</span> <span class="n">mut</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid mutation </span><span class="si">{codonmut}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># define some colors for plotting</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_mutation_type_colors</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s1">&#39;nonsynonymous&#39;</span><span class="p">:</span><span class="n">CBPALETTE</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                <span class="s1">&#39;synonymous&#39;</span><span class="p">:</span><span class="n">CBPALETTE</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span>
                <span class="s1">&#39;stop&#39;</span><span class="p">:</span><span class="n">CBPALETTE</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
                <span class="p">}</span>


<div class="viewcode-block" id="CodonVariantTable.samples"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.samples">[docs]</a>    <span class="k">def</span> <span class="nf">samples</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">library</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;List of all samples for `library`.</span>

<span class="sd">        Args:</span>
<span class="sd">            `library` (str)</span>
<span class="sd">                Valid `library` for the :class:`CodonVariantTable`.</span>

<span class="sd">        Returns:</span>
<span class="sd">            List of all samples for which barcode counts have</span>
<span class="sd">            been added.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_samples</span><span class="p">[</span><span class="n">library</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid `library` </span><span class="si">{library}</span><span class="s2">&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="CodonVariantTable.addSampleCounts"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.addSampleCounts">[docs]</a>    <span class="k">def</span> <span class="nf">addSampleCounts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">library</span><span class="p">,</span> <span class="n">sample</span><span class="p">,</span> <span class="n">barcodecounts</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Add variant counts for a sample to `variant_count_df`.</span>

<span class="sd">        Args:</span>
<span class="sd">            `library` (str)</span>
<span class="sd">                Valid `library` for the :class:`CodonVariantTable`.</span>
<span class="sd">            `sample` (str)</span>
<span class="sd">                Sample name, must **not** already be in</span>
<span class="sd">                :class:`CodonVariantTable.samples` for `library`.</span>
<span class="sd">            `barcodecounts` (pandas DataFrame)</span>
<span class="sd">                Gives counts for each variant by barcode. Must</span>
<span class="sd">                have columns named &quot;barcode&quot; and &quot;count&quot;. The</span>
<span class="sd">                &quot;barcode&quot; column must contain all the barcodes</span>
<span class="sd">                in :class:`CodonVariantTable.valid_barcodes` for</span>
<span class="sd">                `library`. Such data frames are returned</span>
<span class="sd">                by :class:`dms_tools2.barcodes.IlluminaBarcodeParser.parse`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">library</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">libraries</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid library </span><span class="si">{library}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">sample</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">samples</span><span class="p">(</span><span class="n">library</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;`library` </span><span class="si">{library}</span><span class="s2"> already &quot;</span>
                             <span class="n">f</span><span class="s2">&quot;has `sample` </span><span class="si">{sample}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">req_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">set</span><span class="p">(</span><span class="n">barcodecounts</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span><span class="o">.</span><span class="n">issuperset</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">req_cols</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;`barcodecounts` lacks columns </span><span class="si">{req_cols}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">barcodecounts</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">barcodecounts</span><span class="o">.</span><span class="n">barcode</span><span class="o">.</span><span class="n">unique</span><span class="p">())):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;`barcodecounts` has non-unique barcodes&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">set</span><span class="p">(</span><span class="n">barcodecounts</span><span class="o">.</span><span class="n">barcode</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_barcodes</span><span class="p">(</span><span class="n">library</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;barcodes in `barcodecounts` do not match &quot;</span>
                             <span class="n">f</span><span class="s2">&quot;those expected for `library` </span><span class="si">{library}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_samples</span><span class="p">[</span><span class="n">library</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>

        <span class="n">df</span> <span class="o">=</span> <span class="p">(</span><span class="n">barcodecounts</span>
              <span class="p">[</span><span class="n">req_cols</span><span class="p">]</span>
              <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">library</span><span class="o">=</span><span class="n">library</span><span class="p">,</span> <span class="n">sample</span><span class="o">=</span><span class="n">sample</span><span class="p">)</span>
              <span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">barcode_variant_df</span><span class="p">,</span>
                     <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span><span class="p">,</span>
                     <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;barcode&#39;</span><span class="p">],</span>
                     <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                     <span class="n">validate</span><span class="o">=</span><span class="s1">&#39;one_to_one&#39;</span><span class="p">)</span>
              <span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">variant_count_df</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">variant_count_df</span> <span class="o">=</span> <span class="n">df</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">variant_count_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
                              <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">variant_count_df</span><span class="p">,</span> <span class="n">df</span><span class="p">],</span>
                              <span class="n">axis</span><span class="o">=</span><span class="s1">&#39;index&#39;</span><span class="p">,</span>
                              <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                              <span class="n">sort</span><span class="o">=</span><span class="kc">False</span>
                              <span class="p">)</span>

        <span class="c1"># samples in order added after ordering by library, getting</span>
        <span class="c1"># unique ones as here: https://stackoverflow.com/a/39835527</span>
        <span class="n">unique_samples</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">collections</span><span class="o">.</span><span class="n">OrderedDict</span><span class="o">.</span><span class="n">fromkeys</span><span class="p">(</span>
                <span class="n">itertools</span><span class="o">.</span><span class="n">chain</span><span class="o">.</span><span class="n">from_iterable</span><span class="p">(</span>
                    <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">samples</span><span class="p">(</span><span class="n">lib</span><span class="p">)</span> <span class="k">for</span> <span class="n">lib</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">libraries</span><span class="p">])</span>
                <span class="p">))</span>

        <span class="c1"># make library and sample categorical and sort</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">variant_count_df</span> <span class="o">=</span> <span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">variant_count_df</span>
                <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">library</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                                <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
                                    <span class="n">x</span><span class="o">.</span><span class="n">library</span><span class="p">,</span>
                                    <span class="n">categories</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">libraries</span><span class="p">,</span>
                                    <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span>
                                    <span class="p">),</span>
                        <span class="n">sample</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                               <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
                                    <span class="n">x</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">],</span>
                                    <span class="n">categories</span><span class="o">=</span><span class="n">unique_samples</span><span class="p">,</span>
                                    <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span>
                                    <span class="p">)</span>
                         <span class="p">)</span>
                <span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">],</span>
                             <span class="n">ascending</span><span class="o">=</span><span class="p">[</span><span class="kc">True</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">])</span>
                <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                <span class="p">)</span></div>


<div class="viewcode-block" id="CodonVariantTable.valid_barcodes"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.valid_barcodes">[docs]</a>    <span class="k">def</span> <span class="nf">valid_barcodes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">library</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Set of valid barcodes for `library`.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">library</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">libraries</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid `library` </span><span class="si">{library}</span><span class="s2">; &quot;</span>
                             <span class="n">f</span><span class="s2">&quot;valid libraries are </span><span class="si">{self.libraries}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_valid_barcodes</span><span class="p">[</span><span class="n">library</span><span class="p">]</span></div>


<div class="viewcode-block" id="CodonVariantTable.func_scores"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.func_scores">[docs]</a>    <span class="k">def</span> <span class="nf">func_scores</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">preselection</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span>
                <span class="n">pseudocount</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="s2">&quot;barcode&quot;</span><span class="p">,</span>
                <span class="n">combine_libs</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">syn_as_wt</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">logbase</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                <span class="n">permit_zero_wt</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">permit_self_comp</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get data frame with functional scores for variants.</span>

<span class="sd">        The functional score is calculated from the change in counts</span>
<span class="sd">        for a variant pre- and post-selection using the formula in</span>
<span class="sd">        `Otwinoski et al (2018) &lt;https://doi.org/10.1073/pnas.1804015115&gt;`_.</span>
<span class="sd">        Specifically, if :math:`n^v_{pre}` and :math:`n^v_{post}` are</span>
<span class="sd">        the counts of variant :math:`v` pre- and post-selection, and</span>
<span class="sd">        if :math:`n^{wt}_{pre}` and :math:`n^{wt}_{post}` are</span>
<span class="sd">        the summed counts of **all** wildtype variants pre- and post-</span>
<span class="sd">        selection, then the functional score of the variant is</span>
<span class="sd">        :math:`f_v = \log_b\left(\\frac{n^v_{post} / n^{wt}_{post}}{n^v_{pre} / n^{wt}_{pre}}\\right)`</span>
<span class="sd">        and the variance due to Poisson sampling statistics is</span>
<span class="sd">        :math:`\\frac{1}{\left(\ln b\\right)^2}\left(\sigma^2_v = \\frac{1}{n^v_{post}} + \\frac{1}{n^{wt}_{post}} + \\frac{1}{n^v_{pre}} + \\frac{1}{n^{wt}_{pre}}\\right)`</span>
<span class="sd">        where :math:`b` is logarithm base (see `logbase` argument).</span>
<span class="sd">        For both calculations, a pseudocount (see `pseudocount` argument)</span>
<span class="sd">        is added to each count first. The wildtype counts are computed</span>
<span class="sd">        across all **fully wildtype** variants (see `syn_as_wt` for</span>
<span class="sd">        how this is defined).</span>

<span class="sd">        Args:</span>
<span class="sd">            `preselection` (str or dict)</span>
<span class="sd">                The pre-selection sample. If the same for all post-</span>
<span class="sd">                selection samples, then provide the name of the</span>
<span class="sd">                pre-selection sample. If it differs among post-selection</span>
<span class="sd">                samples, then provide a dict keyed by each post-selection</span>
<span class="sd">                sample with the pre-selection sample being the value.</span>
<span class="sd">            `pseudocount` (float)</span>
<span class="sd">                The pseudocount added to each count.</span>
<span class="sd">            `by` (str)</span>
<span class="sd">                Compute effects for each &quot;barcode&quot;, set of</span>
<span class="sd">                &quot;aa_substitutions&quot;, or set of &quot;codon_substitutions&quot;.</span>
<span class="sd">                In the latter two cases, all barcodes with each</span>
<span class="sd">                set of substitutions are combined (see `combine_libs`).</span>
<span class="sd">                Note that if you use &quot;aa_substitutions&quot; then it may</span>
<span class="sd">                be more sensible to set `syn_as_wt` to `True`.</span>
<span class="sd">            `syn_as_wt` (bool)</span>
<span class="sd">                In the formula for the functional scores, consider</span>
<span class="sd">                variants with only synonymous mutations when determining</span>
<span class="sd">                wildtype counts? If `False`, only variants with **no**</span>
<span class="sd">                mutations of any type contribute to wildtype counts.</span>
<span class="sd">            `combine_libs` (bool)</span>
<span class="sd">                If `by` is &quot;aa_substitutions&quot; or &quot;codon_substitutions&quot;,</span>
<span class="sd">                do we combine across libraries as well as barcodes?</span>
<span class="sd">            `logbase` (float)</span>
<span class="sd">                Base for logarithm when calculating functional score.</span>
<span class="sd">            `permit_zero_wt` (bool)</span>
<span class="sd">                If the wildtype counts are zero for any sample, do</span>
<span class="sd">                we raise an error or permit the calculation to proceed</span>
<span class="sd">                just using the pseudocount?</span>
<span class="sd">            `permit_self_comp` (bool)</span>
<span class="sd">                Permit comparisons where the same sample is used as</span>
<span class="sd">                the pre- and post-selection?</span>

<span class="sd">        Returns:</span>
<span class="sd">            A pandas Data Frame with the following columns:</span>
<span class="sd">              - &quot;library&quot;: the library (&quot;all libraries&quot; if `combine_libs`)</span>
<span class="sd">              - &quot;pre_sample&quot;: the pre-selection sample</span>
<span class="sd">              - &quot;post_sample&quot;: the post-selection sample</span>
<span class="sd">              - the value corresponding to the grouping we are using</span>
<span class="sd">                to compute effects (the value of `by`)</span>
<span class="sd">              - &quot;func_score&quot;: the functional score</span>
<span class="sd">              - &quot;func_score_var&quot;: variance on the functional score</span>
<span class="sd">              - &quot;pre_count&quot;: pre-selection counts</span>
<span class="sd">              - &quot;post_count: post-selection counts</span>
<span class="sd">              - &quot;pre_count_wt&quot;: pre-selection counts for all wildtype</span>
<span class="sd">              - &quot;post_count_wt&quot;: post-selection counts for all wildtype</span>
<span class="sd">              - &quot;pseudocount&quot;: the pseudocount value</span>
<span class="sd">              - as many of &quot;aa_substitutions&quot;, &quot;n_aa_substitutions&quot;,</span>
<span class="sd">                &quot;codon_substitutions&quot;, and &quot;n_codon_substitutions&quot;</span>
<span class="sd">                as can be retained given the value of `by`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">ordered_samples</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">variant_count_df</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">preselection</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="c1"># make `preselection` into dict</span>
            <span class="n">preselection</span> <span class="o">=</span> <span class="p">{</span><span class="n">s</span><span class="p">:</span><span class="n">preselection</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">ordered_samples</span>
                            <span class="k">if</span> <span class="n">s</span> <span class="o">!=</span> <span class="n">preselection</span> <span class="ow">or</span> <span class="n">permit_self_comp</span><span class="p">}</span>
        <span class="k">elif</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">preselection</span><span class="p">,</span> <span class="nb">dict</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;`preselection` not str or dict&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">permit_self_comp</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="n">pre</span> <span class="o">==</span> <span class="n">post</span> <span class="k">for</span> <span class="n">pre</span><span class="p">,</span> <span class="n">post</span> <span class="ow">in</span> <span class="n">preselection</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;`permit_self_comp` is False but there&#39;</span>
                                 <span class="s1">&#39; are identical pre and post samples&#39;</span><span class="p">)</span>

        <span class="c1"># all samples of interest</span>
        <span class="n">samples</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">preselection</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">preselection</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">samples</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">ordered_samples</span><span class="p">)):</span>
            <span class="n">extra_samples</span> <span class="o">=</span> <span class="n">samples</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">ordered_samples</span><span class="p">)</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid samples: </span><span class="si">{extra_samples}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># get data frame with samples of interest</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">variant_count_df</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;no sample variant counts have been added&#39;</span><span class="p">)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">variant_count_df</span>
              <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;sample in @samples&#39;</span><span class="p">)</span>
              <span class="p">)</span>

        <span class="k">if</span> <span class="n">combine_libs</span> <span class="ow">and</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">libraries</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">):</span>
            <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="n">s</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">samples</span><span class="p">(</span><span class="n">lib</span><span class="p">)</span> <span class="k">for</span> <span class="n">lib</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">libraries</span>
                                              <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">samples</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;cannot use `combine_libs`, not every &#39;</span>
                                 <span class="n">f</span><span class="s2">&quot;library has every sample: </span><span class="si">{samples}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">df</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">addMergedLibraries</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
                  <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;library == &quot;all libraries&quot;&#39;</span><span class="p">)</span>
                  <span class="p">)</span>

        <span class="c1"># get wildtype counts for each sample and library</span>
        <span class="k">if</span> <span class="n">syn_as_wt</span><span class="p">:</span>
            <span class="n">wt_col</span> <span class="o">=</span> <span class="s1">&#39;n_aa_substitutions&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">wt_col</span> <span class="o">=</span> <span class="s1">&#39;n_codon_substitutions&#39;</span>
        <span class="n">wt_counts</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">df</span>
                <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">count</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;count&#39;</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="mi">0</span> <span class="o">==</span> <span class="n">x</span><span class="p">[</span><span class="n">wt_col</span><span class="p">])</span>
                                        <span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;int&#39;</span><span class="p">))</span>
                <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">])</span>
                <span class="o">.</span><span class="n">aggregate</span><span class="p">({</span><span class="s1">&#39;count&#39;</span><span class="p">:</span><span class="s1">&#39;sum&#39;</span><span class="p">})</span>
                <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
                <span class="p">)</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">wt_counts</span><span class="p">[</span><span class="s1">&#39;count&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">()</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">permit_zero_wt</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;no wildtype counts:</span><span class="se">\n</span><span class="si">{wt_counts}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># sum counts in groups specified by `by`</span>
        <span class="n">group_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;codon_substitutions&#39;</span><span class="p">,</span> <span class="s1">&#39;n_codon_substitutions&#39;</span><span class="p">,</span>
                      <span class="s1">&#39;aa_substitutions&#39;</span><span class="p">,</span> <span class="s1">&#39;n_aa_substitutions&#39;</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">by</span> <span class="ow">in</span> <span class="p">{</span><span class="s1">&#39;aa_substitutions&#39;</span><span class="p">,</span> <span class="s1">&#39;codon_substitutions&#39;</span><span class="p">}:</span>
            <span class="n">group_cols</span> <span class="o">=</span> <span class="n">group_cols</span><span class="p">[</span><span class="n">group_cols</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">by</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span> <span class="p">:</span> <span class="p">]</span>
        <span class="k">elif</span> <span class="n">by</span> <span class="o">!=</span> <span class="s1">&#39;barcode&#39;</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid `by` of </span><span class="si">{by}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="p">(</span><span class="n">df</span>
              <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="n">by</span><span class="p">]</span> <span class="o">+</span> <span class="n">group_cols</span><span class="p">)</span>
              <span class="o">.</span><span class="n">aggregate</span><span class="p">({</span><span class="s1">&#39;count&#39;</span><span class="p">:</span><span class="s1">&#39;sum&#39;</span><span class="p">})</span>
              <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
              <span class="p">)</span>

        <span class="c1"># get data frame with pre- and post-selection samples / counts</span>
        <span class="n">df_func_scores</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">post_sample</span> <span class="ow">in</span> <span class="n">ordered_samples</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">post_sample</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">preselection</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">pre_sample</span> <span class="o">=</span> <span class="n">preselection</span><span class="p">[</span><span class="n">post_sample</span><span class="p">]</span>
            <span class="n">sample_dfs</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">stype</span><span class="p">,</span> <span class="n">s</span> <span class="ow">in</span> <span class="p">[(</span><span class="s1">&#39;pre&#39;</span><span class="p">,</span> <span class="n">pre_sample</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;post&#39;</span><span class="p">,</span> <span class="n">post_sample</span><span class="p">)]:</span>
                <span class="n">sample_dfs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="n">df</span>
                        <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;sample == @s&#39;</span><span class="p">)</span>
                        <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;count&#39;</span><span class="p">:</span><span class="n">f</span><span class="s2">&quot;</span><span class="si">{stype}</span><span class="s2">_count&quot;</span><span class="p">})</span>
                        <span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">wt_counts</span>
                               <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;count&#39;</span><span class="p">:</span><span class="n">f</span><span class="s2">&quot;</span><span class="si">{stype}</span><span class="s2">_count_wt&quot;</span><span class="p">}),</span>
                               <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span><span class="p">,</span> <span class="n">validate</span><span class="o">=</span><span class="s1">&#39;many_to_one&#39;</span>
                               <span class="p">)</span>
                        <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;sample&#39;</span><span class="p">:</span><span class="n">f</span><span class="s2">&quot;</span><span class="si">{stype}</span><span class="s2">_sample&quot;</span><span class="p">})</span>
                        <span class="p">)</span>
            <span class="n">df_func_scores</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">sample_dfs</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">sample_dfs</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                             <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span><span class="p">,</span> <span class="n">validate</span><span class="o">=</span><span class="s1">&#39;1:1&#39;</span><span class="p">)</span>
                    <span class="p">)</span>
        <span class="n">df_func_scores</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">df_func_scores</span><span class="p">,</span>
                                   <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="c1"># check pseudocount</span>
        <span class="k">if</span> <span class="n">pseudocount</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;`pseudocount` is &lt; 0: </span><span class="si">{pseudocount}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="p">(</span><span class="n">pseudocount</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">any</span><span class="p">((</span><span class="n">df_func_scores</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">()</span>
                                    <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;pre_count&#39;</span><span class="p">,</span> <span class="s1">&#39;post_count&#39;</span><span class="p">,</span>
                                    <span class="s1">&#39;pre_count_wt&#39;</span><span class="p">,</span> <span class="s1">&#39;post_count_wt&#39;</span><span class="p">]):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;some counts are zero, you must use &#39;</span>
                             <span class="s1">&#39;`pseudocount` &gt; 0&#39;</span><span class="p">)</span>

        <span class="c1"># calculate functional score and variance</span>
        <span class="n">df_func_scores</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">df_func_scores</span>
                <span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                    <span class="n">pseudocount</span><span class="o">=</span><span class="n">pseudocount</span><span class="p">,</span>
                    <span class="n">func_score</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                               <span class="n">scipy</span><span class="o">.</span><span class="n">log</span><span class="p">(</span>
                                    <span class="p">((</span><span class="n">x</span><span class="o">.</span><span class="n">post_count</span> <span class="o">+</span> <span class="n">x</span><span class="o">.</span><span class="n">pseudocount</span><span class="p">)</span> <span class="o">/</span>
                                     <span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">post_count_wt</span> <span class="o">+</span> <span class="n">x</span><span class="o">.</span><span class="n">pseudocount</span><span class="p">))</span> <span class="o">/</span>
                                    <span class="p">((</span><span class="n">x</span><span class="o">.</span><span class="n">pre_count</span> <span class="o">+</span> <span class="n">x</span><span class="o">.</span><span class="n">pseudocount</span><span class="p">)</span> <span class="o">/</span>
                                     <span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">pre_count_wt</span> <span class="o">+</span> <span class="n">x</span><span class="o">.</span><span class="n">pseudocount</span><span class="p">))</span>
                               <span class="p">)</span> <span class="o">/</span> <span class="n">scipy</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">logbase</span><span class="p">),</span>
                    <span class="n">func_score_var</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                               <span class="p">(</span><span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">post_count</span> <span class="o">+</span> <span class="n">x</span><span class="o">.</span><span class="n">pseudocount</span><span class="p">)</span> <span class="o">+</span>
                                <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">post_count_wt</span> <span class="o">+</span> <span class="n">x</span><span class="o">.</span><span class="n">pseudocount</span><span class="p">)</span> <span class="o">+</span>
                                <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">pre_count</span> <span class="o">+</span> <span class="n">x</span><span class="o">.</span><span class="n">pseudocount</span><span class="p">)</span> <span class="o">+</span>
                                <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">pre_count_wt</span> <span class="o">+</span> <span class="n">x</span><span class="o">.</span><span class="n">pseudocount</span><span class="p">)</span>
                               <span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">scipy</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">logbase</span><span class="p">)</span><span class="o">**</span><span class="mi">2</span><span class="p">)</span>
                    <span class="p">)</span>
                <span class="c1"># set column order in data frame</span>
                <span class="p">[[</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;pre_sample&#39;</span><span class="p">,</span> <span class="s1">&#39;post_sample&#39;</span><span class="p">,</span> <span class="n">by</span><span class="p">,</span>
                  <span class="s1">&#39;func_score&#39;</span><span class="p">,</span> <span class="s1">&#39;func_score_var&#39;</span><span class="p">,</span> <span class="s1">&#39;pre_count&#39;</span><span class="p">,</span>
                  <span class="s1">&#39;post_count&#39;</span><span class="p">,</span> <span class="s1">&#39;pre_count_wt&#39;</span><span class="p">,</span> <span class="s1">&#39;post_count_wt&#39;</span><span class="p">,</span>
                  <span class="s1">&#39;pseudocount&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">group_cols</span><span class="p">]</span>
                <span class="p">)</span>

        <span class="k">return</span> <span class="n">df_func_scores</span></div>


<div class="viewcode-block" id="CodonVariantTable.n_variants_df"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.n_variants_df">[docs]</a>    <span class="k">def</span> <span class="nf">n_variants_df</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">libraries</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span>
                      <span class="n">min_support</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">variant_type</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span>
                      <span class="n">mut_type</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Number of variants per library / sample.</span>

<span class="sd">        Args:</span>
<span class="sd">            `variant_type` (&quot;single&quot; or &quot;all&quot;)</span>
<span class="sd">                Include just variants with &lt;= 1 `mut_type` mutation.</span>
<span class="sd">            `mut_type` (&quot;aa&quot;, &quot;codon&quot;, or `None`)</span>
<span class="sd">                If `variant_type` is single, set to &quot;codon&quot; or &quot;aa&quot;</span>
<span class="sd">                to indicate how to filter for single mutants.</span>
<span class="sd">            All other args:</span>
<span class="sd">                Same meaning as for</span>
<span class="sd">                :class:`CodonVariantTable.plotNumMutsHistogram`.</span>

<span class="sd">        Returns:</span>
<span class="sd">            DataFrame giving number of variants per library /</span>
<span class="sd">            sample.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span><span class="p">,</span> <span class="n">nlibraries</span><span class="p">,</span> <span class="n">nsamples</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_getPlotData</span><span class="p">(</span><span class="n">libraries</span><span class="p">,</span>
                                                     <span class="n">samples</span><span class="p">,</span>
                                                     <span class="n">min_support</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">variant_type</span> <span class="o">==</span> <span class="s1">&#39;single&#39;</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">mut_type</span> <span class="ow">in</span> <span class="p">{</span><span class="s1">&#39;aa&#39;</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">}:</span>
                <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;n_</span><span class="si">{mut_type}</span><span class="s2">_substitutions &lt;= 1&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;`mut_type` must be &quot;aa&quot; or &quot;single&quot;&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">variant_type</span> <span class="o">!=</span> <span class="s1">&#39;all&#39;</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid `variant_type` </span><span class="si">{variant_type}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="p">(</span><span class="n">df</span>
                <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">])</span>
                <span class="o">.</span><span class="n">aggregate</span><span class="p">({</span><span class="s1">&#39;count&#39;</span><span class="p">:</span><span class="s1">&#39;sum&#39;</span><span class="p">})</span>
                <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
                <span class="p">)</span></div>


<div class="viewcode-block" id="CodonVariantTable.mutCounts"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.mutCounts">[docs]</a>    <span class="k">def</span> <span class="nf">mutCounts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">variant_type</span><span class="p">,</span> <span class="n">mut_type</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span>
            <span class="n">libraries</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">min_support</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get counts of each individual mutation.</span>

<span class="sd">        Args:</span>
<span class="sd">            `variant_type` (&quot;single&quot; or &quot;all&quot;)</span>
<span class="sd">                Include just single mutants, or all mutants?</span>
<span class="sd">            Other args:</span>
<span class="sd">                Same meaning as for</span>
<span class="sd">                :class:`CodonVariantTable.plotNumMutsHistogram`</span>

<span class="sd">        Returns:</span>
<span class="sd">            A tidy data frame with columns named &quot;library&quot;,</span>
<span class="sd">            &quot;sample&quot;, &quot;mutation&quot;, &quot;count&quot;, &quot;mutation_type&quot;,</span>
<span class="sd">            and &quot;site&quot;.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span><span class="p">,</span> <span class="n">nlibraries</span><span class="p">,</span> <span class="n">nsamples</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_getPlotData</span><span class="p">(</span><span class="n">libraries</span><span class="p">,</span>
                                                     <span class="n">samples</span><span class="p">,</span>
                                                     <span class="n">min_support</span><span class="p">)</span>

        <span class="n">samplelist</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="n">librarylist</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">mut_type</span> <span class="o">==</span> <span class="s1">&#39;codon&#39;</span><span class="p">:</span>
            <span class="n">wts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">codons</span>
            <span class="n">chars</span> <span class="o">=</span> <span class="n">CODONS</span>
            <span class="n">mutation_types</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;nonsynonymous&#39;</span><span class="p">,</span> <span class="s1">&#39;synonymous&#39;</span><span class="p">,</span> <span class="s1">&#39;stop&#39;</span><span class="p">]</span>
        <span class="k">elif</span> <span class="n">mut_type</span> <span class="o">==</span> <span class="s1">&#39;aa&#39;</span><span class="p">:</span>
            <span class="n">wts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">aas</span>
            <span class="n">chars</span> <span class="o">=</span> <span class="n">AAS_WITHSTOP</span>
            <span class="n">mutation_types</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;nonsynonymous&#39;</span><span class="p">,</span> <span class="s1">&#39;stop&#39;</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid mut_type </span><span class="si">{mut_type}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># data frame listing all mutations with count 0</span>
        <span class="n">mut_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">r</span><span class="p">,</span> <span class="n">wt</span> <span class="ow">in</span> <span class="n">wts</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">chars</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">mut</span> <span class="o">!=</span> <span class="n">wt</span><span class="p">:</span>
                    <span class="n">mut_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{wt}{r}{mut}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">all_muts</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span>
                    <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;mutation&#39;</span><span class="p">:</span><span class="n">mut_list</span><span class="p">,</span>
                                     <span class="s1">&#39;library&#39;</span><span class="p">:</span><span class="n">library</span><span class="p">,</span>
                                     <span class="s1">&#39;sample&#39;</span><span class="p">:</span><span class="n">sample</span><span class="p">,</span>
                                     <span class="s1">&#39;count&#39;</span><span class="p">:</span><span class="mi">0</span><span class="p">})</span>
                    <span class="k">for</span> <span class="n">library</span><span class="p">,</span> <span class="n">sample</span> <span class="ow">in</span>
                    <span class="n">itertools</span><span class="o">.</span><span class="n">product</span><span class="p">(</span><span class="n">librarylist</span><span class="p">,</span> <span class="n">samplelist</span><span class="p">)])</span>

        <span class="k">if</span> <span class="n">variant_type</span> <span class="o">==</span> <span class="s1">&#39;single&#39;</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;n_</span><span class="si">{mut_type}</span><span class="s1">_substitutions == 1&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">variant_type</span> <span class="o">==</span> <span class="s1">&#39;all&#39;</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;n_</span><span class="si">{mut_type}</span><span class="s1">_substitutions &gt;= 1&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid variant_type </span><span class="si">{variant_type}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">def</span> <span class="nf">_classify_mutation</span><span class="p">(</span><span class="n">mut_str</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">mut_type</span> <span class="o">==</span> <span class="s1">&#39;aa&#39;</span><span class="p">:</span>
                <span class="n">m</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s1">&#39;^(?P&lt;wt&gt;[A-Z\*])\d+(?P&lt;mut&gt;[A-Z\*])$&#39;</span><span class="p">,</span>
                             <span class="n">mut_str</span><span class="p">)</span>
                <span class="n">wt_aa</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;wt&#39;</span><span class="p">)</span>
                <span class="n">mut_aa</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;mut&#39;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">m</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s1">&#39;^(?P&lt;wt&gt;[ACTG]</span><span class="si">{3}</span><span class="s1">)\d+(?P&lt;mut&gt;[ACTG]</span><span class="si">{3}</span><span class="s1">)$&#39;</span><span class="p">,</span>
                             <span class="n">mut_str</span><span class="p">)</span>
                <span class="n">wt_aa</span> <span class="o">=</span> <span class="n">CODON_TO_AA</span><span class="p">[</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;wt&#39;</span><span class="p">)]</span>
                <span class="n">mut_aa</span> <span class="o">=</span> <span class="n">CODON_TO_AA</span><span class="p">[</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;mut&#39;</span><span class="p">)]</span>
            <span class="k">if</span> <span class="n">wt_aa</span> <span class="o">==</span> <span class="n">mut_aa</span><span class="p">:</span>
                <span class="k">return</span> <span class="s1">&#39;synonymous&#39;</span>
            <span class="k">elif</span> <span class="n">mut_aa</span> <span class="o">==</span> <span class="s1">&#39;*&#39;</span><span class="p">:</span>
                <span class="k">return</span> <span class="s1">&#39;stop&#39;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">return</span> <span class="s1">&#39;nonsynonymous&#39;</span>

        <span class="k">def</span> <span class="nf">_get_site</span><span class="p">(</span><span class="n">mut_str</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">mut_type</span> <span class="o">==</span> <span class="s1">&#39;aa&#39;</span><span class="p">:</span>
                <span class="n">m</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s1">&#39;^[A-Z\*](?P&lt;site&gt;\d+)[A-Z\*]$&#39;</span><span class="p">,</span> <span class="n">mut_str</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">m</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s1">&#39;^[ACTG]</span><span class="si">{3}</span><span class="s1">(?P&lt;site&gt;\d+)[ACTG]</span><span class="si">{3}</span><span class="s1">$&#39;</span><span class="p">,</span> <span class="n">mut_str</span><span class="p">)</span>
            <span class="n">site</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;site&#39;</span><span class="p">))</span>
            <span class="k">assert</span> <span class="n">site</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sites</span>
            <span class="k">return</span> <span class="n">site</span>

        <span class="n">df</span> <span class="o">=</span> <span class="p">(</span><span class="n">df</span>
              <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span>
                  <span class="p">{</span><span class="n">f</span><span class="s2">&quot;</span><span class="si">{mut_type}</span><span class="s2">_substitutions&quot;</span><span class="p">:</span><span class="s2">&quot;mutation&quot;</span><span class="p">}</span>
                  <span class="p">)</span>
              <span class="p">[[</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;mutation&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">]]</span>
              <span class="o">.</span><span class="n">pipe</span><span class="p">(</span><span class="n">tidy_split</span><span class="p">,</span> <span class="n">column</span><span class="o">=</span><span class="s1">&#39;mutation&#39;</span><span class="p">)</span>
              <span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">all_muts</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;outer&#39;</span><span class="p">)</span>
              <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;mutation&#39;</span><span class="p">])</span>
              <span class="o">.</span><span class="n">aggregate</span><span class="p">({</span><span class="s1">&#39;count&#39;</span><span class="p">:</span><span class="s1">&#39;sum&#39;</span><span class="p">})</span>
              <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
              <span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                <span class="n">library</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                         <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
                          <span class="n">x</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">],</span>
                          <span class="n">librarylist</span><span class="p">,</span>
                          <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
                <span class="n">sample</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                         <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
                          <span class="n">x</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">],</span>
                          <span class="n">samplelist</span><span class="p">,</span>
                          <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
                <span class="n">mutation_type</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                         <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
                          <span class="n">x</span><span class="p">[</span><span class="s1">&#39;mutation&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">_classify_mutation</span><span class="p">),</span>
                          <span class="n">mutation_types</span><span class="p">,</span>
                          <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
                <span class="n">site</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;mutation&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">_get_site</span><span class="p">),</span>
                <span class="p">)</span>
              <span class="o">.</span><span class="n">sort_values</span><span class="p">(</span>
                <span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">,</span> <span class="s1">&#39;mutation&#39;</span><span class="p">],</span>
                <span class="n">ascending</span><span class="o">=</span><span class="p">[</span><span class="kc">True</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">])</span>
              <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
              <span class="p">)</span>

        <span class="k">return</span> <span class="n">df</span></div>


<div class="viewcode-block" id="CodonVariantTable.plotMutHeatmap"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.plotMutHeatmap">[docs]</a>    <span class="k">def</span> <span class="nf">plotMutHeatmap</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">variant_type</span><span class="p">,</span> <span class="n">mut_type</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span>
            <span class="n">count_or_frequency</span><span class="o">=</span><span class="s1">&#39;frequency&#39;</span><span class="p">,</span>
            <span class="n">libraries</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">plotfile</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">orientation</span><span class="o">=</span><span class="s1">&#39;h&#39;</span><span class="p">,</span> <span class="n">widthscale</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">heightscale</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
            <span class="n">min_support</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Heatmap of mutation counts or frequencies.</span>

<span class="sd">        Args:</span>
<span class="sd">            `count_or_frequency` (&quot;count&quot; or &quot;frequency&quot;)</span>
<span class="sd">                Plot mutation counts or frequencies?</span>
<span class="sd">            All other args</span>
<span class="sd">                Same meaning as for</span>
<span class="sd">                :meth:`CodonVariantTable.plotCumulMutCoverage`.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A `plotnine &lt;https://plotnine.readthedocs.io&gt;`_</span>
<span class="sd">            plot; can be displayed in a Jupyter notebook with `p.draw()`.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutCounts</span><span class="p">(</span><span class="n">variant_type</span><span class="p">,</span> <span class="n">mut_type</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="n">samples</span><span class="p">,</span>
                            <span class="n">libraries</span><span class="o">=</span><span class="n">libraries</span><span class="p">,</span> <span class="n">min_support</span><span class="o">=</span><span class="n">min_support</span><span class="p">)</span>

        <span class="n">n_variants</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">n_variants_df</span><span class="p">(</span><span class="n">libraries</span><span class="o">=</span><span class="n">libraries</span><span class="p">,</span>
                                         <span class="n">samples</span><span class="o">=</span><span class="n">samples</span><span class="p">,</span>
                                         <span class="n">min_support</span><span class="o">=</span><span class="n">min_support</span><span class="p">,</span>
                                         <span class="n">variant_type</span><span class="o">=</span><span class="n">variant_type</span><span class="p">,</span>
                                         <span class="n">mut_type</span><span class="o">=</span><span class="n">mut_type</span><span class="p">)</span>
                      <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;count&#39;</span><span class="p">:</span><span class="s1">&#39;nseqs&#39;</span><span class="p">})</span>
                      <span class="p">)</span>

        <span class="c1"># order amino acids by Kyte-Doolittle hydrophobicity,</span>
        <span class="c1"># codons by the amino acid they encode</span>
        <span class="n">aa_order</span> <span class="o">=</span> <span class="p">[</span><span class="n">tup</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">tup</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span>
                    <span class="n">Bio</span><span class="o">.</span><span class="n">SeqUtils</span><span class="o">.</span><span class="n">ProtParamData</span><span class="o">.</span><span class="n">kd</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span>
                    <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">tup</span><span class="p">:</span> <span class="n">tup</span><span class="p">[</span><span class="mi">1</span><span class="p">])]</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;*&#39;</span><span class="p">]</span>
        <span class="n">codon_order</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">itertools</span><span class="o">.</span><span class="n">chain</span><span class="o">.</span><span class="n">from_iterable</span><span class="p">(</span>
                        <span class="p">[</span><span class="n">AA_TO_CODONS</span><span class="p">[</span><span class="n">aa</span><span class="p">]</span> <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">aa_order</span><span class="p">]))</span>

        <span class="n">df</span> <span class="o">=</span> <span class="p">(</span><span class="n">df</span>
              <span class="p">[[</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;mutation&#39;</span><span class="p">,</span> <span class="s1">&#39;site&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">]]</span>
              <span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">n_variants</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">])</span>
              <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">frequency</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;count&#39;</span><span class="p">]</span> <span class="o">/</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;nseqs&#39;</span><span class="p">],</span>
                      <span class="n">mut_char</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                        <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
                         <span class="n">x</span><span class="o">.</span><span class="n">mutation</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span>
                            <span class="s1">&#39;^[A-Z\*]+\d+(?P&lt;mut_char&gt;[A-Z\*]+)$&#39;</span><span class="p">)</span>
                            <span class="o">.</span><span class="n">mut_char</span><span class="p">,</span>
                         <span class="p">{</span><span class="s1">&#39;aa&#39;</span><span class="p">:</span><span class="n">aa_order</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">:</span><span class="n">codon_order</span><span class="p">}[</span><span class="n">mut_type</span><span class="p">],</span>
                         <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                      <span class="p">)</span>
              <span class="p">)</span>

        <span class="k">if</span> <span class="n">count_or_frequency</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">{</span><span class="s1">&#39;count&#39;</span><span class="p">,</span> <span class="s1">&#39;frequency&#39;</span><span class="p">}:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid count_or_frequency &quot;</span>
                             <span class="n">f</span><span class="s2">&quot;</span><span class="si">{count_or_frequency}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">nlibraries</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>
        <span class="n">nsamples</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>

        <span class="k">if</span> <span class="n">mut_type</span> <span class="o">==</span> <span class="s1">&#39;codon&#39;</span><span class="p">:</span>
            <span class="n">height_per</span> <span class="o">=</span> <span class="mf">5.5</span>
            <span class="n">mut_desc</span> <span class="o">=</span> <span class="s1">&#39;codon&#39;</span>
        <span class="k">elif</span> <span class="n">mut_type</span> <span class="o">==</span> <span class="s1">&#39;aa&#39;</span><span class="p">:</span>
            <span class="n">height_per</span> <span class="o">=</span> <span class="mf">1.7</span>
            <span class="n">mut_desc</span> <span class="o">=</span> <span class="s1">&#39;amino acid&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid mut_type </span><span class="si">{mut_type}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">orientation</span> <span class="o">==</span> <span class="s1">&#39;h&#39;</span><span class="p">:</span>
            <span class="n">facet_str</span> <span class="o">=</span> <span class="s1">&#39;sample ~ library&#39;</span>
            <span class="n">width</span> <span class="o">=</span> <span class="n">widthscale</span> <span class="o">*</span> <span class="p">(</span><span class="mf">1.6</span> <span class="o">+</span> <span class="mf">3.5</span> <span class="o">*</span> <span class="n">nlibraries</span><span class="p">)</span>
            <span class="n">height</span> <span class="o">=</span> <span class="n">heightscale</span> <span class="o">*</span> <span class="p">(</span><span class="mf">0.8</span> <span class="o">+</span> <span class="n">height_per</span> <span class="o">*</span> <span class="n">nsamples</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">orientation</span> <span class="o">==</span> <span class="s1">&#39;v&#39;</span><span class="p">:</span>
            <span class="n">facet_str</span> <span class="o">=</span> <span class="s1">&#39;library ~ sample&#39;</span>
            <span class="n">width</span> <span class="o">=</span> <span class="n">widthscale</span> <span class="o">*</span> <span class="p">(</span><span class="mf">1.6</span> <span class="o">+</span> <span class="mf">3.5</span> <span class="o">*</span> <span class="n">nsamples</span><span class="p">)</span>
            <span class="n">height</span> <span class="o">=</span> <span class="n">heightscale</span> <span class="o">*</span> <span class="p">(</span><span class="mf">0.8</span> <span class="o">+</span> <span class="n">height_per</span> <span class="o">*</span> <span class="n">nlibraries</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid `orientation` </span><span class="si">{orientation}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">p</span> <span class="o">=</span> <span class="p">(</span><span class="n">ggplot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="s1">&#39;site&#39;</span><span class="p">,</span> <span class="s1">&#39;mut_char&#39;</span><span class="p">,</span>
                            <span class="n">fill</span><span class="o">=</span><span class="n">count_or_frequency</span><span class="p">))</span> <span class="o">+</span>
             <span class="n">geom_tile</span><span class="p">()</span> <span class="o">+</span>
             <span class="n">facet_grid</span><span class="p">(</span><span class="n">facet_str</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="n">width</span><span class="p">,</span> <span class="n">height</span><span class="p">),</span>
                   <span class="n">legend_key</span><span class="o">=</span><span class="n">element_blank</span><span class="p">(),</span>
                   <span class="n">axis_text_y</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>
                   <span class="p">)</span> <span class="o">+</span>
             <span class="n">scale_x_continuous</span><span class="p">(</span>
                <span class="n">name</span><span class="o">=</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{mut_desc}</span><span class="s1"> site&#39;</span><span class="p">,</span>
                <span class="n">limits</span><span class="o">=</span><span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sites</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="nb">max</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sites</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span>
                <span class="n">expand</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
                <span class="p">)</span> <span class="o">+</span>
             <span class="n">ylab</span><span class="p">(</span><span class="n">mut_desc</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">scale_fill_cmap</span><span class="p">(</span><span class="s1">&#39;gnuplot&#39;</span><span class="p">)</span>
             <span class="p">)</span>


        <span class="k">if</span> <span class="n">plotfile</span><span class="p">:</span>
            <span class="n">p</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">plotfile</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="n">height</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="n">width</span><span class="p">,</span>
                   <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">limitsize</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">p</span></div>


<div class="viewcode-block" id="CodonVariantTable.plotMutFreqs"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.plotMutFreqs">[docs]</a>    <span class="k">def</span> <span class="nf">plotMutFreqs</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">variant_type</span><span class="p">,</span> <span class="n">mut_type</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span>
            <span class="n">libraries</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">plotfile</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">orientation</span><span class="o">=</span><span class="s1">&#39;h&#39;</span><span class="p">,</span> <span class="n">widthscale</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">heightscale</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
            <span class="n">min_support</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Mutation frequency along length of gene.</span>

<span class="sd">        Args:</span>
<span class="sd">            Args have same meaning as for</span>
<span class="sd">            :meth:`CodonVariantTable.plotCumulMutCoverage`.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A `plotnine &lt;https://plotnine.readthedocs.io&gt;`_</span>
<span class="sd">            plot; can be displayed in a Jupyter notebook with `p.draw()`.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutCounts</span><span class="p">(</span><span class="n">variant_type</span><span class="p">,</span> <span class="n">mut_type</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="n">samples</span><span class="p">,</span>
                            <span class="n">libraries</span><span class="o">=</span><span class="n">libraries</span><span class="p">,</span> <span class="n">min_support</span><span class="o">=</span><span class="n">min_support</span><span class="p">)</span>

        <span class="n">n_variants</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">n_variants_df</span><span class="p">(</span><span class="n">libraries</span><span class="o">=</span><span class="n">libraries</span><span class="p">,</span>
                                         <span class="n">samples</span><span class="o">=</span><span class="n">samples</span><span class="p">,</span>
                                         <span class="n">min_support</span><span class="o">=</span><span class="n">min_support</span><span class="p">,</span>
                                         <span class="n">variant_type</span><span class="o">=</span><span class="n">variant_type</span><span class="p">,</span>
                                         <span class="n">mut_type</span><span class="o">=</span><span class="n">mut_type</span><span class="p">)</span>
                      <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;count&#39;</span><span class="p">:</span><span class="s1">&#39;nseqs&#39;</span><span class="p">})</span>
                      <span class="p">)</span>

        <span class="n">df</span> <span class="o">=</span> <span class="p">(</span><span class="n">df</span>
              <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;mutation_type&#39;</span><span class="p">,</span> <span class="s1">&#39;site&#39;</span><span class="p">])</span>
              <span class="o">.</span><span class="n">aggregate</span><span class="p">({</span><span class="s1">&#39;count&#39;</span><span class="p">:</span><span class="s1">&#39;sum&#39;</span><span class="p">})</span>
              <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
              <span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">n_variants</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">])</span>
              <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">freq</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;count&#39;</span><span class="p">]</span> <span class="o">/</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;nseqs&#39;</span><span class="p">])</span>
              <span class="p">)</span>

        <span class="n">nlibraries</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>
        <span class="n">nsamples</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>

        <span class="k">if</span> <span class="n">orientation</span> <span class="o">==</span> <span class="s1">&#39;h&#39;</span><span class="p">:</span>
            <span class="n">facet_str</span> <span class="o">=</span> <span class="s1">&#39;sample ~ library&#39;</span>
            <span class="n">width</span> <span class="o">=</span> <span class="n">widthscale</span> <span class="o">*</span> <span class="p">(</span><span class="mf">1.6</span> <span class="o">+</span> <span class="mf">1.8</span> <span class="o">*</span> <span class="n">nlibraries</span><span class="p">)</span>
            <span class="n">height</span> <span class="o">=</span> <span class="n">heightscale</span> <span class="o">*</span> <span class="p">(</span><span class="mf">0.8</span> <span class="o">+</span> <span class="mi">1</span> <span class="o">*</span> <span class="n">nsamples</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">orientation</span> <span class="o">==</span> <span class="s1">&#39;v&#39;</span><span class="p">:</span>
            <span class="n">facet_str</span> <span class="o">=</span> <span class="s1">&#39;library ~ sample&#39;</span>
            <span class="n">width</span> <span class="o">=</span> <span class="n">widthscale</span> <span class="o">*</span> <span class="p">(</span><span class="mf">1.6</span> <span class="o">+</span> <span class="mf">1.8</span> <span class="o">*</span> <span class="n">nsamples</span><span class="p">)</span>
            <span class="n">height</span> <span class="o">=</span> <span class="n">heightscale</span> <span class="o">*</span> <span class="p">(</span><span class="mf">0.8</span> <span class="o">+</span> <span class="mi">1</span> <span class="o">*</span> <span class="n">nlibraries</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid `orientation` </span><span class="si">{orientation}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">mut_type</span> <span class="o">==</span> <span class="s1">&#39;aa&#39;</span><span class="p">:</span>
            <span class="n">mut_desc</span> <span class="o">=</span> <span class="s1">&#39;amino-acid&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">mut_desc</span> <span class="o">=</span> <span class="n">mut_type</span>

        <span class="k">if</span> <span class="n">height</span> <span class="o">&lt;</span> <span class="mi">3</span><span class="p">:</span>
            <span class="n">ylabel</span> <span class="o">=</span> <span class="p">(</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{mut_desc}</span><span class="s1"> mutation</span><span class="se">\n</span><span class="s1">frequency &#39;</span>
                      <span class="n">f</span><span class="s1">&#39;(</span><span class="si">{variant_type}</span><span class="s1"> mutants)&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">ylabel</span> <span class="o">=</span> <span class="p">(</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{mut_desc}</span><span class="s1"> mutation frequency &#39;</span>
                      <span class="n">f</span><span class="s1">&#39;(</span><span class="si">{variant_type}</span><span class="s1"> mutants)&#39;</span><span class="p">)</span>

        <span class="n">p</span> <span class="o">=</span> <span class="p">(</span><span class="n">ggplot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="s1">&#39;site&#39;</span><span class="p">,</span> <span class="s1">&#39;freq&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;mutation_type&#39;</span><span class="p">))</span> <span class="o">+</span>
             <span class="n">geom_step</span><span class="p">()</span> <span class="o">+</span>
             <span class="n">scale_color_manual</span><span class="p">(</span>
                <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_mutation_type_colors</span><span class="p">[</span><span class="n">m</span><span class="p">]</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span>
                 <span class="n">df</span><span class="o">.</span><span class="n">mutation_type</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">sort_values</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()],</span>
                <span class="n">name</span><span class="o">=</span><span class="s1">&#39;mutation type&#39;</span>
                <span class="p">)</span> <span class="o">+</span>
             <span class="n">scale_x_continuous</span><span class="p">(</span>
                <span class="n">name</span><span class="o">=</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{mut_desc}</span><span class="s1"> site&#39;</span><span class="p">,</span>
                <span class="n">limits</span><span class="o">=</span><span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sites</span><span class="p">),</span> <span class="nb">max</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sites</span><span class="p">))</span>
                <span class="p">)</span> <span class="o">+</span>
             <span class="n">ylab</span><span class="p">(</span><span class="n">ylabel</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">facet_grid</span><span class="p">(</span><span class="n">facet_str</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="n">width</span><span class="p">,</span> <span class="n">height</span><span class="p">),</span>
                   <span class="n">legend_key</span><span class="o">=</span><span class="n">element_blank</span><span class="p">(),</span>
                   <span class="n">legend_text</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="mi">11</span><span class="p">)</span>
                   <span class="p">)</span>
             <span class="p">)</span>

        <span class="k">if</span> <span class="n">plotfile</span><span class="p">:</span>
            <span class="n">p</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">plotfile</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="n">height</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="n">width</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">p</span></div>


<div class="viewcode-block" id="CodonVariantTable.plotCumulVariantCounts"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.plotCumulVariantCounts">[docs]</a>    <span class="k">def</span> <span class="nf">plotCumulVariantCounts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">variant_type</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span>
            <span class="n">libraries</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">plotfile</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">orientation</span><span class="o">=</span><span class="s1">&#39;h&#39;</span><span class="p">,</span> <span class="n">widthscale</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">heightscale</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
            <span class="n">min_support</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">mut_type</span><span class="o">=</span><span class="s1">&#39;aa&#39;</span><span class="p">,</span> <span class="n">tot_variants_hline</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Plot number variants with &gt;= that each number of counts.</span>

<span class="sd">        Args:</span>
<span class="sd">            `variant_type` (&quot;single&quot; or &quot;all&quot;)</span>
<span class="sd">                Include just variants with &lt;= one mutation of type,</span>
<span class="sd">                `mut_type` or all mutants?</span>
<span class="sd">            `tot_variants_hline` (bool)</span>
<span class="sd">                Include a dotted horizontal line indicating the total</span>
<span class="sd">                number of variants (useful since ones not observed</span>
<span class="sd">                at all will not apparent on plot).</span>
<span class="sd">            Other args:</span>
<span class="sd">                Same meaning as for</span>
<span class="sd">                :meth:`CodonVariantTable.plotNumMutsHistogram`.</span>

<span class="sd">        Returns:</span>
<span class="sd">            `plotnine &lt;https://plotnine.readthedocs.io&gt;`_</span>
<span class="sd">            plot; can be displayed in a Jupyter notebook with `p.draw()`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span><span class="p">,</span> <span class="n">nlibraries</span><span class="p">,</span> <span class="n">nsamples</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_getPlotData</span><span class="p">(</span><span class="n">libraries</span><span class="p">,</span>
                                                     <span class="n">samples</span><span class="p">,</span>
                                                     <span class="n">min_support</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">variant_type</span> <span class="o">==</span> <span class="s1">&#39;single&#39;</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">mut_type</span> <span class="o">==</span> <span class="s1">&#39;aa&#39;</span><span class="p">:</span>
                <span class="n">mutstr</span> <span class="o">=</span> <span class="s1">&#39;amino acid&#39;</span>
            <span class="k">elif</span> <span class="n">mut_type</span> <span class="o">==</span> <span class="s1">&#39;codon&#39;</span><span class="p">:</span>
                <span class="n">mutstr</span> <span class="o">=</span> <span class="n">mut_type</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid `mut_type` </span><span class="si">{mut_type}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">ylabel</span> <span class="o">=</span> <span class="n">f</span><span class="s2">&quot;single </span><span class="si">{mutstr}</span><span class="s2"> variants with &gt;= this many counts&quot;</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;n_</span><span class="si">{mut_type}</span><span class="s2">_substitutions &lt;= 1&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">variant_type</span> <span class="o">==</span> <span class="s1">&#39;all&#39;</span><span class="p">:</span>
            <span class="n">ylabel</span> <span class="o">=</span> <span class="s1">&#39;variants with &gt;= this many counts&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid `variant_type` </span><span class="si">{variant_type}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">orientation</span> <span class="o">==</span> <span class="s1">&#39;h&#39;</span><span class="p">:</span>
            <span class="n">facet_str</span> <span class="o">=</span> <span class="s1">&#39;sample ~ library&#39;</span>
            <span class="n">width</span> <span class="o">=</span> <span class="n">widthscale</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="mf">1.5</span> <span class="o">*</span> <span class="n">nlibraries</span><span class="p">)</span>
            <span class="n">height</span> <span class="o">=</span> <span class="n">heightscale</span> <span class="o">*</span> <span class="p">(</span><span class="mf">0.6</span> <span class="o">+</span> <span class="mf">1.5</span> <span class="o">*</span> <span class="n">nsamples</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">orientation</span> <span class="o">==</span> <span class="s1">&#39;v&#39;</span><span class="p">:</span>
            <span class="n">facet_str</span> <span class="o">=</span> <span class="s1">&#39;library ~ sample&#39;</span>
            <span class="n">width</span> <span class="o">=</span> <span class="n">widthscale</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="mf">1.5</span> <span class="o">*</span> <span class="n">nsamples</span><span class="p">)</span>
            <span class="n">height</span> <span class="o">=</span> <span class="n">heightscale</span> <span class="o">*</span> <span class="p">(</span><span class="mf">0.6</span> <span class="o">+</span> <span class="mf">1.5</span> <span class="o">*</span> <span class="n">nlibraries</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid `orientation` </span><span class="si">{orientation}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">df</span> <span class="o">=</span> <span class="p">(</span><span class="n">getCumulVariantsByCount</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">group_cols</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">])</span>
              <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;count &gt; 0&#39;</span><span class="p">)</span>
              <span class="p">)</span>

        <span class="n">p</span> <span class="o">=</span> <span class="p">(</span><span class="n">ggplot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="s1">&#39;count&#39;</span><span class="p">,</span> <span class="s1">&#39;nvariants&#39;</span><span class="p">))</span> <span class="o">+</span>
             <span class="n">geom_step</span><span class="p">()</span> <span class="o">+</span>
             <span class="n">facet_grid</span><span class="p">(</span><span class="n">facet_str</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">xlab</span><span class="p">(</span><span class="s1">&#39;number of counts&#39;</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">ylab</span><span class="p">(</span><span class="n">ylabel</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">scale_x_log10</span><span class="p">(</span><span class="n">labels</span><span class="o">=</span><span class="n">latexSciNot</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">scale_y_continuous</span><span class="p">(</span><span class="n">labels</span><span class="o">=</span><span class="n">latexSciNot</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="n">width</span><span class="p">,</span> <span class="n">height</span><span class="p">))</span>
             <span class="p">)</span>

        <span class="k">if</span> <span class="n">tot_variants_hline</span><span class="p">:</span>
            <span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="n">geom_hline</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">yintercept</span><span class="o">=</span><span class="s1">&#39;total_variants&#39;</span><span class="p">),</span>
                               <span class="n">linetype</span><span class="o">=</span><span class="s1">&#39;dashed&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">CBPALETTE</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

        <span class="k">if</span> <span class="n">plotfile</span><span class="p">:</span>
            <span class="n">p</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">plotfile</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="n">height</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="n">width</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">p</span></div>


<div class="viewcode-block" id="CodonVariantTable.plotCumulMutCoverage"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.plotCumulMutCoverage">[docs]</a>    <span class="k">def</span> <span class="nf">plotCumulMutCoverage</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">variant_type</span><span class="p">,</span> <span class="n">mut_type</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span>
            <span class="n">libraries</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">plotfile</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">orientation</span><span class="o">=</span><span class="s1">&#39;h&#39;</span><span class="p">,</span> <span class="n">widthscale</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">heightscale</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
            <span class="n">min_support</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">max_count</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Fraction of mutation seen &lt;= some number of times.</span>

<span class="sd">        Args:</span>
<span class="sd">            `variant_type` (&quot;single&quot; or &quot;all&quot;)</span>
<span class="sd">                Include just single mutants, or all mutants?</span>
<span class="sd">                Mutations are counted relative to `mut_type`.</span>
<span class="sd">            `max_count` (`None` or int)</span>
<span class="sd">                Plot cumulative fraction plot out to this</span>
<span class="sd">                number of observations of mutation. If `None`,</span>
<span class="sd">                a reasonable value is automatically determined.</span>
<span class="sd">            Other args:</span>
<span class="sd">                Same meaning as for</span>
<span class="sd">                :class:`CodonVariantTable.plotNumMutsHistogram`</span>

<span class="sd">        Returns:</span>
<span class="sd">            A `plotnine &lt;https://plotnine.readthedocs.io&gt;`_</span>
<span class="sd">            plot; can be displayed in a Jupyter notebook with `p.draw()`.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mutCounts</span><span class="p">(</span><span class="n">variant_type</span><span class="p">,</span> <span class="n">mut_type</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="n">samples</span><span class="p">,</span>
                            <span class="n">libraries</span><span class="o">=</span><span class="n">libraries</span><span class="p">,</span> <span class="n">min_support</span><span class="o">=</span><span class="n">min_support</span><span class="p">)</span>

        <span class="c1"># add one to counts to plot fraction found &lt; this many</span>
        <span class="c1"># as stat_ecdf by default does &lt;=</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">count</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;count&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">max_count</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">max_count</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;count&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">quantile</span><span class="p">(</span><span class="mf">0.75</span><span class="p">)</span>

        <span class="n">nlibraries</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>
        <span class="n">nsamples</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>

        <span class="k">if</span> <span class="n">orientation</span> <span class="o">==</span> <span class="s1">&#39;h&#39;</span><span class="p">:</span>
            <span class="n">facet_str</span> <span class="o">=</span> <span class="s1">&#39;sample ~ library&#39;</span>
            <span class="n">width</span> <span class="o">=</span> <span class="n">widthscale</span> <span class="o">*</span> <span class="p">(</span><span class="mf">1.6</span> <span class="o">+</span> <span class="mf">1.3</span> <span class="o">*</span> <span class="n">nlibraries</span><span class="p">)</span>
            <span class="n">height</span> <span class="o">=</span> <span class="n">heightscale</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="mf">1.2</span> <span class="o">*</span> <span class="n">nsamples</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">orientation</span> <span class="o">==</span> <span class="s1">&#39;v&#39;</span><span class="p">:</span>
            <span class="n">facet_str</span> <span class="o">=</span> <span class="s1">&#39;library ~ sample&#39;</span>
            <span class="n">width</span> <span class="o">=</span> <span class="n">widthscale</span> <span class="o">*</span> <span class="p">(</span><span class="mf">1.6</span> <span class="o">+</span> <span class="mf">1.3</span> <span class="o">*</span> <span class="n">nsamples</span><span class="p">)</span>
            <span class="n">height</span> <span class="o">=</span> <span class="n">heightscale</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="mf">1.2</span> <span class="o">*</span> <span class="n">nlibraries</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid `orientation` </span><span class="si">{orientation}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">width</span> <span class="o">&gt;</span> <span class="mi">4</span><span class="p">:</span>
            <span class="n">xlabel</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;counts among </span><span class="si">{variant_type}</span><span class="s1"> mutants&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">xlabel</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;counts among</span><span class="se">\n</span><span class="si">{variant_type}</span><span class="s1"> mutants&#39;</span>

        <span class="n">mut_desc</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;aa&#39;</span><span class="p">:</span><span class="s1">&#39;amino-acid&#39;</span><span class="p">,</span> <span class="s1">&#39;codon&#39;</span><span class="p">:</span><span class="s1">&#39;codon&#39;</span><span class="p">}[</span><span class="n">mut_type</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">height</span> <span class="o">&gt;</span> <span class="mi">3</span><span class="p">:</span>
            <span class="n">ylabel</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;frac </span><span class="si">{mut_desc}</span><span class="s1"> mutations found &lt; this many times&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">ylabel</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;frac </span><span class="si">{mut_desc}</span><span class="s1"> mutations</span><span class="se">\n</span><span class="s1">found &lt; this many times&#39;</span>

        <span class="n">p</span> <span class="o">=</span> <span class="p">(</span><span class="n">ggplot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="s1">&#39;count&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;mutation_type&#39;</span><span class="p">))</span> <span class="o">+</span>
             <span class="n">stat_ecdf</span><span class="p">(</span><span class="n">geom</span><span class="o">=</span><span class="s1">&#39;step&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mf">0.75</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">coord_cartesian</span><span class="p">(</span><span class="n">xlim</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">max_count</span><span class="p">))</span> <span class="o">+</span>
             <span class="n">scale_color_manual</span><span class="p">(</span>
                <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_mutation_type_colors</span><span class="p">[</span><span class="n">m</span><span class="p">]</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span>
                 <span class="n">df</span><span class="o">.</span><span class="n">mutation_type</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">sort_values</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()],</span>
                <span class="n">name</span><span class="o">=</span><span class="s1">&#39;mutation type&#39;</span>
                <span class="p">)</span> <span class="o">+</span>
             <span class="n">xlab</span><span class="p">(</span><span class="n">xlabel</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">ylab</span><span class="p">(</span><span class="n">ylabel</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">facet_grid</span><span class="p">(</span><span class="n">facet_str</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="n">width</span><span class="p">,</span> <span class="n">height</span><span class="p">),</span>
                   <span class="n">legend_key</span><span class="o">=</span><span class="n">element_blank</span><span class="p">(),</span>
                   <span class="n">legend_text</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="mi">11</span><span class="p">),</span>
                   <span class="n">axis_text_x</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="mi">90</span><span class="p">),</span>
                   <span class="p">)</span>
             <span class="p">)</span>

        <span class="k">if</span> <span class="n">plotfile</span><span class="p">:</span>
            <span class="n">p</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">plotfile</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="n">height</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="n">width</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">p</span></div>


<div class="viewcode-block" id="CodonVariantTable.plotNumCodonMutsByType"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.plotNumCodonMutsByType">[docs]</a>    <span class="k">def</span> <span class="nf">plotNumCodonMutsByType</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">variant_type</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span>
            <span class="n">libraries</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">plotfile</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">orientation</span><span class="o">=</span><span class="s1">&#39;h&#39;</span><span class="p">,</span> <span class="n">widthscale</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">heightscale</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
            <span class="n">min_support</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Nonsynonymous, synonymous, stop mutations per variant.</span>

<span class="sd">        Args:</span>
<span class="sd">            `variant_type` (&quot;single&quot; or &quot;all&quot;)</span>
<span class="sd">                Include just single-codon mutants and wildtype,</span>
<span class="sd">                or include all mutants.</span>
<span class="sd">            Other args:</span>
<span class="sd">                Same meaning as for</span>
<span class="sd">                :class:`CodonVariantTable.plotNumMutsHistogram`</span>

<span class="sd">        Returns:</span>
<span class="sd">            A `plotnine &lt;https://plotnine.readthedocs.io&gt;`_</span>
<span class="sd">            plot; can be displayed in a Jupyter notebook with `p.draw()`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span><span class="p">,</span> <span class="n">nlibraries</span><span class="p">,</span> <span class="n">nsamples</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_getPlotData</span><span class="p">(</span><span class="n">libraries</span><span class="p">,</span>
                                                     <span class="n">samples</span><span class="p">,</span>
                                                     <span class="n">min_support</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">variant_type</span> <span class="o">==</span> <span class="s1">&#39;single&#39;</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;n_codon_substitutions &lt;= 1&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">variant_type</span> <span class="o">!=</span> <span class="s1">&#39;all&#39;</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid variant_type </span><span class="si">{variant_type}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">orientation</span> <span class="o">==</span> <span class="s1">&#39;h&#39;</span><span class="p">:</span>
            <span class="n">facet_str</span> <span class="o">=</span> <span class="s1">&#39;sample ~ library&#39;</span>
            <span class="n">width</span> <span class="o">=</span> <span class="n">widthscale</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="mf">1.4</span> <span class="o">*</span> <span class="n">nlibraries</span><span class="p">)</span>
            <span class="n">height</span> <span class="o">=</span> <span class="n">heightscale</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="mf">1.3</span> <span class="o">*</span> <span class="n">nsamples</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">orientation</span> <span class="o">==</span> <span class="s1">&#39;v&#39;</span><span class="p">:</span>
            <span class="n">facet_str</span> <span class="o">=</span> <span class="s1">&#39;library ~ sample&#39;</span>
            <span class="n">width</span> <span class="o">=</span> <span class="n">widthscale</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="mf">1.4</span> <span class="o">*</span> <span class="n">nsamples</span><span class="p">)</span>
            <span class="n">height</span> <span class="o">=</span> <span class="n">heightscale</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="mf">1.3</span> <span class="o">*</span> <span class="n">nlibraries</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid `orientation` </span><span class="si">{orientation}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">height</span> <span class="o">&gt;</span> <span class="mi">3</span><span class="p">:</span>
            <span class="n">ylabel</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;mutations per variant (</span><span class="si">{variant_type}</span><span class="s1"> mutants)&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">ylabel</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;mutations per variant</span><span class="se">\n</span><span class="s1">(</span><span class="si">{variant_type}</span><span class="s1"> mutants)&#39;</span>

        <span class="n">codon_mut_types</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;nonsynonymous&#39;</span><span class="p">,</span> <span class="s1">&#39;synonymous&#39;</span><span class="p">,</span> <span class="s1">&#39;stop&#39;</span><span class="p">]</span>

        <span class="c1"># mutations from stop to another amino-acid counted as nonsyn</span>
        <span class="n">df</span> <span class="o">=</span> <span class="p">(</span><span class="n">df</span>
              <span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                <span class="n">synonymous</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">n_codon_substitutions</span> <span class="o">-</span>
                                     <span class="n">x</span><span class="o">.</span><span class="n">n_aa_substitutions</span><span class="p">,</span>
                <span class="n">stop</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">aa_substitutions</span><span class="o">.</span><span class="n">str</span>
                               <span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="s1">&#39;[A-Z]\d+\*&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="nb">len</span><span class="p">),</span>
                <span class="n">nonsynonymous</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">n_codon_substitutions</span> <span class="o">-</span>
                                        <span class="n">x</span><span class="o">.</span><span class="n">synonymous</span> <span class="o">-</span> <span class="n">x</span><span class="o">.</span><span class="n">stop</span>
                <span class="p">)</span>
              <span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">],</span>
                    <span class="n">value_vars</span><span class="o">=</span><span class="n">codon_mut_types</span><span class="p">,</span>
                    <span class="n">var_name</span><span class="o">=</span><span class="s1">&#39;mutation_type&#39;</span><span class="p">,</span>
                    <span class="n">value_name</span><span class="o">=</span><span class="s1">&#39;num_muts&#39;</span><span class="p">)</span>
              <span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                  <span class="n">mutation_type</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                                <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
                                 <span class="n">x</span><span class="o">.</span><span class="n">mutation_type</span><span class="p">,</span>
                                 <span class="n">categories</span><span class="o">=</span><span class="n">codon_mut_types</span><span class="p">,</span>
                                 <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
                  <span class="n">num_muts_count</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">num_muts</span> <span class="o">*</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;count&#39;</span><span class="p">]</span>
                  <span class="p">)</span>
              <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;mutation_type&#39;</span><span class="p">])</span>
              <span class="o">.</span><span class="n">aggregate</span><span class="p">({</span><span class="s1">&#39;num_muts_count&#39;</span><span class="p">:</span><span class="s1">&#39;sum&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">:</span><span class="s1">&#39;sum&#39;</span><span class="p">})</span>
              <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
              <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">number</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">num_muts_count</span> <span class="o">/</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;count&#39;</span><span class="p">])</span>
              <span class="p">)</span>

        <span class="n">p</span> <span class="o">=</span> <span class="p">(</span><span class="n">ggplot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="s1">&#39;mutation_type&#39;</span><span class="p">,</span> <span class="s1">&#39;number&#39;</span><span class="p">,</span>
                            <span class="n">fill</span><span class="o">=</span><span class="s1">&#39;mutation_type&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;number&#39;</span><span class="p">))</span> <span class="o">+</span>
             <span class="n">geom_bar</span><span class="p">(</span><span class="n">stat</span><span class="o">=</span><span class="s1">&#39;identity&#39;</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">geom_text</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;bottom&#39;</span><span class="p">,</span> <span class="n">format_string</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">{0:.3f}</span><span class="s1">&#39;</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">facet_grid</span><span class="p">(</span><span class="n">facet_str</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">scale_y_continuous</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">ylabel</span><span class="p">,</span>
                                <span class="n">expand</span><span class="o">=</span><span class="p">(</span><span class="mf">0.03</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mf">0.12</span><span class="p">,</span> <span class="mi">0</span><span class="p">))</span> <span class="o">+</span>
             <span class="n">scale_fill_manual</span><span class="p">(</span>
                <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_mutation_type_colors</span><span class="p">[</span><span class="n">m</span><span class="p">]</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span>
                 <span class="n">df</span><span class="o">.</span><span class="n">mutation_type</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">sort_values</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()]</span>
                <span class="p">)</span> <span class="o">+</span>
             <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="n">width</span><span class="p">,</span> <span class="n">height</span><span class="p">),</span>
                   <span class="n">axis_title_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">(),</span>
                   <span class="n">axis_text_x</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="mi">90</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">11</span><span class="p">),</span>
                   <span class="n">legend_position</span><span class="o">=</span><span class="s1">&#39;none&#39;</span><span class="p">)</span>
             <span class="p">)</span>

        <span class="k">if</span> <span class="n">plotfile</span><span class="p">:</span>
            <span class="n">p</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">plotfile</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="n">height</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="n">width</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">p</span></div>


<div class="viewcode-block" id="CodonVariantTable.plotNumMutsHistogram"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.plotNumMutsHistogram">[docs]</a>    <span class="k">def</span> <span class="nf">plotNumMutsHistogram</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">mut_type</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span>
            <span class="n">libraries</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">plotfile</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">orientation</span><span class="o">=</span><span class="s1">&#39;h&#39;</span><span class="p">,</span> <span class="n">widthscale</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">heightscale</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
            <span class="n">min_support</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">max_muts</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Plot histograms of number of mutations per variant.</span>

<span class="sd">        Args:</span>
<span class="sd">            `mut_type` (str)</span>
<span class="sd">                Type of mutation: &quot;codon&quot; or &quot;aa&quot;.</span>
<span class="sd">            `libraries` (&quot;all&quot;, &quot;all_only&quot;, or list)</span>
<span class="sd">                Set to &quot;all&quot; to include all libraries including</span>
<span class="sd">                a merge of the libraries, &quot;all_only&quot; to only</span>
<span class="sd">                include the merge of the libraries, or a list</span>
<span class="sd">                of libraries.</span>
<span class="sd">            `samples` (the str &quot;all&quot;, `None`, or list)</span>
<span class="sd">                Set to &quot;all&quot; to include all samples, `None` to</span>
<span class="sd">                just count each barcoded variant once, or specify</span>
<span class="sd">                a list of samples.</span>
<span class="sd">            `plotfile` (`None` or str)</span>
<span class="sd">                Name of file to which we save plot.</span>
<span class="sd">            `orientation` (the str &#39;h&#39; or &#39;v&#39;)</span>
<span class="sd">                Do we facet libraries horizontally or vertically?</span>
<span class="sd">            `widthscale` (float or int)</span>
<span class="sd">                Expand width of plot by this much.</span>
<span class="sd">            `heightscale` (float or int)</span>
<span class="sd">                Expand height of plot by this much.</span>
<span class="sd">            `min_support` (int)</span>
<span class="sd">                Only include variants with at least this large</span>
<span class="sd">                of a variant call support.</span>
<span class="sd">            `max_muts` (int or `None`)</span>
<span class="sd">                In histogram, group together all variants with</span>
<span class="sd">                &gt;= this many mutations; set to `None` for no</span>
<span class="sd">                cutoff.</span>

<span class="sd">        Returns:</span>
<span class="sd">            `plotnine &lt;https://plotnine.readthedocs.io&gt;`_</span>
<span class="sd">            plot; can be displayed in a Jupyter notebook with `p.draw()`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span><span class="p">,</span> <span class="n">nlibraries</span><span class="p">,</span> <span class="n">nsamples</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_getPlotData</span><span class="p">(</span><span class="n">libraries</span><span class="p">,</span>
                                                     <span class="n">samples</span><span class="p">,</span>
                                                     <span class="n">min_support</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">mut_type</span> <span class="o">==</span> <span class="s1">&#39;aa&#39;</span><span class="p">:</span>
            <span class="n">mut_col</span> <span class="o">=</span> <span class="s1">&#39;n_aa_substitutions&#39;</span>
            <span class="n">xlabel</span> <span class="o">=</span> <span class="s1">&#39;amino-acid mutations&#39;</span>
        <span class="k">elif</span> <span class="n">mut_type</span> <span class="o">==</span> <span class="s1">&#39;codon&#39;</span><span class="p">:</span>
            <span class="n">mut_col</span> <span class="o">=</span> <span class="s1">&#39;n_codon_substitutions&#39;</span>
            <span class="n">xlabel</span> <span class="o">=</span> <span class="s1">&#39;codon mutations&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid mut_type </span><span class="si">{mut_type}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">orientation</span> <span class="o">==</span> <span class="s1">&#39;h&#39;</span><span class="p">:</span>
            <span class="n">facet_str</span> <span class="o">=</span> <span class="s1">&#39;sample ~ library&#39;</span>
            <span class="n">width</span> <span class="o">=</span> <span class="n">widthscale</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="mf">1.5</span> <span class="o">*</span> <span class="n">nlibraries</span><span class="p">)</span>
            <span class="n">height</span> <span class="o">=</span> <span class="n">heightscale</span> <span class="o">*</span> <span class="p">(</span><span class="mf">0.6</span> <span class="o">+</span> <span class="mf">1.5</span> <span class="o">*</span> <span class="n">nsamples</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">orientation</span> <span class="o">==</span> <span class="s1">&#39;v&#39;</span><span class="p">:</span>
            <span class="n">facet_str</span> <span class="o">=</span> <span class="s1">&#39;library ~ sample&#39;</span>
            <span class="n">width</span> <span class="o">=</span> <span class="n">widthscale</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="mf">1.5</span> <span class="o">*</span> <span class="n">nsamples</span><span class="p">)</span>
            <span class="n">height</span> <span class="o">=</span> <span class="n">heightscale</span> <span class="o">*</span> <span class="p">(</span><span class="mf">0.6</span> <span class="o">+</span> <span class="mf">1.5</span> <span class="o">*</span> <span class="n">nlibraries</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid `orientation` </span><span class="si">{orientation}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">df</span><span class="p">[</span><span class="n">mut_col</span><span class="p">]</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">mut_col</span><span class="p">],</span> <span class="kc">None</span><span class="p">,</span> <span class="n">max_muts</span><span class="p">)</span>

        <span class="n">df</span> <span class="o">=</span> <span class="p">(</span><span class="n">df</span>
              <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="n">mut_col</span><span class="p">])</span>
              <span class="o">.</span><span class="n">aggregate</span><span class="p">({</span><span class="s1">&#39;count&#39;</span><span class="p">:</span><span class="s1">&#39;sum&#39;</span><span class="p">})</span>
              <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
              <span class="p">)</span>

        <span class="n">p</span> <span class="o">=</span> <span class="p">(</span><span class="n">ggplot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="n">mut_col</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">))</span> <span class="o">+</span>
             <span class="n">geom_bar</span><span class="p">(</span><span class="n">stat</span><span class="o">=</span><span class="s1">&#39;identity&#39;</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">facet_grid</span><span class="p">(</span><span class="n">facet_str</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">xlab</span><span class="p">(</span><span class="n">xlabel</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">scale_y_continuous</span><span class="p">(</span><span class="n">labels</span><span class="o">=</span><span class="n">latexSciNot</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="n">width</span><span class="p">,</span> <span class="n">height</span><span class="p">))</span>
             <span class="p">)</span>

        <span class="k">if</span> <span class="n">plotfile</span><span class="p">:</span>
            <span class="n">p</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">plotfile</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="n">height</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="n">width</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">p</span></div>


<div class="viewcode-block" id="CodonVariantTable.writeCodonCounts"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.writeCodonCounts">[docs]</a>    <span class="k">def</span> <span class="nf">writeCodonCounts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">single_or_all</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span>
                         <span class="n">outdir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">include_all_libs</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Writes codon counts files for all libraries and samples.</span>

<span class="sd">        Files are written in</span>
<span class="sd">        `format &lt;https://jbloomlab.github.io/dms_tools2/dms2_bcsubamp.html#id12&gt;`_</span>
<span class="sd">        produced by</span>
<span class="sd">        `dms2_bcsubamp &lt;https://jbloomlab.github.io/dms_tools2/dms2_bcsubamp.html&gt;`_.</span>

<span class="sd">        Args:</span>
<span class="sd">            `single_or_all` (&quot;single&quot; or &quot;all&quot;)</span>
<span class="sd">                If &quot;single&quot;, then counts are just from single-codon</span>
<span class="sd">                mutants and wildtype, and we count all wildtype codons</span>
<span class="sd">                and just mutated codon for single-codon mutants. If</span>
<span class="sd">                &quot;all&quot;, we count all codons for all variants at all</span>
<span class="sd">                sites. This is appropriate if enrichment of each mutation</span>
<span class="sd">                is supposed to represent its effect for &quot;single&quot;, and if</span>
<span class="sd">                enrichment of mutation is supposed to represent its</span>
<span class="sd">                average effect across genetic backgrounds in the library</span>
<span class="sd">                for &quot;all&quot; provided mutations are Poisson distributed.</span>
<span class="sd">            `outdir` (`None` or str)</span>
<span class="sd">                Name of directory into which we write counts,</span>
<span class="sd">                created if it does not exist. Use `None` to</span>
<span class="sd">                write to current directory.</span>
<span class="sd">            `include_all_libs` (bool)</span>
<span class="sd">                Include data for a library (named &quot;all-libraries&quot;)</span>
<span class="sd">                that has the summmed data for all individual libraries</span>
<span class="sd">                if there are multiple libraries?</span>

<span class="sd">        Returns:</span>
<span class="sd">            Pandas data frame with columns &quot;library&quot;, &quot;sample&quot;,</span>
<span class="sd">            and &quot;countfile&quot;. The &quot;countfile&quot; columns gives name of the</span>
<span class="sd">            createad CSV file, ``&lt;library&gt;_&lt;sample&gt;_codoncounts.csv``.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">codonmatch</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;^(?P&lt;wt&gt;{&quot;|&quot;.join(CODONS)})&#39;</span>
                                 <span class="s1">&#39;(?P&lt;r&gt;\d+)&#39;</span>
                                <span class="n">f</span><span class="s1">&#39;(?P&lt;mut&gt;{&quot;|&quot;.join(CODONS)})$&#39;</span>
                                <span class="p">)</span>
        <span class="k">def</span> <span class="nf">_parseCodonMut</span><span class="p">(</span><span class="n">mutstr</span><span class="p">):</span>
            <span class="n">m</span> <span class="o">=</span> <span class="n">codonmatch</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">mutstr</span><span class="p">)</span>
            <span class="k">return</span> <span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;wt&#39;</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">)),</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;mut&#39;</span><span class="p">))</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">variant_count_df</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;no samples with counts&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">single_or_all</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">{</span><span class="s1">&#39;single&#39;</span><span class="p">,</span> <span class="s1">&#39;all&#39;</span><span class="p">}:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid `single_or_all` </span><span class="si">{single_or_all}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">outdir</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">outdir</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>

        <span class="k">if</span> <span class="n">include_all_libs</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">addMergedLibraries</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">variant_count_df</span><span class="p">,</span>
                                         <span class="n">all_lib</span><span class="o">=</span><span class="s1">&#39;all-libraries&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">variant_count_df</span>

        <span class="n">countfiles</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">liblist</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">samplelist</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">lib</span><span class="p">,</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">product</span><span class="p">(</span>
                            <span class="n">df</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span>
                            <span class="n">df</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
                            <span class="p">):</span>

            <span class="n">i_df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;library == @lib &amp; sample == @sample&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">i_df</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">continue</span> <span class="c1"># no data for this library and sample</span>

            <span class="n">countfile</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span>
                            <span class="n">f</span><span class="s1">&#39;</span><span class="si">{lib}</span><span class="s1">_</span><span class="si">{sample}</span><span class="s1">_codoncounts.csv&#39;</span><span class="p">)</span>
            <span class="n">countfiles</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">countfile</span><span class="p">)</span>
            <span class="n">liblist</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lib</span><span class="p">)</span>
            <span class="n">samplelist</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>

            <span class="n">codoncounts</span> <span class="o">=</span> <span class="p">{</span><span class="n">codon</span><span class="p">:[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sites</span><span class="p">)</span> <span class="k">for</span> <span class="n">codon</span>
                           <span class="ow">in</span> <span class="n">CODONS</span><span class="p">}</span>

            <span class="k">if</span> <span class="n">single_or_all</span> <span class="o">==</span> <span class="s1">&#39;single&#39;</span><span class="p">:</span>
                <span class="n">n_wt</span> <span class="o">=</span> <span class="p">(</span><span class="n">i_df</span>
                        <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;n_codon_substitutions == 0&#39;</span><span class="p">)</span>
                        <span class="p">[</span><span class="s1">&#39;count&#39;</span><span class="p">]</span>
                        <span class="o">.</span><span class="n">sum</span><span class="p">()</span>
                        <span class="p">)</span>
                <span class="k">for</span> <span class="n">isite</span><span class="p">,</span> <span class="n">site</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sites</span><span class="p">):</span>
                    <span class="n">codoncounts</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">codons</span><span class="p">[</span><span class="n">site</span><span class="p">]][</span><span class="n">isite</span><span class="p">]</span> <span class="o">+=</span> <span class="n">n_wt</span>
                <span class="k">for</span> <span class="n">mut</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="p">(</span><span class="n">i_df</span>
                                   <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;n_codon_substitutions == 1&#39;</span><span class="p">)</span>
                                   <span class="p">[[</span><span class="s1">&#39;codon_substitutions&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">]]</span>
                                   <span class="o">.</span><span class="n">itertuples</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
                                   <span class="p">):</span>
                    <span class="n">wtcodon</span><span class="p">,</span> <span class="n">r</span><span class="p">,</span> <span class="n">mutcodon</span> <span class="o">=</span> <span class="n">_parseCodonMut</span><span class="p">(</span><span class="n">mut</span><span class="p">)</span>
                    <span class="n">codoncounts</span><span class="p">[</span><span class="n">mutcodon</span><span class="p">][</span><span class="n">r</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">+=</span> <span class="n">count</span>

            <span class="k">elif</span> <span class="n">single_or_all</span> <span class="o">==</span> <span class="s1">&#39;all&#39;</span><span class="p">:</span>
                <span class="n">n_wt</span> <span class="o">=</span> <span class="n">i_df</span><span class="p">[</span><span class="s1">&#39;count&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
                <span class="k">for</span> <span class="n">isite</span><span class="p">,</span> <span class="n">site</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sites</span><span class="p">):</span>
                    <span class="n">codoncounts</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">codons</span><span class="p">[</span><span class="n">site</span><span class="p">]][</span><span class="n">isite</span><span class="p">]</span> <span class="o">+=</span> <span class="n">n_wt</span>
                <span class="k">for</span> <span class="n">muts</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="p">(</span><span class="n">i_df</span>
                                   <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;n_codon_substitutions &gt; 0&#39;</span><span class="p">)</span>
                                   <span class="p">[[</span><span class="s1">&#39;codon_substitutions&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">]]</span>
                                   <span class="o">.</span><span class="n">itertuples</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
                                   <span class="p">):</span>
                    <span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">muts</span><span class="o">.</span><span class="n">split</span><span class="p">():</span>
                        <span class="n">wtcodon</span><span class="p">,</span> <span class="n">r</span><span class="p">,</span> <span class="n">mutcodon</span> <span class="o">=</span> <span class="n">_parseCodonMut</span><span class="p">(</span><span class="n">mut</span><span class="p">)</span>
                        <span class="n">codoncounts</span><span class="p">[</span><span class="n">mutcodon</span><span class="p">][</span><span class="n">r</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">+=</span> <span class="n">count</span>
                        <span class="n">codoncounts</span><span class="p">[</span><span class="n">wtcodon</span><span class="p">][</span><span class="n">r</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">-=</span> <span class="n">count</span>

            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid `single_or_all` </span><span class="si">{single_or_all}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="n">counts_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">collections</span><span class="o">.</span><span class="n">OrderedDict</span><span class="p">(</span>
                         <span class="p">[(</span><span class="s1">&#39;site&#39;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">sites</span><span class="p">),</span>
                          <span class="p">(</span><span class="s1">&#39;wildtype&#39;</span><span class="p">,</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">codons</span><span class="p">[</span><span class="n">r</span><span class="p">]</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sites</span><span class="p">])]</span> <span class="o">+</span>
                         <span class="p">[(</span><span class="n">codon</span><span class="p">,</span> <span class="n">codoncounts</span><span class="p">[</span><span class="n">codon</span><span class="p">])</span> <span class="k">for</span> <span class="n">codon</span> <span class="ow">in</span> <span class="n">CODONS</span><span class="p">]</span>
                         <span class="p">))</span>
            <span class="n">counts_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">countfile</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="k">assert</span> <span class="nb">all</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">,</span> <span class="n">countfiles</span><span class="p">))</span>

        <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;library&#39;</span><span class="p">:</span><span class="n">liblist</span><span class="p">,</span>
                             <span class="s1">&#39;sample&#39;</span><span class="p">:</span><span class="n">samplelist</span><span class="p">,</span>
                             <span class="s1">&#39;countfile&#39;</span><span class="p">:</span><span class="n">countfiles</span><span class="p">})</span></div>

<div class="viewcode-block" id="CodonVariantTable.classifyVariants"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.classifyVariants">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">classifyVariants</span><span class="p">(</span><span class="n">df</span><span class="p">,</span>
                         <span class="o">*</span><span class="p">,</span>
                         <span class="n">variant_class_col</span><span class="o">=</span><span class="s1">&#39;variant_class&#39;</span><span class="p">,</span>
                         <span class="n">max_aa</span><span class="o">=</span><span class="mi">2</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Classifies codon variants in `df`.</span>

<span class="sd">        Args:</span>
<span class="sd">            `df` (pandas DataFrame)</span>
<span class="sd">                Must have columns named &#39;aa_substitutions&#39;,</span>
<span class="sd">                &#39;n_aa_substitutions&#39;, and &#39;n_codon_substitutions&#39;.</span>
<span class="sd">                For instance, a data frame of this type can</span>
<span class="sd">                be obtained via the `variant_count_df` or</span>
<span class="sd">                `barcode_variant_df` of a :class:`CodonVariantTable`,</span>
<span class="sd">                or via :meth:`CodonVariantTable.func_scores`.</span>
<span class="sd">            `variant_class_col` (str)</span>
<span class="sd">                Name of column added to `df` that contains</span>
<span class="sd">                variant classification. Overwritten if already exists.</span>
<span class="sd">            `max_aa` (int)</span>
<span class="sd">                When classifying variants, group all with &gt;=</span>
<span class="sd">                this many amino-acid mutations.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A copy of `df` with the column specified by</span>
<span class="sd">            `variant_class_col` classifying variants as:</span>

<span class="sd">              - &#39;wildtype&#39;: no codon mutations</span>

<span class="sd">              - &#39;synonymous&#39;: only synonymous codon mutations</span>

<span class="sd">              - &#39;stop&#39;: at least one stop-codon mutation</span>

<span class="sd">              - &#39;{n_aa} nonsynonymous&#39; where `n_aa` is the</span>
<span class="sd">                number of amino-acid mutations, or is &#39;&gt;{max_aa}&#39;</span>
<span class="sd">                if there are more than `max_aa` such mutations.</span>

<span class="sd">        &gt;&gt;&gt; df = pd.DataFrame.from_records(</span>
<span class="sd">        ...         [(&#39;AAA&#39;, &#39;&#39;, 0, 0),</span>
<span class="sd">        ...          (&#39;AAG&#39;, &#39;&#39;, 0, 1),</span>
<span class="sd">        ...          (&#39;ATA&#39;, &#39;M1* G5K&#39;, 2, 3),</span>
<span class="sd">        ...          (&#39;GAA&#39;, &#39;G5H&#39;, 1, 2),</span>
<span class="sd">        ...          (&#39;CTT&#39;, &#39;M1C G5C&#39;, 2, 3),</span>
<span class="sd">        ...          (&#39;CTT&#39;, &#39;M1A L3T G5C&#39;, 3, 3),</span>
<span class="sd">        ...          ],</span>
<span class="sd">        ...         columns=[&#39;barcode&#39;, &#39;aa_substitutions&#39;,</span>
<span class="sd">        ...                  &#39;n_aa_substitutions&#39;, &#39;n_codon_substitutions&#39;]</span>
<span class="sd">        ...         )</span>
<span class="sd">        &gt;&gt;&gt; CodonVariantTable.classifyVariants(df)</span>
<span class="sd">          barcode aa_substitutions  n_aa_substitutions  n_codon_substitutions      variant_class</span>
<span class="sd">        0     AAA                                    0                      0           wildtype</span>
<span class="sd">        1     AAG                                    0                      1         synonymous</span>
<span class="sd">        2     ATA          M1* G5K                   2                      3               stop</span>
<span class="sd">        3     GAA              G5H                   1                      2    1 nonsynonymous</span>
<span class="sd">        4     CTT          M1C G5C                   2                      3  &gt;=2 nonsynonymous</span>
<span class="sd">        5     CTT      M1A L3T G5C                   3                      3  &gt;=2 nonsynonymous</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">req_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;aa_substitutions&#39;</span><span class="p">,</span> <span class="s1">&#39;n_aa_substitutions&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;n_codon_substitutions&#39;</span><span class="p">]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">req_cols</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="nb">set</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;`df` does not have columns </span><span class="si">{req_cols}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">def</span> <span class="nf">_classify_func</span><span class="p">(</span><span class="n">row</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;n_codon_substitutions&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">return</span> <span class="s1">&#39;wildtype&#39;</span>
            <span class="k">elif</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;n_aa_substitutions&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">return</span> <span class="s1">&#39;synonymous&#39;</span>
            <span class="k">elif</span> <span class="s1">&#39;*&#39;</span> <span class="ow">in</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;aa_substitutions&#39;</span><span class="p">]:</span>
                <span class="k">return</span> <span class="s1">&#39;stop&#39;</span>
            <span class="k">elif</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;n_aa_substitutions&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">max_aa</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">f</span><span class="s2">&quot;</span><span class="si">{row[&#39;n_aa_substitutions&#39;]}</span><span class="s2"> nonsynonymous&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">f</span><span class="s2">&quot;&gt;=</span><span class="si">{max_aa}</span><span class="s2"> nonsynonymous&quot;</span>

        <span class="k">return</span> <span class="n">df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="o">**</span><span class="p">{</span><span class="n">variant_class_col</span><span class="p">:</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                                               <span class="n">x</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">_classify_func</span><span class="p">,</span>
                                                       <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
                            <span class="p">})</span></div>


<div class="viewcode-block" id="CodonVariantTable.addMergedLibraries"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.addMergedLibraries">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">addMergedLibraries</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">all_lib</span><span class="o">=</span><span class="s1">&#39;all libraries&#39;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Add data to `df` for all libraries merged.</span>

<span class="sd">        Args:</span>
<span class="sd">            `df` (pandas DataFrame)</span>
<span class="sd">                DataFrame that includes columns named</span>
<span class="sd">                &quot;library&quot; and &quot;barcode&quot;.</span>
<span class="sd">            `all_lib` (str)</span>
<span class="sd">                Name given to library that is merge of all</span>
<span class="sd">                other libraries.</span>

<span class="sd">        Returns:</span>
<span class="sd">            If `df` only has data for one library, just returns</span>
<span class="sd">            `df`. Otherwise returns a copy of `df` that has a</span>
<span class="sd">            new library with the name given by `all_lib`</span>
<span class="sd">            and contains the data for all individual libraries,</span>
<span class="sd">            with the &quot;barcode&quot; column giving the original</span>
<span class="sd">            library name followed by a hyphen and the barcode.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">libs</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">library</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">libs</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">df</span>

        <span class="k">if</span> <span class="n">all_lib</span> <span class="ow">in</span> <span class="n">libs</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;library </span><span class="si">{all_lib}</span><span class="s2"> already exists&quot;</span><span class="p">)</span>

        <span class="n">df</span> <span class="o">=</span> <span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">df</span><span class="p">,</span>
                             <span class="n">df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                                <span class="n">barcode</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                                    <span class="n">x</span><span class="o">.</span><span class="n">library</span><span class="o">.</span><span class="n">str</span>
                                     <span class="o">.</span><span class="n">cat</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">barcode</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">),</span>
                                <span class="n">library</span><span class="o">=</span><span class="n">all_lib</span><span class="p">)</span>
                             <span class="p">],</span>
                            <span class="n">axis</span><span class="o">=</span><span class="s1">&#39;index&#39;</span><span class="p">,</span>
                            <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                            <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
              <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">library</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                              <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
                               <span class="n">x</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">],</span>
                               <span class="n">categories</span><span class="o">=</span><span class="n">libs</span> <span class="o">+</span> <span class="p">[</span><span class="n">all_lib</span><span class="p">],</span>
                               <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                      <span class="p">)</span>
              <span class="p">)</span>

        <span class="k">return</span> <span class="n">df</span></div>


    <span class="k">def</span> <span class="nf">_getPlotData</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">libraries</span><span class="p">,</span> <span class="n">samples</span><span class="p">,</span> <span class="n">min_support</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Gets data to plot from library and sample filters.</span>

<span class="sd">        Args:</span>
<span class="sd">            `libraries`, `samples`, `min_support` have meaning as</span>
<span class="sd">            for :class:`CodonVariantTable.plotNumMutsHistogram`.</span>

<span class="sd">        Returns:</span>
<span class="sd">            The 3-tuple `(df, nlibraries, nsamples)` where:</span>

<span class="sd">                - `df`: DataFrame with data to plot.</span>

<span class="sd">                - `nlibraries`: number of libraries being plotted.</span>

<span class="sd">                - `nsamples`: number of samples being plotted.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">samples</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">barcode_variant_df</span>
                  <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">sample</span><span class="o">=</span><span class="s1">&#39;barcoded variants&#39;</span><span class="p">)</span>
                  <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">count</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
                  <span class="p">)</span>
        <span class="k">elif</span> <span class="n">samples</span> <span class="o">==</span> <span class="s1">&#39;all&#39;</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">variant_count_df</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;no samples have been added&#39;</span><span class="p">)</span>
            <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">variant_count_df</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">samples</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="n">all_samples</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">itertools</span><span class="o">.</span><span class="n">chain</span><span class="o">.</span><span class="n">from_iterable</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">samples</span><span class="p">(</span><span class="n">lib</span><span class="p">)</span> <span class="k">for</span> <span class="n">lib</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">libraries</span><span class="p">))</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">all_samples</span><span class="o">.</span><span class="n">issuperset</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">samples</span><span class="p">)):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid sample(s) in </span><span class="si">{samples}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">samples</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">samples</span><span class="p">)):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;duplicate samples in </span><span class="si">{samples}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">variant_count_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;sample in @samples&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid `samples` </span><span class="si">{samples}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;variant_call_support &gt;= @min_support&#39;</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;no samples </span><span class="si">{samples}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">nsamples</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>

        <span class="k">if</span> <span class="n">libraries</span> <span class="o">==</span> <span class="s1">&#39;all&#39;</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">addMergedLibraries</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">libraries</span> <span class="o">==</span> <span class="s1">&#39;all_only&#39;</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">addMergedLibraries</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
                  <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;library == &quot;all libraries&quot;&#39;</span><span class="p">)</span>
                  <span class="p">)</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">libraries</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">libraries</span><span class="p">)</span><span class="o">.</span><span class="n">issuperset</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">libraries</span><span class="p">)):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid library in </span><span class="si">{libraries}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">libraries</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">libraries</span><span class="p">)):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;duplicate library in </span><span class="si">{libraries}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;library in @libraries&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid `libraries` </span><span class="si">{libraries}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;no libraries </span><span class="si">{libraries}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">nlibraries</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>

        <span class="k">return</span> <span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">nlibraries</span><span class="p">,</span> <span class="n">nsamples</span><span class="p">)</span>


<div class="viewcode-block" id="CodonVariantTable.codonToAAMuts"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.CodonVariantTable.codonToAAMuts">[docs]</a>    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">codonToAAMuts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">codon_mut_str</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Converts string of codon mutations to amino-acid mutations.</span>

<span class="sd">        Args:</span>
<span class="sd">            `codon_mut_str` (str)</span>
<span class="sd">                Codon mutations, delimited by a space and in</span>
<span class="sd">                1, 2, ... numbering.</span>

<span class="sd">        Returns:</span>
<span class="sd">            String with amino acid mutations in 1, 2, ... numbering.</span>

<span class="sd">        &gt;&gt;&gt; CodonVariantTable.codonToAAMuts(&#39;ATG1GTG GGA2GGC TGA3AGA&#39;)</span>
<span class="sd">        &#39;M1V *3R&#39;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">aa_muts</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">codon_mut_str</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">():</span>
            <span class="n">m</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s1">&#39;^(?P&lt;wt&gt;[ATGC]</span><span class="si">{3}</span><span class="s1">)(?P&lt;r&gt;\d+)(?P&lt;mut&gt;[ATGC]</span><span class="si">{3}</span><span class="s1">)$&#39;</span><span class="p">,</span>
                         <span class="n">mut</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">m</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid mutation </span><span class="si">{mut}</span><span class="s2"> in </span><span class="si">{codon_mut_str}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">r</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">aa_muts</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;duplicate codon mutation for </span><span class="si">{r}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">wt_codon</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;wt&#39;</span><span class="p">)</span>
            <span class="n">mut_codon</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;mut&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">wt_codon</span> <span class="o">==</span> <span class="n">mut_codon</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid mutation </span><span class="si">{mut}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">wt_aa</span> <span class="o">=</span> <span class="n">CODON_TO_AA</span><span class="p">[</span><span class="n">wt_codon</span><span class="p">]</span>
            <span class="n">mut_aa</span> <span class="o">=</span> <span class="n">CODON_TO_AA</span><span class="p">[</span><span class="n">mut_codon</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">wt_aa</span> <span class="o">!=</span> <span class="n">mut_aa</span><span class="p">:</span>
                <span class="n">aa_muts</span><span class="p">[</span><span class="n">r</span><span class="p">]</span> <span class="o">=</span> <span class="n">f</span><span class="s2">&quot;</span><span class="si">{wt_aa}{r}{mut_aa}</span><span class="s2">&quot;</span>

        <span class="k">return</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">mut_str</span> <span class="k">for</span> <span class="n">r</span><span class="p">,</span> <span class="n">mut_str</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">aa_muts</span><span class="o">.</span><span class="n">items</span><span class="p">())])</span></div>


    <span class="k">def</span> <span class="nf">_sortCodonMuts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">mut_str</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Sort space-delimited codon mutations and make uppercase.</span>

<span class="sd">        &gt;&gt;&gt; geneseq = &#39;ATGGGATGA&#39;</span>
<span class="sd">        &gt;&gt;&gt; with tempfile.NamedTemporaryFile(mode=&#39;w&#39;) as f:</span>
<span class="sd">        ...     _ = f.write(&#39;library,barcode,substitutions,variant_call_support&#39;)</span>
<span class="sd">        ...     f.flush()</span>
<span class="sd">        ...     variants = CodonVariantTable(</span>
<span class="sd">        ...                 barcode_variant_file=f.name,</span>
<span class="sd">        ...                 geneseq=geneseq</span>
<span class="sd">        ...                 )</span>
<span class="sd">        &gt;&gt;&gt; variants._sortCodonMuts(&#39;GGA2CGT ATG1GTG&#39;)</span>
<span class="sd">        &#39;ATG1GTG GGA2CGT&#39;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">muts</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">mut_str</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">():</span>
            <span class="n">m</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s1">&#39;^(?P&lt;wt&gt;[ATCG]</span><span class="si">{3}</span><span class="s1">)(?P&lt;r&gt;\d+)(?P&lt;mut&gt;[ACTG]</span><span class="si">{3}</span><span class="s1">)$&#39;</span><span class="p">,</span>
                         <span class="n">mut</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">m</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid codon mutation </span><span class="si">{mut}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">wt_codon</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;wt&#39;</span><span class="p">)</span>
            <span class="n">r</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">))</span>
            <span class="n">mut_codon</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;mut&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">wt_codon</span> <span class="o">==</span> <span class="n">mut_codon</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid codon mutation </span><span class="si">{mut}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">r</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sites</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid site in codon mutation </span><span class="si">{mut}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">wt_codon</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">codons</span><span class="p">[</span><span class="n">r</span><span class="p">]:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid wt in codon mutation </span><span class="si">{mut}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">muts</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;duplicate mutation at codon </span><span class="si">{mut}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">muts</span><span class="p">[</span><span class="n">r</span><span class="p">]</span> <span class="o">=</span> <span class="n">mut</span>
        <span class="k">return</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">mut</span> <span class="k">for</span> <span class="n">r</span><span class="p">,</span> <span class="n">mut</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">muts</span><span class="o">.</span><span class="n">items</span><span class="p">()))</span>


    <span class="k">def</span> <span class="nf">_ntToCodonMuts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">nt_mut_str</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Converts string of nucleotide mutations to codon mutations.</span>

<span class="sd">        Args:</span>
<span class="sd">            `nt_mut_str` (str)</span>
<span class="sd">                Nucleotide mutations, delimited by a space and in</span>
<span class="sd">                1, 2, ... numbering.</span>

<span class="sd">        Returns:</span>
<span class="sd">            String with codon mutations in 1, 2, ... numbering of</span>
<span class="sd">            codon sites.</span>

<span class="sd">        &gt;&gt;&gt; geneseq = &#39;ATGGGATGA&#39;</span>
<span class="sd">        &gt;&gt;&gt; with tempfile.NamedTemporaryFile(mode=&#39;w&#39;) as f:</span>
<span class="sd">        ...     _ = f.write(&#39;library,barcode,substitutions,variant_call_support&#39;)</span>
<span class="sd">        ...     f.flush()</span>
<span class="sd">        ...     variants = CodonVariantTable(</span>
<span class="sd">        ...                 barcode_variant_file=f.name,</span>
<span class="sd">        ...                 geneseq=geneseq</span>
<span class="sd">        ...                 )</span>
<span class="sd">        &gt;&gt;&gt; variants._ntToCodonMuts(&#39;A1G G4C A6T&#39;)</span>
<span class="sd">        &#39;ATG1GTG GGA2CGT&#39;</span>
<span class="sd">        &gt;&gt;&gt; variants._ntToCodonMuts(&#39;G4C A6T A1G&#39;)</span>
<span class="sd">        &#39;ATG1GTG GGA2CGT&#39;</span>
<span class="sd">        &gt;&gt;&gt; variants._ntToCodonMuts(&#39;A1G G4C G6T&#39;)</span>
<span class="sd">        Traceback (most recent call last):</span>
<span class="sd">        ...</span>
<span class="sd">        ValueError: nucleotide 6 should be A not G</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">mut_codons</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">nt_mut_str</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">():</span>
            <span class="n">m</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s1">&#39;^(?P&lt;wt&gt;[ATCG])(?P&lt;i&gt;\d+)(?P&lt;mut&gt;[ATCG])$&#39;</span><span class="p">,</span> <span class="n">mut</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">m</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid mutation </span><span class="si">{mut}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">wt_nt</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;wt&#39;</span><span class="p">)</span>
            <span class="n">i</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;i&#39;</span><span class="p">))</span>
            <span class="n">mut_nt</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;mut&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">wt_nt</span> <span class="o">==</span> <span class="n">mut_nt</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid mutation </span><span class="si">{mut}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">&gt;</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">geneseq</span><span class="p">)</span> <span class="ow">or</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid nucleotide site </span><span class="si">{i}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">geneseq</span><span class="p">[</span><span class="n">i</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="n">wt_nt</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;nucleotide </span><span class="si">{i}</span><span class="s2"> should be &quot;</span>
                                 <span class="n">f</span><span class="s2">&quot;</span><span class="si">{self.geneseq[i - 1]}</span><span class="s2"> not </span><span class="si">{wt_nt}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">icodon</span> <span class="o">=</span> <span class="p">(</span><span class="n">i</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">//</span> <span class="mi">3</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="n">i_nt</span> <span class="o">=</span> <span class="p">(</span><span class="n">i</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">%</span> <span class="mi">3</span>
            <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">codons</span><span class="p">[</span><span class="n">icodon</span><span class="p">][</span><span class="n">i_nt</span><span class="p">]</span> <span class="o">==</span> <span class="n">wt_nt</span>
            <span class="k">if</span> <span class="n">i_nt</span> <span class="ow">in</span> <span class="n">mut_codons</span><span class="p">[</span><span class="n">icodon</span><span class="p">]:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;duplicate mutations </span><span class="si">{i_nt}</span><span class="s2"> in </span><span class="si">{icodon}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">mut_codons</span><span class="p">[</span><span class="n">icodon</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">((</span><span class="n">i_nt</span><span class="p">,</span> <span class="n">mut_nt</span><span class="p">))</span>

        <span class="n">codon_mut_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">r</span><span class="p">,</span> <span class="n">r_muts</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">mut_codons</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
            <span class="n">wt_codon</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">codons</span><span class="p">[</span><span class="n">r</span><span class="p">]</span>
            <span class="n">mut_codon</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">wt_codon</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">mut_nt</span> <span class="ow">in</span> <span class="n">r_muts</span><span class="p">:</span>
                <span class="n">mut_codon</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">mut_nt</span>
            <span class="n">codon_mut_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;</span><span class="si">{wt_codon}{r}</span><span class="s2">{&#39;&#39;.join(mut_codon)}&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">codon_mut_list</span><span class="p">)</span></div>



<div class="viewcode-block" id="simulateSampleCounts"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.simulateSampleCounts">[docs]</a><span class="k">def</span> <span class="nf">simulateSampleCounts</span><span class="p">(</span><span class="o">*</span><span class="p">,</span>
                         <span class="n">variants</span><span class="p">,</span>
                         <span class="n">phenotype_func</span><span class="p">,</span>
                         <span class="n">variant_error_rate</span><span class="p">,</span>
                         <span class="n">pre_sample</span><span class="p">,</span>
                         <span class="n">post_samples</span><span class="p">,</span>
                         <span class="n">pre_sample_name</span><span class="o">=</span><span class="s1">&#39;pre-selection&#39;</span><span class="p">,</span>
                         <span class="n">seed</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Simulate pre- and post-selection variant counts.</span>

<span class="sd">    Simulate variant counts for experiment where barcodes</span>
<span class="sd">    are sequenced pre- and post-selection.</span>

<span class="sd">    Args:</span>
<span class="sd">        `variants` (:class:`CodonVariantTable`)</span>
<span class="sd">           Holds variants used in simulation.</span>
<span class="sd">        `phenotype_func` (function)</span>
<span class="sd">            Takes a row from `variants.barcode_variant_df`</span>
<span class="sd">            and returns the phenotype. Typically this is</span>
<span class="sd">            calculated from the &quot;aa_substitutions&quot; or</span>
<span class="sd">            &quot;codon_substitutions&quot; column. The phenotype is a</span>
<span class="sd">            number &gt;= 0, and represents the expected enrichment of</span>
<span class="sd">            the variant relative to wildtype post-selection (values</span>
<span class="sd">            &gt; 1 indicate beneficial). For instance, you could pass</span>
<span class="sd">            :meth:`PhenotypeSimulator.observedPhenotype`.</span>
<span class="sd">        `variant_error_rate` (float)</span>
<span class="sd">            Rate at which variants in `variants` are</span>
<span class="sd">            mis-called. Provide the probability that a</span>
<span class="sd">            variant has a spuriously called (or missing)</span>
<span class="sd">            codon mutation: with this probability, each</span>
<span class="sd">            variant then has a random codon mutation added</span>
<span class="sd">            or taken away before being passed to</span>
<span class="sd">            `phenotype_func`. Designed to simulate the fact</span>
<span class="sd">            that the method used to link barcodes to variants</span>
<span class="sd">            is probably not perfect.</span>
<span class="sd">        `pre_sample` (pandas DataFrame or dict)</span>
<span class="sd">            The counts of each variant pre-selection. You can</span>
<span class="sd">            specify counts or have them simulated:</span>

<span class="sd">                - To specify counts, provide a data frame with</span>
<span class="sd">                  columns &quot;library&quot;, &quot;barcode&quot;, and &quot;count&quot;</span>
<span class="sd">                  where &quot;count&quot; is the pre-selection counts.</span>

<span class="sd">                - To simulate, provide dict with keys &quot;total_count&quot;</span>
<span class="sd">                  and &quot;uniformity&quot;. For each library, we simulate</span>
<span class="sd">                  pre-selection counts as a draw of &quot;total_count&quot;</span>
<span class="sd">                  counts from a multinomial parameterized by</span>
<span class="sd">                  pre-selection frequences drawn from a Dirichlet</span>
<span class="sd">                  distribution with a concentration parameter of</span>
<span class="sd">                  of &quot;uniformity&quot; (larger uniformity values mean</span>
<span class="sd">                  more even libraries; 5 is reasonable value).</span>

<span class="sd">        `post_samples` (dict)</span>
<span class="sd">            Nested dict indicating post-selection samples.</span>
<span class="sd">            Keyed by name of each post-selection sample, with</span>
<span class="sd">            value being another dict. The post-selection counts</span>
<span class="sd">            are drawn from a multinomial parameterized</span>
<span class="sd">            by the pre-selection frequencies (frequencies calculated from</span>
<span class="sd">            counts if `pre_sample` gives counts, Dirichlet frequencies</span>
<span class="sd">            if `pre_sample` is simulated). Add noise by the</span>
<span class="sd">            following parameters provided in the dict for each sample:</span>

<span class="sd">                - &quot;total_count&quot;: total overall counts per library.</span>

<span class="sd">                - &quot;noise&quot;: add additional noise to selection by</span>
<span class="sd">                  multiplying phenotype times a random variable</span>
<span class="sd">                  drawn from a normal distribution with mean 1</span>
<span class="sd">                  and this standard deviation (truncated at lower</span>
<span class="sd">                  end to zero). Set noise to 0 for no noise.</span>

<span class="sd">                - &quot;bottleneck&quot;: put the pre-selection frequencies</span>
<span class="sd">                  through a bottleneck of this size, then re-calcuate</span>
<span class="sd">                  initial frequencies that selection acts upon. Set</span>
<span class="sd">                  to `None` for no bottleneck.</span>

<span class="sd">        `pre_sample_name` (str)</span>
<span class="sd">            Name used for the pre-selection sample.</span>
<span class="sd">        `seed` (None or int)</span>
<span class="sd">            If not `None`, random number seed set before executing</span>
<span class="sd">            function (sets `scipy.random.seed`).</span>

<span class="sd">    Returns:</span>
<span class="sd">        A pandas DataFrame with the following columns:</span>

<span class="sd">            - &quot;library&quot;</span>

<span class="sd">            - &quot;barcode&quot;</span>

<span class="sd">            - &quot;sample&quot;</span>

<span class="sd">            - &quot;count&quot;</span>

<span class="sd">        The first two columns indicate the library and barcode</span>
<span class="sd">        for each variant as in the `barcode_variant_df` attribute</span>
<span class="sd">        of `variants`, the &quot;sample&quot; and &quot;count&quot; columns give counts</span>
<span class="sd">        for each sample.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">seed</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">scipy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">pre_sample_name</span> <span class="ow">in</span> <span class="n">post_samples</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;`pre_sample_name` is in `post_samples`&#39;</span><span class="p">)</span>

    <span class="c1">#-----------------------------------------</span>
    <span class="c1"># internal function</span>
    <span class="k">def</span> <span class="nf">_add_variant_errors</span><span class="p">(</span><span class="n">codon_substitutions</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Add errors to variant according to `variant_error_rate`.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">scipy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">()</span> <span class="o">&lt;</span> <span class="n">variant_error_rate</span><span class="p">:</span>
            <span class="n">muts</span> <span class="o">=</span> <span class="n">codon_substitutions</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">muts</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">scipy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">()</span> <span class="o">&lt;</span> <span class="mf">0.5</span><span class="p">:</span>
                <span class="c1"># add mutation</span>
                <span class="n">mutatedsites</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span>
                        <span class="nb">int</span><span class="p">,</span>
                        <span class="p">[</span><span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s1">&#39;^[ATGC]</span><span class="si">{3}</span><span class="s1">(?P&lt;site&gt;\d+)[ATGC]</span><span class="si">{3}</span><span class="s1">$&#39;</span><span class="p">,</span>
                                  <span class="n">mut</span><span class="p">)</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;site&#39;</span><span class="p">)</span>
                         <span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">muts</span><span class="p">]</span>
                        <span class="p">))</span>
                <span class="n">unmutatedsites</span> <span class="o">=</span> <span class="p">[</span><span class="n">r</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">sites</span>
                                  <span class="k">if</span> <span class="n">r</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">mutatedsites</span><span class="p">]</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">unmutatedsites</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s2">&quot;variant already has all mutations&quot;</span><span class="p">)</span>
                <span class="n">errorsite</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">unmutatedsites</span><span class="p">)</span>
                <span class="n">wtcodon</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">codons</span><span class="p">[</span><span class="n">errorsite</span><span class="p">]</span>
                <span class="n">mutcodon</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">([</span><span class="n">c</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">CODONS</span>
                                                <span class="k">if</span> <span class="n">c</span> <span class="o">!=</span> <span class="n">wtcodon</span><span class="p">])</span>
                <span class="n">muts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{wtcodon}{errorsite}{mutcodon}</span><span class="s1">&#39;</span><span class="p">)</span>
                <span class="k">return</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">muts</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># remove mutation</span>
                <span class="n">muts</span> <span class="o">=</span> <span class="n">muts</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">scipy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">muts</span><span class="p">)))</span>
                <span class="k">return</span> <span class="n">muts</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">codon_substitutions</span>
    <span class="c1">#-----------------------------------------</span>

    <span class="n">barcode_variant_df</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">variants</span><span class="o">.</span><span class="n">barcode_variant_df</span>
        <span class="p">[[</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;codon_substitutions&#39;</span><span class="p">]]</span>
        <span class="o">.</span><span class="n">assign</span><span class="p">(</span>
            <span class="n">codon_substitutions</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">codon_substitutions</span>
                                <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">_add_variant_errors</span><span class="p">),</span>
            <span class="n">aa_substitutions</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">codon_substitutions</span>
                             <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">CodonVariantTable</span><span class="o">.</span><span class="n">codonToAAMuts</span><span class="p">),</span>
            <span class="n">phenotype</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">phenotype_func</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="p">[[</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;phenotype&#39;</span><span class="p">]]</span>
        <span class="p">)</span>

    <span class="n">libraries</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">libraries</span>

    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">pre_sample</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
        <span class="c1"># pre-sample counts specified</span>
        <span class="n">req_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">set</span><span class="p">(</span><span class="n">req_cols</span><span class="p">)</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">pre_sample</span><span class="o">.</span><span class="n">columns</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;pre_sample lacks cols </span><span class="si">{req_cols}</span><span class="s2">:&quot;</span>
                             <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="si">{pre_sample}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;barcode&#39;</span><span class="p">]</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">pre_sample</span><span class="p">[</span><span class="n">cols</span><span class="p">]</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">cols</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span> <span class="o">!=</span>
            <span class="n">barcode_variant_df</span><span class="p">[</span><span class="n">cols</span><span class="p">]</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">cols</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
            <span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;pre_sample DataFrame lacks required &quot;</span>
                             <span class="s2">&quot;library and barcode columns&quot;</span><span class="p">)</span>
        <span class="n">barcode_variant_df</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">barcode_variant_df</span>
                <span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">pre_sample</span><span class="p">[</span><span class="n">req_cols</span><span class="p">],</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;barcode&#39;</span><span class="p">])</span>
                <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;count&#39;</span><span class="p">:</span><span class="n">pre_sample_name</span><span class="p">})</span>
                <span class="p">)</span>
        <span class="c1"># &quot;true&quot; pre-selection freqs are just input counts</span>
        <span class="n">nperlib</span> <span class="o">=</span> <span class="p">(</span><span class="n">barcode_variant_df</span>
                   <span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;library&#39;</span><span class="p">)</span>
                   <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="s1">&#39;total_count&#39;</span><span class="p">)</span>
                   <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
                   <span class="p">)</span>
        <span class="n">barcode_variant_df</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">barcode_variant_df</span>
                <span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">nperlib</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s1">&#39;library&#39;</span><span class="p">)</span>
                <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">pre_freqs</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="n">pre_sample_name</span><span class="p">]</span> <span class="o">/</span>
                                            <span class="n">x</span><span class="o">.</span><span class="n">total_count</span><span class="p">)</span>
                <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="s1">&#39;total_count&#39;</span><span class="p">)</span>
                <span class="p">)</span>

    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">pre_sample</span><span class="p">,</span> <span class="nb">dict</span><span class="p">):</span>
        <span class="n">pre_req_keys</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;uniformity&#39;</span><span class="p">,</span> <span class="s1">&#39;total_count&#39;</span><span class="p">}</span>
        <span class="k">if</span> <span class="nb">set</span><span class="p">(</span><span class="n">pre_sample</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="o">!=</span> <span class="n">pre_req_keys</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;pre_sample lacks required keys </span><span class="si">{pre_req_keys}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">pre_df_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">lib</span> <span class="ow">in</span> <span class="n">libraries</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">barcode_variant_df</span>
                <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;library == @lib&#39;</span><span class="p">)</span>
                <span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                    <span class="n">pre_freq</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">scipy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">dirichlet</span><span class="p">(</span>
                             <span class="n">pre_sample</span><span class="p">[</span><span class="s1">&#39;uniformity&#39;</span><span class="p">]</span> <span class="o">*</span>
                             <span class="n">scipy</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">))),</span>
                    <span class="n">count</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">scipy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">multinomial</span><span class="p">(</span>
                            <span class="n">pre_sample</span><span class="p">[</span><span class="s1">&#39;total_count&#39;</span><span class="p">],</span> <span class="n">x</span><span class="o">.</span><span class="n">pre_freq</span><span class="p">),</span>
                    <span class="n">sample</span><span class="o">=</span><span class="n">pre_sample_name</span>
                    <span class="p">)</span>
                <span class="p">)</span>
            <span class="n">pre_df_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
        <span class="n">barcode_variant_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">pre_df_list</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;pre_sample not DataFrame / dict: &quot;</span>
                         <span class="n">f</span><span class="s2">&quot;</span><span class="si">{pre_sample}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">,</span>
            <span class="s1">&#39;pre_freq&#39;</span><span class="p">,</span> <span class="s1">&#39;phenotype&#39;</span><span class="p">]</span>
    <span class="k">assert</span> <span class="nb">set</span><span class="p">(</span><span class="n">barcode_variant_df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">==</span> <span class="nb">set</span><span class="p">(</span><span class="n">cols</span><span class="p">),</span> <span class="p">(</span>
            <span class="n">f</span><span class="s2">&quot;cols = {set(cols)}</span><span class="se">\n</span><span class="s2">barcode_variant_df.columns = &quot;</span>
            <span class="n">f</span><span class="s2">&quot;{set(barcode_variant_df.columns)}&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">cols</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">post_samples</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;post_samples can&#39;t have key </span><span class="si">{col}</span><span class="s2">; &quot;</span>
                             <span class="s2">&quot;choose another sample name&quot;</span><span class="p">)</span>

    <span class="n">df_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">barcode_variant_df</span><span class="p">[</span><span class="n">cols</span><span class="p">[</span> <span class="p">:</span> <span class="mi">4</span><span class="p">]]]</span>

    <span class="k">def</span> <span class="nf">_bottleneck_freqs</span><span class="p">(</span><span class="n">pre_freq</span><span class="p">,</span> <span class="n">bottleneck</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">bottleneck</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">pre_freq</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">scipy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">multinomial</span><span class="p">(</span><span class="n">bottleneck</span><span class="p">,</span> <span class="n">pre_freq</span><span class="p">)</span> <span class="o">/</span> <span class="n">bottleneck</span>

    <span class="n">post_req_keys</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;bottleneck&#39;</span><span class="p">,</span> <span class="s1">&#39;noise&#39;</span><span class="p">,</span> <span class="s1">&#39;total_count&#39;</span><span class="p">}</span>
    <span class="k">for</span> <span class="n">lib</span><span class="p">,</span> <span class="p">(</span><span class="n">sample</span><span class="p">,</span> <span class="n">sample_dict</span><span class="p">)</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">product</span><span class="p">(</span>
            <span class="n">libraries</span><span class="p">,</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">post_samples</span><span class="o">.</span><span class="n">items</span><span class="p">())):</span>

        <span class="k">if</span> <span class="nb">set</span><span class="p">(</span><span class="n">sample_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="o">!=</span> <span class="n">post_req_keys</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;post_samples </span><span class="si">{sample}</span><span class="s2"> lacks keys </span><span class="si">{post_req_keys}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">lib_df</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">barcode_variant_df</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;library == @lib&#39;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">assign</span><span class="p">(</span>
                <span class="n">sample</span><span class="o">=</span><span class="n">sample</span><span class="p">,</span>
                <span class="c1"># simulated pre-selection freqs after bottleneck</span>
                <span class="n">bottleneck_freq</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                                <span class="n">_bottleneck_freqs</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">pre_freq</span><span class="p">,</span>
                                                  <span class="n">sample_dict</span><span class="p">[</span><span class="s1">&#39;bottleneck&#39;</span><span class="p">]),</span>
                <span class="c1"># post-selection freqs with noise</span>
                <span class="n">noise</span><span class="o">=</span><span class="n">scipy</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">scipy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">sample_dict</span><span class="p">[</span><span class="s1">&#39;noise&#39;</span><span class="p">]),</span>
                                 <span class="mi">0</span><span class="p">,</span> <span class="kc">None</span><span class="p">),</span>
                <span class="n">post_freq_nonorm</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                            <span class="n">x</span><span class="o">.</span><span class="n">bottleneck_freq</span> <span class="o">*</span> <span class="n">x</span><span class="o">.</span><span class="n">phenotype</span> <span class="o">*</span> <span class="n">x</span><span class="o">.</span><span class="n">noise</span><span class="p">,</span>
                <span class="n">post_freq</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">post_freq_nonorm</span> <span class="o">/</span>
                            <span class="n">x</span><span class="o">.</span><span class="n">post_freq_nonorm</span><span class="o">.</span><span class="n">sum</span><span class="p">(),</span>
                <span class="c1"># post-selection counts simulated from frequencies</span>
                <span class="n">count</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">scipy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">multinomial</span><span class="p">(</span>
                            <span class="n">sample_dict</span><span class="p">[</span><span class="s1">&#39;total_count&#39;</span><span class="p">],</span> <span class="n">x</span><span class="o">.</span><span class="n">post_freq</span><span class="p">)</span>
                <span class="p">)</span>
            <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;post_counts&#39;</span><span class="p">:</span><span class="n">sample</span><span class="p">})</span>
            <span class="p">[[</span><span class="s1">&#39;library&#39;</span><span class="p">,</span> <span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">]]</span>
            <span class="p">)</span>

        <span class="n">df_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lib_df</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">df_list</span><span class="p">)</span></div>


<div class="viewcode-block" id="PhenotypeSimulator"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.PhenotypeSimulator">[docs]</a><span class="k">class</span> <span class="nc">PhenotypeSimulator</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;Simulates phenotypes of variants under plausible model.</span>

<span class="sd">    Mutational effects on latent phenotype are simulated to follow</span>
<span class="sd">    compound normal distribution; latent phenotype maps to observed</span>
<span class="sd">    phenotype via sigmoid. This distinction between latent and</span>
<span class="sd">    observed phenotype parallel the &quot;global epistasis&quot; models of</span>
<span class="sd">    `Otwinoski et al &lt;https://doi.org/10.1073/pnas.1804015115&gt;`_ and</span>
<span class="sd">    `Sailer and Harms &lt;http://www.genetics.org/content/205/3/1079&gt;`_.</span>

<span class="sd">    Initialize with a codon sequence and random number seed.</span>
<span class="sd">    The effects of mutations on the latent phenotype are then drawn</span>
<span class="sd">    from a compound normal distribution biased to negative values.</span>

<span class="sd">    To calculate latent and observed phenotypes, pass rows of a</span>
<span class="sd">    :meth:`CodonVariantTable.barcode_variant_df` to</span>
<span class="sd">    :meth:`PhenotypeSimulator.latentPhenotype` or</span>
<span class="sd">    :meth:`PhenotypeSimulator.observedPhenotype`.</span>

<span class="sd">    Args:</span>
<span class="sd">        `geneseq` (str)</span>
<span class="sd">            Codon sequence of wild-type gene.</span>
<span class="sd">        `seed` (int)</span>
<span class="sd">            Random number seed.</span>
<span class="sd">        `wt_latent` (float)</span>
<span class="sd">            Latent phenotype of wildtype.</span>
<span class="sd">        `norm_weights` (list of tuples)</span>
<span class="sd">            Specifies compound normal distribution of mutational</span>
<span class="sd">            effects on latent phenotype. Each tuple is</span>
<span class="sd">            `(weight, mean, sd)`, giving weight, mean, and standard</span>
<span class="sd">            deviation of each Gaussian in compound normal.</span>
<span class="sd">        `stop_effect` (float)</span>
<span class="sd">            Effect of stop codon at any position.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        `wt_latent` (float)</span>
<span class="sd">            Value for wildtype latent phenotype.</span>
<span class="sd">        `muteffects` (dict)</span>
<span class="sd">            Effects on latent phenotype of each amino-acid mutation.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">geneseq</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">wt_latent</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                 <span class="n">norm_weights</span><span class="o">=</span><span class="p">[(</span><span class="mf">0.4</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.5</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="p">(</span><span class="mf">0.6</span><span class="p">,</span> <span class="o">-</span><span class="mi">5</span><span class="p">,</span> <span class="mf">2.5</span><span class="p">)],</span>
                 <span class="n">stop_effect</span><span class="o">=-</span><span class="mi">10</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;See main class docstring for how to initialize.&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">wt_latent</span> <span class="o">=</span> <span class="n">wt_latent</span>

        <span class="c1"># simulate muteffects from compound normal distribution</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">muteffects</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">scipy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
        <span class="n">weights</span><span class="p">,</span> <span class="n">means</span><span class="p">,</span> <span class="n">sds</span> <span class="o">=</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">norm_weights</span><span class="p">)</span>
        <span class="n">cumweights</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">weights</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">icodon</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">geneseq</span><span class="p">)</span> <span class="o">//</span> <span class="mi">3</span><span class="p">):</span>
            <span class="n">wt_aa</span> <span class="o">=</span> <span class="n">CODON_TO_AA</span><span class="p">[</span><span class="n">geneseq</span><span class="p">[</span><span class="mi">3</span> <span class="o">*</span> <span class="n">icodon</span> <span class="p">:</span> <span class="mi">3</span> <span class="o">*</span> <span class="n">icodon</span> <span class="o">+</span> <span class="mi">3</span><span class="p">]]</span>
            <span class="k">for</span> <span class="n">mut_aa</span> <span class="ow">in</span> <span class="n">AAS_WITHSTOP</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">mut_aa</span> <span class="o">!=</span> <span class="n">wt_aa</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">mut_aa</span> <span class="o">==</span> <span class="s1">&#39;*&#39;</span><span class="p">:</span>
                        <span class="n">muteffect</span> <span class="o">=</span> <span class="n">stop_effect</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="c1"># choose Gaussian from compound normal</span>
                        <span class="n">i</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">argmin</span><span class="p">(</span><span class="n">cumweights</span> <span class="o">&lt;</span> <span class="n">scipy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">())</span>
                        <span class="c1"># draw mutational effect from chosen Gaussian</span>
                        <span class="n">muteffect</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">means</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">sds</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">muteffects</span><span class="p">[</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{wt_aa}</span><span class="s1">{icodon + 1}</span><span class="si">{mut_aa}</span><span class="s1">&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">muteffect</span>

<div class="viewcode-block" id="PhenotypeSimulator.latentPhenotype"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.PhenotypeSimulator.latentPhenotype">[docs]</a>    <span class="k">def</span> <span class="nf">latentPhenotype</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">v</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns latent phenotype of a variant.</span>

<span class="sd">        Args:</span>
<span class="sd">            `v` (dict or row of pandas DataFrame)</span>
<span class="sd">                Must have key &#39;aa_substitutions&#39; that gives</span>
<span class="sd">                space de-limited list of amino-acid mutations.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">wt_latent</span> <span class="o">+</span> <span class="nb">sum</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">muteffects</span><span class="p">[</span><span class="n">m</span><span class="p">]</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span>
                                     <span class="n">v</span><span class="p">[</span><span class="s1">&#39;aa_substitutions&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">()])</span></div>

<div class="viewcode-block" id="PhenotypeSimulator.observedPhenotype"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.PhenotypeSimulator.observedPhenotype">[docs]</a>    <span class="k">def</span> <span class="nf">observedPhenotype</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">v</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Like `latentPhenotype` but returns observed phenotype.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">latentToObservedPhenotype</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">latentPhenotype</span><span class="p">(</span><span class="n">v</span><span class="p">))</span></div>

<div class="viewcode-block" id="PhenotypeSimulator.latentToObservedPhenotype"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.PhenotypeSimulator.latentToObservedPhenotype">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">latentToObservedPhenotype</span><span class="p">(</span><span class="n">latent</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Returns observed phenotype from latent phenotype.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="mi">1</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">math</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="o">-</span><span class="n">latent</span> <span class="o">-</span> <span class="mi">3</span><span class="p">))</span></div>

<div class="viewcode-block" id="PhenotypeSimulator.plotLatentVersusObservedPhenotype"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.PhenotypeSimulator.plotLatentVersusObservedPhenotype">[docs]</a>    <span class="k">def</span> <span class="nf">plotLatentVersusObservedPhenotype</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span>
            <span class="n">latent_min</span><span class="o">=-</span><span class="mi">15</span><span class="p">,</span> <span class="n">latent_max</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">npoints</span><span class="o">=</span><span class="mi">200</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Plots observed phenotype as function of latent phenotype.</span>

<span class="sd">        Plot includes a vertical line at wildtype latent phenotype.</span>

<span class="sd">        Args:</span>
<span class="sd">            `latent_min` (float)</span>
<span class="sd">                Smallest value of latent phenotype on plot.</span>
<span class="sd">            `latent_max` (float)</span>
<span class="sd">                Largest value of latent phenotype on plot.</span>
<span class="sd">            `npoints` (int)</span>
<span class="sd">                Plot a line fit to this many points.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A `plotnine &lt;https://plotnine.readthedocs.io&gt;`_</span>
<span class="sd">            plot; can be displayed in a Jupyter notebook with `p.draw()`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">latent</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="n">latent_min</span><span class="p">,</span> <span class="n">latent_max</span><span class="p">,</span> <span class="n">npoints</span><span class="p">)</span>
        <span class="n">p</span> <span class="o">=</span> <span class="p">(</span><span class="n">ggplot</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="nb">dict</span><span class="p">(</span><span class="n">latent</span><span class="o">=</span><span class="n">latent</span><span class="p">))</span>
                       <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">observed</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">latent</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
                                        <span class="bp">self</span><span class="o">.</span><span class="n">latentToObservedPhenotype</span><span class="p">)),</span>
                    <span class="n">aes</span><span class="p">(</span><span class="s1">&#39;latent&#39;</span><span class="p">,</span> <span class="s1">&#39;observed&#39;</span><span class="p">)</span>
                    <span class="p">)</span> <span class="o">+</span>
             <span class="n">geom_line</span><span class="p">()</span> <span class="o">+</span>
             <span class="n">geom_vline</span><span class="p">(</span><span class="n">xintercept</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">wt_latent</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">CBPALETTE</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                        <span class="n">linetype</span><span class="o">=</span><span class="s1">&#39;dashed&#39;</span><span class="p">)</span> <span class="o">+</span>
            <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mf">3.5</span><span class="p">,</span> <span class="mf">2.5</span><span class="p">))</span> <span class="o">+</span>
            <span class="n">xlab</span><span class="p">(</span><span class="s1">&#39;latent phenotype&#39;</span><span class="p">)</span> <span class="o">+</span>
            <span class="n">ylab</span><span class="p">(</span><span class="s1">&#39;observed phenotype&#39;</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="k">return</span> <span class="n">p</span></div>

<div class="viewcode-block" id="PhenotypeSimulator.plotMutsHistogram"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.PhenotypeSimulator.plotMutsHistogram">[docs]</a>    <span class="k">def</span> <span class="nf">plotMutsHistogram</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">latent_or_observed</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span>
            <span class="n">mutant_order</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="mi">30</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Plots distribution of phenotype for all mutants.</span>

<span class="sd">        Plot includes a vertical line at wildtype phenotype.</span>

<span class="sd">        Args:</span>
<span class="sd">            `latent_or_observed` (&quot;latent&quot; or &quot;observed&quot;)</span>
<span class="sd">                Which type of phenotype to plot.</span>
<span class="sd">            `mutant_order` (int)</span>
<span class="sd">                Plot mutations of this order. Currently only works</span>
<span class="sd">                for 1 (single mutants).</span>
<span class="sd">            `bins` (int)</span>
<span class="sd">                Number of bins in histogram.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A `plotnine &lt;https://plotnine.readthedocs.io&gt;`_</span>
<span class="sd">            plot; can be displayed in a Jupyter notebook with `p.draw()`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">mutant_order</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;only implemented for `mutant_order` of 1&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">latent_or_observed</span> <span class="o">==</span> <span class="s1">&#39;latent&#39;</span><span class="p">:</span>
            <span class="n">phenoFunc</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">latentPhenotype</span>
        <span class="k">elif</span> <span class="n">latent_or_observed</span> <span class="o">==</span> <span class="s1">&#39;observed&#39;</span><span class="p">:</span>
            <span class="n">phenoFunc</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">observedPhenotype</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;invalid value of `latent_or_observed`&#39;</span><span class="p">)</span>
        <span class="n">phenotypes</span> <span class="o">=</span> <span class="p">[</span><span class="n">phenoFunc</span><span class="p">({</span><span class="s1">&#39;aa_substitutions&#39;</span><span class="p">:</span><span class="n">m</span><span class="p">})</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span>
                      <span class="bp">self</span><span class="o">.</span><span class="n">muteffects</span><span class="o">.</span><span class="n">keys</span><span class="p">()]</span>
        <span class="n">p</span> <span class="o">=</span> <span class="p">(</span><span class="n">ggplot</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;phenotype&#39;</span><span class="p">:</span><span class="n">phenotypes</span><span class="p">}),</span>
                    <span class="n">aes</span><span class="p">(</span><span class="s1">&#39;phenotype&#39;</span><span class="p">))</span> <span class="o">+</span>
             <span class="n">geom_histogram</span><span class="p">(</span><span class="n">bins</span><span class="o">=</span><span class="n">bins</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mf">3.5</span><span class="p">,</span> <span class="mf">2.5</span><span class="p">))</span> <span class="o">+</span>
             <span class="n">ylab</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;number of </span><span class="si">{mutant_order}</span><span class="s2">-mutants&quot;</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">xlab</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;</span><span class="si">{latent_or_observed}</span><span class="s2"> phenotype&quot;</span><span class="p">)</span> <span class="o">+</span>
             <span class="n">geom_vline</span><span class="p">(</span><span class="n">xintercept</span><span class="o">=</span><span class="n">phenoFunc</span><span class="p">({</span><span class="s1">&#39;aa_substitutions&#39;</span><span class="p">:</span><span class="s1">&#39;&#39;</span><span class="p">}),</span>
                        <span class="n">color</span><span class="o">=</span><span class="n">CBPALETTE</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">linetype</span><span class="o">=</span><span class="s1">&#39;dashed&#39;</span><span class="p">)</span>
             <span class="p">)</span>
        <span class="k">return</span> <span class="n">p</span></div></div>



<div class="viewcode-block" id="tidy_split"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.tidy_split">[docs]</a><span class="k">def</span> <span class="nf">tidy_split</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">column</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="n">keep</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Split values of a column and expand so new DataFrame has one split</span>
<span class="sd">    value per row. Filters rows where the column is missing.</span>

<span class="sd">    Taken from https://stackoverflow.com/a/39946744</span>

<span class="sd">    Args:</span>
<span class="sd">        df : pandas DataFrame</span>
<span class="sd">            dataframe with the column to split and expand</span>
<span class="sd">        column : str</span>
<span class="sd">            the column to split and expand</span>
<span class="sd">        sep : str</span>
<span class="sd">            the string used to split the column&#39;s values</span>
<span class="sd">        keep : bool</span>
<span class="sd">            whether to retain the presplit value as it&#39;s own row</span>

<span class="sd">    Returns:</span>
<span class="sd">        pandas DataFrame</span>
<span class="sd">            Returns a dataframe with the same columns as `df`.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">indexes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
    <span class="n">new_values</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="n">column</span><span class="p">])</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">presplit</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">column</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)):</span>
        <span class="n">values</span> <span class="o">=</span> <span class="n">presplit</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">sep</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">keep</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">values</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">indexes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
            <span class="n">new_values</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">presplit</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">values</span><span class="p">:</span>
            <span class="n">indexes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
            <span class="n">new_values</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
    <span class="n">new_df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">indexes</span><span class="p">,</span> <span class="p">:]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">new_df</span><span class="p">[</span><span class="n">column</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_values</span>
    <span class="k">return</span> <span class="n">new_df</span></div>


<div class="viewcode-block" id="rarefyBarcodes"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.rarefyBarcodes">[docs]</a><span class="k">def</span> <span class="nf">rarefyBarcodes</span><span class="p">(</span><span class="n">barcodecounts</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span>
                   <span class="n">barcodecol</span><span class="o">=</span><span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="n">countcol</span><span class="o">=</span><span class="s1">&#39;count&#39;</span><span class="p">,</span>
                   <span class="n">maxpoints</span><span class="o">=</span><span class="mf">1e5</span><span class="p">,</span> <span class="n">logspace</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Rarefaction curve of barcode observations.</span>

<span class="sd">    Uses analytical formula for rarefaction defined</span>
<span class="sd">    `here &lt;https://en.wikipedia.org/wiki/Rarefaction_(ecology)#Derivation&gt;`_.</span>

<span class="sd">    Args:</span>
<span class="sd">        `barcodecounts` (pandas DataFrame)</span>
<span class="sd">            Has columns with names matching `barcodecol` and `countcol`.</span>
<span class="sd">        `barcodecol` (str)</span>
<span class="sd">            Name of column listing all unique barcodes.</span>
<span class="sd">        `countcol` (str)</span>
<span class="sd">            Name of column with observed counts of each barcode.</span>
<span class="sd">        `maxpoints` (int)</span>
<span class="sd">            Only calculate rarefaction curve at this many points.</span>
<span class="sd">            The benefit is that it can become very costly to calculate</span>
<span class="sd">            the curve for many points.</span>
<span class="sd">        `logspace` (bool)</span>
<span class="sd">            Logarithmically space the `maxpoint` points. If False,</span>
<span class="sd">            space them linearly.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A pandas DataFrame with the columns `ncounts` and</span>
<span class="sd">        `nbarcodes`, giving the number of unique barcodes</span>
<span class="sd">        observed for each total number observed counts.</span>

<span class="sd">    Here is a small example:</span>

<span class="sd">    &gt;&gt;&gt; barcodecounts = pd.DataFrame({&#39;barcode&#39;:[&#39;A&#39;, &#39;G&#39;, &#39;C&#39;, &#39;T&#39;],</span>
<span class="sd">    ...                               &#39;count&#39;:  [  4,   2,   1,   0]})</span>
<span class="sd">    &gt;&gt;&gt; rarefaction_curve = rarefyBarcodes(barcodecounts)</span>
<span class="sd">    &gt;&gt;&gt; rarefaction_curve</span>
<span class="sd">       ncounts  nbarcodes</span>
<span class="sd">    0        1   1.000000</span>
<span class="sd">    1        2   1.666667</span>
<span class="sd">    2        3   2.114286</span>
<span class="sd">    3        4   2.428571</span>
<span class="sd">    4        5   2.666667</span>
<span class="sd">    5        6   2.857143</span>
<span class="sd">    6        7   3.000000</span>

<span class="sd">    Verify that the result from this example matches what is</span>
<span class="sd">    obtained by random sampling:</span>

<span class="sd">    &gt;&gt;&gt; random.seed(1)</span>
<span class="sd">    &gt;&gt;&gt; barcodelist = []</span>
<span class="sd">    &gt;&gt;&gt; for tup in barcodecounts.itertuples(index=False):</span>
<span class="sd">    ...     barcodelist += [tup.barcode] * tup.count</span>
<span class="sd">    &gt;&gt;&gt; nrand = 10000</span>
<span class="sd">    &gt;&gt;&gt; ncounts = list(range(1, barcodecounts[&#39;count&#39;].sum() + 1))</span>
<span class="sd">    &gt;&gt;&gt; nbarcodes = []</span>
<span class="sd">    &gt;&gt;&gt; for ncount in ncounts:</span>
<span class="sd">    ...     nbarcodes.append(sum(len(set(random.sample(barcodelist, ncount)))</span>
<span class="sd">    ...                      for _ in range(nrand)) / nrand)</span>
<span class="sd">    &gt;&gt;&gt; sim_rarefaction_curve = pd.DataFrame(dict(ncounts=ncounts,</span>
<span class="sd">    ...                                           nbarcodes=nbarcodes))</span>
<span class="sd">    &gt;&gt;&gt; scipy.allclose(rarefaction_curve, sim_rarefaction_curve, atol=1e-2)</span>
<span class="sd">    True</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">barcodecounts</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">barcodecounts</span><span class="p">[</span><span class="n">barcodecol</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;non-unique barcodes in `barcodecounts`&#39;</span><span class="p">)</span>

    <span class="c1"># follow nomenclature at</span>
    <span class="c1"># https://en.wikipedia.org/wiki/Rarefaction_(ecology)#Derivation</span>
    <span class="n">Ni</span> <span class="o">=</span> <span class="n">barcodecounts</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">barcodecol</span><span class="p">)[</span><span class="n">countcol</span><span class="p">]</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span>
    <span class="n">N</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">Ni</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
    <span class="n">K</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">barcodecounts</span><span class="p">)</span>
    <span class="n">Mj</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">Counter</span><span class="p">(</span><span class="n">Ni</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

    <span class="n">Nk</span><span class="p">,</span> <span class="n">num</span> <span class="o">=</span> <span class="nb">map</span><span class="p">(</span><span class="n">scipy</span><span class="o">.</span><span class="n">array</span><span class="p">,</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">Mj</span><span class="o">.</span><span class="n">items</span><span class="p">()))</span>

    <span class="c1"># use simplification that (N - Ni)Cr(n) / (N)Cr(n) =</span>
    <span class="c1"># [(N - Ni)! * (N - n)!] / [N! * (N - Ni - n)!]</span>
    <span class="c1">#</span>
    <span class="c1"># Also use fact that gamma(x + 1) = x!</span>
    <span class="n">nbarcodes</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">lnFactorial_N</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">special</span><span class="o">.</span><span class="n">gammaln</span><span class="p">(</span><span class="n">N</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">logspace</span> <span class="ow">and</span> <span class="n">N</span> <span class="o">&gt;</span> <span class="n">maxpoints</span><span class="p">:</span>
        <span class="n">ncounts</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">scipy</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">scipy</span><span class="o">.</span><span class="n">logspace</span><span class="p">(</span>
                       <span class="n">math</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="mi">1</span><span class="p">),</span> <span class="n">math</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">N</span><span class="p">),</span>
                       <span class="n">num</span><span class="o">=</span><span class="nb">min</span><span class="p">(</span><span class="n">N</span><span class="p">,</span> <span class="n">maxpoints</span><span class="p">))</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;int&#39;</span><span class="p">)))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">ncounts</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">scipy</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">scipy</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span>
                       <span class="mi">1</span><span class="p">,</span> <span class="n">N</span><span class="p">,</span> <span class="n">num</span><span class="o">=</span><span class="nb">min</span><span class="p">(</span><span class="n">N</span><span class="p">,</span> <span class="n">maxpoints</span><span class="p">))</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;int&#39;</span><span class="p">)))</span>
    <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">ncounts</span><span class="p">:</span>
        <span class="n">lnFactorial_N_minus_n</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">special</span><span class="o">.</span><span class="n">gammaln</span><span class="p">(</span><span class="n">N</span> <span class="o">-</span> <span class="n">n</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">i</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">nonzero</span><span class="p">(</span><span class="n">N</span> <span class="o">-</span> <span class="n">Nk</span> <span class="o">-</span> <span class="n">n</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">)</span> <span class="c1"># indices where this is true</span>
        <span class="n">nbarcodes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="n">K</span> <span class="o">-</span> <span class="p">(</span><span class="n">num</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">*</span> <span class="n">scipy</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span>
                            <span class="n">scipy</span><span class="o">.</span><span class="n">special</span><span class="o">.</span><span class="n">gammaln</span><span class="p">(</span><span class="n">N</span> <span class="o">-</span> <span class="n">Nk</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">+</span>
                            <span class="n">lnFactorial_N_minus_n</span> <span class="o">-</span>
                            <span class="n">lnFactorial_N</span> <span class="o">-</span>
                            <span class="n">scipy</span><span class="o">.</span><span class="n">special</span><span class="o">.</span><span class="n">gammaln</span><span class="p">(</span><span class="n">N</span> <span class="o">-</span> <span class="n">Nk</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">-</span> <span class="n">n</span> <span class="o">+</span> <span class="mi">1</span><span class="p">))</span>
                    <span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
                <span class="p">)</span>
    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="nb">dict</span><span class="p">(</span><span class="n">ncounts</span><span class="o">=</span><span class="n">ncounts</span><span class="p">,</span> <span class="n">nbarcodes</span><span class="o">=</span><span class="n">nbarcodes</span><span class="p">))</span></div>


<div class="viewcode-block" id="getCumulVariantsByCount"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.getCumulVariantsByCount">[docs]</a><span class="k">def</span> <span class="nf">getCumulVariantsByCount</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">group_cols</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                            <span class="n">group_cols_as_str</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Get number of variants with observed &gt;= each number of times.</span>

<span class="sd">    Args:</span>
<span class="sd">        `df` (pandas DataFrame)</span>
<span class="sd">            A row for each variant, and a column named &quot;count&quot; giving</span>
<span class="sd">            how many times variant observed. Can have other columns.</span>
<span class="sd">        `group_cols` (None or list)</span>
<span class="sd">            Group by these columns and analyze each group separately.</span>
<span class="sd">        `group_cols_as_str` (bool)</span>
<span class="sd">            Explicitly convert any `group_cols` columns to str type.</span>
<span class="sd">            For some reason, this is needed if you are calling in ``R``</span>
<span class="sd">            using `reticulate &lt;https://rstudio.github.io/reticulate/&gt;`_.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A pandas Data Frame with columns named &quot;count&quot;, &quot;nvariants&quot;,</span>
<span class="sd">        and &quot;total_variants&quot; (plus any columns in `group_cols`).</span>
<span class="sd">        For each value of &quot;count&quot;, &quot;nvariants&quot; gives the number of</span>
<span class="sd">        variants with &gt;= that many counts. The &quot;total_variants&quot;</span>
<span class="sd">        column gives the total number of variants.</span>

<span class="sd">    Here is an example. First, create input Data Frame:</span>

<span class="sd">    &gt;&gt;&gt; df = pd.DataFrame(dict(</span>
<span class="sd">    ...          sample= [&#39;a&#39;, &#39;a&#39;, &#39;b&#39;, &#39;b&#39;, &#39;a&#39;, &#39;a&#39;],</span>
<span class="sd">    ...          count=  [  9,   0,   1,   4,   3,   3]))</span>

<span class="sd">    Now apply function, first **not** grouping by sample. Note</span>
<span class="sd">    how the sample column is therefore simply ignored and dropped:</span>

<span class="sd">    &gt;&gt;&gt; getCumulVariantsByCount(df)</span>
<span class="sd">       count  nvariants  total_variants</span>
<span class="sd">    0      9          1               6</span>
<span class="sd">    1      4          2               6</span>
<span class="sd">    2      3          4               6</span>
<span class="sd">    3      1          5               6</span>
<span class="sd">    4      0          6               6</span>

<span class="sd">    Now apply function, grouping by sample. Note how each sample</span>
<span class="sd">    is now analyzed separately:</span>

<span class="sd">    &gt;&gt;&gt; getCumulVariantsByCount(df, group_cols=[&#39;sample&#39;])</span>
<span class="sd">      sample  count  nvariants  total_variants</span>
<span class="sd">    0      a      9          1               4</span>
<span class="sd">    1      a      3          3               4</span>
<span class="sd">    2      a      0          4               4</span>
<span class="sd">    3      b      4          1               2</span>
<span class="sd">    4      b      1          2               2</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="s1">&#39;count&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;df does not have a column named &quot;count&quot;&#39;</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">group_cols</span><span class="p">:</span>
        <span class="n">drop_group_cols</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">group_cols</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;dummy_col&#39;</span><span class="p">]</span>
        <span class="n">df</span><span class="p">[</span><span class="n">group_cols</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="s1">&#39;_&#39;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">drop_group_cols</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">group_cols</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="n">group_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">group_cols</span><span class="p">]</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="nb">set</span><span class="p">(</span><span class="n">group_cols</span><span class="p">)</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">)):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;invalid `group_cols` </span><span class="si">{group_cols}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">df</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">df</span>

        <span class="c1"># get number of variants with each count</span>
        <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">nvariants</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">group_cols</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;count&#39;</span><span class="p">])</span>
        <span class="o">.</span><span class="n">aggregate</span><span class="p">({</span><span class="s1">&#39;nvariants&#39;</span><span class="p">:</span><span class="s1">&#39;count&#39;</span><span class="p">})</span>
        <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
        <span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">group_cols</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;count&#39;</span><span class="p">],</span>
                     <span class="n">ascending</span><span class="o">=</span><span class="p">[</span><span class="kc">True</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">group_cols</span><span class="p">)</span> <span class="o">+</span> <span class="p">[</span><span class="kc">False</span><span class="p">])</span>
        <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># make nvariants cumulative number with &lt;= number of counts</span>
        <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">nvariants</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">group_cols</span><span class="p">)</span>
                                    <span class="p">[</span><span class="s1">&#39;nvariants&#39;</span><span class="p">]</span>
                                    <span class="o">.</span><span class="n">cumsum</span><span class="p">())</span>

        <span class="c1"># add new column that is total number of variants</span>
        <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">total_variants</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">group_cols</span><span class="p">)</span>
                                         <span class="p">[</span><span class="s1">&#39;nvariants&#39;</span><span class="p">]</span>
                                         <span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="s1">&#39;max&#39;</span><span class="p">))</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">drop_group_cols</span><span class="p">:</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">group_cols</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="s1">&#39;columns&#39;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">group_cols_as_str</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">group_cols</span><span class="p">:</span>
            <span class="n">df</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;str&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">df</span></div>


<div class="viewcode-block" id="codonSubsToSeq"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.codonSubsToSeq">[docs]</a><span class="k">def</span> <span class="nf">codonSubsToSeq</span><span class="p">(</span><span class="n">wildtype</span><span class="p">,</span> <span class="n">codon_subs</span><span class="p">,</span> <span class="n">return_aa</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">aa_subs</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Convert codon substitutions to sequence.</span>

<span class="sd">    Args:</span>
<span class="sd">        `wildtype` (str)</span>
<span class="sd">            The wildtype sequence</span>
<span class="sd">        `codon_subs` (str)</span>
<span class="sd">            String of space delimited codon substitutions, in the format:</span>
<span class="sd">            OldCodonSiteNewCodon</span>
<span class="sd">        `return_aa` (bool)</span>
<span class="sd">            Specify whether to return sequence as nucleotide or amino acid.</span>
<span class="sd">            Default is nucleotide.</span>
<span class="sd">        `aa_subs` (bool)</span>
<span class="sd">            Specify whether the substitutions are in amino acid form</span>
<span class="sd">            rather than codon. Default is codon. `return_aa` must be True</span>
<span class="sd">            in order for `aa_subs` to be True, since there are numerous</span>
<span class="sd">            possible nucleotide sequences for an amino acid sequence.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A str of the sequence with all the codon substitutions</span>

<span class="sd">    Here is an example:</span>

<span class="sd">    &gt;&gt;&gt; codonSubsToSeq(&#39;ATGGAACAA&#39;, &#39;&#39;)</span>
<span class="sd">    &#39;ATGGAACAA&#39;</span>
<span class="sd">    &gt;&gt;&gt; codonSubsToSeq(&#39;ATGGAACAA&#39;, &#39;GAA2CAG&#39;)</span>
<span class="sd">    &#39;ATGCAGCAA&#39;</span>
<span class="sd">    &gt;&gt;&gt; codonSubsToSeq(&#39;ATGGAACAA&#39;, &#39;ATG1GGG GAA2CAG&#39;)</span>
<span class="sd">    &#39;GGGCAGCAA&#39;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Make sure you are not trying to convert amino acids to codons</span>
    <span class="k">if</span> <span class="n">aa_subs</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">return_aa</span><span class="o">==</span><span class="kc">True</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Cannot return nucleotide sequence using aa subs&#39;</span><span class="p">)</span>
    <span class="c1"># Make sure the wildtype sequence is divisible into codons</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">wildtype</span><span class="p">)</span> <span class="o">%</span> <span class="mi">3</span>  <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;`wildtype` not divisible by 3&#39;</span><span class="p">)</span>

    <span class="n">codon_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">wildtype</span><span class="p">[</span><span class="mi">3</span> <span class="o">*</span> <span class="n">r</span> <span class="p">:</span> <span class="mi">3</span> <span class="o">*</span> <span class="n">r</span> <span class="o">+</span> <span class="mi">3</span><span class="p">]</span> <span class="k">for</span>
                  <span class="n">r</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">wildtype</span><span class="p">)</span> <span class="o">//</span> <span class="mi">3</span><span class="p">)]</span>

    <span class="c1"># Create parser for the codon substitutions</span>
    <span class="k">if</span> <span class="n">aa_subs</span><span class="p">:</span>
        <span class="n">submatcher</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s1">&#39;^(?P&lt;wt&gt;.)&#39;</span>
                                <span class="s1">&#39;(?P&lt;site&gt;\d+)&#39;</span>
                                <span class="s1">&#39;(?P&lt;mut&gt;.)$&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">aa_subs</span><span class="p">:</span>
        <span class="n">submatcher</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s1">&#39;^(?P&lt;wt&gt;[ATGC]</span><span class="si">{3}</span><span class="s1">)&#39;</span>
                                <span class="s1">&#39;(?P&lt;site&gt;\d+)&#39;</span>
                                <span class="s1">&#39;(?P&lt;mut&gt;[ATGC]</span><span class="si">{3}</span><span class="s1">)$&#39;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">codon</span> <span class="ow">in</span> <span class="n">codon_subs</span><span class="o">.</span><span class="n">split</span><span class="p">():</span>
        <span class="n">m</span> <span class="o">=</span> <span class="n">submatcher</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">codon</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">m</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Invalid codon substitution </span><span class="si">{codon}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">site</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;site&#39;</span><span class="p">))</span> <span class="o">-</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="mi">0</span> <span class="o">&lt;=</span> <span class="n">site</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">codon_list</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Codon site out of bounds&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">aa_subs</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;wt&#39;</span><span class="p">)</span> <span class="o">!=</span> <span class="n">codon_list</span><span class="p">[</span><span class="n">site</span><span class="p">]:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Invalid wildtype codon in </span><span class="si">{codon}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;wt&#39;</span><span class="p">)</span> <span class="o">!=</span> <span class="n">CODON_TO_AA</span><span class="p">[</span><span class="n">codon_list</span><span class="p">[</span><span class="n">site</span><span class="p">]]:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Invalid wildtype aa in </span><span class="si">{codon}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;wt&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;mut&#39;</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;wildtype and mutant the same in </span><span class="si">{codon}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">aa_subs</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;mut&#39;</span><span class="p">)</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">CODON_TO_AA</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;Invalid mutant codon in </span><span class="si">{codon}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;mut&#39;</span><span class="p">)</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">AAS_WITHSTOP</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;Invalid mutant aa in </span><span class="si">{codon}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="c1"># Change to the mutant codon</span>
        <span class="n">sub</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;mut&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">aa_subs</span><span class="p">:</span>
            <span class="n">sub</span> <span class="o">=</span> <span class="n">AA_TO_CODONS</span><span class="p">[</span><span class="n">sub</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">codon_list</span><span class="p">[</span><span class="n">site</span><span class="p">]</span> <span class="o">=</span> <span class="n">sub</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">codon_list</span><span class="p">[</span><span class="n">site</span><span class="p">]</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;mut&#39;</span><span class="p">)</span>

    <span class="c1"># Return the sequence</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">return_aa</span><span class="p">:</span>
        <span class="k">return</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">codon_list</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">CODON_TO_AA</span><span class="p">[</span><span class="n">codon</span><span class="p">]</span> <span class="k">for</span> <span class="n">codon</span> <span class="ow">in</span> <span class="n">codon_list</span><span class="p">)</span></div>


<div class="viewcode-block" id="func_score_to_gpm"><a class="viewcode-back" href="../../dms_tools2.codonvarianttable.html#dms_tools2.codonvarianttable.func_score_to_gpm">[docs]</a><span class="k">def</span> <span class="nf">func_score_to_gpm</span><span class="p">(</span><span class="n">func_scores_df</span><span class="p">,</span> <span class="n">wildtype</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;func_score&#39;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Generate a genotype phenotype map from a functional score dataframe.</span>

<span class="sd">    Args:</span>
<span class="sd">        `func_scores_df` (functional score dataframe)</span>
<span class="sd">            A functional score dataframe (from the</span>
<span class="sd">            :meth:`CodonVariantTable.func_scores` method), narrowed down to one</span>
<span class="sd">            post sample condition, typically with something like:</span>
<span class="sd">            `func_scores_df.query(&#39;library == @library &amp; sample == @sample&#39;)`</span>
<span class="sd">        `wildtype` (str)</span>
<span class="sd">            A string containing the wildtype sequence</span>
<span class="sd">        `metric` (str)</span>
<span class="sd">            A string specifying which metric to use as a phenotype</span>

<span class="sd">    Returns:</span>
<span class="sd">         A genotype phenotype map object from the Harms lab</span>
<span class="sd">         `gpmap package &lt;https://github.com/harmslab/gpmap&gt;`_</span>
<span class="sd">         to be used as in input into epistasis models in the epistasis package.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Put the phenotypes into a list</span>
    <span class="n">phenotypes</span> <span class="o">=</span> <span class="n">func_scores_df</span><span class="p">[</span><span class="n">metric</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

    <span class="c1"># Get standard deviations for each phenotype</span>
    <span class="n">var</span> <span class="o">=</span> <span class="n">func_scores_df</span><span class="p">[</span><span class="s1">&#39;func_score_var&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">stdev</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">var</span><span class="p">)</span>

    <span class="c1"># Get codon substitutions in a list</span>
    <span class="n">substitutions</span> <span class="o">=</span> <span class="n">func_scores_df</span><span class="p">[</span><span class="s1">&#39;aa_substitutions&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

    <span class="c1"># Get a list of genotypes</span>
    <span class="n">genotypes</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">subs</span> <span class="ow">in</span> <span class="n">substitutions</span><span class="p">:</span>
        <span class="n">genotype</span> <span class="o">=</span> <span class="n">codonSubsToSeq</span><span class="p">(</span><span class="n">wildtype</span><span class="p">,</span> <span class="n">subs</span><span class="p">,</span> <span class="n">return_aa</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">aa_subs</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">genotypes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">genotype</span><span class="p">)</span>

    <span class="c1"># Get the wildtype amino acid sequence</span>
    <span class="n">wildtype</span> <span class="o">=</span> <span class="n">codonSubsToSeq</span><span class="p">(</span><span class="n">wildtype</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">return_aa</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Generate the genotype phenotype map</span>
    <span class="n">gpm</span> <span class="o">=</span> <span class="n">gpmap</span><span class="o">.</span><span class="n">GenotypePhenotypeMap</span><span class="p">(</span><span class="n">wildtype</span><span class="o">=</span><span class="n">wildtype</span><span class="p">,</span> <span class="n">genotypes</span><span class="o">=</span><span class="n">genotypes</span><span class="p">,</span>
                               <span class="n">phenotypes</span><span class="o">=</span><span class="n">phenotypes</span><span class="p">,</span> <span class="n">stdeviations</span><span class="o">=</span><span class="n">stdev</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">gpm</span></div>




<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">doctest</span>
    <span class="n">doctest</span><span class="o">.</span><span class="n">testmod</span><span class="p">()</span>
</pre></div>

          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="../../index.html">
    <img class="logo" src="../../_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">dms_tools2</h1>
    
  </a>
</p>



<p class="blurb">Tools for analyzing deep mutational scanning data.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=dms_tools2&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a class="badge" href="https://travis-ci.org/jbloomlab/dms_tools2">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../examples.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../bcsubamp.html">Barcoded-subamplicon sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../prefs.html">Amino-acid preferences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../diffsel.html">Differential selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../fracsurvive.html">Fraction surviving</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../programs.html">Executable programs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api.html">Python API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../citations.html">Citations</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../index.html">Documentation overview</a><ul>
  <li><a href="../index.html">Module code</a><ul>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017--2019, the Bloom lab.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
    </div>

    
    <a href="https://github.com/jbloomlab/dms_tools2" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>