
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>dms_tools2.seqnumbering &#8212; dms_tools2 2.4.0 documentation</title>
    <link rel="stylesheet" href="../../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
   
  <link rel="stylesheet" href="../../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <h1>Source code for dms_tools2.seqnumbering</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">===============</span>
<span class="sd">seqnumbering</span>
<span class="sd">===============</span>

<span class="sd">Deals with sequence numbering conversions.</span>
<span class="sd">&quot;&quot;&quot;</span>


<span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">import</span> <span class="nn">tempfile</span>
<span class="kn">import</span> <span class="nn">re</span>

<span class="kn">import</span> <span class="nn">Bio.SeqIO</span>
<span class="kn">import</span> <span class="nn">Bio.SeqUtils</span>

<span class="kn">from</span> <span class="nn">dms_tools2</span> <span class="k">import</span> <span class="n">NTS</span>


<div class="viewcode-block" id="TranscriptConverter"><a class="viewcode-back" href="../../dms_tools2.seqnumbering.html#dms_tools2.seqnumbering.TranscriptConverter">[docs]</a><span class="k">class</span> <span class="nc">TranscriptConverter</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;Convert sites in mRNA transcript to chromosome or CDS sites.</span>

<span class="sd">    In mRNA sequencing, we identify mutations with respect</span>
<span class="sd">    to their position in the mRNA. But we may want map them</span>
<span class="sd">    to the corresponding numbers in the entire chromosome </span>
<span class="sd">    (gene segment in case of a segmented virus, or entire</span>
<span class="sd">    viral genome in the case of a non-segmented virus), or</span>
<span class="sd">    to CDSs encoded on that chromosome.</span>

<span class="sd">    For all site numbers used as input and output for this</span>
<span class="sd">    class, numbering is assumed to be 1, 2, .... Note that</span>
<span class="sd">    this is different than the 0, 1, ... numbering used</span>
<span class="sd">    by Python for strings.</span>

<span class="sd">    Args:</span>
<span class="sd">        `genbankfile` (str)</span>
<span class="sd">            Genbank file with one or more loci, each of which</span>
<span class="sd">            should be a separate chromosome. The features</span>
<span class="sd">            of relevance are called `mRNA` and `CDS`.</span>
<span class="sd">            Each of these features should have a qualifier</span>
<span class="sd">            called `label` that gives its name. </span>
<span class="sd">        `ignore_other_features` (bool)</span>
<span class="sd">            If `genbankfile` contains features not `mRNA`</span>
<span class="sd">            or `CDS`, ignore them or raise error?</span>
<span class="sd">        `to_upper` (bool)</span>
<span class="sd">            Convert all sequences to upper case letters?</span>

<span class="sd">    Attributes:</span>
<span class="sd">        `chromosomes` (dict)</span>
<span class="sd">            Keyed by chromosome names, values are</span>
<span class="sd">            `Bio.SeqRecord.SeqRecord` for chromosomes.</span>
<span class="sd">        `mRNAs` (dict)</span>
<span class="sd">            Keyed by mRNA names, values are</span>
<span class="sd">            `Bio.SeqFeature.SeqFeature` for mRNA.</span>
<span class="sd">        `CDSs` (dict)</span>
<span class="sd">            Keyed by CDS names, values are</span>
<span class="sd">            `Bio.SeqFeature.SeqFeature` for CDS.</span>
<span class="sd">        `mRNA_chromosome` (dict)</span>
<span class="sd">            Keyed by mRNA name, values is name of chromosome</span>
<span class="sd">            containing mRNA.</span>
<span class="sd">        `chromosome_CDSs` (dict)</span>
<span class="sd">            Keyed by chromosome names, values are names of</span>
<span class="sd">            CDSs on chromosome.</span>

<span class="sd">    Specify example `genbankfile` contents with required information.</span>
<span class="sd">    In this file, the chromosome is `fluNS`, and it encodes two mRNAs</span>
<span class="sd">    and two CDSs (for `fluNS1` and `fluNS2`):</span>

<span class="sd">    &gt;&gt;&gt; genbank_text = &#39;&#39;&#39;</span>
<span class="sd">    ... LOCUS       fluNS                    890 bp    DNA              UNK 01-JAN-1980</span>
<span class="sd">    ... FEATURES             Location/Qualifiers</span>
<span class="sd">    ...      mRNA            2..868</span>
<span class="sd">    ...                      /label=&quot;fluNS1&quot;</span>
<span class="sd">    ...      mRNA            join(2..56,529..868)</span>
<span class="sd">    ...                      /label=&quot;fluNS2&quot;</span>
<span class="sd">    ...      CDS             27..719</span>
<span class="sd">    ...                      /label=&quot;fluNS1&quot;</span>
<span class="sd">    ...      CDS             join(27..56,529..864)</span>
<span class="sd">    ...                      /label=&quot;fluNS2&quot;</span>
<span class="sd">    ... ORIGIN</span>
<span class="sd">    ...         1 agcaaaagca gggtgacaaa gacataatgg atccaaacac tgtgtcaagc tttcaggtag</span>
<span class="sd">    ...        61 attgctttct ttggcatgtc cgcaaaagag ttgcagacca agaactaggt gatgccccat</span>
<span class="sd">    ...       121 tccttgatcg gcttcgccga gatcagaagt ccctaagagg aagaggcagc actcttggtc</span>
<span class="sd">    ...       181 tggacatcga aacagccacc cgtgctggaa agcaaatagt ggagcggatt ctgaaggaag</span>
<span class="sd">    ...       241 aatctgatga ggcactcaaa atgaccatgg cctctgtacc tgcatcgcgc tacctaactg</span>
<span class="sd">    ...       301 acatgactct tgaggaaatg tcaaggcact ggttcatgct catgcccaag cagaaagtgg</span>
<span class="sd">    ...       361 caggccctct ttgtatcaga atggaccagg cgatcatgga taagaacatc atactgaaag</span>
<span class="sd">    ...       421 cgaacttcag tgtgattttt gaccggctgg agactctaat attactaagg gccttcaccg</span>
<span class="sd">    ...       481 aagaggggac aattgttggc gaaatttcac cactgccctc tcttccagga catactgatg</span>
<span class="sd">    ...       541 aggatgtcaa aaatgcagtt ggggtcctca tcggaggact tgaatggaat aataacacag</span>
<span class="sd">    ...       601 ttcgagtctc tgaaactcta cagagattcg cttggagaag cagtaatgag aatgggagac</span>
<span class="sd">    ...       661 ctccactcac tccaaaacag aaacggaaaa tggcgggaac aattaggtca gaagtttgaa</span>
<span class="sd">    ...       721 gaaataaggt ggttgattga agaagtgaga cacagactga agataacaga gaatagtttt</span>
<span class="sd">    ...       781 gagcaaataa catttatgca agccttacaa ctattgcttg aagtggagca agagataaga</span>
<span class="sd">    ...       841 actttctcgt ttcagcttat ttaataataa aaaacaccct tgtttctact</span>
<span class="sd">    ... //</span>
<span class="sd">    ... &#39;&#39;&#39;</span>

<span class="sd">    Now initialize a :class:`TranscriptConverter`:</span>

<span class="sd">    &gt;&gt;&gt; with tempfile.NamedTemporaryFile(mode=&#39;r+&#39;) as genbankfile:</span>
<span class="sd">    ...     _ = genbankfile.write(genbank_text)</span>
<span class="sd">    ...     genbankfile.flush()</span>
<span class="sd">    ...     _ = genbankfile.seek(0)</span>
<span class="sd">    ...     converter = TranscriptConverter(genbankfile)</span>

<span class="sd">    Confirm resulting :class:`TranscriptConverter` contains one</span>
<span class="sd">    chromosome (`fluNS`) with the expected to mRNAs and CDSs:</span>

<span class="sd">    &gt;&gt;&gt; list(converter.chromosomes.keys())</span>
<span class="sd">    [&#39;fluNS&#39;]</span>
<span class="sd">    &gt;&gt;&gt; sorted(converter.CDSs.keys())</span>
<span class="sd">    [&#39;fluNS1&#39;, &#39;fluNS2&#39;]</span>
<span class="sd">    &gt;&gt;&gt; sorted(converter.mRNAs.keys())</span>
<span class="sd">    [&#39;fluNS1&#39;, &#39;fluNS2&#39;]</span>
<span class="sd">    &gt;&gt;&gt; converter.mRNA_chromosome[&#39;fluNS1&#39;]</span>
<span class="sd">    &#39;fluNS&#39;</span>
<span class="sd">    &gt;&gt;&gt; converter.mRNA_chromosome[&#39;fluNS2&#39;]</span>
<span class="sd">    &#39;fluNS&#39;</span>
<span class="sd">    &gt;&gt;&gt; converter.chromosome_CDSs[&#39;fluNS&#39;]</span>
<span class="sd">    [&#39;fluNS1&#39;, &#39;fluNS2&#39;]</span>

<span class="sd">    Get site in chromosome (`fluNS`) that corresponds to a</span>
<span class="sd">    position in the `fluNS1` mRNA, and then do the same for</span>
<span class="sd">    the `fluNS2` mRNA, using</span>
<span class="sd">    :class:`TranscriptConverter.i_mRNAtoChromosome`. Then</span>
<span class="sd">    check nucleotide identities with</span>
<span class="sd">    :class:`TranscriptConverter.ntIdentity`:</span>

<span class="sd">    &gt;&gt;&gt; converter.i_mRNAtoChromosome(&#39;fluNS1&#39;, 60)</span>
<span class="sd">    61</span>
<span class="sd">    &gt;&gt;&gt; converter.ntIdentity(&#39;fluNS&#39;, 61)</span>
<span class="sd">    &#39;A&#39;</span>
<span class="sd">    &gt;&gt;&gt; converter.i_mRNAtoChromosome(&#39;fluNS2&#39;, 58)</span>
<span class="sd">    531</span>
<span class="sd">    &gt;&gt;&gt; converter.ntIdentity(&#39;fluNS&#39;, 531)</span>
<span class="sd">    &#39;C&#39;</span>

<span class="sd">    Do the same for substrings of `fluNS1` and `fluNS2`:</span>

<span class="sd">    &gt;&gt;&gt; converter.i_mRNAtoChromosome(&#39;fluNS1&#39;, 2, mRNAfragment=&#39;GATTGCTTTCT&#39;)</span>
<span class="sd">    61</span>
<span class="sd">    &gt;&gt;&gt; converter.i_mRNAtoChromosome(&#39;fluNS2&#39;, 9, mRNAfragment=&#39;TTTCAGGACATACTGATG&#39;)</span>
<span class="sd">    531</span>

<span class="sd">    Get amino-acid substitutions caused by chromosome point mutation.</span>
<span class="sd">    We can specify point mutation as string or tuple, and get</span>
<span class="sd">    output with 3 or 1-letter amino-acid codes:</span>

<span class="sd">    &gt;&gt;&gt; converter.aaSubstitutions(&#39;fluNS&#39;, &#39;A61T&#39;)</span>
<span class="sd">    &#39;fluNS1-Asp12Val&#39;</span>
<span class="sd">    &gt;&gt;&gt; converter.aaSubstitutions(&#39;fluNS&#39;, (&#39;A&#39;, 61, &#39;T&#39;))</span>
<span class="sd">    &#39;fluNS1-Asp12Val&#39;</span>
<span class="sd">    &gt;&gt;&gt; converter.aaSubstitutions(&#39;fluNS&#39;, &#39;A61T&#39;, aa_3letter=False)</span>
<span class="sd">    &#39;fluNS1-D12V&#39;</span>

<span class="sd">    Now look at a point mutation that affects `fluNS1` and `fluNS2`, but</span>
<span class="sd">    only causes an amino-acid substitution in the former (is synonymous</span>
<span class="sd">    in the latter):</span>

<span class="sd">    &gt;&gt;&gt; converter.aaSubstitutions(&#39;fluNS&#39;, &#39;C531T&#39;)</span>
<span class="sd">    &#39;fluNS1-His169Tyr&#39;</span>

<span class="sd">    Now mutation that causes amino-acid substitutions in</span>
<span class="sd">    `fluNS1` and `fluNS2`:</span>

<span class="sd">    &gt;&gt;&gt; converter.aaSubstitutions(&#39;fluNS&#39;, &#39;A532T&#39;)</span>
<span class="sd">    &#39;fluNS1-His169Leu_fluNS2-Ile12Leu&#39;</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genbankfile</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span>
            <span class="n">ignore_other_features</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">to_upper</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;See main class docstring.&quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">to_upper</span> <span class="o">=</span> <span class="n">to_upper</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span> <span class="o">=</span> <span class="p">{</span><span class="n">c</span><span class="o">.</span><span class="n">name</span><span class="p">:</span><span class="n">c</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span>
                <span class="n">Bio</span><span class="o">.</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">genbankfile</span><span class="p">,</span> <span class="s1">&#39;genbank&#39;</span><span class="p">)}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">CDSs</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mRNAs</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mRNA_chromosome</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">chromosome_CDSs</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">to_upper</span><span class="p">:</span>
                <span class="n">c</span><span class="o">.</span><span class="n">seq</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">c</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">feature</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">{</span><span class="s1">&#39;mRNA&#39;</span><span class="p">,</span> <span class="s1">&#39;CDS&#39;</span><span class="p">}:</span>
                    <span class="n">name</span> <span class="o">=</span> <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">]</span>
                    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">name</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">:</span>
                        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;multiple &#39;label&#39; qualifiers&quot;</span><span class="p">)</span>
                    <span class="n">feature</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="k">if</span> <span class="n">feature</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s1">&#39;mRNA&#39;</span><span class="p">:</span>
                        <span class="n">d</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mRNAs</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">mRNA_chromosome</span><span class="p">[</span><span class="n">feature</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">name</span>
                    <span class="k">elif</span> <span class="n">feature</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s1">&#39;CDS&#39;</span><span class="p">:</span>
                        <span class="n">d</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">CDSs</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">chromosome_CDSs</span><span class="p">[</span><span class="n">c</span><span class="o">.</span><span class="n">name</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s1">&#39;should not get here&#39;</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">feature</span><span class="o">.</span><span class="n">name</span> <span class="ow">in</span> <span class="n">d</span><span class="p">:</span>
                        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;duplicate </span><span class="si">{0}</span><span class="s2"> </span><span class="si">{1}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                                <span class="n">feature</span><span class="o">.</span><span class="n">type</span><span class="p">,</span> <span class="n">feature</span><span class="o">.</span><span class="n">name</span><span class="p">))</span>
                    <span class="n">d</span><span class="p">[</span><span class="n">feature</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">feature</span>
                <span class="k">elif</span> <span class="ow">not</span> <span class="n">ignore_other_features</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;unrecognized feature type </span><span class="si">{0}</span><span class="s2">&quot;</span>
                            <span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">type</span><span class="p">))</span>


<div class="viewcode-block" id="TranscriptConverter.i_mRNAtoChromosome"><a class="viewcode-back" href="../../dms_tools2.seqnumbering.html#dms_tools2.seqnumbering.TranscriptConverter.i_mRNAtoChromosome">[docs]</a>    <span class="k">def</span> <span class="nf">i_mRNAtoChromosome</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">mRNA</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">mRNAfragment</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Convert site number in mRNA to number in chromosome.</span>

<span class="sd">        Args:</span>
<span class="sd">            `mRNA` (str)</span>
<span class="sd">                Name of a valid mRNA.</span>
<span class="sd">            `i` (int)</span>
<span class="sd">                Site number in mRNA.</span>
<span class="sd">            `mRNAfragment` (str or `None`)</span>
<span class="sd">                Substring sequence of `mRNA`. In this case, `i`</span>
<span class="sd">                is taken as the site in this substring of the</span>
<span class="sd">                `mRNA`. Useful because sometimes mutations may</span>
<span class="sd">                be called in a fragment of the full mRNA. Set</span>
<span class="sd">                to `None` if `i` is site in full mRNA.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Site number in chromosome that contains `mRNA`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">chromosome</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">mRNA_chromosome</span><span class="p">[</span><span class="n">mRNA</span><span class="p">]]</span>
            <span class="n">mRNA_feature</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mRNAs</span><span class="p">[</span><span class="n">mRNA</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;invalid `mRNA` </span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">mRNA</span><span class="p">))</span>

        <span class="n">mRNA_seq</span> <span class="o">=</span> <span class="n">mRNA_feature</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">chromosome</span><span class="p">)</span><span class="o">.</span><span class="n">seq</span>

        <span class="k">if</span> <span class="n">mRNAfragment</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">to_upper</span><span class="p">:</span>
                <span class="n">mRNAfragment</span> <span class="o">=</span> <span class="n">mRNAfragment</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
            <span class="n">n</span> <span class="o">=</span> <span class="n">mRNA_seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">mRNAfragment</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">n</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">i</span> <span class="o">+=</span> <span class="n">mRNA_seq</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="n">mRNAfragment</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">n</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;`mRNA` </span><span class="si">{0}</span><span class="s2"> does contains </span><span class="si">{1}</span><span class="s2"> &quot;</span>
                        <span class="s2">&quot;copies of `mRNAfragment`&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">mRNA</span><span class="p">,</span> <span class="n">n</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;`mRNA` </span><span class="si">{0}</span><span class="s2"> does not contain &quot;</span>
                        <span class="s2">&quot;specified `mRNAfragment`&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">mRNA</span><span class="p">))</span>

        <span class="n">mRNA_sites</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">mRNA_feature</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="mi">1</span> <span class="o">&lt;=</span> <span class="n">i</span> <span class="o">&lt;=</span> <span class="nb">len</span><span class="p">(</span><span class="n">mRNA_sites</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;`i` of </span><span class="si">{0}</span><span class="s2"> not valid site in `mRNA` </span><span class="si">{1}</span><span class="s2">&quot;</span>
                    <span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">mRNA</span><span class="p">))</span>

        <span class="k">return</span> <span class="n">mRNA_sites</span><span class="p">[</span><span class="n">i</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span></div>


<div class="viewcode-block" id="TranscriptConverter.ntIdentity"><a class="viewcode-back" href="../../dms_tools2.seqnumbering.html#dms_tools2.seqnumbering.TranscriptConverter.ntIdentity">[docs]</a>    <span class="k">def</span> <span class="nf">ntIdentity</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chromosome</span><span class="p">,</span> <span class="n">i</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Gets identity at site in chromosome.</span>

<span class="sd">        Args:</span>
<span class="sd">            `chromosome` (str)</span>
<span class="sd">                Name of chromosome.</span>
<span class="sd">            `i` (int)</span>
<span class="sd">                Site in chromosome.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Nucleotide in `chromosome` at site `i`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">chromosome_seq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">[</span><span class="n">chromosome</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;`chromosome` </span><span class="si">{0}</span><span class="s2"> does not exist&quot;</span>
                    <span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chromosome</span><span class="p">))</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="mi">1</span> <span class="o">&lt;=</span> <span class="n">i</span> <span class="o">&lt;=</span> <span class="nb">len</span><span class="p">(</span><span class="n">chromosome_seq</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;`i` of </span><span class="si">{0}</span><span class="s2"> not in `chromosome` </span><span class="si">{1}</span><span class="s2">&quot;</span>
                    <span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">chromosome</span><span class="p">))</span>

        <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="n">chromosome_seq</span><span class="p">[</span><span class="n">i</span> <span class="o">-</span> <span class="mi">1</span><span class="p">])</span></div>


<div class="viewcode-block" id="TranscriptConverter.aaSubstitutions"><a class="viewcode-back" href="../../dms_tools2.seqnumbering.html#dms_tools2.seqnumbering.TranscriptConverter.aaSubstitutions">[docs]</a>    <span class="k">def</span> <span class="nf">aaSubstitutions</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chromosome</span><span class="p">,</span> <span class="n">mutation</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span>
            <span class="n">aa_3letter</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Amino-acid substitutions from point mutation to chromosome.</span>

<span class="sd">        Gets all amino-acid substitutions in CDSs caused by a</span>
<span class="sd">        point mutation to a chromosome.</span>

<span class="sd">        Args:</span>
<span class="sd">            `chromosome` (str)</span>
<span class="sd">                Name of chromosome.</span>
<span class="sd">            `mutation` (str or 3-tuple)</span>
<span class="sd">                Point mutation in chromosome based numbering. A str</span>
<span class="sd">                like &quot;A50T&quot;, or tuple `(wt, i, mut)` where `wt` is</span>
<span class="sd">                wildtype nt, `i` is site, and `mut` is mutant nt.</span>
<span class="sd">            `aa_3letter` (bool)</span>
<span class="sd">                Use 3-letter rather than 1-letter amino-acid codes?</span>

<span class="sd">        Returns:</span>
<span class="sd">            A string giving the amino-acid mutations:</span>
<span class="sd">            </span>
<span class="sd">                - If no mutations, returns empty str.</span>
<span class="sd">                </span>
<span class="sd">                - If one mutation, will be str like &#39;fluNS1-Asp12Val&#39;.</span>

<span class="sd">                - If several mutations, will be str like</span>
<span class="sd">                  &#39;fluNS1-His169Leu_fluNS2-Ile12Leu&#39;.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">cds_names</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosome_CDSs</span><span class="p">[</span><span class="n">chromosome</span><span class="p">]</span>
            <span class="n">chromosome_seq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">[</span><span class="n">chromosome</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;invalid `chromosome` </span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chromosome</span><span class="p">))</span>

        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">mutation</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="n">m</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s1">&#39;^(?P&lt;wt&gt;[</span><span class="si">{0}</span><span class="s1">])(?P&lt;i&gt;\d+)(?P&lt;mut&gt;[</span><span class="si">{0}</span><span class="s1">])$&#39;</span>
                    <span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">NTS</span><span class="p">)),</span> <span class="n">mutation</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">m</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;cannot parse mutation </span><span class="si">{0}</span><span class="s2">&quot;</span>
                        <span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">mutation</span><span class="p">))</span>
            <span class="n">mutation</span> <span class="o">=</span> <span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;wt&#39;</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;i&#39;</span><span class="p">)),</span> <span class="n">m</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="s1">&#39;mut&#39;</span><span class="p">))</span>

        <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">mutation</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">3</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">mutation</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">NTS</span>
                    <span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">mutation</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">NTS</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;invalid mutation </span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">mutation</span><span class="p">))</span>

        <span class="p">(</span><span class="n">wt</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="n">mut</span><span class="p">)</span> <span class="o">=</span> <span class="n">mutation</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="mi">1</span> <span class="o">&lt;=</span> <span class="n">i</span> <span class="o">&lt;=</span> <span class="nb">len</span><span class="p">(</span><span class="n">chromosome_seq</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;`i` of </span><span class="si">{0}</span><span class="s2"> is not valid site in &quot;</span>
                    <span class="s2">&quot;`chromosome` </span><span class="si">{1}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">chromosome</span><span class="p">))</span>

        <span class="k">if</span> <span class="n">wt</span> <span class="o">!=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ntIdentity</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">i</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Invalid wildtype nucleotide at site &quot;</span>
                    <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2"> of </span><span class="si">{1}</span><span class="s2">. Specified </span><span class="si">{2}</span><span class="s2">, should be </span><span class="si">{3}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">i</span><span class="p">,</span> <span class="n">chromosome</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">ntIdentity</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">i</span><span class="p">)))</span>
        
        <span class="k">if</span> <span class="n">wt</span> <span class="o">==</span> <span class="n">mut</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;wildtype and mutant nt are the same&quot;</span><span class="p">)</span>

        <span class="n">cds_w_mut</span> <span class="o">=</span> <span class="p">[</span><span class="n">cds</span> <span class="k">for</span> <span class="n">cds</span> <span class="ow">in</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">CDSs</span><span class="p">[</span><span class="n">cds_name</span><span class="p">]</span>
                <span class="k">for</span> <span class="n">cds_name</span> <span class="ow">in</span> <span class="n">cds_names</span><span class="p">]</span> <span class="k">if</span> <span class="p">(</span><span class="n">i</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="ow">in</span> <span class="n">cds</span><span class="p">]</span>

        <span class="n">mutstring</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">mut_chromosome</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">for</span> <span class="n">cds_name</span> <span class="ow">in</span> <span class="n">cds_names</span><span class="p">:</span>
            <span class="n">cds</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">CDSs</span><span class="p">[</span><span class="n">cds_name</span><span class="p">]</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">i</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="ow">in</span> <span class="n">cds</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">mut_chromosome</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">mut_chromosome</span> <span class="o">=</span> <span class="n">chromosome_seq</span><span class="o">.</span><span class="n">tomutable</span><span class="p">()</span>
                    <span class="n">mut_chromosome</span><span class="p">[</span><span class="n">i</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">mut</span>
                <span class="n">wtprot</span> <span class="o">=</span> <span class="n">cds</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">chromosome_seq</span><span class="p">)</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
                <span class="n">mutprot</span> <span class="o">=</span> <span class="n">cds</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">mut_chromosome</span><span class="p">)</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
                <span class="n">aa_mut</span> <span class="o">=</span> <span class="p">[(</span><span class="n">wt</span><span class="p">,</span> <span class="n">r0</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">mut</span><span class="p">)</span> <span class="k">for</span> <span class="n">r0</span><span class="p">,</span> <span class="p">(</span><span class="n">wt</span><span class="p">,</span> <span class="n">mut</span><span class="p">)</span> <span class="ow">in</span>
                        <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">wtprot</span><span class="p">,</span> <span class="n">mutprot</span><span class="p">))</span> <span class="k">if</span> <span class="n">wt</span> <span class="o">!=</span> <span class="n">mut</span><span class="p">]</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">aa_mut</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;&gt;1 amino-acid mutation&quot;</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">aa_mut</span><span class="p">:</span>
                    <span class="n">wt_aa</span><span class="p">,</span> <span class="n">r</span><span class="p">,</span> <span class="n">mut_aa</span> <span class="o">=</span> <span class="n">aa_mut</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="k">if</span> <span class="n">aa_3letter</span><span class="p">:</span>
                        <span class="n">wt_aa</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqUtils</span><span class="o">.</span><span class="n">seq3</span><span class="p">(</span><span class="n">wt_aa</span><span class="p">)</span>
                        <span class="n">mut_aa</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqUtils</span><span class="o">.</span><span class="n">seq3</span><span class="p">(</span><span class="n">mut_aa</span><span class="p">)</span>
                    <span class="n">mutstring</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{0}</span><span class="s1">-</span><span class="si">{1}{2}{3}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cds_name</span><span class="p">,</span>
                            <span class="n">wt_aa</span><span class="p">,</span> <span class="n">r</span><span class="p">,</span> <span class="n">mut_aa</span><span class="p">))</span>

        <span class="k">return</span> <span class="s1">&#39;_&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">mutstring</span><span class="p">)</span></div></div>



<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">doctest</span>
    <span class="n">doctest</span><span class="o">.</span><span class="n">testmod</span><span class="p">()</span>
</pre></div>

          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="../../index.html">
    <img class="logo" src="../../_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">dms_tools2</h1>
    
  </a>
</p>



<p class="blurb">Tools for analyzing deep mutational scanning data.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=dms_tools2&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a class="badge" href="https://travis-ci.org/jbloomlab/dms_tools2">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/dms_tools2.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../examples.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../bcsubamp.html">Barcoded-subamplicon sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../prefs.html">Amino-acid preferences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../diffsel.html">Differential selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../fracsurvive.html">Fraction surviving</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../programs.html">Executable programs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api.html">Python API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../citations.html">Citations</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../index.html">Documentation overview</a><ul>
  <li><a href="../index.html">Module code</a><ul>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017--2018, the Bloom lab.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
    </div>

    
    <a href="https://github.com/jbloomlab/dms_tools2" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>